// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleRatesSetpoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VehicleRatesSetpointRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  roll: primitiveTypes.float32,
  pitch: primitiveTypes.float32,
  yaw: primitiveTypes.float32,
  thrust_body: ArrayType(primitiveTypes.float32, 3),
  reset_integral: primitiveTypes.bool,
});
const VehicleRatesSetpointRefArray = ArrayType(VehicleRatesSetpointRefStruct);
const VehicleRatesSetpointRefStructArray = StructType({
  data: VehicleRatesSetpointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleRatesSetpointWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleRatesSetpointRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._rollIntialized = true;
      this._pitchIntialized = true;
      this._yawIntialized = true;
      this._wrapperFields.thrust_body = Float32Wrapper.createArray();
      this._wrapperFields.thrust_body.copy(other._wrapperFields.thrust_body);
      this._reset_integralIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleRatesSetpointRefStruct();
    this._timestampIntialized = false;
    this._rollIntialized = false;
    this._pitchIntialized = false;
    this._yawIntialized = false;
    this._wrapperFields.thrust_body = Float32Wrapper.createArray();
    this._reset_integralIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleRatesSetpointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleRatesSetpointArrayWrapper;
  }
  static get ArrayType() {
    return VehicleRatesSetpointArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleRatesSetpointRefStructArray
  }
  static get refObjectType() {
    return VehicleRatesSetpointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleRatesSetpoint');
      }
      if (checkConsistency && !this._rollIntialized) {
        throw new TypeError('Invalid argument: roll in VehicleRatesSetpoint');
      }
      if (checkConsistency && !this._pitchIntialized) {
        throw new TypeError('Invalid argument: pitch in VehicleRatesSetpoint');
      }
      if (checkConsistency && !this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in VehicleRatesSetpoint');
      }
      if (checkConsistency && !this._reset_integralIntialized) {
        throw new TypeError('Invalid argument: reset_integral in VehicleRatesSetpoint');
      }
    this._refObject.thrust_body = Array.from(this._wrapperFields.thrust_body.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._rollIntialized = true;
    this._refObject.roll = refObject.roll;
    this._pitchIntialized = true;
    this._refObject.pitch = refObject.pitch;
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
    this._wrapperFields.thrust_body.fill(refObject.thrust_body.toArray());
    this._reset_integralIntialized = true;
    this._refObject.reset_integral = refObject.reset_integral;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleRatesSetpointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleRatesSetpoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get roll() {
    if (!this._rollIntialized) {
      return undefined;
    }
    return this._refObject.roll;
  }
  set roll(value) {
    this._rollIntialized = true;
    this._refObject.roll = value;
  }
  get pitch() {
    if (!this._pitchIntialized) {
      return undefined;
    }
    return this._refObject.pitch;
  }
  set pitch(value) {
    this._pitchIntialized = true;
    this._refObject.pitch = value;
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  get thrust_body() {
    return this._wrapperFields['thrust_body'].data;
  }
  set thrust_body(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['thrust_body'].fill(value);
  }
  get reset_integral() {
    if (!this._reset_integralIntialized) {
      return undefined;
    }
    return this._refObject.reset_integral;
  }
  set reset_integral(value) {
    this._reset_integralIntialized = true;
    this._refObject.reset_integral = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleRatesSetpointRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._rollIntialized = true;
    this._pitchIntialized = true;
    this._yawIntialized = true;
    this._wrapperFields.thrust_body.fill(refObject.thrust_body.toArray());
    this._reset_integralIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleRatesSetpointRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._rollIntialized = true;
    this._pitchIntialized = true;
    this._yawIntialized = true;
    this._wrapperFields.thrust_body.copy(other._wrapperFields.thrust_body);
    this._reset_integralIntialized = true;
  }
  static get classType() {
    return VehicleRatesSetpointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"roll","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"thrust_body","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_integral","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleRatesSetpoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleRatesSetpoint"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","roll","pitch","yaw","thrust_body","reset_integral"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleRatesSetpointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleRatesSetpointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleRatesSetpointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleRatesSetpointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleRatesSetpointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleRatesSetpointRefArray(size);
    this._refObject = new VehicleRatesSetpointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleRatesSetpointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleRatesSetpointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleRatesSetpointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleRatesSetpointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleRatesSetpointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleRatesSetpointArrayWrapper;
  }
}
module.exports = VehicleRatesSetpointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "roll",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "thrust_body",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_integral",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleRatesSetpoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleRatesSetpoint"
}
*/
