// rclnodejs/generated/px4_msgs/px4_msgs__msg__EstimatorStatusFlags.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EstimatorStatusFlagsRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  control_status_changes: primitiveTypes.uint32,
  cs_tilt_align: primitiveTypes.bool,
  cs_yaw_align: primitiveTypes.bool,
  cs_gps: primitiveTypes.bool,
  cs_opt_flow: primitiveTypes.bool,
  cs_mag_hdg: primitiveTypes.bool,
  cs_mag_3d: primitiveTypes.bool,
  cs_mag_dec: primitiveTypes.bool,
  cs_in_air: primitiveTypes.bool,
  cs_wind: primitiveTypes.bool,
  cs_baro_hgt: primitiveTypes.bool,
  cs_rng_hgt: primitiveTypes.bool,
  cs_gps_hgt: primitiveTypes.bool,
  cs_ev_pos: primitiveTypes.bool,
  cs_ev_yaw: primitiveTypes.bool,
  cs_ev_hgt: primitiveTypes.bool,
  cs_fuse_beta: primitiveTypes.bool,
  cs_mag_field_disturbed: primitiveTypes.bool,
  cs_fixed_wing: primitiveTypes.bool,
  cs_mag_fault: primitiveTypes.bool,
  cs_fuse_aspd: primitiveTypes.bool,
  cs_gnd_effect: primitiveTypes.bool,
  cs_rng_stuck: primitiveTypes.bool,
  cs_gps_yaw: primitiveTypes.bool,
  cs_mag_aligned_in_flight: primitiveTypes.bool,
  cs_ev_vel: primitiveTypes.bool,
  cs_synthetic_mag_z: primitiveTypes.bool,
  cs_vehicle_at_rest: primitiveTypes.bool,
  cs_gps_yaw_fault: primitiveTypes.bool,
  cs_rng_fault: primitiveTypes.bool,
  cs_inertial_dead_reckoning: primitiveTypes.bool,
  cs_wind_dead_reckoning: primitiveTypes.bool,
  cs_rng_kin_consistent: primitiveTypes.bool,
  cs_fake_pos: primitiveTypes.bool,
  cs_fake_hgt: primitiveTypes.bool,
  cs_gravity_vector: primitiveTypes.bool,
  fault_status_changes: primitiveTypes.uint32,
  fs_bad_mag_x: primitiveTypes.bool,
  fs_bad_mag_y: primitiveTypes.bool,
  fs_bad_mag_z: primitiveTypes.bool,
  fs_bad_hdg: primitiveTypes.bool,
  fs_bad_mag_decl: primitiveTypes.bool,
  fs_bad_airspeed: primitiveTypes.bool,
  fs_bad_sideslip: primitiveTypes.bool,
  fs_bad_optflow_x: primitiveTypes.bool,
  fs_bad_optflow_y: primitiveTypes.bool,
  fs_bad_vel_n: primitiveTypes.bool,
  fs_bad_vel_e: primitiveTypes.bool,
  fs_bad_vel_d: primitiveTypes.bool,
  fs_bad_pos_n: primitiveTypes.bool,
  fs_bad_pos_e: primitiveTypes.bool,
  fs_bad_pos_d: primitiveTypes.bool,
  fs_bad_acc_bias: primitiveTypes.bool,
  fs_bad_acc_vertical: primitiveTypes.bool,
  fs_bad_acc_clipping: primitiveTypes.bool,
  innovation_fault_status_changes: primitiveTypes.uint32,
  reject_hor_vel: primitiveTypes.bool,
  reject_ver_vel: primitiveTypes.bool,
  reject_hor_pos: primitiveTypes.bool,
  reject_ver_pos: primitiveTypes.bool,
  reject_yaw: primitiveTypes.bool,
  reject_airspeed: primitiveTypes.bool,
  reject_sideslip: primitiveTypes.bool,
  reject_hagl: primitiveTypes.bool,
  reject_optflow_x: primitiveTypes.bool,
  reject_optflow_y: primitiveTypes.bool,
});
const EstimatorStatusFlagsRefArray = ArrayType(EstimatorStatusFlagsRefStruct);
const EstimatorStatusFlagsRefStructArray = StructType({
  data: EstimatorStatusFlagsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EstimatorStatusFlagsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EstimatorStatusFlagsRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._control_status_changesIntialized = true;
      this._cs_tilt_alignIntialized = true;
      this._cs_yaw_alignIntialized = true;
      this._cs_gpsIntialized = true;
      this._cs_opt_flowIntialized = true;
      this._cs_mag_hdgIntialized = true;
      this._cs_mag_3dIntialized = true;
      this._cs_mag_decIntialized = true;
      this._cs_in_airIntialized = true;
      this._cs_windIntialized = true;
      this._cs_baro_hgtIntialized = true;
      this._cs_rng_hgtIntialized = true;
      this._cs_gps_hgtIntialized = true;
      this._cs_ev_posIntialized = true;
      this._cs_ev_yawIntialized = true;
      this._cs_ev_hgtIntialized = true;
      this._cs_fuse_betaIntialized = true;
      this._cs_mag_field_disturbedIntialized = true;
      this._cs_fixed_wingIntialized = true;
      this._cs_mag_faultIntialized = true;
      this._cs_fuse_aspdIntialized = true;
      this._cs_gnd_effectIntialized = true;
      this._cs_rng_stuckIntialized = true;
      this._cs_gps_yawIntialized = true;
      this._cs_mag_aligned_in_flightIntialized = true;
      this._cs_ev_velIntialized = true;
      this._cs_synthetic_mag_zIntialized = true;
      this._cs_vehicle_at_restIntialized = true;
      this._cs_gps_yaw_faultIntialized = true;
      this._cs_rng_faultIntialized = true;
      this._cs_inertial_dead_reckoningIntialized = true;
      this._cs_wind_dead_reckoningIntialized = true;
      this._cs_rng_kin_consistentIntialized = true;
      this._cs_fake_posIntialized = true;
      this._cs_fake_hgtIntialized = true;
      this._cs_gravity_vectorIntialized = true;
      this._fault_status_changesIntialized = true;
      this._fs_bad_mag_xIntialized = true;
      this._fs_bad_mag_yIntialized = true;
      this._fs_bad_mag_zIntialized = true;
      this._fs_bad_hdgIntialized = true;
      this._fs_bad_mag_declIntialized = true;
      this._fs_bad_airspeedIntialized = true;
      this._fs_bad_sideslipIntialized = true;
      this._fs_bad_optflow_xIntialized = true;
      this._fs_bad_optflow_yIntialized = true;
      this._fs_bad_vel_nIntialized = true;
      this._fs_bad_vel_eIntialized = true;
      this._fs_bad_vel_dIntialized = true;
      this._fs_bad_pos_nIntialized = true;
      this._fs_bad_pos_eIntialized = true;
      this._fs_bad_pos_dIntialized = true;
      this._fs_bad_acc_biasIntialized = true;
      this._fs_bad_acc_verticalIntialized = true;
      this._fs_bad_acc_clippingIntialized = true;
      this._innovation_fault_status_changesIntialized = true;
      this._reject_hor_velIntialized = true;
      this._reject_ver_velIntialized = true;
      this._reject_hor_posIntialized = true;
      this._reject_ver_posIntialized = true;
      this._reject_yawIntialized = true;
      this._reject_airspeedIntialized = true;
      this._reject_sideslipIntialized = true;
      this._reject_haglIntialized = true;
      this._reject_optflow_xIntialized = true;
      this._reject_optflow_yIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EstimatorStatusFlagsRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._control_status_changesIntialized = false;
    this._cs_tilt_alignIntialized = false;
    this._cs_yaw_alignIntialized = false;
    this._cs_gpsIntialized = false;
    this._cs_opt_flowIntialized = false;
    this._cs_mag_hdgIntialized = false;
    this._cs_mag_3dIntialized = false;
    this._cs_mag_decIntialized = false;
    this._cs_in_airIntialized = false;
    this._cs_windIntialized = false;
    this._cs_baro_hgtIntialized = false;
    this._cs_rng_hgtIntialized = false;
    this._cs_gps_hgtIntialized = false;
    this._cs_ev_posIntialized = false;
    this._cs_ev_yawIntialized = false;
    this._cs_ev_hgtIntialized = false;
    this._cs_fuse_betaIntialized = false;
    this._cs_mag_field_disturbedIntialized = false;
    this._cs_fixed_wingIntialized = false;
    this._cs_mag_faultIntialized = false;
    this._cs_fuse_aspdIntialized = false;
    this._cs_gnd_effectIntialized = false;
    this._cs_rng_stuckIntialized = false;
    this._cs_gps_yawIntialized = false;
    this._cs_mag_aligned_in_flightIntialized = false;
    this._cs_ev_velIntialized = false;
    this._cs_synthetic_mag_zIntialized = false;
    this._cs_vehicle_at_restIntialized = false;
    this._cs_gps_yaw_faultIntialized = false;
    this._cs_rng_faultIntialized = false;
    this._cs_inertial_dead_reckoningIntialized = false;
    this._cs_wind_dead_reckoningIntialized = false;
    this._cs_rng_kin_consistentIntialized = false;
    this._cs_fake_posIntialized = false;
    this._cs_fake_hgtIntialized = false;
    this._cs_gravity_vectorIntialized = false;
    this._fault_status_changesIntialized = false;
    this._fs_bad_mag_xIntialized = false;
    this._fs_bad_mag_yIntialized = false;
    this._fs_bad_mag_zIntialized = false;
    this._fs_bad_hdgIntialized = false;
    this._fs_bad_mag_declIntialized = false;
    this._fs_bad_airspeedIntialized = false;
    this._fs_bad_sideslipIntialized = false;
    this._fs_bad_optflow_xIntialized = false;
    this._fs_bad_optflow_yIntialized = false;
    this._fs_bad_vel_nIntialized = false;
    this._fs_bad_vel_eIntialized = false;
    this._fs_bad_vel_dIntialized = false;
    this._fs_bad_pos_nIntialized = false;
    this._fs_bad_pos_eIntialized = false;
    this._fs_bad_pos_dIntialized = false;
    this._fs_bad_acc_biasIntialized = false;
    this._fs_bad_acc_verticalIntialized = false;
    this._fs_bad_acc_clippingIntialized = false;
    this._innovation_fault_status_changesIntialized = false;
    this._reject_hor_velIntialized = false;
    this._reject_ver_velIntialized = false;
    this._reject_hor_posIntialized = false;
    this._reject_ver_posIntialized = false;
    this._reject_yawIntialized = false;
    this._reject_airspeedIntialized = false;
    this._reject_sideslipIntialized = false;
    this._reject_haglIntialized = false;
    this._reject_optflow_xIntialized = false;
    this._reject_optflow_yIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EstimatorStatusFlagsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EstimatorStatusFlagsArrayWrapper;
  }
  static get ArrayType() {
    return EstimatorStatusFlagsArrayWrapper;
  }
  static get refObjectArrayType() {
    return EstimatorStatusFlagsRefStructArray
  }
  static get refObjectType() {
    return EstimatorStatusFlagsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._control_status_changesIntialized) {
        throw new TypeError('Invalid argument: control_status_changes in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_tilt_alignIntialized) {
        throw new TypeError('Invalid argument: cs_tilt_align in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_yaw_alignIntialized) {
        throw new TypeError('Invalid argument: cs_yaw_align in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_gpsIntialized) {
        throw new TypeError('Invalid argument: cs_gps in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_opt_flowIntialized) {
        throw new TypeError('Invalid argument: cs_opt_flow in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_mag_hdgIntialized) {
        throw new TypeError('Invalid argument: cs_mag_hdg in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_mag_3dIntialized) {
        throw new TypeError('Invalid argument: cs_mag_3d in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_mag_decIntialized) {
        throw new TypeError('Invalid argument: cs_mag_dec in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_in_airIntialized) {
        throw new TypeError('Invalid argument: cs_in_air in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_windIntialized) {
        throw new TypeError('Invalid argument: cs_wind in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_baro_hgtIntialized) {
        throw new TypeError('Invalid argument: cs_baro_hgt in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_rng_hgtIntialized) {
        throw new TypeError('Invalid argument: cs_rng_hgt in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_gps_hgtIntialized) {
        throw new TypeError('Invalid argument: cs_gps_hgt in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_ev_posIntialized) {
        throw new TypeError('Invalid argument: cs_ev_pos in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_ev_yawIntialized) {
        throw new TypeError('Invalid argument: cs_ev_yaw in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_ev_hgtIntialized) {
        throw new TypeError('Invalid argument: cs_ev_hgt in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_fuse_betaIntialized) {
        throw new TypeError('Invalid argument: cs_fuse_beta in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_mag_field_disturbedIntialized) {
        throw new TypeError('Invalid argument: cs_mag_field_disturbed in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_fixed_wingIntialized) {
        throw new TypeError('Invalid argument: cs_fixed_wing in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_mag_faultIntialized) {
        throw new TypeError('Invalid argument: cs_mag_fault in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_fuse_aspdIntialized) {
        throw new TypeError('Invalid argument: cs_fuse_aspd in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_gnd_effectIntialized) {
        throw new TypeError('Invalid argument: cs_gnd_effect in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_rng_stuckIntialized) {
        throw new TypeError('Invalid argument: cs_rng_stuck in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_gps_yawIntialized) {
        throw new TypeError('Invalid argument: cs_gps_yaw in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_mag_aligned_in_flightIntialized) {
        throw new TypeError('Invalid argument: cs_mag_aligned_in_flight in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_ev_velIntialized) {
        throw new TypeError('Invalid argument: cs_ev_vel in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_synthetic_mag_zIntialized) {
        throw new TypeError('Invalid argument: cs_synthetic_mag_z in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_vehicle_at_restIntialized) {
        throw new TypeError('Invalid argument: cs_vehicle_at_rest in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_gps_yaw_faultIntialized) {
        throw new TypeError('Invalid argument: cs_gps_yaw_fault in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_rng_faultIntialized) {
        throw new TypeError('Invalid argument: cs_rng_fault in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_inertial_dead_reckoningIntialized) {
        throw new TypeError('Invalid argument: cs_inertial_dead_reckoning in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_wind_dead_reckoningIntialized) {
        throw new TypeError('Invalid argument: cs_wind_dead_reckoning in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_rng_kin_consistentIntialized) {
        throw new TypeError('Invalid argument: cs_rng_kin_consistent in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_fake_posIntialized) {
        throw new TypeError('Invalid argument: cs_fake_pos in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_fake_hgtIntialized) {
        throw new TypeError('Invalid argument: cs_fake_hgt in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._cs_gravity_vectorIntialized) {
        throw new TypeError('Invalid argument: cs_gravity_vector in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fault_status_changesIntialized) {
        throw new TypeError('Invalid argument: fault_status_changes in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_mag_xIntialized) {
        throw new TypeError('Invalid argument: fs_bad_mag_x in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_mag_yIntialized) {
        throw new TypeError('Invalid argument: fs_bad_mag_y in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_mag_zIntialized) {
        throw new TypeError('Invalid argument: fs_bad_mag_z in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_hdgIntialized) {
        throw new TypeError('Invalid argument: fs_bad_hdg in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_mag_declIntialized) {
        throw new TypeError('Invalid argument: fs_bad_mag_decl in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_airspeedIntialized) {
        throw new TypeError('Invalid argument: fs_bad_airspeed in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_sideslipIntialized) {
        throw new TypeError('Invalid argument: fs_bad_sideslip in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_optflow_xIntialized) {
        throw new TypeError('Invalid argument: fs_bad_optflow_x in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_optflow_yIntialized) {
        throw new TypeError('Invalid argument: fs_bad_optflow_y in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_vel_nIntialized) {
        throw new TypeError('Invalid argument: fs_bad_vel_n in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_vel_eIntialized) {
        throw new TypeError('Invalid argument: fs_bad_vel_e in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_vel_dIntialized) {
        throw new TypeError('Invalid argument: fs_bad_vel_d in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_pos_nIntialized) {
        throw new TypeError('Invalid argument: fs_bad_pos_n in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_pos_eIntialized) {
        throw new TypeError('Invalid argument: fs_bad_pos_e in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_pos_dIntialized) {
        throw new TypeError('Invalid argument: fs_bad_pos_d in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_acc_biasIntialized) {
        throw new TypeError('Invalid argument: fs_bad_acc_bias in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_acc_verticalIntialized) {
        throw new TypeError('Invalid argument: fs_bad_acc_vertical in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._fs_bad_acc_clippingIntialized) {
        throw new TypeError('Invalid argument: fs_bad_acc_clipping in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._innovation_fault_status_changesIntialized) {
        throw new TypeError('Invalid argument: innovation_fault_status_changes in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_hor_velIntialized) {
        throw new TypeError('Invalid argument: reject_hor_vel in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_ver_velIntialized) {
        throw new TypeError('Invalid argument: reject_ver_vel in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_hor_posIntialized) {
        throw new TypeError('Invalid argument: reject_hor_pos in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_ver_posIntialized) {
        throw new TypeError('Invalid argument: reject_ver_pos in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_yawIntialized) {
        throw new TypeError('Invalid argument: reject_yaw in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_airspeedIntialized) {
        throw new TypeError('Invalid argument: reject_airspeed in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_sideslipIntialized) {
        throw new TypeError('Invalid argument: reject_sideslip in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_haglIntialized) {
        throw new TypeError('Invalid argument: reject_hagl in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_optflow_xIntialized) {
        throw new TypeError('Invalid argument: reject_optflow_x in EstimatorStatusFlags');
      }
      if (checkConsistency && !this._reject_optflow_yIntialized) {
        throw new TypeError('Invalid argument: reject_optflow_y in EstimatorStatusFlags');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._control_status_changesIntialized = true;
    this._refObject.control_status_changes = refObject.control_status_changes;
    this._cs_tilt_alignIntialized = true;
    this._refObject.cs_tilt_align = refObject.cs_tilt_align;
    this._cs_yaw_alignIntialized = true;
    this._refObject.cs_yaw_align = refObject.cs_yaw_align;
    this._cs_gpsIntialized = true;
    this._refObject.cs_gps = refObject.cs_gps;
    this._cs_opt_flowIntialized = true;
    this._refObject.cs_opt_flow = refObject.cs_opt_flow;
    this._cs_mag_hdgIntialized = true;
    this._refObject.cs_mag_hdg = refObject.cs_mag_hdg;
    this._cs_mag_3dIntialized = true;
    this._refObject.cs_mag_3d = refObject.cs_mag_3d;
    this._cs_mag_decIntialized = true;
    this._refObject.cs_mag_dec = refObject.cs_mag_dec;
    this._cs_in_airIntialized = true;
    this._refObject.cs_in_air = refObject.cs_in_air;
    this._cs_windIntialized = true;
    this._refObject.cs_wind = refObject.cs_wind;
    this._cs_baro_hgtIntialized = true;
    this._refObject.cs_baro_hgt = refObject.cs_baro_hgt;
    this._cs_rng_hgtIntialized = true;
    this._refObject.cs_rng_hgt = refObject.cs_rng_hgt;
    this._cs_gps_hgtIntialized = true;
    this._refObject.cs_gps_hgt = refObject.cs_gps_hgt;
    this._cs_ev_posIntialized = true;
    this._refObject.cs_ev_pos = refObject.cs_ev_pos;
    this._cs_ev_yawIntialized = true;
    this._refObject.cs_ev_yaw = refObject.cs_ev_yaw;
    this._cs_ev_hgtIntialized = true;
    this._refObject.cs_ev_hgt = refObject.cs_ev_hgt;
    this._cs_fuse_betaIntialized = true;
    this._refObject.cs_fuse_beta = refObject.cs_fuse_beta;
    this._cs_mag_field_disturbedIntialized = true;
    this._refObject.cs_mag_field_disturbed = refObject.cs_mag_field_disturbed;
    this._cs_fixed_wingIntialized = true;
    this._refObject.cs_fixed_wing = refObject.cs_fixed_wing;
    this._cs_mag_faultIntialized = true;
    this._refObject.cs_mag_fault = refObject.cs_mag_fault;
    this._cs_fuse_aspdIntialized = true;
    this._refObject.cs_fuse_aspd = refObject.cs_fuse_aspd;
    this._cs_gnd_effectIntialized = true;
    this._refObject.cs_gnd_effect = refObject.cs_gnd_effect;
    this._cs_rng_stuckIntialized = true;
    this._refObject.cs_rng_stuck = refObject.cs_rng_stuck;
    this._cs_gps_yawIntialized = true;
    this._refObject.cs_gps_yaw = refObject.cs_gps_yaw;
    this._cs_mag_aligned_in_flightIntialized = true;
    this._refObject.cs_mag_aligned_in_flight = refObject.cs_mag_aligned_in_flight;
    this._cs_ev_velIntialized = true;
    this._refObject.cs_ev_vel = refObject.cs_ev_vel;
    this._cs_synthetic_mag_zIntialized = true;
    this._refObject.cs_synthetic_mag_z = refObject.cs_synthetic_mag_z;
    this._cs_vehicle_at_restIntialized = true;
    this._refObject.cs_vehicle_at_rest = refObject.cs_vehicle_at_rest;
    this._cs_gps_yaw_faultIntialized = true;
    this._refObject.cs_gps_yaw_fault = refObject.cs_gps_yaw_fault;
    this._cs_rng_faultIntialized = true;
    this._refObject.cs_rng_fault = refObject.cs_rng_fault;
    this._cs_inertial_dead_reckoningIntialized = true;
    this._refObject.cs_inertial_dead_reckoning = refObject.cs_inertial_dead_reckoning;
    this._cs_wind_dead_reckoningIntialized = true;
    this._refObject.cs_wind_dead_reckoning = refObject.cs_wind_dead_reckoning;
    this._cs_rng_kin_consistentIntialized = true;
    this._refObject.cs_rng_kin_consistent = refObject.cs_rng_kin_consistent;
    this._cs_fake_posIntialized = true;
    this._refObject.cs_fake_pos = refObject.cs_fake_pos;
    this._cs_fake_hgtIntialized = true;
    this._refObject.cs_fake_hgt = refObject.cs_fake_hgt;
    this._cs_gravity_vectorIntialized = true;
    this._refObject.cs_gravity_vector = refObject.cs_gravity_vector;
    this._fault_status_changesIntialized = true;
    this._refObject.fault_status_changes = refObject.fault_status_changes;
    this._fs_bad_mag_xIntialized = true;
    this._refObject.fs_bad_mag_x = refObject.fs_bad_mag_x;
    this._fs_bad_mag_yIntialized = true;
    this._refObject.fs_bad_mag_y = refObject.fs_bad_mag_y;
    this._fs_bad_mag_zIntialized = true;
    this._refObject.fs_bad_mag_z = refObject.fs_bad_mag_z;
    this._fs_bad_hdgIntialized = true;
    this._refObject.fs_bad_hdg = refObject.fs_bad_hdg;
    this._fs_bad_mag_declIntialized = true;
    this._refObject.fs_bad_mag_decl = refObject.fs_bad_mag_decl;
    this._fs_bad_airspeedIntialized = true;
    this._refObject.fs_bad_airspeed = refObject.fs_bad_airspeed;
    this._fs_bad_sideslipIntialized = true;
    this._refObject.fs_bad_sideslip = refObject.fs_bad_sideslip;
    this._fs_bad_optflow_xIntialized = true;
    this._refObject.fs_bad_optflow_x = refObject.fs_bad_optflow_x;
    this._fs_bad_optflow_yIntialized = true;
    this._refObject.fs_bad_optflow_y = refObject.fs_bad_optflow_y;
    this._fs_bad_vel_nIntialized = true;
    this._refObject.fs_bad_vel_n = refObject.fs_bad_vel_n;
    this._fs_bad_vel_eIntialized = true;
    this._refObject.fs_bad_vel_e = refObject.fs_bad_vel_e;
    this._fs_bad_vel_dIntialized = true;
    this._refObject.fs_bad_vel_d = refObject.fs_bad_vel_d;
    this._fs_bad_pos_nIntialized = true;
    this._refObject.fs_bad_pos_n = refObject.fs_bad_pos_n;
    this._fs_bad_pos_eIntialized = true;
    this._refObject.fs_bad_pos_e = refObject.fs_bad_pos_e;
    this._fs_bad_pos_dIntialized = true;
    this._refObject.fs_bad_pos_d = refObject.fs_bad_pos_d;
    this._fs_bad_acc_biasIntialized = true;
    this._refObject.fs_bad_acc_bias = refObject.fs_bad_acc_bias;
    this._fs_bad_acc_verticalIntialized = true;
    this._refObject.fs_bad_acc_vertical = refObject.fs_bad_acc_vertical;
    this._fs_bad_acc_clippingIntialized = true;
    this._refObject.fs_bad_acc_clipping = refObject.fs_bad_acc_clipping;
    this._innovation_fault_status_changesIntialized = true;
    this._refObject.innovation_fault_status_changes = refObject.innovation_fault_status_changes;
    this._reject_hor_velIntialized = true;
    this._refObject.reject_hor_vel = refObject.reject_hor_vel;
    this._reject_ver_velIntialized = true;
    this._refObject.reject_ver_vel = refObject.reject_ver_vel;
    this._reject_hor_posIntialized = true;
    this._refObject.reject_hor_pos = refObject.reject_hor_pos;
    this._reject_ver_posIntialized = true;
    this._refObject.reject_ver_pos = refObject.reject_ver_pos;
    this._reject_yawIntialized = true;
    this._refObject.reject_yaw = refObject.reject_yaw;
    this._reject_airspeedIntialized = true;
    this._refObject.reject_airspeed = refObject.reject_airspeed;
    this._reject_sideslipIntialized = true;
    this._refObject.reject_sideslip = refObject.reject_sideslip;
    this._reject_haglIntialized = true;
    this._refObject.reject_hagl = refObject.reject_hagl;
    this._reject_optflow_xIntialized = true;
    this._refObject.reject_optflow_x = refObject.reject_optflow_x;
    this._reject_optflow_yIntialized = true;
    this._refObject.reject_optflow_y = refObject.reject_optflow_y;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    EstimatorStatusFlagsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'EstimatorStatusFlags'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get control_status_changes() {
    if (!this._control_status_changesIntialized) {
      return undefined;
    }
    return this._refObject.control_status_changes;
  }
  set control_status_changes(value) {
    this._control_status_changesIntialized = true;
    this._refObject.control_status_changes = value;
  }
  get cs_tilt_align() {
    if (!this._cs_tilt_alignIntialized) {
      return undefined;
    }
    return this._refObject.cs_tilt_align;
  }
  set cs_tilt_align(value) {
    this._cs_tilt_alignIntialized = true;
    this._refObject.cs_tilt_align = value;
  }
  get cs_yaw_align() {
    if (!this._cs_yaw_alignIntialized) {
      return undefined;
    }
    return this._refObject.cs_yaw_align;
  }
  set cs_yaw_align(value) {
    this._cs_yaw_alignIntialized = true;
    this._refObject.cs_yaw_align = value;
  }
  get cs_gps() {
    if (!this._cs_gpsIntialized) {
      return undefined;
    }
    return this._refObject.cs_gps;
  }
  set cs_gps(value) {
    this._cs_gpsIntialized = true;
    this._refObject.cs_gps = value;
  }
  get cs_opt_flow() {
    if (!this._cs_opt_flowIntialized) {
      return undefined;
    }
    return this._refObject.cs_opt_flow;
  }
  set cs_opt_flow(value) {
    this._cs_opt_flowIntialized = true;
    this._refObject.cs_opt_flow = value;
  }
  get cs_mag_hdg() {
    if (!this._cs_mag_hdgIntialized) {
      return undefined;
    }
    return this._refObject.cs_mag_hdg;
  }
  set cs_mag_hdg(value) {
    this._cs_mag_hdgIntialized = true;
    this._refObject.cs_mag_hdg = value;
  }
  get cs_mag_3d() {
    if (!this._cs_mag_3dIntialized) {
      return undefined;
    }
    return this._refObject.cs_mag_3d;
  }
  set cs_mag_3d(value) {
    this._cs_mag_3dIntialized = true;
    this._refObject.cs_mag_3d = value;
  }
  get cs_mag_dec() {
    if (!this._cs_mag_decIntialized) {
      return undefined;
    }
    return this._refObject.cs_mag_dec;
  }
  set cs_mag_dec(value) {
    this._cs_mag_decIntialized = true;
    this._refObject.cs_mag_dec = value;
  }
  get cs_in_air() {
    if (!this._cs_in_airIntialized) {
      return undefined;
    }
    return this._refObject.cs_in_air;
  }
  set cs_in_air(value) {
    this._cs_in_airIntialized = true;
    this._refObject.cs_in_air = value;
  }
  get cs_wind() {
    if (!this._cs_windIntialized) {
      return undefined;
    }
    return this._refObject.cs_wind;
  }
  set cs_wind(value) {
    this._cs_windIntialized = true;
    this._refObject.cs_wind = value;
  }
  get cs_baro_hgt() {
    if (!this._cs_baro_hgtIntialized) {
      return undefined;
    }
    return this._refObject.cs_baro_hgt;
  }
  set cs_baro_hgt(value) {
    this._cs_baro_hgtIntialized = true;
    this._refObject.cs_baro_hgt = value;
  }
  get cs_rng_hgt() {
    if (!this._cs_rng_hgtIntialized) {
      return undefined;
    }
    return this._refObject.cs_rng_hgt;
  }
  set cs_rng_hgt(value) {
    this._cs_rng_hgtIntialized = true;
    this._refObject.cs_rng_hgt = value;
  }
  get cs_gps_hgt() {
    if (!this._cs_gps_hgtIntialized) {
      return undefined;
    }
    return this._refObject.cs_gps_hgt;
  }
  set cs_gps_hgt(value) {
    this._cs_gps_hgtIntialized = true;
    this._refObject.cs_gps_hgt = value;
  }
  get cs_ev_pos() {
    if (!this._cs_ev_posIntialized) {
      return undefined;
    }
    return this._refObject.cs_ev_pos;
  }
  set cs_ev_pos(value) {
    this._cs_ev_posIntialized = true;
    this._refObject.cs_ev_pos = value;
  }
  get cs_ev_yaw() {
    if (!this._cs_ev_yawIntialized) {
      return undefined;
    }
    return this._refObject.cs_ev_yaw;
  }
  set cs_ev_yaw(value) {
    this._cs_ev_yawIntialized = true;
    this._refObject.cs_ev_yaw = value;
  }
  get cs_ev_hgt() {
    if (!this._cs_ev_hgtIntialized) {
      return undefined;
    }
    return this._refObject.cs_ev_hgt;
  }
  set cs_ev_hgt(value) {
    this._cs_ev_hgtIntialized = true;
    this._refObject.cs_ev_hgt = value;
  }
  get cs_fuse_beta() {
    if (!this._cs_fuse_betaIntialized) {
      return undefined;
    }
    return this._refObject.cs_fuse_beta;
  }
  set cs_fuse_beta(value) {
    this._cs_fuse_betaIntialized = true;
    this._refObject.cs_fuse_beta = value;
  }
  get cs_mag_field_disturbed() {
    if (!this._cs_mag_field_disturbedIntialized) {
      return undefined;
    }
    return this._refObject.cs_mag_field_disturbed;
  }
  set cs_mag_field_disturbed(value) {
    this._cs_mag_field_disturbedIntialized = true;
    this._refObject.cs_mag_field_disturbed = value;
  }
  get cs_fixed_wing() {
    if (!this._cs_fixed_wingIntialized) {
      return undefined;
    }
    return this._refObject.cs_fixed_wing;
  }
  set cs_fixed_wing(value) {
    this._cs_fixed_wingIntialized = true;
    this._refObject.cs_fixed_wing = value;
  }
  get cs_mag_fault() {
    if (!this._cs_mag_faultIntialized) {
      return undefined;
    }
    return this._refObject.cs_mag_fault;
  }
  set cs_mag_fault(value) {
    this._cs_mag_faultIntialized = true;
    this._refObject.cs_mag_fault = value;
  }
  get cs_fuse_aspd() {
    if (!this._cs_fuse_aspdIntialized) {
      return undefined;
    }
    return this._refObject.cs_fuse_aspd;
  }
  set cs_fuse_aspd(value) {
    this._cs_fuse_aspdIntialized = true;
    this._refObject.cs_fuse_aspd = value;
  }
  get cs_gnd_effect() {
    if (!this._cs_gnd_effectIntialized) {
      return undefined;
    }
    return this._refObject.cs_gnd_effect;
  }
  set cs_gnd_effect(value) {
    this._cs_gnd_effectIntialized = true;
    this._refObject.cs_gnd_effect = value;
  }
  get cs_rng_stuck() {
    if (!this._cs_rng_stuckIntialized) {
      return undefined;
    }
    return this._refObject.cs_rng_stuck;
  }
  set cs_rng_stuck(value) {
    this._cs_rng_stuckIntialized = true;
    this._refObject.cs_rng_stuck = value;
  }
  get cs_gps_yaw() {
    if (!this._cs_gps_yawIntialized) {
      return undefined;
    }
    return this._refObject.cs_gps_yaw;
  }
  set cs_gps_yaw(value) {
    this._cs_gps_yawIntialized = true;
    this._refObject.cs_gps_yaw = value;
  }
  get cs_mag_aligned_in_flight() {
    if (!this._cs_mag_aligned_in_flightIntialized) {
      return undefined;
    }
    return this._refObject.cs_mag_aligned_in_flight;
  }
  set cs_mag_aligned_in_flight(value) {
    this._cs_mag_aligned_in_flightIntialized = true;
    this._refObject.cs_mag_aligned_in_flight = value;
  }
  get cs_ev_vel() {
    if (!this._cs_ev_velIntialized) {
      return undefined;
    }
    return this._refObject.cs_ev_vel;
  }
  set cs_ev_vel(value) {
    this._cs_ev_velIntialized = true;
    this._refObject.cs_ev_vel = value;
  }
  get cs_synthetic_mag_z() {
    if (!this._cs_synthetic_mag_zIntialized) {
      return undefined;
    }
    return this._refObject.cs_synthetic_mag_z;
  }
  set cs_synthetic_mag_z(value) {
    this._cs_synthetic_mag_zIntialized = true;
    this._refObject.cs_synthetic_mag_z = value;
  }
  get cs_vehicle_at_rest() {
    if (!this._cs_vehicle_at_restIntialized) {
      return undefined;
    }
    return this._refObject.cs_vehicle_at_rest;
  }
  set cs_vehicle_at_rest(value) {
    this._cs_vehicle_at_restIntialized = true;
    this._refObject.cs_vehicle_at_rest = value;
  }
  get cs_gps_yaw_fault() {
    if (!this._cs_gps_yaw_faultIntialized) {
      return undefined;
    }
    return this._refObject.cs_gps_yaw_fault;
  }
  set cs_gps_yaw_fault(value) {
    this._cs_gps_yaw_faultIntialized = true;
    this._refObject.cs_gps_yaw_fault = value;
  }
  get cs_rng_fault() {
    if (!this._cs_rng_faultIntialized) {
      return undefined;
    }
    return this._refObject.cs_rng_fault;
  }
  set cs_rng_fault(value) {
    this._cs_rng_faultIntialized = true;
    this._refObject.cs_rng_fault = value;
  }
  get cs_inertial_dead_reckoning() {
    if (!this._cs_inertial_dead_reckoningIntialized) {
      return undefined;
    }
    return this._refObject.cs_inertial_dead_reckoning;
  }
  set cs_inertial_dead_reckoning(value) {
    this._cs_inertial_dead_reckoningIntialized = true;
    this._refObject.cs_inertial_dead_reckoning = value;
  }
  get cs_wind_dead_reckoning() {
    if (!this._cs_wind_dead_reckoningIntialized) {
      return undefined;
    }
    return this._refObject.cs_wind_dead_reckoning;
  }
  set cs_wind_dead_reckoning(value) {
    this._cs_wind_dead_reckoningIntialized = true;
    this._refObject.cs_wind_dead_reckoning = value;
  }
  get cs_rng_kin_consistent() {
    if (!this._cs_rng_kin_consistentIntialized) {
      return undefined;
    }
    return this._refObject.cs_rng_kin_consistent;
  }
  set cs_rng_kin_consistent(value) {
    this._cs_rng_kin_consistentIntialized = true;
    this._refObject.cs_rng_kin_consistent = value;
  }
  get cs_fake_pos() {
    if (!this._cs_fake_posIntialized) {
      return undefined;
    }
    return this._refObject.cs_fake_pos;
  }
  set cs_fake_pos(value) {
    this._cs_fake_posIntialized = true;
    this._refObject.cs_fake_pos = value;
  }
  get cs_fake_hgt() {
    if (!this._cs_fake_hgtIntialized) {
      return undefined;
    }
    return this._refObject.cs_fake_hgt;
  }
  set cs_fake_hgt(value) {
    this._cs_fake_hgtIntialized = true;
    this._refObject.cs_fake_hgt = value;
  }
  get cs_gravity_vector() {
    if (!this._cs_gravity_vectorIntialized) {
      return undefined;
    }
    return this._refObject.cs_gravity_vector;
  }
  set cs_gravity_vector(value) {
    this._cs_gravity_vectorIntialized = true;
    this._refObject.cs_gravity_vector = value;
  }
  get fault_status_changes() {
    if (!this._fault_status_changesIntialized) {
      return undefined;
    }
    return this._refObject.fault_status_changes;
  }
  set fault_status_changes(value) {
    this._fault_status_changesIntialized = true;
    this._refObject.fault_status_changes = value;
  }
  get fs_bad_mag_x() {
    if (!this._fs_bad_mag_xIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_mag_x;
  }
  set fs_bad_mag_x(value) {
    this._fs_bad_mag_xIntialized = true;
    this._refObject.fs_bad_mag_x = value;
  }
  get fs_bad_mag_y() {
    if (!this._fs_bad_mag_yIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_mag_y;
  }
  set fs_bad_mag_y(value) {
    this._fs_bad_mag_yIntialized = true;
    this._refObject.fs_bad_mag_y = value;
  }
  get fs_bad_mag_z() {
    if (!this._fs_bad_mag_zIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_mag_z;
  }
  set fs_bad_mag_z(value) {
    this._fs_bad_mag_zIntialized = true;
    this._refObject.fs_bad_mag_z = value;
  }
  get fs_bad_hdg() {
    if (!this._fs_bad_hdgIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_hdg;
  }
  set fs_bad_hdg(value) {
    this._fs_bad_hdgIntialized = true;
    this._refObject.fs_bad_hdg = value;
  }
  get fs_bad_mag_decl() {
    if (!this._fs_bad_mag_declIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_mag_decl;
  }
  set fs_bad_mag_decl(value) {
    this._fs_bad_mag_declIntialized = true;
    this._refObject.fs_bad_mag_decl = value;
  }
  get fs_bad_airspeed() {
    if (!this._fs_bad_airspeedIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_airspeed;
  }
  set fs_bad_airspeed(value) {
    this._fs_bad_airspeedIntialized = true;
    this._refObject.fs_bad_airspeed = value;
  }
  get fs_bad_sideslip() {
    if (!this._fs_bad_sideslipIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_sideslip;
  }
  set fs_bad_sideslip(value) {
    this._fs_bad_sideslipIntialized = true;
    this._refObject.fs_bad_sideslip = value;
  }
  get fs_bad_optflow_x() {
    if (!this._fs_bad_optflow_xIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_optflow_x;
  }
  set fs_bad_optflow_x(value) {
    this._fs_bad_optflow_xIntialized = true;
    this._refObject.fs_bad_optflow_x = value;
  }
  get fs_bad_optflow_y() {
    if (!this._fs_bad_optflow_yIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_optflow_y;
  }
  set fs_bad_optflow_y(value) {
    this._fs_bad_optflow_yIntialized = true;
    this._refObject.fs_bad_optflow_y = value;
  }
  get fs_bad_vel_n() {
    if (!this._fs_bad_vel_nIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_vel_n;
  }
  set fs_bad_vel_n(value) {
    this._fs_bad_vel_nIntialized = true;
    this._refObject.fs_bad_vel_n = value;
  }
  get fs_bad_vel_e() {
    if (!this._fs_bad_vel_eIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_vel_e;
  }
  set fs_bad_vel_e(value) {
    this._fs_bad_vel_eIntialized = true;
    this._refObject.fs_bad_vel_e = value;
  }
  get fs_bad_vel_d() {
    if (!this._fs_bad_vel_dIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_vel_d;
  }
  set fs_bad_vel_d(value) {
    this._fs_bad_vel_dIntialized = true;
    this._refObject.fs_bad_vel_d = value;
  }
  get fs_bad_pos_n() {
    if (!this._fs_bad_pos_nIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_pos_n;
  }
  set fs_bad_pos_n(value) {
    this._fs_bad_pos_nIntialized = true;
    this._refObject.fs_bad_pos_n = value;
  }
  get fs_bad_pos_e() {
    if (!this._fs_bad_pos_eIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_pos_e;
  }
  set fs_bad_pos_e(value) {
    this._fs_bad_pos_eIntialized = true;
    this._refObject.fs_bad_pos_e = value;
  }
  get fs_bad_pos_d() {
    if (!this._fs_bad_pos_dIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_pos_d;
  }
  set fs_bad_pos_d(value) {
    this._fs_bad_pos_dIntialized = true;
    this._refObject.fs_bad_pos_d = value;
  }
  get fs_bad_acc_bias() {
    if (!this._fs_bad_acc_biasIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_acc_bias;
  }
  set fs_bad_acc_bias(value) {
    this._fs_bad_acc_biasIntialized = true;
    this._refObject.fs_bad_acc_bias = value;
  }
  get fs_bad_acc_vertical() {
    if (!this._fs_bad_acc_verticalIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_acc_vertical;
  }
  set fs_bad_acc_vertical(value) {
    this._fs_bad_acc_verticalIntialized = true;
    this._refObject.fs_bad_acc_vertical = value;
  }
  get fs_bad_acc_clipping() {
    if (!this._fs_bad_acc_clippingIntialized) {
      return undefined;
    }
    return this._refObject.fs_bad_acc_clipping;
  }
  set fs_bad_acc_clipping(value) {
    this._fs_bad_acc_clippingIntialized = true;
    this._refObject.fs_bad_acc_clipping = value;
  }
  get innovation_fault_status_changes() {
    if (!this._innovation_fault_status_changesIntialized) {
      return undefined;
    }
    return this._refObject.innovation_fault_status_changes;
  }
  set innovation_fault_status_changes(value) {
    this._innovation_fault_status_changesIntialized = true;
    this._refObject.innovation_fault_status_changes = value;
  }
  get reject_hor_vel() {
    if (!this._reject_hor_velIntialized) {
      return undefined;
    }
    return this._refObject.reject_hor_vel;
  }
  set reject_hor_vel(value) {
    this._reject_hor_velIntialized = true;
    this._refObject.reject_hor_vel = value;
  }
  get reject_ver_vel() {
    if (!this._reject_ver_velIntialized) {
      return undefined;
    }
    return this._refObject.reject_ver_vel;
  }
  set reject_ver_vel(value) {
    this._reject_ver_velIntialized = true;
    this._refObject.reject_ver_vel = value;
  }
  get reject_hor_pos() {
    if (!this._reject_hor_posIntialized) {
      return undefined;
    }
    return this._refObject.reject_hor_pos;
  }
  set reject_hor_pos(value) {
    this._reject_hor_posIntialized = true;
    this._refObject.reject_hor_pos = value;
  }
  get reject_ver_pos() {
    if (!this._reject_ver_posIntialized) {
      return undefined;
    }
    return this._refObject.reject_ver_pos;
  }
  set reject_ver_pos(value) {
    this._reject_ver_posIntialized = true;
    this._refObject.reject_ver_pos = value;
  }
  get reject_yaw() {
    if (!this._reject_yawIntialized) {
      return undefined;
    }
    return this._refObject.reject_yaw;
  }
  set reject_yaw(value) {
    this._reject_yawIntialized = true;
    this._refObject.reject_yaw = value;
  }
  get reject_airspeed() {
    if (!this._reject_airspeedIntialized) {
      return undefined;
    }
    return this._refObject.reject_airspeed;
  }
  set reject_airspeed(value) {
    this._reject_airspeedIntialized = true;
    this._refObject.reject_airspeed = value;
  }
  get reject_sideslip() {
    if (!this._reject_sideslipIntialized) {
      return undefined;
    }
    return this._refObject.reject_sideslip;
  }
  set reject_sideslip(value) {
    this._reject_sideslipIntialized = true;
    this._refObject.reject_sideslip = value;
  }
  get reject_hagl() {
    if (!this._reject_haglIntialized) {
      return undefined;
    }
    return this._refObject.reject_hagl;
  }
  set reject_hagl(value) {
    this._reject_haglIntialized = true;
    this._refObject.reject_hagl = value;
  }
  get reject_optflow_x() {
    if (!this._reject_optflow_xIntialized) {
      return undefined;
    }
    return this._refObject.reject_optflow_x;
  }
  set reject_optflow_x(value) {
    this._reject_optflow_xIntialized = true;
    this._refObject.reject_optflow_x = value;
  }
  get reject_optflow_y() {
    if (!this._reject_optflow_yIntialized) {
      return undefined;
    }
    return this._refObject.reject_optflow_y;
  }
  set reject_optflow_y(value) {
    this._reject_optflow_yIntialized = true;
    this._refObject.reject_optflow_y = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EstimatorStatusFlagsRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._control_status_changesIntialized = true;
    this._cs_tilt_alignIntialized = true;
    this._cs_yaw_alignIntialized = true;
    this._cs_gpsIntialized = true;
    this._cs_opt_flowIntialized = true;
    this._cs_mag_hdgIntialized = true;
    this._cs_mag_3dIntialized = true;
    this._cs_mag_decIntialized = true;
    this._cs_in_airIntialized = true;
    this._cs_windIntialized = true;
    this._cs_baro_hgtIntialized = true;
    this._cs_rng_hgtIntialized = true;
    this._cs_gps_hgtIntialized = true;
    this._cs_ev_posIntialized = true;
    this._cs_ev_yawIntialized = true;
    this._cs_ev_hgtIntialized = true;
    this._cs_fuse_betaIntialized = true;
    this._cs_mag_field_disturbedIntialized = true;
    this._cs_fixed_wingIntialized = true;
    this._cs_mag_faultIntialized = true;
    this._cs_fuse_aspdIntialized = true;
    this._cs_gnd_effectIntialized = true;
    this._cs_rng_stuckIntialized = true;
    this._cs_gps_yawIntialized = true;
    this._cs_mag_aligned_in_flightIntialized = true;
    this._cs_ev_velIntialized = true;
    this._cs_synthetic_mag_zIntialized = true;
    this._cs_vehicle_at_restIntialized = true;
    this._cs_gps_yaw_faultIntialized = true;
    this._cs_rng_faultIntialized = true;
    this._cs_inertial_dead_reckoningIntialized = true;
    this._cs_wind_dead_reckoningIntialized = true;
    this._cs_rng_kin_consistentIntialized = true;
    this._cs_fake_posIntialized = true;
    this._cs_fake_hgtIntialized = true;
    this._cs_gravity_vectorIntialized = true;
    this._fault_status_changesIntialized = true;
    this._fs_bad_mag_xIntialized = true;
    this._fs_bad_mag_yIntialized = true;
    this._fs_bad_mag_zIntialized = true;
    this._fs_bad_hdgIntialized = true;
    this._fs_bad_mag_declIntialized = true;
    this._fs_bad_airspeedIntialized = true;
    this._fs_bad_sideslipIntialized = true;
    this._fs_bad_optflow_xIntialized = true;
    this._fs_bad_optflow_yIntialized = true;
    this._fs_bad_vel_nIntialized = true;
    this._fs_bad_vel_eIntialized = true;
    this._fs_bad_vel_dIntialized = true;
    this._fs_bad_pos_nIntialized = true;
    this._fs_bad_pos_eIntialized = true;
    this._fs_bad_pos_dIntialized = true;
    this._fs_bad_acc_biasIntialized = true;
    this._fs_bad_acc_verticalIntialized = true;
    this._fs_bad_acc_clippingIntialized = true;
    this._innovation_fault_status_changesIntialized = true;
    this._reject_hor_velIntialized = true;
    this._reject_ver_velIntialized = true;
    this._reject_hor_posIntialized = true;
    this._reject_ver_posIntialized = true;
    this._reject_yawIntialized = true;
    this._reject_airspeedIntialized = true;
    this._reject_sideslipIntialized = true;
    this._reject_haglIntialized = true;
    this._reject_optflow_xIntialized = true;
    this._reject_optflow_yIntialized = true;
  }
  copy(other) {
    this._refObject = new EstimatorStatusFlagsRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._control_status_changesIntialized = true;
    this._cs_tilt_alignIntialized = true;
    this._cs_yaw_alignIntialized = true;
    this._cs_gpsIntialized = true;
    this._cs_opt_flowIntialized = true;
    this._cs_mag_hdgIntialized = true;
    this._cs_mag_3dIntialized = true;
    this._cs_mag_decIntialized = true;
    this._cs_in_airIntialized = true;
    this._cs_windIntialized = true;
    this._cs_baro_hgtIntialized = true;
    this._cs_rng_hgtIntialized = true;
    this._cs_gps_hgtIntialized = true;
    this._cs_ev_posIntialized = true;
    this._cs_ev_yawIntialized = true;
    this._cs_ev_hgtIntialized = true;
    this._cs_fuse_betaIntialized = true;
    this._cs_mag_field_disturbedIntialized = true;
    this._cs_fixed_wingIntialized = true;
    this._cs_mag_faultIntialized = true;
    this._cs_fuse_aspdIntialized = true;
    this._cs_gnd_effectIntialized = true;
    this._cs_rng_stuckIntialized = true;
    this._cs_gps_yawIntialized = true;
    this._cs_mag_aligned_in_flightIntialized = true;
    this._cs_ev_velIntialized = true;
    this._cs_synthetic_mag_zIntialized = true;
    this._cs_vehicle_at_restIntialized = true;
    this._cs_gps_yaw_faultIntialized = true;
    this._cs_rng_faultIntialized = true;
    this._cs_inertial_dead_reckoningIntialized = true;
    this._cs_wind_dead_reckoningIntialized = true;
    this._cs_rng_kin_consistentIntialized = true;
    this._cs_fake_posIntialized = true;
    this._cs_fake_hgtIntialized = true;
    this._cs_gravity_vectorIntialized = true;
    this._fault_status_changesIntialized = true;
    this._fs_bad_mag_xIntialized = true;
    this._fs_bad_mag_yIntialized = true;
    this._fs_bad_mag_zIntialized = true;
    this._fs_bad_hdgIntialized = true;
    this._fs_bad_mag_declIntialized = true;
    this._fs_bad_airspeedIntialized = true;
    this._fs_bad_sideslipIntialized = true;
    this._fs_bad_optflow_xIntialized = true;
    this._fs_bad_optflow_yIntialized = true;
    this._fs_bad_vel_nIntialized = true;
    this._fs_bad_vel_eIntialized = true;
    this._fs_bad_vel_dIntialized = true;
    this._fs_bad_pos_nIntialized = true;
    this._fs_bad_pos_eIntialized = true;
    this._fs_bad_pos_dIntialized = true;
    this._fs_bad_acc_biasIntialized = true;
    this._fs_bad_acc_verticalIntialized = true;
    this._fs_bad_acc_clippingIntialized = true;
    this._innovation_fault_status_changesIntialized = true;
    this._reject_hor_velIntialized = true;
    this._reject_ver_velIntialized = true;
    this._reject_hor_posIntialized = true;
    this._reject_ver_posIntialized = true;
    this._reject_yawIntialized = true;
    this._reject_airspeedIntialized = true;
    this._reject_sideslipIntialized = true;
    this._reject_haglIntialized = true;
    this._reject_optflow_xIntialized = true;
    this._reject_optflow_yIntialized = true;
  }
  static get classType() {
    return EstimatorStatusFlagsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_status_changes","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_tilt_align","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_yaw_align","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_gps","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_opt_flow","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_mag_hdg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_mag_3d","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_mag_dec","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_in_air","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_wind","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_baro_hgt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_rng_hgt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_gps_hgt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_ev_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_ev_yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_ev_hgt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_fuse_beta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_mag_field_disturbed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_fixed_wing","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_mag_fault","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_fuse_aspd","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_gnd_effect","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_rng_stuck","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_gps_yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_mag_aligned_in_flight","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_ev_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_synthetic_mag_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_vehicle_at_rest","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_gps_yaw_fault","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_rng_fault","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_inertial_dead_reckoning","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_wind_dead_reckoning","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_rng_kin_consistent","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_fake_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_fake_hgt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cs_gravity_vector","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fault_status_changes","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_mag_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_mag_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_mag_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_hdg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_mag_decl","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_airspeed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_sideslip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_optflow_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_optflow_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_vel_n","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_vel_e","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_vel_d","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_pos_n","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_pos_e","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_pos_d","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_acc_bias","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_acc_vertical","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fs_bad_acc_clipping","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innovation_fault_status_changes","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_hor_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_ver_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_hor_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_ver_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_airspeed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_sideslip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_hagl","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_optflow_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reject_optflow_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"EstimatorStatusFlags","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EstimatorStatusFlags"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","control_status_changes","cs_tilt_align","cs_yaw_align","cs_gps","cs_opt_flow","cs_mag_hdg","cs_mag_3d","cs_mag_dec","cs_in_air","cs_wind","cs_baro_hgt","cs_rng_hgt","cs_gps_hgt","cs_ev_pos","cs_ev_yaw","cs_ev_hgt","cs_fuse_beta","cs_mag_field_disturbed","cs_fixed_wing","cs_mag_fault","cs_fuse_aspd","cs_gnd_effect","cs_rng_stuck","cs_gps_yaw","cs_mag_aligned_in_flight","cs_ev_vel","cs_synthetic_mag_z","cs_vehicle_at_rest","cs_gps_yaw_fault","cs_rng_fault","cs_inertial_dead_reckoning","cs_wind_dead_reckoning","cs_rng_kin_consistent","cs_fake_pos","cs_fake_hgt","cs_gravity_vector","fault_status_changes","fs_bad_mag_x","fs_bad_mag_y","fs_bad_mag_z","fs_bad_hdg","fs_bad_mag_decl","fs_bad_airspeed","fs_bad_sideslip","fs_bad_optflow_x","fs_bad_optflow_y","fs_bad_vel_n","fs_bad_vel_e","fs_bad_vel_d","fs_bad_pos_n","fs_bad_pos_e","fs_bad_pos_d","fs_bad_acc_bias","fs_bad_acc_vertical","fs_bad_acc_clipping","innovation_fault_status_changes","reject_hor_vel","reject_ver_vel","reject_hor_pos","reject_ver_pos","reject_yaw","reject_airspeed","reject_sideslip","reject_hagl","reject_optflow_x","reject_optflow_y"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EstimatorStatusFlagsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EstimatorStatusFlagsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EstimatorStatusFlagsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorStatusFlagsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorStatusFlagsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EstimatorStatusFlagsRefArray(size);
    this._refObject = new EstimatorStatusFlagsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EstimatorStatusFlagsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EstimatorStatusFlagsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EstimatorStatusFlagsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EstimatorStatusFlagsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EstimatorStatusFlagsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EstimatorStatusFlagsArrayWrapper;
  }
}
module.exports = EstimatorStatusFlagsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "control_status_changes",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_tilt_align",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_yaw_align",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_gps",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_opt_flow",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_mag_hdg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_mag_3d",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_mag_dec",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_in_air",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_wind",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_baro_hgt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_rng_hgt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_gps_hgt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_ev_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_ev_yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_ev_hgt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_fuse_beta",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_mag_field_disturbed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_fixed_wing",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_mag_fault",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_fuse_aspd",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_gnd_effect",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_rng_stuck",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_gps_yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_mag_aligned_in_flight",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_ev_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_synthetic_mag_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_vehicle_at_rest",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_gps_yaw_fault",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_rng_fault",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_inertial_dead_reckoning",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_wind_dead_reckoning",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_rng_kin_consistent",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_fake_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_fake_hgt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cs_gravity_vector",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fault_status_changes",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_mag_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_mag_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_mag_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_hdg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_mag_decl",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_airspeed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_sideslip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_optflow_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_optflow_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_vel_n",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_vel_e",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_vel_d",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_pos_n",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_pos_e",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_pos_d",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_acc_bias",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_acc_vertical",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fs_bad_acc_clipping",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innovation_fault_status_changes",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_hor_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_ver_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_hor_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_ver_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_airspeed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_sideslip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_hagl",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_optflow_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reject_optflow_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "EstimatorStatusFlags",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EstimatorStatusFlags"
}
*/
