// rclnodejs/generated/px4_msgs/px4_msgs__msg__EstimatorStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const EstimatorStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  output_tracking_error: ArrayType(primitiveTypes.float32, 3),
  gps_check_fail_flags: primitiveTypes.uint16,
  control_mode_flags: primitiveTypes.uint64,
  filter_fault_flags: primitiveTypes.uint32,
  pos_horiz_accuracy: primitiveTypes.float32,
  pos_vert_accuracy: primitiveTypes.float32,
  innovation_check_flags: primitiveTypes.uint16,
  mag_test_ratio: primitiveTypes.float32,
  vel_test_ratio: primitiveTypes.float32,
  pos_test_ratio: primitiveTypes.float32,
  hgt_test_ratio: primitiveTypes.float32,
  tas_test_ratio: primitiveTypes.float32,
  hagl_test_ratio: primitiveTypes.float32,
  beta_test_ratio: primitiveTypes.float32,
  solution_status_flags: primitiveTypes.uint16,
  reset_count_vel_ne: primitiveTypes.uint8,
  reset_count_vel_d: primitiveTypes.uint8,
  reset_count_pos_ne: primitiveTypes.uint8,
  reset_count_pod_d: primitiveTypes.uint8,
  reset_count_quat: primitiveTypes.uint8,
  time_slip: primitiveTypes.float32,
  pre_flt_fail_innov_heading: primitiveTypes.bool,
  pre_flt_fail_innov_vel_horiz: primitiveTypes.bool,
  pre_flt_fail_innov_vel_vert: primitiveTypes.bool,
  pre_flt_fail_innov_height: primitiveTypes.bool,
  pre_flt_fail_mag_field_disturbed: primitiveTypes.bool,
  accel_device_id: primitiveTypes.uint32,
  gyro_device_id: primitiveTypes.uint32,
  baro_device_id: primitiveTypes.uint32,
  mag_device_id: primitiveTypes.uint32,
  health_flags: primitiveTypes.uint8,
  timeout_flags: primitiveTypes.uint8,
});
const EstimatorStatusRefArray = ArrayType(EstimatorStatusRefStruct);
const EstimatorStatusRefStructArray = StructType({
  data: EstimatorStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EstimatorStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EstimatorStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._wrapperFields.output_tracking_error = Float32Wrapper.createArray();
      this._wrapperFields.output_tracking_error.copy(other._wrapperFields.output_tracking_error);
      this._gps_check_fail_flagsIntialized = true;
      this._control_mode_flagsIntialized = true;
      this._filter_fault_flagsIntialized = true;
      this._pos_horiz_accuracyIntialized = true;
      this._pos_vert_accuracyIntialized = true;
      this._innovation_check_flagsIntialized = true;
      this._mag_test_ratioIntialized = true;
      this._vel_test_ratioIntialized = true;
      this._pos_test_ratioIntialized = true;
      this._hgt_test_ratioIntialized = true;
      this._tas_test_ratioIntialized = true;
      this._hagl_test_ratioIntialized = true;
      this._beta_test_ratioIntialized = true;
      this._solution_status_flagsIntialized = true;
      this._reset_count_vel_neIntialized = true;
      this._reset_count_vel_dIntialized = true;
      this._reset_count_pos_neIntialized = true;
      this._reset_count_pod_dIntialized = true;
      this._reset_count_quatIntialized = true;
      this._time_slipIntialized = true;
      this._pre_flt_fail_innov_headingIntialized = true;
      this._pre_flt_fail_innov_vel_horizIntialized = true;
      this._pre_flt_fail_innov_vel_vertIntialized = true;
      this._pre_flt_fail_innov_heightIntialized = true;
      this._pre_flt_fail_mag_field_disturbedIntialized = true;
      this._accel_device_idIntialized = true;
      this._gyro_device_idIntialized = true;
      this._baro_device_idIntialized = true;
      this._mag_device_idIntialized = true;
      this._health_flagsIntialized = true;
      this._timeout_flagsIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EstimatorStatusRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._wrapperFields.output_tracking_error = Float32Wrapper.createArray();
    this._gps_check_fail_flagsIntialized = false;
    this._control_mode_flagsIntialized = false;
    this._filter_fault_flagsIntialized = false;
    this._pos_horiz_accuracyIntialized = false;
    this._pos_vert_accuracyIntialized = false;
    this._innovation_check_flagsIntialized = false;
    this._mag_test_ratioIntialized = false;
    this._vel_test_ratioIntialized = false;
    this._pos_test_ratioIntialized = false;
    this._hgt_test_ratioIntialized = false;
    this._tas_test_ratioIntialized = false;
    this._hagl_test_ratioIntialized = false;
    this._beta_test_ratioIntialized = false;
    this._solution_status_flagsIntialized = false;
    this._reset_count_vel_neIntialized = false;
    this._reset_count_vel_dIntialized = false;
    this._reset_count_pos_neIntialized = false;
    this._reset_count_pod_dIntialized = false;
    this._reset_count_quatIntialized = false;
    this._time_slipIntialized = false;
    this._pre_flt_fail_innov_headingIntialized = false;
    this._pre_flt_fail_innov_vel_horizIntialized = false;
    this._pre_flt_fail_innov_vel_vertIntialized = false;
    this._pre_flt_fail_innov_heightIntialized = false;
    this._pre_flt_fail_mag_field_disturbedIntialized = false;
    this._accel_device_idIntialized = false;
    this._gyro_device_idIntialized = false;
    this._baro_device_idIntialized = false;
    this._mag_device_idIntialized = false;
    this._health_flagsIntialized = false;
    this._timeout_flagsIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EstimatorStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EstimatorStatusArrayWrapper;
  }
  static get ArrayType() {
    return EstimatorStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return EstimatorStatusRefStructArray
  }
  static get refObjectType() {
    return EstimatorStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in EstimatorStatus');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in EstimatorStatus');
      }
      if (checkConsistency && !this._gps_check_fail_flagsIntialized) {
        throw new TypeError('Invalid argument: gps_check_fail_flags in EstimatorStatus');
      }
      if (checkConsistency && !this._control_mode_flagsIntialized) {
        throw new TypeError('Invalid argument: control_mode_flags in EstimatorStatus');
      }
      if (checkConsistency && !this._filter_fault_flagsIntialized) {
        throw new TypeError('Invalid argument: filter_fault_flags in EstimatorStatus');
      }
      if (checkConsistency && !this._pos_horiz_accuracyIntialized) {
        throw new TypeError('Invalid argument: pos_horiz_accuracy in EstimatorStatus');
      }
      if (checkConsistency && !this._pos_vert_accuracyIntialized) {
        throw new TypeError('Invalid argument: pos_vert_accuracy in EstimatorStatus');
      }
      if (checkConsistency && !this._innovation_check_flagsIntialized) {
        throw new TypeError('Invalid argument: innovation_check_flags in EstimatorStatus');
      }
      if (checkConsistency && !this._mag_test_ratioIntialized) {
        throw new TypeError('Invalid argument: mag_test_ratio in EstimatorStatus');
      }
      if (checkConsistency && !this._vel_test_ratioIntialized) {
        throw new TypeError('Invalid argument: vel_test_ratio in EstimatorStatus');
      }
      if (checkConsistency && !this._pos_test_ratioIntialized) {
        throw new TypeError('Invalid argument: pos_test_ratio in EstimatorStatus');
      }
      if (checkConsistency && !this._hgt_test_ratioIntialized) {
        throw new TypeError('Invalid argument: hgt_test_ratio in EstimatorStatus');
      }
      if (checkConsistency && !this._tas_test_ratioIntialized) {
        throw new TypeError('Invalid argument: tas_test_ratio in EstimatorStatus');
      }
      if (checkConsistency && !this._hagl_test_ratioIntialized) {
        throw new TypeError('Invalid argument: hagl_test_ratio in EstimatorStatus');
      }
      if (checkConsistency && !this._beta_test_ratioIntialized) {
        throw new TypeError('Invalid argument: beta_test_ratio in EstimatorStatus');
      }
      if (checkConsistency && !this._solution_status_flagsIntialized) {
        throw new TypeError('Invalid argument: solution_status_flags in EstimatorStatus');
      }
      if (checkConsistency && !this._reset_count_vel_neIntialized) {
        throw new TypeError('Invalid argument: reset_count_vel_ne in EstimatorStatus');
      }
      if (checkConsistency && !this._reset_count_vel_dIntialized) {
        throw new TypeError('Invalid argument: reset_count_vel_d in EstimatorStatus');
      }
      if (checkConsistency && !this._reset_count_pos_neIntialized) {
        throw new TypeError('Invalid argument: reset_count_pos_ne in EstimatorStatus');
      }
      if (checkConsistency && !this._reset_count_pod_dIntialized) {
        throw new TypeError('Invalid argument: reset_count_pod_d in EstimatorStatus');
      }
      if (checkConsistency && !this._reset_count_quatIntialized) {
        throw new TypeError('Invalid argument: reset_count_quat in EstimatorStatus');
      }
      if (checkConsistency && !this._time_slipIntialized) {
        throw new TypeError('Invalid argument: time_slip in EstimatorStatus');
      }
      if (checkConsistency && !this._pre_flt_fail_innov_headingIntialized) {
        throw new TypeError('Invalid argument: pre_flt_fail_innov_heading in EstimatorStatus');
      }
      if (checkConsistency && !this._pre_flt_fail_innov_vel_horizIntialized) {
        throw new TypeError('Invalid argument: pre_flt_fail_innov_vel_horiz in EstimatorStatus');
      }
      if (checkConsistency && !this._pre_flt_fail_innov_vel_vertIntialized) {
        throw new TypeError('Invalid argument: pre_flt_fail_innov_vel_vert in EstimatorStatus');
      }
      if (checkConsistency && !this._pre_flt_fail_innov_heightIntialized) {
        throw new TypeError('Invalid argument: pre_flt_fail_innov_height in EstimatorStatus');
      }
      if (checkConsistency && !this._pre_flt_fail_mag_field_disturbedIntialized) {
        throw new TypeError('Invalid argument: pre_flt_fail_mag_field_disturbed in EstimatorStatus');
      }
      if (checkConsistency && !this._accel_device_idIntialized) {
        throw new TypeError('Invalid argument: accel_device_id in EstimatorStatus');
      }
      if (checkConsistency && !this._gyro_device_idIntialized) {
        throw new TypeError('Invalid argument: gyro_device_id in EstimatorStatus');
      }
      if (checkConsistency && !this._baro_device_idIntialized) {
        throw new TypeError('Invalid argument: baro_device_id in EstimatorStatus');
      }
      if (checkConsistency && !this._mag_device_idIntialized) {
        throw new TypeError('Invalid argument: mag_device_id in EstimatorStatus');
      }
      if (checkConsistency && !this._health_flagsIntialized) {
        throw new TypeError('Invalid argument: health_flags in EstimatorStatus');
      }
      if (checkConsistency && !this._timeout_flagsIntialized) {
        throw new TypeError('Invalid argument: timeout_flags in EstimatorStatus');
      }
    this._refObject.output_tracking_error = Array.from(this._wrapperFields.output_tracking_error.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._wrapperFields.output_tracking_error.fill(refObject.output_tracking_error.toArray());
    this._gps_check_fail_flagsIntialized = true;
    this._refObject.gps_check_fail_flags = refObject.gps_check_fail_flags;
    this._control_mode_flagsIntialized = true;
    this._refObject.control_mode_flags = refObject.control_mode_flags;
    this._filter_fault_flagsIntialized = true;
    this._refObject.filter_fault_flags = refObject.filter_fault_flags;
    this._pos_horiz_accuracyIntialized = true;
    this._refObject.pos_horiz_accuracy = refObject.pos_horiz_accuracy;
    this._pos_vert_accuracyIntialized = true;
    this._refObject.pos_vert_accuracy = refObject.pos_vert_accuracy;
    this._innovation_check_flagsIntialized = true;
    this._refObject.innovation_check_flags = refObject.innovation_check_flags;
    this._mag_test_ratioIntialized = true;
    this._refObject.mag_test_ratio = refObject.mag_test_ratio;
    this._vel_test_ratioIntialized = true;
    this._refObject.vel_test_ratio = refObject.vel_test_ratio;
    this._pos_test_ratioIntialized = true;
    this._refObject.pos_test_ratio = refObject.pos_test_ratio;
    this._hgt_test_ratioIntialized = true;
    this._refObject.hgt_test_ratio = refObject.hgt_test_ratio;
    this._tas_test_ratioIntialized = true;
    this._refObject.tas_test_ratio = refObject.tas_test_ratio;
    this._hagl_test_ratioIntialized = true;
    this._refObject.hagl_test_ratio = refObject.hagl_test_ratio;
    this._beta_test_ratioIntialized = true;
    this._refObject.beta_test_ratio = refObject.beta_test_ratio;
    this._solution_status_flagsIntialized = true;
    this._refObject.solution_status_flags = refObject.solution_status_flags;
    this._reset_count_vel_neIntialized = true;
    this._refObject.reset_count_vel_ne = refObject.reset_count_vel_ne;
    this._reset_count_vel_dIntialized = true;
    this._refObject.reset_count_vel_d = refObject.reset_count_vel_d;
    this._reset_count_pos_neIntialized = true;
    this._refObject.reset_count_pos_ne = refObject.reset_count_pos_ne;
    this._reset_count_pod_dIntialized = true;
    this._refObject.reset_count_pod_d = refObject.reset_count_pod_d;
    this._reset_count_quatIntialized = true;
    this._refObject.reset_count_quat = refObject.reset_count_quat;
    this._time_slipIntialized = true;
    this._refObject.time_slip = refObject.time_slip;
    this._pre_flt_fail_innov_headingIntialized = true;
    this._refObject.pre_flt_fail_innov_heading = refObject.pre_flt_fail_innov_heading;
    this._pre_flt_fail_innov_vel_horizIntialized = true;
    this._refObject.pre_flt_fail_innov_vel_horiz = refObject.pre_flt_fail_innov_vel_horiz;
    this._pre_flt_fail_innov_vel_vertIntialized = true;
    this._refObject.pre_flt_fail_innov_vel_vert = refObject.pre_flt_fail_innov_vel_vert;
    this._pre_flt_fail_innov_heightIntialized = true;
    this._refObject.pre_flt_fail_innov_height = refObject.pre_flt_fail_innov_height;
    this._pre_flt_fail_mag_field_disturbedIntialized = true;
    this._refObject.pre_flt_fail_mag_field_disturbed = refObject.pre_flt_fail_mag_field_disturbed;
    this._accel_device_idIntialized = true;
    this._refObject.accel_device_id = refObject.accel_device_id;
    this._gyro_device_idIntialized = true;
    this._refObject.gyro_device_id = refObject.gyro_device_id;
    this._baro_device_idIntialized = true;
    this._refObject.baro_device_id = refObject.baro_device_id;
    this._mag_device_idIntialized = true;
    this._refObject.mag_device_id = refObject.mag_device_id;
    this._health_flagsIntialized = true;
    this._refObject.health_flags = refObject.health_flags;
    this._timeout_flagsIntialized = true;
    this._refObject.timeout_flags = refObject.timeout_flags;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    EstimatorStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'EstimatorStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get output_tracking_error() {
    return this._wrapperFields['output_tracking_error'].data;
  }
  set output_tracking_error(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['output_tracking_error'].fill(value);
  }
  get gps_check_fail_flags() {
    if (!this._gps_check_fail_flagsIntialized) {
      return undefined;
    }
    return this._refObject.gps_check_fail_flags;
  }
  set gps_check_fail_flags(value) {
    this._gps_check_fail_flagsIntialized = true;
    this._refObject.gps_check_fail_flags = value;
  }
  get control_mode_flags() {
    if (!this._control_mode_flagsIntialized) {
      return undefined;
    }
    return this._refObject.control_mode_flags;
  }
  set control_mode_flags(value) {
    this._control_mode_flagsIntialized = true;
    this._refObject.control_mode_flags = value;
  }
  get filter_fault_flags() {
    if (!this._filter_fault_flagsIntialized) {
      return undefined;
    }
    return this._refObject.filter_fault_flags;
  }
  set filter_fault_flags(value) {
    this._filter_fault_flagsIntialized = true;
    this._refObject.filter_fault_flags = value;
  }
  get pos_horiz_accuracy() {
    if (!this._pos_horiz_accuracyIntialized) {
      return undefined;
    }
    return this._refObject.pos_horiz_accuracy;
  }
  set pos_horiz_accuracy(value) {
    this._pos_horiz_accuracyIntialized = true;
    this._refObject.pos_horiz_accuracy = value;
  }
  get pos_vert_accuracy() {
    if (!this._pos_vert_accuracyIntialized) {
      return undefined;
    }
    return this._refObject.pos_vert_accuracy;
  }
  set pos_vert_accuracy(value) {
    this._pos_vert_accuracyIntialized = true;
    this._refObject.pos_vert_accuracy = value;
  }
  get innovation_check_flags() {
    if (!this._innovation_check_flagsIntialized) {
      return undefined;
    }
    return this._refObject.innovation_check_flags;
  }
  set innovation_check_flags(value) {
    this._innovation_check_flagsIntialized = true;
    this._refObject.innovation_check_flags = value;
  }
  get mag_test_ratio() {
    if (!this._mag_test_ratioIntialized) {
      return undefined;
    }
    return this._refObject.mag_test_ratio;
  }
  set mag_test_ratio(value) {
    this._mag_test_ratioIntialized = true;
    this._refObject.mag_test_ratio = value;
  }
  get vel_test_ratio() {
    if (!this._vel_test_ratioIntialized) {
      return undefined;
    }
    return this._refObject.vel_test_ratio;
  }
  set vel_test_ratio(value) {
    this._vel_test_ratioIntialized = true;
    this._refObject.vel_test_ratio = value;
  }
  get pos_test_ratio() {
    if (!this._pos_test_ratioIntialized) {
      return undefined;
    }
    return this._refObject.pos_test_ratio;
  }
  set pos_test_ratio(value) {
    this._pos_test_ratioIntialized = true;
    this._refObject.pos_test_ratio = value;
  }
  get hgt_test_ratio() {
    if (!this._hgt_test_ratioIntialized) {
      return undefined;
    }
    return this._refObject.hgt_test_ratio;
  }
  set hgt_test_ratio(value) {
    this._hgt_test_ratioIntialized = true;
    this._refObject.hgt_test_ratio = value;
  }
  get tas_test_ratio() {
    if (!this._tas_test_ratioIntialized) {
      return undefined;
    }
    return this._refObject.tas_test_ratio;
  }
  set tas_test_ratio(value) {
    this._tas_test_ratioIntialized = true;
    this._refObject.tas_test_ratio = value;
  }
  get hagl_test_ratio() {
    if (!this._hagl_test_ratioIntialized) {
      return undefined;
    }
    return this._refObject.hagl_test_ratio;
  }
  set hagl_test_ratio(value) {
    this._hagl_test_ratioIntialized = true;
    this._refObject.hagl_test_ratio = value;
  }
  get beta_test_ratio() {
    if (!this._beta_test_ratioIntialized) {
      return undefined;
    }
    return this._refObject.beta_test_ratio;
  }
  set beta_test_ratio(value) {
    this._beta_test_ratioIntialized = true;
    this._refObject.beta_test_ratio = value;
  }
  get solution_status_flags() {
    if (!this._solution_status_flagsIntialized) {
      return undefined;
    }
    return this._refObject.solution_status_flags;
  }
  set solution_status_flags(value) {
    this._solution_status_flagsIntialized = true;
    this._refObject.solution_status_flags = value;
  }
  get reset_count_vel_ne() {
    if (!this._reset_count_vel_neIntialized) {
      return undefined;
    }
    return this._refObject.reset_count_vel_ne;
  }
  set reset_count_vel_ne(value) {
    this._reset_count_vel_neIntialized = true;
    this._refObject.reset_count_vel_ne = value;
  }
  get reset_count_vel_d() {
    if (!this._reset_count_vel_dIntialized) {
      return undefined;
    }
    return this._refObject.reset_count_vel_d;
  }
  set reset_count_vel_d(value) {
    this._reset_count_vel_dIntialized = true;
    this._refObject.reset_count_vel_d = value;
  }
  get reset_count_pos_ne() {
    if (!this._reset_count_pos_neIntialized) {
      return undefined;
    }
    return this._refObject.reset_count_pos_ne;
  }
  set reset_count_pos_ne(value) {
    this._reset_count_pos_neIntialized = true;
    this._refObject.reset_count_pos_ne = value;
  }
  get reset_count_pod_d() {
    if (!this._reset_count_pod_dIntialized) {
      return undefined;
    }
    return this._refObject.reset_count_pod_d;
  }
  set reset_count_pod_d(value) {
    this._reset_count_pod_dIntialized = true;
    this._refObject.reset_count_pod_d = value;
  }
  get reset_count_quat() {
    if (!this._reset_count_quatIntialized) {
      return undefined;
    }
    return this._refObject.reset_count_quat;
  }
  set reset_count_quat(value) {
    this._reset_count_quatIntialized = true;
    this._refObject.reset_count_quat = value;
  }
  get time_slip() {
    if (!this._time_slipIntialized) {
      return undefined;
    }
    return this._refObject.time_slip;
  }
  set time_slip(value) {
    this._time_slipIntialized = true;
    this._refObject.time_slip = value;
  }
  get pre_flt_fail_innov_heading() {
    if (!this._pre_flt_fail_innov_headingIntialized) {
      return undefined;
    }
    return this._refObject.pre_flt_fail_innov_heading;
  }
  set pre_flt_fail_innov_heading(value) {
    this._pre_flt_fail_innov_headingIntialized = true;
    this._refObject.pre_flt_fail_innov_heading = value;
  }
  get pre_flt_fail_innov_vel_horiz() {
    if (!this._pre_flt_fail_innov_vel_horizIntialized) {
      return undefined;
    }
    return this._refObject.pre_flt_fail_innov_vel_horiz;
  }
  set pre_flt_fail_innov_vel_horiz(value) {
    this._pre_flt_fail_innov_vel_horizIntialized = true;
    this._refObject.pre_flt_fail_innov_vel_horiz = value;
  }
  get pre_flt_fail_innov_vel_vert() {
    if (!this._pre_flt_fail_innov_vel_vertIntialized) {
      return undefined;
    }
    return this._refObject.pre_flt_fail_innov_vel_vert;
  }
  set pre_flt_fail_innov_vel_vert(value) {
    this._pre_flt_fail_innov_vel_vertIntialized = true;
    this._refObject.pre_flt_fail_innov_vel_vert = value;
  }
  get pre_flt_fail_innov_height() {
    if (!this._pre_flt_fail_innov_heightIntialized) {
      return undefined;
    }
    return this._refObject.pre_flt_fail_innov_height;
  }
  set pre_flt_fail_innov_height(value) {
    this._pre_flt_fail_innov_heightIntialized = true;
    this._refObject.pre_flt_fail_innov_height = value;
  }
  get pre_flt_fail_mag_field_disturbed() {
    if (!this._pre_flt_fail_mag_field_disturbedIntialized) {
      return undefined;
    }
    return this._refObject.pre_flt_fail_mag_field_disturbed;
  }
  set pre_flt_fail_mag_field_disturbed(value) {
    this._pre_flt_fail_mag_field_disturbedIntialized = true;
    this._refObject.pre_flt_fail_mag_field_disturbed = value;
  }
  get accel_device_id() {
    if (!this._accel_device_idIntialized) {
      return undefined;
    }
    return this._refObject.accel_device_id;
  }
  set accel_device_id(value) {
    this._accel_device_idIntialized = true;
    this._refObject.accel_device_id = value;
  }
  get gyro_device_id() {
    if (!this._gyro_device_idIntialized) {
      return undefined;
    }
    return this._refObject.gyro_device_id;
  }
  set gyro_device_id(value) {
    this._gyro_device_idIntialized = true;
    this._refObject.gyro_device_id = value;
  }
  get baro_device_id() {
    if (!this._baro_device_idIntialized) {
      return undefined;
    }
    return this._refObject.baro_device_id;
  }
  set baro_device_id(value) {
    this._baro_device_idIntialized = true;
    this._refObject.baro_device_id = value;
  }
  get mag_device_id() {
    if (!this._mag_device_idIntialized) {
      return undefined;
    }
    return this._refObject.mag_device_id;
  }
  set mag_device_id(value) {
    this._mag_device_idIntialized = true;
    this._refObject.mag_device_id = value;
  }
  get health_flags() {
    if (!this._health_flagsIntialized) {
      return undefined;
    }
    return this._refObject.health_flags;
  }
  set health_flags(value) {
    this._health_flagsIntialized = true;
    this._refObject.health_flags = value;
  }
  get timeout_flags() {
    if (!this._timeout_flagsIntialized) {
      return undefined;
    }
    return this._refObject.timeout_flags;
  }
  set timeout_flags(value) {
    this._timeout_flagsIntialized = true;
    this._refObject.timeout_flags = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EstimatorStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._wrapperFields.output_tracking_error.fill(refObject.output_tracking_error.toArray());
    this._gps_check_fail_flagsIntialized = true;
    this._control_mode_flagsIntialized = true;
    this._filter_fault_flagsIntialized = true;
    this._pos_horiz_accuracyIntialized = true;
    this._pos_vert_accuracyIntialized = true;
    this._innovation_check_flagsIntialized = true;
    this._mag_test_ratioIntialized = true;
    this._vel_test_ratioIntialized = true;
    this._pos_test_ratioIntialized = true;
    this._hgt_test_ratioIntialized = true;
    this._tas_test_ratioIntialized = true;
    this._hagl_test_ratioIntialized = true;
    this._beta_test_ratioIntialized = true;
    this._solution_status_flagsIntialized = true;
    this._reset_count_vel_neIntialized = true;
    this._reset_count_vel_dIntialized = true;
    this._reset_count_pos_neIntialized = true;
    this._reset_count_pod_dIntialized = true;
    this._reset_count_quatIntialized = true;
    this._time_slipIntialized = true;
    this._pre_flt_fail_innov_headingIntialized = true;
    this._pre_flt_fail_innov_vel_horizIntialized = true;
    this._pre_flt_fail_innov_vel_vertIntialized = true;
    this._pre_flt_fail_innov_heightIntialized = true;
    this._pre_flt_fail_mag_field_disturbedIntialized = true;
    this._accel_device_idIntialized = true;
    this._gyro_device_idIntialized = true;
    this._baro_device_idIntialized = true;
    this._mag_device_idIntialized = true;
    this._health_flagsIntialized = true;
    this._timeout_flagsIntialized = true;
  }
  copy(other) {
    this._refObject = new EstimatorStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._wrapperFields.output_tracking_error.copy(other._wrapperFields.output_tracking_error);
    this._gps_check_fail_flagsIntialized = true;
    this._control_mode_flagsIntialized = true;
    this._filter_fault_flagsIntialized = true;
    this._pos_horiz_accuracyIntialized = true;
    this._pos_vert_accuracyIntialized = true;
    this._innovation_check_flagsIntialized = true;
    this._mag_test_ratioIntialized = true;
    this._vel_test_ratioIntialized = true;
    this._pos_test_ratioIntialized = true;
    this._hgt_test_ratioIntialized = true;
    this._tas_test_ratioIntialized = true;
    this._hagl_test_ratioIntialized = true;
    this._beta_test_ratioIntialized = true;
    this._solution_status_flagsIntialized = true;
    this._reset_count_vel_neIntialized = true;
    this._reset_count_vel_dIntialized = true;
    this._reset_count_pos_neIntialized = true;
    this._reset_count_pod_dIntialized = true;
    this._reset_count_quatIntialized = true;
    this._time_slipIntialized = true;
    this._pre_flt_fail_innov_headingIntialized = true;
    this._pre_flt_fail_innov_vel_horizIntialized = true;
    this._pre_flt_fail_innov_vel_vertIntialized = true;
    this._pre_flt_fail_innov_heightIntialized = true;
    this._pre_flt_fail_mag_field_disturbedIntialized = true;
    this._accel_device_idIntialized = true;
    this._gyro_device_idIntialized = true;
    this._baro_device_idIntialized = true;
    this._mag_device_idIntialized = true;
    this._health_flagsIntialized = true;
    this._timeout_flagsIntialized = true;
  }
  static get classType() {
    return EstimatorStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"GPS_CHECK_FAIL_GPS_FIX","value":0},{"type":"uint8","name":"GPS_CHECK_FAIL_MIN_SAT_COUNT","value":1},{"type":"uint8","name":"GPS_CHECK_FAIL_MAX_PDOP","value":2},{"type":"uint8","name":"GPS_CHECK_FAIL_MAX_HORZ_ERR","value":3},{"type":"uint8","name":"GPS_CHECK_FAIL_MAX_VERT_ERR","value":4},{"type":"uint8","name":"GPS_CHECK_FAIL_MAX_SPD_ERR","value":5},{"type":"uint8","name":"GPS_CHECK_FAIL_MAX_HORZ_DRIFT","value":6},{"type":"uint8","name":"GPS_CHECK_FAIL_MAX_VERT_DRIFT","value":7},{"type":"uint8","name":"GPS_CHECK_FAIL_MAX_HORZ_SPD_ERR","value":8},{"type":"uint8","name":"GPS_CHECK_FAIL_MAX_VERT_SPD_ERR","value":9},{"type":"uint8","name":"CS_TILT_ALIGN","value":0},{"type":"uint8","name":"CS_YAW_ALIGN","value":1},{"type":"uint8","name":"CS_GPS","value":2},{"type":"uint8","name":"CS_OPT_FLOW","value":3},{"type":"uint8","name":"CS_MAG_HDG","value":4},{"type":"uint8","name":"CS_MAG_3D","value":5},{"type":"uint8","name":"CS_MAG_DEC","value":6},{"type":"uint8","name":"CS_IN_AIR","value":7},{"type":"uint8","name":"CS_WIND","value":8},{"type":"uint8","name":"CS_BARO_HGT","value":9},{"type":"uint8","name":"CS_RNG_HGT","value":10},{"type":"uint8","name":"CS_GPS_HGT","value":11},{"type":"uint8","name":"CS_EV_POS","value":12},{"type":"uint8","name":"CS_EV_YAW","value":13},{"type":"uint8","name":"CS_EV_HGT","value":14},{"type":"uint8","name":"CS_BETA","value":15},{"type":"uint8","name":"CS_MAG_FIELD","value":16},{"type":"uint8","name":"CS_FIXED_WING","value":17},{"type":"uint8","name":"CS_MAG_FAULT","value":18},{"type":"uint8","name":"CS_ASPD","value":19},{"type":"uint8","name":"CS_GND_EFFECT","value":20},{"type":"uint8","name":"CS_RNG_STUCK","value":21},{"type":"uint8","name":"CS_GPS_YAW","value":22},{"type":"uint8","name":"CS_MAG_ALIGNED","value":23},{"type":"uint8","name":"CS_EV_VEL","value":24},{"type":"uint8","name":"CS_SYNTHETIC_MAG_Z","value":25},{"type":"uint8","name":"CS_VEHICLE_AT_REST","value":26},{"type":"uint8","name":"CS_GPS_YAW_FAULT","value":27},{"type":"uint8","name":"CS_RNG_FAULT","value":28}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"output_tracking_error","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_check_fail_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_mode_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"filter_fault_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pos_horiz_accuracy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pos_vert_accuracy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innovation_check_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_test_ratio","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vel_test_ratio","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pos_test_ratio","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hgt_test_ratio","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tas_test_ratio","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hagl_test_ratio","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"beta_test_ratio","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"solution_status_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_count_vel_ne","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_count_vel_d","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_count_pos_ne","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_count_pod_d","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_count_quat","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_slip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pre_flt_fail_innov_heading","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pre_flt_fail_innov_vel_horiz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pre_flt_fail_innov_vel_vert","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pre_flt_fail_innov_height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pre_flt_fail_mag_field_disturbed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baro_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"health_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timeout_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"EstimatorStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EstimatorStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","output_tracking_error","gps_check_fail_flags","control_mode_flags","filter_fault_flags","pos_horiz_accuracy","pos_vert_accuracy","innovation_check_flags","mag_test_ratio","vel_test_ratio","pos_test_ratio","hgt_test_ratio","tas_test_ratio","hagl_test_ratio","beta_test_ratio","solution_status_flags","reset_count_vel_ne","reset_count_vel_d","reset_count_pos_ne","reset_count_pod_d","reset_count_quat","time_slip","pre_flt_fail_innov_heading","pre_flt_fail_innov_vel_horiz","pre_flt_fail_innov_vel_vert","pre_flt_fail_innov_height","pre_flt_fail_mag_field_disturbed","accel_device_id","gyro_device_id","baro_device_id","mag_device_id","health_flags","timeout_flags"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EstimatorStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EstimatorStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EstimatorStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EstimatorStatusRefArray(size);
    this._refObject = new EstimatorStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EstimatorStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EstimatorStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EstimatorStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EstimatorStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EstimatorStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EstimatorStatusArrayWrapper;
  }
}
// Define constants (39 in total)
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_GPS_FIX", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_MIN_SAT_COUNT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_MAX_PDOP", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_MAX_HORZ_ERR", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_MAX_VERT_ERR", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_MAX_SPD_ERR", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_MAX_HORZ_DRIFT", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_MAX_VERT_DRIFT", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_MAX_HORZ_SPD_ERR", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "GPS_CHECK_FAIL_MAX_VERT_SPD_ERR", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_TILT_ALIGN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_YAW_ALIGN", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_GPS", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_OPT_FLOW", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_MAG_HDG", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_MAG_3D", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_MAG_DEC", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_IN_AIR", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_WIND", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_BARO_HGT", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_RNG_HGT", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_GPS_HGT", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_EV_POS", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_EV_YAW", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_EV_HGT", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_BETA", {value: 15, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_MAG_FIELD", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_FIXED_WING", {value: 17, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_MAG_FAULT", {value: 18, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_ASPD", {value: 19, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_GND_EFFECT", {value: 20, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_RNG_STUCK", {value: 21, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_GPS_YAW", {value: 22, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_MAG_ALIGNED", {value: 23, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_EV_VEL", {value: 24, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_SYNTHETIC_MAG_Z", {value: 25, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_VEHICLE_AT_REST", {value: 26, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_GPS_YAW_FAULT", {value: 27, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EstimatorStatusWrapper, "CS_RNG_FAULT", {value: 28, writable: false, enumerable: true, configurable: true});
module.exports = EstimatorStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_GPS_FIX",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_MIN_SAT_COUNT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_MAX_PDOP",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_MAX_HORZ_ERR",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_MAX_VERT_ERR",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_MAX_SPD_ERR",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_MAX_HORZ_DRIFT",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_MAX_VERT_DRIFT",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_MAX_HORZ_SPD_ERR",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "GPS_CHECK_FAIL_MAX_VERT_SPD_ERR",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "CS_TILT_ALIGN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "CS_YAW_ALIGN",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "CS_GPS",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "CS_OPT_FLOW",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "CS_MAG_HDG",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "CS_MAG_3D",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "CS_MAG_DEC",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "CS_IN_AIR",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "CS_WIND",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "CS_BARO_HGT",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "CS_RNG_HGT",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "CS_GPS_HGT",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "CS_EV_POS",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "CS_EV_YAW",
      "value": 13
    },
    {
      "type": "uint8",
      "name": "CS_EV_HGT",
      "value": 14
    },
    {
      "type": "uint8",
      "name": "CS_BETA",
      "value": 15
    },
    {
      "type": "uint8",
      "name": "CS_MAG_FIELD",
      "value": 16
    },
    {
      "type": "uint8",
      "name": "CS_FIXED_WING",
      "value": 17
    },
    {
      "type": "uint8",
      "name": "CS_MAG_FAULT",
      "value": 18
    },
    {
      "type": "uint8",
      "name": "CS_ASPD",
      "value": 19
    },
    {
      "type": "uint8",
      "name": "CS_GND_EFFECT",
      "value": 20
    },
    {
      "type": "uint8",
      "name": "CS_RNG_STUCK",
      "value": 21
    },
    {
      "type": "uint8",
      "name": "CS_GPS_YAW",
      "value": 22
    },
    {
      "type": "uint8",
      "name": "CS_MAG_ALIGNED",
      "value": 23
    },
    {
      "type": "uint8",
      "name": "CS_EV_VEL",
      "value": 24
    },
    {
      "type": "uint8",
      "name": "CS_SYNTHETIC_MAG_Z",
      "value": 25
    },
    {
      "type": "uint8",
      "name": "CS_VEHICLE_AT_REST",
      "value": 26
    },
    {
      "type": "uint8",
      "name": "CS_GPS_YAW_FAULT",
      "value": 27
    },
    {
      "type": "uint8",
      "name": "CS_RNG_FAULT",
      "value": 28
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "output_tracking_error",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_check_fail_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "control_mode_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "filter_fault_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pos_horiz_accuracy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pos_vert_accuracy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innovation_check_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_test_ratio",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vel_test_ratio",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pos_test_ratio",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hgt_test_ratio",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tas_test_ratio",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hagl_test_ratio",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "beta_test_ratio",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "solution_status_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_count_vel_ne",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_count_vel_d",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_count_pos_ne",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_count_pod_d",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_count_quat",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_slip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pre_flt_fail_innov_heading",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pre_flt_fail_innov_vel_horiz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pre_flt_fail_innov_vel_vert",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pre_flt_fail_innov_height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pre_flt_fail_mag_field_disturbed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baro_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "health_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timeout_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "EstimatorStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EstimatorStatus"
}
*/
