// rclnodejs/generated/px4_msgs/px4_msgs__msg__PositionControllerLandingStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PositionControllerLandingStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  lateral_touchdown_offset: primitiveTypes.float32,
  flaring: primitiveTypes.bool,
  abort_status: primitiveTypes.uint8,
});
const PositionControllerLandingStatusRefArray = ArrayType(PositionControllerLandingStatusRefStruct);
const PositionControllerLandingStatusRefStructArray = StructType({
  data: PositionControllerLandingStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PositionControllerLandingStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PositionControllerLandingStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._lateral_touchdown_offsetIntialized = true;
      this._flaringIntialized = true;
      this._abort_statusIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PositionControllerLandingStatusRefStruct();
    this._timestampIntialized = false;
    this._lateral_touchdown_offsetIntialized = false;
    this._flaringIntialized = false;
    this._abort_statusIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PositionControllerLandingStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PositionControllerLandingStatusArrayWrapper;
  }
  static get ArrayType() {
    return PositionControllerLandingStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return PositionControllerLandingStatusRefStructArray
  }
  static get refObjectType() {
    return PositionControllerLandingStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in PositionControllerLandingStatus');
      }
      if (checkConsistency && !this._lateral_touchdown_offsetIntialized) {
        throw new TypeError('Invalid argument: lateral_touchdown_offset in PositionControllerLandingStatus');
      }
      if (checkConsistency && !this._flaringIntialized) {
        throw new TypeError('Invalid argument: flaring in PositionControllerLandingStatus');
      }
      if (checkConsistency && !this._abort_statusIntialized) {
        throw new TypeError('Invalid argument: abort_status in PositionControllerLandingStatus');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._lateral_touchdown_offsetIntialized = true;
    this._refObject.lateral_touchdown_offset = refObject.lateral_touchdown_offset;
    this._flaringIntialized = true;
    this._refObject.flaring = refObject.flaring;
    this._abort_statusIntialized = true;
    this._refObject.abort_status = refObject.abort_status;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PositionControllerLandingStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'PositionControllerLandingStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get lateral_touchdown_offset() {
    if (!this._lateral_touchdown_offsetIntialized) {
      return undefined;
    }
    return this._refObject.lateral_touchdown_offset;
  }
  set lateral_touchdown_offset(value) {
    this._lateral_touchdown_offsetIntialized = true;
    this._refObject.lateral_touchdown_offset = value;
  }
  get flaring() {
    if (!this._flaringIntialized) {
      return undefined;
    }
    return this._refObject.flaring;
  }
  set flaring(value) {
    this._flaringIntialized = true;
    this._refObject.flaring = value;
  }
  get abort_status() {
    if (!this._abort_statusIntialized) {
      return undefined;
    }
    return this._refObject.abort_status;
  }
  set abort_status(value) {
    this._abort_statusIntialized = true;
    this._refObject.abort_status = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PositionControllerLandingStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._lateral_touchdown_offsetIntialized = true;
    this._flaringIntialized = true;
    this._abort_statusIntialized = true;
  }
  copy(other) {
    this._refObject = new PositionControllerLandingStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._lateral_touchdown_offsetIntialized = true;
    this._flaringIntialized = true;
    this._abort_statusIntialized = true;
  }
  static get classType() {
    return PositionControllerLandingStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"NOT_ABORTED","value":0},{"type":"uint8","name":"ABORTED_BY_OPERATOR","value":1},{"type":"uint8","name":"TERRAIN_NOT_FOUND","value":2},{"type":"uint8","name":"TERRAIN_TIMEOUT","value":3},{"type":"uint8","name":"UNKNOWN_ABORT_CRITERION","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lateral_touchdown_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flaring","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"abort_status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"PositionControllerLandingStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PositionControllerLandingStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","lateral_touchdown_offset","flaring","abort_status"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PositionControllerLandingStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PositionControllerLandingStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PositionControllerLandingStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PositionControllerLandingStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PositionControllerLandingStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PositionControllerLandingStatusRefArray(size);
    this._refObject = new PositionControllerLandingStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PositionControllerLandingStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PositionControllerLandingStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PositionControllerLandingStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PositionControllerLandingStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PositionControllerLandingStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PositionControllerLandingStatusArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(PositionControllerLandingStatusWrapper, "NOT_ABORTED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PositionControllerLandingStatusWrapper, "ABORTED_BY_OPERATOR", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PositionControllerLandingStatusWrapper, "TERRAIN_NOT_FOUND", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PositionControllerLandingStatusWrapper, "TERRAIN_TIMEOUT", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PositionControllerLandingStatusWrapper, "UNKNOWN_ABORT_CRITERION", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = PositionControllerLandingStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "NOT_ABORTED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ABORTED_BY_OPERATOR",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "TERRAIN_NOT_FOUND",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "TERRAIN_TIMEOUT",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "UNKNOWN_ABORT_CRITERION",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lateral_touchdown_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flaring",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "abort_status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "PositionControllerLandingStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PositionControllerLandingStatus"
}
*/
