// rclnodejs/generated/px4_msgs/px4_msgs__msg__QshellRetval.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const QshellRetvalRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  return_value: primitiveTypes.int32,
  return_sequence: primitiveTypes.uint32,
});
const QshellRetvalRefArray = ArrayType(QshellRetvalRefStruct);
const QshellRetvalRefStructArray = StructType({
  data: QshellRetvalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class QshellRetvalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new QshellRetvalRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._return_valueIntialized = true;
      this._return_sequenceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new QshellRetvalRefStruct();
    this._timestampIntialized = false;
    this._return_valueIntialized = false;
    this._return_sequenceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new QshellRetvalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new QshellRetvalArrayWrapper;
  }
  static get ArrayType() {
    return QshellRetvalArrayWrapper;
  }
  static get refObjectArrayType() {
    return QshellRetvalRefStructArray
  }
  static get refObjectType() {
    return QshellRetvalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in QshellRetval');
      }
      if (checkConsistency && !this._return_valueIntialized) {
        throw new TypeError('Invalid argument: return_value in QshellRetval');
      }
      if (checkConsistency && !this._return_sequenceIntialized) {
        throw new TypeError('Invalid argument: return_sequence in QshellRetval');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._return_valueIntialized = true;
    this._refObject.return_value = refObject.return_value;
    this._return_sequenceIntialized = true;
    this._refObject.return_sequence = refObject.return_sequence;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    QshellRetvalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'QshellRetval'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get return_value() {
    if (!this._return_valueIntialized) {
      return undefined;
    }
    return this._refObject.return_value;
  }
  set return_value(value) {
    this._return_valueIntialized = true;
    this._refObject.return_value = value;
  }
  get return_sequence() {
    if (!this._return_sequenceIntialized) {
      return undefined;
    }
    return this._refObject.return_sequence;
  }
  set return_sequence(value) {
    this._return_sequenceIntialized = true;
    this._refObject.return_sequence = value;
  }
  copyRefObject(refObject) {
    this._refObject = new QshellRetvalRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._return_valueIntialized = true;
    this._return_sequenceIntialized = true;
  }
  copy(other) {
    this._refObject = new QshellRetvalRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._return_valueIntialized = true;
    this._return_sequenceIntialized = true;
  }
  static get classType() {
    return QshellRetvalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"return_value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"return_sequence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"QshellRetval","stringUpperBound":null,"isPrimitiveType":false},"msgName":"QshellRetval"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","return_value","return_sequence"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class QshellRetvalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof QshellRetvalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new QshellRetvalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QshellRetvalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QshellRetvalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new QshellRetvalRefArray(size);
    this._refObject = new QshellRetvalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new QshellRetvalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof QshellRetvalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "QshellRetvalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      QshellRetvalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return QshellRetvalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return QshellRetvalArrayWrapper;
  }
}
module.exports = QshellRetvalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "return_value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "return_sequence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "QshellRetval",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "QshellRetval"
}
*/
