// rclnodejs/generated/px4_msgs/px4_msgs__msg__GimbalManagerSetManualControl.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GimbalManagerSetManualControlRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  origin_sysid: primitiveTypes.uint8,
  origin_compid: primitiveTypes.uint8,
  target_system: primitiveTypes.uint8,
  target_component: primitiveTypes.uint8,
  flags: primitiveTypes.uint32,
  gimbal_device_id: primitiveTypes.uint8,
  pitch: primitiveTypes.float32,
  yaw: primitiveTypes.float32,
  pitch_rate: primitiveTypes.float32,
  yaw_rate: primitiveTypes.float32,
});
const GimbalManagerSetManualControlRefArray = ArrayType(GimbalManagerSetManualControlRefStruct);
const GimbalManagerSetManualControlRefStructArray = StructType({
  data: GimbalManagerSetManualControlRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GimbalManagerSetManualControlWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GimbalManagerSetManualControlRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._origin_sysidIntialized = true;
      this._origin_compidIntialized = true;
      this._target_systemIntialized = true;
      this._target_componentIntialized = true;
      this._flagsIntialized = true;
      this._gimbal_device_idIntialized = true;
      this._pitchIntialized = true;
      this._yawIntialized = true;
      this._pitch_rateIntialized = true;
      this._yaw_rateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GimbalManagerSetManualControlRefStruct();
    this._timestampIntialized = false;
    this._origin_sysidIntialized = false;
    this._origin_compidIntialized = false;
    this._target_systemIntialized = false;
    this._target_componentIntialized = false;
    this._flagsIntialized = false;
    this._gimbal_device_idIntialized = false;
    this._pitchIntialized = false;
    this._yawIntialized = false;
    this._pitch_rateIntialized = false;
    this._yaw_rateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GimbalManagerSetManualControlWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GimbalManagerSetManualControlArrayWrapper;
  }
  static get ArrayType() {
    return GimbalManagerSetManualControlArrayWrapper;
  }
  static get refObjectArrayType() {
    return GimbalManagerSetManualControlRefStructArray
  }
  static get refObjectType() {
    return GimbalManagerSetManualControlRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._origin_sysidIntialized) {
        throw new TypeError('Invalid argument: origin_sysid in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._origin_compidIntialized) {
        throw new TypeError('Invalid argument: origin_compid in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._target_systemIntialized) {
        throw new TypeError('Invalid argument: target_system in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._target_componentIntialized) {
        throw new TypeError('Invalid argument: target_component in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._flagsIntialized) {
        throw new TypeError('Invalid argument: flags in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._gimbal_device_idIntialized) {
        throw new TypeError('Invalid argument: gimbal_device_id in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._pitchIntialized) {
        throw new TypeError('Invalid argument: pitch in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._pitch_rateIntialized) {
        throw new TypeError('Invalid argument: pitch_rate in GimbalManagerSetManualControl');
      }
      if (checkConsistency && !this._yaw_rateIntialized) {
        throw new TypeError('Invalid argument: yaw_rate in GimbalManagerSetManualControl');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._origin_sysidIntialized = true;
    this._refObject.origin_sysid = refObject.origin_sysid;
    this._origin_compidIntialized = true;
    this._refObject.origin_compid = refObject.origin_compid;
    this._target_systemIntialized = true;
    this._refObject.target_system = refObject.target_system;
    this._target_componentIntialized = true;
    this._refObject.target_component = refObject.target_component;
    this._flagsIntialized = true;
    this._refObject.flags = refObject.flags;
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = refObject.gimbal_device_id;
    this._pitchIntialized = true;
    this._refObject.pitch = refObject.pitch;
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
    this._pitch_rateIntialized = true;
    this._refObject.pitch_rate = refObject.pitch_rate;
    this._yaw_rateIntialized = true;
    this._refObject.yaw_rate = refObject.yaw_rate;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GimbalManagerSetManualControlWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'GimbalManagerSetManualControl'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get origin_sysid() {
    if (!this._origin_sysidIntialized) {
      return undefined;
    }
    return this._refObject.origin_sysid;
  }
  set origin_sysid(value) {
    this._origin_sysidIntialized = true;
    this._refObject.origin_sysid = value;
  }
  get origin_compid() {
    if (!this._origin_compidIntialized) {
      return undefined;
    }
    return this._refObject.origin_compid;
  }
  set origin_compid(value) {
    this._origin_compidIntialized = true;
    this._refObject.origin_compid = value;
  }
  get target_system() {
    if (!this._target_systemIntialized) {
      return undefined;
    }
    return this._refObject.target_system;
  }
  set target_system(value) {
    this._target_systemIntialized = true;
    this._refObject.target_system = value;
  }
  get target_component() {
    if (!this._target_componentIntialized) {
      return undefined;
    }
    return this._refObject.target_component;
  }
  set target_component(value) {
    this._target_componentIntialized = true;
    this._refObject.target_component = value;
  }
  get flags() {
    if (!this._flagsIntialized) {
      return undefined;
    }
    return this._refObject.flags;
  }
  set flags(value) {
    this._flagsIntialized = true;
    this._refObject.flags = value;
  }
  get gimbal_device_id() {
    if (!this._gimbal_device_idIntialized) {
      return undefined;
    }
    return this._refObject.gimbal_device_id;
  }
  set gimbal_device_id(value) {
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = value;
  }
  get pitch() {
    if (!this._pitchIntialized) {
      return undefined;
    }
    return this._refObject.pitch;
  }
  set pitch(value) {
    this._pitchIntialized = true;
    this._refObject.pitch = value;
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  get pitch_rate() {
    if (!this._pitch_rateIntialized) {
      return undefined;
    }
    return this._refObject.pitch_rate;
  }
  set pitch_rate(value) {
    this._pitch_rateIntialized = true;
    this._refObject.pitch_rate = value;
  }
  get yaw_rate() {
    if (!this._yaw_rateIntialized) {
      return undefined;
    }
    return this._refObject.yaw_rate;
  }
  set yaw_rate(value) {
    this._yaw_rateIntialized = true;
    this._refObject.yaw_rate = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GimbalManagerSetManualControlRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._origin_sysidIntialized = true;
    this._origin_compidIntialized = true;
    this._target_systemIntialized = true;
    this._target_componentIntialized = true;
    this._flagsIntialized = true;
    this._gimbal_device_idIntialized = true;
    this._pitchIntialized = true;
    this._yawIntialized = true;
    this._pitch_rateIntialized = true;
    this._yaw_rateIntialized = true;
  }
  copy(other) {
    this._refObject = new GimbalManagerSetManualControlRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._origin_sysidIntialized = true;
    this._origin_compidIntialized = true;
    this._target_systemIntialized = true;
    this._target_componentIntialized = true;
    this._flagsIntialized = true;
    this._gimbal_device_idIntialized = true;
    this._pitchIntialized = true;
    this._yawIntialized = true;
    this._pitch_rateIntialized = true;
    this._yaw_rateIntialized = true;
  }
  static get classType() {
    return GimbalManagerSetManualControlWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_RETRACT","value":1},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_NEUTRAL","value":2},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_ROLL_LOCK","value":4},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_PITCH_LOCK","value":8},{"type":"uint32","name":"GIMBAL_MANAGER_FLAGS_YAW_LOCK","value":16}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"origin_sysid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"origin_compid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"target_system","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"target_component","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gimbal_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"GimbalManagerSetManualControl","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GimbalManagerSetManualControl"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","origin_sysid","origin_compid","target_system","target_component","flags","gimbal_device_id","pitch","yaw","pitch_rate","yaw_rate"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GimbalManagerSetManualControlArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GimbalManagerSetManualControlWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GimbalManagerSetManualControlWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerSetManualControlArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerSetManualControlArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GimbalManagerSetManualControlRefArray(size);
    this._refObject = new GimbalManagerSetManualControlRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GimbalManagerSetManualControlWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GimbalManagerSetManualControlArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GimbalManagerSetManualControlArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GimbalManagerSetManualControlWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GimbalManagerSetManualControlWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GimbalManagerSetManualControlArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(GimbalManagerSetManualControlWrapper, "GIMBAL_MANAGER_FLAGS_RETRACT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerSetManualControlWrapper, "GIMBAL_MANAGER_FLAGS_NEUTRAL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerSetManualControlWrapper, "GIMBAL_MANAGER_FLAGS_ROLL_LOCK", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerSetManualControlWrapper, "GIMBAL_MANAGER_FLAGS_PITCH_LOCK", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalManagerSetManualControlWrapper, "GIMBAL_MANAGER_FLAGS_YAW_LOCK", {value: 16, writable: false, enumerable: true, configurable: true});
module.exports = GimbalManagerSetManualControlWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_RETRACT",
      "value": 1
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_NEUTRAL",
      "value": 2
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_ROLL_LOCK",
      "value": 4
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_PITCH_LOCK",
      "value": 8
    },
    {
      "type": "uint32",
      "name": "GIMBAL_MANAGER_FLAGS_YAW_LOCK",
      "value": 16
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "origin_sysid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "origin_compid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "target_system",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "target_component",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gimbal_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "GimbalManagerSetManualControl",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GimbalManagerSetManualControl"
}
*/
