// rclnodejs/generated/px4_msgs/px4_msgs__msg__GpsDump.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const GpsDumpRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  instance: primitiveTypes.uint8,
  len: primitiveTypes.uint8,
  data: ArrayType(primitiveTypes.uint8, 79),
});
const GpsDumpRefArray = ArrayType(GpsDumpRefStruct);
const GpsDumpRefStructArray = StructType({
  data: GpsDumpRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GpsDumpWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GpsDumpRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._instanceIntialized = true;
      this._lenIntialized = true;
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GpsDumpRefStruct();
    this._timestampIntialized = false;
    this._instanceIntialized = false;
    this._lenIntialized = false;
    this._wrapperFields.data = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GpsDumpWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GpsDumpArrayWrapper;
  }
  static get ArrayType() {
    return GpsDumpArrayWrapper;
  }
  static get refObjectArrayType() {
    return GpsDumpRefStructArray
  }
  static get refObjectType() {
    return GpsDumpRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in GpsDump');
      }
      if (checkConsistency && !this._instanceIntialized) {
        throw new TypeError('Invalid argument: instance in GpsDump');
      }
      if (checkConsistency && !this._lenIntialized) {
        throw new TypeError('Invalid argument: len in GpsDump');
      }
    this._refObject.data = Array.from(this._wrapperFields.data.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._instanceIntialized = true;
    this._refObject.instance = refObject.instance;
    this._lenIntialized = true;
    this._refObject.len = refObject.len;
    this._wrapperFields.data.fill(refObject.data.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GpsDumpWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'GpsDump'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get instance() {
    if (!this._instanceIntialized) {
      return undefined;
    }
    return this._refObject.instance;
  }
  set instance(value) {
    this._instanceIntialized = true;
    this._refObject.instance = value;
  }
  get len() {
    if (!this._lenIntialized) {
      return undefined;
    }
    return this._refObject.len;
  }
  set len(value) {
    this._lenIntialized = true;
    this._refObject.len = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    if (value.length !== 79) {
      throw new RangeError('The length of the array must be 79.');
    }
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GpsDumpRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._instanceIntialized = true;
    this._lenIntialized = true;
    this._wrapperFields.data.fill(refObject.data.toArray());
  }
  copy(other) {
    this._refObject = new GpsDumpRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._instanceIntialized = true;
    this._lenIntialized = true;
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return GpsDumpWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ORB_QUEUE_LENGTH","value":8}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"instance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"len","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":79,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"GpsDump","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GpsDump"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","instance","len","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GpsDumpArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GpsDumpWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GpsDumpWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GpsDumpArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GpsDumpArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GpsDumpRefArray(size);
    this._refObject = new GpsDumpRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GpsDumpWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GpsDumpArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GpsDumpArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GpsDumpWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GpsDumpWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GpsDumpArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(GpsDumpWrapper, "ORB_QUEUE_LENGTH", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = GpsDumpWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ORB_QUEUE_LENGTH",
      "value": 8
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "instance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "len",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": 79,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "GpsDump",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GpsDump"
}
*/
