// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleOdometry.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VehicleOdometryRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  pose_frame: primitiveTypes.uint8,
  position: ArrayType(primitiveTypes.float32, 3),
  q: ArrayType(primitiveTypes.float32, 4),
  velocity_frame: primitiveTypes.uint8,
  velocity: ArrayType(primitiveTypes.float32, 3),
  angular_velocity: ArrayType(primitiveTypes.float32, 3),
  position_variance: ArrayType(primitiveTypes.float32, 3),
  orientation_variance: ArrayType(primitiveTypes.float32, 3),
  velocity_variance: ArrayType(primitiveTypes.float32, 3),
  reset_counter: primitiveTypes.uint8,
  quality: primitiveTypes.int8,
});
const VehicleOdometryRefArray = ArrayType(VehicleOdometryRefStruct);
const VehicleOdometryRefStructArray = StructType({
  data: VehicleOdometryRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleOdometryWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleOdometryRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._pose_frameIntialized = true;
      this._wrapperFields.position = Float32Wrapper.createArray();
      this._wrapperFields.position.copy(other._wrapperFields.position);
      this._wrapperFields.q = Float32Wrapper.createArray();
      this._wrapperFields.q.copy(other._wrapperFields.q);
      this._velocity_frameIntialized = true;
      this._wrapperFields.velocity = Float32Wrapper.createArray();
      this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
      this._wrapperFields.angular_velocity = Float32Wrapper.createArray();
      this._wrapperFields.angular_velocity.copy(other._wrapperFields.angular_velocity);
      this._wrapperFields.position_variance = Float32Wrapper.createArray();
      this._wrapperFields.position_variance.copy(other._wrapperFields.position_variance);
      this._wrapperFields.orientation_variance = Float32Wrapper.createArray();
      this._wrapperFields.orientation_variance.copy(other._wrapperFields.orientation_variance);
      this._wrapperFields.velocity_variance = Float32Wrapper.createArray();
      this._wrapperFields.velocity_variance.copy(other._wrapperFields.velocity_variance);
      this._reset_counterIntialized = true;
      this._qualityIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleOdometryRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._pose_frameIntialized = false;
    this._wrapperFields.position = Float32Wrapper.createArray();
    this._wrapperFields.q = Float32Wrapper.createArray();
    this._velocity_frameIntialized = false;
    this._wrapperFields.velocity = Float32Wrapper.createArray();
    this._wrapperFields.angular_velocity = Float32Wrapper.createArray();
    this._wrapperFields.position_variance = Float32Wrapper.createArray();
    this._wrapperFields.orientation_variance = Float32Wrapper.createArray();
    this._wrapperFields.velocity_variance = Float32Wrapper.createArray();
    this._reset_counterIntialized = false;
    this._qualityIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleOdometryWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleOdometryArrayWrapper;
  }
  static get ArrayType() {
    return VehicleOdometryArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleOdometryRefStructArray
  }
  static get refObjectType() {
    return VehicleOdometryRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleOdometry');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in VehicleOdometry');
      }
      if (checkConsistency && !this._pose_frameIntialized) {
        throw new TypeError('Invalid argument: pose_frame in VehicleOdometry');
      }
      if (checkConsistency && !this._velocity_frameIntialized) {
        throw new TypeError('Invalid argument: velocity_frame in VehicleOdometry');
      }
      if (checkConsistency && !this._reset_counterIntialized) {
        throw new TypeError('Invalid argument: reset_counter in VehicleOdometry');
      }
      if (checkConsistency && !this._qualityIntialized) {
        throw new TypeError('Invalid argument: quality in VehicleOdometry');
      }
    this._refObject.position = Array.from(this._wrapperFields.position.data);
    this._refObject.q = Array.from(this._wrapperFields.q.data);
    this._refObject.velocity = Array.from(this._wrapperFields.velocity.data);
    this._refObject.angular_velocity = Array.from(this._wrapperFields.angular_velocity.data);
    this._refObject.position_variance = Array.from(this._wrapperFields.position_variance.data);
    this._refObject.orientation_variance = Array.from(this._wrapperFields.orientation_variance.data);
    this._refObject.velocity_variance = Array.from(this._wrapperFields.velocity_variance.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._pose_frameIntialized = true;
    this._refObject.pose_frame = refObject.pose_frame;
    this._wrapperFields.position.fill(refObject.position.toArray());
    this._wrapperFields.q.fill(refObject.q.toArray());
    this._velocity_frameIntialized = true;
    this._refObject.velocity_frame = refObject.velocity_frame;
    this._wrapperFields.velocity.fill(refObject.velocity.toArray());
    this._wrapperFields.angular_velocity.fill(refObject.angular_velocity.toArray());
    this._wrapperFields.position_variance.fill(refObject.position_variance.toArray());
    this._wrapperFields.orientation_variance.fill(refObject.orientation_variance.toArray());
    this._wrapperFields.velocity_variance.fill(refObject.velocity_variance.toArray());
    this._reset_counterIntialized = true;
    this._refObject.reset_counter = refObject.reset_counter;
    this._qualityIntialized = true;
    this._refObject.quality = refObject.quality;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleOdometryWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleOdometry'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get pose_frame() {
    if (!this._pose_frameIntialized) {
      return undefined;
    }
    return this._refObject.pose_frame;
  }
  set pose_frame(value) {
    this._pose_frameIntialized = true;
    this._refObject.pose_frame = value;
  }
  get position() {
    return this._wrapperFields['position'].data;
  }
  set position(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['position'].fill(value);
  }
  get q() {
    return this._wrapperFields['q'].data;
  }
  set q(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields['q'].fill(value);
  }
  get velocity_frame() {
    if (!this._velocity_frameIntialized) {
      return undefined;
    }
    return this._refObject.velocity_frame;
  }
  set velocity_frame(value) {
    this._velocity_frameIntialized = true;
    this._refObject.velocity_frame = value;
  }
  get velocity() {
    return this._wrapperFields['velocity'].data;
  }
  set velocity(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['velocity'].fill(value);
  }
  get angular_velocity() {
    return this._wrapperFields['angular_velocity'].data;
  }
  set angular_velocity(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['angular_velocity'].fill(value);
  }
  get position_variance() {
    return this._wrapperFields['position_variance'].data;
  }
  set position_variance(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['position_variance'].fill(value);
  }
  get orientation_variance() {
    return this._wrapperFields['orientation_variance'].data;
  }
  set orientation_variance(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['orientation_variance'].fill(value);
  }
  get velocity_variance() {
    return this._wrapperFields['velocity_variance'].data;
  }
  set velocity_variance(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['velocity_variance'].fill(value);
  }
  get reset_counter() {
    if (!this._reset_counterIntialized) {
      return undefined;
    }
    return this._refObject.reset_counter;
  }
  set reset_counter(value) {
    this._reset_counterIntialized = true;
    this._refObject.reset_counter = value;
  }
  get quality() {
    if (!this._qualityIntialized) {
      return undefined;
    }
    return this._refObject.quality;
  }
  set quality(value) {
    this._qualityIntialized = true;
    this._refObject.quality = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleOdometryRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._pose_frameIntialized = true;
    this._wrapperFields.position.fill(refObject.position.toArray());
    this._wrapperFields.q.fill(refObject.q.toArray());
    this._velocity_frameIntialized = true;
    this._wrapperFields.velocity.fill(refObject.velocity.toArray());
    this._wrapperFields.angular_velocity.fill(refObject.angular_velocity.toArray());
    this._wrapperFields.position_variance.fill(refObject.position_variance.toArray());
    this._wrapperFields.orientation_variance.fill(refObject.orientation_variance.toArray());
    this._wrapperFields.velocity_variance.fill(refObject.velocity_variance.toArray());
    this._reset_counterIntialized = true;
    this._qualityIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleOdometryRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._pose_frameIntialized = true;
    this._wrapperFields.position.copy(other._wrapperFields.position);
    this._wrapperFields.q.copy(other._wrapperFields.q);
    this._velocity_frameIntialized = true;
    this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
    this._wrapperFields.angular_velocity.copy(other._wrapperFields.angular_velocity);
    this._wrapperFields.position_variance.copy(other._wrapperFields.position_variance);
    this._wrapperFields.orientation_variance.copy(other._wrapperFields.orientation_variance);
    this._wrapperFields.velocity_variance.copy(other._wrapperFields.velocity_variance);
    this._reset_counterIntialized = true;
    this._qualityIntialized = true;
  }
  static get classType() {
    return VehicleOdometryWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"POSE_FRAME_UNKNOWN","value":0},{"type":"uint8","name":"POSE_FRAME_NED","value":1},{"type":"uint8","name":"POSE_FRAME_FRD","value":2},{"type":"uint8","name":"VELOCITY_FRAME_UNKNOWN","value":0},{"type":"uint8","name":"VELOCITY_FRAME_NED","value":1},{"type":"uint8","name":"VELOCITY_FRAME_FRD","value":2},{"type":"uint8","name":"VELOCITY_FRAME_BODY_FRD","value":3}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"q","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular_velocity","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position_variance","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"orientation_variance","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity_variance","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"quality","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleOdometry","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleOdometry"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","pose_frame","position","q","velocity_frame","velocity","angular_velocity","position_variance","orientation_variance","velocity_variance","reset_counter","quality"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleOdometryArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleOdometryWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleOdometryWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleOdometryArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleOdometryArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleOdometryRefArray(size);
    this._refObject = new VehicleOdometryRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleOdometryWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleOdometryArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleOdometryArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleOdometryWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleOdometryWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleOdometryArrayWrapper;
  }
}
// Define constants (7 in total)
Object.defineProperty(VehicleOdometryWrapper, "POSE_FRAME_UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleOdometryWrapper, "POSE_FRAME_NED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleOdometryWrapper, "POSE_FRAME_FRD", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleOdometryWrapper, "VELOCITY_FRAME_UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleOdometryWrapper, "VELOCITY_FRAME_NED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleOdometryWrapper, "VELOCITY_FRAME_FRD", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleOdometryWrapper, "VELOCITY_FRAME_BODY_FRD", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = VehicleOdometryWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "POSE_FRAME_UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "POSE_FRAME_NED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "POSE_FRAME_FRD",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "VELOCITY_FRAME_UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "VELOCITY_FRAME_NED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "VELOCITY_FRAME_FRD",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "VELOCITY_FRAME_BODY_FRD",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pose_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "q",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocity_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocity",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angular_velocity",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position_variance",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "orientation_variance",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocity_variance",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "quality",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleOdometry",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleOdometry"
}
*/
