// rclnodejs/generated/px4_msgs/px4_msgs__msg__IrlockReport.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const IrlockReportRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  signature: primitiveTypes.uint16,
  pos_x: primitiveTypes.float32,
  pos_y: primitiveTypes.float32,
  size_x: primitiveTypes.float32,
  size_y: primitiveTypes.float32,
});
const IrlockReportRefArray = ArrayType(IrlockReportRefStruct);
const IrlockReportRefStructArray = StructType({
  data: IrlockReportRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class IrlockReportWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new IrlockReportRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._signatureIntialized = true;
      this._pos_xIntialized = true;
      this._pos_yIntialized = true;
      this._size_xIntialized = true;
      this._size_yIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new IrlockReportRefStruct();
    this._timestampIntialized = false;
    this._signatureIntialized = false;
    this._pos_xIntialized = false;
    this._pos_yIntialized = false;
    this._size_xIntialized = false;
    this._size_yIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new IrlockReportWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IrlockReportArrayWrapper;
  }
  static get ArrayType() {
    return IrlockReportArrayWrapper;
  }
  static get refObjectArrayType() {
    return IrlockReportRefStructArray
  }
  static get refObjectType() {
    return IrlockReportRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in IrlockReport');
      }
      if (checkConsistency && !this._signatureIntialized) {
        throw new TypeError('Invalid argument: signature in IrlockReport');
      }
      if (checkConsistency && !this._pos_xIntialized) {
        throw new TypeError('Invalid argument: pos_x in IrlockReport');
      }
      if (checkConsistency && !this._pos_yIntialized) {
        throw new TypeError('Invalid argument: pos_y in IrlockReport');
      }
      if (checkConsistency && !this._size_xIntialized) {
        throw new TypeError('Invalid argument: size_x in IrlockReport');
      }
      if (checkConsistency && !this._size_yIntialized) {
        throw new TypeError('Invalid argument: size_y in IrlockReport');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._signatureIntialized = true;
    this._refObject.signature = refObject.signature;
    this._pos_xIntialized = true;
    this._refObject.pos_x = refObject.pos_x;
    this._pos_yIntialized = true;
    this._refObject.pos_y = refObject.pos_y;
    this._size_xIntialized = true;
    this._refObject.size_x = refObject.size_x;
    this._size_yIntialized = true;
    this._refObject.size_y = refObject.size_y;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    IrlockReportWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'IrlockReport'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get signature() {
    if (!this._signatureIntialized) {
      return undefined;
    }
    return this._refObject.signature;
  }
  set signature(value) {
    this._signatureIntialized = true;
    this._refObject.signature = value;
  }
  get pos_x() {
    if (!this._pos_xIntialized) {
      return undefined;
    }
    return this._refObject.pos_x;
  }
  set pos_x(value) {
    this._pos_xIntialized = true;
    this._refObject.pos_x = value;
  }
  get pos_y() {
    if (!this._pos_yIntialized) {
      return undefined;
    }
    return this._refObject.pos_y;
  }
  set pos_y(value) {
    this._pos_yIntialized = true;
    this._refObject.pos_y = value;
  }
  get size_x() {
    if (!this._size_xIntialized) {
      return undefined;
    }
    return this._refObject.size_x;
  }
  set size_x(value) {
    this._size_xIntialized = true;
    this._refObject.size_x = value;
  }
  get size_y() {
    if (!this._size_yIntialized) {
      return undefined;
    }
    return this._refObject.size_y;
  }
  set size_y(value) {
    this._size_yIntialized = true;
    this._refObject.size_y = value;
  }
  copyRefObject(refObject) {
    this._refObject = new IrlockReportRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._signatureIntialized = true;
    this._pos_xIntialized = true;
    this._pos_yIntialized = true;
    this._size_xIntialized = true;
    this._size_yIntialized = true;
  }
  copy(other) {
    this._refObject = new IrlockReportRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._signatureIntialized = true;
    this._pos_xIntialized = true;
    this._pos_yIntialized = true;
    this._size_xIntialized = true;
    this._size_yIntialized = true;
  }
  static get classType() {
    return IrlockReportWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"signature","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pos_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pos_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"size_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"IrlockReport","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IrlockReport"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","signature","pos_x","pos_y","size_x","size_y"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class IrlockReportArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IrlockReportWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IrlockReportWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IrlockReportArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IrlockReportArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new IrlockReportRefArray(size);
    this._refObject = new IrlockReportRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IrlockReportWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IrlockReportArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IrlockReportArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IrlockReportWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IrlockReportWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IrlockReportArrayWrapper;
  }
}
module.exports = IrlockReportWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "signature",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pos_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pos_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "size_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "size_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "IrlockReport",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "IrlockReport"
}
*/
