// rclnodejs/generated/px4_msgs/px4_msgs__msg__Wind.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const WindRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  windspeed_north: primitiveTypes.float32,
  windspeed_east: primitiveTypes.float32,
  variance_north: primitiveTypes.float32,
  variance_east: primitiveTypes.float32,
  tas_innov: primitiveTypes.float32,
  tas_innov_var: primitiveTypes.float32,
  beta_innov: primitiveTypes.float32,
  beta_innov_var: primitiveTypes.float32,
});
const WindRefArray = ArrayType(WindRefStruct);
const WindRefStructArray = StructType({
  data: WindRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WindWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WindRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._windspeed_northIntialized = true;
      this._windspeed_eastIntialized = true;
      this._variance_northIntialized = true;
      this._variance_eastIntialized = true;
      this._tas_innovIntialized = true;
      this._tas_innov_varIntialized = true;
      this._beta_innovIntialized = true;
      this._beta_innov_varIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WindRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._windspeed_northIntialized = false;
    this._windspeed_eastIntialized = false;
    this._variance_northIntialized = false;
    this._variance_eastIntialized = false;
    this._tas_innovIntialized = false;
    this._tas_innov_varIntialized = false;
    this._beta_innovIntialized = false;
    this._beta_innov_varIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new WindWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WindArrayWrapper;
  }
  static get ArrayType() {
    return WindArrayWrapper;
  }
  static get refObjectArrayType() {
    return WindRefStructArray
  }
  static get refObjectType() {
    return WindRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in Wind');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in Wind');
      }
      if (checkConsistency && !this._windspeed_northIntialized) {
        throw new TypeError('Invalid argument: windspeed_north in Wind');
      }
      if (checkConsistency && !this._windspeed_eastIntialized) {
        throw new TypeError('Invalid argument: windspeed_east in Wind');
      }
      if (checkConsistency && !this._variance_northIntialized) {
        throw new TypeError('Invalid argument: variance_north in Wind');
      }
      if (checkConsistency && !this._variance_eastIntialized) {
        throw new TypeError('Invalid argument: variance_east in Wind');
      }
      if (checkConsistency && !this._tas_innovIntialized) {
        throw new TypeError('Invalid argument: tas_innov in Wind');
      }
      if (checkConsistency && !this._tas_innov_varIntialized) {
        throw new TypeError('Invalid argument: tas_innov_var in Wind');
      }
      if (checkConsistency && !this._beta_innovIntialized) {
        throw new TypeError('Invalid argument: beta_innov in Wind');
      }
      if (checkConsistency && !this._beta_innov_varIntialized) {
        throw new TypeError('Invalid argument: beta_innov_var in Wind');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._windspeed_northIntialized = true;
    this._refObject.windspeed_north = refObject.windspeed_north;
    this._windspeed_eastIntialized = true;
    this._refObject.windspeed_east = refObject.windspeed_east;
    this._variance_northIntialized = true;
    this._refObject.variance_north = refObject.variance_north;
    this._variance_eastIntialized = true;
    this._refObject.variance_east = refObject.variance_east;
    this._tas_innovIntialized = true;
    this._refObject.tas_innov = refObject.tas_innov;
    this._tas_innov_varIntialized = true;
    this._refObject.tas_innov_var = refObject.tas_innov_var;
    this._beta_innovIntialized = true;
    this._refObject.beta_innov = refObject.beta_innov;
    this._beta_innov_varIntialized = true;
    this._refObject.beta_innov_var = refObject.beta_innov_var;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    WindWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'Wind'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get windspeed_north() {
    if (!this._windspeed_northIntialized) {
      return undefined;
    }
    return this._refObject.windspeed_north;
  }
  set windspeed_north(value) {
    this._windspeed_northIntialized = true;
    this._refObject.windspeed_north = value;
  }
  get windspeed_east() {
    if (!this._windspeed_eastIntialized) {
      return undefined;
    }
    return this._refObject.windspeed_east;
  }
  set windspeed_east(value) {
    this._windspeed_eastIntialized = true;
    this._refObject.windspeed_east = value;
  }
  get variance_north() {
    if (!this._variance_northIntialized) {
      return undefined;
    }
    return this._refObject.variance_north;
  }
  set variance_north(value) {
    this._variance_northIntialized = true;
    this._refObject.variance_north = value;
  }
  get variance_east() {
    if (!this._variance_eastIntialized) {
      return undefined;
    }
    return this._refObject.variance_east;
  }
  set variance_east(value) {
    this._variance_eastIntialized = true;
    this._refObject.variance_east = value;
  }
  get tas_innov() {
    if (!this._tas_innovIntialized) {
      return undefined;
    }
    return this._refObject.tas_innov;
  }
  set tas_innov(value) {
    this._tas_innovIntialized = true;
    this._refObject.tas_innov = value;
  }
  get tas_innov_var() {
    if (!this._tas_innov_varIntialized) {
      return undefined;
    }
    return this._refObject.tas_innov_var;
  }
  set tas_innov_var(value) {
    this._tas_innov_varIntialized = true;
    this._refObject.tas_innov_var = value;
  }
  get beta_innov() {
    if (!this._beta_innovIntialized) {
      return undefined;
    }
    return this._refObject.beta_innov;
  }
  set beta_innov(value) {
    this._beta_innovIntialized = true;
    this._refObject.beta_innov = value;
  }
  get beta_innov_var() {
    if (!this._beta_innov_varIntialized) {
      return undefined;
    }
    return this._refObject.beta_innov_var;
  }
  set beta_innov_var(value) {
    this._beta_innov_varIntialized = true;
    this._refObject.beta_innov_var = value;
  }
  copyRefObject(refObject) {
    this._refObject = new WindRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._windspeed_northIntialized = true;
    this._windspeed_eastIntialized = true;
    this._variance_northIntialized = true;
    this._variance_eastIntialized = true;
    this._tas_innovIntialized = true;
    this._tas_innov_varIntialized = true;
    this._beta_innovIntialized = true;
    this._beta_innov_varIntialized = true;
  }
  copy(other) {
    this._refObject = new WindRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._windspeed_northIntialized = true;
    this._windspeed_eastIntialized = true;
    this._variance_northIntialized = true;
    this._variance_eastIntialized = true;
    this._tas_innovIntialized = true;
    this._tas_innov_varIntialized = true;
    this._beta_innovIntialized = true;
    this._beta_innov_varIntialized = true;
  }
  static get classType() {
    return WindWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"windspeed_north","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"windspeed_east","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"variance_north","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"variance_east","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tas_innov","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tas_innov_var","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"beta_innov","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"beta_innov_var","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"Wind","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Wind"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","windspeed_north","windspeed_east","variance_north","variance_east","tas_innov","tas_innov_var","beta_innov","beta_innov_var"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WindArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WindWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WindWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WindArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WindArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WindRefArray(size);
    this._refObject = new WindRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WindWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WindArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WindArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WindWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WindWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WindArrayWrapper;
  }
}
module.exports = WindWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "windspeed_north",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "windspeed_east",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "variance_north",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "variance_east",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tas_innov",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tas_innov_var",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "beta_innov",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "beta_innov_var",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "Wind",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Wind"
}
*/
