// rclnodejs/generated/px4_msgs/px4_msgs__msg__TakeoffStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TakeoffStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  takeoff_state: primitiveTypes.uint8,
  tilt_limit: primitiveTypes.float32,
});
const TakeoffStatusRefArray = ArrayType(TakeoffStatusRefStruct);
const TakeoffStatusRefStructArray = StructType({
  data: TakeoffStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TakeoffStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TakeoffStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._takeoff_stateIntialized = true;
      this._tilt_limitIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TakeoffStatusRefStruct();
    this._timestampIntialized = false;
    this._takeoff_stateIntialized = false;
    this._tilt_limitIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new TakeoffStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TakeoffStatusArrayWrapper;
  }
  static get ArrayType() {
    return TakeoffStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return TakeoffStatusRefStructArray
  }
  static get refObjectType() {
    return TakeoffStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in TakeoffStatus');
      }
      if (checkConsistency && !this._takeoff_stateIntialized) {
        throw new TypeError('Invalid argument: takeoff_state in TakeoffStatus');
      }
      if (checkConsistency && !this._tilt_limitIntialized) {
        throw new TypeError('Invalid argument: tilt_limit in TakeoffStatus');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._takeoff_stateIntialized = true;
    this._refObject.takeoff_state = refObject.takeoff_state;
    this._tilt_limitIntialized = true;
    this._refObject.tilt_limit = refObject.tilt_limit;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    TakeoffStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'TakeoffStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get takeoff_state() {
    if (!this._takeoff_stateIntialized) {
      return undefined;
    }
    return this._refObject.takeoff_state;
  }
  set takeoff_state(value) {
    this._takeoff_stateIntialized = true;
    this._refObject.takeoff_state = value;
  }
  get tilt_limit() {
    if (!this._tilt_limitIntialized) {
      return undefined;
    }
    return this._refObject.tilt_limit;
  }
  set tilt_limit(value) {
    this._tilt_limitIntialized = true;
    this._refObject.tilt_limit = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TakeoffStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._takeoff_stateIntialized = true;
    this._tilt_limitIntialized = true;
  }
  copy(other) {
    this._refObject = new TakeoffStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._takeoff_stateIntialized = true;
    this._tilt_limitIntialized = true;
  }
  static get classType() {
    return TakeoffStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"TAKEOFF_STATE_UNINITIALIZED","value":0},{"type":"uint8","name":"TAKEOFF_STATE_DISARMED","value":1},{"type":"uint8","name":"TAKEOFF_STATE_SPOOLUP","value":2},{"type":"uint8","name":"TAKEOFF_STATE_READY_FOR_TAKEOFF","value":3},{"type":"uint8","name":"TAKEOFF_STATE_RAMPUP","value":4},{"type":"uint8","name":"TAKEOFF_STATE_FLIGHT","value":5}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"takeoff_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tilt_limit","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"TakeoffStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TakeoffStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","takeoff_state","tilt_limit"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TakeoffStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TakeoffStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TakeoffStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TakeoffStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TakeoffStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TakeoffStatusRefArray(size);
    this._refObject = new TakeoffStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TakeoffStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TakeoffStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TakeoffStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TakeoffStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TakeoffStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TakeoffStatusArrayWrapper;
  }
}
// Define constants (6 in total)
Object.defineProperty(TakeoffStatusWrapper, "TAKEOFF_STATE_UNINITIALIZED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(TakeoffStatusWrapper, "TAKEOFF_STATE_DISARMED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(TakeoffStatusWrapper, "TAKEOFF_STATE_SPOOLUP", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(TakeoffStatusWrapper, "TAKEOFF_STATE_READY_FOR_TAKEOFF", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(TakeoffStatusWrapper, "TAKEOFF_STATE_RAMPUP", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(TakeoffStatusWrapper, "TAKEOFF_STATE_FLIGHT", {value: 5, writable: false, enumerable: true, configurable: true});
module.exports = TakeoffStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "TAKEOFF_STATE_UNINITIALIZED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "TAKEOFF_STATE_DISARMED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "TAKEOFF_STATE_SPOOLUP",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "TAKEOFF_STATE_READY_FOR_TAKEOFF",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "TAKEOFF_STATE_RAMPUP",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "TAKEOFF_STATE_FLIGHT",
      "value": 5
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "takeoff_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tilt_limit",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "TakeoffStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TakeoffStatus"
}
*/
