// rclnodejs/generated/px4_msgs/px4_msgs__msg__RateCtrlStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RateCtrlStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  rollspeed_integ: primitiveTypes.float32,
  pitchspeed_integ: primitiveTypes.float32,
  yawspeed_integ: primitiveTypes.float32,
  wheel_rate_integ: primitiveTypes.float32,
});
const RateCtrlStatusRefArray = ArrayType(RateCtrlStatusRefStruct);
const RateCtrlStatusRefStructArray = StructType({
  data: RateCtrlStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RateCtrlStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RateCtrlStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._rollspeed_integIntialized = true;
      this._pitchspeed_integIntialized = true;
      this._yawspeed_integIntialized = true;
      this._wheel_rate_integIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RateCtrlStatusRefStruct();
    this._timestampIntialized = false;
    this._rollspeed_integIntialized = false;
    this._pitchspeed_integIntialized = false;
    this._yawspeed_integIntialized = false;
    this._wheel_rate_integIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new RateCtrlStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RateCtrlStatusArrayWrapper;
  }
  static get ArrayType() {
    return RateCtrlStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return RateCtrlStatusRefStructArray
  }
  static get refObjectType() {
    return RateCtrlStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in RateCtrlStatus');
      }
      if (checkConsistency && !this._rollspeed_integIntialized) {
        throw new TypeError('Invalid argument: rollspeed_integ in RateCtrlStatus');
      }
      if (checkConsistency && !this._pitchspeed_integIntialized) {
        throw new TypeError('Invalid argument: pitchspeed_integ in RateCtrlStatus');
      }
      if (checkConsistency && !this._yawspeed_integIntialized) {
        throw new TypeError('Invalid argument: yawspeed_integ in RateCtrlStatus');
      }
      if (checkConsistency && !this._wheel_rate_integIntialized) {
        throw new TypeError('Invalid argument: wheel_rate_integ in RateCtrlStatus');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._rollspeed_integIntialized = true;
    this._refObject.rollspeed_integ = refObject.rollspeed_integ;
    this._pitchspeed_integIntialized = true;
    this._refObject.pitchspeed_integ = refObject.pitchspeed_integ;
    this._yawspeed_integIntialized = true;
    this._refObject.yawspeed_integ = refObject.yawspeed_integ;
    this._wheel_rate_integIntialized = true;
    this._refObject.wheel_rate_integ = refObject.wheel_rate_integ;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    RateCtrlStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'RateCtrlStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get rollspeed_integ() {
    if (!this._rollspeed_integIntialized) {
      return undefined;
    }
    return this._refObject.rollspeed_integ;
  }
  set rollspeed_integ(value) {
    this._rollspeed_integIntialized = true;
    this._refObject.rollspeed_integ = value;
  }
  get pitchspeed_integ() {
    if (!this._pitchspeed_integIntialized) {
      return undefined;
    }
    return this._refObject.pitchspeed_integ;
  }
  set pitchspeed_integ(value) {
    this._pitchspeed_integIntialized = true;
    this._refObject.pitchspeed_integ = value;
  }
  get yawspeed_integ() {
    if (!this._yawspeed_integIntialized) {
      return undefined;
    }
    return this._refObject.yawspeed_integ;
  }
  set yawspeed_integ(value) {
    this._yawspeed_integIntialized = true;
    this._refObject.yawspeed_integ = value;
  }
  get wheel_rate_integ() {
    if (!this._wheel_rate_integIntialized) {
      return undefined;
    }
    return this._refObject.wheel_rate_integ;
  }
  set wheel_rate_integ(value) {
    this._wheel_rate_integIntialized = true;
    this._refObject.wheel_rate_integ = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RateCtrlStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._rollspeed_integIntialized = true;
    this._pitchspeed_integIntialized = true;
    this._yawspeed_integIntialized = true;
    this._wheel_rate_integIntialized = true;
  }
  copy(other) {
    this._refObject = new RateCtrlStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._rollspeed_integIntialized = true;
    this._pitchspeed_integIntialized = true;
    this._yawspeed_integIntialized = true;
    this._wheel_rate_integIntialized = true;
  }
  static get classType() {
    return RateCtrlStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rollspeed_integ","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitchspeed_integ","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yawspeed_integ","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"wheel_rate_integ","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"RateCtrlStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RateCtrlStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","rollspeed_integ","pitchspeed_integ","yawspeed_integ","wheel_rate_integ"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RateCtrlStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RateCtrlStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RateCtrlStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RateCtrlStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RateCtrlStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RateCtrlStatusRefArray(size);
    this._refObject = new RateCtrlStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RateCtrlStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RateCtrlStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RateCtrlStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RateCtrlStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RateCtrlStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RateCtrlStatusArrayWrapper;
  }
}
module.exports = RateCtrlStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rollspeed_integ",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitchspeed_integ",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yawspeed_integ",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "wheel_rate_integ",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "RateCtrlStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RateCtrlStatus"
}
*/
