// rclnodejs/generated/px4_msgs/px4_msgs__msg__CameraTrigger.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CameraTriggerRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_utc: primitiveTypes.uint64,
  seq: primitiveTypes.uint32,
  feedback: primitiveTypes.bool,
});
const CameraTriggerRefArray = ArrayType(CameraTriggerRefStruct);
const CameraTriggerRefStructArray = StructType({
  data: CameraTriggerRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CameraTriggerWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CameraTriggerRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_utcIntialized = true;
      this._seqIntialized = true;
      this._feedbackIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CameraTriggerRefStruct();
    this._timestampIntialized = false;
    this._timestamp_utcIntialized = false;
    this._seqIntialized = false;
    this._feedbackIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CameraTriggerWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CameraTriggerArrayWrapper;
  }
  static get ArrayType() {
    return CameraTriggerArrayWrapper;
  }
  static get refObjectArrayType() {
    return CameraTriggerRefStructArray
  }
  static get refObjectType() {
    return CameraTriggerRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in CameraTrigger');
      }
      if (checkConsistency && !this._timestamp_utcIntialized) {
        throw new TypeError('Invalid argument: timestamp_utc in CameraTrigger');
      }
      if (checkConsistency && !this._seqIntialized) {
        throw new TypeError('Invalid argument: seq in CameraTrigger');
      }
      if (checkConsistency && !this._feedbackIntialized) {
        throw new TypeError('Invalid argument: feedback in CameraTrigger');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_utcIntialized = true;
    this._refObject.timestamp_utc = refObject.timestamp_utc;
    this._seqIntialized = true;
    this._refObject.seq = refObject.seq;
    this._feedbackIntialized = true;
    this._refObject.feedback = refObject.feedback;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CameraTriggerWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'CameraTrigger'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_utc() {
    if (!this._timestamp_utcIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_utc;
  }
  set timestamp_utc(value) {
    this._timestamp_utcIntialized = true;
    this._refObject.timestamp_utc = value;
  }
  get seq() {
    if (!this._seqIntialized) {
      return undefined;
    }
    return this._refObject.seq;
  }
  set seq(value) {
    this._seqIntialized = true;
    this._refObject.seq = value;
  }
  get feedback() {
    if (!this._feedbackIntialized) {
      return undefined;
    }
    return this._refObject.feedback;
  }
  set feedback(value) {
    this._feedbackIntialized = true;
    this._refObject.feedback = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CameraTriggerRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_utcIntialized = true;
    this._seqIntialized = true;
    this._feedbackIntialized = true;
  }
  copy(other) {
    this._refObject = new CameraTriggerRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_utcIntialized = true;
    this._seqIntialized = true;
    this._feedbackIntialized = true;
  }
  static get classType() {
    return CameraTriggerWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint32","name":"ORB_QUEUE_LENGTH","value":2}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_utc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"seq","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"feedback","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"CameraTrigger","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CameraTrigger"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_utc","seq","feedback"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CameraTriggerArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CameraTriggerWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CameraTriggerWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraTriggerArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraTriggerArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CameraTriggerRefArray(size);
    this._refObject = new CameraTriggerRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CameraTriggerWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CameraTriggerArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CameraTriggerArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CameraTriggerWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CameraTriggerWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CameraTriggerArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(CameraTriggerWrapper, "ORB_QUEUE_LENGTH", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = CameraTriggerWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint32",
      "name": "ORB_QUEUE_LENGTH",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_utc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "seq",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "feedback",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "CameraTrigger",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CameraTrigger"
}
*/
