// rclnodejs/generated/px4_msgs/px4_msgs__msg__ActuatorControlsStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const ActuatorControlsStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  control_power: ArrayType(primitiveTypes.float32, 3),
});
const ActuatorControlsStatusRefArray = ArrayType(ActuatorControlsStatusRefStruct);
const ActuatorControlsStatusRefStructArray = StructType({
  data: ActuatorControlsStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ActuatorControlsStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ActuatorControlsStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._wrapperFields.control_power = Float32Wrapper.createArray();
      this._wrapperFields.control_power.copy(other._wrapperFields.control_power);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ActuatorControlsStatusRefStruct();
    this._timestampIntialized = false;
    this._wrapperFields.control_power = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ActuatorControlsStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ActuatorControlsStatusArrayWrapper;
  }
  static get ArrayType() {
    return ActuatorControlsStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return ActuatorControlsStatusRefStructArray
  }
  static get refObjectType() {
    return ActuatorControlsStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in ActuatorControlsStatus');
      }
    this._refObject.control_power = Array.from(this._wrapperFields.control_power.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._wrapperFields.control_power.fill(refObject.control_power.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ActuatorControlsStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'ActuatorControlsStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get control_power() {
    return this._wrapperFields['control_power'].data;
  }
  set control_power(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['control_power'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ActuatorControlsStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.control_power.fill(refObject.control_power.toArray());
  }
  copy(other) {
    this._refObject = new ActuatorControlsStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.control_power.copy(other._wrapperFields.control_power);
  }
  static get classType() {
    return ActuatorControlsStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_power","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"ActuatorControlsStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ActuatorControlsStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","control_power"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ActuatorControlsStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ActuatorControlsStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ActuatorControlsStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorControlsStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorControlsStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ActuatorControlsStatusRefArray(size);
    this._refObject = new ActuatorControlsStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ActuatorControlsStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ActuatorControlsStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ActuatorControlsStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ActuatorControlsStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ActuatorControlsStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ActuatorControlsStatusArrayWrapper;
  }
}
module.exports = ActuatorControlsStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "control_power",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "ActuatorControlsStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ActuatorControlsStatus"
}
*/
