// rclnodejs/generated/px4_msgs/px4_msgs__msg__OffboardControlMode.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const OffboardControlModeRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  position: primitiveTypes.bool,
  velocity: primitiveTypes.bool,
  acceleration: primitiveTypes.bool,
  attitude: primitiveTypes.bool,
  body_rate: primitiveTypes.bool,
  actuator: primitiveTypes.bool,
});
const OffboardControlModeRefArray = ArrayType(OffboardControlModeRefStruct);
const OffboardControlModeRefStructArray = StructType({
  data: OffboardControlModeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OffboardControlModeWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OffboardControlModeRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._positionIntialized = true;
      this._velocityIntialized = true;
      this._accelerationIntialized = true;
      this._attitudeIntialized = true;
      this._body_rateIntialized = true;
      this._actuatorIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OffboardControlModeRefStruct();
    this._timestampIntialized = false;
    this._positionIntialized = false;
    this._velocityIntialized = false;
    this._accelerationIntialized = false;
    this._attitudeIntialized = false;
    this._body_rateIntialized = false;
    this._actuatorIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new OffboardControlModeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OffboardControlModeArrayWrapper;
  }
  static get ArrayType() {
    return OffboardControlModeArrayWrapper;
  }
  static get refObjectArrayType() {
    return OffboardControlModeRefStructArray
  }
  static get refObjectType() {
    return OffboardControlModeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in OffboardControlMode');
      }
      if (checkConsistency && !this._positionIntialized) {
        throw new TypeError('Invalid argument: position in OffboardControlMode');
      }
      if (checkConsistency && !this._velocityIntialized) {
        throw new TypeError('Invalid argument: velocity in OffboardControlMode');
      }
      if (checkConsistency && !this._accelerationIntialized) {
        throw new TypeError('Invalid argument: acceleration in OffboardControlMode');
      }
      if (checkConsistency && !this._attitudeIntialized) {
        throw new TypeError('Invalid argument: attitude in OffboardControlMode');
      }
      if (checkConsistency && !this._body_rateIntialized) {
        throw new TypeError('Invalid argument: body_rate in OffboardControlMode');
      }
      if (checkConsistency && !this._actuatorIntialized) {
        throw new TypeError('Invalid argument: actuator in OffboardControlMode');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._positionIntialized = true;
    this._refObject.position = refObject.position;
    this._velocityIntialized = true;
    this._refObject.velocity = refObject.velocity;
    this._accelerationIntialized = true;
    this._refObject.acceleration = refObject.acceleration;
    this._attitudeIntialized = true;
    this._refObject.attitude = refObject.attitude;
    this._body_rateIntialized = true;
    this._refObject.body_rate = refObject.body_rate;
    this._actuatorIntialized = true;
    this._refObject.actuator = refObject.actuator;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    OffboardControlModeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'OffboardControlMode'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get position() {
    if (!this._positionIntialized) {
      return undefined;
    }
    return this._refObject.position;
  }
  set position(value) {
    this._positionIntialized = true;
    this._refObject.position = value;
  }
  get velocity() {
    if (!this._velocityIntialized) {
      return undefined;
    }
    return this._refObject.velocity;
  }
  set velocity(value) {
    this._velocityIntialized = true;
    this._refObject.velocity = value;
  }
  get acceleration() {
    if (!this._accelerationIntialized) {
      return undefined;
    }
    return this._refObject.acceleration;
  }
  set acceleration(value) {
    this._accelerationIntialized = true;
    this._refObject.acceleration = value;
  }
  get attitude() {
    if (!this._attitudeIntialized) {
      return undefined;
    }
    return this._refObject.attitude;
  }
  set attitude(value) {
    this._attitudeIntialized = true;
    this._refObject.attitude = value;
  }
  get body_rate() {
    if (!this._body_rateIntialized) {
      return undefined;
    }
    return this._refObject.body_rate;
  }
  set body_rate(value) {
    this._body_rateIntialized = true;
    this._refObject.body_rate = value;
  }
  get actuator() {
    if (!this._actuatorIntialized) {
      return undefined;
    }
    return this._refObject.actuator;
  }
  set actuator(value) {
    this._actuatorIntialized = true;
    this._refObject.actuator = value;
  }
  copyRefObject(refObject) {
    this._refObject = new OffboardControlModeRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._positionIntialized = true;
    this._velocityIntialized = true;
    this._accelerationIntialized = true;
    this._attitudeIntialized = true;
    this._body_rateIntialized = true;
    this._actuatorIntialized = true;
  }
  copy(other) {
    this._refObject = new OffboardControlModeRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._positionIntialized = true;
    this._velocityIntialized = true;
    this._accelerationIntialized = true;
    this._attitudeIntialized = true;
    this._body_rateIntialized = true;
    this._actuatorIntialized = true;
  }
  static get classType() {
    return OffboardControlModeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"acceleration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"attitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"body_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"actuator","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"OffboardControlMode","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OffboardControlMode"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","position","velocity","acceleration","attitude","body_rate","actuator"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OffboardControlModeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OffboardControlModeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OffboardControlModeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OffboardControlModeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OffboardControlModeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OffboardControlModeRefArray(size);
    this._refObject = new OffboardControlModeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OffboardControlModeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OffboardControlModeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OffboardControlModeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OffboardControlModeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OffboardControlModeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OffboardControlModeArrayWrapper;
  }
}
module.exports = OffboardControlModeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "acceleration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "attitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "body_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "actuator",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "OffboardControlMode",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OffboardControlMode"
}
*/
