// rclnodejs/generated/px4_msgs/px4_msgs__msg__PwmInput.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PwmInputRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  error_count: primitiveTypes.uint64,
  pulse_width: primitiveTypes.uint32,
  period: primitiveTypes.uint32,
});
const PwmInputRefArray = ArrayType(PwmInputRefStruct);
const PwmInputRefStructArray = StructType({
  data: PwmInputRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PwmInputWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PwmInputRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._error_countIntialized = true;
      this._pulse_widthIntialized = true;
      this._periodIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PwmInputRefStruct();
    this._timestampIntialized = false;
    this._error_countIntialized = false;
    this._pulse_widthIntialized = false;
    this._periodIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PwmInputWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PwmInputArrayWrapper;
  }
  static get ArrayType() {
    return PwmInputArrayWrapper;
  }
  static get refObjectArrayType() {
    return PwmInputRefStructArray
  }
  static get refObjectType() {
    return PwmInputRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in PwmInput');
      }
      if (checkConsistency && !this._error_countIntialized) {
        throw new TypeError('Invalid argument: error_count in PwmInput');
      }
      if (checkConsistency && !this._pulse_widthIntialized) {
        throw new TypeError('Invalid argument: pulse_width in PwmInput');
      }
      if (checkConsistency && !this._periodIntialized) {
        throw new TypeError('Invalid argument: period in PwmInput');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._error_countIntialized = true;
    this._refObject.error_count = refObject.error_count;
    this._pulse_widthIntialized = true;
    this._refObject.pulse_width = refObject.pulse_width;
    this._periodIntialized = true;
    this._refObject.period = refObject.period;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PwmInputWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'PwmInput'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get error_count() {
    if (!this._error_countIntialized) {
      return undefined;
    }
    return this._refObject.error_count;
  }
  set error_count(value) {
    this._error_countIntialized = true;
    this._refObject.error_count = value;
  }
  get pulse_width() {
    if (!this._pulse_widthIntialized) {
      return undefined;
    }
    return this._refObject.pulse_width;
  }
  set pulse_width(value) {
    this._pulse_widthIntialized = true;
    this._refObject.pulse_width = value;
  }
  get period() {
    if (!this._periodIntialized) {
      return undefined;
    }
    return this._refObject.period;
  }
  set period(value) {
    this._periodIntialized = true;
    this._refObject.period = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PwmInputRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._error_countIntialized = true;
    this._pulse_widthIntialized = true;
    this._periodIntialized = true;
  }
  copy(other) {
    this._refObject = new PwmInputRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._error_countIntialized = true;
    this._pulse_widthIntialized = true;
    this._periodIntialized = true;
  }
  static get classType() {
    return PwmInputWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pulse_width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"period","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"PwmInput","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PwmInput"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","error_count","pulse_width","period"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PwmInputArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PwmInputWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PwmInputWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PwmInputArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PwmInputArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PwmInputRefArray(size);
    this._refObject = new PwmInputRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PwmInputWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PwmInputArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PwmInputArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PwmInputWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PwmInputWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PwmInputArrayWrapper;
  }
}
module.exports = PwmInputWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pulse_width",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "period",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "PwmInput",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PwmInput"
}
*/
