// rclnodejs/generated/px4_msgs/px4_msgs__msg__Px4ioStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt16Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt16.js');
const Px4ioStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  free_memory_bytes: primitiveTypes.uint16,
  voltage_v: primitiveTypes.float32,
  rssi_v: primitiveTypes.float32,
  status_arm_sync: primitiveTypes.bool,
  status_failsafe: primitiveTypes.bool,
  status_fmu_initialized: primitiveTypes.bool,
  status_fmu_ok: primitiveTypes.bool,
  status_init_ok: primitiveTypes.bool,
  status_outputs_armed: primitiveTypes.bool,
  status_raw_pwm: primitiveTypes.bool,
  status_rc_ok: primitiveTypes.bool,
  status_rc_dsm: primitiveTypes.bool,
  status_rc_ppm: primitiveTypes.bool,
  status_rc_sbus: primitiveTypes.bool,
  status_rc_st24: primitiveTypes.bool,
  status_rc_sumd: primitiveTypes.bool,
  status_safety_button_event: primitiveTypes.bool,
  alarm_pwm_error: primitiveTypes.bool,
  alarm_rc_lost: primitiveTypes.bool,
  arming_failsafe_custom: primitiveTypes.bool,
  arming_fmu_armed: primitiveTypes.bool,
  arming_fmu_prearmed: primitiveTypes.bool,
  arming_force_failsafe: primitiveTypes.bool,
  arming_io_arm_ok: primitiveTypes.bool,
  arming_lockdown: primitiveTypes.bool,
  arming_termination_failsafe: primitiveTypes.bool,
  pwm: ArrayType(primitiveTypes.uint16, 8),
  pwm_disarmed: ArrayType(primitiveTypes.uint16, 8),
  pwm_failsafe: ArrayType(primitiveTypes.uint16, 8),
  pwm_rate_hz: ArrayType(primitiveTypes.uint16, 8),
  raw_inputs: ArrayType(primitiveTypes.uint16, 18),
});
const Px4ioStatusRefArray = ArrayType(Px4ioStatusRefStruct);
const Px4ioStatusRefStructArray = StructType({
  data: Px4ioStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Px4ioStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Px4ioStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._free_memory_bytesIntialized = true;
      this._voltage_vIntialized = true;
      this._rssi_vIntialized = true;
      this._status_arm_syncIntialized = true;
      this._status_failsafeIntialized = true;
      this._status_fmu_initializedIntialized = true;
      this._status_fmu_okIntialized = true;
      this._status_init_okIntialized = true;
      this._status_outputs_armedIntialized = true;
      this._status_raw_pwmIntialized = true;
      this._status_rc_okIntialized = true;
      this._status_rc_dsmIntialized = true;
      this._status_rc_ppmIntialized = true;
      this._status_rc_sbusIntialized = true;
      this._status_rc_st24Intialized = true;
      this._status_rc_sumdIntialized = true;
      this._status_safety_button_eventIntialized = true;
      this._alarm_pwm_errorIntialized = true;
      this._alarm_rc_lostIntialized = true;
      this._arming_failsafe_customIntialized = true;
      this._arming_fmu_armedIntialized = true;
      this._arming_fmu_prearmedIntialized = true;
      this._arming_force_failsafeIntialized = true;
      this._arming_io_arm_okIntialized = true;
      this._arming_lockdownIntialized = true;
      this._arming_termination_failsafeIntialized = true;
      this._wrapperFields.pwm = UInt16Wrapper.createArray();
      this._wrapperFields.pwm.copy(other._wrapperFields.pwm);
      this._wrapperFields.pwm_disarmed = UInt16Wrapper.createArray();
      this._wrapperFields.pwm_disarmed.copy(other._wrapperFields.pwm_disarmed);
      this._wrapperFields.pwm_failsafe = UInt16Wrapper.createArray();
      this._wrapperFields.pwm_failsafe.copy(other._wrapperFields.pwm_failsafe);
      this._wrapperFields.pwm_rate_hz = UInt16Wrapper.createArray();
      this._wrapperFields.pwm_rate_hz.copy(other._wrapperFields.pwm_rate_hz);
      this._wrapperFields.raw_inputs = UInt16Wrapper.createArray();
      this._wrapperFields.raw_inputs.copy(other._wrapperFields.raw_inputs);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Px4ioStatusRefStruct();
    this._timestampIntialized = false;
    this._free_memory_bytesIntialized = false;
    this._voltage_vIntialized = false;
    this._rssi_vIntialized = false;
    this._status_arm_syncIntialized = false;
    this._status_failsafeIntialized = false;
    this._status_fmu_initializedIntialized = false;
    this._status_fmu_okIntialized = false;
    this._status_init_okIntialized = false;
    this._status_outputs_armedIntialized = false;
    this._status_raw_pwmIntialized = false;
    this._status_rc_okIntialized = false;
    this._status_rc_dsmIntialized = false;
    this._status_rc_ppmIntialized = false;
    this._status_rc_sbusIntialized = false;
    this._status_rc_st24Intialized = false;
    this._status_rc_sumdIntialized = false;
    this._status_safety_button_eventIntialized = false;
    this._alarm_pwm_errorIntialized = false;
    this._alarm_rc_lostIntialized = false;
    this._arming_failsafe_customIntialized = false;
    this._arming_fmu_armedIntialized = false;
    this._arming_fmu_prearmedIntialized = false;
    this._arming_force_failsafeIntialized = false;
    this._arming_io_arm_okIntialized = false;
    this._arming_lockdownIntialized = false;
    this._arming_termination_failsafeIntialized = false;
    this._wrapperFields.pwm = UInt16Wrapper.createArray();
    this._wrapperFields.pwm_disarmed = UInt16Wrapper.createArray();
    this._wrapperFields.pwm_failsafe = UInt16Wrapper.createArray();
    this._wrapperFields.pwm_rate_hz = UInt16Wrapper.createArray();
    this._wrapperFields.raw_inputs = UInt16Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new Px4ioStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Px4ioStatusArrayWrapper;
  }
  static get ArrayType() {
    return Px4ioStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return Px4ioStatusRefStructArray
  }
  static get refObjectType() {
    return Px4ioStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in Px4ioStatus');
      }
      if (checkConsistency && !this._free_memory_bytesIntialized) {
        throw new TypeError('Invalid argument: free_memory_bytes in Px4ioStatus');
      }
      if (checkConsistency && !this._voltage_vIntialized) {
        throw new TypeError('Invalid argument: voltage_v in Px4ioStatus');
      }
      if (checkConsistency && !this._rssi_vIntialized) {
        throw new TypeError('Invalid argument: rssi_v in Px4ioStatus');
      }
      if (checkConsistency && !this._status_arm_syncIntialized) {
        throw new TypeError('Invalid argument: status_arm_sync in Px4ioStatus');
      }
      if (checkConsistency && !this._status_failsafeIntialized) {
        throw new TypeError('Invalid argument: status_failsafe in Px4ioStatus');
      }
      if (checkConsistency && !this._status_fmu_initializedIntialized) {
        throw new TypeError('Invalid argument: status_fmu_initialized in Px4ioStatus');
      }
      if (checkConsistency && !this._status_fmu_okIntialized) {
        throw new TypeError('Invalid argument: status_fmu_ok in Px4ioStatus');
      }
      if (checkConsistency && !this._status_init_okIntialized) {
        throw new TypeError('Invalid argument: status_init_ok in Px4ioStatus');
      }
      if (checkConsistency && !this._status_outputs_armedIntialized) {
        throw new TypeError('Invalid argument: status_outputs_armed in Px4ioStatus');
      }
      if (checkConsistency && !this._status_raw_pwmIntialized) {
        throw new TypeError('Invalid argument: status_raw_pwm in Px4ioStatus');
      }
      if (checkConsistency && !this._status_rc_okIntialized) {
        throw new TypeError('Invalid argument: status_rc_ok in Px4ioStatus');
      }
      if (checkConsistency && !this._status_rc_dsmIntialized) {
        throw new TypeError('Invalid argument: status_rc_dsm in Px4ioStatus');
      }
      if (checkConsistency && !this._status_rc_ppmIntialized) {
        throw new TypeError('Invalid argument: status_rc_ppm in Px4ioStatus');
      }
      if (checkConsistency && !this._status_rc_sbusIntialized) {
        throw new TypeError('Invalid argument: status_rc_sbus in Px4ioStatus');
      }
      if (checkConsistency && !this._status_rc_st24Intialized) {
        throw new TypeError('Invalid argument: status_rc_st24 in Px4ioStatus');
      }
      if (checkConsistency && !this._status_rc_sumdIntialized) {
        throw new TypeError('Invalid argument: status_rc_sumd in Px4ioStatus');
      }
      if (checkConsistency && !this._status_safety_button_eventIntialized) {
        throw new TypeError('Invalid argument: status_safety_button_event in Px4ioStatus');
      }
      if (checkConsistency && !this._alarm_pwm_errorIntialized) {
        throw new TypeError('Invalid argument: alarm_pwm_error in Px4ioStatus');
      }
      if (checkConsistency && !this._alarm_rc_lostIntialized) {
        throw new TypeError('Invalid argument: alarm_rc_lost in Px4ioStatus');
      }
      if (checkConsistency && !this._arming_failsafe_customIntialized) {
        throw new TypeError('Invalid argument: arming_failsafe_custom in Px4ioStatus');
      }
      if (checkConsistency && !this._arming_fmu_armedIntialized) {
        throw new TypeError('Invalid argument: arming_fmu_armed in Px4ioStatus');
      }
      if (checkConsistency && !this._arming_fmu_prearmedIntialized) {
        throw new TypeError('Invalid argument: arming_fmu_prearmed in Px4ioStatus');
      }
      if (checkConsistency && !this._arming_force_failsafeIntialized) {
        throw new TypeError('Invalid argument: arming_force_failsafe in Px4ioStatus');
      }
      if (checkConsistency && !this._arming_io_arm_okIntialized) {
        throw new TypeError('Invalid argument: arming_io_arm_ok in Px4ioStatus');
      }
      if (checkConsistency && !this._arming_lockdownIntialized) {
        throw new TypeError('Invalid argument: arming_lockdown in Px4ioStatus');
      }
      if (checkConsistency && !this._arming_termination_failsafeIntialized) {
        throw new TypeError('Invalid argument: arming_termination_failsafe in Px4ioStatus');
      }
    this._refObject.pwm = Array.from(this._wrapperFields.pwm.data);
    this._refObject.pwm_disarmed = Array.from(this._wrapperFields.pwm_disarmed.data);
    this._refObject.pwm_failsafe = Array.from(this._wrapperFields.pwm_failsafe.data);
    this._refObject.pwm_rate_hz = Array.from(this._wrapperFields.pwm_rate_hz.data);
    this._refObject.raw_inputs = Array.from(this._wrapperFields.raw_inputs.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._free_memory_bytesIntialized = true;
    this._refObject.free_memory_bytes = refObject.free_memory_bytes;
    this._voltage_vIntialized = true;
    this._refObject.voltage_v = refObject.voltage_v;
    this._rssi_vIntialized = true;
    this._refObject.rssi_v = refObject.rssi_v;
    this._status_arm_syncIntialized = true;
    this._refObject.status_arm_sync = refObject.status_arm_sync;
    this._status_failsafeIntialized = true;
    this._refObject.status_failsafe = refObject.status_failsafe;
    this._status_fmu_initializedIntialized = true;
    this._refObject.status_fmu_initialized = refObject.status_fmu_initialized;
    this._status_fmu_okIntialized = true;
    this._refObject.status_fmu_ok = refObject.status_fmu_ok;
    this._status_init_okIntialized = true;
    this._refObject.status_init_ok = refObject.status_init_ok;
    this._status_outputs_armedIntialized = true;
    this._refObject.status_outputs_armed = refObject.status_outputs_armed;
    this._status_raw_pwmIntialized = true;
    this._refObject.status_raw_pwm = refObject.status_raw_pwm;
    this._status_rc_okIntialized = true;
    this._refObject.status_rc_ok = refObject.status_rc_ok;
    this._status_rc_dsmIntialized = true;
    this._refObject.status_rc_dsm = refObject.status_rc_dsm;
    this._status_rc_ppmIntialized = true;
    this._refObject.status_rc_ppm = refObject.status_rc_ppm;
    this._status_rc_sbusIntialized = true;
    this._refObject.status_rc_sbus = refObject.status_rc_sbus;
    this._status_rc_st24Intialized = true;
    this._refObject.status_rc_st24 = refObject.status_rc_st24;
    this._status_rc_sumdIntialized = true;
    this._refObject.status_rc_sumd = refObject.status_rc_sumd;
    this._status_safety_button_eventIntialized = true;
    this._refObject.status_safety_button_event = refObject.status_safety_button_event;
    this._alarm_pwm_errorIntialized = true;
    this._refObject.alarm_pwm_error = refObject.alarm_pwm_error;
    this._alarm_rc_lostIntialized = true;
    this._refObject.alarm_rc_lost = refObject.alarm_rc_lost;
    this._arming_failsafe_customIntialized = true;
    this._refObject.arming_failsafe_custom = refObject.arming_failsafe_custom;
    this._arming_fmu_armedIntialized = true;
    this._refObject.arming_fmu_armed = refObject.arming_fmu_armed;
    this._arming_fmu_prearmedIntialized = true;
    this._refObject.arming_fmu_prearmed = refObject.arming_fmu_prearmed;
    this._arming_force_failsafeIntialized = true;
    this._refObject.arming_force_failsafe = refObject.arming_force_failsafe;
    this._arming_io_arm_okIntialized = true;
    this._refObject.arming_io_arm_ok = refObject.arming_io_arm_ok;
    this._arming_lockdownIntialized = true;
    this._refObject.arming_lockdown = refObject.arming_lockdown;
    this._arming_termination_failsafeIntialized = true;
    this._refObject.arming_termination_failsafe = refObject.arming_termination_failsafe;
    this._wrapperFields.pwm.fill(refObject.pwm.toArray());
    this._wrapperFields.pwm_disarmed.fill(refObject.pwm_disarmed.toArray());
    this._wrapperFields.pwm_failsafe.fill(refObject.pwm_failsafe.toArray());
    this._wrapperFields.pwm_rate_hz.fill(refObject.pwm_rate_hz.toArray());
    this._wrapperFields.raw_inputs.fill(refObject.raw_inputs.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    Px4ioStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'Px4ioStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get free_memory_bytes() {
    if (!this._free_memory_bytesIntialized) {
      return undefined;
    }
    return this._refObject.free_memory_bytes;
  }
  set free_memory_bytes(value) {
    this._free_memory_bytesIntialized = true;
    this._refObject.free_memory_bytes = value;
  }
  get voltage_v() {
    if (!this._voltage_vIntialized) {
      return undefined;
    }
    return this._refObject.voltage_v;
  }
  set voltage_v(value) {
    this._voltage_vIntialized = true;
    this._refObject.voltage_v = value;
  }
  get rssi_v() {
    if (!this._rssi_vIntialized) {
      return undefined;
    }
    return this._refObject.rssi_v;
  }
  set rssi_v(value) {
    this._rssi_vIntialized = true;
    this._refObject.rssi_v = value;
  }
  get status_arm_sync() {
    if (!this._status_arm_syncIntialized) {
      return undefined;
    }
    return this._refObject.status_arm_sync;
  }
  set status_arm_sync(value) {
    this._status_arm_syncIntialized = true;
    this._refObject.status_arm_sync = value;
  }
  get status_failsafe() {
    if (!this._status_failsafeIntialized) {
      return undefined;
    }
    return this._refObject.status_failsafe;
  }
  set status_failsafe(value) {
    this._status_failsafeIntialized = true;
    this._refObject.status_failsafe = value;
  }
  get status_fmu_initialized() {
    if (!this._status_fmu_initializedIntialized) {
      return undefined;
    }
    return this._refObject.status_fmu_initialized;
  }
  set status_fmu_initialized(value) {
    this._status_fmu_initializedIntialized = true;
    this._refObject.status_fmu_initialized = value;
  }
  get status_fmu_ok() {
    if (!this._status_fmu_okIntialized) {
      return undefined;
    }
    return this._refObject.status_fmu_ok;
  }
  set status_fmu_ok(value) {
    this._status_fmu_okIntialized = true;
    this._refObject.status_fmu_ok = value;
  }
  get status_init_ok() {
    if (!this._status_init_okIntialized) {
      return undefined;
    }
    return this._refObject.status_init_ok;
  }
  set status_init_ok(value) {
    this._status_init_okIntialized = true;
    this._refObject.status_init_ok = value;
  }
  get status_outputs_armed() {
    if (!this._status_outputs_armedIntialized) {
      return undefined;
    }
    return this._refObject.status_outputs_armed;
  }
  set status_outputs_armed(value) {
    this._status_outputs_armedIntialized = true;
    this._refObject.status_outputs_armed = value;
  }
  get status_raw_pwm() {
    if (!this._status_raw_pwmIntialized) {
      return undefined;
    }
    return this._refObject.status_raw_pwm;
  }
  set status_raw_pwm(value) {
    this._status_raw_pwmIntialized = true;
    this._refObject.status_raw_pwm = value;
  }
  get status_rc_ok() {
    if (!this._status_rc_okIntialized) {
      return undefined;
    }
    return this._refObject.status_rc_ok;
  }
  set status_rc_ok(value) {
    this._status_rc_okIntialized = true;
    this._refObject.status_rc_ok = value;
  }
  get status_rc_dsm() {
    if (!this._status_rc_dsmIntialized) {
      return undefined;
    }
    return this._refObject.status_rc_dsm;
  }
  set status_rc_dsm(value) {
    this._status_rc_dsmIntialized = true;
    this._refObject.status_rc_dsm = value;
  }
  get status_rc_ppm() {
    if (!this._status_rc_ppmIntialized) {
      return undefined;
    }
    return this._refObject.status_rc_ppm;
  }
  set status_rc_ppm(value) {
    this._status_rc_ppmIntialized = true;
    this._refObject.status_rc_ppm = value;
  }
  get status_rc_sbus() {
    if (!this._status_rc_sbusIntialized) {
      return undefined;
    }
    return this._refObject.status_rc_sbus;
  }
  set status_rc_sbus(value) {
    this._status_rc_sbusIntialized = true;
    this._refObject.status_rc_sbus = value;
  }
  get status_rc_st24() {
    if (!this._status_rc_st24Intialized) {
      return undefined;
    }
    return this._refObject.status_rc_st24;
  }
  set status_rc_st24(value) {
    this._status_rc_st24Intialized = true;
    this._refObject.status_rc_st24 = value;
  }
  get status_rc_sumd() {
    if (!this._status_rc_sumdIntialized) {
      return undefined;
    }
    return this._refObject.status_rc_sumd;
  }
  set status_rc_sumd(value) {
    this._status_rc_sumdIntialized = true;
    this._refObject.status_rc_sumd = value;
  }
  get status_safety_button_event() {
    if (!this._status_safety_button_eventIntialized) {
      return undefined;
    }
    return this._refObject.status_safety_button_event;
  }
  set status_safety_button_event(value) {
    this._status_safety_button_eventIntialized = true;
    this._refObject.status_safety_button_event = value;
  }
  get alarm_pwm_error() {
    if (!this._alarm_pwm_errorIntialized) {
      return undefined;
    }
    return this._refObject.alarm_pwm_error;
  }
  set alarm_pwm_error(value) {
    this._alarm_pwm_errorIntialized = true;
    this._refObject.alarm_pwm_error = value;
  }
  get alarm_rc_lost() {
    if (!this._alarm_rc_lostIntialized) {
      return undefined;
    }
    return this._refObject.alarm_rc_lost;
  }
  set alarm_rc_lost(value) {
    this._alarm_rc_lostIntialized = true;
    this._refObject.alarm_rc_lost = value;
  }
  get arming_failsafe_custom() {
    if (!this._arming_failsafe_customIntialized) {
      return undefined;
    }
    return this._refObject.arming_failsafe_custom;
  }
  set arming_failsafe_custom(value) {
    this._arming_failsafe_customIntialized = true;
    this._refObject.arming_failsafe_custom = value;
  }
  get arming_fmu_armed() {
    if (!this._arming_fmu_armedIntialized) {
      return undefined;
    }
    return this._refObject.arming_fmu_armed;
  }
  set arming_fmu_armed(value) {
    this._arming_fmu_armedIntialized = true;
    this._refObject.arming_fmu_armed = value;
  }
  get arming_fmu_prearmed() {
    if (!this._arming_fmu_prearmedIntialized) {
      return undefined;
    }
    return this._refObject.arming_fmu_prearmed;
  }
  set arming_fmu_prearmed(value) {
    this._arming_fmu_prearmedIntialized = true;
    this._refObject.arming_fmu_prearmed = value;
  }
  get arming_force_failsafe() {
    if (!this._arming_force_failsafeIntialized) {
      return undefined;
    }
    return this._refObject.arming_force_failsafe;
  }
  set arming_force_failsafe(value) {
    this._arming_force_failsafeIntialized = true;
    this._refObject.arming_force_failsafe = value;
  }
  get arming_io_arm_ok() {
    if (!this._arming_io_arm_okIntialized) {
      return undefined;
    }
    return this._refObject.arming_io_arm_ok;
  }
  set arming_io_arm_ok(value) {
    this._arming_io_arm_okIntialized = true;
    this._refObject.arming_io_arm_ok = value;
  }
  get arming_lockdown() {
    if (!this._arming_lockdownIntialized) {
      return undefined;
    }
    return this._refObject.arming_lockdown;
  }
  set arming_lockdown(value) {
    this._arming_lockdownIntialized = true;
    this._refObject.arming_lockdown = value;
  }
  get arming_termination_failsafe() {
    if (!this._arming_termination_failsafeIntialized) {
      return undefined;
    }
    return this._refObject.arming_termination_failsafe;
  }
  set arming_termination_failsafe(value) {
    this._arming_termination_failsafeIntialized = true;
    this._refObject.arming_termination_failsafe = value;
  }
  get pwm() {
    return this._wrapperFields['pwm'].data;
  }
  set pwm(value) {
    if (value.length !== 8) {
      throw new RangeError('The length of the array must be 8.');
    }
    this._wrapperFields['pwm'].fill(value);
  }
  get pwm_disarmed() {
    return this._wrapperFields['pwm_disarmed'].data;
  }
  set pwm_disarmed(value) {
    if (value.length !== 8) {
      throw new RangeError('The length of the array must be 8.');
    }
    this._wrapperFields['pwm_disarmed'].fill(value);
  }
  get pwm_failsafe() {
    return this._wrapperFields['pwm_failsafe'].data;
  }
  set pwm_failsafe(value) {
    if (value.length !== 8) {
      throw new RangeError('The length of the array must be 8.');
    }
    this._wrapperFields['pwm_failsafe'].fill(value);
  }
  get pwm_rate_hz() {
    return this._wrapperFields['pwm_rate_hz'].data;
  }
  set pwm_rate_hz(value) {
    if (value.length !== 8) {
      throw new RangeError('The length of the array must be 8.');
    }
    this._wrapperFields['pwm_rate_hz'].fill(value);
  }
  get raw_inputs() {
    return this._wrapperFields['raw_inputs'].data;
  }
  set raw_inputs(value) {
    if (value.length !== 18) {
      throw new RangeError('The length of the array must be 18.');
    }
    this._wrapperFields['raw_inputs'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new Px4ioStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._free_memory_bytesIntialized = true;
    this._voltage_vIntialized = true;
    this._rssi_vIntialized = true;
    this._status_arm_syncIntialized = true;
    this._status_failsafeIntialized = true;
    this._status_fmu_initializedIntialized = true;
    this._status_fmu_okIntialized = true;
    this._status_init_okIntialized = true;
    this._status_outputs_armedIntialized = true;
    this._status_raw_pwmIntialized = true;
    this._status_rc_okIntialized = true;
    this._status_rc_dsmIntialized = true;
    this._status_rc_ppmIntialized = true;
    this._status_rc_sbusIntialized = true;
    this._status_rc_st24Intialized = true;
    this._status_rc_sumdIntialized = true;
    this._status_safety_button_eventIntialized = true;
    this._alarm_pwm_errorIntialized = true;
    this._alarm_rc_lostIntialized = true;
    this._arming_failsafe_customIntialized = true;
    this._arming_fmu_armedIntialized = true;
    this._arming_fmu_prearmedIntialized = true;
    this._arming_force_failsafeIntialized = true;
    this._arming_io_arm_okIntialized = true;
    this._arming_lockdownIntialized = true;
    this._arming_termination_failsafeIntialized = true;
    this._wrapperFields.pwm.fill(refObject.pwm.toArray());
    this._wrapperFields.pwm_disarmed.fill(refObject.pwm_disarmed.toArray());
    this._wrapperFields.pwm_failsafe.fill(refObject.pwm_failsafe.toArray());
    this._wrapperFields.pwm_rate_hz.fill(refObject.pwm_rate_hz.toArray());
    this._wrapperFields.raw_inputs.fill(refObject.raw_inputs.toArray());
  }
  copy(other) {
    this._refObject = new Px4ioStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._free_memory_bytesIntialized = true;
    this._voltage_vIntialized = true;
    this._rssi_vIntialized = true;
    this._status_arm_syncIntialized = true;
    this._status_failsafeIntialized = true;
    this._status_fmu_initializedIntialized = true;
    this._status_fmu_okIntialized = true;
    this._status_init_okIntialized = true;
    this._status_outputs_armedIntialized = true;
    this._status_raw_pwmIntialized = true;
    this._status_rc_okIntialized = true;
    this._status_rc_dsmIntialized = true;
    this._status_rc_ppmIntialized = true;
    this._status_rc_sbusIntialized = true;
    this._status_rc_st24Intialized = true;
    this._status_rc_sumdIntialized = true;
    this._status_safety_button_eventIntialized = true;
    this._alarm_pwm_errorIntialized = true;
    this._alarm_rc_lostIntialized = true;
    this._arming_failsafe_customIntialized = true;
    this._arming_fmu_armedIntialized = true;
    this._arming_fmu_prearmedIntialized = true;
    this._arming_force_failsafeIntialized = true;
    this._arming_io_arm_okIntialized = true;
    this._arming_lockdownIntialized = true;
    this._arming_termination_failsafeIntialized = true;
    this._wrapperFields.pwm.copy(other._wrapperFields.pwm);
    this._wrapperFields.pwm_disarmed.copy(other._wrapperFields.pwm_disarmed);
    this._wrapperFields.pwm_failsafe.copy(other._wrapperFields.pwm_failsafe);
    this._wrapperFields.pwm_rate_hz.copy(other._wrapperFields.pwm_rate_hz);
    this._wrapperFields.raw_inputs.copy(other._wrapperFields.raw_inputs);
  }
  static get classType() {
    return Px4ioStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"free_memory_bytes","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"voltage_v","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rssi_v","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_arm_sync","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_failsafe","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_fmu_initialized","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_fmu_ok","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_init_ok","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_outputs_armed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_raw_pwm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_rc_ok","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_rc_dsm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_rc_ppm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_rc_sbus","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_rc_st24","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_rc_sumd","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_safety_button_event","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"alarm_pwm_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"alarm_rc_lost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_failsafe_custom","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_fmu_armed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_fmu_prearmed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_force_failsafe","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_io_arm_ok","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_lockdown","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_termination_failsafe","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pwm","type":{"isArray":true,"arraySize":8,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pwm_disarmed","type":{"isArray":true,"arraySize":8,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pwm_failsafe","type":{"isArray":true,"arraySize":8,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pwm_rate_hz","type":{"isArray":true,"arraySize":8,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"raw_inputs","type":{"isArray":true,"arraySize":18,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"Px4ioStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Px4ioStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","free_memory_bytes","voltage_v","rssi_v","status_arm_sync","status_failsafe","status_fmu_initialized","status_fmu_ok","status_init_ok","status_outputs_armed","status_raw_pwm","status_rc_ok","status_rc_dsm","status_rc_ppm","status_rc_sbus","status_rc_st24","status_rc_sumd","status_safety_button_event","alarm_pwm_error","alarm_rc_lost","arming_failsafe_custom","arming_fmu_armed","arming_fmu_prearmed","arming_force_failsafe","arming_io_arm_ok","arming_lockdown","arming_termination_failsafe","pwm","pwm_disarmed","pwm_failsafe","pwm_rate_hz","raw_inputs"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Px4ioStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Px4ioStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Px4ioStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Px4ioStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Px4ioStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Px4ioStatusRefArray(size);
    this._refObject = new Px4ioStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Px4ioStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Px4ioStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Px4ioStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Px4ioStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Px4ioStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Px4ioStatusArrayWrapper;
  }
}
module.exports = Px4ioStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "free_memory_bytes",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "voltage_v",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rssi_v",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_arm_sync",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_failsafe",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_fmu_initialized",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_fmu_ok",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_init_ok",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_outputs_armed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_raw_pwm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_rc_ok",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_rc_dsm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_rc_ppm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_rc_sbus",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_rc_st24",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_rc_sumd",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_safety_button_event",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "alarm_pwm_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "alarm_rc_lost",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_failsafe_custom",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_fmu_armed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_fmu_prearmed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_force_failsafe",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_io_arm_ok",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_lockdown",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_termination_failsafe",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pwm",
      "type": {
        "isArray": true,
        "arraySize": 8,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pwm_disarmed",
      "type": {
        "isArray": true,
        "arraySize": 8,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pwm_failsafe",
      "type": {
        "isArray": true,
        "arraySize": 8,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pwm_rate_hz",
      "type": {
        "isArray": true,
        "arraySize": 8,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "raw_inputs",
      "type": {
        "isArray": true,
        "arraySize": 18,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "Px4ioStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Px4ioStatus"
}
*/
