// rclnodejs/generated/px4_msgs/px4_msgs__msg__IridiumsbdStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const IridiumsbdStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  last_heartbeat: primitiveTypes.uint64,
  tx_buf_write_index: primitiveTypes.uint16,
  rx_buf_read_index: primitiveTypes.uint16,
  rx_buf_end_index: primitiveTypes.uint16,
  failed_sbd_sessions: primitiveTypes.uint16,
  successful_sbd_sessions: primitiveTypes.uint16,
  num_tx_buf_reset: primitiveTypes.uint16,
  signal_quality: primitiveTypes.uint8,
  state: primitiveTypes.uint8,
  ring_pending: primitiveTypes.bool,
  tx_buf_write_pending: primitiveTypes.bool,
  tx_session_pending: primitiveTypes.bool,
  rx_read_pending: primitiveTypes.bool,
  rx_session_pending: primitiveTypes.bool,
});
const IridiumsbdStatusRefArray = ArrayType(IridiumsbdStatusRefStruct);
const IridiumsbdStatusRefStructArray = StructType({
  data: IridiumsbdStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class IridiumsbdStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new IridiumsbdStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._last_heartbeatIntialized = true;
      this._tx_buf_write_indexIntialized = true;
      this._rx_buf_read_indexIntialized = true;
      this._rx_buf_end_indexIntialized = true;
      this._failed_sbd_sessionsIntialized = true;
      this._successful_sbd_sessionsIntialized = true;
      this._num_tx_buf_resetIntialized = true;
      this._signal_qualityIntialized = true;
      this._stateIntialized = true;
      this._ring_pendingIntialized = true;
      this._tx_buf_write_pendingIntialized = true;
      this._tx_session_pendingIntialized = true;
      this._rx_read_pendingIntialized = true;
      this._rx_session_pendingIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new IridiumsbdStatusRefStruct();
    this._timestampIntialized = false;
    this._last_heartbeatIntialized = false;
    this._tx_buf_write_indexIntialized = false;
    this._rx_buf_read_indexIntialized = false;
    this._rx_buf_end_indexIntialized = false;
    this._failed_sbd_sessionsIntialized = false;
    this._successful_sbd_sessionsIntialized = false;
    this._num_tx_buf_resetIntialized = false;
    this._signal_qualityIntialized = false;
    this._stateIntialized = false;
    this._ring_pendingIntialized = false;
    this._tx_buf_write_pendingIntialized = false;
    this._tx_session_pendingIntialized = false;
    this._rx_read_pendingIntialized = false;
    this._rx_session_pendingIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new IridiumsbdStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IridiumsbdStatusArrayWrapper;
  }
  static get ArrayType() {
    return IridiumsbdStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return IridiumsbdStatusRefStructArray
  }
  static get refObjectType() {
    return IridiumsbdStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in IridiumsbdStatus');
      }
      if (checkConsistency && !this._last_heartbeatIntialized) {
        throw new TypeError('Invalid argument: last_heartbeat in IridiumsbdStatus');
      }
      if (checkConsistency && !this._tx_buf_write_indexIntialized) {
        throw new TypeError('Invalid argument: tx_buf_write_index in IridiumsbdStatus');
      }
      if (checkConsistency && !this._rx_buf_read_indexIntialized) {
        throw new TypeError('Invalid argument: rx_buf_read_index in IridiumsbdStatus');
      }
      if (checkConsistency && !this._rx_buf_end_indexIntialized) {
        throw new TypeError('Invalid argument: rx_buf_end_index in IridiumsbdStatus');
      }
      if (checkConsistency && !this._failed_sbd_sessionsIntialized) {
        throw new TypeError('Invalid argument: failed_sbd_sessions in IridiumsbdStatus');
      }
      if (checkConsistency && !this._successful_sbd_sessionsIntialized) {
        throw new TypeError('Invalid argument: successful_sbd_sessions in IridiumsbdStatus');
      }
      if (checkConsistency && !this._num_tx_buf_resetIntialized) {
        throw new TypeError('Invalid argument: num_tx_buf_reset in IridiumsbdStatus');
      }
      if (checkConsistency && !this._signal_qualityIntialized) {
        throw new TypeError('Invalid argument: signal_quality in IridiumsbdStatus');
      }
      if (checkConsistency && !this._stateIntialized) {
        throw new TypeError('Invalid argument: state in IridiumsbdStatus');
      }
      if (checkConsistency && !this._ring_pendingIntialized) {
        throw new TypeError('Invalid argument: ring_pending in IridiumsbdStatus');
      }
      if (checkConsistency && !this._tx_buf_write_pendingIntialized) {
        throw new TypeError('Invalid argument: tx_buf_write_pending in IridiumsbdStatus');
      }
      if (checkConsistency && !this._tx_session_pendingIntialized) {
        throw new TypeError('Invalid argument: tx_session_pending in IridiumsbdStatus');
      }
      if (checkConsistency && !this._rx_read_pendingIntialized) {
        throw new TypeError('Invalid argument: rx_read_pending in IridiumsbdStatus');
      }
      if (checkConsistency && !this._rx_session_pendingIntialized) {
        throw new TypeError('Invalid argument: rx_session_pending in IridiumsbdStatus');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._last_heartbeatIntialized = true;
    this._refObject.last_heartbeat = refObject.last_heartbeat;
    this._tx_buf_write_indexIntialized = true;
    this._refObject.tx_buf_write_index = refObject.tx_buf_write_index;
    this._rx_buf_read_indexIntialized = true;
    this._refObject.rx_buf_read_index = refObject.rx_buf_read_index;
    this._rx_buf_end_indexIntialized = true;
    this._refObject.rx_buf_end_index = refObject.rx_buf_end_index;
    this._failed_sbd_sessionsIntialized = true;
    this._refObject.failed_sbd_sessions = refObject.failed_sbd_sessions;
    this._successful_sbd_sessionsIntialized = true;
    this._refObject.successful_sbd_sessions = refObject.successful_sbd_sessions;
    this._num_tx_buf_resetIntialized = true;
    this._refObject.num_tx_buf_reset = refObject.num_tx_buf_reset;
    this._signal_qualityIntialized = true;
    this._refObject.signal_quality = refObject.signal_quality;
    this._stateIntialized = true;
    this._refObject.state = refObject.state;
    this._ring_pendingIntialized = true;
    this._refObject.ring_pending = refObject.ring_pending;
    this._tx_buf_write_pendingIntialized = true;
    this._refObject.tx_buf_write_pending = refObject.tx_buf_write_pending;
    this._tx_session_pendingIntialized = true;
    this._refObject.tx_session_pending = refObject.tx_session_pending;
    this._rx_read_pendingIntialized = true;
    this._refObject.rx_read_pending = refObject.rx_read_pending;
    this._rx_session_pendingIntialized = true;
    this._refObject.rx_session_pending = refObject.rx_session_pending;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    IridiumsbdStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'IridiumsbdStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get last_heartbeat() {
    if (!this._last_heartbeatIntialized) {
      return undefined;
    }
    return this._refObject.last_heartbeat;
  }
  set last_heartbeat(value) {
    this._last_heartbeatIntialized = true;
    this._refObject.last_heartbeat = value;
  }
  get tx_buf_write_index() {
    if (!this._tx_buf_write_indexIntialized) {
      return undefined;
    }
    return this._refObject.tx_buf_write_index;
  }
  set tx_buf_write_index(value) {
    this._tx_buf_write_indexIntialized = true;
    this._refObject.tx_buf_write_index = value;
  }
  get rx_buf_read_index() {
    if (!this._rx_buf_read_indexIntialized) {
      return undefined;
    }
    return this._refObject.rx_buf_read_index;
  }
  set rx_buf_read_index(value) {
    this._rx_buf_read_indexIntialized = true;
    this._refObject.rx_buf_read_index = value;
  }
  get rx_buf_end_index() {
    if (!this._rx_buf_end_indexIntialized) {
      return undefined;
    }
    return this._refObject.rx_buf_end_index;
  }
  set rx_buf_end_index(value) {
    this._rx_buf_end_indexIntialized = true;
    this._refObject.rx_buf_end_index = value;
  }
  get failed_sbd_sessions() {
    if (!this._failed_sbd_sessionsIntialized) {
      return undefined;
    }
    return this._refObject.failed_sbd_sessions;
  }
  set failed_sbd_sessions(value) {
    this._failed_sbd_sessionsIntialized = true;
    this._refObject.failed_sbd_sessions = value;
  }
  get successful_sbd_sessions() {
    if (!this._successful_sbd_sessionsIntialized) {
      return undefined;
    }
    return this._refObject.successful_sbd_sessions;
  }
  set successful_sbd_sessions(value) {
    this._successful_sbd_sessionsIntialized = true;
    this._refObject.successful_sbd_sessions = value;
  }
  get num_tx_buf_reset() {
    if (!this._num_tx_buf_resetIntialized) {
      return undefined;
    }
    return this._refObject.num_tx_buf_reset;
  }
  set num_tx_buf_reset(value) {
    this._num_tx_buf_resetIntialized = true;
    this._refObject.num_tx_buf_reset = value;
  }
  get signal_quality() {
    if (!this._signal_qualityIntialized) {
      return undefined;
    }
    return this._refObject.signal_quality;
  }
  set signal_quality(value) {
    this._signal_qualityIntialized = true;
    this._refObject.signal_quality = value;
  }
  get state() {
    if (!this._stateIntialized) {
      return undefined;
    }
    return this._refObject.state;
  }
  set state(value) {
    this._stateIntialized = true;
    this._refObject.state = value;
  }
  get ring_pending() {
    if (!this._ring_pendingIntialized) {
      return undefined;
    }
    return this._refObject.ring_pending;
  }
  set ring_pending(value) {
    this._ring_pendingIntialized = true;
    this._refObject.ring_pending = value;
  }
  get tx_buf_write_pending() {
    if (!this._tx_buf_write_pendingIntialized) {
      return undefined;
    }
    return this._refObject.tx_buf_write_pending;
  }
  set tx_buf_write_pending(value) {
    this._tx_buf_write_pendingIntialized = true;
    this._refObject.tx_buf_write_pending = value;
  }
  get tx_session_pending() {
    if (!this._tx_session_pendingIntialized) {
      return undefined;
    }
    return this._refObject.tx_session_pending;
  }
  set tx_session_pending(value) {
    this._tx_session_pendingIntialized = true;
    this._refObject.tx_session_pending = value;
  }
  get rx_read_pending() {
    if (!this._rx_read_pendingIntialized) {
      return undefined;
    }
    return this._refObject.rx_read_pending;
  }
  set rx_read_pending(value) {
    this._rx_read_pendingIntialized = true;
    this._refObject.rx_read_pending = value;
  }
  get rx_session_pending() {
    if (!this._rx_session_pendingIntialized) {
      return undefined;
    }
    return this._refObject.rx_session_pending;
  }
  set rx_session_pending(value) {
    this._rx_session_pendingIntialized = true;
    this._refObject.rx_session_pending = value;
  }
  copyRefObject(refObject) {
    this._refObject = new IridiumsbdStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._last_heartbeatIntialized = true;
    this._tx_buf_write_indexIntialized = true;
    this._rx_buf_read_indexIntialized = true;
    this._rx_buf_end_indexIntialized = true;
    this._failed_sbd_sessionsIntialized = true;
    this._successful_sbd_sessionsIntialized = true;
    this._num_tx_buf_resetIntialized = true;
    this._signal_qualityIntialized = true;
    this._stateIntialized = true;
    this._ring_pendingIntialized = true;
    this._tx_buf_write_pendingIntialized = true;
    this._tx_session_pendingIntialized = true;
    this._rx_read_pendingIntialized = true;
    this._rx_session_pendingIntialized = true;
  }
  copy(other) {
    this._refObject = new IridiumsbdStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._last_heartbeatIntialized = true;
    this._tx_buf_write_indexIntialized = true;
    this._rx_buf_read_indexIntialized = true;
    this._rx_buf_end_indexIntialized = true;
    this._failed_sbd_sessionsIntialized = true;
    this._successful_sbd_sessionsIntialized = true;
    this._num_tx_buf_resetIntialized = true;
    this._signal_qualityIntialized = true;
    this._stateIntialized = true;
    this._ring_pendingIntialized = true;
    this._tx_buf_write_pendingIntialized = true;
    this._tx_session_pendingIntialized = true;
    this._rx_read_pendingIntialized = true;
    this._rx_session_pendingIntialized = true;
  }
  static get classType() {
    return IridiumsbdStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"last_heartbeat","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tx_buf_write_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rx_buf_read_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rx_buf_end_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"failed_sbd_sessions","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"successful_sbd_sessions","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"num_tx_buf_reset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"signal_quality","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ring_pending","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tx_buf_write_pending","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tx_session_pending","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rx_read_pending","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rx_session_pending","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"IridiumsbdStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IridiumsbdStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","last_heartbeat","tx_buf_write_index","rx_buf_read_index","rx_buf_end_index","failed_sbd_sessions","successful_sbd_sessions","num_tx_buf_reset","signal_quality","state","ring_pending","tx_buf_write_pending","tx_session_pending","rx_read_pending","rx_session_pending"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class IridiumsbdStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IridiumsbdStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IridiumsbdStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IridiumsbdStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IridiumsbdStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new IridiumsbdStatusRefArray(size);
    this._refObject = new IridiumsbdStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IridiumsbdStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IridiumsbdStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IridiumsbdStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IridiumsbdStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IridiumsbdStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IridiumsbdStatusArrayWrapper;
  }
}
module.exports = IridiumsbdStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "last_heartbeat",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tx_buf_write_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rx_buf_read_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rx_buf_end_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "failed_sbd_sessions",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "successful_sbd_sessions",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "num_tx_buf_reset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "signal_quality",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ring_pending",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tx_buf_write_pending",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tx_session_pending",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rx_read_pending",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rx_session_pending",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "IridiumsbdStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "IridiumsbdStatus"
}
*/
