// rclnodejs/generated/px4_msgs/px4_msgs__msg__Event.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const EventRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  id: primitiveTypes.uint32,
  event_sequence: primitiveTypes.uint16,
  arguments: ArrayType(primitiveTypes.uint8, 25),
  log_levels: primitiveTypes.uint8,
});
const EventRefArray = ArrayType(EventRefStruct);
const EventRefStructArray = StructType({
  data: EventRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EventWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EventRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._idIntialized = true;
      this._event_sequenceIntialized = true;
      this._wrapperFields.arguments = UInt8Wrapper.createArray();
      this._wrapperFields.arguments.copy(other._wrapperFields.arguments);
      this._log_levelsIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EventRefStruct();
    this._timestampIntialized = false;
    this._idIntialized = false;
    this._event_sequenceIntialized = false;
    this._wrapperFields.arguments = UInt8Wrapper.createArray();
    this._log_levelsIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EventWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EventArrayWrapper;
  }
  static get ArrayType() {
    return EventArrayWrapper;
  }
  static get refObjectArrayType() {
    return EventRefStructArray
  }
  static get refObjectType() {
    return EventRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in Event');
      }
      if (checkConsistency && !this._idIntialized) {
        throw new TypeError('Invalid argument: id in Event');
      }
      if (checkConsistency && !this._event_sequenceIntialized) {
        throw new TypeError('Invalid argument: event_sequence in Event');
      }
      if (checkConsistency && !this._log_levelsIntialized) {
        throw new TypeError('Invalid argument: log_levels in Event');
      }
    this._refObject.arguments = Array.from(this._wrapperFields.arguments.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._event_sequenceIntialized = true;
    this._refObject.event_sequence = refObject.event_sequence;
    this._wrapperFields.arguments.fill(refObject.arguments.toArray());
    this._log_levelsIntialized = true;
    this._refObject.log_levels = refObject.log_levels;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    EventWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'Event'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get event_sequence() {
    if (!this._event_sequenceIntialized) {
      return undefined;
    }
    return this._refObject.event_sequence;
  }
  set event_sequence(value) {
    this._event_sequenceIntialized = true;
    this._refObject.event_sequence = value;
  }
  get arguments() {
    return this._wrapperFields['arguments'].data;
  }
  set arguments(value) {
    if (value.length !== 25) {
      throw new RangeError('The length of the array must be 25.');
    }
    this._wrapperFields['arguments'].fill(value);
  }
  get log_levels() {
    if (!this._log_levelsIntialized) {
      return undefined;
    }
    return this._refObject.log_levels;
  }
  set log_levels(value) {
    this._log_levelsIntialized = true;
    this._refObject.log_levels = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EventRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._idIntialized = true;
    this._event_sequenceIntialized = true;
    this._wrapperFields.arguments.fill(refObject.arguments.toArray());
    this._log_levelsIntialized = true;
  }
  copy(other) {
    this._refObject = new EventRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._idIntialized = true;
    this._event_sequenceIntialized = true;
    this._wrapperFields.arguments.copy(other._wrapperFields.arguments);
    this._log_levelsIntialized = true;
  }
  static get classType() {
    return EventWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ORB_QUEUE_LENGTH","value":16}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"event_sequence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arguments","type":{"isArray":true,"arraySize":25,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"log_levels","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"Event","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Event"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","id","event_sequence","arguments","log_levels"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EventArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EventWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EventWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EventArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EventArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EventRefArray(size);
    this._refObject = new EventRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EventWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EventArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EventArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EventWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EventWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EventArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(EventWrapper, "ORB_QUEUE_LENGTH", {value: 16, writable: false, enumerable: true, configurable: true});
module.exports = EventWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ORB_QUEUE_LENGTH",
      "value": 16
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "event_sequence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arguments",
      "type": {
        "isArray": true,
        "arraySize": 25,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "log_levels",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "Event",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Event"
}
*/
