// rclnodejs/generated/px4_msgs/px4_msgs__msg__LandingGear.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LandingGearRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  landing_gear: primitiveTypes.int8,
});
const LandingGearRefArray = ArrayType(LandingGearRefStruct);
const LandingGearRefStructArray = StructType({
  data: LandingGearRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LandingGearWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LandingGearRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._landing_gearIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LandingGearRefStruct();
    this._timestampIntialized = false;
    this._landing_gearIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LandingGearWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LandingGearArrayWrapper;
  }
  static get ArrayType() {
    return LandingGearArrayWrapper;
  }
  static get refObjectArrayType() {
    return LandingGearRefStructArray
  }
  static get refObjectType() {
    return LandingGearRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in LandingGear');
      }
      if (checkConsistency && !this._landing_gearIntialized) {
        throw new TypeError('Invalid argument: landing_gear in LandingGear');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._landing_gearIntialized = true;
    this._refObject.landing_gear = refObject.landing_gear;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LandingGearWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'LandingGear'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get landing_gear() {
    if (!this._landing_gearIntialized) {
      return undefined;
    }
    return this._refObject.landing_gear;
  }
  set landing_gear(value) {
    this._landing_gearIntialized = true;
    this._refObject.landing_gear = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LandingGearRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._landing_gearIntialized = true;
  }
  copy(other) {
    this._refObject = new LandingGearRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._landing_gearIntialized = true;
  }
  static get classType() {
    return LandingGearWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"GEAR_UP","value":1},{"type":"int8","name":"GEAR_DOWN","value":-1},{"type":"int8","name":"GEAR_KEEP","value":0}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"landing_gear","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"LandingGear","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LandingGear"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","landing_gear"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LandingGearArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LandingGearWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LandingGearWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandingGearArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandingGearArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LandingGearRefArray(size);
    this._refObject = new LandingGearRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LandingGearWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LandingGearArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LandingGearArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LandingGearWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LandingGearWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LandingGearArrayWrapper;
  }
}
// Define constants (3 in total)
Object.defineProperty(LandingGearWrapper, "GEAR_UP", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingGearWrapper, "GEAR_DOWN", {value: -1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LandingGearWrapper, "GEAR_KEEP", {value: 0, writable: false, enumerable: true, configurable: true});
module.exports = LandingGearWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "int8",
      "name": "GEAR_UP",
      "value": 1
    },
    {
      "type": "int8",
      "name": "GEAR_DOWN",
      "value": -1
    },
    {
      "type": "int8",
      "name": "GEAR_KEEP",
      "value": 0
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "landing_gear",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "LandingGear",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LandingGear"
}
*/
