// rclnodejs/generated/px4_msgs/px4_msgs__msg__EstimatorEventFlags.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EstimatorEventFlagsRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  information_event_changes: primitiveTypes.uint32,
  gps_checks_passed: primitiveTypes.bool,
  reset_vel_to_gps: primitiveTypes.bool,
  reset_vel_to_flow: primitiveTypes.bool,
  reset_vel_to_vision: primitiveTypes.bool,
  reset_vel_to_zero: primitiveTypes.bool,
  reset_pos_to_last_known: primitiveTypes.bool,
  reset_pos_to_gps: primitiveTypes.bool,
  reset_pos_to_vision: primitiveTypes.bool,
  starting_gps_fusion: primitiveTypes.bool,
  starting_vision_pos_fusion: primitiveTypes.bool,
  starting_vision_vel_fusion: primitiveTypes.bool,
  starting_vision_yaw_fusion: primitiveTypes.bool,
  yaw_aligned_to_imu_gps: primitiveTypes.bool,
  reset_hgt_to_baro: primitiveTypes.bool,
  reset_hgt_to_gps: primitiveTypes.bool,
  reset_hgt_to_rng: primitiveTypes.bool,
  reset_hgt_to_ev: primitiveTypes.bool,
  warning_event_changes: primitiveTypes.uint32,
  gps_quality_poor: primitiveTypes.bool,
  gps_fusion_timout: primitiveTypes.bool,
  gps_data_stopped: primitiveTypes.bool,
  gps_data_stopped_using_alternate: primitiveTypes.bool,
  height_sensor_timeout: primitiveTypes.bool,
  stopping_navigation: primitiveTypes.bool,
  invalid_accel_bias_cov_reset: primitiveTypes.bool,
  bad_yaw_using_gps_course: primitiveTypes.bool,
  stopping_mag_use: primitiveTypes.bool,
  vision_data_stopped: primitiveTypes.bool,
  emergency_yaw_reset_mag_stopped: primitiveTypes.bool,
  emergency_yaw_reset_gps_yaw_stopped: primitiveTypes.bool,
});
const EstimatorEventFlagsRefArray = ArrayType(EstimatorEventFlagsRefStruct);
const EstimatorEventFlagsRefStructArray = StructType({
  data: EstimatorEventFlagsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EstimatorEventFlagsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EstimatorEventFlagsRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._information_event_changesIntialized = true;
      this._gps_checks_passedIntialized = true;
      this._reset_vel_to_gpsIntialized = true;
      this._reset_vel_to_flowIntialized = true;
      this._reset_vel_to_visionIntialized = true;
      this._reset_vel_to_zeroIntialized = true;
      this._reset_pos_to_last_knownIntialized = true;
      this._reset_pos_to_gpsIntialized = true;
      this._reset_pos_to_visionIntialized = true;
      this._starting_gps_fusionIntialized = true;
      this._starting_vision_pos_fusionIntialized = true;
      this._starting_vision_vel_fusionIntialized = true;
      this._starting_vision_yaw_fusionIntialized = true;
      this._yaw_aligned_to_imu_gpsIntialized = true;
      this._reset_hgt_to_baroIntialized = true;
      this._reset_hgt_to_gpsIntialized = true;
      this._reset_hgt_to_rngIntialized = true;
      this._reset_hgt_to_evIntialized = true;
      this._warning_event_changesIntialized = true;
      this._gps_quality_poorIntialized = true;
      this._gps_fusion_timoutIntialized = true;
      this._gps_data_stoppedIntialized = true;
      this._gps_data_stopped_using_alternateIntialized = true;
      this._height_sensor_timeoutIntialized = true;
      this._stopping_navigationIntialized = true;
      this._invalid_accel_bias_cov_resetIntialized = true;
      this._bad_yaw_using_gps_courseIntialized = true;
      this._stopping_mag_useIntialized = true;
      this._vision_data_stoppedIntialized = true;
      this._emergency_yaw_reset_mag_stoppedIntialized = true;
      this._emergency_yaw_reset_gps_yaw_stoppedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EstimatorEventFlagsRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._information_event_changesIntialized = false;
    this._gps_checks_passedIntialized = false;
    this._reset_vel_to_gpsIntialized = false;
    this._reset_vel_to_flowIntialized = false;
    this._reset_vel_to_visionIntialized = false;
    this._reset_vel_to_zeroIntialized = false;
    this._reset_pos_to_last_knownIntialized = false;
    this._reset_pos_to_gpsIntialized = false;
    this._reset_pos_to_visionIntialized = false;
    this._starting_gps_fusionIntialized = false;
    this._starting_vision_pos_fusionIntialized = false;
    this._starting_vision_vel_fusionIntialized = false;
    this._starting_vision_yaw_fusionIntialized = false;
    this._yaw_aligned_to_imu_gpsIntialized = false;
    this._reset_hgt_to_baroIntialized = false;
    this._reset_hgt_to_gpsIntialized = false;
    this._reset_hgt_to_rngIntialized = false;
    this._reset_hgt_to_evIntialized = false;
    this._warning_event_changesIntialized = false;
    this._gps_quality_poorIntialized = false;
    this._gps_fusion_timoutIntialized = false;
    this._gps_data_stoppedIntialized = false;
    this._gps_data_stopped_using_alternateIntialized = false;
    this._height_sensor_timeoutIntialized = false;
    this._stopping_navigationIntialized = false;
    this._invalid_accel_bias_cov_resetIntialized = false;
    this._bad_yaw_using_gps_courseIntialized = false;
    this._stopping_mag_useIntialized = false;
    this._vision_data_stoppedIntialized = false;
    this._emergency_yaw_reset_mag_stoppedIntialized = false;
    this._emergency_yaw_reset_gps_yaw_stoppedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EstimatorEventFlagsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EstimatorEventFlagsArrayWrapper;
  }
  static get ArrayType() {
    return EstimatorEventFlagsArrayWrapper;
  }
  static get refObjectArrayType() {
    return EstimatorEventFlagsRefStructArray
  }
  static get refObjectType() {
    return EstimatorEventFlagsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in EstimatorEventFlags');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in EstimatorEventFlags');
      }
      if (checkConsistency && !this._information_event_changesIntialized) {
        throw new TypeError('Invalid argument: information_event_changes in EstimatorEventFlags');
      }
      if (checkConsistency && !this._gps_checks_passedIntialized) {
        throw new TypeError('Invalid argument: gps_checks_passed in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_vel_to_gpsIntialized) {
        throw new TypeError('Invalid argument: reset_vel_to_gps in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_vel_to_flowIntialized) {
        throw new TypeError('Invalid argument: reset_vel_to_flow in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_vel_to_visionIntialized) {
        throw new TypeError('Invalid argument: reset_vel_to_vision in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_vel_to_zeroIntialized) {
        throw new TypeError('Invalid argument: reset_vel_to_zero in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_pos_to_last_knownIntialized) {
        throw new TypeError('Invalid argument: reset_pos_to_last_known in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_pos_to_gpsIntialized) {
        throw new TypeError('Invalid argument: reset_pos_to_gps in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_pos_to_visionIntialized) {
        throw new TypeError('Invalid argument: reset_pos_to_vision in EstimatorEventFlags');
      }
      if (checkConsistency && !this._starting_gps_fusionIntialized) {
        throw new TypeError('Invalid argument: starting_gps_fusion in EstimatorEventFlags');
      }
      if (checkConsistency && !this._starting_vision_pos_fusionIntialized) {
        throw new TypeError('Invalid argument: starting_vision_pos_fusion in EstimatorEventFlags');
      }
      if (checkConsistency && !this._starting_vision_vel_fusionIntialized) {
        throw new TypeError('Invalid argument: starting_vision_vel_fusion in EstimatorEventFlags');
      }
      if (checkConsistency && !this._starting_vision_yaw_fusionIntialized) {
        throw new TypeError('Invalid argument: starting_vision_yaw_fusion in EstimatorEventFlags');
      }
      if (checkConsistency && !this._yaw_aligned_to_imu_gpsIntialized) {
        throw new TypeError('Invalid argument: yaw_aligned_to_imu_gps in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_hgt_to_baroIntialized) {
        throw new TypeError('Invalid argument: reset_hgt_to_baro in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_hgt_to_gpsIntialized) {
        throw new TypeError('Invalid argument: reset_hgt_to_gps in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_hgt_to_rngIntialized) {
        throw new TypeError('Invalid argument: reset_hgt_to_rng in EstimatorEventFlags');
      }
      if (checkConsistency && !this._reset_hgt_to_evIntialized) {
        throw new TypeError('Invalid argument: reset_hgt_to_ev in EstimatorEventFlags');
      }
      if (checkConsistency && !this._warning_event_changesIntialized) {
        throw new TypeError('Invalid argument: warning_event_changes in EstimatorEventFlags');
      }
      if (checkConsistency && !this._gps_quality_poorIntialized) {
        throw new TypeError('Invalid argument: gps_quality_poor in EstimatorEventFlags');
      }
      if (checkConsistency && !this._gps_fusion_timoutIntialized) {
        throw new TypeError('Invalid argument: gps_fusion_timout in EstimatorEventFlags');
      }
      if (checkConsistency && !this._gps_data_stoppedIntialized) {
        throw new TypeError('Invalid argument: gps_data_stopped in EstimatorEventFlags');
      }
      if (checkConsistency && !this._gps_data_stopped_using_alternateIntialized) {
        throw new TypeError('Invalid argument: gps_data_stopped_using_alternate in EstimatorEventFlags');
      }
      if (checkConsistency && !this._height_sensor_timeoutIntialized) {
        throw new TypeError('Invalid argument: height_sensor_timeout in EstimatorEventFlags');
      }
      if (checkConsistency && !this._stopping_navigationIntialized) {
        throw new TypeError('Invalid argument: stopping_navigation in EstimatorEventFlags');
      }
      if (checkConsistency && !this._invalid_accel_bias_cov_resetIntialized) {
        throw new TypeError('Invalid argument: invalid_accel_bias_cov_reset in EstimatorEventFlags');
      }
      if (checkConsistency && !this._bad_yaw_using_gps_courseIntialized) {
        throw new TypeError('Invalid argument: bad_yaw_using_gps_course in EstimatorEventFlags');
      }
      if (checkConsistency && !this._stopping_mag_useIntialized) {
        throw new TypeError('Invalid argument: stopping_mag_use in EstimatorEventFlags');
      }
      if (checkConsistency && !this._vision_data_stoppedIntialized) {
        throw new TypeError('Invalid argument: vision_data_stopped in EstimatorEventFlags');
      }
      if (checkConsistency && !this._emergency_yaw_reset_mag_stoppedIntialized) {
        throw new TypeError('Invalid argument: emergency_yaw_reset_mag_stopped in EstimatorEventFlags');
      }
      if (checkConsistency && !this._emergency_yaw_reset_gps_yaw_stoppedIntialized) {
        throw new TypeError('Invalid argument: emergency_yaw_reset_gps_yaw_stopped in EstimatorEventFlags');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._information_event_changesIntialized = true;
    this._refObject.information_event_changes = refObject.information_event_changes;
    this._gps_checks_passedIntialized = true;
    this._refObject.gps_checks_passed = refObject.gps_checks_passed;
    this._reset_vel_to_gpsIntialized = true;
    this._refObject.reset_vel_to_gps = refObject.reset_vel_to_gps;
    this._reset_vel_to_flowIntialized = true;
    this._refObject.reset_vel_to_flow = refObject.reset_vel_to_flow;
    this._reset_vel_to_visionIntialized = true;
    this._refObject.reset_vel_to_vision = refObject.reset_vel_to_vision;
    this._reset_vel_to_zeroIntialized = true;
    this._refObject.reset_vel_to_zero = refObject.reset_vel_to_zero;
    this._reset_pos_to_last_knownIntialized = true;
    this._refObject.reset_pos_to_last_known = refObject.reset_pos_to_last_known;
    this._reset_pos_to_gpsIntialized = true;
    this._refObject.reset_pos_to_gps = refObject.reset_pos_to_gps;
    this._reset_pos_to_visionIntialized = true;
    this._refObject.reset_pos_to_vision = refObject.reset_pos_to_vision;
    this._starting_gps_fusionIntialized = true;
    this._refObject.starting_gps_fusion = refObject.starting_gps_fusion;
    this._starting_vision_pos_fusionIntialized = true;
    this._refObject.starting_vision_pos_fusion = refObject.starting_vision_pos_fusion;
    this._starting_vision_vel_fusionIntialized = true;
    this._refObject.starting_vision_vel_fusion = refObject.starting_vision_vel_fusion;
    this._starting_vision_yaw_fusionIntialized = true;
    this._refObject.starting_vision_yaw_fusion = refObject.starting_vision_yaw_fusion;
    this._yaw_aligned_to_imu_gpsIntialized = true;
    this._refObject.yaw_aligned_to_imu_gps = refObject.yaw_aligned_to_imu_gps;
    this._reset_hgt_to_baroIntialized = true;
    this._refObject.reset_hgt_to_baro = refObject.reset_hgt_to_baro;
    this._reset_hgt_to_gpsIntialized = true;
    this._refObject.reset_hgt_to_gps = refObject.reset_hgt_to_gps;
    this._reset_hgt_to_rngIntialized = true;
    this._refObject.reset_hgt_to_rng = refObject.reset_hgt_to_rng;
    this._reset_hgt_to_evIntialized = true;
    this._refObject.reset_hgt_to_ev = refObject.reset_hgt_to_ev;
    this._warning_event_changesIntialized = true;
    this._refObject.warning_event_changes = refObject.warning_event_changes;
    this._gps_quality_poorIntialized = true;
    this._refObject.gps_quality_poor = refObject.gps_quality_poor;
    this._gps_fusion_timoutIntialized = true;
    this._refObject.gps_fusion_timout = refObject.gps_fusion_timout;
    this._gps_data_stoppedIntialized = true;
    this._refObject.gps_data_stopped = refObject.gps_data_stopped;
    this._gps_data_stopped_using_alternateIntialized = true;
    this._refObject.gps_data_stopped_using_alternate = refObject.gps_data_stopped_using_alternate;
    this._height_sensor_timeoutIntialized = true;
    this._refObject.height_sensor_timeout = refObject.height_sensor_timeout;
    this._stopping_navigationIntialized = true;
    this._refObject.stopping_navigation = refObject.stopping_navigation;
    this._invalid_accel_bias_cov_resetIntialized = true;
    this._refObject.invalid_accel_bias_cov_reset = refObject.invalid_accel_bias_cov_reset;
    this._bad_yaw_using_gps_courseIntialized = true;
    this._refObject.bad_yaw_using_gps_course = refObject.bad_yaw_using_gps_course;
    this._stopping_mag_useIntialized = true;
    this._refObject.stopping_mag_use = refObject.stopping_mag_use;
    this._vision_data_stoppedIntialized = true;
    this._refObject.vision_data_stopped = refObject.vision_data_stopped;
    this._emergency_yaw_reset_mag_stoppedIntialized = true;
    this._refObject.emergency_yaw_reset_mag_stopped = refObject.emergency_yaw_reset_mag_stopped;
    this._emergency_yaw_reset_gps_yaw_stoppedIntialized = true;
    this._refObject.emergency_yaw_reset_gps_yaw_stopped = refObject.emergency_yaw_reset_gps_yaw_stopped;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    EstimatorEventFlagsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'EstimatorEventFlags'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get information_event_changes() {
    if (!this._information_event_changesIntialized) {
      return undefined;
    }
    return this._refObject.information_event_changes;
  }
  set information_event_changes(value) {
    this._information_event_changesIntialized = true;
    this._refObject.information_event_changes = value;
  }
  get gps_checks_passed() {
    if (!this._gps_checks_passedIntialized) {
      return undefined;
    }
    return this._refObject.gps_checks_passed;
  }
  set gps_checks_passed(value) {
    this._gps_checks_passedIntialized = true;
    this._refObject.gps_checks_passed = value;
  }
  get reset_vel_to_gps() {
    if (!this._reset_vel_to_gpsIntialized) {
      return undefined;
    }
    return this._refObject.reset_vel_to_gps;
  }
  set reset_vel_to_gps(value) {
    this._reset_vel_to_gpsIntialized = true;
    this._refObject.reset_vel_to_gps = value;
  }
  get reset_vel_to_flow() {
    if (!this._reset_vel_to_flowIntialized) {
      return undefined;
    }
    return this._refObject.reset_vel_to_flow;
  }
  set reset_vel_to_flow(value) {
    this._reset_vel_to_flowIntialized = true;
    this._refObject.reset_vel_to_flow = value;
  }
  get reset_vel_to_vision() {
    if (!this._reset_vel_to_visionIntialized) {
      return undefined;
    }
    return this._refObject.reset_vel_to_vision;
  }
  set reset_vel_to_vision(value) {
    this._reset_vel_to_visionIntialized = true;
    this._refObject.reset_vel_to_vision = value;
  }
  get reset_vel_to_zero() {
    if (!this._reset_vel_to_zeroIntialized) {
      return undefined;
    }
    return this._refObject.reset_vel_to_zero;
  }
  set reset_vel_to_zero(value) {
    this._reset_vel_to_zeroIntialized = true;
    this._refObject.reset_vel_to_zero = value;
  }
  get reset_pos_to_last_known() {
    if (!this._reset_pos_to_last_knownIntialized) {
      return undefined;
    }
    return this._refObject.reset_pos_to_last_known;
  }
  set reset_pos_to_last_known(value) {
    this._reset_pos_to_last_knownIntialized = true;
    this._refObject.reset_pos_to_last_known = value;
  }
  get reset_pos_to_gps() {
    if (!this._reset_pos_to_gpsIntialized) {
      return undefined;
    }
    return this._refObject.reset_pos_to_gps;
  }
  set reset_pos_to_gps(value) {
    this._reset_pos_to_gpsIntialized = true;
    this._refObject.reset_pos_to_gps = value;
  }
  get reset_pos_to_vision() {
    if (!this._reset_pos_to_visionIntialized) {
      return undefined;
    }
    return this._refObject.reset_pos_to_vision;
  }
  set reset_pos_to_vision(value) {
    this._reset_pos_to_visionIntialized = true;
    this._refObject.reset_pos_to_vision = value;
  }
  get starting_gps_fusion() {
    if (!this._starting_gps_fusionIntialized) {
      return undefined;
    }
    return this._refObject.starting_gps_fusion;
  }
  set starting_gps_fusion(value) {
    this._starting_gps_fusionIntialized = true;
    this._refObject.starting_gps_fusion = value;
  }
  get starting_vision_pos_fusion() {
    if (!this._starting_vision_pos_fusionIntialized) {
      return undefined;
    }
    return this._refObject.starting_vision_pos_fusion;
  }
  set starting_vision_pos_fusion(value) {
    this._starting_vision_pos_fusionIntialized = true;
    this._refObject.starting_vision_pos_fusion = value;
  }
  get starting_vision_vel_fusion() {
    if (!this._starting_vision_vel_fusionIntialized) {
      return undefined;
    }
    return this._refObject.starting_vision_vel_fusion;
  }
  set starting_vision_vel_fusion(value) {
    this._starting_vision_vel_fusionIntialized = true;
    this._refObject.starting_vision_vel_fusion = value;
  }
  get starting_vision_yaw_fusion() {
    if (!this._starting_vision_yaw_fusionIntialized) {
      return undefined;
    }
    return this._refObject.starting_vision_yaw_fusion;
  }
  set starting_vision_yaw_fusion(value) {
    this._starting_vision_yaw_fusionIntialized = true;
    this._refObject.starting_vision_yaw_fusion = value;
  }
  get yaw_aligned_to_imu_gps() {
    if (!this._yaw_aligned_to_imu_gpsIntialized) {
      return undefined;
    }
    return this._refObject.yaw_aligned_to_imu_gps;
  }
  set yaw_aligned_to_imu_gps(value) {
    this._yaw_aligned_to_imu_gpsIntialized = true;
    this._refObject.yaw_aligned_to_imu_gps = value;
  }
  get reset_hgt_to_baro() {
    if (!this._reset_hgt_to_baroIntialized) {
      return undefined;
    }
    return this._refObject.reset_hgt_to_baro;
  }
  set reset_hgt_to_baro(value) {
    this._reset_hgt_to_baroIntialized = true;
    this._refObject.reset_hgt_to_baro = value;
  }
  get reset_hgt_to_gps() {
    if (!this._reset_hgt_to_gpsIntialized) {
      return undefined;
    }
    return this._refObject.reset_hgt_to_gps;
  }
  set reset_hgt_to_gps(value) {
    this._reset_hgt_to_gpsIntialized = true;
    this._refObject.reset_hgt_to_gps = value;
  }
  get reset_hgt_to_rng() {
    if (!this._reset_hgt_to_rngIntialized) {
      return undefined;
    }
    return this._refObject.reset_hgt_to_rng;
  }
  set reset_hgt_to_rng(value) {
    this._reset_hgt_to_rngIntialized = true;
    this._refObject.reset_hgt_to_rng = value;
  }
  get reset_hgt_to_ev() {
    if (!this._reset_hgt_to_evIntialized) {
      return undefined;
    }
    return this._refObject.reset_hgt_to_ev;
  }
  set reset_hgt_to_ev(value) {
    this._reset_hgt_to_evIntialized = true;
    this._refObject.reset_hgt_to_ev = value;
  }
  get warning_event_changes() {
    if (!this._warning_event_changesIntialized) {
      return undefined;
    }
    return this._refObject.warning_event_changes;
  }
  set warning_event_changes(value) {
    this._warning_event_changesIntialized = true;
    this._refObject.warning_event_changes = value;
  }
  get gps_quality_poor() {
    if (!this._gps_quality_poorIntialized) {
      return undefined;
    }
    return this._refObject.gps_quality_poor;
  }
  set gps_quality_poor(value) {
    this._gps_quality_poorIntialized = true;
    this._refObject.gps_quality_poor = value;
  }
  get gps_fusion_timout() {
    if (!this._gps_fusion_timoutIntialized) {
      return undefined;
    }
    return this._refObject.gps_fusion_timout;
  }
  set gps_fusion_timout(value) {
    this._gps_fusion_timoutIntialized = true;
    this._refObject.gps_fusion_timout = value;
  }
  get gps_data_stopped() {
    if (!this._gps_data_stoppedIntialized) {
      return undefined;
    }
    return this._refObject.gps_data_stopped;
  }
  set gps_data_stopped(value) {
    this._gps_data_stoppedIntialized = true;
    this._refObject.gps_data_stopped = value;
  }
  get gps_data_stopped_using_alternate() {
    if (!this._gps_data_stopped_using_alternateIntialized) {
      return undefined;
    }
    return this._refObject.gps_data_stopped_using_alternate;
  }
  set gps_data_stopped_using_alternate(value) {
    this._gps_data_stopped_using_alternateIntialized = true;
    this._refObject.gps_data_stopped_using_alternate = value;
  }
  get height_sensor_timeout() {
    if (!this._height_sensor_timeoutIntialized) {
      return undefined;
    }
    return this._refObject.height_sensor_timeout;
  }
  set height_sensor_timeout(value) {
    this._height_sensor_timeoutIntialized = true;
    this._refObject.height_sensor_timeout = value;
  }
  get stopping_navigation() {
    if (!this._stopping_navigationIntialized) {
      return undefined;
    }
    return this._refObject.stopping_navigation;
  }
  set stopping_navigation(value) {
    this._stopping_navigationIntialized = true;
    this._refObject.stopping_navigation = value;
  }
  get invalid_accel_bias_cov_reset() {
    if (!this._invalid_accel_bias_cov_resetIntialized) {
      return undefined;
    }
    return this._refObject.invalid_accel_bias_cov_reset;
  }
  set invalid_accel_bias_cov_reset(value) {
    this._invalid_accel_bias_cov_resetIntialized = true;
    this._refObject.invalid_accel_bias_cov_reset = value;
  }
  get bad_yaw_using_gps_course() {
    if (!this._bad_yaw_using_gps_courseIntialized) {
      return undefined;
    }
    return this._refObject.bad_yaw_using_gps_course;
  }
  set bad_yaw_using_gps_course(value) {
    this._bad_yaw_using_gps_courseIntialized = true;
    this._refObject.bad_yaw_using_gps_course = value;
  }
  get stopping_mag_use() {
    if (!this._stopping_mag_useIntialized) {
      return undefined;
    }
    return this._refObject.stopping_mag_use;
  }
  set stopping_mag_use(value) {
    this._stopping_mag_useIntialized = true;
    this._refObject.stopping_mag_use = value;
  }
  get vision_data_stopped() {
    if (!this._vision_data_stoppedIntialized) {
      return undefined;
    }
    return this._refObject.vision_data_stopped;
  }
  set vision_data_stopped(value) {
    this._vision_data_stoppedIntialized = true;
    this._refObject.vision_data_stopped = value;
  }
  get emergency_yaw_reset_mag_stopped() {
    if (!this._emergency_yaw_reset_mag_stoppedIntialized) {
      return undefined;
    }
    return this._refObject.emergency_yaw_reset_mag_stopped;
  }
  set emergency_yaw_reset_mag_stopped(value) {
    this._emergency_yaw_reset_mag_stoppedIntialized = true;
    this._refObject.emergency_yaw_reset_mag_stopped = value;
  }
  get emergency_yaw_reset_gps_yaw_stopped() {
    if (!this._emergency_yaw_reset_gps_yaw_stoppedIntialized) {
      return undefined;
    }
    return this._refObject.emergency_yaw_reset_gps_yaw_stopped;
  }
  set emergency_yaw_reset_gps_yaw_stopped(value) {
    this._emergency_yaw_reset_gps_yaw_stoppedIntialized = true;
    this._refObject.emergency_yaw_reset_gps_yaw_stopped = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EstimatorEventFlagsRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._information_event_changesIntialized = true;
    this._gps_checks_passedIntialized = true;
    this._reset_vel_to_gpsIntialized = true;
    this._reset_vel_to_flowIntialized = true;
    this._reset_vel_to_visionIntialized = true;
    this._reset_vel_to_zeroIntialized = true;
    this._reset_pos_to_last_knownIntialized = true;
    this._reset_pos_to_gpsIntialized = true;
    this._reset_pos_to_visionIntialized = true;
    this._starting_gps_fusionIntialized = true;
    this._starting_vision_pos_fusionIntialized = true;
    this._starting_vision_vel_fusionIntialized = true;
    this._starting_vision_yaw_fusionIntialized = true;
    this._yaw_aligned_to_imu_gpsIntialized = true;
    this._reset_hgt_to_baroIntialized = true;
    this._reset_hgt_to_gpsIntialized = true;
    this._reset_hgt_to_rngIntialized = true;
    this._reset_hgt_to_evIntialized = true;
    this._warning_event_changesIntialized = true;
    this._gps_quality_poorIntialized = true;
    this._gps_fusion_timoutIntialized = true;
    this._gps_data_stoppedIntialized = true;
    this._gps_data_stopped_using_alternateIntialized = true;
    this._height_sensor_timeoutIntialized = true;
    this._stopping_navigationIntialized = true;
    this._invalid_accel_bias_cov_resetIntialized = true;
    this._bad_yaw_using_gps_courseIntialized = true;
    this._stopping_mag_useIntialized = true;
    this._vision_data_stoppedIntialized = true;
    this._emergency_yaw_reset_mag_stoppedIntialized = true;
    this._emergency_yaw_reset_gps_yaw_stoppedIntialized = true;
  }
  copy(other) {
    this._refObject = new EstimatorEventFlagsRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._information_event_changesIntialized = true;
    this._gps_checks_passedIntialized = true;
    this._reset_vel_to_gpsIntialized = true;
    this._reset_vel_to_flowIntialized = true;
    this._reset_vel_to_visionIntialized = true;
    this._reset_vel_to_zeroIntialized = true;
    this._reset_pos_to_last_knownIntialized = true;
    this._reset_pos_to_gpsIntialized = true;
    this._reset_pos_to_visionIntialized = true;
    this._starting_gps_fusionIntialized = true;
    this._starting_vision_pos_fusionIntialized = true;
    this._starting_vision_vel_fusionIntialized = true;
    this._starting_vision_yaw_fusionIntialized = true;
    this._yaw_aligned_to_imu_gpsIntialized = true;
    this._reset_hgt_to_baroIntialized = true;
    this._reset_hgt_to_gpsIntialized = true;
    this._reset_hgt_to_rngIntialized = true;
    this._reset_hgt_to_evIntialized = true;
    this._warning_event_changesIntialized = true;
    this._gps_quality_poorIntialized = true;
    this._gps_fusion_timoutIntialized = true;
    this._gps_data_stoppedIntialized = true;
    this._gps_data_stopped_using_alternateIntialized = true;
    this._height_sensor_timeoutIntialized = true;
    this._stopping_navigationIntialized = true;
    this._invalid_accel_bias_cov_resetIntialized = true;
    this._bad_yaw_using_gps_courseIntialized = true;
    this._stopping_mag_useIntialized = true;
    this._vision_data_stoppedIntialized = true;
    this._emergency_yaw_reset_mag_stoppedIntialized = true;
    this._emergency_yaw_reset_gps_yaw_stoppedIntialized = true;
  }
  static get classType() {
    return EstimatorEventFlagsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"information_event_changes","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_checks_passed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_vel_to_gps","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_vel_to_flow","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_vel_to_vision","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_vel_to_zero","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_pos_to_last_known","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_pos_to_gps","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_pos_to_vision","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"starting_gps_fusion","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"starting_vision_pos_fusion","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"starting_vision_vel_fusion","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"starting_vision_yaw_fusion","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_aligned_to_imu_gps","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_hgt_to_baro","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_hgt_to_gps","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_hgt_to_rng","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_hgt_to_ev","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"warning_event_changes","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_quality_poor","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_fusion_timout","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_data_stopped","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_data_stopped_using_alternate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"height_sensor_timeout","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stopping_navigation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"invalid_accel_bias_cov_reset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bad_yaw_using_gps_course","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stopping_mag_use","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vision_data_stopped","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"emergency_yaw_reset_mag_stopped","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"emergency_yaw_reset_gps_yaw_stopped","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"EstimatorEventFlags","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EstimatorEventFlags"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","information_event_changes","gps_checks_passed","reset_vel_to_gps","reset_vel_to_flow","reset_vel_to_vision","reset_vel_to_zero","reset_pos_to_last_known","reset_pos_to_gps","reset_pos_to_vision","starting_gps_fusion","starting_vision_pos_fusion","starting_vision_vel_fusion","starting_vision_yaw_fusion","yaw_aligned_to_imu_gps","reset_hgt_to_baro","reset_hgt_to_gps","reset_hgt_to_rng","reset_hgt_to_ev","warning_event_changes","gps_quality_poor","gps_fusion_timout","gps_data_stopped","gps_data_stopped_using_alternate","height_sensor_timeout","stopping_navigation","invalid_accel_bias_cov_reset","bad_yaw_using_gps_course","stopping_mag_use","vision_data_stopped","emergency_yaw_reset_mag_stopped","emergency_yaw_reset_gps_yaw_stopped"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EstimatorEventFlagsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EstimatorEventFlagsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EstimatorEventFlagsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorEventFlagsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorEventFlagsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EstimatorEventFlagsRefArray(size);
    this._refObject = new EstimatorEventFlagsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EstimatorEventFlagsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EstimatorEventFlagsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EstimatorEventFlagsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EstimatorEventFlagsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EstimatorEventFlagsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EstimatorEventFlagsArrayWrapper;
  }
}
module.exports = EstimatorEventFlagsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "information_event_changes",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_checks_passed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_vel_to_gps",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_vel_to_flow",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_vel_to_vision",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_vel_to_zero",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_pos_to_last_known",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_pos_to_gps",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_pos_to_vision",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "starting_gps_fusion",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "starting_vision_pos_fusion",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "starting_vision_vel_fusion",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "starting_vision_yaw_fusion",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_aligned_to_imu_gps",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_hgt_to_baro",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_hgt_to_gps",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_hgt_to_rng",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_hgt_to_ev",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "warning_event_changes",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_quality_poor",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_fusion_timout",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_data_stopped",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_data_stopped_using_alternate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "height_sensor_timeout",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stopping_navigation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "invalid_accel_bias_cov_reset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "bad_yaw_using_gps_course",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stopping_mag_use",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vision_data_stopped",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "emergency_yaw_reset_mag_stopped",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "emergency_yaw_reset_gps_yaw_stopped",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "EstimatorEventFlags",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EstimatorEventFlags"
}
*/
