// rclnodejs/generated/px4_msgs/px4_msgs__msg__EstimatorBias3d.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const EstimatorBias3dRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  device_id: primitiveTypes.uint32,
  bias: ArrayType(primitiveTypes.float32, 3),
  bias_var: ArrayType(primitiveTypes.float32, 3),
  innov: ArrayType(primitiveTypes.float32, 3),
  innov_var: ArrayType(primitiveTypes.float32, 3),
  innov_test_ratio: ArrayType(primitiveTypes.float32, 3),
});
const EstimatorBias3dRefArray = ArrayType(EstimatorBias3dRefStruct);
const EstimatorBias3dRefStructArray = StructType({
  data: EstimatorBias3dRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EstimatorBias3dWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EstimatorBias3dRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._device_idIntialized = true;
      this._wrapperFields.bias = Float32Wrapper.createArray();
      this._wrapperFields.bias.copy(other._wrapperFields.bias);
      this._wrapperFields.bias_var = Float32Wrapper.createArray();
      this._wrapperFields.bias_var.copy(other._wrapperFields.bias_var);
      this._wrapperFields.innov = Float32Wrapper.createArray();
      this._wrapperFields.innov.copy(other._wrapperFields.innov);
      this._wrapperFields.innov_var = Float32Wrapper.createArray();
      this._wrapperFields.innov_var.copy(other._wrapperFields.innov_var);
      this._wrapperFields.innov_test_ratio = Float32Wrapper.createArray();
      this._wrapperFields.innov_test_ratio.copy(other._wrapperFields.innov_test_ratio);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EstimatorBias3dRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._device_idIntialized = false;
    this._wrapperFields.bias = Float32Wrapper.createArray();
    this._wrapperFields.bias_var = Float32Wrapper.createArray();
    this._wrapperFields.innov = Float32Wrapper.createArray();
    this._wrapperFields.innov_var = Float32Wrapper.createArray();
    this._wrapperFields.innov_test_ratio = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new EstimatorBias3dWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EstimatorBias3dArrayWrapper;
  }
  static get ArrayType() {
    return EstimatorBias3dArrayWrapper;
  }
  static get refObjectArrayType() {
    return EstimatorBias3dRefStructArray
  }
  static get refObjectType() {
    return EstimatorBias3dRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in EstimatorBias3d');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in EstimatorBias3d');
      }
      if (checkConsistency && !this._device_idIntialized) {
        throw new TypeError('Invalid argument: device_id in EstimatorBias3d');
      }
    this._refObject.bias = Array.from(this._wrapperFields.bias.data);
    this._refObject.bias_var = Array.from(this._wrapperFields.bias_var.data);
    this._refObject.innov = Array.from(this._wrapperFields.innov.data);
    this._refObject.innov_var = Array.from(this._wrapperFields.innov_var.data);
    this._refObject.innov_test_ratio = Array.from(this._wrapperFields.innov_test_ratio.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._device_idIntialized = true;
    this._refObject.device_id = refObject.device_id;
    this._wrapperFields.bias.fill(refObject.bias.toArray());
    this._wrapperFields.bias_var.fill(refObject.bias_var.toArray());
    this._wrapperFields.innov.fill(refObject.innov.toArray());
    this._wrapperFields.innov_var.fill(refObject.innov_var.toArray());
    this._wrapperFields.innov_test_ratio.fill(refObject.innov_test_ratio.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    EstimatorBias3dWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'EstimatorBias3d'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get device_id() {
    if (!this._device_idIntialized) {
      return undefined;
    }
    return this._refObject.device_id;
  }
  set device_id(value) {
    this._device_idIntialized = true;
    this._refObject.device_id = value;
  }
  get bias() {
    return this._wrapperFields['bias'].data;
  }
  set bias(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['bias'].fill(value);
  }
  get bias_var() {
    return this._wrapperFields['bias_var'].data;
  }
  set bias_var(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['bias_var'].fill(value);
  }
  get innov() {
    return this._wrapperFields['innov'].data;
  }
  set innov(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['innov'].fill(value);
  }
  get innov_var() {
    return this._wrapperFields['innov_var'].data;
  }
  set innov_var(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['innov_var'].fill(value);
  }
  get innov_test_ratio() {
    return this._wrapperFields['innov_test_ratio'].data;
  }
  set innov_test_ratio(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['innov_test_ratio'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new EstimatorBias3dRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._device_idIntialized = true;
    this._wrapperFields.bias.fill(refObject.bias.toArray());
    this._wrapperFields.bias_var.fill(refObject.bias_var.toArray());
    this._wrapperFields.innov.fill(refObject.innov.toArray());
    this._wrapperFields.innov_var.fill(refObject.innov_var.toArray());
    this._wrapperFields.innov_test_ratio.fill(refObject.innov_test_ratio.toArray());
  }
  copy(other) {
    this._refObject = new EstimatorBias3dRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._device_idIntialized = true;
    this._wrapperFields.bias.copy(other._wrapperFields.bias);
    this._wrapperFields.bias_var.copy(other._wrapperFields.bias_var);
    this._wrapperFields.innov.copy(other._wrapperFields.innov);
    this._wrapperFields.innov_var.copy(other._wrapperFields.innov_var);
    this._wrapperFields.innov_test_ratio.copy(other._wrapperFields.innov_test_ratio);
  }
  static get classType() {
    return EstimatorBias3dWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bias","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bias_var","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov_var","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov_test_ratio","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"EstimatorBias3d","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EstimatorBias3d"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","device_id","bias","bias_var","innov","innov_var","innov_test_ratio"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EstimatorBias3dArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EstimatorBias3dWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EstimatorBias3dWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorBias3dArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorBias3dArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EstimatorBias3dRefArray(size);
    this._refObject = new EstimatorBias3dRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EstimatorBias3dWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EstimatorBias3dArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EstimatorBias3dArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EstimatorBias3dWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EstimatorBias3dWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EstimatorBias3dArrayWrapper;
  }
}
module.exports = EstimatorBias3dWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "bias",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "bias_var",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov_var",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov_test_ratio",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "EstimatorBias3d",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EstimatorBias3d"
}
*/
