// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleImu.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VehicleImuRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  accel_device_id: primitiveTypes.uint32,
  gyro_device_id: primitiveTypes.uint32,
  delta_angle: ArrayType(primitiveTypes.float32, 3),
  delta_velocity: ArrayType(primitiveTypes.float32, 3),
  delta_angle_dt: primitiveTypes.uint16,
  delta_velocity_dt: primitiveTypes.uint16,
  delta_angle_clipping: primitiveTypes.uint8,
  delta_velocity_clipping: primitiveTypes.uint8,
  accel_calibration_count: primitiveTypes.uint8,
  gyro_calibration_count: primitiveTypes.uint8,
});
const VehicleImuRefArray = ArrayType(VehicleImuRefStruct);
const VehicleImuRefStructArray = StructType({
  data: VehicleImuRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleImuWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleImuRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._accel_device_idIntialized = true;
      this._gyro_device_idIntialized = true;
      this._wrapperFields.delta_angle = Float32Wrapper.createArray();
      this._wrapperFields.delta_angle.copy(other._wrapperFields.delta_angle);
      this._wrapperFields.delta_velocity = Float32Wrapper.createArray();
      this._wrapperFields.delta_velocity.copy(other._wrapperFields.delta_velocity);
      this._delta_angle_dtIntialized = true;
      this._delta_velocity_dtIntialized = true;
      this._delta_angle_clippingIntialized = true;
      this._delta_velocity_clippingIntialized = true;
      this._accel_calibration_countIntialized = true;
      this._gyro_calibration_countIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleImuRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._accel_device_idIntialized = false;
    this._gyro_device_idIntialized = false;
    this._wrapperFields.delta_angle = Float32Wrapper.createArray();
    this._wrapperFields.delta_velocity = Float32Wrapper.createArray();
    this._delta_angle_dtIntialized = false;
    this._delta_velocity_dtIntialized = false;
    this._delta_angle_clippingIntialized = false;
    this._delta_velocity_clippingIntialized = false;
    this._accel_calibration_countIntialized = false;
    this._gyro_calibration_countIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleImuWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleImuArrayWrapper;
  }
  static get ArrayType() {
    return VehicleImuArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleImuRefStructArray
  }
  static get refObjectType() {
    return VehicleImuRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleImu');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in VehicleImu');
      }
      if (checkConsistency && !this._accel_device_idIntialized) {
        throw new TypeError('Invalid argument: accel_device_id in VehicleImu');
      }
      if (checkConsistency && !this._gyro_device_idIntialized) {
        throw new TypeError('Invalid argument: gyro_device_id in VehicleImu');
      }
      if (checkConsistency && !this._delta_angle_dtIntialized) {
        throw new TypeError('Invalid argument: delta_angle_dt in VehicleImu');
      }
      if (checkConsistency && !this._delta_velocity_dtIntialized) {
        throw new TypeError('Invalid argument: delta_velocity_dt in VehicleImu');
      }
      if (checkConsistency && !this._delta_angle_clippingIntialized) {
        throw new TypeError('Invalid argument: delta_angle_clipping in VehicleImu');
      }
      if (checkConsistency && !this._delta_velocity_clippingIntialized) {
        throw new TypeError('Invalid argument: delta_velocity_clipping in VehicleImu');
      }
      if (checkConsistency && !this._accel_calibration_countIntialized) {
        throw new TypeError('Invalid argument: accel_calibration_count in VehicleImu');
      }
      if (checkConsistency && !this._gyro_calibration_countIntialized) {
        throw new TypeError('Invalid argument: gyro_calibration_count in VehicleImu');
      }
    this._refObject.delta_angle = Array.from(this._wrapperFields.delta_angle.data);
    this._refObject.delta_velocity = Array.from(this._wrapperFields.delta_velocity.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._accel_device_idIntialized = true;
    this._refObject.accel_device_id = refObject.accel_device_id;
    this._gyro_device_idIntialized = true;
    this._refObject.gyro_device_id = refObject.gyro_device_id;
    this._wrapperFields.delta_angle.fill(refObject.delta_angle.toArray());
    this._wrapperFields.delta_velocity.fill(refObject.delta_velocity.toArray());
    this._delta_angle_dtIntialized = true;
    this._refObject.delta_angle_dt = refObject.delta_angle_dt;
    this._delta_velocity_dtIntialized = true;
    this._refObject.delta_velocity_dt = refObject.delta_velocity_dt;
    this._delta_angle_clippingIntialized = true;
    this._refObject.delta_angle_clipping = refObject.delta_angle_clipping;
    this._delta_velocity_clippingIntialized = true;
    this._refObject.delta_velocity_clipping = refObject.delta_velocity_clipping;
    this._accel_calibration_countIntialized = true;
    this._refObject.accel_calibration_count = refObject.accel_calibration_count;
    this._gyro_calibration_countIntialized = true;
    this._refObject.gyro_calibration_count = refObject.gyro_calibration_count;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleImuWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleImu'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get accel_device_id() {
    if (!this._accel_device_idIntialized) {
      return undefined;
    }
    return this._refObject.accel_device_id;
  }
  set accel_device_id(value) {
    this._accel_device_idIntialized = true;
    this._refObject.accel_device_id = value;
  }
  get gyro_device_id() {
    if (!this._gyro_device_idIntialized) {
      return undefined;
    }
    return this._refObject.gyro_device_id;
  }
  set gyro_device_id(value) {
    this._gyro_device_idIntialized = true;
    this._refObject.gyro_device_id = value;
  }
  get delta_angle() {
    return this._wrapperFields['delta_angle'].data;
  }
  set delta_angle(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['delta_angle'].fill(value);
  }
  get delta_velocity() {
    return this._wrapperFields['delta_velocity'].data;
  }
  set delta_velocity(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['delta_velocity'].fill(value);
  }
  get delta_angle_dt() {
    if (!this._delta_angle_dtIntialized) {
      return undefined;
    }
    return this._refObject.delta_angle_dt;
  }
  set delta_angle_dt(value) {
    this._delta_angle_dtIntialized = true;
    this._refObject.delta_angle_dt = value;
  }
  get delta_velocity_dt() {
    if (!this._delta_velocity_dtIntialized) {
      return undefined;
    }
    return this._refObject.delta_velocity_dt;
  }
  set delta_velocity_dt(value) {
    this._delta_velocity_dtIntialized = true;
    this._refObject.delta_velocity_dt = value;
  }
  get delta_angle_clipping() {
    if (!this._delta_angle_clippingIntialized) {
      return undefined;
    }
    return this._refObject.delta_angle_clipping;
  }
  set delta_angle_clipping(value) {
    this._delta_angle_clippingIntialized = true;
    this._refObject.delta_angle_clipping = value;
  }
  get delta_velocity_clipping() {
    if (!this._delta_velocity_clippingIntialized) {
      return undefined;
    }
    return this._refObject.delta_velocity_clipping;
  }
  set delta_velocity_clipping(value) {
    this._delta_velocity_clippingIntialized = true;
    this._refObject.delta_velocity_clipping = value;
  }
  get accel_calibration_count() {
    if (!this._accel_calibration_countIntialized) {
      return undefined;
    }
    return this._refObject.accel_calibration_count;
  }
  set accel_calibration_count(value) {
    this._accel_calibration_countIntialized = true;
    this._refObject.accel_calibration_count = value;
  }
  get gyro_calibration_count() {
    if (!this._gyro_calibration_countIntialized) {
      return undefined;
    }
    return this._refObject.gyro_calibration_count;
  }
  set gyro_calibration_count(value) {
    this._gyro_calibration_countIntialized = true;
    this._refObject.gyro_calibration_count = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleImuRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._accel_device_idIntialized = true;
    this._gyro_device_idIntialized = true;
    this._wrapperFields.delta_angle.fill(refObject.delta_angle.toArray());
    this._wrapperFields.delta_velocity.fill(refObject.delta_velocity.toArray());
    this._delta_angle_dtIntialized = true;
    this._delta_velocity_dtIntialized = true;
    this._delta_angle_clippingIntialized = true;
    this._delta_velocity_clippingIntialized = true;
    this._accel_calibration_countIntialized = true;
    this._gyro_calibration_countIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleImuRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._accel_device_idIntialized = true;
    this._gyro_device_idIntialized = true;
    this._wrapperFields.delta_angle.copy(other._wrapperFields.delta_angle);
    this._wrapperFields.delta_velocity.copy(other._wrapperFields.delta_velocity);
    this._delta_angle_dtIntialized = true;
    this._delta_velocity_dtIntialized = true;
    this._delta_angle_clippingIntialized = true;
    this._delta_velocity_clippingIntialized = true;
    this._accel_calibration_countIntialized = true;
    this._gyro_calibration_countIntialized = true;
  }
  static get classType() {
    return VehicleImuWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"CLIPPING_X","value":1},{"type":"uint8","name":"CLIPPING_Y","value":2},{"type":"uint8","name":"CLIPPING_Z","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_angle","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_velocity","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_angle_dt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_velocity_dt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_angle_clipping","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_velocity_clipping","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_calibration_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_calibration_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleImu","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleImu"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","accel_device_id","gyro_device_id","delta_angle","delta_velocity","delta_angle_dt","delta_velocity_dt","delta_angle_clipping","delta_velocity_clipping","accel_calibration_count","gyro_calibration_count"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleImuArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleImuWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleImuWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleImuArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleImuArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleImuRefArray(size);
    this._refObject = new VehicleImuRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleImuWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleImuArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleImuArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleImuWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleImuWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleImuArrayWrapper;
  }
}
// Define constants (3 in total)
Object.defineProperty(VehicleImuWrapper, "CLIPPING_X", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleImuWrapper, "CLIPPING_Y", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleImuWrapper, "CLIPPING_Z", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = VehicleImuWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "CLIPPING_X",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "CLIPPING_Y",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "CLIPPING_Z",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_angle",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_velocity",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_angle_dt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_velocity_dt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_angle_clipping",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_velocity_clipping",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_calibration_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_calibration_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleImu",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleImu"
}
*/
