// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleOpticalFlowVel.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VehicleOpticalFlowVelRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  vel_body: ArrayType(primitiveTypes.float32, 2),
  vel_ne: ArrayType(primitiveTypes.float32, 2),
  flow_uncompensated_integral: ArrayType(primitiveTypes.float32, 2),
  flow_compensated_integral: ArrayType(primitiveTypes.float32, 2),
  gyro_rate: ArrayType(primitiveTypes.float32, 3),
  gyro_rate_integral: ArrayType(primitiveTypes.float32, 3),
});
const VehicleOpticalFlowVelRefArray = ArrayType(VehicleOpticalFlowVelRefStruct);
const VehicleOpticalFlowVelRefStructArray = StructType({
  data: VehicleOpticalFlowVelRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleOpticalFlowVelWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleOpticalFlowVelRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._wrapperFields.vel_body = Float32Wrapper.createArray();
      this._wrapperFields.vel_body.copy(other._wrapperFields.vel_body);
      this._wrapperFields.vel_ne = Float32Wrapper.createArray();
      this._wrapperFields.vel_ne.copy(other._wrapperFields.vel_ne);
      this._wrapperFields.flow_uncompensated_integral = Float32Wrapper.createArray();
      this._wrapperFields.flow_uncompensated_integral.copy(other._wrapperFields.flow_uncompensated_integral);
      this._wrapperFields.flow_compensated_integral = Float32Wrapper.createArray();
      this._wrapperFields.flow_compensated_integral.copy(other._wrapperFields.flow_compensated_integral);
      this._wrapperFields.gyro_rate = Float32Wrapper.createArray();
      this._wrapperFields.gyro_rate.copy(other._wrapperFields.gyro_rate);
      this._wrapperFields.gyro_rate_integral = Float32Wrapper.createArray();
      this._wrapperFields.gyro_rate_integral.copy(other._wrapperFields.gyro_rate_integral);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleOpticalFlowVelRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._wrapperFields.vel_body = Float32Wrapper.createArray();
    this._wrapperFields.vel_ne = Float32Wrapper.createArray();
    this._wrapperFields.flow_uncompensated_integral = Float32Wrapper.createArray();
    this._wrapperFields.flow_compensated_integral = Float32Wrapper.createArray();
    this._wrapperFields.gyro_rate = Float32Wrapper.createArray();
    this._wrapperFields.gyro_rate_integral = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new VehicleOpticalFlowVelWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleOpticalFlowVelArrayWrapper;
  }
  static get ArrayType() {
    return VehicleOpticalFlowVelArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleOpticalFlowVelRefStructArray
  }
  static get refObjectType() {
    return VehicleOpticalFlowVelRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleOpticalFlowVel');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in VehicleOpticalFlowVel');
      }
    this._refObject.vel_body = Array.from(this._wrapperFields.vel_body.data);
    this._refObject.vel_ne = Array.from(this._wrapperFields.vel_ne.data);
    this._refObject.flow_uncompensated_integral = Array.from(this._wrapperFields.flow_uncompensated_integral.data);
    this._refObject.flow_compensated_integral = Array.from(this._wrapperFields.flow_compensated_integral.data);
    this._refObject.gyro_rate = Array.from(this._wrapperFields.gyro_rate.data);
    this._refObject.gyro_rate_integral = Array.from(this._wrapperFields.gyro_rate_integral.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._wrapperFields.vel_body.fill(refObject.vel_body.toArray());
    this._wrapperFields.vel_ne.fill(refObject.vel_ne.toArray());
    this._wrapperFields.flow_uncompensated_integral.fill(refObject.flow_uncompensated_integral.toArray());
    this._wrapperFields.flow_compensated_integral.fill(refObject.flow_compensated_integral.toArray());
    this._wrapperFields.gyro_rate.fill(refObject.gyro_rate.toArray());
    this._wrapperFields.gyro_rate_integral.fill(refObject.gyro_rate_integral.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleOpticalFlowVelWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleOpticalFlowVel'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get vel_body() {
    return this._wrapperFields['vel_body'].data;
  }
  set vel_body(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['vel_body'].fill(value);
  }
  get vel_ne() {
    return this._wrapperFields['vel_ne'].data;
  }
  set vel_ne(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['vel_ne'].fill(value);
  }
  get flow_uncompensated_integral() {
    return this._wrapperFields['flow_uncompensated_integral'].data;
  }
  set flow_uncompensated_integral(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['flow_uncompensated_integral'].fill(value);
  }
  get flow_compensated_integral() {
    return this._wrapperFields['flow_compensated_integral'].data;
  }
  set flow_compensated_integral(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['flow_compensated_integral'].fill(value);
  }
  get gyro_rate() {
    return this._wrapperFields['gyro_rate'].data;
  }
  set gyro_rate(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['gyro_rate'].fill(value);
  }
  get gyro_rate_integral() {
    return this._wrapperFields['gyro_rate_integral'].data;
  }
  set gyro_rate_integral(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['gyro_rate_integral'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleOpticalFlowVelRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._wrapperFields.vel_body.fill(refObject.vel_body.toArray());
    this._wrapperFields.vel_ne.fill(refObject.vel_ne.toArray());
    this._wrapperFields.flow_uncompensated_integral.fill(refObject.flow_uncompensated_integral.toArray());
    this._wrapperFields.flow_compensated_integral.fill(refObject.flow_compensated_integral.toArray());
    this._wrapperFields.gyro_rate.fill(refObject.gyro_rate.toArray());
    this._wrapperFields.gyro_rate_integral.fill(refObject.gyro_rate_integral.toArray());
  }
  copy(other) {
    this._refObject = new VehicleOpticalFlowVelRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._wrapperFields.vel_body.copy(other._wrapperFields.vel_body);
    this._wrapperFields.vel_ne.copy(other._wrapperFields.vel_ne);
    this._wrapperFields.flow_uncompensated_integral.copy(other._wrapperFields.flow_uncompensated_integral);
    this._wrapperFields.flow_compensated_integral.copy(other._wrapperFields.flow_compensated_integral);
    this._wrapperFields.gyro_rate.copy(other._wrapperFields.gyro_rate);
    this._wrapperFields.gyro_rate_integral.copy(other._wrapperFields.gyro_rate_integral);
  }
  static get classType() {
    return VehicleOpticalFlowVelWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vel_body","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vel_ne","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flow_uncompensated_integral","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flow_compensated_integral","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_rate","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_rate_integral","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleOpticalFlowVel","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleOpticalFlowVel"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","vel_body","vel_ne","flow_uncompensated_integral","flow_compensated_integral","gyro_rate","gyro_rate_integral"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleOpticalFlowVelArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleOpticalFlowVelWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleOpticalFlowVelWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleOpticalFlowVelArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleOpticalFlowVelArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleOpticalFlowVelRefArray(size);
    this._refObject = new VehicleOpticalFlowVelRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleOpticalFlowVelWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleOpticalFlowVelArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleOpticalFlowVelArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleOpticalFlowVelWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleOpticalFlowVelWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleOpticalFlowVelArrayWrapper;
  }
}
module.exports = VehicleOpticalFlowVelWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vel_body",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vel_ne",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flow_uncompensated_integral",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flow_compensated_integral",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_rate",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_rate_integral",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleOpticalFlowVel",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleOpticalFlowVel"
}
*/
