// rclnodejs/generated/px4_msgs/px4_msgs__msg__SensorUwb.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SensorUwbRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  sessionid: primitiveTypes.uint32,
  time_offset: primitiveTypes.uint32,
  counter: primitiveTypes.uint32,
  mac: primitiveTypes.uint16,
  mac_dest: primitiveTypes.uint16,
  status: primitiveTypes.uint16,
  nlos: primitiveTypes.uint8,
  distance: primitiveTypes.float32,
  aoa_azimuth_dev: primitiveTypes.float32,
  aoa_elevation_dev: primitiveTypes.float32,
  aoa_azimuth_resp: primitiveTypes.float32,
  aoa_elevation_resp: primitiveTypes.float32,
  aoa_azimuth_fom: primitiveTypes.uint8,
  aoa_elevation_fom: primitiveTypes.uint8,
  aoa_dest_azimuth_fom: primitiveTypes.uint8,
  aoa_dest_elevation_fom: primitiveTypes.uint8,
  orientation: primitiveTypes.uint8,
  offset_x: primitiveTypes.float32,
  offset_y: primitiveTypes.float32,
  offset_z: primitiveTypes.float32,
});
const SensorUwbRefArray = ArrayType(SensorUwbRefStruct);
const SensorUwbRefStructArray = StructType({
  data: SensorUwbRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SensorUwbWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SensorUwbRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._sessionidIntialized = true;
      this._time_offsetIntialized = true;
      this._counterIntialized = true;
      this._macIntialized = true;
      this._mac_destIntialized = true;
      this._statusIntialized = true;
      this._nlosIntialized = true;
      this._distanceIntialized = true;
      this._aoa_azimuth_devIntialized = true;
      this._aoa_elevation_devIntialized = true;
      this._aoa_azimuth_respIntialized = true;
      this._aoa_elevation_respIntialized = true;
      this._aoa_azimuth_fomIntialized = true;
      this._aoa_elevation_fomIntialized = true;
      this._aoa_dest_azimuth_fomIntialized = true;
      this._aoa_dest_elevation_fomIntialized = true;
      this._orientationIntialized = true;
      this._offset_xIntialized = true;
      this._offset_yIntialized = true;
      this._offset_zIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SensorUwbRefStruct();
    this._timestampIntialized = false;
    this._sessionidIntialized = false;
    this._time_offsetIntialized = false;
    this._counterIntialized = false;
    this._macIntialized = false;
    this._mac_destIntialized = false;
    this._statusIntialized = false;
    this._nlosIntialized = false;
    this._distanceIntialized = false;
    this._aoa_azimuth_devIntialized = false;
    this._aoa_elevation_devIntialized = false;
    this._aoa_azimuth_respIntialized = false;
    this._aoa_elevation_respIntialized = false;
    this._aoa_azimuth_fomIntialized = false;
    this._aoa_elevation_fomIntialized = false;
    this._aoa_dest_azimuth_fomIntialized = false;
    this._aoa_dest_elevation_fomIntialized = false;
    this._orientationIntialized = false;
    this._offset_xIntialized = false;
    this._offset_yIntialized = false;
    this._offset_zIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SensorUwbWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SensorUwbArrayWrapper;
  }
  static get ArrayType() {
    return SensorUwbArrayWrapper;
  }
  static get refObjectArrayType() {
    return SensorUwbRefStructArray
  }
  static get refObjectType() {
    return SensorUwbRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in SensorUwb');
      }
      if (checkConsistency && !this._sessionidIntialized) {
        throw new TypeError('Invalid argument: sessionid in SensorUwb');
      }
      if (checkConsistency && !this._time_offsetIntialized) {
        throw new TypeError('Invalid argument: time_offset in SensorUwb');
      }
      if (checkConsistency && !this._counterIntialized) {
        throw new TypeError('Invalid argument: counter in SensorUwb');
      }
      if (checkConsistency && !this._macIntialized) {
        throw new TypeError('Invalid argument: mac in SensorUwb');
      }
      if (checkConsistency && !this._mac_destIntialized) {
        throw new TypeError('Invalid argument: mac_dest in SensorUwb');
      }
      if (checkConsistency && !this._statusIntialized) {
        throw new TypeError('Invalid argument: status in SensorUwb');
      }
      if (checkConsistency && !this._nlosIntialized) {
        throw new TypeError('Invalid argument: nlos in SensorUwb');
      }
      if (checkConsistency && !this._distanceIntialized) {
        throw new TypeError('Invalid argument: distance in SensorUwb');
      }
      if (checkConsistency && !this._aoa_azimuth_devIntialized) {
        throw new TypeError('Invalid argument: aoa_azimuth_dev in SensorUwb');
      }
      if (checkConsistency && !this._aoa_elevation_devIntialized) {
        throw new TypeError('Invalid argument: aoa_elevation_dev in SensorUwb');
      }
      if (checkConsistency && !this._aoa_azimuth_respIntialized) {
        throw new TypeError('Invalid argument: aoa_azimuth_resp in SensorUwb');
      }
      if (checkConsistency && !this._aoa_elevation_respIntialized) {
        throw new TypeError('Invalid argument: aoa_elevation_resp in SensorUwb');
      }
      if (checkConsistency && !this._aoa_azimuth_fomIntialized) {
        throw new TypeError('Invalid argument: aoa_azimuth_fom in SensorUwb');
      }
      if (checkConsistency && !this._aoa_elevation_fomIntialized) {
        throw new TypeError('Invalid argument: aoa_elevation_fom in SensorUwb');
      }
      if (checkConsistency && !this._aoa_dest_azimuth_fomIntialized) {
        throw new TypeError('Invalid argument: aoa_dest_azimuth_fom in SensorUwb');
      }
      if (checkConsistency && !this._aoa_dest_elevation_fomIntialized) {
        throw new TypeError('Invalid argument: aoa_dest_elevation_fom in SensorUwb');
      }
      if (checkConsistency && !this._orientationIntialized) {
        throw new TypeError('Invalid argument: orientation in SensorUwb');
      }
      if (checkConsistency && !this._offset_xIntialized) {
        throw new TypeError('Invalid argument: offset_x in SensorUwb');
      }
      if (checkConsistency && !this._offset_yIntialized) {
        throw new TypeError('Invalid argument: offset_y in SensorUwb');
      }
      if (checkConsistency && !this._offset_zIntialized) {
        throw new TypeError('Invalid argument: offset_z in SensorUwb');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._sessionidIntialized = true;
    this._refObject.sessionid = refObject.sessionid;
    this._time_offsetIntialized = true;
    this._refObject.time_offset = refObject.time_offset;
    this._counterIntialized = true;
    this._refObject.counter = refObject.counter;
    this._macIntialized = true;
    this._refObject.mac = refObject.mac;
    this._mac_destIntialized = true;
    this._refObject.mac_dest = refObject.mac_dest;
    this._statusIntialized = true;
    this._refObject.status = refObject.status;
    this._nlosIntialized = true;
    this._refObject.nlos = refObject.nlos;
    this._distanceIntialized = true;
    this._refObject.distance = refObject.distance;
    this._aoa_azimuth_devIntialized = true;
    this._refObject.aoa_azimuth_dev = refObject.aoa_azimuth_dev;
    this._aoa_elevation_devIntialized = true;
    this._refObject.aoa_elevation_dev = refObject.aoa_elevation_dev;
    this._aoa_azimuth_respIntialized = true;
    this._refObject.aoa_azimuth_resp = refObject.aoa_azimuth_resp;
    this._aoa_elevation_respIntialized = true;
    this._refObject.aoa_elevation_resp = refObject.aoa_elevation_resp;
    this._aoa_azimuth_fomIntialized = true;
    this._refObject.aoa_azimuth_fom = refObject.aoa_azimuth_fom;
    this._aoa_elevation_fomIntialized = true;
    this._refObject.aoa_elevation_fom = refObject.aoa_elevation_fom;
    this._aoa_dest_azimuth_fomIntialized = true;
    this._refObject.aoa_dest_azimuth_fom = refObject.aoa_dest_azimuth_fom;
    this._aoa_dest_elevation_fomIntialized = true;
    this._refObject.aoa_dest_elevation_fom = refObject.aoa_dest_elevation_fom;
    this._orientationIntialized = true;
    this._refObject.orientation = refObject.orientation;
    this._offset_xIntialized = true;
    this._refObject.offset_x = refObject.offset_x;
    this._offset_yIntialized = true;
    this._refObject.offset_y = refObject.offset_y;
    this._offset_zIntialized = true;
    this._refObject.offset_z = refObject.offset_z;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SensorUwbWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'SensorUwb'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get sessionid() {
    if (!this._sessionidIntialized) {
      return undefined;
    }
    return this._refObject.sessionid;
  }
  set sessionid(value) {
    this._sessionidIntialized = true;
    this._refObject.sessionid = value;
  }
  get time_offset() {
    if (!this._time_offsetIntialized) {
      return undefined;
    }
    return this._refObject.time_offset;
  }
  set time_offset(value) {
    this._time_offsetIntialized = true;
    this._refObject.time_offset = value;
  }
  get counter() {
    if (!this._counterIntialized) {
      return undefined;
    }
    return this._refObject.counter;
  }
  set counter(value) {
    this._counterIntialized = true;
    this._refObject.counter = value;
  }
  get mac() {
    if (!this._macIntialized) {
      return undefined;
    }
    return this._refObject.mac;
  }
  set mac(value) {
    this._macIntialized = true;
    this._refObject.mac = value;
  }
  get mac_dest() {
    if (!this._mac_destIntialized) {
      return undefined;
    }
    return this._refObject.mac_dest;
  }
  set mac_dest(value) {
    this._mac_destIntialized = true;
    this._refObject.mac_dest = value;
  }
  get status() {
    if (!this._statusIntialized) {
      return undefined;
    }
    return this._refObject.status;
  }
  set status(value) {
    this._statusIntialized = true;
    this._refObject.status = value;
  }
  get nlos() {
    if (!this._nlosIntialized) {
      return undefined;
    }
    return this._refObject.nlos;
  }
  set nlos(value) {
    this._nlosIntialized = true;
    this._refObject.nlos = value;
  }
  get distance() {
    if (!this._distanceIntialized) {
      return undefined;
    }
    return this._refObject.distance;
  }
  set distance(value) {
    this._distanceIntialized = true;
    this._refObject.distance = value;
  }
  get aoa_azimuth_dev() {
    if (!this._aoa_azimuth_devIntialized) {
      return undefined;
    }
    return this._refObject.aoa_azimuth_dev;
  }
  set aoa_azimuth_dev(value) {
    this._aoa_azimuth_devIntialized = true;
    this._refObject.aoa_azimuth_dev = value;
  }
  get aoa_elevation_dev() {
    if (!this._aoa_elevation_devIntialized) {
      return undefined;
    }
    return this._refObject.aoa_elevation_dev;
  }
  set aoa_elevation_dev(value) {
    this._aoa_elevation_devIntialized = true;
    this._refObject.aoa_elevation_dev = value;
  }
  get aoa_azimuth_resp() {
    if (!this._aoa_azimuth_respIntialized) {
      return undefined;
    }
    return this._refObject.aoa_azimuth_resp;
  }
  set aoa_azimuth_resp(value) {
    this._aoa_azimuth_respIntialized = true;
    this._refObject.aoa_azimuth_resp = value;
  }
  get aoa_elevation_resp() {
    if (!this._aoa_elevation_respIntialized) {
      return undefined;
    }
    return this._refObject.aoa_elevation_resp;
  }
  set aoa_elevation_resp(value) {
    this._aoa_elevation_respIntialized = true;
    this._refObject.aoa_elevation_resp = value;
  }
  get aoa_azimuth_fom() {
    if (!this._aoa_azimuth_fomIntialized) {
      return undefined;
    }
    return this._refObject.aoa_azimuth_fom;
  }
  set aoa_azimuth_fom(value) {
    this._aoa_azimuth_fomIntialized = true;
    this._refObject.aoa_azimuth_fom = value;
  }
  get aoa_elevation_fom() {
    if (!this._aoa_elevation_fomIntialized) {
      return undefined;
    }
    return this._refObject.aoa_elevation_fom;
  }
  set aoa_elevation_fom(value) {
    this._aoa_elevation_fomIntialized = true;
    this._refObject.aoa_elevation_fom = value;
  }
  get aoa_dest_azimuth_fom() {
    if (!this._aoa_dest_azimuth_fomIntialized) {
      return undefined;
    }
    return this._refObject.aoa_dest_azimuth_fom;
  }
  set aoa_dest_azimuth_fom(value) {
    this._aoa_dest_azimuth_fomIntialized = true;
    this._refObject.aoa_dest_azimuth_fom = value;
  }
  get aoa_dest_elevation_fom() {
    if (!this._aoa_dest_elevation_fomIntialized) {
      return undefined;
    }
    return this._refObject.aoa_dest_elevation_fom;
  }
  set aoa_dest_elevation_fom(value) {
    this._aoa_dest_elevation_fomIntialized = true;
    this._refObject.aoa_dest_elevation_fom = value;
  }
  get orientation() {
    if (!this._orientationIntialized) {
      return undefined;
    }
    return this._refObject.orientation;
  }
  set orientation(value) {
    this._orientationIntialized = true;
    this._refObject.orientation = value;
  }
  get offset_x() {
    if (!this._offset_xIntialized) {
      return undefined;
    }
    return this._refObject.offset_x;
  }
  set offset_x(value) {
    this._offset_xIntialized = true;
    this._refObject.offset_x = value;
  }
  get offset_y() {
    if (!this._offset_yIntialized) {
      return undefined;
    }
    return this._refObject.offset_y;
  }
  set offset_y(value) {
    this._offset_yIntialized = true;
    this._refObject.offset_y = value;
  }
  get offset_z() {
    if (!this._offset_zIntialized) {
      return undefined;
    }
    return this._refObject.offset_z;
  }
  set offset_z(value) {
    this._offset_zIntialized = true;
    this._refObject.offset_z = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SensorUwbRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._sessionidIntialized = true;
    this._time_offsetIntialized = true;
    this._counterIntialized = true;
    this._macIntialized = true;
    this._mac_destIntialized = true;
    this._statusIntialized = true;
    this._nlosIntialized = true;
    this._distanceIntialized = true;
    this._aoa_azimuth_devIntialized = true;
    this._aoa_elevation_devIntialized = true;
    this._aoa_azimuth_respIntialized = true;
    this._aoa_elevation_respIntialized = true;
    this._aoa_azimuth_fomIntialized = true;
    this._aoa_elevation_fomIntialized = true;
    this._aoa_dest_azimuth_fomIntialized = true;
    this._aoa_dest_elevation_fomIntialized = true;
    this._orientationIntialized = true;
    this._offset_xIntialized = true;
    this._offset_yIntialized = true;
    this._offset_zIntialized = true;
  }
  copy(other) {
    this._refObject = new SensorUwbRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._sessionidIntialized = true;
    this._time_offsetIntialized = true;
    this._counterIntialized = true;
    this._macIntialized = true;
    this._mac_destIntialized = true;
    this._statusIntialized = true;
    this._nlosIntialized = true;
    this._distanceIntialized = true;
    this._aoa_azimuth_devIntialized = true;
    this._aoa_elevation_devIntialized = true;
    this._aoa_azimuth_respIntialized = true;
    this._aoa_elevation_respIntialized = true;
    this._aoa_azimuth_fomIntialized = true;
    this._aoa_elevation_fomIntialized = true;
    this._aoa_dest_azimuth_fomIntialized = true;
    this._aoa_dest_elevation_fomIntialized = true;
    this._orientationIntialized = true;
    this._offset_xIntialized = true;
    this._offset_yIntialized = true;
    this._offset_zIntialized = true;
  }
  static get classType() {
    return SensorUwbWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sessionid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mac","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mac_dest","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nlos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aoa_azimuth_dev","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aoa_elevation_dev","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aoa_azimuth_resp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aoa_elevation_resp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aoa_azimuth_fom","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aoa_elevation_fom","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aoa_dest_azimuth_fom","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aoa_dest_elevation_fom","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"orientation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"offset_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"offset_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"offset_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"SensorUwb","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SensorUwb"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","sessionid","time_offset","counter","mac","mac_dest","status","nlos","distance","aoa_azimuth_dev","aoa_elevation_dev","aoa_azimuth_resp","aoa_elevation_resp","aoa_azimuth_fom","aoa_elevation_fom","aoa_dest_azimuth_fom","aoa_dest_elevation_fom","orientation","offset_x","offset_y","offset_z"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SensorUwbArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SensorUwbWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SensorUwbWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorUwbArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorUwbArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SensorUwbRefArray(size);
    this._refObject = new SensorUwbRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SensorUwbWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SensorUwbArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SensorUwbArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SensorUwbWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SensorUwbWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SensorUwbArrayWrapper;
  }
}
module.exports = SensorUwbWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sessionid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mac",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mac_dest",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nlos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aoa_azimuth_dev",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aoa_elevation_dev",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aoa_azimuth_resp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aoa_elevation_resp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aoa_azimuth_fom",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aoa_elevation_fom",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aoa_dest_azimuth_fom",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aoa_dest_elevation_fom",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "orientation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "offset_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "offset_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "offset_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "SensorUwb",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SensorUwb"
}
*/
