// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleAttitudeSetpoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VehicleAttitudeSetpointRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  roll_body: primitiveTypes.float32,
  pitch_body: primitiveTypes.float32,
  yaw_body: primitiveTypes.float32,
  yaw_sp_move_rate: primitiveTypes.float32,
  q_d: ArrayType(primitiveTypes.float32, 4),
  thrust_body: ArrayType(primitiveTypes.float32, 3),
  reset_integral: primitiveTypes.bool,
  fw_control_yaw_wheel: primitiveTypes.bool,
});
const VehicleAttitudeSetpointRefArray = ArrayType(VehicleAttitudeSetpointRefStruct);
const VehicleAttitudeSetpointRefStructArray = StructType({
  data: VehicleAttitudeSetpointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleAttitudeSetpointWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleAttitudeSetpointRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._roll_bodyIntialized = true;
      this._pitch_bodyIntialized = true;
      this._yaw_bodyIntialized = true;
      this._yaw_sp_move_rateIntialized = true;
      this._wrapperFields.q_d = Float32Wrapper.createArray();
      this._wrapperFields.q_d.copy(other._wrapperFields.q_d);
      this._wrapperFields.thrust_body = Float32Wrapper.createArray();
      this._wrapperFields.thrust_body.copy(other._wrapperFields.thrust_body);
      this._reset_integralIntialized = true;
      this._fw_control_yaw_wheelIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleAttitudeSetpointRefStruct();
    this._timestampIntialized = false;
    this._roll_bodyIntialized = false;
    this._pitch_bodyIntialized = false;
    this._yaw_bodyIntialized = false;
    this._yaw_sp_move_rateIntialized = false;
    this._wrapperFields.q_d = Float32Wrapper.createArray();
    this._wrapperFields.thrust_body = Float32Wrapper.createArray();
    this._reset_integralIntialized = false;
    this._fw_control_yaw_wheelIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleAttitudeSetpointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleAttitudeSetpointArrayWrapper;
  }
  static get ArrayType() {
    return VehicleAttitudeSetpointArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleAttitudeSetpointRefStructArray
  }
  static get refObjectType() {
    return VehicleAttitudeSetpointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleAttitudeSetpoint');
      }
      if (checkConsistency && !this._roll_bodyIntialized) {
        throw new TypeError('Invalid argument: roll_body in VehicleAttitudeSetpoint');
      }
      if (checkConsistency && !this._pitch_bodyIntialized) {
        throw new TypeError('Invalid argument: pitch_body in VehicleAttitudeSetpoint');
      }
      if (checkConsistency && !this._yaw_bodyIntialized) {
        throw new TypeError('Invalid argument: yaw_body in VehicleAttitudeSetpoint');
      }
      if (checkConsistency && !this._yaw_sp_move_rateIntialized) {
        throw new TypeError('Invalid argument: yaw_sp_move_rate in VehicleAttitudeSetpoint');
      }
      if (checkConsistency && !this._reset_integralIntialized) {
        throw new TypeError('Invalid argument: reset_integral in VehicleAttitudeSetpoint');
      }
      if (checkConsistency && !this._fw_control_yaw_wheelIntialized) {
        throw new TypeError('Invalid argument: fw_control_yaw_wheel in VehicleAttitudeSetpoint');
      }
    this._refObject.q_d = Array.from(this._wrapperFields.q_d.data);
    this._refObject.thrust_body = Array.from(this._wrapperFields.thrust_body.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._roll_bodyIntialized = true;
    this._refObject.roll_body = refObject.roll_body;
    this._pitch_bodyIntialized = true;
    this._refObject.pitch_body = refObject.pitch_body;
    this._yaw_bodyIntialized = true;
    this._refObject.yaw_body = refObject.yaw_body;
    this._yaw_sp_move_rateIntialized = true;
    this._refObject.yaw_sp_move_rate = refObject.yaw_sp_move_rate;
    this._wrapperFields.q_d.fill(refObject.q_d.toArray());
    this._wrapperFields.thrust_body.fill(refObject.thrust_body.toArray());
    this._reset_integralIntialized = true;
    this._refObject.reset_integral = refObject.reset_integral;
    this._fw_control_yaw_wheelIntialized = true;
    this._refObject.fw_control_yaw_wheel = refObject.fw_control_yaw_wheel;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleAttitudeSetpointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleAttitudeSetpoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get roll_body() {
    if (!this._roll_bodyIntialized) {
      return undefined;
    }
    return this._refObject.roll_body;
  }
  set roll_body(value) {
    this._roll_bodyIntialized = true;
    this._refObject.roll_body = value;
  }
  get pitch_body() {
    if (!this._pitch_bodyIntialized) {
      return undefined;
    }
    return this._refObject.pitch_body;
  }
  set pitch_body(value) {
    this._pitch_bodyIntialized = true;
    this._refObject.pitch_body = value;
  }
  get yaw_body() {
    if (!this._yaw_bodyIntialized) {
      return undefined;
    }
    return this._refObject.yaw_body;
  }
  set yaw_body(value) {
    this._yaw_bodyIntialized = true;
    this._refObject.yaw_body = value;
  }
  get yaw_sp_move_rate() {
    if (!this._yaw_sp_move_rateIntialized) {
      return undefined;
    }
    return this._refObject.yaw_sp_move_rate;
  }
  set yaw_sp_move_rate(value) {
    this._yaw_sp_move_rateIntialized = true;
    this._refObject.yaw_sp_move_rate = value;
  }
  get q_d() {
    return this._wrapperFields['q_d'].data;
  }
  set q_d(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields['q_d'].fill(value);
  }
  get thrust_body() {
    return this._wrapperFields['thrust_body'].data;
  }
  set thrust_body(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['thrust_body'].fill(value);
  }
  get reset_integral() {
    if (!this._reset_integralIntialized) {
      return undefined;
    }
    return this._refObject.reset_integral;
  }
  set reset_integral(value) {
    this._reset_integralIntialized = true;
    this._refObject.reset_integral = value;
  }
  get fw_control_yaw_wheel() {
    if (!this._fw_control_yaw_wheelIntialized) {
      return undefined;
    }
    return this._refObject.fw_control_yaw_wheel;
  }
  set fw_control_yaw_wheel(value) {
    this._fw_control_yaw_wheelIntialized = true;
    this._refObject.fw_control_yaw_wheel = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleAttitudeSetpointRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._roll_bodyIntialized = true;
    this._pitch_bodyIntialized = true;
    this._yaw_bodyIntialized = true;
    this._yaw_sp_move_rateIntialized = true;
    this._wrapperFields.q_d.fill(refObject.q_d.toArray());
    this._wrapperFields.thrust_body.fill(refObject.thrust_body.toArray());
    this._reset_integralIntialized = true;
    this._fw_control_yaw_wheelIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleAttitudeSetpointRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._roll_bodyIntialized = true;
    this._pitch_bodyIntialized = true;
    this._yaw_bodyIntialized = true;
    this._yaw_sp_move_rateIntialized = true;
    this._wrapperFields.q_d.copy(other._wrapperFields.q_d);
    this._wrapperFields.thrust_body.copy(other._wrapperFields.thrust_body);
    this._reset_integralIntialized = true;
    this._fw_control_yaw_wheelIntialized = true;
  }
  static get classType() {
    return VehicleAttitudeSetpointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"roll_body","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch_body","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_body","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_sp_move_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"q_d","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"thrust_body","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reset_integral","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fw_control_yaw_wheel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleAttitudeSetpoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleAttitudeSetpoint"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","roll_body","pitch_body","yaw_body","yaw_sp_move_rate","q_d","thrust_body","reset_integral","fw_control_yaw_wheel"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleAttitudeSetpointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleAttitudeSetpointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleAttitudeSetpointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleAttitudeSetpointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleAttitudeSetpointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleAttitudeSetpointRefArray(size);
    this._refObject = new VehicleAttitudeSetpointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleAttitudeSetpointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleAttitudeSetpointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleAttitudeSetpointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleAttitudeSetpointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleAttitudeSetpointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleAttitudeSetpointArrayWrapper;
  }
}
module.exports = VehicleAttitudeSetpointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "roll_body",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch_body",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_body",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_sp_move_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "q_d",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "thrust_body",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reset_integral",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fw_control_yaw_wheel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleAttitudeSetpoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleAttitudeSetpoint"
}
*/
