// rclnodejs/generated/px4_msgs/px4_msgs__msg__OrbTestMedium.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const OrbTestMediumRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  val: primitiveTypes.int32,
  junk: ArrayType(primitiveTypes.uint8, 64),
});
const OrbTestMediumRefArray = ArrayType(OrbTestMediumRefStruct);
const OrbTestMediumRefStructArray = StructType({
  data: OrbTestMediumRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OrbTestMediumWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OrbTestMediumRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._valIntialized = true;
      this._wrapperFields.junk = UInt8Wrapper.createArray();
      this._wrapperFields.junk.copy(other._wrapperFields.junk);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OrbTestMediumRefStruct();
    this._timestampIntialized = false;
    this._valIntialized = false;
    this._wrapperFields.junk = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new OrbTestMediumWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OrbTestMediumArrayWrapper;
  }
  static get ArrayType() {
    return OrbTestMediumArrayWrapper;
  }
  static get refObjectArrayType() {
    return OrbTestMediumRefStructArray
  }
  static get refObjectType() {
    return OrbTestMediumRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in OrbTestMedium');
      }
      if (checkConsistency && !this._valIntialized) {
        throw new TypeError('Invalid argument: val in OrbTestMedium');
      }
    this._refObject.junk = Array.from(this._wrapperFields.junk.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._valIntialized = true;
    this._refObject.val = refObject.val;
    this._wrapperFields.junk.fill(refObject.junk.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    OrbTestMediumWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'OrbTestMedium'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get val() {
    if (!this._valIntialized) {
      return undefined;
    }
    return this._refObject.val;
  }
  set val(value) {
    this._valIntialized = true;
    this._refObject.val = value;
  }
  get junk() {
    return this._wrapperFields['junk'].data;
  }
  set junk(value) {
    if (value.length !== 64) {
      throw new RangeError('The length of the array must be 64.');
    }
    this._wrapperFields['junk'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new OrbTestMediumRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._valIntialized = true;
    this._wrapperFields.junk.fill(refObject.junk.toArray());
  }
  copy(other) {
    this._refObject = new OrbTestMediumRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._valIntialized = true;
    this._wrapperFields.junk.copy(other._wrapperFields.junk);
  }
  static get classType() {
    return OrbTestMediumWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"val","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"junk","type":{"isArray":true,"arraySize":64,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"OrbTestMedium","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OrbTestMedium"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","val","junk"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OrbTestMediumArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OrbTestMediumWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OrbTestMediumWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OrbTestMediumArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OrbTestMediumArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OrbTestMediumRefArray(size);
    this._refObject = new OrbTestMediumRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OrbTestMediumWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OrbTestMediumArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OrbTestMediumArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OrbTestMediumWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OrbTestMediumWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OrbTestMediumArrayWrapper;
  }
}
module.exports = OrbTestMediumWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "val",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "junk",
      "type": {
        "isArray": true,
        "arraySize": 64,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "OrbTestMedium",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OrbTestMedium"
}
*/
