// rclnodejs/generated/px4_msgs/px4_msgs__msg__SensorGyroFft.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const SensorGyroFftRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  device_id: primitiveTypes.uint32,
  sensor_sample_rate_hz: primitiveTypes.float32,
  resolution_hz: primitiveTypes.float32,
  peak_frequencies_x: ArrayType(primitiveTypes.float32, 3),
  peak_frequencies_y: ArrayType(primitiveTypes.float32, 3),
  peak_frequencies_z: ArrayType(primitiveTypes.float32, 3),
  peak_snr_x: ArrayType(primitiveTypes.float32, 3),
  peak_snr_y: ArrayType(primitiveTypes.float32, 3),
  peak_snr_z: ArrayType(primitiveTypes.float32, 3),
});
const SensorGyroFftRefArray = ArrayType(SensorGyroFftRefStruct);
const SensorGyroFftRefStructArray = StructType({
  data: SensorGyroFftRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SensorGyroFftWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SensorGyroFftRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._device_idIntialized = true;
      this._sensor_sample_rate_hzIntialized = true;
      this._resolution_hzIntialized = true;
      this._wrapperFields.peak_frequencies_x = Float32Wrapper.createArray();
      this._wrapperFields.peak_frequencies_x.copy(other._wrapperFields.peak_frequencies_x);
      this._wrapperFields.peak_frequencies_y = Float32Wrapper.createArray();
      this._wrapperFields.peak_frequencies_y.copy(other._wrapperFields.peak_frequencies_y);
      this._wrapperFields.peak_frequencies_z = Float32Wrapper.createArray();
      this._wrapperFields.peak_frequencies_z.copy(other._wrapperFields.peak_frequencies_z);
      this._wrapperFields.peak_snr_x = Float32Wrapper.createArray();
      this._wrapperFields.peak_snr_x.copy(other._wrapperFields.peak_snr_x);
      this._wrapperFields.peak_snr_y = Float32Wrapper.createArray();
      this._wrapperFields.peak_snr_y.copy(other._wrapperFields.peak_snr_y);
      this._wrapperFields.peak_snr_z = Float32Wrapper.createArray();
      this._wrapperFields.peak_snr_z.copy(other._wrapperFields.peak_snr_z);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SensorGyroFftRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._device_idIntialized = false;
    this._sensor_sample_rate_hzIntialized = false;
    this._resolution_hzIntialized = false;
    this._wrapperFields.peak_frequencies_x = Float32Wrapper.createArray();
    this._wrapperFields.peak_frequencies_y = Float32Wrapper.createArray();
    this._wrapperFields.peak_frequencies_z = Float32Wrapper.createArray();
    this._wrapperFields.peak_snr_x = Float32Wrapper.createArray();
    this._wrapperFields.peak_snr_y = Float32Wrapper.createArray();
    this._wrapperFields.peak_snr_z = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SensorGyroFftWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SensorGyroFftArrayWrapper;
  }
  static get ArrayType() {
    return SensorGyroFftArrayWrapper;
  }
  static get refObjectArrayType() {
    return SensorGyroFftRefStructArray
  }
  static get refObjectType() {
    return SensorGyroFftRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in SensorGyroFft');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in SensorGyroFft');
      }
      if (checkConsistency && !this._device_idIntialized) {
        throw new TypeError('Invalid argument: device_id in SensorGyroFft');
      }
      if (checkConsistency && !this._sensor_sample_rate_hzIntialized) {
        throw new TypeError('Invalid argument: sensor_sample_rate_hz in SensorGyroFft');
      }
      if (checkConsistency && !this._resolution_hzIntialized) {
        throw new TypeError('Invalid argument: resolution_hz in SensorGyroFft');
      }
    this._refObject.peak_frequencies_x = Array.from(this._wrapperFields.peak_frequencies_x.data);
    this._refObject.peak_frequencies_y = Array.from(this._wrapperFields.peak_frequencies_y.data);
    this._refObject.peak_frequencies_z = Array.from(this._wrapperFields.peak_frequencies_z.data);
    this._refObject.peak_snr_x = Array.from(this._wrapperFields.peak_snr_x.data);
    this._refObject.peak_snr_y = Array.from(this._wrapperFields.peak_snr_y.data);
    this._refObject.peak_snr_z = Array.from(this._wrapperFields.peak_snr_z.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._device_idIntialized = true;
    this._refObject.device_id = refObject.device_id;
    this._sensor_sample_rate_hzIntialized = true;
    this._refObject.sensor_sample_rate_hz = refObject.sensor_sample_rate_hz;
    this._resolution_hzIntialized = true;
    this._refObject.resolution_hz = refObject.resolution_hz;
    this._wrapperFields.peak_frequencies_x.fill(refObject.peak_frequencies_x.toArray());
    this._wrapperFields.peak_frequencies_y.fill(refObject.peak_frequencies_y.toArray());
    this._wrapperFields.peak_frequencies_z.fill(refObject.peak_frequencies_z.toArray());
    this._wrapperFields.peak_snr_x.fill(refObject.peak_snr_x.toArray());
    this._wrapperFields.peak_snr_y.fill(refObject.peak_snr_y.toArray());
    this._wrapperFields.peak_snr_z.fill(refObject.peak_snr_z.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SensorGyroFftWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'SensorGyroFft'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get device_id() {
    if (!this._device_idIntialized) {
      return undefined;
    }
    return this._refObject.device_id;
  }
  set device_id(value) {
    this._device_idIntialized = true;
    this._refObject.device_id = value;
  }
  get sensor_sample_rate_hz() {
    if (!this._sensor_sample_rate_hzIntialized) {
      return undefined;
    }
    return this._refObject.sensor_sample_rate_hz;
  }
  set sensor_sample_rate_hz(value) {
    this._sensor_sample_rate_hzIntialized = true;
    this._refObject.sensor_sample_rate_hz = value;
  }
  get resolution_hz() {
    if (!this._resolution_hzIntialized) {
      return undefined;
    }
    return this._refObject.resolution_hz;
  }
  set resolution_hz(value) {
    this._resolution_hzIntialized = true;
    this._refObject.resolution_hz = value;
  }
  get peak_frequencies_x() {
    return this._wrapperFields['peak_frequencies_x'].data;
  }
  set peak_frequencies_x(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['peak_frequencies_x'].fill(value);
  }
  get peak_frequencies_y() {
    return this._wrapperFields['peak_frequencies_y'].data;
  }
  set peak_frequencies_y(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['peak_frequencies_y'].fill(value);
  }
  get peak_frequencies_z() {
    return this._wrapperFields['peak_frequencies_z'].data;
  }
  set peak_frequencies_z(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['peak_frequencies_z'].fill(value);
  }
  get peak_snr_x() {
    return this._wrapperFields['peak_snr_x'].data;
  }
  set peak_snr_x(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['peak_snr_x'].fill(value);
  }
  get peak_snr_y() {
    return this._wrapperFields['peak_snr_y'].data;
  }
  set peak_snr_y(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['peak_snr_y'].fill(value);
  }
  get peak_snr_z() {
    return this._wrapperFields['peak_snr_z'].data;
  }
  set peak_snr_z(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['peak_snr_z'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SensorGyroFftRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._device_idIntialized = true;
    this._sensor_sample_rate_hzIntialized = true;
    this._resolution_hzIntialized = true;
    this._wrapperFields.peak_frequencies_x.fill(refObject.peak_frequencies_x.toArray());
    this._wrapperFields.peak_frequencies_y.fill(refObject.peak_frequencies_y.toArray());
    this._wrapperFields.peak_frequencies_z.fill(refObject.peak_frequencies_z.toArray());
    this._wrapperFields.peak_snr_x.fill(refObject.peak_snr_x.toArray());
    this._wrapperFields.peak_snr_y.fill(refObject.peak_snr_y.toArray());
    this._wrapperFields.peak_snr_z.fill(refObject.peak_snr_z.toArray());
  }
  copy(other) {
    this._refObject = new SensorGyroFftRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._device_idIntialized = true;
    this._sensor_sample_rate_hzIntialized = true;
    this._resolution_hzIntialized = true;
    this._wrapperFields.peak_frequencies_x.copy(other._wrapperFields.peak_frequencies_x);
    this._wrapperFields.peak_frequencies_y.copy(other._wrapperFields.peak_frequencies_y);
    this._wrapperFields.peak_frequencies_z.copy(other._wrapperFields.peak_frequencies_z);
    this._wrapperFields.peak_snr_x.copy(other._wrapperFields.peak_snr_x);
    this._wrapperFields.peak_snr_y.copy(other._wrapperFields.peak_snr_y);
    this._wrapperFields.peak_snr_z.copy(other._wrapperFields.peak_snr_z);
  }
  static get classType() {
    return SensorGyroFftWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sensor_sample_rate_hz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"resolution_hz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"peak_frequencies_x","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"peak_frequencies_y","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"peak_frequencies_z","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"peak_snr_x","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"peak_snr_y","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"peak_snr_z","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"SensorGyroFft","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SensorGyroFft"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","device_id","sensor_sample_rate_hz","resolution_hz","peak_frequencies_x","peak_frequencies_y","peak_frequencies_z","peak_snr_x","peak_snr_y","peak_snr_z"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SensorGyroFftArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SensorGyroFftWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SensorGyroFftWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorGyroFftArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorGyroFftArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SensorGyroFftRefArray(size);
    this._refObject = new SensorGyroFftRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SensorGyroFftWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SensorGyroFftArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SensorGyroFftArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SensorGyroFftWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SensorGyroFftWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SensorGyroFftArrayWrapper;
  }
}
module.exports = SensorGyroFftWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sensor_sample_rate_hz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "resolution_hz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "peak_frequencies_x",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "peak_frequencies_y",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "peak_frequencies_z",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "peak_snr_x",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "peak_snr_y",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "peak_snr_z",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "SensorGyroFft",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SensorGyroFft"
}
*/
