// rclnodejs/generated/px4_msgs/px4_msgs__msg__GpsInjectData.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const GpsInjectDataRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  device_id: primitiveTypes.uint32,
  len: primitiveTypes.uint16,
  flags: primitiveTypes.uint8,
  data: ArrayType(primitiveTypes.uint8, 300),
});
const GpsInjectDataRefArray = ArrayType(GpsInjectDataRefStruct);
const GpsInjectDataRefStructArray = StructType({
  data: GpsInjectDataRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GpsInjectDataWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GpsInjectDataRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._device_idIntialized = true;
      this._lenIntialized = true;
      this._flagsIntialized = true;
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GpsInjectDataRefStruct();
    this._timestampIntialized = false;
    this._device_idIntialized = false;
    this._lenIntialized = false;
    this._flagsIntialized = false;
    this._wrapperFields.data = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GpsInjectDataWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GpsInjectDataArrayWrapper;
  }
  static get ArrayType() {
    return GpsInjectDataArrayWrapper;
  }
  static get refObjectArrayType() {
    return GpsInjectDataRefStructArray
  }
  static get refObjectType() {
    return GpsInjectDataRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in GpsInjectData');
      }
      if (checkConsistency && !this._device_idIntialized) {
        throw new TypeError('Invalid argument: device_id in GpsInjectData');
      }
      if (checkConsistency && !this._lenIntialized) {
        throw new TypeError('Invalid argument: len in GpsInjectData');
      }
      if (checkConsistency && !this._flagsIntialized) {
        throw new TypeError('Invalid argument: flags in GpsInjectData');
      }
    this._refObject.data = Array.from(this._wrapperFields.data.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._device_idIntialized = true;
    this._refObject.device_id = refObject.device_id;
    this._lenIntialized = true;
    this._refObject.len = refObject.len;
    this._flagsIntialized = true;
    this._refObject.flags = refObject.flags;
    this._wrapperFields.data.fill(refObject.data.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GpsInjectDataWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'GpsInjectData'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get device_id() {
    if (!this._device_idIntialized) {
      return undefined;
    }
    return this._refObject.device_id;
  }
  set device_id(value) {
    this._device_idIntialized = true;
    this._refObject.device_id = value;
  }
  get len() {
    if (!this._lenIntialized) {
      return undefined;
    }
    return this._refObject.len;
  }
  set len(value) {
    this._lenIntialized = true;
    this._refObject.len = value;
  }
  get flags() {
    if (!this._flagsIntialized) {
      return undefined;
    }
    return this._refObject.flags;
  }
  set flags(value) {
    this._flagsIntialized = true;
    this._refObject.flags = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    if (value.length !== 300) {
      throw new RangeError('The length of the array must be 300.');
    }
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GpsInjectDataRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._device_idIntialized = true;
    this._lenIntialized = true;
    this._flagsIntialized = true;
    this._wrapperFields.data.fill(refObject.data.toArray());
  }
  copy(other) {
    this._refObject = new GpsInjectDataRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._device_idIntialized = true;
    this._lenIntialized = true;
    this._flagsIntialized = true;
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return GpsInjectDataWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ORB_QUEUE_LENGTH","value":8},{"type":"uint8","name":"MAX_INSTANCES","value":2}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"len","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":300,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"GpsInjectData","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GpsInjectData"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","device_id","len","flags","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GpsInjectDataArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GpsInjectDataWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GpsInjectDataWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GpsInjectDataArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GpsInjectDataArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GpsInjectDataRefArray(size);
    this._refObject = new GpsInjectDataRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GpsInjectDataWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GpsInjectDataArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GpsInjectDataArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GpsInjectDataWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GpsInjectDataWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GpsInjectDataArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(GpsInjectDataWrapper, "ORB_QUEUE_LENGTH", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GpsInjectDataWrapper, "MAX_INSTANCES", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = GpsInjectDataWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ORB_QUEUE_LENGTH",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "MAX_INSTANCES",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "len",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": 300,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "GpsInjectData",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GpsInjectData"
}
*/
