// rclnodejs/generated/px4_msgs/px4_msgs__msg__AirspeedWind.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const AirspeedWindRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  windspeed_north: primitiveTypes.float32,
  windspeed_east: primitiveTypes.float32,
  variance_north: primitiveTypes.float32,
  variance_east: primitiveTypes.float32,
  tas_innov: primitiveTypes.float32,
  tas_innov_var: primitiveTypes.float32,
  tas_scale_raw: primitiveTypes.float32,
  tas_scale_raw_var: primitiveTypes.float32,
  tas_scale_validated: primitiveTypes.float32,
  beta_innov: primitiveTypes.float32,
  beta_innov_var: primitiveTypes.float32,
  source: primitiveTypes.uint8,
});
const AirspeedWindRefArray = ArrayType(AirspeedWindRefStruct);
const AirspeedWindRefStructArray = StructType({
  data: AirspeedWindRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AirspeedWindWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AirspeedWindRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._windspeed_northIntialized = true;
      this._windspeed_eastIntialized = true;
      this._variance_northIntialized = true;
      this._variance_eastIntialized = true;
      this._tas_innovIntialized = true;
      this._tas_innov_varIntialized = true;
      this._tas_scale_rawIntialized = true;
      this._tas_scale_raw_varIntialized = true;
      this._tas_scale_validatedIntialized = true;
      this._beta_innovIntialized = true;
      this._beta_innov_varIntialized = true;
      this._sourceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AirspeedWindRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._windspeed_northIntialized = false;
    this._windspeed_eastIntialized = false;
    this._variance_northIntialized = false;
    this._variance_eastIntialized = false;
    this._tas_innovIntialized = false;
    this._tas_innov_varIntialized = false;
    this._tas_scale_rawIntialized = false;
    this._tas_scale_raw_varIntialized = false;
    this._tas_scale_validatedIntialized = false;
    this._beta_innovIntialized = false;
    this._beta_innov_varIntialized = false;
    this._sourceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new AirspeedWindWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AirspeedWindArrayWrapper;
  }
  static get ArrayType() {
    return AirspeedWindArrayWrapper;
  }
  static get refObjectArrayType() {
    return AirspeedWindRefStructArray
  }
  static get refObjectType() {
    return AirspeedWindRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in AirspeedWind');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in AirspeedWind');
      }
      if (checkConsistency && !this._windspeed_northIntialized) {
        throw new TypeError('Invalid argument: windspeed_north in AirspeedWind');
      }
      if (checkConsistency && !this._windspeed_eastIntialized) {
        throw new TypeError('Invalid argument: windspeed_east in AirspeedWind');
      }
      if (checkConsistency && !this._variance_northIntialized) {
        throw new TypeError('Invalid argument: variance_north in AirspeedWind');
      }
      if (checkConsistency && !this._variance_eastIntialized) {
        throw new TypeError('Invalid argument: variance_east in AirspeedWind');
      }
      if (checkConsistency && !this._tas_innovIntialized) {
        throw new TypeError('Invalid argument: tas_innov in AirspeedWind');
      }
      if (checkConsistency && !this._tas_innov_varIntialized) {
        throw new TypeError('Invalid argument: tas_innov_var in AirspeedWind');
      }
      if (checkConsistency && !this._tas_scale_rawIntialized) {
        throw new TypeError('Invalid argument: tas_scale_raw in AirspeedWind');
      }
      if (checkConsistency && !this._tas_scale_raw_varIntialized) {
        throw new TypeError('Invalid argument: tas_scale_raw_var in AirspeedWind');
      }
      if (checkConsistency && !this._tas_scale_validatedIntialized) {
        throw new TypeError('Invalid argument: tas_scale_validated in AirspeedWind');
      }
      if (checkConsistency && !this._beta_innovIntialized) {
        throw new TypeError('Invalid argument: beta_innov in AirspeedWind');
      }
      if (checkConsistency && !this._beta_innov_varIntialized) {
        throw new TypeError('Invalid argument: beta_innov_var in AirspeedWind');
      }
      if (checkConsistency && !this._sourceIntialized) {
        throw new TypeError('Invalid argument: source in AirspeedWind');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._windspeed_northIntialized = true;
    this._refObject.windspeed_north = refObject.windspeed_north;
    this._windspeed_eastIntialized = true;
    this._refObject.windspeed_east = refObject.windspeed_east;
    this._variance_northIntialized = true;
    this._refObject.variance_north = refObject.variance_north;
    this._variance_eastIntialized = true;
    this._refObject.variance_east = refObject.variance_east;
    this._tas_innovIntialized = true;
    this._refObject.tas_innov = refObject.tas_innov;
    this._tas_innov_varIntialized = true;
    this._refObject.tas_innov_var = refObject.tas_innov_var;
    this._tas_scale_rawIntialized = true;
    this._refObject.tas_scale_raw = refObject.tas_scale_raw;
    this._tas_scale_raw_varIntialized = true;
    this._refObject.tas_scale_raw_var = refObject.tas_scale_raw_var;
    this._tas_scale_validatedIntialized = true;
    this._refObject.tas_scale_validated = refObject.tas_scale_validated;
    this._beta_innovIntialized = true;
    this._refObject.beta_innov = refObject.beta_innov;
    this._beta_innov_varIntialized = true;
    this._refObject.beta_innov_var = refObject.beta_innov_var;
    this._sourceIntialized = true;
    this._refObject.source = refObject.source;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    AirspeedWindWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'AirspeedWind'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get windspeed_north() {
    if (!this._windspeed_northIntialized) {
      return undefined;
    }
    return this._refObject.windspeed_north;
  }
  set windspeed_north(value) {
    this._windspeed_northIntialized = true;
    this._refObject.windspeed_north = value;
  }
  get windspeed_east() {
    if (!this._windspeed_eastIntialized) {
      return undefined;
    }
    return this._refObject.windspeed_east;
  }
  set windspeed_east(value) {
    this._windspeed_eastIntialized = true;
    this._refObject.windspeed_east = value;
  }
  get variance_north() {
    if (!this._variance_northIntialized) {
      return undefined;
    }
    return this._refObject.variance_north;
  }
  set variance_north(value) {
    this._variance_northIntialized = true;
    this._refObject.variance_north = value;
  }
  get variance_east() {
    if (!this._variance_eastIntialized) {
      return undefined;
    }
    return this._refObject.variance_east;
  }
  set variance_east(value) {
    this._variance_eastIntialized = true;
    this._refObject.variance_east = value;
  }
  get tas_innov() {
    if (!this._tas_innovIntialized) {
      return undefined;
    }
    return this._refObject.tas_innov;
  }
  set tas_innov(value) {
    this._tas_innovIntialized = true;
    this._refObject.tas_innov = value;
  }
  get tas_innov_var() {
    if (!this._tas_innov_varIntialized) {
      return undefined;
    }
    return this._refObject.tas_innov_var;
  }
  set tas_innov_var(value) {
    this._tas_innov_varIntialized = true;
    this._refObject.tas_innov_var = value;
  }
  get tas_scale_raw() {
    if (!this._tas_scale_rawIntialized) {
      return undefined;
    }
    return this._refObject.tas_scale_raw;
  }
  set tas_scale_raw(value) {
    this._tas_scale_rawIntialized = true;
    this._refObject.tas_scale_raw = value;
  }
  get tas_scale_raw_var() {
    if (!this._tas_scale_raw_varIntialized) {
      return undefined;
    }
    return this._refObject.tas_scale_raw_var;
  }
  set tas_scale_raw_var(value) {
    this._tas_scale_raw_varIntialized = true;
    this._refObject.tas_scale_raw_var = value;
  }
  get tas_scale_validated() {
    if (!this._tas_scale_validatedIntialized) {
      return undefined;
    }
    return this._refObject.tas_scale_validated;
  }
  set tas_scale_validated(value) {
    this._tas_scale_validatedIntialized = true;
    this._refObject.tas_scale_validated = value;
  }
  get beta_innov() {
    if (!this._beta_innovIntialized) {
      return undefined;
    }
    return this._refObject.beta_innov;
  }
  set beta_innov(value) {
    this._beta_innovIntialized = true;
    this._refObject.beta_innov = value;
  }
  get beta_innov_var() {
    if (!this._beta_innov_varIntialized) {
      return undefined;
    }
    return this._refObject.beta_innov_var;
  }
  set beta_innov_var(value) {
    this._beta_innov_varIntialized = true;
    this._refObject.beta_innov_var = value;
  }
  get source() {
    if (!this._sourceIntialized) {
      return undefined;
    }
    return this._refObject.source;
  }
  set source(value) {
    this._sourceIntialized = true;
    this._refObject.source = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AirspeedWindRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._windspeed_northIntialized = true;
    this._windspeed_eastIntialized = true;
    this._variance_northIntialized = true;
    this._variance_eastIntialized = true;
    this._tas_innovIntialized = true;
    this._tas_innov_varIntialized = true;
    this._tas_scale_rawIntialized = true;
    this._tas_scale_raw_varIntialized = true;
    this._tas_scale_validatedIntialized = true;
    this._beta_innovIntialized = true;
    this._beta_innov_varIntialized = true;
    this._sourceIntialized = true;
  }
  copy(other) {
    this._refObject = new AirspeedWindRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._windspeed_northIntialized = true;
    this._windspeed_eastIntialized = true;
    this._variance_northIntialized = true;
    this._variance_eastIntialized = true;
    this._tas_innovIntialized = true;
    this._tas_innov_varIntialized = true;
    this._tas_scale_rawIntialized = true;
    this._tas_scale_raw_varIntialized = true;
    this._tas_scale_validatedIntialized = true;
    this._beta_innovIntialized = true;
    this._beta_innov_varIntialized = true;
    this._sourceIntialized = true;
  }
  static get classType() {
    return AirspeedWindWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"SOURCE_AS_BETA_ONLY","value":0},{"type":"uint8","name":"SOURCE_AS_SENSOR_1","value":1},{"type":"uint8","name":"SOURCE_AS_SENSOR_2","value":2},{"type":"uint8","name":"SOURCE_AS_SENSOR_3","value":3}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"windspeed_north","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"windspeed_east","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"variance_north","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"variance_east","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tas_innov","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tas_innov_var","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tas_scale_raw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tas_scale_raw_var","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tas_scale_validated","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"beta_innov","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"beta_innov_var","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"source","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"AirspeedWind","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AirspeedWind"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","windspeed_north","windspeed_east","variance_north","variance_east","tas_innov","tas_innov_var","tas_scale_raw","tas_scale_raw_var","tas_scale_validated","beta_innov","beta_innov_var","source"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AirspeedWindArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AirspeedWindWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AirspeedWindWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AirspeedWindArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AirspeedWindArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AirspeedWindRefArray(size);
    this._refObject = new AirspeedWindRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AirspeedWindWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AirspeedWindArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AirspeedWindArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AirspeedWindWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AirspeedWindWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AirspeedWindArrayWrapper;
  }
}
// Define constants (4 in total)
Object.defineProperty(AirspeedWindWrapper, "SOURCE_AS_BETA_ONLY", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AirspeedWindWrapper, "SOURCE_AS_SENSOR_1", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AirspeedWindWrapper, "SOURCE_AS_SENSOR_2", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AirspeedWindWrapper, "SOURCE_AS_SENSOR_3", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = AirspeedWindWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "SOURCE_AS_BETA_ONLY",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "SOURCE_AS_SENSOR_1",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "SOURCE_AS_SENSOR_2",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "SOURCE_AS_SENSOR_3",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "windspeed_north",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "windspeed_east",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "variance_north",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "variance_east",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tas_innov",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tas_innov_var",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tas_scale_raw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tas_scale_raw_var",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tas_scale_validated",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "beta_innov",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "beta_innov_var",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "source",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "AirspeedWind",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "AirspeedWind"
}
*/
