// rclnodejs/generated/px4_msgs/px4_msgs__msg__SensorPreflightMag.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SensorPreflightMagRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  mag_inconsistency_angle: primitiveTypes.float32,
});
const SensorPreflightMagRefArray = ArrayType(SensorPreflightMagRefStruct);
const SensorPreflightMagRefStructArray = StructType({
  data: SensorPreflightMagRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SensorPreflightMagWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SensorPreflightMagRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._mag_inconsistency_angleIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SensorPreflightMagRefStruct();
    this._timestampIntialized = false;
    this._mag_inconsistency_angleIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SensorPreflightMagWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SensorPreflightMagArrayWrapper;
  }
  static get ArrayType() {
    return SensorPreflightMagArrayWrapper;
  }
  static get refObjectArrayType() {
    return SensorPreflightMagRefStructArray
  }
  static get refObjectType() {
    return SensorPreflightMagRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in SensorPreflightMag');
      }
      if (checkConsistency && !this._mag_inconsistency_angleIntialized) {
        throw new TypeError('Invalid argument: mag_inconsistency_angle in SensorPreflightMag');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._mag_inconsistency_angleIntialized = true;
    this._refObject.mag_inconsistency_angle = refObject.mag_inconsistency_angle;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SensorPreflightMagWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'SensorPreflightMag'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get mag_inconsistency_angle() {
    if (!this._mag_inconsistency_angleIntialized) {
      return undefined;
    }
    return this._refObject.mag_inconsistency_angle;
  }
  set mag_inconsistency_angle(value) {
    this._mag_inconsistency_angleIntialized = true;
    this._refObject.mag_inconsistency_angle = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SensorPreflightMagRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._mag_inconsistency_angleIntialized = true;
  }
  copy(other) {
    this._refObject = new SensorPreflightMagRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._mag_inconsistency_angleIntialized = true;
  }
  static get classType() {
    return SensorPreflightMagWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_inconsistency_angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"SensorPreflightMag","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SensorPreflightMag"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","mag_inconsistency_angle"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SensorPreflightMagArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SensorPreflightMagWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SensorPreflightMagWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorPreflightMagArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorPreflightMagArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SensorPreflightMagRefArray(size);
    this._refObject = new SensorPreflightMagRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SensorPreflightMagWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SensorPreflightMagArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SensorPreflightMagArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SensorPreflightMagWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SensorPreflightMagWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SensorPreflightMagArrayWrapper;
  }
}
module.exports = SensorPreflightMagWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_inconsistency_angle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "SensorPreflightMag",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SensorPreflightMag"
}
*/
