// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleLocalPositionSetpoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VehicleLocalPositionSetpointRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  x: primitiveTypes.float32,
  y: primitiveTypes.float32,
  z: primitiveTypes.float32,
  vx: primitiveTypes.float32,
  vy: primitiveTypes.float32,
  vz: primitiveTypes.float32,
  acceleration: ArrayType(primitiveTypes.float32, 3),
  thrust: ArrayType(primitiveTypes.float32, 3),
  yaw: primitiveTypes.float32,
  yawspeed: primitiveTypes.float32,
});
const VehicleLocalPositionSetpointRefArray = ArrayType(VehicleLocalPositionSetpointRefStruct);
const VehicleLocalPositionSetpointRefStructArray = StructType({
  data: VehicleLocalPositionSetpointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleLocalPositionSetpointWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleLocalPositionSetpointRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._xIntialized = true;
      this._yIntialized = true;
      this._zIntialized = true;
      this._vxIntialized = true;
      this._vyIntialized = true;
      this._vzIntialized = true;
      this._wrapperFields.acceleration = Float32Wrapper.createArray();
      this._wrapperFields.acceleration.copy(other._wrapperFields.acceleration);
      this._wrapperFields.thrust = Float32Wrapper.createArray();
      this._wrapperFields.thrust.copy(other._wrapperFields.thrust);
      this._yawIntialized = true;
      this._yawspeedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleLocalPositionSetpointRefStruct();
    this._timestampIntialized = false;
    this._xIntialized = false;
    this._yIntialized = false;
    this._zIntialized = false;
    this._vxIntialized = false;
    this._vyIntialized = false;
    this._vzIntialized = false;
    this._wrapperFields.acceleration = Float32Wrapper.createArray();
    this._wrapperFields.thrust = Float32Wrapper.createArray();
    this._yawIntialized = false;
    this._yawspeedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleLocalPositionSetpointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleLocalPositionSetpointArrayWrapper;
  }
  static get ArrayType() {
    return VehicleLocalPositionSetpointArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleLocalPositionSetpointRefStructArray
  }
  static get refObjectType() {
    return VehicleLocalPositionSetpointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleLocalPositionSetpoint');
      }
      if (checkConsistency && !this._xIntialized) {
        throw new TypeError('Invalid argument: x in VehicleLocalPositionSetpoint');
      }
      if (checkConsistency && !this._yIntialized) {
        throw new TypeError('Invalid argument: y in VehicleLocalPositionSetpoint');
      }
      if (checkConsistency && !this._zIntialized) {
        throw new TypeError('Invalid argument: z in VehicleLocalPositionSetpoint');
      }
      if (checkConsistency && !this._vxIntialized) {
        throw new TypeError('Invalid argument: vx in VehicleLocalPositionSetpoint');
      }
      if (checkConsistency && !this._vyIntialized) {
        throw new TypeError('Invalid argument: vy in VehicleLocalPositionSetpoint');
      }
      if (checkConsistency && !this._vzIntialized) {
        throw new TypeError('Invalid argument: vz in VehicleLocalPositionSetpoint');
      }
      if (checkConsistency && !this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in VehicleLocalPositionSetpoint');
      }
      if (checkConsistency && !this._yawspeedIntialized) {
        throw new TypeError('Invalid argument: yawspeed in VehicleLocalPositionSetpoint');
      }
    this._refObject.acceleration = Array.from(this._wrapperFields.acceleration.data);
    this._refObject.thrust = Array.from(this._wrapperFields.thrust.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._zIntialized = true;
    this._refObject.z = refObject.z;
    this._vxIntialized = true;
    this._refObject.vx = refObject.vx;
    this._vyIntialized = true;
    this._refObject.vy = refObject.vy;
    this._vzIntialized = true;
    this._refObject.vz = refObject.vz;
    this._wrapperFields.acceleration.fill(refObject.acceleration.toArray());
    this._wrapperFields.thrust.fill(refObject.thrust.toArray());
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
    this._yawspeedIntialized = true;
    this._refObject.yawspeed = refObject.yawspeed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleLocalPositionSetpointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleLocalPositionSetpoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get z() {
    if (!this._zIntialized) {
      return undefined;
    }
    return this._refObject.z;
  }
  set z(value) {
    this._zIntialized = true;
    this._refObject.z = value;
  }
  get vx() {
    if (!this._vxIntialized) {
      return undefined;
    }
    return this._refObject.vx;
  }
  set vx(value) {
    this._vxIntialized = true;
    this._refObject.vx = value;
  }
  get vy() {
    if (!this._vyIntialized) {
      return undefined;
    }
    return this._refObject.vy;
  }
  set vy(value) {
    this._vyIntialized = true;
    this._refObject.vy = value;
  }
  get vz() {
    if (!this._vzIntialized) {
      return undefined;
    }
    return this._refObject.vz;
  }
  set vz(value) {
    this._vzIntialized = true;
    this._refObject.vz = value;
  }
  get acceleration() {
    return this._wrapperFields['acceleration'].data;
  }
  set acceleration(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['acceleration'].fill(value);
  }
  get thrust() {
    return this._wrapperFields['thrust'].data;
  }
  set thrust(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['thrust'].fill(value);
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  get yawspeed() {
    if (!this._yawspeedIntialized) {
      return undefined;
    }
    return this._refObject.yawspeed;
  }
  set yawspeed(value) {
    this._yawspeedIntialized = true;
    this._refObject.yawspeed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleLocalPositionSetpointRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._vxIntialized = true;
    this._vyIntialized = true;
    this._vzIntialized = true;
    this._wrapperFields.acceleration.fill(refObject.acceleration.toArray());
    this._wrapperFields.thrust.fill(refObject.thrust.toArray());
    this._yawIntialized = true;
    this._yawspeedIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleLocalPositionSetpointRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._vxIntialized = true;
    this._vyIntialized = true;
    this._vzIntialized = true;
    this._wrapperFields.acceleration.copy(other._wrapperFields.acceleration);
    this._wrapperFields.thrust.copy(other._wrapperFields.thrust);
    this._yawIntialized = true;
    this._yawspeedIntialized = true;
  }
  static get classType() {
    return VehicleLocalPositionSetpointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vx","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"acceleration","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"thrust","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yawspeed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleLocalPositionSetpoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleLocalPositionSetpoint"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","x","y","z","vx","vy","vz","acceleration","thrust","yaw","yawspeed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleLocalPositionSetpointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleLocalPositionSetpointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleLocalPositionSetpointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleLocalPositionSetpointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleLocalPositionSetpointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleLocalPositionSetpointRefArray(size);
    this._refObject = new VehicleLocalPositionSetpointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleLocalPositionSetpointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleLocalPositionSetpointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleLocalPositionSetpointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleLocalPositionSetpointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleLocalPositionSetpointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleLocalPositionSetpointArrayWrapper;
  }
}
module.exports = VehicleLocalPositionSetpointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vx",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "acceleration",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "thrust",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yawspeed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleLocalPositionSetpoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleLocalPositionSetpoint"
}
*/
