// rclnodejs/generated/px4_msgs/px4_msgs__msg__VtolVehicleStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const VtolVehicleStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  vehicle_vtol_state: primitiveTypes.uint8,
  fixed_wing_system_failure: primitiveTypes.bool,
});
const VtolVehicleStatusRefArray = ArrayType(VtolVehicleStatusRefStruct);
const VtolVehicleStatusRefStructArray = StructType({
  data: VtolVehicleStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VtolVehicleStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VtolVehicleStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._vehicle_vtol_stateIntialized = true;
      this._fixed_wing_system_failureIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VtolVehicleStatusRefStruct();
    this._timestampIntialized = false;
    this._vehicle_vtol_stateIntialized = false;
    this._fixed_wing_system_failureIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VtolVehicleStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VtolVehicleStatusArrayWrapper;
  }
  static get ArrayType() {
    return VtolVehicleStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return VtolVehicleStatusRefStructArray
  }
  static get refObjectType() {
    return VtolVehicleStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VtolVehicleStatus');
      }
      if (checkConsistency && !this._vehicle_vtol_stateIntialized) {
        throw new TypeError('Invalid argument: vehicle_vtol_state in VtolVehicleStatus');
      }
      if (checkConsistency && !this._fixed_wing_system_failureIntialized) {
        throw new TypeError('Invalid argument: fixed_wing_system_failure in VtolVehicleStatus');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._vehicle_vtol_stateIntialized = true;
    this._refObject.vehicle_vtol_state = refObject.vehicle_vtol_state;
    this._fixed_wing_system_failureIntialized = true;
    this._refObject.fixed_wing_system_failure = refObject.fixed_wing_system_failure;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VtolVehicleStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VtolVehicleStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get vehicle_vtol_state() {
    if (!this._vehicle_vtol_stateIntialized) {
      return undefined;
    }
    return this._refObject.vehicle_vtol_state;
  }
  set vehicle_vtol_state(value) {
    this._vehicle_vtol_stateIntialized = true;
    this._refObject.vehicle_vtol_state = value;
  }
  get fixed_wing_system_failure() {
    if (!this._fixed_wing_system_failureIntialized) {
      return undefined;
    }
    return this._refObject.fixed_wing_system_failure;
  }
  set fixed_wing_system_failure(value) {
    this._fixed_wing_system_failureIntialized = true;
    this._refObject.fixed_wing_system_failure = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VtolVehicleStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._vehicle_vtol_stateIntialized = true;
    this._fixed_wing_system_failureIntialized = true;
  }
  copy(other) {
    this._refObject = new VtolVehicleStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._vehicle_vtol_stateIntialized = true;
    this._fixed_wing_system_failureIntialized = true;
  }
  static get classType() {
    return VtolVehicleStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"VEHICLE_VTOL_STATE_UNDEFINED","value":0},{"type":"uint8","name":"VEHICLE_VTOL_STATE_TRANSITION_TO_FW","value":1},{"type":"uint8","name":"VEHICLE_VTOL_STATE_TRANSITION_TO_MC","value":2},{"type":"uint8","name":"VEHICLE_VTOL_STATE_MC","value":3},{"type":"uint8","name":"VEHICLE_VTOL_STATE_FW","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vehicle_vtol_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fixed_wing_system_failure","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VtolVehicleStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VtolVehicleStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","vehicle_vtol_state","fixed_wing_system_failure"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VtolVehicleStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VtolVehicleStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VtolVehicleStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VtolVehicleStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VtolVehicleStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VtolVehicleStatusRefArray(size);
    this._refObject = new VtolVehicleStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VtolVehicleStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VtolVehicleStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VtolVehicleStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VtolVehicleStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VtolVehicleStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VtolVehicleStatusArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(VtolVehicleStatusWrapper, "VEHICLE_VTOL_STATE_UNDEFINED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VtolVehicleStatusWrapper, "VEHICLE_VTOL_STATE_TRANSITION_TO_FW", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VtolVehicleStatusWrapper, "VEHICLE_VTOL_STATE_TRANSITION_TO_MC", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VtolVehicleStatusWrapper, "VEHICLE_VTOL_STATE_MC", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VtolVehicleStatusWrapper, "VEHICLE_VTOL_STATE_FW", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = VtolVehicleStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "VEHICLE_VTOL_STATE_UNDEFINED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "VEHICLE_VTOL_STATE_TRANSITION_TO_FW",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "VEHICLE_VTOL_STATE_TRANSITION_TO_MC",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "VEHICLE_VTOL_STATE_MC",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "VEHICLE_VTOL_STATE_FW",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vehicle_vtol_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fixed_wing_system_failure",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VtolVehicleStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VtolVehicleStatus"
}
*/
