// rclnodejs/generated/px4_msgs/px4_msgs__msg__EstimatorSensorBias.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const EstimatorSensorBiasRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  gyro_device_id: primitiveTypes.uint32,
  gyro_bias: ArrayType(primitiveTypes.float32, 3),
  gyro_bias_limit: primitiveTypes.float32,
  gyro_bias_variance: ArrayType(primitiveTypes.float32, 3),
  gyro_bias_valid: primitiveTypes.bool,
  gyro_bias_stable: primitiveTypes.bool,
  accel_device_id: primitiveTypes.uint32,
  accel_bias: ArrayType(primitiveTypes.float32, 3),
  accel_bias_limit: primitiveTypes.float32,
  accel_bias_variance: ArrayType(primitiveTypes.float32, 3),
  accel_bias_valid: primitiveTypes.bool,
  accel_bias_stable: primitiveTypes.bool,
  mag_device_id: primitiveTypes.uint32,
  mag_bias: ArrayType(primitiveTypes.float32, 3),
  mag_bias_limit: primitiveTypes.float32,
  mag_bias_variance: ArrayType(primitiveTypes.float32, 3),
  mag_bias_valid: primitiveTypes.bool,
  mag_bias_stable: primitiveTypes.bool,
});
const EstimatorSensorBiasRefArray = ArrayType(EstimatorSensorBiasRefStruct);
const EstimatorSensorBiasRefStructArray = StructType({
  data: EstimatorSensorBiasRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EstimatorSensorBiasWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EstimatorSensorBiasRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._gyro_device_idIntialized = true;
      this._wrapperFields.gyro_bias = Float32Wrapper.createArray();
      this._wrapperFields.gyro_bias.copy(other._wrapperFields.gyro_bias);
      this._gyro_bias_limitIntialized = true;
      this._wrapperFields.gyro_bias_variance = Float32Wrapper.createArray();
      this._wrapperFields.gyro_bias_variance.copy(other._wrapperFields.gyro_bias_variance);
      this._gyro_bias_validIntialized = true;
      this._gyro_bias_stableIntialized = true;
      this._accel_device_idIntialized = true;
      this._wrapperFields.accel_bias = Float32Wrapper.createArray();
      this._wrapperFields.accel_bias.copy(other._wrapperFields.accel_bias);
      this._accel_bias_limitIntialized = true;
      this._wrapperFields.accel_bias_variance = Float32Wrapper.createArray();
      this._wrapperFields.accel_bias_variance.copy(other._wrapperFields.accel_bias_variance);
      this._accel_bias_validIntialized = true;
      this._accel_bias_stableIntialized = true;
      this._mag_device_idIntialized = true;
      this._wrapperFields.mag_bias = Float32Wrapper.createArray();
      this._wrapperFields.mag_bias.copy(other._wrapperFields.mag_bias);
      this._mag_bias_limitIntialized = true;
      this._wrapperFields.mag_bias_variance = Float32Wrapper.createArray();
      this._wrapperFields.mag_bias_variance.copy(other._wrapperFields.mag_bias_variance);
      this._mag_bias_validIntialized = true;
      this._mag_bias_stableIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EstimatorSensorBiasRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._gyro_device_idIntialized = false;
    this._wrapperFields.gyro_bias = Float32Wrapper.createArray();
    this._gyro_bias_limitIntialized = false;
    this._wrapperFields.gyro_bias_variance = Float32Wrapper.createArray();
    this._gyro_bias_validIntialized = false;
    this._gyro_bias_stableIntialized = false;
    this._accel_device_idIntialized = false;
    this._wrapperFields.accel_bias = Float32Wrapper.createArray();
    this._accel_bias_limitIntialized = false;
    this._wrapperFields.accel_bias_variance = Float32Wrapper.createArray();
    this._accel_bias_validIntialized = false;
    this._accel_bias_stableIntialized = false;
    this._mag_device_idIntialized = false;
    this._wrapperFields.mag_bias = Float32Wrapper.createArray();
    this._mag_bias_limitIntialized = false;
    this._wrapperFields.mag_bias_variance = Float32Wrapper.createArray();
    this._mag_bias_validIntialized = false;
    this._mag_bias_stableIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EstimatorSensorBiasWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EstimatorSensorBiasArrayWrapper;
  }
  static get ArrayType() {
    return EstimatorSensorBiasArrayWrapper;
  }
  static get refObjectArrayType() {
    return EstimatorSensorBiasRefStructArray
  }
  static get refObjectType() {
    return EstimatorSensorBiasRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in EstimatorSensorBias');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in EstimatorSensorBias');
      }
      if (checkConsistency && !this._gyro_device_idIntialized) {
        throw new TypeError('Invalid argument: gyro_device_id in EstimatorSensorBias');
      }
      if (checkConsistency && !this._gyro_bias_limitIntialized) {
        throw new TypeError('Invalid argument: gyro_bias_limit in EstimatorSensorBias');
      }
      if (checkConsistency && !this._gyro_bias_validIntialized) {
        throw new TypeError('Invalid argument: gyro_bias_valid in EstimatorSensorBias');
      }
      if (checkConsistency && !this._gyro_bias_stableIntialized) {
        throw new TypeError('Invalid argument: gyro_bias_stable in EstimatorSensorBias');
      }
      if (checkConsistency && !this._accel_device_idIntialized) {
        throw new TypeError('Invalid argument: accel_device_id in EstimatorSensorBias');
      }
      if (checkConsistency && !this._accel_bias_limitIntialized) {
        throw new TypeError('Invalid argument: accel_bias_limit in EstimatorSensorBias');
      }
      if (checkConsistency && !this._accel_bias_validIntialized) {
        throw new TypeError('Invalid argument: accel_bias_valid in EstimatorSensorBias');
      }
      if (checkConsistency && !this._accel_bias_stableIntialized) {
        throw new TypeError('Invalid argument: accel_bias_stable in EstimatorSensorBias');
      }
      if (checkConsistency && !this._mag_device_idIntialized) {
        throw new TypeError('Invalid argument: mag_device_id in EstimatorSensorBias');
      }
      if (checkConsistency && !this._mag_bias_limitIntialized) {
        throw new TypeError('Invalid argument: mag_bias_limit in EstimatorSensorBias');
      }
      if (checkConsistency && !this._mag_bias_validIntialized) {
        throw new TypeError('Invalid argument: mag_bias_valid in EstimatorSensorBias');
      }
      if (checkConsistency && !this._mag_bias_stableIntialized) {
        throw new TypeError('Invalid argument: mag_bias_stable in EstimatorSensorBias');
      }
    this._refObject.gyro_bias = Array.from(this._wrapperFields.gyro_bias.data);
    this._refObject.gyro_bias_variance = Array.from(this._wrapperFields.gyro_bias_variance.data);
    this._refObject.accel_bias = Array.from(this._wrapperFields.accel_bias.data);
    this._refObject.accel_bias_variance = Array.from(this._wrapperFields.accel_bias_variance.data);
    this._refObject.mag_bias = Array.from(this._wrapperFields.mag_bias.data);
    this._refObject.mag_bias_variance = Array.from(this._wrapperFields.mag_bias_variance.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._gyro_device_idIntialized = true;
    this._refObject.gyro_device_id = refObject.gyro_device_id;
    this._wrapperFields.gyro_bias.fill(refObject.gyro_bias.toArray());
    this._gyro_bias_limitIntialized = true;
    this._refObject.gyro_bias_limit = refObject.gyro_bias_limit;
    this._wrapperFields.gyro_bias_variance.fill(refObject.gyro_bias_variance.toArray());
    this._gyro_bias_validIntialized = true;
    this._refObject.gyro_bias_valid = refObject.gyro_bias_valid;
    this._gyro_bias_stableIntialized = true;
    this._refObject.gyro_bias_stable = refObject.gyro_bias_stable;
    this._accel_device_idIntialized = true;
    this._refObject.accel_device_id = refObject.accel_device_id;
    this._wrapperFields.accel_bias.fill(refObject.accel_bias.toArray());
    this._accel_bias_limitIntialized = true;
    this._refObject.accel_bias_limit = refObject.accel_bias_limit;
    this._wrapperFields.accel_bias_variance.fill(refObject.accel_bias_variance.toArray());
    this._accel_bias_validIntialized = true;
    this._refObject.accel_bias_valid = refObject.accel_bias_valid;
    this._accel_bias_stableIntialized = true;
    this._refObject.accel_bias_stable = refObject.accel_bias_stable;
    this._mag_device_idIntialized = true;
    this._refObject.mag_device_id = refObject.mag_device_id;
    this._wrapperFields.mag_bias.fill(refObject.mag_bias.toArray());
    this._mag_bias_limitIntialized = true;
    this._refObject.mag_bias_limit = refObject.mag_bias_limit;
    this._wrapperFields.mag_bias_variance.fill(refObject.mag_bias_variance.toArray());
    this._mag_bias_validIntialized = true;
    this._refObject.mag_bias_valid = refObject.mag_bias_valid;
    this._mag_bias_stableIntialized = true;
    this._refObject.mag_bias_stable = refObject.mag_bias_stable;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    EstimatorSensorBiasWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'EstimatorSensorBias'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get gyro_device_id() {
    if (!this._gyro_device_idIntialized) {
      return undefined;
    }
    return this._refObject.gyro_device_id;
  }
  set gyro_device_id(value) {
    this._gyro_device_idIntialized = true;
    this._refObject.gyro_device_id = value;
  }
  get gyro_bias() {
    return this._wrapperFields['gyro_bias'].data;
  }
  set gyro_bias(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['gyro_bias'].fill(value);
  }
  get gyro_bias_limit() {
    if (!this._gyro_bias_limitIntialized) {
      return undefined;
    }
    return this._refObject.gyro_bias_limit;
  }
  set gyro_bias_limit(value) {
    this._gyro_bias_limitIntialized = true;
    this._refObject.gyro_bias_limit = value;
  }
  get gyro_bias_variance() {
    return this._wrapperFields['gyro_bias_variance'].data;
  }
  set gyro_bias_variance(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['gyro_bias_variance'].fill(value);
  }
  get gyro_bias_valid() {
    if (!this._gyro_bias_validIntialized) {
      return undefined;
    }
    return this._refObject.gyro_bias_valid;
  }
  set gyro_bias_valid(value) {
    this._gyro_bias_validIntialized = true;
    this._refObject.gyro_bias_valid = value;
  }
  get gyro_bias_stable() {
    if (!this._gyro_bias_stableIntialized) {
      return undefined;
    }
    return this._refObject.gyro_bias_stable;
  }
  set gyro_bias_stable(value) {
    this._gyro_bias_stableIntialized = true;
    this._refObject.gyro_bias_stable = value;
  }
  get accel_device_id() {
    if (!this._accel_device_idIntialized) {
      return undefined;
    }
    return this._refObject.accel_device_id;
  }
  set accel_device_id(value) {
    this._accel_device_idIntialized = true;
    this._refObject.accel_device_id = value;
  }
  get accel_bias() {
    return this._wrapperFields['accel_bias'].data;
  }
  set accel_bias(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['accel_bias'].fill(value);
  }
  get accel_bias_limit() {
    if (!this._accel_bias_limitIntialized) {
      return undefined;
    }
    return this._refObject.accel_bias_limit;
  }
  set accel_bias_limit(value) {
    this._accel_bias_limitIntialized = true;
    this._refObject.accel_bias_limit = value;
  }
  get accel_bias_variance() {
    return this._wrapperFields['accel_bias_variance'].data;
  }
  set accel_bias_variance(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['accel_bias_variance'].fill(value);
  }
  get accel_bias_valid() {
    if (!this._accel_bias_validIntialized) {
      return undefined;
    }
    return this._refObject.accel_bias_valid;
  }
  set accel_bias_valid(value) {
    this._accel_bias_validIntialized = true;
    this._refObject.accel_bias_valid = value;
  }
  get accel_bias_stable() {
    if (!this._accel_bias_stableIntialized) {
      return undefined;
    }
    return this._refObject.accel_bias_stable;
  }
  set accel_bias_stable(value) {
    this._accel_bias_stableIntialized = true;
    this._refObject.accel_bias_stable = value;
  }
  get mag_device_id() {
    if (!this._mag_device_idIntialized) {
      return undefined;
    }
    return this._refObject.mag_device_id;
  }
  set mag_device_id(value) {
    this._mag_device_idIntialized = true;
    this._refObject.mag_device_id = value;
  }
  get mag_bias() {
    return this._wrapperFields['mag_bias'].data;
  }
  set mag_bias(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['mag_bias'].fill(value);
  }
  get mag_bias_limit() {
    if (!this._mag_bias_limitIntialized) {
      return undefined;
    }
    return this._refObject.mag_bias_limit;
  }
  set mag_bias_limit(value) {
    this._mag_bias_limitIntialized = true;
    this._refObject.mag_bias_limit = value;
  }
  get mag_bias_variance() {
    return this._wrapperFields['mag_bias_variance'].data;
  }
  set mag_bias_variance(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['mag_bias_variance'].fill(value);
  }
  get mag_bias_valid() {
    if (!this._mag_bias_validIntialized) {
      return undefined;
    }
    return this._refObject.mag_bias_valid;
  }
  set mag_bias_valid(value) {
    this._mag_bias_validIntialized = true;
    this._refObject.mag_bias_valid = value;
  }
  get mag_bias_stable() {
    if (!this._mag_bias_stableIntialized) {
      return undefined;
    }
    return this._refObject.mag_bias_stable;
  }
  set mag_bias_stable(value) {
    this._mag_bias_stableIntialized = true;
    this._refObject.mag_bias_stable = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EstimatorSensorBiasRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._gyro_device_idIntialized = true;
    this._wrapperFields.gyro_bias.fill(refObject.gyro_bias.toArray());
    this._gyro_bias_limitIntialized = true;
    this._wrapperFields.gyro_bias_variance.fill(refObject.gyro_bias_variance.toArray());
    this._gyro_bias_validIntialized = true;
    this._gyro_bias_stableIntialized = true;
    this._accel_device_idIntialized = true;
    this._wrapperFields.accel_bias.fill(refObject.accel_bias.toArray());
    this._accel_bias_limitIntialized = true;
    this._wrapperFields.accel_bias_variance.fill(refObject.accel_bias_variance.toArray());
    this._accel_bias_validIntialized = true;
    this._accel_bias_stableIntialized = true;
    this._mag_device_idIntialized = true;
    this._wrapperFields.mag_bias.fill(refObject.mag_bias.toArray());
    this._mag_bias_limitIntialized = true;
    this._wrapperFields.mag_bias_variance.fill(refObject.mag_bias_variance.toArray());
    this._mag_bias_validIntialized = true;
    this._mag_bias_stableIntialized = true;
  }
  copy(other) {
    this._refObject = new EstimatorSensorBiasRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._gyro_device_idIntialized = true;
    this._wrapperFields.gyro_bias.copy(other._wrapperFields.gyro_bias);
    this._gyro_bias_limitIntialized = true;
    this._wrapperFields.gyro_bias_variance.copy(other._wrapperFields.gyro_bias_variance);
    this._gyro_bias_validIntialized = true;
    this._gyro_bias_stableIntialized = true;
    this._accel_device_idIntialized = true;
    this._wrapperFields.accel_bias.copy(other._wrapperFields.accel_bias);
    this._accel_bias_limitIntialized = true;
    this._wrapperFields.accel_bias_variance.copy(other._wrapperFields.accel_bias_variance);
    this._accel_bias_validIntialized = true;
    this._accel_bias_stableIntialized = true;
    this._mag_device_idIntialized = true;
    this._wrapperFields.mag_bias.copy(other._wrapperFields.mag_bias);
    this._mag_bias_limitIntialized = true;
    this._wrapperFields.mag_bias_variance.copy(other._wrapperFields.mag_bias_variance);
    this._mag_bias_validIntialized = true;
    this._mag_bias_stableIntialized = true;
  }
  static get classType() {
    return EstimatorSensorBiasWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_bias","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_bias_limit","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_bias_variance","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_bias_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gyro_bias_stable","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_bias","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_bias_limit","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_bias_variance","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_bias_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"accel_bias_stable","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_bias","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_bias_limit","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_bias_variance","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_bias_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_bias_stable","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"EstimatorSensorBias","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EstimatorSensorBias"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","gyro_device_id","gyro_bias","gyro_bias_limit","gyro_bias_variance","gyro_bias_valid","gyro_bias_stable","accel_device_id","accel_bias","accel_bias_limit","accel_bias_variance","accel_bias_valid","accel_bias_stable","mag_device_id","mag_bias","mag_bias_limit","mag_bias_variance","mag_bias_valid","mag_bias_stable"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EstimatorSensorBiasArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EstimatorSensorBiasWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EstimatorSensorBiasWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorSensorBiasArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorSensorBiasArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EstimatorSensorBiasRefArray(size);
    this._refObject = new EstimatorSensorBiasRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EstimatorSensorBiasWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EstimatorSensorBiasArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EstimatorSensorBiasArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EstimatorSensorBiasWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EstimatorSensorBiasWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EstimatorSensorBiasArrayWrapper;
  }
}
module.exports = EstimatorSensorBiasWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_bias",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_bias_limit",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_bias_variance",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_bias_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gyro_bias_stable",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_bias",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_bias_limit",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_bias_variance",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_bias_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "accel_bias_stable",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_bias",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_bias_limit",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_bias_variance",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_bias_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_bias_stable",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "EstimatorSensorBias",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EstimatorSensorBias"
}
*/
