// rclnodejs/generated/px4_msgs/px4_msgs__msg__YawEstimatorStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const YawEstimatorStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  yaw_composite: primitiveTypes.float32,
  yaw_variance: primitiveTypes.float32,
  yaw_composite_valid: primitiveTypes.bool,
  yaw: ArrayType(primitiveTypes.float32, 5),
  innov_vn: ArrayType(primitiveTypes.float32, 5),
  innov_ve: ArrayType(primitiveTypes.float32, 5),
  weight: ArrayType(primitiveTypes.float32, 5),
});
const YawEstimatorStatusRefArray = ArrayType(YawEstimatorStatusRefStruct);
const YawEstimatorStatusRefStructArray = StructType({
  data: YawEstimatorStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class YawEstimatorStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new YawEstimatorStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._yaw_compositeIntialized = true;
      this._yaw_varianceIntialized = true;
      this._yaw_composite_validIntialized = true;
      this._wrapperFields.yaw = Float32Wrapper.createArray();
      this._wrapperFields.yaw.copy(other._wrapperFields.yaw);
      this._wrapperFields.innov_vn = Float32Wrapper.createArray();
      this._wrapperFields.innov_vn.copy(other._wrapperFields.innov_vn);
      this._wrapperFields.innov_ve = Float32Wrapper.createArray();
      this._wrapperFields.innov_ve.copy(other._wrapperFields.innov_ve);
      this._wrapperFields.weight = Float32Wrapper.createArray();
      this._wrapperFields.weight.copy(other._wrapperFields.weight);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new YawEstimatorStatusRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._yaw_compositeIntialized = false;
    this._yaw_varianceIntialized = false;
    this._yaw_composite_validIntialized = false;
    this._wrapperFields.yaw = Float32Wrapper.createArray();
    this._wrapperFields.innov_vn = Float32Wrapper.createArray();
    this._wrapperFields.innov_ve = Float32Wrapper.createArray();
    this._wrapperFields.weight = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new YawEstimatorStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new YawEstimatorStatusArrayWrapper;
  }
  static get ArrayType() {
    return YawEstimatorStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return YawEstimatorStatusRefStructArray
  }
  static get refObjectType() {
    return YawEstimatorStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in YawEstimatorStatus');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in YawEstimatorStatus');
      }
      if (checkConsistency && !this._yaw_compositeIntialized) {
        throw new TypeError('Invalid argument: yaw_composite in YawEstimatorStatus');
      }
      if (checkConsistency && !this._yaw_varianceIntialized) {
        throw new TypeError('Invalid argument: yaw_variance in YawEstimatorStatus');
      }
      if (checkConsistency && !this._yaw_composite_validIntialized) {
        throw new TypeError('Invalid argument: yaw_composite_valid in YawEstimatorStatus');
      }
    this._refObject.yaw = Array.from(this._wrapperFields.yaw.data);
    this._refObject.innov_vn = Array.from(this._wrapperFields.innov_vn.data);
    this._refObject.innov_ve = Array.from(this._wrapperFields.innov_ve.data);
    this._refObject.weight = Array.from(this._wrapperFields.weight.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._yaw_compositeIntialized = true;
    this._refObject.yaw_composite = refObject.yaw_composite;
    this._yaw_varianceIntialized = true;
    this._refObject.yaw_variance = refObject.yaw_variance;
    this._yaw_composite_validIntialized = true;
    this._refObject.yaw_composite_valid = refObject.yaw_composite_valid;
    this._wrapperFields.yaw.fill(refObject.yaw.toArray());
    this._wrapperFields.innov_vn.fill(refObject.innov_vn.toArray());
    this._wrapperFields.innov_ve.fill(refObject.innov_ve.toArray());
    this._wrapperFields.weight.fill(refObject.weight.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    YawEstimatorStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'YawEstimatorStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get yaw_composite() {
    if (!this._yaw_compositeIntialized) {
      return undefined;
    }
    return this._refObject.yaw_composite;
  }
  set yaw_composite(value) {
    this._yaw_compositeIntialized = true;
    this._refObject.yaw_composite = value;
  }
  get yaw_variance() {
    if (!this._yaw_varianceIntialized) {
      return undefined;
    }
    return this._refObject.yaw_variance;
  }
  set yaw_variance(value) {
    this._yaw_varianceIntialized = true;
    this._refObject.yaw_variance = value;
  }
  get yaw_composite_valid() {
    if (!this._yaw_composite_validIntialized) {
      return undefined;
    }
    return this._refObject.yaw_composite_valid;
  }
  set yaw_composite_valid(value) {
    this._yaw_composite_validIntialized = true;
    this._refObject.yaw_composite_valid = value;
  }
  get yaw() {
    return this._wrapperFields['yaw'].data;
  }
  set yaw(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields['yaw'].fill(value);
  }
  get innov_vn() {
    return this._wrapperFields['innov_vn'].data;
  }
  set innov_vn(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields['innov_vn'].fill(value);
  }
  get innov_ve() {
    return this._wrapperFields['innov_ve'].data;
  }
  set innov_ve(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields['innov_ve'].fill(value);
  }
  get weight() {
    return this._wrapperFields['weight'].data;
  }
  set weight(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields['weight'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new YawEstimatorStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._yaw_compositeIntialized = true;
    this._yaw_varianceIntialized = true;
    this._yaw_composite_validIntialized = true;
    this._wrapperFields.yaw.fill(refObject.yaw.toArray());
    this._wrapperFields.innov_vn.fill(refObject.innov_vn.toArray());
    this._wrapperFields.innov_ve.fill(refObject.innov_ve.toArray());
    this._wrapperFields.weight.fill(refObject.weight.toArray());
  }
  copy(other) {
    this._refObject = new YawEstimatorStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._yaw_compositeIntialized = true;
    this._yaw_varianceIntialized = true;
    this._yaw_composite_validIntialized = true;
    this._wrapperFields.yaw.copy(other._wrapperFields.yaw);
    this._wrapperFields.innov_vn.copy(other._wrapperFields.innov_vn);
    this._wrapperFields.innov_ve.copy(other._wrapperFields.innov_ve);
    this._wrapperFields.weight.copy(other._wrapperFields.weight);
  }
  static get classType() {
    return YawEstimatorStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_composite","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_variance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_composite_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov_vn","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov_ve","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"weight","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"YawEstimatorStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"YawEstimatorStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","yaw_composite","yaw_variance","yaw_composite_valid","yaw","innov_vn","innov_ve","weight"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class YawEstimatorStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof YawEstimatorStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new YawEstimatorStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to YawEstimatorStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to YawEstimatorStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new YawEstimatorStatusRefArray(size);
    this._refObject = new YawEstimatorStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new YawEstimatorStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof YawEstimatorStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "YawEstimatorStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      YawEstimatorStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return YawEstimatorStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return YawEstimatorStatusArrayWrapper;
  }
}
module.exports = YawEstimatorStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_composite",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_variance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_composite_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov_vn",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov_ve",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "weight",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "YawEstimatorStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "YawEstimatorStatus"
}
*/
