// rclnodejs/generated/px4_msgs/px4_msgs__msg__ActuatorOutputs.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const ActuatorOutputsRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  noutputs: primitiveTypes.uint32,
  output: ArrayType(primitiveTypes.float32, 16),
});
const ActuatorOutputsRefArray = ArrayType(ActuatorOutputsRefStruct);
const ActuatorOutputsRefStructArray = StructType({
  data: ActuatorOutputsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ActuatorOutputsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ActuatorOutputsRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._noutputsIntialized = true;
      this._wrapperFields.output = Float32Wrapper.createArray();
      this._wrapperFields.output.copy(other._wrapperFields.output);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ActuatorOutputsRefStruct();
    this._timestampIntialized = false;
    this._noutputsIntialized = false;
    this._wrapperFields.output = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ActuatorOutputsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ActuatorOutputsArrayWrapper;
  }
  static get ArrayType() {
    return ActuatorOutputsArrayWrapper;
  }
  static get refObjectArrayType() {
    return ActuatorOutputsRefStructArray
  }
  static get refObjectType() {
    return ActuatorOutputsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in ActuatorOutputs');
      }
      if (checkConsistency && !this._noutputsIntialized) {
        throw new TypeError('Invalid argument: noutputs in ActuatorOutputs');
      }
    this._refObject.output = Array.from(this._wrapperFields.output.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._noutputsIntialized = true;
    this._refObject.noutputs = refObject.noutputs;
    this._wrapperFields.output.fill(refObject.output.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ActuatorOutputsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'ActuatorOutputs'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get noutputs() {
    if (!this._noutputsIntialized) {
      return undefined;
    }
    return this._refObject.noutputs;
  }
  set noutputs(value) {
    this._noutputsIntialized = true;
    this._refObject.noutputs = value;
  }
  get output() {
    return this._wrapperFields['output'].data;
  }
  set output(value) {
    if (value.length !== 16) {
      throw new RangeError('The length of the array must be 16.');
    }
    this._wrapperFields['output'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ActuatorOutputsRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._noutputsIntialized = true;
    this._wrapperFields.output.fill(refObject.output.toArray());
  }
  copy(other) {
    this._refObject = new ActuatorOutputsRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._noutputsIntialized = true;
    this._wrapperFields.output.copy(other._wrapperFields.output);
  }
  static get classType() {
    return ActuatorOutputsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"NUM_ACTUATOR_OUTPUTS","value":16},{"type":"uint8","name":"NUM_ACTUATOR_OUTPUT_GROUPS","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"noutputs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"output","type":{"isArray":true,"arraySize":16,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"ActuatorOutputs","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ActuatorOutputs"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","noutputs","output"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ActuatorOutputsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ActuatorOutputsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ActuatorOutputsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorOutputsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorOutputsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ActuatorOutputsRefArray(size);
    this._refObject = new ActuatorOutputsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ActuatorOutputsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ActuatorOutputsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ActuatorOutputsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ActuatorOutputsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ActuatorOutputsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ActuatorOutputsArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(ActuatorOutputsWrapper, "NUM_ACTUATOR_OUTPUTS", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorOutputsWrapper, "NUM_ACTUATOR_OUTPUT_GROUPS", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = ActuatorOutputsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "NUM_ACTUATOR_OUTPUTS",
      "value": 16
    },
    {
      "type": "uint8",
      "name": "NUM_ACTUATOR_OUTPUT_GROUPS",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "noutputs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "output",
      "type": {
        "isArray": true,
        "arraySize": 16,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "ActuatorOutputs",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ActuatorOutputs"
}
*/
