// rclnodejs/generated/px4_msgs/px4_msgs__msg__LedControl.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LedControlRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  led_mask: primitiveTypes.uint8,
  color: primitiveTypes.uint8,
  mode: primitiveTypes.uint8,
  num_blinks: primitiveTypes.uint8,
  priority: primitiveTypes.uint8,
});
const LedControlRefArray = ArrayType(LedControlRefStruct);
const LedControlRefStructArray = StructType({
  data: LedControlRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LedControlWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LedControlRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._led_maskIntialized = true;
      this._colorIntialized = true;
      this._modeIntialized = true;
      this._num_blinksIntialized = true;
      this._priorityIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LedControlRefStruct();
    this._timestampIntialized = false;
    this._led_maskIntialized = false;
    this._colorIntialized = false;
    this._modeIntialized = false;
    this._num_blinksIntialized = false;
    this._priorityIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LedControlWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LedControlArrayWrapper;
  }
  static get ArrayType() {
    return LedControlArrayWrapper;
  }
  static get refObjectArrayType() {
    return LedControlRefStructArray
  }
  static get refObjectType() {
    return LedControlRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in LedControl');
      }
      if (checkConsistency && !this._led_maskIntialized) {
        throw new TypeError('Invalid argument: led_mask in LedControl');
      }
      if (checkConsistency && !this._colorIntialized) {
        throw new TypeError('Invalid argument: color in LedControl');
      }
      if (checkConsistency && !this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in LedControl');
      }
      if (checkConsistency && !this._num_blinksIntialized) {
        throw new TypeError('Invalid argument: num_blinks in LedControl');
      }
      if (checkConsistency && !this._priorityIntialized) {
        throw new TypeError('Invalid argument: priority in LedControl');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._led_maskIntialized = true;
    this._refObject.led_mask = refObject.led_mask;
    this._colorIntialized = true;
    this._refObject.color = refObject.color;
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._num_blinksIntialized = true;
    this._refObject.num_blinks = refObject.num_blinks;
    this._priorityIntialized = true;
    this._refObject.priority = refObject.priority;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LedControlWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'LedControl'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get led_mask() {
    if (!this._led_maskIntialized) {
      return undefined;
    }
    return this._refObject.led_mask;
  }
  set led_mask(value) {
    this._led_maskIntialized = true;
    this._refObject.led_mask = value;
  }
  get color() {
    if (!this._colorIntialized) {
      return undefined;
    }
    return this._refObject.color;
  }
  set color(value) {
    this._colorIntialized = true;
    this._refObject.color = value;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get num_blinks() {
    if (!this._num_blinksIntialized) {
      return undefined;
    }
    return this._refObject.num_blinks;
  }
  set num_blinks(value) {
    this._num_blinksIntialized = true;
    this._refObject.num_blinks = value;
  }
  get priority() {
    if (!this._priorityIntialized) {
      return undefined;
    }
    return this._refObject.priority;
  }
  set priority(value) {
    this._priorityIntialized = true;
    this._refObject.priority = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LedControlRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._led_maskIntialized = true;
    this._colorIntialized = true;
    this._modeIntialized = true;
    this._num_blinksIntialized = true;
    this._priorityIntialized = true;
  }
  copy(other) {
    this._refObject = new LedControlRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._led_maskIntialized = true;
    this._colorIntialized = true;
    this._modeIntialized = true;
    this._num_blinksIntialized = true;
    this._priorityIntialized = true;
  }
  static get classType() {
    return LedControlWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"COLOR_OFF","value":0},{"type":"uint8","name":"COLOR_RED","value":1},{"type":"uint8","name":"COLOR_GREEN","value":2},{"type":"uint8","name":"COLOR_BLUE","value":3},{"type":"uint8","name":"COLOR_YELLOW","value":4},{"type":"uint8","name":"COLOR_PURPLE","value":5},{"type":"uint8","name":"COLOR_AMBER","value":6},{"type":"uint8","name":"COLOR_CYAN","value":7},{"type":"uint8","name":"COLOR_WHITE","value":8},{"type":"uint8","name":"MODE_OFF","value":0},{"type":"uint8","name":"MODE_ON","value":1},{"type":"uint8","name":"MODE_DISABLED","value":2},{"type":"uint8","name":"MODE_BLINK_SLOW","value":3},{"type":"uint8","name":"MODE_BLINK_NORMAL","value":4},{"type":"uint8","name":"MODE_BLINK_FAST","value":5},{"type":"uint8","name":"MODE_BREATHE","value":6},{"type":"uint8","name":"MODE_FLASH","value":7},{"type":"uint8","name":"MAX_PRIORITY","value":2},{"type":"uint8","name":"ORB_QUEUE_LENGTH","value":8}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"led_mask","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"color","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"num_blinks","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"priority","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"LedControl","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LedControl"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","led_mask","color","mode","num_blinks","priority"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LedControlArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LedControlWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LedControlWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LedControlArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LedControlArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LedControlRefArray(size);
    this._refObject = new LedControlRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LedControlWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LedControlArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LedControlArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LedControlWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LedControlWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LedControlArrayWrapper;
  }
}
// Define constants (19 in total)
Object.defineProperty(LedControlWrapper, "COLOR_OFF", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "COLOR_RED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "COLOR_GREEN", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "COLOR_BLUE", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "COLOR_YELLOW", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "COLOR_PURPLE", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "COLOR_AMBER", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "COLOR_CYAN", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "COLOR_WHITE", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "MODE_OFF", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "MODE_ON", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "MODE_DISABLED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "MODE_BLINK_SLOW", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "MODE_BLINK_NORMAL", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "MODE_BLINK_FAST", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "MODE_BREATHE", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "MODE_FLASH", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "MAX_PRIORITY", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LedControlWrapper, "ORB_QUEUE_LENGTH", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = LedControlWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "COLOR_OFF",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "COLOR_RED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "COLOR_GREEN",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "COLOR_BLUE",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "COLOR_YELLOW",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "COLOR_PURPLE",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "COLOR_AMBER",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "COLOR_CYAN",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "COLOR_WHITE",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "MODE_OFF",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "MODE_ON",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "MODE_DISABLED",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "MODE_BLINK_SLOW",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "MODE_BLINK_NORMAL",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "MODE_BLINK_FAST",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "MODE_BREATHE",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "MODE_FLASH",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "MAX_PRIORITY",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ORB_QUEUE_LENGTH",
      "value": 8
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "led_mask",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "color",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "num_blinks",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "priority",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "LedControl",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LedControl"
}
*/
