// rclnodejs/generated/px4_msgs/px4_msgs__msg__FailsafeFlags.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const FailsafeFlagsRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  mode_req_angular_velocity: primitiveTypes.uint32,
  mode_req_attitude: primitiveTypes.uint32,
  mode_req_local_alt: primitiveTypes.uint32,
  mode_req_local_position: primitiveTypes.uint32,
  mode_req_local_position_relaxed: primitiveTypes.uint32,
  mode_req_global_position: primitiveTypes.uint32,
  mode_req_mission: primitiveTypes.uint32,
  mode_req_offboard_signal: primitiveTypes.uint32,
  mode_req_home_position: primitiveTypes.uint32,
  mode_req_wind_and_flight_time_compliance: primitiveTypes.uint32,
  mode_req_prevent_arming: primitiveTypes.uint32,
  mode_req_manual_control: primitiveTypes.uint32,
  mode_req_other: primitiveTypes.uint32,
  angular_velocity_invalid: primitiveTypes.bool,
  attitude_invalid: primitiveTypes.bool,
  local_altitude_invalid: primitiveTypes.bool,
  local_position_invalid: primitiveTypes.bool,
  local_position_invalid_relaxed: primitiveTypes.bool,
  local_velocity_invalid: primitiveTypes.bool,
  global_position_invalid: primitiveTypes.bool,
  auto_mission_missing: primitiveTypes.bool,
  offboard_control_signal_lost: primitiveTypes.bool,
  home_position_invalid: primitiveTypes.bool,
  manual_control_signal_lost: primitiveTypes.bool,
  gcs_connection_lost: primitiveTypes.bool,
  battery_warning: primitiveTypes.uint8,
  battery_low_remaining_time: primitiveTypes.bool,
  battery_unhealthy: primitiveTypes.bool,
  primary_geofence_breached: primitiveTypes.bool,
  mission_failure: primitiveTypes.bool,
  vtol_fixed_wing_system_failure: primitiveTypes.bool,
  wind_limit_exceeded: primitiveTypes.bool,
  flight_time_limit_exceeded: primitiveTypes.bool,
  local_position_accuracy_low: primitiveTypes.bool,
  fd_critical_failure: primitiveTypes.bool,
  fd_esc_arming_failure: primitiveTypes.bool,
  fd_imbalanced_prop: primitiveTypes.bool,
  fd_motor_failure: primitiveTypes.bool,
});
const FailsafeFlagsRefArray = ArrayType(FailsafeFlagsRefStruct);
const FailsafeFlagsRefStructArray = StructType({
  data: FailsafeFlagsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FailsafeFlagsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FailsafeFlagsRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._mode_req_angular_velocityIntialized = true;
      this._mode_req_attitudeIntialized = true;
      this._mode_req_local_altIntialized = true;
      this._mode_req_local_positionIntialized = true;
      this._mode_req_local_position_relaxedIntialized = true;
      this._mode_req_global_positionIntialized = true;
      this._mode_req_missionIntialized = true;
      this._mode_req_offboard_signalIntialized = true;
      this._mode_req_home_positionIntialized = true;
      this._mode_req_wind_and_flight_time_complianceIntialized = true;
      this._mode_req_prevent_armingIntialized = true;
      this._mode_req_manual_controlIntialized = true;
      this._mode_req_otherIntialized = true;
      this._angular_velocity_invalidIntialized = true;
      this._attitude_invalidIntialized = true;
      this._local_altitude_invalidIntialized = true;
      this._local_position_invalidIntialized = true;
      this._local_position_invalid_relaxedIntialized = true;
      this._local_velocity_invalidIntialized = true;
      this._global_position_invalidIntialized = true;
      this._auto_mission_missingIntialized = true;
      this._offboard_control_signal_lostIntialized = true;
      this._home_position_invalidIntialized = true;
      this._manual_control_signal_lostIntialized = true;
      this._gcs_connection_lostIntialized = true;
      this._battery_warningIntialized = true;
      this._battery_low_remaining_timeIntialized = true;
      this._battery_unhealthyIntialized = true;
      this._primary_geofence_breachedIntialized = true;
      this._mission_failureIntialized = true;
      this._vtol_fixed_wing_system_failureIntialized = true;
      this._wind_limit_exceededIntialized = true;
      this._flight_time_limit_exceededIntialized = true;
      this._local_position_accuracy_lowIntialized = true;
      this._fd_critical_failureIntialized = true;
      this._fd_esc_arming_failureIntialized = true;
      this._fd_imbalanced_propIntialized = true;
      this._fd_motor_failureIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FailsafeFlagsRefStruct();
    this._timestampIntialized = false;
    this._mode_req_angular_velocityIntialized = false;
    this._mode_req_attitudeIntialized = false;
    this._mode_req_local_altIntialized = false;
    this._mode_req_local_positionIntialized = false;
    this._mode_req_local_position_relaxedIntialized = false;
    this._mode_req_global_positionIntialized = false;
    this._mode_req_missionIntialized = false;
    this._mode_req_offboard_signalIntialized = false;
    this._mode_req_home_positionIntialized = false;
    this._mode_req_wind_and_flight_time_complianceIntialized = false;
    this._mode_req_prevent_armingIntialized = false;
    this._mode_req_manual_controlIntialized = false;
    this._mode_req_otherIntialized = false;
    this._angular_velocity_invalidIntialized = false;
    this._attitude_invalidIntialized = false;
    this._local_altitude_invalidIntialized = false;
    this._local_position_invalidIntialized = false;
    this._local_position_invalid_relaxedIntialized = false;
    this._local_velocity_invalidIntialized = false;
    this._global_position_invalidIntialized = false;
    this._auto_mission_missingIntialized = false;
    this._offboard_control_signal_lostIntialized = false;
    this._home_position_invalidIntialized = false;
    this._manual_control_signal_lostIntialized = false;
    this._gcs_connection_lostIntialized = false;
    this._battery_warningIntialized = false;
    this._battery_low_remaining_timeIntialized = false;
    this._battery_unhealthyIntialized = false;
    this._primary_geofence_breachedIntialized = false;
    this._mission_failureIntialized = false;
    this._vtol_fixed_wing_system_failureIntialized = false;
    this._wind_limit_exceededIntialized = false;
    this._flight_time_limit_exceededIntialized = false;
    this._local_position_accuracy_lowIntialized = false;
    this._fd_critical_failureIntialized = false;
    this._fd_esc_arming_failureIntialized = false;
    this._fd_imbalanced_propIntialized = false;
    this._fd_motor_failureIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new FailsafeFlagsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FailsafeFlagsArrayWrapper;
  }
  static get ArrayType() {
    return FailsafeFlagsArrayWrapper;
  }
  static get refObjectArrayType() {
    return FailsafeFlagsRefStructArray
  }
  static get refObjectType() {
    return FailsafeFlagsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_angular_velocityIntialized) {
        throw new TypeError('Invalid argument: mode_req_angular_velocity in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_attitudeIntialized) {
        throw new TypeError('Invalid argument: mode_req_attitude in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_local_altIntialized) {
        throw new TypeError('Invalid argument: mode_req_local_alt in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_local_positionIntialized) {
        throw new TypeError('Invalid argument: mode_req_local_position in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_local_position_relaxedIntialized) {
        throw new TypeError('Invalid argument: mode_req_local_position_relaxed in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_global_positionIntialized) {
        throw new TypeError('Invalid argument: mode_req_global_position in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_missionIntialized) {
        throw new TypeError('Invalid argument: mode_req_mission in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_offboard_signalIntialized) {
        throw new TypeError('Invalid argument: mode_req_offboard_signal in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_home_positionIntialized) {
        throw new TypeError('Invalid argument: mode_req_home_position in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_wind_and_flight_time_complianceIntialized) {
        throw new TypeError('Invalid argument: mode_req_wind_and_flight_time_compliance in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_prevent_armingIntialized) {
        throw new TypeError('Invalid argument: mode_req_prevent_arming in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_manual_controlIntialized) {
        throw new TypeError('Invalid argument: mode_req_manual_control in FailsafeFlags');
      }
      if (checkConsistency && !this._mode_req_otherIntialized) {
        throw new TypeError('Invalid argument: mode_req_other in FailsafeFlags');
      }
      if (checkConsistency && !this._angular_velocity_invalidIntialized) {
        throw new TypeError('Invalid argument: angular_velocity_invalid in FailsafeFlags');
      }
      if (checkConsistency && !this._attitude_invalidIntialized) {
        throw new TypeError('Invalid argument: attitude_invalid in FailsafeFlags');
      }
      if (checkConsistency && !this._local_altitude_invalidIntialized) {
        throw new TypeError('Invalid argument: local_altitude_invalid in FailsafeFlags');
      }
      if (checkConsistency && !this._local_position_invalidIntialized) {
        throw new TypeError('Invalid argument: local_position_invalid in FailsafeFlags');
      }
      if (checkConsistency && !this._local_position_invalid_relaxedIntialized) {
        throw new TypeError('Invalid argument: local_position_invalid_relaxed in FailsafeFlags');
      }
      if (checkConsistency && !this._local_velocity_invalidIntialized) {
        throw new TypeError('Invalid argument: local_velocity_invalid in FailsafeFlags');
      }
      if (checkConsistency && !this._global_position_invalidIntialized) {
        throw new TypeError('Invalid argument: global_position_invalid in FailsafeFlags');
      }
      if (checkConsistency && !this._auto_mission_missingIntialized) {
        throw new TypeError('Invalid argument: auto_mission_missing in FailsafeFlags');
      }
      if (checkConsistency && !this._offboard_control_signal_lostIntialized) {
        throw new TypeError('Invalid argument: offboard_control_signal_lost in FailsafeFlags');
      }
      if (checkConsistency && !this._home_position_invalidIntialized) {
        throw new TypeError('Invalid argument: home_position_invalid in FailsafeFlags');
      }
      if (checkConsistency && !this._manual_control_signal_lostIntialized) {
        throw new TypeError('Invalid argument: manual_control_signal_lost in FailsafeFlags');
      }
      if (checkConsistency && !this._gcs_connection_lostIntialized) {
        throw new TypeError('Invalid argument: gcs_connection_lost in FailsafeFlags');
      }
      if (checkConsistency && !this._battery_warningIntialized) {
        throw new TypeError('Invalid argument: battery_warning in FailsafeFlags');
      }
      if (checkConsistency && !this._battery_low_remaining_timeIntialized) {
        throw new TypeError('Invalid argument: battery_low_remaining_time in FailsafeFlags');
      }
      if (checkConsistency && !this._battery_unhealthyIntialized) {
        throw new TypeError('Invalid argument: battery_unhealthy in FailsafeFlags');
      }
      if (checkConsistency && !this._primary_geofence_breachedIntialized) {
        throw new TypeError('Invalid argument: primary_geofence_breached in FailsafeFlags');
      }
      if (checkConsistency && !this._mission_failureIntialized) {
        throw new TypeError('Invalid argument: mission_failure in FailsafeFlags');
      }
      if (checkConsistency && !this._vtol_fixed_wing_system_failureIntialized) {
        throw new TypeError('Invalid argument: vtol_fixed_wing_system_failure in FailsafeFlags');
      }
      if (checkConsistency && !this._wind_limit_exceededIntialized) {
        throw new TypeError('Invalid argument: wind_limit_exceeded in FailsafeFlags');
      }
      if (checkConsistency && !this._flight_time_limit_exceededIntialized) {
        throw new TypeError('Invalid argument: flight_time_limit_exceeded in FailsafeFlags');
      }
      if (checkConsistency && !this._local_position_accuracy_lowIntialized) {
        throw new TypeError('Invalid argument: local_position_accuracy_low in FailsafeFlags');
      }
      if (checkConsistency && !this._fd_critical_failureIntialized) {
        throw new TypeError('Invalid argument: fd_critical_failure in FailsafeFlags');
      }
      if (checkConsistency && !this._fd_esc_arming_failureIntialized) {
        throw new TypeError('Invalid argument: fd_esc_arming_failure in FailsafeFlags');
      }
      if (checkConsistency && !this._fd_imbalanced_propIntialized) {
        throw new TypeError('Invalid argument: fd_imbalanced_prop in FailsafeFlags');
      }
      if (checkConsistency && !this._fd_motor_failureIntialized) {
        throw new TypeError('Invalid argument: fd_motor_failure in FailsafeFlags');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._mode_req_angular_velocityIntialized = true;
    this._refObject.mode_req_angular_velocity = refObject.mode_req_angular_velocity;
    this._mode_req_attitudeIntialized = true;
    this._refObject.mode_req_attitude = refObject.mode_req_attitude;
    this._mode_req_local_altIntialized = true;
    this._refObject.mode_req_local_alt = refObject.mode_req_local_alt;
    this._mode_req_local_positionIntialized = true;
    this._refObject.mode_req_local_position = refObject.mode_req_local_position;
    this._mode_req_local_position_relaxedIntialized = true;
    this._refObject.mode_req_local_position_relaxed = refObject.mode_req_local_position_relaxed;
    this._mode_req_global_positionIntialized = true;
    this._refObject.mode_req_global_position = refObject.mode_req_global_position;
    this._mode_req_missionIntialized = true;
    this._refObject.mode_req_mission = refObject.mode_req_mission;
    this._mode_req_offboard_signalIntialized = true;
    this._refObject.mode_req_offboard_signal = refObject.mode_req_offboard_signal;
    this._mode_req_home_positionIntialized = true;
    this._refObject.mode_req_home_position = refObject.mode_req_home_position;
    this._mode_req_wind_and_flight_time_complianceIntialized = true;
    this._refObject.mode_req_wind_and_flight_time_compliance = refObject.mode_req_wind_and_flight_time_compliance;
    this._mode_req_prevent_armingIntialized = true;
    this._refObject.mode_req_prevent_arming = refObject.mode_req_prevent_arming;
    this._mode_req_manual_controlIntialized = true;
    this._refObject.mode_req_manual_control = refObject.mode_req_manual_control;
    this._mode_req_otherIntialized = true;
    this._refObject.mode_req_other = refObject.mode_req_other;
    this._angular_velocity_invalidIntialized = true;
    this._refObject.angular_velocity_invalid = refObject.angular_velocity_invalid;
    this._attitude_invalidIntialized = true;
    this._refObject.attitude_invalid = refObject.attitude_invalid;
    this._local_altitude_invalidIntialized = true;
    this._refObject.local_altitude_invalid = refObject.local_altitude_invalid;
    this._local_position_invalidIntialized = true;
    this._refObject.local_position_invalid = refObject.local_position_invalid;
    this._local_position_invalid_relaxedIntialized = true;
    this._refObject.local_position_invalid_relaxed = refObject.local_position_invalid_relaxed;
    this._local_velocity_invalidIntialized = true;
    this._refObject.local_velocity_invalid = refObject.local_velocity_invalid;
    this._global_position_invalidIntialized = true;
    this._refObject.global_position_invalid = refObject.global_position_invalid;
    this._auto_mission_missingIntialized = true;
    this._refObject.auto_mission_missing = refObject.auto_mission_missing;
    this._offboard_control_signal_lostIntialized = true;
    this._refObject.offboard_control_signal_lost = refObject.offboard_control_signal_lost;
    this._home_position_invalidIntialized = true;
    this._refObject.home_position_invalid = refObject.home_position_invalid;
    this._manual_control_signal_lostIntialized = true;
    this._refObject.manual_control_signal_lost = refObject.manual_control_signal_lost;
    this._gcs_connection_lostIntialized = true;
    this._refObject.gcs_connection_lost = refObject.gcs_connection_lost;
    this._battery_warningIntialized = true;
    this._refObject.battery_warning = refObject.battery_warning;
    this._battery_low_remaining_timeIntialized = true;
    this._refObject.battery_low_remaining_time = refObject.battery_low_remaining_time;
    this._battery_unhealthyIntialized = true;
    this._refObject.battery_unhealthy = refObject.battery_unhealthy;
    this._primary_geofence_breachedIntialized = true;
    this._refObject.primary_geofence_breached = refObject.primary_geofence_breached;
    this._mission_failureIntialized = true;
    this._refObject.mission_failure = refObject.mission_failure;
    this._vtol_fixed_wing_system_failureIntialized = true;
    this._refObject.vtol_fixed_wing_system_failure = refObject.vtol_fixed_wing_system_failure;
    this._wind_limit_exceededIntialized = true;
    this._refObject.wind_limit_exceeded = refObject.wind_limit_exceeded;
    this._flight_time_limit_exceededIntialized = true;
    this._refObject.flight_time_limit_exceeded = refObject.flight_time_limit_exceeded;
    this._local_position_accuracy_lowIntialized = true;
    this._refObject.local_position_accuracy_low = refObject.local_position_accuracy_low;
    this._fd_critical_failureIntialized = true;
    this._refObject.fd_critical_failure = refObject.fd_critical_failure;
    this._fd_esc_arming_failureIntialized = true;
    this._refObject.fd_esc_arming_failure = refObject.fd_esc_arming_failure;
    this._fd_imbalanced_propIntialized = true;
    this._refObject.fd_imbalanced_prop = refObject.fd_imbalanced_prop;
    this._fd_motor_failureIntialized = true;
    this._refObject.fd_motor_failure = refObject.fd_motor_failure;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    FailsafeFlagsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'FailsafeFlags'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get mode_req_angular_velocity() {
    if (!this._mode_req_angular_velocityIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_angular_velocity;
  }
  set mode_req_angular_velocity(value) {
    this._mode_req_angular_velocityIntialized = true;
    this._refObject.mode_req_angular_velocity = value;
  }
  get mode_req_attitude() {
    if (!this._mode_req_attitudeIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_attitude;
  }
  set mode_req_attitude(value) {
    this._mode_req_attitudeIntialized = true;
    this._refObject.mode_req_attitude = value;
  }
  get mode_req_local_alt() {
    if (!this._mode_req_local_altIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_local_alt;
  }
  set mode_req_local_alt(value) {
    this._mode_req_local_altIntialized = true;
    this._refObject.mode_req_local_alt = value;
  }
  get mode_req_local_position() {
    if (!this._mode_req_local_positionIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_local_position;
  }
  set mode_req_local_position(value) {
    this._mode_req_local_positionIntialized = true;
    this._refObject.mode_req_local_position = value;
  }
  get mode_req_local_position_relaxed() {
    if (!this._mode_req_local_position_relaxedIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_local_position_relaxed;
  }
  set mode_req_local_position_relaxed(value) {
    this._mode_req_local_position_relaxedIntialized = true;
    this._refObject.mode_req_local_position_relaxed = value;
  }
  get mode_req_global_position() {
    if (!this._mode_req_global_positionIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_global_position;
  }
  set mode_req_global_position(value) {
    this._mode_req_global_positionIntialized = true;
    this._refObject.mode_req_global_position = value;
  }
  get mode_req_mission() {
    if (!this._mode_req_missionIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_mission;
  }
  set mode_req_mission(value) {
    this._mode_req_missionIntialized = true;
    this._refObject.mode_req_mission = value;
  }
  get mode_req_offboard_signal() {
    if (!this._mode_req_offboard_signalIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_offboard_signal;
  }
  set mode_req_offboard_signal(value) {
    this._mode_req_offboard_signalIntialized = true;
    this._refObject.mode_req_offboard_signal = value;
  }
  get mode_req_home_position() {
    if (!this._mode_req_home_positionIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_home_position;
  }
  set mode_req_home_position(value) {
    this._mode_req_home_positionIntialized = true;
    this._refObject.mode_req_home_position = value;
  }
  get mode_req_wind_and_flight_time_compliance() {
    if (!this._mode_req_wind_and_flight_time_complianceIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_wind_and_flight_time_compliance;
  }
  set mode_req_wind_and_flight_time_compliance(value) {
    this._mode_req_wind_and_flight_time_complianceIntialized = true;
    this._refObject.mode_req_wind_and_flight_time_compliance = value;
  }
  get mode_req_prevent_arming() {
    if (!this._mode_req_prevent_armingIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_prevent_arming;
  }
  set mode_req_prevent_arming(value) {
    this._mode_req_prevent_armingIntialized = true;
    this._refObject.mode_req_prevent_arming = value;
  }
  get mode_req_manual_control() {
    if (!this._mode_req_manual_controlIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_manual_control;
  }
  set mode_req_manual_control(value) {
    this._mode_req_manual_controlIntialized = true;
    this._refObject.mode_req_manual_control = value;
  }
  get mode_req_other() {
    if (!this._mode_req_otherIntialized) {
      return undefined;
    }
    return this._refObject.mode_req_other;
  }
  set mode_req_other(value) {
    this._mode_req_otherIntialized = true;
    this._refObject.mode_req_other = value;
  }
  get angular_velocity_invalid() {
    if (!this._angular_velocity_invalidIntialized) {
      return undefined;
    }
    return this._refObject.angular_velocity_invalid;
  }
  set angular_velocity_invalid(value) {
    this._angular_velocity_invalidIntialized = true;
    this._refObject.angular_velocity_invalid = value;
  }
  get attitude_invalid() {
    if (!this._attitude_invalidIntialized) {
      return undefined;
    }
    return this._refObject.attitude_invalid;
  }
  set attitude_invalid(value) {
    this._attitude_invalidIntialized = true;
    this._refObject.attitude_invalid = value;
  }
  get local_altitude_invalid() {
    if (!this._local_altitude_invalidIntialized) {
      return undefined;
    }
    return this._refObject.local_altitude_invalid;
  }
  set local_altitude_invalid(value) {
    this._local_altitude_invalidIntialized = true;
    this._refObject.local_altitude_invalid = value;
  }
  get local_position_invalid() {
    if (!this._local_position_invalidIntialized) {
      return undefined;
    }
    return this._refObject.local_position_invalid;
  }
  set local_position_invalid(value) {
    this._local_position_invalidIntialized = true;
    this._refObject.local_position_invalid = value;
  }
  get local_position_invalid_relaxed() {
    if (!this._local_position_invalid_relaxedIntialized) {
      return undefined;
    }
    return this._refObject.local_position_invalid_relaxed;
  }
  set local_position_invalid_relaxed(value) {
    this._local_position_invalid_relaxedIntialized = true;
    this._refObject.local_position_invalid_relaxed = value;
  }
  get local_velocity_invalid() {
    if (!this._local_velocity_invalidIntialized) {
      return undefined;
    }
    return this._refObject.local_velocity_invalid;
  }
  set local_velocity_invalid(value) {
    this._local_velocity_invalidIntialized = true;
    this._refObject.local_velocity_invalid = value;
  }
  get global_position_invalid() {
    if (!this._global_position_invalidIntialized) {
      return undefined;
    }
    return this._refObject.global_position_invalid;
  }
  set global_position_invalid(value) {
    this._global_position_invalidIntialized = true;
    this._refObject.global_position_invalid = value;
  }
  get auto_mission_missing() {
    if (!this._auto_mission_missingIntialized) {
      return undefined;
    }
    return this._refObject.auto_mission_missing;
  }
  set auto_mission_missing(value) {
    this._auto_mission_missingIntialized = true;
    this._refObject.auto_mission_missing = value;
  }
  get offboard_control_signal_lost() {
    if (!this._offboard_control_signal_lostIntialized) {
      return undefined;
    }
    return this._refObject.offboard_control_signal_lost;
  }
  set offboard_control_signal_lost(value) {
    this._offboard_control_signal_lostIntialized = true;
    this._refObject.offboard_control_signal_lost = value;
  }
  get home_position_invalid() {
    if (!this._home_position_invalidIntialized) {
      return undefined;
    }
    return this._refObject.home_position_invalid;
  }
  set home_position_invalid(value) {
    this._home_position_invalidIntialized = true;
    this._refObject.home_position_invalid = value;
  }
  get manual_control_signal_lost() {
    if (!this._manual_control_signal_lostIntialized) {
      return undefined;
    }
    return this._refObject.manual_control_signal_lost;
  }
  set manual_control_signal_lost(value) {
    this._manual_control_signal_lostIntialized = true;
    this._refObject.manual_control_signal_lost = value;
  }
  get gcs_connection_lost() {
    if (!this._gcs_connection_lostIntialized) {
      return undefined;
    }
    return this._refObject.gcs_connection_lost;
  }
  set gcs_connection_lost(value) {
    this._gcs_connection_lostIntialized = true;
    this._refObject.gcs_connection_lost = value;
  }
  get battery_warning() {
    if (!this._battery_warningIntialized) {
      return undefined;
    }
    return this._refObject.battery_warning;
  }
  set battery_warning(value) {
    this._battery_warningIntialized = true;
    this._refObject.battery_warning = value;
  }
  get battery_low_remaining_time() {
    if (!this._battery_low_remaining_timeIntialized) {
      return undefined;
    }
    return this._refObject.battery_low_remaining_time;
  }
  set battery_low_remaining_time(value) {
    this._battery_low_remaining_timeIntialized = true;
    this._refObject.battery_low_remaining_time = value;
  }
  get battery_unhealthy() {
    if (!this._battery_unhealthyIntialized) {
      return undefined;
    }
    return this._refObject.battery_unhealthy;
  }
  set battery_unhealthy(value) {
    this._battery_unhealthyIntialized = true;
    this._refObject.battery_unhealthy = value;
  }
  get primary_geofence_breached() {
    if (!this._primary_geofence_breachedIntialized) {
      return undefined;
    }
    return this._refObject.primary_geofence_breached;
  }
  set primary_geofence_breached(value) {
    this._primary_geofence_breachedIntialized = true;
    this._refObject.primary_geofence_breached = value;
  }
  get mission_failure() {
    if (!this._mission_failureIntialized) {
      return undefined;
    }
    return this._refObject.mission_failure;
  }
  set mission_failure(value) {
    this._mission_failureIntialized = true;
    this._refObject.mission_failure = value;
  }
  get vtol_fixed_wing_system_failure() {
    if (!this._vtol_fixed_wing_system_failureIntialized) {
      return undefined;
    }
    return this._refObject.vtol_fixed_wing_system_failure;
  }
  set vtol_fixed_wing_system_failure(value) {
    this._vtol_fixed_wing_system_failureIntialized = true;
    this._refObject.vtol_fixed_wing_system_failure = value;
  }
  get wind_limit_exceeded() {
    if (!this._wind_limit_exceededIntialized) {
      return undefined;
    }
    return this._refObject.wind_limit_exceeded;
  }
  set wind_limit_exceeded(value) {
    this._wind_limit_exceededIntialized = true;
    this._refObject.wind_limit_exceeded = value;
  }
  get flight_time_limit_exceeded() {
    if (!this._flight_time_limit_exceededIntialized) {
      return undefined;
    }
    return this._refObject.flight_time_limit_exceeded;
  }
  set flight_time_limit_exceeded(value) {
    this._flight_time_limit_exceededIntialized = true;
    this._refObject.flight_time_limit_exceeded = value;
  }
  get local_position_accuracy_low() {
    if (!this._local_position_accuracy_lowIntialized) {
      return undefined;
    }
    return this._refObject.local_position_accuracy_low;
  }
  set local_position_accuracy_low(value) {
    this._local_position_accuracy_lowIntialized = true;
    this._refObject.local_position_accuracy_low = value;
  }
  get fd_critical_failure() {
    if (!this._fd_critical_failureIntialized) {
      return undefined;
    }
    return this._refObject.fd_critical_failure;
  }
  set fd_critical_failure(value) {
    this._fd_critical_failureIntialized = true;
    this._refObject.fd_critical_failure = value;
  }
  get fd_esc_arming_failure() {
    if (!this._fd_esc_arming_failureIntialized) {
      return undefined;
    }
    return this._refObject.fd_esc_arming_failure;
  }
  set fd_esc_arming_failure(value) {
    this._fd_esc_arming_failureIntialized = true;
    this._refObject.fd_esc_arming_failure = value;
  }
  get fd_imbalanced_prop() {
    if (!this._fd_imbalanced_propIntialized) {
      return undefined;
    }
    return this._refObject.fd_imbalanced_prop;
  }
  set fd_imbalanced_prop(value) {
    this._fd_imbalanced_propIntialized = true;
    this._refObject.fd_imbalanced_prop = value;
  }
  get fd_motor_failure() {
    if (!this._fd_motor_failureIntialized) {
      return undefined;
    }
    return this._refObject.fd_motor_failure;
  }
  set fd_motor_failure(value) {
    this._fd_motor_failureIntialized = true;
    this._refObject.fd_motor_failure = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FailsafeFlagsRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._mode_req_angular_velocityIntialized = true;
    this._mode_req_attitudeIntialized = true;
    this._mode_req_local_altIntialized = true;
    this._mode_req_local_positionIntialized = true;
    this._mode_req_local_position_relaxedIntialized = true;
    this._mode_req_global_positionIntialized = true;
    this._mode_req_missionIntialized = true;
    this._mode_req_offboard_signalIntialized = true;
    this._mode_req_home_positionIntialized = true;
    this._mode_req_wind_and_flight_time_complianceIntialized = true;
    this._mode_req_prevent_armingIntialized = true;
    this._mode_req_manual_controlIntialized = true;
    this._mode_req_otherIntialized = true;
    this._angular_velocity_invalidIntialized = true;
    this._attitude_invalidIntialized = true;
    this._local_altitude_invalidIntialized = true;
    this._local_position_invalidIntialized = true;
    this._local_position_invalid_relaxedIntialized = true;
    this._local_velocity_invalidIntialized = true;
    this._global_position_invalidIntialized = true;
    this._auto_mission_missingIntialized = true;
    this._offboard_control_signal_lostIntialized = true;
    this._home_position_invalidIntialized = true;
    this._manual_control_signal_lostIntialized = true;
    this._gcs_connection_lostIntialized = true;
    this._battery_warningIntialized = true;
    this._battery_low_remaining_timeIntialized = true;
    this._battery_unhealthyIntialized = true;
    this._primary_geofence_breachedIntialized = true;
    this._mission_failureIntialized = true;
    this._vtol_fixed_wing_system_failureIntialized = true;
    this._wind_limit_exceededIntialized = true;
    this._flight_time_limit_exceededIntialized = true;
    this._local_position_accuracy_lowIntialized = true;
    this._fd_critical_failureIntialized = true;
    this._fd_esc_arming_failureIntialized = true;
    this._fd_imbalanced_propIntialized = true;
    this._fd_motor_failureIntialized = true;
  }
  copy(other) {
    this._refObject = new FailsafeFlagsRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._mode_req_angular_velocityIntialized = true;
    this._mode_req_attitudeIntialized = true;
    this._mode_req_local_altIntialized = true;
    this._mode_req_local_positionIntialized = true;
    this._mode_req_local_position_relaxedIntialized = true;
    this._mode_req_global_positionIntialized = true;
    this._mode_req_missionIntialized = true;
    this._mode_req_offboard_signalIntialized = true;
    this._mode_req_home_positionIntialized = true;
    this._mode_req_wind_and_flight_time_complianceIntialized = true;
    this._mode_req_prevent_armingIntialized = true;
    this._mode_req_manual_controlIntialized = true;
    this._mode_req_otherIntialized = true;
    this._angular_velocity_invalidIntialized = true;
    this._attitude_invalidIntialized = true;
    this._local_altitude_invalidIntialized = true;
    this._local_position_invalidIntialized = true;
    this._local_position_invalid_relaxedIntialized = true;
    this._local_velocity_invalidIntialized = true;
    this._global_position_invalidIntialized = true;
    this._auto_mission_missingIntialized = true;
    this._offboard_control_signal_lostIntialized = true;
    this._home_position_invalidIntialized = true;
    this._manual_control_signal_lostIntialized = true;
    this._gcs_connection_lostIntialized = true;
    this._battery_warningIntialized = true;
    this._battery_low_remaining_timeIntialized = true;
    this._battery_unhealthyIntialized = true;
    this._primary_geofence_breachedIntialized = true;
    this._mission_failureIntialized = true;
    this._vtol_fixed_wing_system_failureIntialized = true;
    this._wind_limit_exceededIntialized = true;
    this._flight_time_limit_exceededIntialized = true;
    this._local_position_accuracy_lowIntialized = true;
    this._fd_critical_failureIntialized = true;
    this._fd_esc_arming_failureIntialized = true;
    this._fd_imbalanced_propIntialized = true;
    this._fd_motor_failureIntialized = true;
  }
  static get classType() {
    return FailsafeFlagsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_angular_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_attitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_local_alt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_local_position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_local_position_relaxed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_global_position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_mission","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_offboard_signal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_home_position","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_wind_and_flight_time_compliance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_prevent_arming","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_manual_control","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode_req_other","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular_velocity_invalid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"attitude_invalid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"local_altitude_invalid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"local_position_invalid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"local_position_invalid_relaxed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"local_velocity_invalid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"global_position_invalid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"auto_mission_missing","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"offboard_control_signal_lost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"home_position_invalid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"manual_control_signal_lost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gcs_connection_lost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"battery_warning","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"battery_low_remaining_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"battery_unhealthy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"primary_geofence_breached","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mission_failure","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vtol_fixed_wing_system_failure","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"wind_limit_exceeded","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flight_time_limit_exceeded","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"local_position_accuracy_low","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fd_critical_failure","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fd_esc_arming_failure","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fd_imbalanced_prop","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fd_motor_failure","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"FailsafeFlags","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FailsafeFlags"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","mode_req_angular_velocity","mode_req_attitude","mode_req_local_alt","mode_req_local_position","mode_req_local_position_relaxed","mode_req_global_position","mode_req_mission","mode_req_offboard_signal","mode_req_home_position","mode_req_wind_and_flight_time_compliance","mode_req_prevent_arming","mode_req_manual_control","mode_req_other","angular_velocity_invalid","attitude_invalid","local_altitude_invalid","local_position_invalid","local_position_invalid_relaxed","local_velocity_invalid","global_position_invalid","auto_mission_missing","offboard_control_signal_lost","home_position_invalid","manual_control_signal_lost","gcs_connection_lost","battery_warning","battery_low_remaining_time","battery_unhealthy","primary_geofence_breached","mission_failure","vtol_fixed_wing_system_failure","wind_limit_exceeded","flight_time_limit_exceeded","local_position_accuracy_low","fd_critical_failure","fd_esc_arming_failure","fd_imbalanced_prop","fd_motor_failure"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FailsafeFlagsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FailsafeFlagsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FailsafeFlagsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FailsafeFlagsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FailsafeFlagsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FailsafeFlagsRefArray(size);
    this._refObject = new FailsafeFlagsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FailsafeFlagsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FailsafeFlagsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FailsafeFlagsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FailsafeFlagsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FailsafeFlagsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FailsafeFlagsArrayWrapper;
  }
}
module.exports = FailsafeFlagsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_angular_velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_attitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_local_alt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_local_position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_local_position_relaxed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_global_position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_mission",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_offboard_signal",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_home_position",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_wind_and_flight_time_compliance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_prevent_arming",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_manual_control",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode_req_other",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angular_velocity_invalid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "attitude_invalid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "local_altitude_invalid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "local_position_invalid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "local_position_invalid_relaxed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "local_velocity_invalid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "global_position_invalid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "auto_mission_missing",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "offboard_control_signal_lost",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "home_position_invalid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "manual_control_signal_lost",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gcs_connection_lost",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "battery_warning",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "battery_low_remaining_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "battery_unhealthy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "primary_geofence_breached",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mission_failure",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vtol_fixed_wing_system_failure",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "wind_limit_exceeded",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flight_time_limit_exceeded",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "local_position_accuracy_low",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fd_critical_failure",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fd_esc_arming_failure",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fd_imbalanced_prop",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fd_motor_failure",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "FailsafeFlags",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FailsafeFlags"
}
*/
