// rclnodejs/generated/px4_msgs/px4_msgs__msg__Mission.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MissionRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  dataman_id: primitiveTypes.uint8,
  count: primitiveTypes.uint16,
  current_seq: primitiveTypes.int32,
});
const MissionRefArray = ArrayType(MissionRefStruct);
const MissionRefStructArray = StructType({
  data: MissionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MissionWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MissionRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._dataman_idIntialized = true;
      this._countIntialized = true;
      this._current_seqIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MissionRefStruct();
    this._timestampIntialized = false;
    this._dataman_idIntialized = false;
    this._countIntialized = false;
    this._current_seqIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MissionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MissionArrayWrapper;
  }
  static get ArrayType() {
    return MissionArrayWrapper;
  }
  static get refObjectArrayType() {
    return MissionRefStructArray
  }
  static get refObjectType() {
    return MissionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in Mission');
      }
      if (checkConsistency && !this._dataman_idIntialized) {
        throw new TypeError('Invalid argument: dataman_id in Mission');
      }
      if (checkConsistency && !this._countIntialized) {
        throw new TypeError('Invalid argument: count in Mission');
      }
      if (checkConsistency && !this._current_seqIntialized) {
        throw new TypeError('Invalid argument: current_seq in Mission');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._dataman_idIntialized = true;
    this._refObject.dataman_id = refObject.dataman_id;
    this._countIntialized = true;
    this._refObject.count = refObject.count;
    this._current_seqIntialized = true;
    this._refObject.current_seq = refObject.current_seq;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    MissionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'Mission'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get dataman_id() {
    if (!this._dataman_idIntialized) {
      return undefined;
    }
    return this._refObject.dataman_id;
  }
  set dataman_id(value) {
    this._dataman_idIntialized = true;
    this._refObject.dataman_id = value;
  }
  get count() {
    if (!this._countIntialized) {
      return undefined;
    }
    return this._refObject.count;
  }
  set count(value) {
    this._countIntialized = true;
    this._refObject.count = value;
  }
  get current_seq() {
    if (!this._current_seqIntialized) {
      return undefined;
    }
    return this._refObject.current_seq;
  }
  set current_seq(value) {
    this._current_seqIntialized = true;
    this._refObject.current_seq = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MissionRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._dataman_idIntialized = true;
    this._countIntialized = true;
    this._current_seqIntialized = true;
  }
  copy(other) {
    this._refObject = new MissionRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._dataman_idIntialized = true;
    this._countIntialized = true;
    this._current_seqIntialized = true;
  }
  static get classType() {
    return MissionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dataman_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_seq","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"Mission","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Mission"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","dataman_id","count","current_seq"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MissionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MissionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MissionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MissionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MissionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MissionRefArray(size);
    this._refObject = new MissionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MissionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MissionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MissionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MissionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MissionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MissionArrayWrapper;
  }
}
module.exports = MissionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dataman_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current_seq",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "Mission",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Mission"
}
*/
