// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleLocalPosition.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VehicleLocalPositionRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  xy_valid: primitiveTypes.bool,
  z_valid: primitiveTypes.bool,
  v_xy_valid: primitiveTypes.bool,
  v_z_valid: primitiveTypes.bool,
  x: primitiveTypes.float32,
  y: primitiveTypes.float32,
  z: primitiveTypes.float32,
  delta_xy: ArrayType(primitiveTypes.float32, 2),
  xy_reset_counter: primitiveTypes.uint8,
  delta_z: primitiveTypes.float32,
  z_reset_counter: primitiveTypes.uint8,
  vx: primitiveTypes.float32,
  vy: primitiveTypes.float32,
  vz: primitiveTypes.float32,
  z_deriv: primitiveTypes.float32,
  delta_vxy: ArrayType(primitiveTypes.float32, 2),
  vxy_reset_counter: primitiveTypes.uint8,
  delta_vz: primitiveTypes.float32,
  vz_reset_counter: primitiveTypes.uint8,
  ax: primitiveTypes.float32,
  ay: primitiveTypes.float32,
  az: primitiveTypes.float32,
  heading: primitiveTypes.float32,
  delta_heading: primitiveTypes.float32,
  heading_reset_counter: primitiveTypes.uint8,
  heading_good_for_control: primitiveTypes.bool,
  xy_global: primitiveTypes.bool,
  z_global: primitiveTypes.bool,
  ref_timestamp: primitiveTypes.uint64,
  ref_lat: primitiveTypes.float64,
  ref_lon: primitiveTypes.float64,
  ref_alt: primitiveTypes.float32,
  dist_bottom: primitiveTypes.float32,
  dist_bottom_valid: primitiveTypes.bool,
  dist_bottom_sensor_bitfield: primitiveTypes.uint8,
  eph: primitiveTypes.float32,
  epv: primitiveTypes.float32,
  evh: primitiveTypes.float32,
  evv: primitiveTypes.float32,
  dead_reckoning: primitiveTypes.bool,
  vxy_max: primitiveTypes.float32,
  vz_max: primitiveTypes.float32,
  hagl_min: primitiveTypes.float32,
  hagl_max: primitiveTypes.float32,
});
const VehicleLocalPositionRefArray = ArrayType(VehicleLocalPositionRefStruct);
const VehicleLocalPositionRefStructArray = StructType({
  data: VehicleLocalPositionRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleLocalPositionWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleLocalPositionRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._xy_validIntialized = true;
      this._z_validIntialized = true;
      this._v_xy_validIntialized = true;
      this._v_z_validIntialized = true;
      this._xIntialized = true;
      this._yIntialized = true;
      this._zIntialized = true;
      this._wrapperFields.delta_xy = Float32Wrapper.createArray();
      this._wrapperFields.delta_xy.copy(other._wrapperFields.delta_xy);
      this._xy_reset_counterIntialized = true;
      this._delta_zIntialized = true;
      this._z_reset_counterIntialized = true;
      this._vxIntialized = true;
      this._vyIntialized = true;
      this._vzIntialized = true;
      this._z_derivIntialized = true;
      this._wrapperFields.delta_vxy = Float32Wrapper.createArray();
      this._wrapperFields.delta_vxy.copy(other._wrapperFields.delta_vxy);
      this._vxy_reset_counterIntialized = true;
      this._delta_vzIntialized = true;
      this._vz_reset_counterIntialized = true;
      this._axIntialized = true;
      this._ayIntialized = true;
      this._azIntialized = true;
      this._headingIntialized = true;
      this._delta_headingIntialized = true;
      this._heading_reset_counterIntialized = true;
      this._heading_good_for_controlIntialized = true;
      this._xy_globalIntialized = true;
      this._z_globalIntialized = true;
      this._ref_timestampIntialized = true;
      this._ref_latIntialized = true;
      this._ref_lonIntialized = true;
      this._ref_altIntialized = true;
      this._dist_bottomIntialized = true;
      this._dist_bottom_validIntialized = true;
      this._dist_bottom_sensor_bitfieldIntialized = true;
      this._ephIntialized = true;
      this._epvIntialized = true;
      this._evhIntialized = true;
      this._evvIntialized = true;
      this._dead_reckoningIntialized = true;
      this._vxy_maxIntialized = true;
      this._vz_maxIntialized = true;
      this._hagl_minIntialized = true;
      this._hagl_maxIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleLocalPositionRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._xy_validIntialized = false;
    this._z_validIntialized = false;
    this._v_xy_validIntialized = false;
    this._v_z_validIntialized = false;
    this._xIntialized = false;
    this._yIntialized = false;
    this._zIntialized = false;
    this._wrapperFields.delta_xy = Float32Wrapper.createArray();
    this._xy_reset_counterIntialized = false;
    this._delta_zIntialized = false;
    this._z_reset_counterIntialized = false;
    this._vxIntialized = false;
    this._vyIntialized = false;
    this._vzIntialized = false;
    this._z_derivIntialized = false;
    this._wrapperFields.delta_vxy = Float32Wrapper.createArray();
    this._vxy_reset_counterIntialized = false;
    this._delta_vzIntialized = false;
    this._vz_reset_counterIntialized = false;
    this._axIntialized = false;
    this._ayIntialized = false;
    this._azIntialized = false;
    this._headingIntialized = false;
    this._delta_headingIntialized = false;
    this._heading_reset_counterIntialized = false;
    this._heading_good_for_controlIntialized = false;
    this._xy_globalIntialized = false;
    this._z_globalIntialized = false;
    this._ref_timestampIntialized = false;
    this._ref_latIntialized = false;
    this._ref_lonIntialized = false;
    this._ref_altIntialized = false;
    this._dist_bottomIntialized = false;
    this._dist_bottom_validIntialized = false;
    this._dist_bottom_sensor_bitfieldIntialized = false;
    this._ephIntialized = false;
    this._epvIntialized = false;
    this._evhIntialized = false;
    this._evvIntialized = false;
    this._dead_reckoningIntialized = false;
    this._vxy_maxIntialized = false;
    this._vz_maxIntialized = false;
    this._hagl_minIntialized = false;
    this._hagl_maxIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleLocalPositionWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleLocalPositionArrayWrapper;
  }
  static get ArrayType() {
    return VehicleLocalPositionArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleLocalPositionRefStructArray
  }
  static get refObjectType() {
    return VehicleLocalPositionRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleLocalPosition');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in VehicleLocalPosition');
      }
      if (checkConsistency && !this._xy_validIntialized) {
        throw new TypeError('Invalid argument: xy_valid in VehicleLocalPosition');
      }
      if (checkConsistency && !this._z_validIntialized) {
        throw new TypeError('Invalid argument: z_valid in VehicleLocalPosition');
      }
      if (checkConsistency && !this._v_xy_validIntialized) {
        throw new TypeError('Invalid argument: v_xy_valid in VehicleLocalPosition');
      }
      if (checkConsistency && !this._v_z_validIntialized) {
        throw new TypeError('Invalid argument: v_z_valid in VehicleLocalPosition');
      }
      if (checkConsistency && !this._xIntialized) {
        throw new TypeError('Invalid argument: x in VehicleLocalPosition');
      }
      if (checkConsistency && !this._yIntialized) {
        throw new TypeError('Invalid argument: y in VehicleLocalPosition');
      }
      if (checkConsistency && !this._zIntialized) {
        throw new TypeError('Invalid argument: z in VehicleLocalPosition');
      }
      if (checkConsistency && !this._xy_reset_counterIntialized) {
        throw new TypeError('Invalid argument: xy_reset_counter in VehicleLocalPosition');
      }
      if (checkConsistency && !this._delta_zIntialized) {
        throw new TypeError('Invalid argument: delta_z in VehicleLocalPosition');
      }
      if (checkConsistency && !this._z_reset_counterIntialized) {
        throw new TypeError('Invalid argument: z_reset_counter in VehicleLocalPosition');
      }
      if (checkConsistency && !this._vxIntialized) {
        throw new TypeError('Invalid argument: vx in VehicleLocalPosition');
      }
      if (checkConsistency && !this._vyIntialized) {
        throw new TypeError('Invalid argument: vy in VehicleLocalPosition');
      }
      if (checkConsistency && !this._vzIntialized) {
        throw new TypeError('Invalid argument: vz in VehicleLocalPosition');
      }
      if (checkConsistency && !this._z_derivIntialized) {
        throw new TypeError('Invalid argument: z_deriv in VehicleLocalPosition');
      }
      if (checkConsistency && !this._vxy_reset_counterIntialized) {
        throw new TypeError('Invalid argument: vxy_reset_counter in VehicleLocalPosition');
      }
      if (checkConsistency && !this._delta_vzIntialized) {
        throw new TypeError('Invalid argument: delta_vz in VehicleLocalPosition');
      }
      if (checkConsistency && !this._vz_reset_counterIntialized) {
        throw new TypeError('Invalid argument: vz_reset_counter in VehicleLocalPosition');
      }
      if (checkConsistency && !this._axIntialized) {
        throw new TypeError('Invalid argument: ax in VehicleLocalPosition');
      }
      if (checkConsistency && !this._ayIntialized) {
        throw new TypeError('Invalid argument: ay in VehicleLocalPosition');
      }
      if (checkConsistency && !this._azIntialized) {
        throw new TypeError('Invalid argument: az in VehicleLocalPosition');
      }
      if (checkConsistency && !this._headingIntialized) {
        throw new TypeError('Invalid argument: heading in VehicleLocalPosition');
      }
      if (checkConsistency && !this._delta_headingIntialized) {
        throw new TypeError('Invalid argument: delta_heading in VehicleLocalPosition');
      }
      if (checkConsistency && !this._heading_reset_counterIntialized) {
        throw new TypeError('Invalid argument: heading_reset_counter in VehicleLocalPosition');
      }
      if (checkConsistency && !this._heading_good_for_controlIntialized) {
        throw new TypeError('Invalid argument: heading_good_for_control in VehicleLocalPosition');
      }
      if (checkConsistency && !this._xy_globalIntialized) {
        throw new TypeError('Invalid argument: xy_global in VehicleLocalPosition');
      }
      if (checkConsistency && !this._z_globalIntialized) {
        throw new TypeError('Invalid argument: z_global in VehicleLocalPosition');
      }
      if (checkConsistency && !this._ref_timestampIntialized) {
        throw new TypeError('Invalid argument: ref_timestamp in VehicleLocalPosition');
      }
      if (checkConsistency && !this._ref_latIntialized) {
        throw new TypeError('Invalid argument: ref_lat in VehicleLocalPosition');
      }
      if (checkConsistency && !this._ref_lonIntialized) {
        throw new TypeError('Invalid argument: ref_lon in VehicleLocalPosition');
      }
      if (checkConsistency && !this._ref_altIntialized) {
        throw new TypeError('Invalid argument: ref_alt in VehicleLocalPosition');
      }
      if (checkConsistency && !this._dist_bottomIntialized) {
        throw new TypeError('Invalid argument: dist_bottom in VehicleLocalPosition');
      }
      if (checkConsistency && !this._dist_bottom_validIntialized) {
        throw new TypeError('Invalid argument: dist_bottom_valid in VehicleLocalPosition');
      }
      if (checkConsistency && !this._dist_bottom_sensor_bitfieldIntialized) {
        throw new TypeError('Invalid argument: dist_bottom_sensor_bitfield in VehicleLocalPosition');
      }
      if (checkConsistency && !this._ephIntialized) {
        throw new TypeError('Invalid argument: eph in VehicleLocalPosition');
      }
      if (checkConsistency && !this._epvIntialized) {
        throw new TypeError('Invalid argument: epv in VehicleLocalPosition');
      }
      if (checkConsistency && !this._evhIntialized) {
        throw new TypeError('Invalid argument: evh in VehicleLocalPosition');
      }
      if (checkConsistency && !this._evvIntialized) {
        throw new TypeError('Invalid argument: evv in VehicleLocalPosition');
      }
      if (checkConsistency && !this._dead_reckoningIntialized) {
        throw new TypeError('Invalid argument: dead_reckoning in VehicleLocalPosition');
      }
      if (checkConsistency && !this._vxy_maxIntialized) {
        throw new TypeError('Invalid argument: vxy_max in VehicleLocalPosition');
      }
      if (checkConsistency && !this._vz_maxIntialized) {
        throw new TypeError('Invalid argument: vz_max in VehicleLocalPosition');
      }
      if (checkConsistency && !this._hagl_minIntialized) {
        throw new TypeError('Invalid argument: hagl_min in VehicleLocalPosition');
      }
      if (checkConsistency && !this._hagl_maxIntialized) {
        throw new TypeError('Invalid argument: hagl_max in VehicleLocalPosition');
      }
    this._refObject.delta_xy = Array.from(this._wrapperFields.delta_xy.data);
    this._refObject.delta_vxy = Array.from(this._wrapperFields.delta_vxy.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._xy_validIntialized = true;
    this._refObject.xy_valid = refObject.xy_valid;
    this._z_validIntialized = true;
    this._refObject.z_valid = refObject.z_valid;
    this._v_xy_validIntialized = true;
    this._refObject.v_xy_valid = refObject.v_xy_valid;
    this._v_z_validIntialized = true;
    this._refObject.v_z_valid = refObject.v_z_valid;
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._zIntialized = true;
    this._refObject.z = refObject.z;
    this._wrapperFields.delta_xy.fill(refObject.delta_xy.toArray());
    this._xy_reset_counterIntialized = true;
    this._refObject.xy_reset_counter = refObject.xy_reset_counter;
    this._delta_zIntialized = true;
    this._refObject.delta_z = refObject.delta_z;
    this._z_reset_counterIntialized = true;
    this._refObject.z_reset_counter = refObject.z_reset_counter;
    this._vxIntialized = true;
    this._refObject.vx = refObject.vx;
    this._vyIntialized = true;
    this._refObject.vy = refObject.vy;
    this._vzIntialized = true;
    this._refObject.vz = refObject.vz;
    this._z_derivIntialized = true;
    this._refObject.z_deriv = refObject.z_deriv;
    this._wrapperFields.delta_vxy.fill(refObject.delta_vxy.toArray());
    this._vxy_reset_counterIntialized = true;
    this._refObject.vxy_reset_counter = refObject.vxy_reset_counter;
    this._delta_vzIntialized = true;
    this._refObject.delta_vz = refObject.delta_vz;
    this._vz_reset_counterIntialized = true;
    this._refObject.vz_reset_counter = refObject.vz_reset_counter;
    this._axIntialized = true;
    this._refObject.ax = refObject.ax;
    this._ayIntialized = true;
    this._refObject.ay = refObject.ay;
    this._azIntialized = true;
    this._refObject.az = refObject.az;
    this._headingIntialized = true;
    this._refObject.heading = refObject.heading;
    this._delta_headingIntialized = true;
    this._refObject.delta_heading = refObject.delta_heading;
    this._heading_reset_counterIntialized = true;
    this._refObject.heading_reset_counter = refObject.heading_reset_counter;
    this._heading_good_for_controlIntialized = true;
    this._refObject.heading_good_for_control = refObject.heading_good_for_control;
    this._xy_globalIntialized = true;
    this._refObject.xy_global = refObject.xy_global;
    this._z_globalIntialized = true;
    this._refObject.z_global = refObject.z_global;
    this._ref_timestampIntialized = true;
    this._refObject.ref_timestamp = refObject.ref_timestamp;
    this._ref_latIntialized = true;
    this._refObject.ref_lat = refObject.ref_lat;
    this._ref_lonIntialized = true;
    this._refObject.ref_lon = refObject.ref_lon;
    this._ref_altIntialized = true;
    this._refObject.ref_alt = refObject.ref_alt;
    this._dist_bottomIntialized = true;
    this._refObject.dist_bottom = refObject.dist_bottom;
    this._dist_bottom_validIntialized = true;
    this._refObject.dist_bottom_valid = refObject.dist_bottom_valid;
    this._dist_bottom_sensor_bitfieldIntialized = true;
    this._refObject.dist_bottom_sensor_bitfield = refObject.dist_bottom_sensor_bitfield;
    this._ephIntialized = true;
    this._refObject.eph = refObject.eph;
    this._epvIntialized = true;
    this._refObject.epv = refObject.epv;
    this._evhIntialized = true;
    this._refObject.evh = refObject.evh;
    this._evvIntialized = true;
    this._refObject.evv = refObject.evv;
    this._dead_reckoningIntialized = true;
    this._refObject.dead_reckoning = refObject.dead_reckoning;
    this._vxy_maxIntialized = true;
    this._refObject.vxy_max = refObject.vxy_max;
    this._vz_maxIntialized = true;
    this._refObject.vz_max = refObject.vz_max;
    this._hagl_minIntialized = true;
    this._refObject.hagl_min = refObject.hagl_min;
    this._hagl_maxIntialized = true;
    this._refObject.hagl_max = refObject.hagl_max;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleLocalPositionWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleLocalPosition'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get xy_valid() {
    if (!this._xy_validIntialized) {
      return undefined;
    }
    return this._refObject.xy_valid;
  }
  set xy_valid(value) {
    this._xy_validIntialized = true;
    this._refObject.xy_valid = value;
  }
  get z_valid() {
    if (!this._z_validIntialized) {
      return undefined;
    }
    return this._refObject.z_valid;
  }
  set z_valid(value) {
    this._z_validIntialized = true;
    this._refObject.z_valid = value;
  }
  get v_xy_valid() {
    if (!this._v_xy_validIntialized) {
      return undefined;
    }
    return this._refObject.v_xy_valid;
  }
  set v_xy_valid(value) {
    this._v_xy_validIntialized = true;
    this._refObject.v_xy_valid = value;
  }
  get v_z_valid() {
    if (!this._v_z_validIntialized) {
      return undefined;
    }
    return this._refObject.v_z_valid;
  }
  set v_z_valid(value) {
    this._v_z_validIntialized = true;
    this._refObject.v_z_valid = value;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get z() {
    if (!this._zIntialized) {
      return undefined;
    }
    return this._refObject.z;
  }
  set z(value) {
    this._zIntialized = true;
    this._refObject.z = value;
  }
  get delta_xy() {
    return this._wrapperFields['delta_xy'].data;
  }
  set delta_xy(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['delta_xy'].fill(value);
  }
  get xy_reset_counter() {
    if (!this._xy_reset_counterIntialized) {
      return undefined;
    }
    return this._refObject.xy_reset_counter;
  }
  set xy_reset_counter(value) {
    this._xy_reset_counterIntialized = true;
    this._refObject.xy_reset_counter = value;
  }
  get delta_z() {
    if (!this._delta_zIntialized) {
      return undefined;
    }
    return this._refObject.delta_z;
  }
  set delta_z(value) {
    this._delta_zIntialized = true;
    this._refObject.delta_z = value;
  }
  get z_reset_counter() {
    if (!this._z_reset_counterIntialized) {
      return undefined;
    }
    return this._refObject.z_reset_counter;
  }
  set z_reset_counter(value) {
    this._z_reset_counterIntialized = true;
    this._refObject.z_reset_counter = value;
  }
  get vx() {
    if (!this._vxIntialized) {
      return undefined;
    }
    return this._refObject.vx;
  }
  set vx(value) {
    this._vxIntialized = true;
    this._refObject.vx = value;
  }
  get vy() {
    if (!this._vyIntialized) {
      return undefined;
    }
    return this._refObject.vy;
  }
  set vy(value) {
    this._vyIntialized = true;
    this._refObject.vy = value;
  }
  get vz() {
    if (!this._vzIntialized) {
      return undefined;
    }
    return this._refObject.vz;
  }
  set vz(value) {
    this._vzIntialized = true;
    this._refObject.vz = value;
  }
  get z_deriv() {
    if (!this._z_derivIntialized) {
      return undefined;
    }
    return this._refObject.z_deriv;
  }
  set z_deriv(value) {
    this._z_derivIntialized = true;
    this._refObject.z_deriv = value;
  }
  get delta_vxy() {
    return this._wrapperFields['delta_vxy'].data;
  }
  set delta_vxy(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['delta_vxy'].fill(value);
  }
  get vxy_reset_counter() {
    if (!this._vxy_reset_counterIntialized) {
      return undefined;
    }
    return this._refObject.vxy_reset_counter;
  }
  set vxy_reset_counter(value) {
    this._vxy_reset_counterIntialized = true;
    this._refObject.vxy_reset_counter = value;
  }
  get delta_vz() {
    if (!this._delta_vzIntialized) {
      return undefined;
    }
    return this._refObject.delta_vz;
  }
  set delta_vz(value) {
    this._delta_vzIntialized = true;
    this._refObject.delta_vz = value;
  }
  get vz_reset_counter() {
    if (!this._vz_reset_counterIntialized) {
      return undefined;
    }
    return this._refObject.vz_reset_counter;
  }
  set vz_reset_counter(value) {
    this._vz_reset_counterIntialized = true;
    this._refObject.vz_reset_counter = value;
  }
  get ax() {
    if (!this._axIntialized) {
      return undefined;
    }
    return this._refObject.ax;
  }
  set ax(value) {
    this._axIntialized = true;
    this._refObject.ax = value;
  }
  get ay() {
    if (!this._ayIntialized) {
      return undefined;
    }
    return this._refObject.ay;
  }
  set ay(value) {
    this._ayIntialized = true;
    this._refObject.ay = value;
  }
  get az() {
    if (!this._azIntialized) {
      return undefined;
    }
    return this._refObject.az;
  }
  set az(value) {
    this._azIntialized = true;
    this._refObject.az = value;
  }
  get heading() {
    if (!this._headingIntialized) {
      return undefined;
    }
    return this._refObject.heading;
  }
  set heading(value) {
    this._headingIntialized = true;
    this._refObject.heading = value;
  }
  get delta_heading() {
    if (!this._delta_headingIntialized) {
      return undefined;
    }
    return this._refObject.delta_heading;
  }
  set delta_heading(value) {
    this._delta_headingIntialized = true;
    this._refObject.delta_heading = value;
  }
  get heading_reset_counter() {
    if (!this._heading_reset_counterIntialized) {
      return undefined;
    }
    return this._refObject.heading_reset_counter;
  }
  set heading_reset_counter(value) {
    this._heading_reset_counterIntialized = true;
    this._refObject.heading_reset_counter = value;
  }
  get heading_good_for_control() {
    if (!this._heading_good_for_controlIntialized) {
      return undefined;
    }
    return this._refObject.heading_good_for_control;
  }
  set heading_good_for_control(value) {
    this._heading_good_for_controlIntialized = true;
    this._refObject.heading_good_for_control = value;
  }
  get xy_global() {
    if (!this._xy_globalIntialized) {
      return undefined;
    }
    return this._refObject.xy_global;
  }
  set xy_global(value) {
    this._xy_globalIntialized = true;
    this._refObject.xy_global = value;
  }
  get z_global() {
    if (!this._z_globalIntialized) {
      return undefined;
    }
    return this._refObject.z_global;
  }
  set z_global(value) {
    this._z_globalIntialized = true;
    this._refObject.z_global = value;
  }
  get ref_timestamp() {
    if (!this._ref_timestampIntialized) {
      return undefined;
    }
    return this._refObject.ref_timestamp;
  }
  set ref_timestamp(value) {
    this._ref_timestampIntialized = true;
    this._refObject.ref_timestamp = value;
  }
  get ref_lat() {
    if (!this._ref_latIntialized) {
      return undefined;
    }
    return this._refObject.ref_lat;
  }
  set ref_lat(value) {
    this._ref_latIntialized = true;
    this._refObject.ref_lat = value;
  }
  get ref_lon() {
    if (!this._ref_lonIntialized) {
      return undefined;
    }
    return this._refObject.ref_lon;
  }
  set ref_lon(value) {
    this._ref_lonIntialized = true;
    this._refObject.ref_lon = value;
  }
  get ref_alt() {
    if (!this._ref_altIntialized) {
      return undefined;
    }
    return this._refObject.ref_alt;
  }
  set ref_alt(value) {
    this._ref_altIntialized = true;
    this._refObject.ref_alt = value;
  }
  get dist_bottom() {
    if (!this._dist_bottomIntialized) {
      return undefined;
    }
    return this._refObject.dist_bottom;
  }
  set dist_bottom(value) {
    this._dist_bottomIntialized = true;
    this._refObject.dist_bottom = value;
  }
  get dist_bottom_valid() {
    if (!this._dist_bottom_validIntialized) {
      return undefined;
    }
    return this._refObject.dist_bottom_valid;
  }
  set dist_bottom_valid(value) {
    this._dist_bottom_validIntialized = true;
    this._refObject.dist_bottom_valid = value;
  }
  get dist_bottom_sensor_bitfield() {
    if (!this._dist_bottom_sensor_bitfieldIntialized) {
      return undefined;
    }
    return this._refObject.dist_bottom_sensor_bitfield;
  }
  set dist_bottom_sensor_bitfield(value) {
    this._dist_bottom_sensor_bitfieldIntialized = true;
    this._refObject.dist_bottom_sensor_bitfield = value;
  }
  get eph() {
    if (!this._ephIntialized) {
      return undefined;
    }
    return this._refObject.eph;
  }
  set eph(value) {
    this._ephIntialized = true;
    this._refObject.eph = value;
  }
  get epv() {
    if (!this._epvIntialized) {
      return undefined;
    }
    return this._refObject.epv;
  }
  set epv(value) {
    this._epvIntialized = true;
    this._refObject.epv = value;
  }
  get evh() {
    if (!this._evhIntialized) {
      return undefined;
    }
    return this._refObject.evh;
  }
  set evh(value) {
    this._evhIntialized = true;
    this._refObject.evh = value;
  }
  get evv() {
    if (!this._evvIntialized) {
      return undefined;
    }
    return this._refObject.evv;
  }
  set evv(value) {
    this._evvIntialized = true;
    this._refObject.evv = value;
  }
  get dead_reckoning() {
    if (!this._dead_reckoningIntialized) {
      return undefined;
    }
    return this._refObject.dead_reckoning;
  }
  set dead_reckoning(value) {
    this._dead_reckoningIntialized = true;
    this._refObject.dead_reckoning = value;
  }
  get vxy_max() {
    if (!this._vxy_maxIntialized) {
      return undefined;
    }
    return this._refObject.vxy_max;
  }
  set vxy_max(value) {
    this._vxy_maxIntialized = true;
    this._refObject.vxy_max = value;
  }
  get vz_max() {
    if (!this._vz_maxIntialized) {
      return undefined;
    }
    return this._refObject.vz_max;
  }
  set vz_max(value) {
    this._vz_maxIntialized = true;
    this._refObject.vz_max = value;
  }
  get hagl_min() {
    if (!this._hagl_minIntialized) {
      return undefined;
    }
    return this._refObject.hagl_min;
  }
  set hagl_min(value) {
    this._hagl_minIntialized = true;
    this._refObject.hagl_min = value;
  }
  get hagl_max() {
    if (!this._hagl_maxIntialized) {
      return undefined;
    }
    return this._refObject.hagl_max;
  }
  set hagl_max(value) {
    this._hagl_maxIntialized = true;
    this._refObject.hagl_max = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleLocalPositionRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._xy_validIntialized = true;
    this._z_validIntialized = true;
    this._v_xy_validIntialized = true;
    this._v_z_validIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._wrapperFields.delta_xy.fill(refObject.delta_xy.toArray());
    this._xy_reset_counterIntialized = true;
    this._delta_zIntialized = true;
    this._z_reset_counterIntialized = true;
    this._vxIntialized = true;
    this._vyIntialized = true;
    this._vzIntialized = true;
    this._z_derivIntialized = true;
    this._wrapperFields.delta_vxy.fill(refObject.delta_vxy.toArray());
    this._vxy_reset_counterIntialized = true;
    this._delta_vzIntialized = true;
    this._vz_reset_counterIntialized = true;
    this._axIntialized = true;
    this._ayIntialized = true;
    this._azIntialized = true;
    this._headingIntialized = true;
    this._delta_headingIntialized = true;
    this._heading_reset_counterIntialized = true;
    this._heading_good_for_controlIntialized = true;
    this._xy_globalIntialized = true;
    this._z_globalIntialized = true;
    this._ref_timestampIntialized = true;
    this._ref_latIntialized = true;
    this._ref_lonIntialized = true;
    this._ref_altIntialized = true;
    this._dist_bottomIntialized = true;
    this._dist_bottom_validIntialized = true;
    this._dist_bottom_sensor_bitfieldIntialized = true;
    this._ephIntialized = true;
    this._epvIntialized = true;
    this._evhIntialized = true;
    this._evvIntialized = true;
    this._dead_reckoningIntialized = true;
    this._vxy_maxIntialized = true;
    this._vz_maxIntialized = true;
    this._hagl_minIntialized = true;
    this._hagl_maxIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleLocalPositionRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._xy_validIntialized = true;
    this._z_validIntialized = true;
    this._v_xy_validIntialized = true;
    this._v_z_validIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._wrapperFields.delta_xy.copy(other._wrapperFields.delta_xy);
    this._xy_reset_counterIntialized = true;
    this._delta_zIntialized = true;
    this._z_reset_counterIntialized = true;
    this._vxIntialized = true;
    this._vyIntialized = true;
    this._vzIntialized = true;
    this._z_derivIntialized = true;
    this._wrapperFields.delta_vxy.copy(other._wrapperFields.delta_vxy);
    this._vxy_reset_counterIntialized = true;
    this._delta_vzIntialized = true;
    this._vz_reset_counterIntialized = true;
    this._axIntialized = true;
    this._ayIntialized = true;
    this._azIntialized = true;
    this._headingIntialized = true;
    this._delta_headingIntialized = true;
    this._heading_reset_counterIntialized = true;
    this._heading_good_for_controlIntialized = true;
    this._xy_globalIntialized = true;
    this._z_globalIntialized = true;
    this._ref_timestampIntialized = true;
    this._ref_latIntialized = true;
    this._ref_lonIntialized = true;
    this._ref_altIntialized = true;
    this._dist_bottomIntialized = true;
    this._dist_bottom_validIntialized = true;
    this._dist_bottom_sensor_bitfieldIntialized = true;
    this._ephIntialized = true;
    this._epvIntialized = true;
    this._evhIntialized = true;
    this._evvIntialized = true;
    this._dead_reckoningIntialized = true;
    this._vxy_maxIntialized = true;
    this._vz_maxIntialized = true;
    this._hagl_minIntialized = true;
    this._hagl_maxIntialized = true;
  }
  static get classType() {
    return VehicleLocalPositionWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"DIST_BOTTOM_SENSOR_NONE","value":0},{"type":"uint8","name":"DIST_BOTTOM_SENSOR_RANGE","value":1},{"type":"uint8","name":"DIST_BOTTOM_SENSOR_FLOW","value":2}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"xy_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"v_xy_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"v_z_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_xy","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"xy_reset_counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z_reset_counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vx","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z_deriv","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_vxy","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vxy_reset_counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_vz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vz_reset_counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ax","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ay","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"az","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"heading","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_heading","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"heading_reset_counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"heading_good_for_control","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"xy_global","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z_global","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ref_timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ref_lat","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ref_lon","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ref_alt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dist_bottom","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dist_bottom_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dist_bottom_sensor_bitfield","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"eph","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"epv","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"evh","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"evv","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dead_reckoning","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vxy_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vz_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hagl_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hagl_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleLocalPosition","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleLocalPosition"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","xy_valid","z_valid","v_xy_valid","v_z_valid","x","y","z","delta_xy","xy_reset_counter","delta_z","z_reset_counter","vx","vy","vz","z_deriv","delta_vxy","vxy_reset_counter","delta_vz","vz_reset_counter","ax","ay","az","heading","delta_heading","heading_reset_counter","heading_good_for_control","xy_global","z_global","ref_timestamp","ref_lat","ref_lon","ref_alt","dist_bottom","dist_bottom_valid","dist_bottom_sensor_bitfield","eph","epv","evh","evv","dead_reckoning","vxy_max","vz_max","hagl_min","hagl_max"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleLocalPositionArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleLocalPositionWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleLocalPositionWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleLocalPositionArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleLocalPositionArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleLocalPositionRefArray(size);
    this._refObject = new VehicleLocalPositionRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleLocalPositionWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleLocalPositionArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleLocalPositionArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleLocalPositionWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleLocalPositionWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleLocalPositionArrayWrapper;
  }
}
// Define constants (3 in total)
Object.defineProperty(VehicleLocalPositionWrapper, "DIST_BOTTOM_SENSOR_NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleLocalPositionWrapper, "DIST_BOTTOM_SENSOR_RANGE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleLocalPositionWrapper, "DIST_BOTTOM_SENSOR_FLOW", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = VehicleLocalPositionWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "DIST_BOTTOM_SENSOR_NONE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "DIST_BOTTOM_SENSOR_RANGE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "DIST_BOTTOM_SENSOR_FLOW",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "xy_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "v_xy_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "v_z_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_xy",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "xy_reset_counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z_reset_counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vx",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z_deriv",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_vxy",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vxy_reset_counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_vz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vz_reset_counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ax",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ay",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "az",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "heading",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_heading",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "heading_reset_counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "heading_good_for_control",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "xy_global",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z_global",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ref_timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ref_lat",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ref_lon",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ref_alt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dist_bottom",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dist_bottom_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dist_bottom_sensor_bitfield",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "eph",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "epv",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "evh",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "evv",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dead_reckoning",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vxy_max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vz_max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hagl_min",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hagl_max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleLocalPosition",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleLocalPosition"
}
*/
