// rclnodejs/generated/px4_msgs/px4_msgs__msg__UlogStream.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const UlogStreamRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  length: primitiveTypes.uint8,
  first_message_offset: primitiveTypes.uint8,
  msg_sequence: primitiveTypes.uint16,
  flags: primitiveTypes.uint8,
  data: ArrayType(primitiveTypes.uint8, 249),
});
const UlogStreamRefArray = ArrayType(UlogStreamRefStruct);
const UlogStreamRefStructArray = StructType({
  data: UlogStreamRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class UlogStreamWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new UlogStreamRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._lengthIntialized = true;
      this._first_message_offsetIntialized = true;
      this._msg_sequenceIntialized = true;
      this._flagsIntialized = true;
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new UlogStreamRefStruct();
    this._timestampIntialized = false;
    this._lengthIntialized = false;
    this._first_message_offsetIntialized = false;
    this._msg_sequenceIntialized = false;
    this._flagsIntialized = false;
    this._wrapperFields.data = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new UlogStreamWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UlogStreamArrayWrapper;
  }
  static get ArrayType() {
    return UlogStreamArrayWrapper;
  }
  static get refObjectArrayType() {
    return UlogStreamRefStructArray
  }
  static get refObjectType() {
    return UlogStreamRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in UlogStream');
      }
      if (checkConsistency && !this._lengthIntialized) {
        throw new TypeError('Invalid argument: length in UlogStream');
      }
      if (checkConsistency && !this._first_message_offsetIntialized) {
        throw new TypeError('Invalid argument: first_message_offset in UlogStream');
      }
      if (checkConsistency && !this._msg_sequenceIntialized) {
        throw new TypeError('Invalid argument: msg_sequence in UlogStream');
      }
      if (checkConsistency && !this._flagsIntialized) {
        throw new TypeError('Invalid argument: flags in UlogStream');
      }
    this._refObject.data = Array.from(this._wrapperFields.data.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._lengthIntialized = true;
    this._refObject.length = refObject.length;
    this._first_message_offsetIntialized = true;
    this._refObject.first_message_offset = refObject.first_message_offset;
    this._msg_sequenceIntialized = true;
    this._refObject.msg_sequence = refObject.msg_sequence;
    this._flagsIntialized = true;
    this._refObject.flags = refObject.flags;
    this._wrapperFields.data.fill(refObject.data.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    UlogStreamWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'UlogStream'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get length() {
    if (!this._lengthIntialized) {
      return undefined;
    }
    return this._refObject.length;
  }
  set length(value) {
    this._lengthIntialized = true;
    this._refObject.length = value;
  }
  get first_message_offset() {
    if (!this._first_message_offsetIntialized) {
      return undefined;
    }
    return this._refObject.first_message_offset;
  }
  set first_message_offset(value) {
    this._first_message_offsetIntialized = true;
    this._refObject.first_message_offset = value;
  }
  get msg_sequence() {
    if (!this._msg_sequenceIntialized) {
      return undefined;
    }
    return this._refObject.msg_sequence;
  }
  set msg_sequence(value) {
    this._msg_sequenceIntialized = true;
    this._refObject.msg_sequence = value;
  }
  get flags() {
    if (!this._flagsIntialized) {
      return undefined;
    }
    return this._refObject.flags;
  }
  set flags(value) {
    this._flagsIntialized = true;
    this._refObject.flags = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    if (value.length !== 249) {
      throw new RangeError('The length of the array must be 249.');
    }
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new UlogStreamRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._lengthIntialized = true;
    this._first_message_offsetIntialized = true;
    this._msg_sequenceIntialized = true;
    this._flagsIntialized = true;
    this._wrapperFields.data.fill(refObject.data.toArray());
  }
  copy(other) {
    this._refObject = new UlogStreamRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._lengthIntialized = true;
    this._first_message_offsetIntialized = true;
    this._msg_sequenceIntialized = true;
    this._flagsIntialized = true;
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return UlogStreamWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"FLAGS_NEED_ACK","value":1},{"type":"uint8","name":"ORB_QUEUE_LENGTH","value":16}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"length","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"first_message_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"msg_sequence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":249,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"UlogStream","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UlogStream"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","length","first_message_offset","msg_sequence","flags","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class UlogStreamArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UlogStreamWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UlogStreamWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UlogStreamArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UlogStreamArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new UlogStreamRefArray(size);
    this._refObject = new UlogStreamRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UlogStreamWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UlogStreamArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UlogStreamArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UlogStreamWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UlogStreamWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UlogStreamArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(UlogStreamWrapper, "FLAGS_NEED_ACK", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(UlogStreamWrapper, "ORB_QUEUE_LENGTH", {value: 16, writable: false, enumerable: true, configurable: true});
module.exports = UlogStreamWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "FLAGS_NEED_ACK",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "ORB_QUEUE_LENGTH",
      "value": 16
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "length",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "first_message_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "msg_sequence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": 249,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "UlogStream",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "UlogStream"
}
*/
