// rclnodejs/generated/px4_msgs/px4_msgs__msg__HealthReport.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HealthReportRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  can_arm_mode_flags: primitiveTypes.uint64,
  can_run_mode_flags: primitiveTypes.uint64,
  health_is_present_flags: primitiveTypes.uint64,
  health_warning_flags: primitiveTypes.uint64,
  health_error_flags: primitiveTypes.uint64,
  arming_check_warning_flags: primitiveTypes.uint64,
  arming_check_error_flags: primitiveTypes.uint64,
});
const HealthReportRefArray = ArrayType(HealthReportRefStruct);
const HealthReportRefStructArray = StructType({
  data: HealthReportRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class HealthReportWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new HealthReportRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._can_arm_mode_flagsIntialized = true;
      this._can_run_mode_flagsIntialized = true;
      this._health_is_present_flagsIntialized = true;
      this._health_warning_flagsIntialized = true;
      this._health_error_flagsIntialized = true;
      this._arming_check_warning_flagsIntialized = true;
      this._arming_check_error_flagsIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new HealthReportRefStruct();
    this._timestampIntialized = false;
    this._can_arm_mode_flagsIntialized = false;
    this._can_run_mode_flagsIntialized = false;
    this._health_is_present_flagsIntialized = false;
    this._health_warning_flagsIntialized = false;
    this._health_error_flagsIntialized = false;
    this._arming_check_warning_flagsIntialized = false;
    this._arming_check_error_flagsIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new HealthReportWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new HealthReportArrayWrapper;
  }
  static get ArrayType() {
    return HealthReportArrayWrapper;
  }
  static get refObjectArrayType() {
    return HealthReportRefStructArray
  }
  static get refObjectType() {
    return HealthReportRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in HealthReport');
      }
      if (checkConsistency && !this._can_arm_mode_flagsIntialized) {
        throw new TypeError('Invalid argument: can_arm_mode_flags in HealthReport');
      }
      if (checkConsistency && !this._can_run_mode_flagsIntialized) {
        throw new TypeError('Invalid argument: can_run_mode_flags in HealthReport');
      }
      if (checkConsistency && !this._health_is_present_flagsIntialized) {
        throw new TypeError('Invalid argument: health_is_present_flags in HealthReport');
      }
      if (checkConsistency && !this._health_warning_flagsIntialized) {
        throw new TypeError('Invalid argument: health_warning_flags in HealthReport');
      }
      if (checkConsistency && !this._health_error_flagsIntialized) {
        throw new TypeError('Invalid argument: health_error_flags in HealthReport');
      }
      if (checkConsistency && !this._arming_check_warning_flagsIntialized) {
        throw new TypeError('Invalid argument: arming_check_warning_flags in HealthReport');
      }
      if (checkConsistency && !this._arming_check_error_flagsIntialized) {
        throw new TypeError('Invalid argument: arming_check_error_flags in HealthReport');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._can_arm_mode_flagsIntialized = true;
    this._refObject.can_arm_mode_flags = refObject.can_arm_mode_flags;
    this._can_run_mode_flagsIntialized = true;
    this._refObject.can_run_mode_flags = refObject.can_run_mode_flags;
    this._health_is_present_flagsIntialized = true;
    this._refObject.health_is_present_flags = refObject.health_is_present_flags;
    this._health_warning_flagsIntialized = true;
    this._refObject.health_warning_flags = refObject.health_warning_flags;
    this._health_error_flagsIntialized = true;
    this._refObject.health_error_flags = refObject.health_error_flags;
    this._arming_check_warning_flagsIntialized = true;
    this._refObject.arming_check_warning_flags = refObject.arming_check_warning_flags;
    this._arming_check_error_flagsIntialized = true;
    this._refObject.arming_check_error_flags = refObject.arming_check_error_flags;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    HealthReportWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'HealthReport'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get can_arm_mode_flags() {
    if (!this._can_arm_mode_flagsIntialized) {
      return undefined;
    }
    return this._refObject.can_arm_mode_flags;
  }
  set can_arm_mode_flags(value) {
    this._can_arm_mode_flagsIntialized = true;
    this._refObject.can_arm_mode_flags = value;
  }
  get can_run_mode_flags() {
    if (!this._can_run_mode_flagsIntialized) {
      return undefined;
    }
    return this._refObject.can_run_mode_flags;
  }
  set can_run_mode_flags(value) {
    this._can_run_mode_flagsIntialized = true;
    this._refObject.can_run_mode_flags = value;
  }
  get health_is_present_flags() {
    if (!this._health_is_present_flagsIntialized) {
      return undefined;
    }
    return this._refObject.health_is_present_flags;
  }
  set health_is_present_flags(value) {
    this._health_is_present_flagsIntialized = true;
    this._refObject.health_is_present_flags = value;
  }
  get health_warning_flags() {
    if (!this._health_warning_flagsIntialized) {
      return undefined;
    }
    return this._refObject.health_warning_flags;
  }
  set health_warning_flags(value) {
    this._health_warning_flagsIntialized = true;
    this._refObject.health_warning_flags = value;
  }
  get health_error_flags() {
    if (!this._health_error_flagsIntialized) {
      return undefined;
    }
    return this._refObject.health_error_flags;
  }
  set health_error_flags(value) {
    this._health_error_flagsIntialized = true;
    this._refObject.health_error_flags = value;
  }
  get arming_check_warning_flags() {
    if (!this._arming_check_warning_flagsIntialized) {
      return undefined;
    }
    return this._refObject.arming_check_warning_flags;
  }
  set arming_check_warning_flags(value) {
    this._arming_check_warning_flagsIntialized = true;
    this._refObject.arming_check_warning_flags = value;
  }
  get arming_check_error_flags() {
    if (!this._arming_check_error_flagsIntialized) {
      return undefined;
    }
    return this._refObject.arming_check_error_flags;
  }
  set arming_check_error_flags(value) {
    this._arming_check_error_flagsIntialized = true;
    this._refObject.arming_check_error_flags = value;
  }
  copyRefObject(refObject) {
    this._refObject = new HealthReportRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._can_arm_mode_flagsIntialized = true;
    this._can_run_mode_flagsIntialized = true;
    this._health_is_present_flagsIntialized = true;
    this._health_warning_flagsIntialized = true;
    this._health_error_flagsIntialized = true;
    this._arming_check_warning_flagsIntialized = true;
    this._arming_check_error_flagsIntialized = true;
  }
  copy(other) {
    this._refObject = new HealthReportRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._can_arm_mode_flagsIntialized = true;
    this._can_run_mode_flagsIntialized = true;
    this._health_is_present_flagsIntialized = true;
    this._health_warning_flagsIntialized = true;
    this._health_error_flagsIntialized = true;
    this._arming_check_warning_flagsIntialized = true;
    this._arming_check_error_flagsIntialized = true;
  }
  static get classType() {
    return HealthReportWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"can_arm_mode_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"can_run_mode_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"health_is_present_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"health_warning_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"health_error_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_check_warning_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_check_error_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"HealthReport","stringUpperBound":null,"isPrimitiveType":false},"msgName":"HealthReport"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","can_arm_mode_flags","can_run_mode_flags","health_is_present_flags","health_warning_flags","health_error_flags","arming_check_warning_flags","arming_check_error_flags"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class HealthReportArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof HealthReportWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new HealthReportWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HealthReportArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to HealthReportArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new HealthReportRefArray(size);
    this._refObject = new HealthReportRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new HealthReportWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof HealthReportArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "HealthReportArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      HealthReportWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return HealthReportWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return HealthReportArrayWrapper;
  }
}
module.exports = HealthReportWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "can_arm_mode_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "can_run_mode_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "health_is_present_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "health_warning_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "health_error_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_check_warning_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_check_error_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "HealthReport",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "HealthReport"
}
*/
