// rclnodejs/generated/px4_msgs/px4_msgs__msg__AdcReport.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int16Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int16.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const AdcReportRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  device_id: primitiveTypes.uint32,
  channel_id: ArrayType(primitiveTypes.int16, 12),
  raw_data: ArrayType(primitiveTypes.int32, 12),
  resolution: primitiveTypes.uint32,
  v_ref: primitiveTypes.float32,
});
const AdcReportRefArray = ArrayType(AdcReportRefStruct);
const AdcReportRefStructArray = StructType({
  data: AdcReportRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AdcReportWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AdcReportRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._device_idIntialized = true;
      this._wrapperFields.channel_id = Int16Wrapper.createArray();
      this._wrapperFields.channel_id.copy(other._wrapperFields.channel_id);
      this._wrapperFields.raw_data = Int32Wrapper.createArray();
      this._wrapperFields.raw_data.copy(other._wrapperFields.raw_data);
      this._resolutionIntialized = true;
      this._v_refIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AdcReportRefStruct();
    this._timestampIntialized = false;
    this._device_idIntialized = false;
    this._wrapperFields.channel_id = Int16Wrapper.createArray();
    this._wrapperFields.raw_data = Int32Wrapper.createArray();
    this._resolutionIntialized = false;
    this._v_refIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new AdcReportWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AdcReportArrayWrapper;
  }
  static get ArrayType() {
    return AdcReportArrayWrapper;
  }
  static get refObjectArrayType() {
    return AdcReportRefStructArray
  }
  static get refObjectType() {
    return AdcReportRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in AdcReport');
      }
      if (checkConsistency && !this._device_idIntialized) {
        throw new TypeError('Invalid argument: device_id in AdcReport');
      }
      if (checkConsistency && !this._resolutionIntialized) {
        throw new TypeError('Invalid argument: resolution in AdcReport');
      }
      if (checkConsistency && !this._v_refIntialized) {
        throw new TypeError('Invalid argument: v_ref in AdcReport');
      }
    this._refObject.channel_id = Array.from(this._wrapperFields.channel_id.data);
    this._refObject.raw_data = Array.from(this._wrapperFields.raw_data.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._device_idIntialized = true;
    this._refObject.device_id = refObject.device_id;
    this._wrapperFields.channel_id.fill(refObject.channel_id.toArray());
    this._wrapperFields.raw_data.fill(refObject.raw_data.toArray());
    this._resolutionIntialized = true;
    this._refObject.resolution = refObject.resolution;
    this._v_refIntialized = true;
    this._refObject.v_ref = refObject.v_ref;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    AdcReportWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'AdcReport'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get device_id() {
    if (!this._device_idIntialized) {
      return undefined;
    }
    return this._refObject.device_id;
  }
  set device_id(value) {
    this._device_idIntialized = true;
    this._refObject.device_id = value;
  }
  get channel_id() {
    return this._wrapperFields['channel_id'].data;
  }
  set channel_id(value) {
    if (value.length !== 12) {
      throw new RangeError('The length of the array must be 12.');
    }
    this._wrapperFields['channel_id'].fill(value);
  }
  get raw_data() {
    return this._wrapperFields['raw_data'].data;
  }
  set raw_data(value) {
    if (value.length !== 12) {
      throw new RangeError('The length of the array must be 12.');
    }
    this._wrapperFields['raw_data'].fill(value);
  }
  get resolution() {
    if (!this._resolutionIntialized) {
      return undefined;
    }
    return this._refObject.resolution;
  }
  set resolution(value) {
    this._resolutionIntialized = true;
    this._refObject.resolution = value;
  }
  get v_ref() {
    if (!this._v_refIntialized) {
      return undefined;
    }
    return this._refObject.v_ref;
  }
  set v_ref(value) {
    this._v_refIntialized = true;
    this._refObject.v_ref = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AdcReportRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._device_idIntialized = true;
    this._wrapperFields.channel_id.fill(refObject.channel_id.toArray());
    this._wrapperFields.raw_data.fill(refObject.raw_data.toArray());
    this._resolutionIntialized = true;
    this._v_refIntialized = true;
  }
  copy(other) {
    this._refObject = new AdcReportRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._device_idIntialized = true;
    this._wrapperFields.channel_id.copy(other._wrapperFields.channel_id);
    this._wrapperFields.raw_data.copy(other._wrapperFields.raw_data);
    this._resolutionIntialized = true;
    this._v_refIntialized = true;
  }
  static get classType() {
    return AdcReportWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"channel_id","type":{"isArray":true,"arraySize":12,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"raw_data","type":{"isArray":true,"arraySize":12,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"resolution","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"v_ref","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"AdcReport","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AdcReport"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","device_id","channel_id","raw_data","resolution","v_ref"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AdcReportArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AdcReportWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AdcReportWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AdcReportArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AdcReportArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AdcReportRefArray(size);
    this._refObject = new AdcReportRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AdcReportWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AdcReportArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AdcReportArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AdcReportWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AdcReportWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AdcReportArrayWrapper;
  }
}
module.exports = AdcReportWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "channel_id",
      "type": {
        "isArray": true,
        "arraySize": 12,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "raw_data",
      "type": {
        "isArray": true,
        "arraySize": 12,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "resolution",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "v_ref",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "AdcReport",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "AdcReport"
}
*/
