// rclnodejs/generated/px4_msgs/px4_msgs__msg__ActuatorArmed.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ActuatorArmedRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  armed: primitiveTypes.bool,
  prearmed: primitiveTypes.bool,
  ready_to_arm: primitiveTypes.bool,
  lockdown: primitiveTypes.bool,
  manual_lockdown: primitiveTypes.bool,
  force_failsafe: primitiveTypes.bool,
  in_esc_calibration_mode: primitiveTypes.bool,
});
const ActuatorArmedRefArray = ArrayType(ActuatorArmedRefStruct);
const ActuatorArmedRefStructArray = StructType({
  data: ActuatorArmedRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ActuatorArmedWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ActuatorArmedRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._armedIntialized = true;
      this._prearmedIntialized = true;
      this._ready_to_armIntialized = true;
      this._lockdownIntialized = true;
      this._manual_lockdownIntialized = true;
      this._force_failsafeIntialized = true;
      this._in_esc_calibration_modeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ActuatorArmedRefStruct();
    this._timestampIntialized = false;
    this._armedIntialized = false;
    this._prearmedIntialized = false;
    this._ready_to_armIntialized = false;
    this._lockdownIntialized = false;
    this._manual_lockdownIntialized = false;
    this._force_failsafeIntialized = false;
    this._in_esc_calibration_modeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ActuatorArmedWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ActuatorArmedArrayWrapper;
  }
  static get ArrayType() {
    return ActuatorArmedArrayWrapper;
  }
  static get refObjectArrayType() {
    return ActuatorArmedRefStructArray
  }
  static get refObjectType() {
    return ActuatorArmedRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in ActuatorArmed');
      }
      if (checkConsistency && !this._armedIntialized) {
        throw new TypeError('Invalid argument: armed in ActuatorArmed');
      }
      if (checkConsistency && !this._prearmedIntialized) {
        throw new TypeError('Invalid argument: prearmed in ActuatorArmed');
      }
      if (checkConsistency && !this._ready_to_armIntialized) {
        throw new TypeError('Invalid argument: ready_to_arm in ActuatorArmed');
      }
      if (checkConsistency && !this._lockdownIntialized) {
        throw new TypeError('Invalid argument: lockdown in ActuatorArmed');
      }
      if (checkConsistency && !this._manual_lockdownIntialized) {
        throw new TypeError('Invalid argument: manual_lockdown in ActuatorArmed');
      }
      if (checkConsistency && !this._force_failsafeIntialized) {
        throw new TypeError('Invalid argument: force_failsafe in ActuatorArmed');
      }
      if (checkConsistency && !this._in_esc_calibration_modeIntialized) {
        throw new TypeError('Invalid argument: in_esc_calibration_mode in ActuatorArmed');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._armedIntialized = true;
    this._refObject.armed = refObject.armed;
    this._prearmedIntialized = true;
    this._refObject.prearmed = refObject.prearmed;
    this._ready_to_armIntialized = true;
    this._refObject.ready_to_arm = refObject.ready_to_arm;
    this._lockdownIntialized = true;
    this._refObject.lockdown = refObject.lockdown;
    this._manual_lockdownIntialized = true;
    this._refObject.manual_lockdown = refObject.manual_lockdown;
    this._force_failsafeIntialized = true;
    this._refObject.force_failsafe = refObject.force_failsafe;
    this._in_esc_calibration_modeIntialized = true;
    this._refObject.in_esc_calibration_mode = refObject.in_esc_calibration_mode;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ActuatorArmedWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'ActuatorArmed'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get armed() {
    if (!this._armedIntialized) {
      return undefined;
    }
    return this._refObject.armed;
  }
  set armed(value) {
    this._armedIntialized = true;
    this._refObject.armed = value;
  }
  get prearmed() {
    if (!this._prearmedIntialized) {
      return undefined;
    }
    return this._refObject.prearmed;
  }
  set prearmed(value) {
    this._prearmedIntialized = true;
    this._refObject.prearmed = value;
  }
  get ready_to_arm() {
    if (!this._ready_to_armIntialized) {
      return undefined;
    }
    return this._refObject.ready_to_arm;
  }
  set ready_to_arm(value) {
    this._ready_to_armIntialized = true;
    this._refObject.ready_to_arm = value;
  }
  get lockdown() {
    if (!this._lockdownIntialized) {
      return undefined;
    }
    return this._refObject.lockdown;
  }
  set lockdown(value) {
    this._lockdownIntialized = true;
    this._refObject.lockdown = value;
  }
  get manual_lockdown() {
    if (!this._manual_lockdownIntialized) {
      return undefined;
    }
    return this._refObject.manual_lockdown;
  }
  set manual_lockdown(value) {
    this._manual_lockdownIntialized = true;
    this._refObject.manual_lockdown = value;
  }
  get force_failsafe() {
    if (!this._force_failsafeIntialized) {
      return undefined;
    }
    return this._refObject.force_failsafe;
  }
  set force_failsafe(value) {
    this._force_failsafeIntialized = true;
    this._refObject.force_failsafe = value;
  }
  get in_esc_calibration_mode() {
    if (!this._in_esc_calibration_modeIntialized) {
      return undefined;
    }
    return this._refObject.in_esc_calibration_mode;
  }
  set in_esc_calibration_mode(value) {
    this._in_esc_calibration_modeIntialized = true;
    this._refObject.in_esc_calibration_mode = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ActuatorArmedRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._armedIntialized = true;
    this._prearmedIntialized = true;
    this._ready_to_armIntialized = true;
    this._lockdownIntialized = true;
    this._manual_lockdownIntialized = true;
    this._force_failsafeIntialized = true;
    this._in_esc_calibration_modeIntialized = true;
  }
  copy(other) {
    this._refObject = new ActuatorArmedRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._armedIntialized = true;
    this._prearmedIntialized = true;
    this._ready_to_armIntialized = true;
    this._lockdownIntialized = true;
    this._manual_lockdownIntialized = true;
    this._force_failsafeIntialized = true;
    this._in_esc_calibration_modeIntialized = true;
  }
  static get classType() {
    return ActuatorArmedWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"armed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"prearmed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ready_to_arm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lockdown","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"manual_lockdown","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"force_failsafe","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"in_esc_calibration_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"ActuatorArmed","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ActuatorArmed"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","armed","prearmed","ready_to_arm","lockdown","manual_lockdown","force_failsafe","in_esc_calibration_mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ActuatorArmedArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ActuatorArmedWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ActuatorArmedWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorArmedArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorArmedArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ActuatorArmedRefArray(size);
    this._refObject = new ActuatorArmedRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ActuatorArmedWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ActuatorArmedArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ActuatorArmedArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ActuatorArmedWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ActuatorArmedWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ActuatorArmedArrayWrapper;
  }
}
module.exports = ActuatorArmedWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "armed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "prearmed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ready_to_arm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lockdown",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "manual_lockdown",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "force_failsafe",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "in_esc_calibration_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "ActuatorArmed",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ActuatorArmed"
}
*/
