// rclnodejs/generated/px4_msgs/px4_msgs__msg__BatteryStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const BatteryStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  connected: primitiveTypes.bool,
  voltage_v: primitiveTypes.float32,
  voltage_filtered_v: primitiveTypes.float32,
  current_a: primitiveTypes.float32,
  current_filtered_a: primitiveTypes.float32,
  current_average_a: primitiveTypes.float32,
  discharged_mah: primitiveTypes.float32,
  remaining: primitiveTypes.float32,
  scale: primitiveTypes.float32,
  time_remaining_s: primitiveTypes.float32,
  temperature: primitiveTypes.float32,
  cell_count: primitiveTypes.uint8,
  source: primitiveTypes.uint8,
  priority: primitiveTypes.uint8,
  capacity: primitiveTypes.uint16,
  cycle_count: primitiveTypes.uint16,
  average_time_to_empty: primitiveTypes.uint16,
  serial_number: primitiveTypes.uint16,
  manufacture_date: primitiveTypes.uint16,
  state_of_health: primitiveTypes.uint16,
  max_error: primitiveTypes.uint16,
  id: primitiveTypes.uint8,
  interface_error: primitiveTypes.uint16,
  voltage_cell_v: ArrayType(primitiveTypes.float32, 14),
  max_cell_voltage_delta: primitiveTypes.float32,
  is_powering_off: primitiveTypes.bool,
  is_required: primitiveTypes.bool,
  faults: primitiveTypes.uint16,
  custom_faults: primitiveTypes.uint32,
  warning: primitiveTypes.uint8,
  mode: primitiveTypes.uint8,
  average_power: primitiveTypes.float32,
  available_energy: primitiveTypes.float32,
  full_charge_capacity_wh: primitiveTypes.float32,
  remaining_capacity_wh: primitiveTypes.float32,
  design_capacity: primitiveTypes.float32,
  average_time_to_full: primitiveTypes.uint16,
  over_discharge_count: primitiveTypes.uint16,
  nominal_voltage: primitiveTypes.float32,
});
const BatteryStatusRefArray = ArrayType(BatteryStatusRefStruct);
const BatteryStatusRefStructArray = StructType({
  data: BatteryStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BatteryStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BatteryStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._connectedIntialized = true;
      this._voltage_vIntialized = true;
      this._voltage_filtered_vIntialized = true;
      this._current_aIntialized = true;
      this._current_filtered_aIntialized = true;
      this._current_average_aIntialized = true;
      this._discharged_mahIntialized = true;
      this._remainingIntialized = true;
      this._scaleIntialized = true;
      this._time_remaining_sIntialized = true;
      this._temperatureIntialized = true;
      this._cell_countIntialized = true;
      this._sourceIntialized = true;
      this._priorityIntialized = true;
      this._capacityIntialized = true;
      this._cycle_countIntialized = true;
      this._average_time_to_emptyIntialized = true;
      this._serial_numberIntialized = true;
      this._manufacture_dateIntialized = true;
      this._state_of_healthIntialized = true;
      this._max_errorIntialized = true;
      this._idIntialized = true;
      this._interface_errorIntialized = true;
      this._wrapperFields.voltage_cell_v = Float32Wrapper.createArray();
      this._wrapperFields.voltage_cell_v.copy(other._wrapperFields.voltage_cell_v);
      this._max_cell_voltage_deltaIntialized = true;
      this._is_powering_offIntialized = true;
      this._is_requiredIntialized = true;
      this._faultsIntialized = true;
      this._custom_faultsIntialized = true;
      this._warningIntialized = true;
      this._modeIntialized = true;
      this._average_powerIntialized = true;
      this._available_energyIntialized = true;
      this._full_charge_capacity_whIntialized = true;
      this._remaining_capacity_whIntialized = true;
      this._design_capacityIntialized = true;
      this._average_time_to_fullIntialized = true;
      this._over_discharge_countIntialized = true;
      this._nominal_voltageIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BatteryStatusRefStruct();
    this._timestampIntialized = false;
    this._connectedIntialized = false;
    this._voltage_vIntialized = false;
    this._voltage_filtered_vIntialized = false;
    this._current_aIntialized = false;
    this._current_filtered_aIntialized = false;
    this._current_average_aIntialized = false;
    this._discharged_mahIntialized = false;
    this._remainingIntialized = false;
    this._scaleIntialized = false;
    this._time_remaining_sIntialized = false;
    this._temperatureIntialized = false;
    this._cell_countIntialized = false;
    this._sourceIntialized = false;
    this._priorityIntialized = false;
    this._capacityIntialized = false;
    this._cycle_countIntialized = false;
    this._average_time_to_emptyIntialized = false;
    this._serial_numberIntialized = false;
    this._manufacture_dateIntialized = false;
    this._state_of_healthIntialized = false;
    this._max_errorIntialized = false;
    this._idIntialized = false;
    this._interface_errorIntialized = false;
    this._wrapperFields.voltage_cell_v = Float32Wrapper.createArray();
    this._max_cell_voltage_deltaIntialized = false;
    this._is_powering_offIntialized = false;
    this._is_requiredIntialized = false;
    this._faultsIntialized = false;
    this._custom_faultsIntialized = false;
    this._warningIntialized = false;
    this._modeIntialized = false;
    this._average_powerIntialized = false;
    this._available_energyIntialized = false;
    this._full_charge_capacity_whIntialized = false;
    this._remaining_capacity_whIntialized = false;
    this._design_capacityIntialized = false;
    this._average_time_to_fullIntialized = false;
    this._over_discharge_countIntialized = false;
    this._nominal_voltageIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new BatteryStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BatteryStatusArrayWrapper;
  }
  static get ArrayType() {
    return BatteryStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return BatteryStatusRefStructArray
  }
  static get refObjectType() {
    return BatteryStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in BatteryStatus');
      }
      if (checkConsistency && !this._connectedIntialized) {
        throw new TypeError('Invalid argument: connected in BatteryStatus');
      }
      if (checkConsistency && !this._voltage_vIntialized) {
        throw new TypeError('Invalid argument: voltage_v in BatteryStatus');
      }
      if (checkConsistency && !this._voltage_filtered_vIntialized) {
        throw new TypeError('Invalid argument: voltage_filtered_v in BatteryStatus');
      }
      if (checkConsistency && !this._current_aIntialized) {
        throw new TypeError('Invalid argument: current_a in BatteryStatus');
      }
      if (checkConsistency && !this._current_filtered_aIntialized) {
        throw new TypeError('Invalid argument: current_filtered_a in BatteryStatus');
      }
      if (checkConsistency && !this._current_average_aIntialized) {
        throw new TypeError('Invalid argument: current_average_a in BatteryStatus');
      }
      if (checkConsistency && !this._discharged_mahIntialized) {
        throw new TypeError('Invalid argument: discharged_mah in BatteryStatus');
      }
      if (checkConsistency && !this._remainingIntialized) {
        throw new TypeError('Invalid argument: remaining in BatteryStatus');
      }
      if (checkConsistency && !this._scaleIntialized) {
        throw new TypeError('Invalid argument: scale in BatteryStatus');
      }
      if (checkConsistency && !this._time_remaining_sIntialized) {
        throw new TypeError('Invalid argument: time_remaining_s in BatteryStatus');
      }
      if (checkConsistency && !this._temperatureIntialized) {
        throw new TypeError('Invalid argument: temperature in BatteryStatus');
      }
      if (checkConsistency && !this._cell_countIntialized) {
        throw new TypeError('Invalid argument: cell_count in BatteryStatus');
      }
      if (checkConsistency && !this._sourceIntialized) {
        throw new TypeError('Invalid argument: source in BatteryStatus');
      }
      if (checkConsistency && !this._priorityIntialized) {
        throw new TypeError('Invalid argument: priority in BatteryStatus');
      }
      if (checkConsistency && !this._capacityIntialized) {
        throw new TypeError('Invalid argument: capacity in BatteryStatus');
      }
      if (checkConsistency && !this._cycle_countIntialized) {
        throw new TypeError('Invalid argument: cycle_count in BatteryStatus');
      }
      if (checkConsistency && !this._average_time_to_emptyIntialized) {
        throw new TypeError('Invalid argument: average_time_to_empty in BatteryStatus');
      }
      if (checkConsistency && !this._serial_numberIntialized) {
        throw new TypeError('Invalid argument: serial_number in BatteryStatus');
      }
      if (checkConsistency && !this._manufacture_dateIntialized) {
        throw new TypeError('Invalid argument: manufacture_date in BatteryStatus');
      }
      if (checkConsistency && !this._state_of_healthIntialized) {
        throw new TypeError('Invalid argument: state_of_health in BatteryStatus');
      }
      if (checkConsistency && !this._max_errorIntialized) {
        throw new TypeError('Invalid argument: max_error in BatteryStatus');
      }
      if (checkConsistency && !this._idIntialized) {
        throw new TypeError('Invalid argument: id in BatteryStatus');
      }
      if (checkConsistency && !this._interface_errorIntialized) {
        throw new TypeError('Invalid argument: interface_error in BatteryStatus');
      }
      if (checkConsistency && !this._max_cell_voltage_deltaIntialized) {
        throw new TypeError('Invalid argument: max_cell_voltage_delta in BatteryStatus');
      }
      if (checkConsistency && !this._is_powering_offIntialized) {
        throw new TypeError('Invalid argument: is_powering_off in BatteryStatus');
      }
      if (checkConsistency && !this._is_requiredIntialized) {
        throw new TypeError('Invalid argument: is_required in BatteryStatus');
      }
      if (checkConsistency && !this._faultsIntialized) {
        throw new TypeError('Invalid argument: faults in BatteryStatus');
      }
      if (checkConsistency && !this._custom_faultsIntialized) {
        throw new TypeError('Invalid argument: custom_faults in BatteryStatus');
      }
      if (checkConsistency && !this._warningIntialized) {
        throw new TypeError('Invalid argument: warning in BatteryStatus');
      }
      if (checkConsistency && !this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in BatteryStatus');
      }
      if (checkConsistency && !this._average_powerIntialized) {
        throw new TypeError('Invalid argument: average_power in BatteryStatus');
      }
      if (checkConsistency && !this._available_energyIntialized) {
        throw new TypeError('Invalid argument: available_energy in BatteryStatus');
      }
      if (checkConsistency && !this._full_charge_capacity_whIntialized) {
        throw new TypeError('Invalid argument: full_charge_capacity_wh in BatteryStatus');
      }
      if (checkConsistency && !this._remaining_capacity_whIntialized) {
        throw new TypeError('Invalid argument: remaining_capacity_wh in BatteryStatus');
      }
      if (checkConsistency && !this._design_capacityIntialized) {
        throw new TypeError('Invalid argument: design_capacity in BatteryStatus');
      }
      if (checkConsistency && !this._average_time_to_fullIntialized) {
        throw new TypeError('Invalid argument: average_time_to_full in BatteryStatus');
      }
      if (checkConsistency && !this._over_discharge_countIntialized) {
        throw new TypeError('Invalid argument: over_discharge_count in BatteryStatus');
      }
      if (checkConsistency && !this._nominal_voltageIntialized) {
        throw new TypeError('Invalid argument: nominal_voltage in BatteryStatus');
      }
    this._refObject.voltage_cell_v = Array.from(this._wrapperFields.voltage_cell_v.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._connectedIntialized = true;
    this._refObject.connected = refObject.connected;
    this._voltage_vIntialized = true;
    this._refObject.voltage_v = refObject.voltage_v;
    this._voltage_filtered_vIntialized = true;
    this._refObject.voltage_filtered_v = refObject.voltage_filtered_v;
    this._current_aIntialized = true;
    this._refObject.current_a = refObject.current_a;
    this._current_filtered_aIntialized = true;
    this._refObject.current_filtered_a = refObject.current_filtered_a;
    this._current_average_aIntialized = true;
    this._refObject.current_average_a = refObject.current_average_a;
    this._discharged_mahIntialized = true;
    this._refObject.discharged_mah = refObject.discharged_mah;
    this._remainingIntialized = true;
    this._refObject.remaining = refObject.remaining;
    this._scaleIntialized = true;
    this._refObject.scale = refObject.scale;
    this._time_remaining_sIntialized = true;
    this._refObject.time_remaining_s = refObject.time_remaining_s;
    this._temperatureIntialized = true;
    this._refObject.temperature = refObject.temperature;
    this._cell_countIntialized = true;
    this._refObject.cell_count = refObject.cell_count;
    this._sourceIntialized = true;
    this._refObject.source = refObject.source;
    this._priorityIntialized = true;
    this._refObject.priority = refObject.priority;
    this._capacityIntialized = true;
    this._refObject.capacity = refObject.capacity;
    this._cycle_countIntialized = true;
    this._refObject.cycle_count = refObject.cycle_count;
    this._average_time_to_emptyIntialized = true;
    this._refObject.average_time_to_empty = refObject.average_time_to_empty;
    this._serial_numberIntialized = true;
    this._refObject.serial_number = refObject.serial_number;
    this._manufacture_dateIntialized = true;
    this._refObject.manufacture_date = refObject.manufacture_date;
    this._state_of_healthIntialized = true;
    this._refObject.state_of_health = refObject.state_of_health;
    this._max_errorIntialized = true;
    this._refObject.max_error = refObject.max_error;
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._interface_errorIntialized = true;
    this._refObject.interface_error = refObject.interface_error;
    this._wrapperFields.voltage_cell_v.fill(refObject.voltage_cell_v.toArray());
    this._max_cell_voltage_deltaIntialized = true;
    this._refObject.max_cell_voltage_delta = refObject.max_cell_voltage_delta;
    this._is_powering_offIntialized = true;
    this._refObject.is_powering_off = refObject.is_powering_off;
    this._is_requiredIntialized = true;
    this._refObject.is_required = refObject.is_required;
    this._faultsIntialized = true;
    this._refObject.faults = refObject.faults;
    this._custom_faultsIntialized = true;
    this._refObject.custom_faults = refObject.custom_faults;
    this._warningIntialized = true;
    this._refObject.warning = refObject.warning;
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._average_powerIntialized = true;
    this._refObject.average_power = refObject.average_power;
    this._available_energyIntialized = true;
    this._refObject.available_energy = refObject.available_energy;
    this._full_charge_capacity_whIntialized = true;
    this._refObject.full_charge_capacity_wh = refObject.full_charge_capacity_wh;
    this._remaining_capacity_whIntialized = true;
    this._refObject.remaining_capacity_wh = refObject.remaining_capacity_wh;
    this._design_capacityIntialized = true;
    this._refObject.design_capacity = refObject.design_capacity;
    this._average_time_to_fullIntialized = true;
    this._refObject.average_time_to_full = refObject.average_time_to_full;
    this._over_discharge_countIntialized = true;
    this._refObject.over_discharge_count = refObject.over_discharge_count;
    this._nominal_voltageIntialized = true;
    this._refObject.nominal_voltage = refObject.nominal_voltage;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    BatteryStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'BatteryStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get connected() {
    if (!this._connectedIntialized) {
      return undefined;
    }
    return this._refObject.connected;
  }
  set connected(value) {
    this._connectedIntialized = true;
    this._refObject.connected = value;
  }
  get voltage_v() {
    if (!this._voltage_vIntialized) {
      return undefined;
    }
    return this._refObject.voltage_v;
  }
  set voltage_v(value) {
    this._voltage_vIntialized = true;
    this._refObject.voltage_v = value;
  }
  get voltage_filtered_v() {
    if (!this._voltage_filtered_vIntialized) {
      return undefined;
    }
    return this._refObject.voltage_filtered_v;
  }
  set voltage_filtered_v(value) {
    this._voltage_filtered_vIntialized = true;
    this._refObject.voltage_filtered_v = value;
  }
  get current_a() {
    if (!this._current_aIntialized) {
      return undefined;
    }
    return this._refObject.current_a;
  }
  set current_a(value) {
    this._current_aIntialized = true;
    this._refObject.current_a = value;
  }
  get current_filtered_a() {
    if (!this._current_filtered_aIntialized) {
      return undefined;
    }
    return this._refObject.current_filtered_a;
  }
  set current_filtered_a(value) {
    this._current_filtered_aIntialized = true;
    this._refObject.current_filtered_a = value;
  }
  get current_average_a() {
    if (!this._current_average_aIntialized) {
      return undefined;
    }
    return this._refObject.current_average_a;
  }
  set current_average_a(value) {
    this._current_average_aIntialized = true;
    this._refObject.current_average_a = value;
  }
  get discharged_mah() {
    if (!this._discharged_mahIntialized) {
      return undefined;
    }
    return this._refObject.discharged_mah;
  }
  set discharged_mah(value) {
    this._discharged_mahIntialized = true;
    this._refObject.discharged_mah = value;
  }
  get remaining() {
    if (!this._remainingIntialized) {
      return undefined;
    }
    return this._refObject.remaining;
  }
  set remaining(value) {
    this._remainingIntialized = true;
    this._refObject.remaining = value;
  }
  get scale() {
    if (!this._scaleIntialized) {
      return undefined;
    }
    return this._refObject.scale;
  }
  set scale(value) {
    this._scaleIntialized = true;
    this._refObject.scale = value;
  }
  get time_remaining_s() {
    if (!this._time_remaining_sIntialized) {
      return undefined;
    }
    return this._refObject.time_remaining_s;
  }
  set time_remaining_s(value) {
    this._time_remaining_sIntialized = true;
    this._refObject.time_remaining_s = value;
  }
  get temperature() {
    if (!this._temperatureIntialized) {
      return undefined;
    }
    return this._refObject.temperature;
  }
  set temperature(value) {
    this._temperatureIntialized = true;
    this._refObject.temperature = value;
  }
  get cell_count() {
    if (!this._cell_countIntialized) {
      return undefined;
    }
    return this._refObject.cell_count;
  }
  set cell_count(value) {
    this._cell_countIntialized = true;
    this._refObject.cell_count = value;
  }
  get source() {
    if (!this._sourceIntialized) {
      return undefined;
    }
    return this._refObject.source;
  }
  set source(value) {
    this._sourceIntialized = true;
    this._refObject.source = value;
  }
  get priority() {
    if (!this._priorityIntialized) {
      return undefined;
    }
    return this._refObject.priority;
  }
  set priority(value) {
    this._priorityIntialized = true;
    this._refObject.priority = value;
  }
  get capacity() {
    if (!this._capacityIntialized) {
      return undefined;
    }
    return this._refObject.capacity;
  }
  set capacity(value) {
    this._capacityIntialized = true;
    this._refObject.capacity = value;
  }
  get cycle_count() {
    if (!this._cycle_countIntialized) {
      return undefined;
    }
    return this._refObject.cycle_count;
  }
  set cycle_count(value) {
    this._cycle_countIntialized = true;
    this._refObject.cycle_count = value;
  }
  get average_time_to_empty() {
    if (!this._average_time_to_emptyIntialized) {
      return undefined;
    }
    return this._refObject.average_time_to_empty;
  }
  set average_time_to_empty(value) {
    this._average_time_to_emptyIntialized = true;
    this._refObject.average_time_to_empty = value;
  }
  get serial_number() {
    if (!this._serial_numberIntialized) {
      return undefined;
    }
    return this._refObject.serial_number;
  }
  set serial_number(value) {
    this._serial_numberIntialized = true;
    this._refObject.serial_number = value;
  }
  get manufacture_date() {
    if (!this._manufacture_dateIntialized) {
      return undefined;
    }
    return this._refObject.manufacture_date;
  }
  set manufacture_date(value) {
    this._manufacture_dateIntialized = true;
    this._refObject.manufacture_date = value;
  }
  get state_of_health() {
    if (!this._state_of_healthIntialized) {
      return undefined;
    }
    return this._refObject.state_of_health;
  }
  set state_of_health(value) {
    this._state_of_healthIntialized = true;
    this._refObject.state_of_health = value;
  }
  get max_error() {
    if (!this._max_errorIntialized) {
      return undefined;
    }
    return this._refObject.max_error;
  }
  set max_error(value) {
    this._max_errorIntialized = true;
    this._refObject.max_error = value;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get interface_error() {
    if (!this._interface_errorIntialized) {
      return undefined;
    }
    return this._refObject.interface_error;
  }
  set interface_error(value) {
    this._interface_errorIntialized = true;
    this._refObject.interface_error = value;
  }
  get voltage_cell_v() {
    return this._wrapperFields['voltage_cell_v'].data;
  }
  set voltage_cell_v(value) {
    if (value.length !== 14) {
      throw new RangeError('The length of the array must be 14.');
    }
    this._wrapperFields['voltage_cell_v'].fill(value);
  }
  get max_cell_voltage_delta() {
    if (!this._max_cell_voltage_deltaIntialized) {
      return undefined;
    }
    return this._refObject.max_cell_voltage_delta;
  }
  set max_cell_voltage_delta(value) {
    this._max_cell_voltage_deltaIntialized = true;
    this._refObject.max_cell_voltage_delta = value;
  }
  get is_powering_off() {
    if (!this._is_powering_offIntialized) {
      return undefined;
    }
    return this._refObject.is_powering_off;
  }
  set is_powering_off(value) {
    this._is_powering_offIntialized = true;
    this._refObject.is_powering_off = value;
  }
  get is_required() {
    if (!this._is_requiredIntialized) {
      return undefined;
    }
    return this._refObject.is_required;
  }
  set is_required(value) {
    this._is_requiredIntialized = true;
    this._refObject.is_required = value;
  }
  get faults() {
    if (!this._faultsIntialized) {
      return undefined;
    }
    return this._refObject.faults;
  }
  set faults(value) {
    this._faultsIntialized = true;
    this._refObject.faults = value;
  }
  get custom_faults() {
    if (!this._custom_faultsIntialized) {
      return undefined;
    }
    return this._refObject.custom_faults;
  }
  set custom_faults(value) {
    this._custom_faultsIntialized = true;
    this._refObject.custom_faults = value;
  }
  get warning() {
    if (!this._warningIntialized) {
      return undefined;
    }
    return this._refObject.warning;
  }
  set warning(value) {
    this._warningIntialized = true;
    this._refObject.warning = value;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get average_power() {
    if (!this._average_powerIntialized) {
      return undefined;
    }
    return this._refObject.average_power;
  }
  set average_power(value) {
    this._average_powerIntialized = true;
    this._refObject.average_power = value;
  }
  get available_energy() {
    if (!this._available_energyIntialized) {
      return undefined;
    }
    return this._refObject.available_energy;
  }
  set available_energy(value) {
    this._available_energyIntialized = true;
    this._refObject.available_energy = value;
  }
  get full_charge_capacity_wh() {
    if (!this._full_charge_capacity_whIntialized) {
      return undefined;
    }
    return this._refObject.full_charge_capacity_wh;
  }
  set full_charge_capacity_wh(value) {
    this._full_charge_capacity_whIntialized = true;
    this._refObject.full_charge_capacity_wh = value;
  }
  get remaining_capacity_wh() {
    if (!this._remaining_capacity_whIntialized) {
      return undefined;
    }
    return this._refObject.remaining_capacity_wh;
  }
  set remaining_capacity_wh(value) {
    this._remaining_capacity_whIntialized = true;
    this._refObject.remaining_capacity_wh = value;
  }
  get design_capacity() {
    if (!this._design_capacityIntialized) {
      return undefined;
    }
    return this._refObject.design_capacity;
  }
  set design_capacity(value) {
    this._design_capacityIntialized = true;
    this._refObject.design_capacity = value;
  }
  get average_time_to_full() {
    if (!this._average_time_to_fullIntialized) {
      return undefined;
    }
    return this._refObject.average_time_to_full;
  }
  set average_time_to_full(value) {
    this._average_time_to_fullIntialized = true;
    this._refObject.average_time_to_full = value;
  }
  get over_discharge_count() {
    if (!this._over_discharge_countIntialized) {
      return undefined;
    }
    return this._refObject.over_discharge_count;
  }
  set over_discharge_count(value) {
    this._over_discharge_countIntialized = true;
    this._refObject.over_discharge_count = value;
  }
  get nominal_voltage() {
    if (!this._nominal_voltageIntialized) {
      return undefined;
    }
    return this._refObject.nominal_voltage;
  }
  set nominal_voltage(value) {
    this._nominal_voltageIntialized = true;
    this._refObject.nominal_voltage = value;
  }
  copyRefObject(refObject) {
    this._refObject = new BatteryStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._connectedIntialized = true;
    this._voltage_vIntialized = true;
    this._voltage_filtered_vIntialized = true;
    this._current_aIntialized = true;
    this._current_filtered_aIntialized = true;
    this._current_average_aIntialized = true;
    this._discharged_mahIntialized = true;
    this._remainingIntialized = true;
    this._scaleIntialized = true;
    this._time_remaining_sIntialized = true;
    this._temperatureIntialized = true;
    this._cell_countIntialized = true;
    this._sourceIntialized = true;
    this._priorityIntialized = true;
    this._capacityIntialized = true;
    this._cycle_countIntialized = true;
    this._average_time_to_emptyIntialized = true;
    this._serial_numberIntialized = true;
    this._manufacture_dateIntialized = true;
    this._state_of_healthIntialized = true;
    this._max_errorIntialized = true;
    this._idIntialized = true;
    this._interface_errorIntialized = true;
    this._wrapperFields.voltage_cell_v.fill(refObject.voltage_cell_v.toArray());
    this._max_cell_voltage_deltaIntialized = true;
    this._is_powering_offIntialized = true;
    this._is_requiredIntialized = true;
    this._faultsIntialized = true;
    this._custom_faultsIntialized = true;
    this._warningIntialized = true;
    this._modeIntialized = true;
    this._average_powerIntialized = true;
    this._available_energyIntialized = true;
    this._full_charge_capacity_whIntialized = true;
    this._remaining_capacity_whIntialized = true;
    this._design_capacityIntialized = true;
    this._average_time_to_fullIntialized = true;
    this._over_discharge_countIntialized = true;
    this._nominal_voltageIntialized = true;
  }
  copy(other) {
    this._refObject = new BatteryStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._connectedIntialized = true;
    this._voltage_vIntialized = true;
    this._voltage_filtered_vIntialized = true;
    this._current_aIntialized = true;
    this._current_filtered_aIntialized = true;
    this._current_average_aIntialized = true;
    this._discharged_mahIntialized = true;
    this._remainingIntialized = true;
    this._scaleIntialized = true;
    this._time_remaining_sIntialized = true;
    this._temperatureIntialized = true;
    this._cell_countIntialized = true;
    this._sourceIntialized = true;
    this._priorityIntialized = true;
    this._capacityIntialized = true;
    this._cycle_countIntialized = true;
    this._average_time_to_emptyIntialized = true;
    this._serial_numberIntialized = true;
    this._manufacture_dateIntialized = true;
    this._state_of_healthIntialized = true;
    this._max_errorIntialized = true;
    this._idIntialized = true;
    this._interface_errorIntialized = true;
    this._wrapperFields.voltage_cell_v.copy(other._wrapperFields.voltage_cell_v);
    this._max_cell_voltage_deltaIntialized = true;
    this._is_powering_offIntialized = true;
    this._is_requiredIntialized = true;
    this._faultsIntialized = true;
    this._custom_faultsIntialized = true;
    this._warningIntialized = true;
    this._modeIntialized = true;
    this._average_powerIntialized = true;
    this._available_energyIntialized = true;
    this._full_charge_capacity_whIntialized = true;
    this._remaining_capacity_whIntialized = true;
    this._design_capacityIntialized = true;
    this._average_time_to_fullIntialized = true;
    this._over_discharge_countIntialized = true;
    this._nominal_voltageIntialized = true;
  }
  static get classType() {
    return BatteryStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"BATTERY_SOURCE_POWER_MODULE","value":0},{"type":"uint8","name":"BATTERY_SOURCE_EXTERNAL","value":1},{"type":"uint8","name":"BATTERY_SOURCE_ESCS","value":2},{"type":"uint8","name":"BATTERY_WARNING_NONE","value":0},{"type":"uint8","name":"BATTERY_WARNING_LOW","value":1},{"type":"uint8","name":"BATTERY_WARNING_CRITICAL","value":2},{"type":"uint8","name":"BATTERY_WARNING_EMERGENCY","value":3},{"type":"uint8","name":"BATTERY_WARNING_FAILED","value":4},{"type":"uint8","name":"BATTERY_STATE_UNHEALTHY","value":6},{"type":"uint8","name":"BATTERY_STATE_CHARGING","value":7},{"type":"uint8","name":"BATTERY_FAULT_DEEP_DISCHARGE","value":0},{"type":"uint8","name":"BATTERY_FAULT_SPIKES","value":1},{"type":"uint8","name":"BATTERY_FAULT_CELL_FAIL","value":2},{"type":"uint8","name":"BATTERY_FAULT_OVER_CURRENT","value":3},{"type":"uint8","name":"BATTERY_FAULT_OVER_TEMPERATURE","value":4},{"type":"uint8","name":"BATTERY_FAULT_UNDER_TEMPERATURE","value":5},{"type":"uint8","name":"BATTERY_FAULT_INCOMPATIBLE_VOLTAGE","value":6},{"type":"uint8","name":"BATTERY_FAULT_INCOMPATIBLE_FIRMWARE","value":7},{"type":"uint8","name":"BATTERY_FAULT_INCOMPATIBLE_MODEL","value":8},{"type":"uint8","name":"BATTERY_FAULT_HARDWARE_FAILURE","value":9},{"type":"uint8","name":"BATTERY_WARNING_OVER_TEMPERATURE","value":10},{"type":"uint8","name":"BATTERY_FAULT_COUNT","value":11},{"type":"uint8","name":"BATTERY_MODE_UNKNOWN","value":0},{"type":"uint8","name":"BATTERY_MODE_AUTO_DISCHARGING","value":1},{"type":"uint8","name":"BATTERY_MODE_HOT_SWAP","value":2},{"type":"uint8","name":"BATTERY_MODE_COUNT","value":3},{"type":"uint8","name":"MAX_INSTANCES","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"connected","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"voltage_v","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"voltage_filtered_v","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_a","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_filtered_a","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"current_average_a","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"discharged_mah","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"remaining","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scale","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_remaining_s","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"temperature","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cell_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"source","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"priority","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"capacity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cycle_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"average_time_to_empty","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"serial_number","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"manufacture_date","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"state_of_health","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"interface_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"voltage_cell_v","type":{"isArray":true,"arraySize":14,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_cell_voltage_delta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_powering_off","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_required","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"faults","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"custom_faults","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"warning","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"average_power","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"available_energy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"full_charge_capacity_wh","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"remaining_capacity_wh","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"design_capacity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"average_time_to_full","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"over_discharge_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nominal_voltage","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"BatteryStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BatteryStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","connected","voltage_v","voltage_filtered_v","current_a","current_filtered_a","current_average_a","discharged_mah","remaining","scale","time_remaining_s","temperature","cell_count","source","priority","capacity","cycle_count","average_time_to_empty","serial_number","manufacture_date","state_of_health","max_error","id","interface_error","voltage_cell_v","max_cell_voltage_delta","is_powering_off","is_required","faults","custom_faults","warning","mode","average_power","available_energy","full_charge_capacity_wh","remaining_capacity_wh","design_capacity","average_time_to_full","over_discharge_count","nominal_voltage"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BatteryStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BatteryStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BatteryStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BatteryStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BatteryStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BatteryStatusRefArray(size);
    this._refObject = new BatteryStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BatteryStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BatteryStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BatteryStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BatteryStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BatteryStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BatteryStatusArrayWrapper;
  }
}
// Define constants (27 in total)
Object.defineProperty(BatteryStatusWrapper, "BATTERY_SOURCE_POWER_MODULE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_SOURCE_EXTERNAL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_SOURCE_ESCS", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_WARNING_NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_WARNING_LOW", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_WARNING_CRITICAL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_WARNING_EMERGENCY", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_WARNING_FAILED", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_STATE_UNHEALTHY", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_STATE_CHARGING", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_DEEP_DISCHARGE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_SPIKES", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_CELL_FAIL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_OVER_CURRENT", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_OVER_TEMPERATURE", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_UNDER_TEMPERATURE", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_INCOMPATIBLE_VOLTAGE", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_INCOMPATIBLE_FIRMWARE", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_INCOMPATIBLE_MODEL", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_HARDWARE_FAILURE", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_WARNING_OVER_TEMPERATURE", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_FAULT_COUNT", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_MODE_UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_MODE_AUTO_DISCHARGING", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_MODE_HOT_SWAP", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "BATTERY_MODE_COUNT", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(BatteryStatusWrapper, "MAX_INSTANCES", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = BatteryStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "BATTERY_SOURCE_POWER_MODULE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "BATTERY_SOURCE_EXTERNAL",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "BATTERY_SOURCE_ESCS",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "BATTERY_WARNING_NONE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "BATTERY_WARNING_LOW",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "BATTERY_WARNING_CRITICAL",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "BATTERY_WARNING_EMERGENCY",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "BATTERY_WARNING_FAILED",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "BATTERY_STATE_UNHEALTHY",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "BATTERY_STATE_CHARGING",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_DEEP_DISCHARGE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_SPIKES",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_CELL_FAIL",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_OVER_CURRENT",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_OVER_TEMPERATURE",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_UNDER_TEMPERATURE",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_INCOMPATIBLE_VOLTAGE",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_INCOMPATIBLE_FIRMWARE",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_INCOMPATIBLE_MODEL",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_HARDWARE_FAILURE",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "BATTERY_WARNING_OVER_TEMPERATURE",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "BATTERY_FAULT_COUNT",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "BATTERY_MODE_UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "BATTERY_MODE_AUTO_DISCHARGING",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "BATTERY_MODE_HOT_SWAP",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "BATTERY_MODE_COUNT",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "MAX_INSTANCES",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "connected",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "voltage_v",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "voltage_filtered_v",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current_a",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current_filtered_a",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "current_average_a",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "discharged_mah",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "remaining",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "scale",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_remaining_s",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "temperature",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cell_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "source",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "priority",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "capacity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cycle_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "average_time_to_empty",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "serial_number",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "manufacture_date",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "state_of_health",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "interface_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "voltage_cell_v",
      "type": {
        "isArray": true,
        "arraySize": 14,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_cell_voltage_delta",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "is_powering_off",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "is_required",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "faults",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "custom_faults",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "warning",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "average_power",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "available_energy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "full_charge_capacity_wh",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "remaining_capacity_wh",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "design_capacity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "average_time_to_full",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "over_discharge_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nominal_voltage",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "BatteryStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BatteryStatus"
}
*/
