// rclnodejs/generated/px4_msgs/px4_msgs__msg__LogMessage.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CharWrapper = require('../../generated/std_msgs/std_msgs__msg__Char.js');
const LogMessageRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  severity: primitiveTypes.uint8,
  text: ArrayType(primitiveTypes.char, 127),
});
const LogMessageRefArray = ArrayType(LogMessageRefStruct);
const LogMessageRefStructArray = StructType({
  data: LogMessageRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LogMessageWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LogMessageRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._severityIntialized = true;
      this._wrapperFields.text = CharWrapper.createArray();
      this._wrapperFields.text.copy(other._wrapperFields.text);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LogMessageRefStruct();
    this._timestampIntialized = false;
    this._severityIntialized = false;
    this._wrapperFields.text = CharWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new LogMessageWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LogMessageArrayWrapper;
  }
  static get ArrayType() {
    return LogMessageArrayWrapper;
  }
  static get refObjectArrayType() {
    return LogMessageRefStructArray
  }
  static get refObjectType() {
    return LogMessageRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in LogMessage');
      }
      if (checkConsistency && !this._severityIntialized) {
        throw new TypeError('Invalid argument: severity in LogMessage');
      }
    this._refObject.text = Array.from(this._wrapperFields.text.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._severityIntialized = true;
    this._refObject.severity = refObject.severity;
    this._wrapperFields.text.fill(refObject.text.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LogMessageWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'LogMessage'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get severity() {
    if (!this._severityIntialized) {
      return undefined;
    }
    return this._refObject.severity;
  }
  set severity(value) {
    this._severityIntialized = true;
    this._refObject.severity = value;
  }
  get text() {
    return this._wrapperFields['text'].data;
  }
  set text(value) {
    if (value.length !== 127) {
      throw new RangeError('The length of the array must be 127.');
    }
    this._wrapperFields['text'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new LogMessageRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._severityIntialized = true;
    this._wrapperFields.text.fill(refObject.text.toArray());
  }
  copy(other) {
    this._refObject = new LogMessageRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._severityIntialized = true;
    this._wrapperFields.text.copy(other._wrapperFields.text);
  }
  static get classType() {
    return LogMessageWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ORB_QUEUE_LENGTH","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"severity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"text","type":{"isArray":true,"arraySize":127,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"char","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"LogMessage","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LogMessage"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","severity","text"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LogMessageArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LogMessageWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LogMessageWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogMessageArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LogMessageArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LogMessageRefArray(size);
    this._refObject = new LogMessageRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LogMessageWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LogMessageArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LogMessageArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LogMessageWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LogMessageWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LogMessageArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(LogMessageWrapper, "ORB_QUEUE_LENGTH", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = LogMessageWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ORB_QUEUE_LENGTH",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "severity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "text",
      "type": {
        "isArray": true,
        "arraySize": 127,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "char",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "LogMessage",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LogMessage"
}
*/
