// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const VehicleStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  armed_time: primitiveTypes.uint64,
  takeoff_time: primitiveTypes.uint64,
  arming_state: primitiveTypes.uint8,
  latest_arming_reason: primitiveTypes.uint8,
  latest_disarming_reason: primitiveTypes.uint8,
  nav_state_timestamp: primitiveTypes.uint64,
  nav_state_user_intention: primitiveTypes.uint8,
  nav_state: primitiveTypes.uint8,
  failure_detector_status: primitiveTypes.uint16,
  hil_state: primitiveTypes.uint8,
  vehicle_type: primitiveTypes.uint8,
  failsafe: primitiveTypes.bool,
  failsafe_and_user_took_over: primitiveTypes.bool,
  gcs_connection_lost: primitiveTypes.bool,
  gcs_connection_lost_counter: primitiveTypes.uint8,
  high_latency_data_link_lost: primitiveTypes.bool,
  is_vtol: primitiveTypes.bool,
  is_vtol_tailsitter: primitiveTypes.bool,
  in_transition_mode: primitiveTypes.bool,
  in_transition_to_fw: primitiveTypes.bool,
  system_type: primitiveTypes.uint8,
  system_id: primitiveTypes.uint8,
  component_id: primitiveTypes.uint8,
  safety_button_available: primitiveTypes.bool,
  safety_off: primitiveTypes.bool,
  power_input_valid: primitiveTypes.bool,
  usb_connected: primitiveTypes.bool,
  open_drone_id_system_present: primitiveTypes.bool,
  open_drone_id_system_healthy: primitiveTypes.bool,
  parachute_system_present: primitiveTypes.bool,
  parachute_system_healthy: primitiveTypes.bool,
  avoidance_system_required: primitiveTypes.bool,
  avoidance_system_valid: primitiveTypes.bool,
  rc_calibration_in_progress: primitiveTypes.bool,
  calibration_enabled: primitiveTypes.bool,
  pre_flight_checks_pass: primitiveTypes.bool,
});
const VehicleStatusRefArray = ArrayType(VehicleStatusRefStruct);
const VehicleStatusRefStructArray = StructType({
  data: VehicleStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._armed_timeIntialized = true;
      this._takeoff_timeIntialized = true;
      this._arming_stateIntialized = true;
      this._latest_arming_reasonIntialized = true;
      this._latest_disarming_reasonIntialized = true;
      this._nav_state_timestampIntialized = true;
      this._nav_state_user_intentionIntialized = true;
      this._nav_stateIntialized = true;
      this._failure_detector_statusIntialized = true;
      this._hil_stateIntialized = true;
      this._vehicle_typeIntialized = true;
      this._failsafeIntialized = true;
      this._failsafe_and_user_took_overIntialized = true;
      this._gcs_connection_lostIntialized = true;
      this._gcs_connection_lost_counterIntialized = true;
      this._high_latency_data_link_lostIntialized = true;
      this._is_vtolIntialized = true;
      this._is_vtol_tailsitterIntialized = true;
      this._in_transition_modeIntialized = true;
      this._in_transition_to_fwIntialized = true;
      this._system_typeIntialized = true;
      this._system_idIntialized = true;
      this._component_idIntialized = true;
      this._safety_button_availableIntialized = true;
      this._safety_offIntialized = true;
      this._power_input_validIntialized = true;
      this._usb_connectedIntialized = true;
      this._open_drone_id_system_presentIntialized = true;
      this._open_drone_id_system_healthyIntialized = true;
      this._parachute_system_presentIntialized = true;
      this._parachute_system_healthyIntialized = true;
      this._avoidance_system_requiredIntialized = true;
      this._avoidance_system_validIntialized = true;
      this._rc_calibration_in_progressIntialized = true;
      this._calibration_enabledIntialized = true;
      this._pre_flight_checks_passIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleStatusRefStruct();
    this._timestampIntialized = false;
    this._armed_timeIntialized = false;
    this._takeoff_timeIntialized = false;
    this._arming_stateIntialized = false;
    this._latest_arming_reasonIntialized = false;
    this._latest_disarming_reasonIntialized = false;
    this._nav_state_timestampIntialized = false;
    this._nav_state_user_intentionIntialized = false;
    this._nav_stateIntialized = false;
    this._failure_detector_statusIntialized = false;
    this._hil_stateIntialized = false;
    this._vehicle_typeIntialized = false;
    this._failsafeIntialized = false;
    this._failsafe_and_user_took_overIntialized = false;
    this._gcs_connection_lostIntialized = false;
    this._gcs_connection_lost_counterIntialized = false;
    this._high_latency_data_link_lostIntialized = false;
    this._is_vtolIntialized = false;
    this._is_vtol_tailsitterIntialized = false;
    this._in_transition_modeIntialized = false;
    this._in_transition_to_fwIntialized = false;
    this._system_typeIntialized = false;
    this._system_idIntialized = false;
    this._component_idIntialized = false;
    this._safety_button_availableIntialized = false;
    this._safety_offIntialized = false;
    this._power_input_validIntialized = false;
    this._usb_connectedIntialized = false;
    this._open_drone_id_system_presentIntialized = false;
    this._open_drone_id_system_healthyIntialized = false;
    this._parachute_system_presentIntialized = false;
    this._parachute_system_healthyIntialized = false;
    this._avoidance_system_requiredIntialized = false;
    this._avoidance_system_validIntialized = false;
    this._rc_calibration_in_progressIntialized = false;
    this._calibration_enabledIntialized = false;
    this._pre_flight_checks_passIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleStatusArrayWrapper;
  }
  static get ArrayType() {
    return VehicleStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleStatusRefStructArray
  }
  static get refObjectType() {
    return VehicleStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleStatus');
      }
      if (checkConsistency && !this._armed_timeIntialized) {
        throw new TypeError('Invalid argument: armed_time in VehicleStatus');
      }
      if (checkConsistency && !this._takeoff_timeIntialized) {
        throw new TypeError('Invalid argument: takeoff_time in VehicleStatus');
      }
      if (checkConsistency && !this._arming_stateIntialized) {
        throw new TypeError('Invalid argument: arming_state in VehicleStatus');
      }
      if (checkConsistency && !this._latest_arming_reasonIntialized) {
        throw new TypeError('Invalid argument: latest_arming_reason in VehicleStatus');
      }
      if (checkConsistency && !this._latest_disarming_reasonIntialized) {
        throw new TypeError('Invalid argument: latest_disarming_reason in VehicleStatus');
      }
      if (checkConsistency && !this._nav_state_timestampIntialized) {
        throw new TypeError('Invalid argument: nav_state_timestamp in VehicleStatus');
      }
      if (checkConsistency && !this._nav_state_user_intentionIntialized) {
        throw new TypeError('Invalid argument: nav_state_user_intention in VehicleStatus');
      }
      if (checkConsistency && !this._nav_stateIntialized) {
        throw new TypeError('Invalid argument: nav_state in VehicleStatus');
      }
      if (checkConsistency && !this._failure_detector_statusIntialized) {
        throw new TypeError('Invalid argument: failure_detector_status in VehicleStatus');
      }
      if (checkConsistency && !this._hil_stateIntialized) {
        throw new TypeError('Invalid argument: hil_state in VehicleStatus');
      }
      if (checkConsistency && !this._vehicle_typeIntialized) {
        throw new TypeError('Invalid argument: vehicle_type in VehicleStatus');
      }
      if (checkConsistency && !this._failsafeIntialized) {
        throw new TypeError('Invalid argument: failsafe in VehicleStatus');
      }
      if (checkConsistency && !this._failsafe_and_user_took_overIntialized) {
        throw new TypeError('Invalid argument: failsafe_and_user_took_over in VehicleStatus');
      }
      if (checkConsistency && !this._gcs_connection_lostIntialized) {
        throw new TypeError('Invalid argument: gcs_connection_lost in VehicleStatus');
      }
      if (checkConsistency && !this._gcs_connection_lost_counterIntialized) {
        throw new TypeError('Invalid argument: gcs_connection_lost_counter in VehicleStatus');
      }
      if (checkConsistency && !this._high_latency_data_link_lostIntialized) {
        throw new TypeError('Invalid argument: high_latency_data_link_lost in VehicleStatus');
      }
      if (checkConsistency && !this._is_vtolIntialized) {
        throw new TypeError('Invalid argument: is_vtol in VehicleStatus');
      }
      if (checkConsistency && !this._is_vtol_tailsitterIntialized) {
        throw new TypeError('Invalid argument: is_vtol_tailsitter in VehicleStatus');
      }
      if (checkConsistency && !this._in_transition_modeIntialized) {
        throw new TypeError('Invalid argument: in_transition_mode in VehicleStatus');
      }
      if (checkConsistency && !this._in_transition_to_fwIntialized) {
        throw new TypeError('Invalid argument: in_transition_to_fw in VehicleStatus');
      }
      if (checkConsistency && !this._system_typeIntialized) {
        throw new TypeError('Invalid argument: system_type in VehicleStatus');
      }
      if (checkConsistency && !this._system_idIntialized) {
        throw new TypeError('Invalid argument: system_id in VehicleStatus');
      }
      if (checkConsistency && !this._component_idIntialized) {
        throw new TypeError('Invalid argument: component_id in VehicleStatus');
      }
      if (checkConsistency && !this._safety_button_availableIntialized) {
        throw new TypeError('Invalid argument: safety_button_available in VehicleStatus');
      }
      if (checkConsistency && !this._safety_offIntialized) {
        throw new TypeError('Invalid argument: safety_off in VehicleStatus');
      }
      if (checkConsistency && !this._power_input_validIntialized) {
        throw new TypeError('Invalid argument: power_input_valid in VehicleStatus');
      }
      if (checkConsistency && !this._usb_connectedIntialized) {
        throw new TypeError('Invalid argument: usb_connected in VehicleStatus');
      }
      if (checkConsistency && !this._open_drone_id_system_presentIntialized) {
        throw new TypeError('Invalid argument: open_drone_id_system_present in VehicleStatus');
      }
      if (checkConsistency && !this._open_drone_id_system_healthyIntialized) {
        throw new TypeError('Invalid argument: open_drone_id_system_healthy in VehicleStatus');
      }
      if (checkConsistency && !this._parachute_system_presentIntialized) {
        throw new TypeError('Invalid argument: parachute_system_present in VehicleStatus');
      }
      if (checkConsistency && !this._parachute_system_healthyIntialized) {
        throw new TypeError('Invalid argument: parachute_system_healthy in VehicleStatus');
      }
      if (checkConsistency && !this._avoidance_system_requiredIntialized) {
        throw new TypeError('Invalid argument: avoidance_system_required in VehicleStatus');
      }
      if (checkConsistency && !this._avoidance_system_validIntialized) {
        throw new TypeError('Invalid argument: avoidance_system_valid in VehicleStatus');
      }
      if (checkConsistency && !this._rc_calibration_in_progressIntialized) {
        throw new TypeError('Invalid argument: rc_calibration_in_progress in VehicleStatus');
      }
      if (checkConsistency && !this._calibration_enabledIntialized) {
        throw new TypeError('Invalid argument: calibration_enabled in VehicleStatus');
      }
      if (checkConsistency && !this._pre_flight_checks_passIntialized) {
        throw new TypeError('Invalid argument: pre_flight_checks_pass in VehicleStatus');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._armed_timeIntialized = true;
    this._refObject.armed_time = refObject.armed_time;
    this._takeoff_timeIntialized = true;
    this._refObject.takeoff_time = refObject.takeoff_time;
    this._arming_stateIntialized = true;
    this._refObject.arming_state = refObject.arming_state;
    this._latest_arming_reasonIntialized = true;
    this._refObject.latest_arming_reason = refObject.latest_arming_reason;
    this._latest_disarming_reasonIntialized = true;
    this._refObject.latest_disarming_reason = refObject.latest_disarming_reason;
    this._nav_state_timestampIntialized = true;
    this._refObject.nav_state_timestamp = refObject.nav_state_timestamp;
    this._nav_state_user_intentionIntialized = true;
    this._refObject.nav_state_user_intention = refObject.nav_state_user_intention;
    this._nav_stateIntialized = true;
    this._refObject.nav_state = refObject.nav_state;
    this._failure_detector_statusIntialized = true;
    this._refObject.failure_detector_status = refObject.failure_detector_status;
    this._hil_stateIntialized = true;
    this._refObject.hil_state = refObject.hil_state;
    this._vehicle_typeIntialized = true;
    this._refObject.vehicle_type = refObject.vehicle_type;
    this._failsafeIntialized = true;
    this._refObject.failsafe = refObject.failsafe;
    this._failsafe_and_user_took_overIntialized = true;
    this._refObject.failsafe_and_user_took_over = refObject.failsafe_and_user_took_over;
    this._gcs_connection_lostIntialized = true;
    this._refObject.gcs_connection_lost = refObject.gcs_connection_lost;
    this._gcs_connection_lost_counterIntialized = true;
    this._refObject.gcs_connection_lost_counter = refObject.gcs_connection_lost_counter;
    this._high_latency_data_link_lostIntialized = true;
    this._refObject.high_latency_data_link_lost = refObject.high_latency_data_link_lost;
    this._is_vtolIntialized = true;
    this._refObject.is_vtol = refObject.is_vtol;
    this._is_vtol_tailsitterIntialized = true;
    this._refObject.is_vtol_tailsitter = refObject.is_vtol_tailsitter;
    this._in_transition_modeIntialized = true;
    this._refObject.in_transition_mode = refObject.in_transition_mode;
    this._in_transition_to_fwIntialized = true;
    this._refObject.in_transition_to_fw = refObject.in_transition_to_fw;
    this._system_typeIntialized = true;
    this._refObject.system_type = refObject.system_type;
    this._system_idIntialized = true;
    this._refObject.system_id = refObject.system_id;
    this._component_idIntialized = true;
    this._refObject.component_id = refObject.component_id;
    this._safety_button_availableIntialized = true;
    this._refObject.safety_button_available = refObject.safety_button_available;
    this._safety_offIntialized = true;
    this._refObject.safety_off = refObject.safety_off;
    this._power_input_validIntialized = true;
    this._refObject.power_input_valid = refObject.power_input_valid;
    this._usb_connectedIntialized = true;
    this._refObject.usb_connected = refObject.usb_connected;
    this._open_drone_id_system_presentIntialized = true;
    this._refObject.open_drone_id_system_present = refObject.open_drone_id_system_present;
    this._open_drone_id_system_healthyIntialized = true;
    this._refObject.open_drone_id_system_healthy = refObject.open_drone_id_system_healthy;
    this._parachute_system_presentIntialized = true;
    this._refObject.parachute_system_present = refObject.parachute_system_present;
    this._parachute_system_healthyIntialized = true;
    this._refObject.parachute_system_healthy = refObject.parachute_system_healthy;
    this._avoidance_system_requiredIntialized = true;
    this._refObject.avoidance_system_required = refObject.avoidance_system_required;
    this._avoidance_system_validIntialized = true;
    this._refObject.avoidance_system_valid = refObject.avoidance_system_valid;
    this._rc_calibration_in_progressIntialized = true;
    this._refObject.rc_calibration_in_progress = refObject.rc_calibration_in_progress;
    this._calibration_enabledIntialized = true;
    this._refObject.calibration_enabled = refObject.calibration_enabled;
    this._pre_flight_checks_passIntialized = true;
    this._refObject.pre_flight_checks_pass = refObject.pre_flight_checks_pass;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get armed_time() {
    if (!this._armed_timeIntialized) {
      return undefined;
    }
    return this._refObject.armed_time;
  }
  set armed_time(value) {
    this._armed_timeIntialized = true;
    this._refObject.armed_time = value;
  }
  get takeoff_time() {
    if (!this._takeoff_timeIntialized) {
      return undefined;
    }
    return this._refObject.takeoff_time;
  }
  set takeoff_time(value) {
    this._takeoff_timeIntialized = true;
    this._refObject.takeoff_time = value;
  }
  get arming_state() {
    if (!this._arming_stateIntialized) {
      return undefined;
    }
    return this._refObject.arming_state;
  }
  set arming_state(value) {
    this._arming_stateIntialized = true;
    this._refObject.arming_state = value;
  }
  get latest_arming_reason() {
    if (!this._latest_arming_reasonIntialized) {
      return undefined;
    }
    return this._refObject.latest_arming_reason;
  }
  set latest_arming_reason(value) {
    this._latest_arming_reasonIntialized = true;
    this._refObject.latest_arming_reason = value;
  }
  get latest_disarming_reason() {
    if (!this._latest_disarming_reasonIntialized) {
      return undefined;
    }
    return this._refObject.latest_disarming_reason;
  }
  set latest_disarming_reason(value) {
    this._latest_disarming_reasonIntialized = true;
    this._refObject.latest_disarming_reason = value;
  }
  get nav_state_timestamp() {
    if (!this._nav_state_timestampIntialized) {
      return undefined;
    }
    return this._refObject.nav_state_timestamp;
  }
  set nav_state_timestamp(value) {
    this._nav_state_timestampIntialized = true;
    this._refObject.nav_state_timestamp = value;
  }
  get nav_state_user_intention() {
    if (!this._nav_state_user_intentionIntialized) {
      return undefined;
    }
    return this._refObject.nav_state_user_intention;
  }
  set nav_state_user_intention(value) {
    this._nav_state_user_intentionIntialized = true;
    this._refObject.nav_state_user_intention = value;
  }
  get nav_state() {
    if (!this._nav_stateIntialized) {
      return undefined;
    }
    return this._refObject.nav_state;
  }
  set nav_state(value) {
    this._nav_stateIntialized = true;
    this._refObject.nav_state = value;
  }
  get failure_detector_status() {
    if (!this._failure_detector_statusIntialized) {
      return undefined;
    }
    return this._refObject.failure_detector_status;
  }
  set failure_detector_status(value) {
    this._failure_detector_statusIntialized = true;
    this._refObject.failure_detector_status = value;
  }
  get hil_state() {
    if (!this._hil_stateIntialized) {
      return undefined;
    }
    return this._refObject.hil_state;
  }
  set hil_state(value) {
    this._hil_stateIntialized = true;
    this._refObject.hil_state = value;
  }
  get vehicle_type() {
    if (!this._vehicle_typeIntialized) {
      return undefined;
    }
    return this._refObject.vehicle_type;
  }
  set vehicle_type(value) {
    this._vehicle_typeIntialized = true;
    this._refObject.vehicle_type = value;
  }
  get failsafe() {
    if (!this._failsafeIntialized) {
      return undefined;
    }
    return this._refObject.failsafe;
  }
  set failsafe(value) {
    this._failsafeIntialized = true;
    this._refObject.failsafe = value;
  }
  get failsafe_and_user_took_over() {
    if (!this._failsafe_and_user_took_overIntialized) {
      return undefined;
    }
    return this._refObject.failsafe_and_user_took_over;
  }
  set failsafe_and_user_took_over(value) {
    this._failsafe_and_user_took_overIntialized = true;
    this._refObject.failsafe_and_user_took_over = value;
  }
  get gcs_connection_lost() {
    if (!this._gcs_connection_lostIntialized) {
      return undefined;
    }
    return this._refObject.gcs_connection_lost;
  }
  set gcs_connection_lost(value) {
    this._gcs_connection_lostIntialized = true;
    this._refObject.gcs_connection_lost = value;
  }
  get gcs_connection_lost_counter() {
    if (!this._gcs_connection_lost_counterIntialized) {
      return undefined;
    }
    return this._refObject.gcs_connection_lost_counter;
  }
  set gcs_connection_lost_counter(value) {
    this._gcs_connection_lost_counterIntialized = true;
    this._refObject.gcs_connection_lost_counter = value;
  }
  get high_latency_data_link_lost() {
    if (!this._high_latency_data_link_lostIntialized) {
      return undefined;
    }
    return this._refObject.high_latency_data_link_lost;
  }
  set high_latency_data_link_lost(value) {
    this._high_latency_data_link_lostIntialized = true;
    this._refObject.high_latency_data_link_lost = value;
  }
  get is_vtol() {
    if (!this._is_vtolIntialized) {
      return undefined;
    }
    return this._refObject.is_vtol;
  }
  set is_vtol(value) {
    this._is_vtolIntialized = true;
    this._refObject.is_vtol = value;
  }
  get is_vtol_tailsitter() {
    if (!this._is_vtol_tailsitterIntialized) {
      return undefined;
    }
    return this._refObject.is_vtol_tailsitter;
  }
  set is_vtol_tailsitter(value) {
    this._is_vtol_tailsitterIntialized = true;
    this._refObject.is_vtol_tailsitter = value;
  }
  get in_transition_mode() {
    if (!this._in_transition_modeIntialized) {
      return undefined;
    }
    return this._refObject.in_transition_mode;
  }
  set in_transition_mode(value) {
    this._in_transition_modeIntialized = true;
    this._refObject.in_transition_mode = value;
  }
  get in_transition_to_fw() {
    if (!this._in_transition_to_fwIntialized) {
      return undefined;
    }
    return this._refObject.in_transition_to_fw;
  }
  set in_transition_to_fw(value) {
    this._in_transition_to_fwIntialized = true;
    this._refObject.in_transition_to_fw = value;
  }
  get system_type() {
    if (!this._system_typeIntialized) {
      return undefined;
    }
    return this._refObject.system_type;
  }
  set system_type(value) {
    this._system_typeIntialized = true;
    this._refObject.system_type = value;
  }
  get system_id() {
    if (!this._system_idIntialized) {
      return undefined;
    }
    return this._refObject.system_id;
  }
  set system_id(value) {
    this._system_idIntialized = true;
    this._refObject.system_id = value;
  }
  get component_id() {
    if (!this._component_idIntialized) {
      return undefined;
    }
    return this._refObject.component_id;
  }
  set component_id(value) {
    this._component_idIntialized = true;
    this._refObject.component_id = value;
  }
  get safety_button_available() {
    if (!this._safety_button_availableIntialized) {
      return undefined;
    }
    return this._refObject.safety_button_available;
  }
  set safety_button_available(value) {
    this._safety_button_availableIntialized = true;
    this._refObject.safety_button_available = value;
  }
  get safety_off() {
    if (!this._safety_offIntialized) {
      return undefined;
    }
    return this._refObject.safety_off;
  }
  set safety_off(value) {
    this._safety_offIntialized = true;
    this._refObject.safety_off = value;
  }
  get power_input_valid() {
    if (!this._power_input_validIntialized) {
      return undefined;
    }
    return this._refObject.power_input_valid;
  }
  set power_input_valid(value) {
    this._power_input_validIntialized = true;
    this._refObject.power_input_valid = value;
  }
  get usb_connected() {
    if (!this._usb_connectedIntialized) {
      return undefined;
    }
    return this._refObject.usb_connected;
  }
  set usb_connected(value) {
    this._usb_connectedIntialized = true;
    this._refObject.usb_connected = value;
  }
  get open_drone_id_system_present() {
    if (!this._open_drone_id_system_presentIntialized) {
      return undefined;
    }
    return this._refObject.open_drone_id_system_present;
  }
  set open_drone_id_system_present(value) {
    this._open_drone_id_system_presentIntialized = true;
    this._refObject.open_drone_id_system_present = value;
  }
  get open_drone_id_system_healthy() {
    if (!this._open_drone_id_system_healthyIntialized) {
      return undefined;
    }
    return this._refObject.open_drone_id_system_healthy;
  }
  set open_drone_id_system_healthy(value) {
    this._open_drone_id_system_healthyIntialized = true;
    this._refObject.open_drone_id_system_healthy = value;
  }
  get parachute_system_present() {
    if (!this._parachute_system_presentIntialized) {
      return undefined;
    }
    return this._refObject.parachute_system_present;
  }
  set parachute_system_present(value) {
    this._parachute_system_presentIntialized = true;
    this._refObject.parachute_system_present = value;
  }
  get parachute_system_healthy() {
    if (!this._parachute_system_healthyIntialized) {
      return undefined;
    }
    return this._refObject.parachute_system_healthy;
  }
  set parachute_system_healthy(value) {
    this._parachute_system_healthyIntialized = true;
    this._refObject.parachute_system_healthy = value;
  }
  get avoidance_system_required() {
    if (!this._avoidance_system_requiredIntialized) {
      return undefined;
    }
    return this._refObject.avoidance_system_required;
  }
  set avoidance_system_required(value) {
    this._avoidance_system_requiredIntialized = true;
    this._refObject.avoidance_system_required = value;
  }
  get avoidance_system_valid() {
    if (!this._avoidance_system_validIntialized) {
      return undefined;
    }
    return this._refObject.avoidance_system_valid;
  }
  set avoidance_system_valid(value) {
    this._avoidance_system_validIntialized = true;
    this._refObject.avoidance_system_valid = value;
  }
  get rc_calibration_in_progress() {
    if (!this._rc_calibration_in_progressIntialized) {
      return undefined;
    }
    return this._refObject.rc_calibration_in_progress;
  }
  set rc_calibration_in_progress(value) {
    this._rc_calibration_in_progressIntialized = true;
    this._refObject.rc_calibration_in_progress = value;
  }
  get calibration_enabled() {
    if (!this._calibration_enabledIntialized) {
      return undefined;
    }
    return this._refObject.calibration_enabled;
  }
  set calibration_enabled(value) {
    this._calibration_enabledIntialized = true;
    this._refObject.calibration_enabled = value;
  }
  get pre_flight_checks_pass() {
    if (!this._pre_flight_checks_passIntialized) {
      return undefined;
    }
    return this._refObject.pre_flight_checks_pass;
  }
  set pre_flight_checks_pass(value) {
    this._pre_flight_checks_passIntialized = true;
    this._refObject.pre_flight_checks_pass = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._armed_timeIntialized = true;
    this._takeoff_timeIntialized = true;
    this._arming_stateIntialized = true;
    this._latest_arming_reasonIntialized = true;
    this._latest_disarming_reasonIntialized = true;
    this._nav_state_timestampIntialized = true;
    this._nav_state_user_intentionIntialized = true;
    this._nav_stateIntialized = true;
    this._failure_detector_statusIntialized = true;
    this._hil_stateIntialized = true;
    this._vehicle_typeIntialized = true;
    this._failsafeIntialized = true;
    this._failsafe_and_user_took_overIntialized = true;
    this._gcs_connection_lostIntialized = true;
    this._gcs_connection_lost_counterIntialized = true;
    this._high_latency_data_link_lostIntialized = true;
    this._is_vtolIntialized = true;
    this._is_vtol_tailsitterIntialized = true;
    this._in_transition_modeIntialized = true;
    this._in_transition_to_fwIntialized = true;
    this._system_typeIntialized = true;
    this._system_idIntialized = true;
    this._component_idIntialized = true;
    this._safety_button_availableIntialized = true;
    this._safety_offIntialized = true;
    this._power_input_validIntialized = true;
    this._usb_connectedIntialized = true;
    this._open_drone_id_system_presentIntialized = true;
    this._open_drone_id_system_healthyIntialized = true;
    this._parachute_system_presentIntialized = true;
    this._parachute_system_healthyIntialized = true;
    this._avoidance_system_requiredIntialized = true;
    this._avoidance_system_validIntialized = true;
    this._rc_calibration_in_progressIntialized = true;
    this._calibration_enabledIntialized = true;
    this._pre_flight_checks_passIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._armed_timeIntialized = true;
    this._takeoff_timeIntialized = true;
    this._arming_stateIntialized = true;
    this._latest_arming_reasonIntialized = true;
    this._latest_disarming_reasonIntialized = true;
    this._nav_state_timestampIntialized = true;
    this._nav_state_user_intentionIntialized = true;
    this._nav_stateIntialized = true;
    this._failure_detector_statusIntialized = true;
    this._hil_stateIntialized = true;
    this._vehicle_typeIntialized = true;
    this._failsafeIntialized = true;
    this._failsafe_and_user_took_overIntialized = true;
    this._gcs_connection_lostIntialized = true;
    this._gcs_connection_lost_counterIntialized = true;
    this._high_latency_data_link_lostIntialized = true;
    this._is_vtolIntialized = true;
    this._is_vtol_tailsitterIntialized = true;
    this._in_transition_modeIntialized = true;
    this._in_transition_to_fwIntialized = true;
    this._system_typeIntialized = true;
    this._system_idIntialized = true;
    this._component_idIntialized = true;
    this._safety_button_availableIntialized = true;
    this._safety_offIntialized = true;
    this._power_input_validIntialized = true;
    this._usb_connectedIntialized = true;
    this._open_drone_id_system_presentIntialized = true;
    this._open_drone_id_system_healthyIntialized = true;
    this._parachute_system_presentIntialized = true;
    this._parachute_system_healthyIntialized = true;
    this._avoidance_system_requiredIntialized = true;
    this._avoidance_system_validIntialized = true;
    this._rc_calibration_in_progressIntialized = true;
    this._calibration_enabledIntialized = true;
    this._pre_flight_checks_passIntialized = true;
  }
  static get classType() {
    return VehicleStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ARMING_STATE_INIT","value":0},{"type":"uint8","name":"ARMING_STATE_STANDBY","value":1},{"type":"uint8","name":"ARMING_STATE_ARMED","value":2},{"type":"uint8","name":"ARMING_STATE_STANDBY_ERROR","value":3},{"type":"uint8","name":"ARMING_STATE_SHUTDOWN","value":4},{"type":"uint8","name":"ARMING_STATE_IN_AIR_RESTORE","value":5},{"type":"uint8","name":"ARMING_STATE_MAX","value":6},{"type":"uint8","name":"ARM_DISARM_REASON_TRANSITION_TO_STANDBY","value":0},{"type":"uint8","name":"ARM_DISARM_REASON_RC_STICK","value":1},{"type":"uint8","name":"ARM_DISARM_REASON_RC_SWITCH","value":2},{"type":"uint8","name":"ARM_DISARM_REASON_COMMAND_INTERNAL","value":3},{"type":"uint8","name":"ARM_DISARM_REASON_COMMAND_EXTERNAL","value":4},{"type":"uint8","name":"ARM_DISARM_REASON_MISSION_START","value":5},{"type":"uint8","name":"ARM_DISARM_REASON_SAFETY_BUTTON","value":6},{"type":"uint8","name":"ARM_DISARM_REASON_AUTO_DISARM_LAND","value":7},{"type":"uint8","name":"ARM_DISARM_REASON_AUTO_DISARM_PREFLIGHT","value":8},{"type":"uint8","name":"ARM_DISARM_REASON_KILL_SWITCH","value":9},{"type":"uint8","name":"ARM_DISARM_REASON_LOCKDOWN","value":10},{"type":"uint8","name":"ARM_DISARM_REASON_FAILURE_DETECTOR","value":11},{"type":"uint8","name":"ARM_DISARM_REASON_SHUTDOWN","value":12},{"type":"uint8","name":"ARM_DISARM_REASON_UNIT_TEST","value":13},{"type":"uint8","name":"NAVIGATION_STATE_MANUAL","value":0},{"type":"uint8","name":"NAVIGATION_STATE_ALTCTL","value":1},{"type":"uint8","name":"NAVIGATION_STATE_POSCTL","value":2},{"type":"uint8","name":"NAVIGATION_STATE_AUTO_MISSION","value":3},{"type":"uint8","name":"NAVIGATION_STATE_AUTO_LOITER","value":4},{"type":"uint8","name":"NAVIGATION_STATE_AUTO_RTL","value":5},{"type":"uint8","name":"NAVIGATION_STATE_UNUSED3","value":8},{"type":"uint8","name":"NAVIGATION_STATE_UNUSED","value":9},{"type":"uint8","name":"NAVIGATION_STATE_ACRO","value":10},{"type":"uint8","name":"NAVIGATION_STATE_UNUSED1","value":11},{"type":"uint8","name":"NAVIGATION_STATE_DESCEND","value":12},{"type":"uint8","name":"NAVIGATION_STATE_TERMINATION","value":13},{"type":"uint8","name":"NAVIGATION_STATE_OFFBOARD","value":14},{"type":"uint8","name":"NAVIGATION_STATE_STAB","value":15},{"type":"uint8","name":"NAVIGATION_STATE_UNUSED2","value":16},{"type":"uint8","name":"NAVIGATION_STATE_AUTO_TAKEOFF","value":17},{"type":"uint8","name":"NAVIGATION_STATE_AUTO_LAND","value":18},{"type":"uint8","name":"NAVIGATION_STATE_AUTO_FOLLOW_TARGET","value":19},{"type":"uint8","name":"NAVIGATION_STATE_AUTO_PRECLAND","value":20},{"type":"uint8","name":"NAVIGATION_STATE_ORBIT","value":21},{"type":"uint8","name":"NAVIGATION_STATE_AUTO_VTOL_TAKEOFF","value":22},{"type":"uint8","name":"NAVIGATION_STATE_MAX","value":23},{"type":"uint16","name":"FAILURE_NONE","value":0},{"type":"uint16","name":"FAILURE_ROLL","value":1},{"type":"uint16","name":"FAILURE_PITCH","value":2},{"type":"uint16","name":"FAILURE_ALT","value":4},{"type":"uint16","name":"FAILURE_EXT","value":8},{"type":"uint16","name":"FAILURE_ARM_ESC","value":16},{"type":"uint16","name":"FAILURE_BATTERY","value":32},{"type":"uint16","name":"FAILURE_IMBALANCED_PROP","value":64},{"type":"uint16","name":"FAILURE_MOTOR","value":128},{"type":"uint8","name":"HIL_STATE_OFF","value":0},{"type":"uint8","name":"HIL_STATE_ON","value":1},{"type":"uint8","name":"VEHICLE_TYPE_UNKNOWN","value":0},{"type":"uint8","name":"VEHICLE_TYPE_ROTARY_WING","value":1},{"type":"uint8","name":"VEHICLE_TYPE_FIXED_WING","value":2},{"type":"uint8","name":"VEHICLE_TYPE_ROVER","value":3},{"type":"uint8","name":"VEHICLE_TYPE_AIRSHIP","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"armed_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"takeoff_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"arming_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"latest_arming_reason","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"latest_disarming_reason","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nav_state_timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nav_state_user_intention","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"nav_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"failure_detector_status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hil_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vehicle_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"failsafe","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"failsafe_and_user_took_over","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gcs_connection_lost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gcs_connection_lost_counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"high_latency_data_link_lost","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_vtol","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_vtol_tailsitter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"in_transition_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"in_transition_to_fw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"system_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"system_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"component_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"safety_button_available","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"safety_off","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"power_input_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"usb_connected","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"open_drone_id_system_present","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"open_drone_id_system_healthy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"parachute_system_present","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"parachute_system_healthy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"avoidance_system_required","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"avoidance_system_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rc_calibration_in_progress","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"calibration_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pre_flight_checks_pass","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","armed_time","takeoff_time","arming_state","latest_arming_reason","latest_disarming_reason","nav_state_timestamp","nav_state_user_intention","nav_state","failure_detector_status","hil_state","vehicle_type","failsafe","failsafe_and_user_took_over","gcs_connection_lost","gcs_connection_lost_counter","high_latency_data_link_lost","is_vtol","is_vtol_tailsitter","in_transition_mode","in_transition_to_fw","system_type","system_id","component_id","safety_button_available","safety_off","power_input_valid","usb_connected","open_drone_id_system_present","open_drone_id_system_healthy","parachute_system_present","parachute_system_healthy","avoidance_system_required","avoidance_system_valid","rc_calibration_in_progress","calibration_enabled","pre_flight_checks_pass"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleStatusRefArray(size);
    this._refObject = new VehicleStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleStatusArrayWrapper;
  }
}
// Define constants (59 in total)
Object.defineProperty(VehicleStatusWrapper, "ARMING_STATE_INIT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARMING_STATE_STANDBY", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARMING_STATE_ARMED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARMING_STATE_STANDBY_ERROR", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARMING_STATE_SHUTDOWN", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARMING_STATE_IN_AIR_RESTORE", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARMING_STATE_MAX", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_TRANSITION_TO_STANDBY", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_RC_STICK", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_RC_SWITCH", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_COMMAND_INTERNAL", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_COMMAND_EXTERNAL", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_MISSION_START", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_SAFETY_BUTTON", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_AUTO_DISARM_LAND", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_AUTO_DISARM_PREFLIGHT", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_KILL_SWITCH", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_LOCKDOWN", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_FAILURE_DETECTOR", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_SHUTDOWN", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "ARM_DISARM_REASON_UNIT_TEST", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_MANUAL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_ALTCTL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_POSCTL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_AUTO_MISSION", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_AUTO_LOITER", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_AUTO_RTL", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_UNUSED3", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_UNUSED", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_ACRO", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_UNUSED1", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_DESCEND", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_TERMINATION", {value: 13, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_OFFBOARD", {value: 14, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_STAB", {value: 15, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_UNUSED2", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_AUTO_TAKEOFF", {value: 17, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_AUTO_LAND", {value: 18, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_AUTO_FOLLOW_TARGET", {value: 19, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_AUTO_PRECLAND", {value: 20, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_ORBIT", {value: 21, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_AUTO_VTOL_TAKEOFF", {value: 22, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "NAVIGATION_STATE_MAX", {value: 23, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "FAILURE_NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "FAILURE_ROLL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "FAILURE_PITCH", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "FAILURE_ALT", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "FAILURE_EXT", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "FAILURE_ARM_ESC", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "FAILURE_BATTERY", {value: 32, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "FAILURE_IMBALANCED_PROP", {value: 64, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "FAILURE_MOTOR", {value: 128, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "HIL_STATE_OFF", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "HIL_STATE_ON", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "VEHICLE_TYPE_UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "VEHICLE_TYPE_ROTARY_WING", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "VEHICLE_TYPE_FIXED_WING", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "VEHICLE_TYPE_ROVER", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleStatusWrapper, "VEHICLE_TYPE_AIRSHIP", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = VehicleStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ARMING_STATE_INIT",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ARMING_STATE_STANDBY",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "ARMING_STATE_ARMED",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ARMING_STATE_STANDBY_ERROR",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "ARMING_STATE_SHUTDOWN",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "ARMING_STATE_IN_AIR_RESTORE",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "ARMING_STATE_MAX",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_TRANSITION_TO_STANDBY",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_RC_STICK",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_RC_SWITCH",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_COMMAND_INTERNAL",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_COMMAND_EXTERNAL",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_MISSION_START",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_SAFETY_BUTTON",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_AUTO_DISARM_LAND",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_AUTO_DISARM_PREFLIGHT",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_KILL_SWITCH",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_LOCKDOWN",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_FAILURE_DETECTOR",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_SHUTDOWN",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "ARM_DISARM_REASON_UNIT_TEST",
      "value": 13
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_MANUAL",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_ALTCTL",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_POSCTL",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_AUTO_MISSION",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_AUTO_LOITER",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_AUTO_RTL",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_UNUSED3",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_UNUSED",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_ACRO",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_UNUSED1",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_DESCEND",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_TERMINATION",
      "value": 13
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_OFFBOARD",
      "value": 14
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_STAB",
      "value": 15
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_UNUSED2",
      "value": 16
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_AUTO_TAKEOFF",
      "value": 17
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_AUTO_LAND",
      "value": 18
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_AUTO_FOLLOW_TARGET",
      "value": 19
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_AUTO_PRECLAND",
      "value": 20
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_ORBIT",
      "value": 21
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_AUTO_VTOL_TAKEOFF",
      "value": 22
    },
    {
      "type": "uint8",
      "name": "NAVIGATION_STATE_MAX",
      "value": 23
    },
    {
      "type": "uint16",
      "name": "FAILURE_NONE",
      "value": 0
    },
    {
      "type": "uint16",
      "name": "FAILURE_ROLL",
      "value": 1
    },
    {
      "type": "uint16",
      "name": "FAILURE_PITCH",
      "value": 2
    },
    {
      "type": "uint16",
      "name": "FAILURE_ALT",
      "value": 4
    },
    {
      "type": "uint16",
      "name": "FAILURE_EXT",
      "value": 8
    },
    {
      "type": "uint16",
      "name": "FAILURE_ARM_ESC",
      "value": 16
    },
    {
      "type": "uint16",
      "name": "FAILURE_BATTERY",
      "value": 32
    },
    {
      "type": "uint16",
      "name": "FAILURE_IMBALANCED_PROP",
      "value": 64
    },
    {
      "type": "uint16",
      "name": "FAILURE_MOTOR",
      "value": 128
    },
    {
      "type": "uint8",
      "name": "HIL_STATE_OFF",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "HIL_STATE_ON",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "VEHICLE_TYPE_UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "VEHICLE_TYPE_ROTARY_WING",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "VEHICLE_TYPE_FIXED_WING",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "VEHICLE_TYPE_ROVER",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "VEHICLE_TYPE_AIRSHIP",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "armed_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "takeoff_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "arming_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "latest_arming_reason",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "latest_disarming_reason",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nav_state_timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nav_state_user_intention",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "nav_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "failure_detector_status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hil_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vehicle_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "failsafe",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "failsafe_and_user_took_over",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gcs_connection_lost",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gcs_connection_lost_counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "high_latency_data_link_lost",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "is_vtol",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "is_vtol_tailsitter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "in_transition_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "in_transition_to_fw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "system_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "system_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "component_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "safety_button_available",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "safety_off",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "power_input_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "usb_connected",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "open_drone_id_system_present",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "open_drone_id_system_healthy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "parachute_system_present",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "parachute_system_healthy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "avoidance_system_required",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "avoidance_system_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rc_calibration_in_progress",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "calibration_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pre_flight_checks_pass",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleStatus"
}
*/
