// rclnodejs/generated/px4_msgs/px4_msgs__msg__SystemPower.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const SystemPowerRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  voltage5v_v: primitiveTypes.float32,
  sensors3v3: ArrayType(primitiveTypes.float32, 4),
  sensors3v3_valid: primitiveTypes.uint8,
  usb_connected: primitiveTypes.uint8,
  brick_valid: primitiveTypes.uint8,
  usb_valid: primitiveTypes.uint8,
  servo_valid: primitiveTypes.uint8,
  periph_5v_oc: primitiveTypes.uint8,
  hipower_5v_oc: primitiveTypes.uint8,
  comp_5v_valid: primitiveTypes.uint8,
  can1_gps1_5v_valid: primitiveTypes.uint8,
});
const SystemPowerRefArray = ArrayType(SystemPowerRefStruct);
const SystemPowerRefStructArray = StructType({
  data: SystemPowerRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SystemPowerWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SystemPowerRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._voltage5v_vIntialized = true;
      this._wrapperFields.sensors3v3 = Float32Wrapper.createArray();
      this._wrapperFields.sensors3v3.copy(other._wrapperFields.sensors3v3);
      this._sensors3v3_validIntialized = true;
      this._usb_connectedIntialized = true;
      this._brick_validIntialized = true;
      this._usb_validIntialized = true;
      this._servo_validIntialized = true;
      this._periph_5v_ocIntialized = true;
      this._hipower_5v_ocIntialized = true;
      this._comp_5v_validIntialized = true;
      this._can1_gps1_5v_validIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SystemPowerRefStruct();
    this._timestampIntialized = false;
    this._voltage5v_vIntialized = false;
    this._wrapperFields.sensors3v3 = Float32Wrapper.createArray();
    this._sensors3v3_validIntialized = false;
    this._usb_connectedIntialized = false;
    this._brick_validIntialized = false;
    this._usb_validIntialized = false;
    this._servo_validIntialized = false;
    this._periph_5v_ocIntialized = false;
    this._hipower_5v_ocIntialized = false;
    this._comp_5v_validIntialized = false;
    this._can1_gps1_5v_validIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SystemPowerWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SystemPowerArrayWrapper;
  }
  static get ArrayType() {
    return SystemPowerArrayWrapper;
  }
  static get refObjectArrayType() {
    return SystemPowerRefStructArray
  }
  static get refObjectType() {
    return SystemPowerRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in SystemPower');
      }
      if (checkConsistency && !this._voltage5v_vIntialized) {
        throw new TypeError('Invalid argument: voltage5v_v in SystemPower');
      }
      if (checkConsistency && !this._sensors3v3_validIntialized) {
        throw new TypeError('Invalid argument: sensors3v3_valid in SystemPower');
      }
      if (checkConsistency && !this._usb_connectedIntialized) {
        throw new TypeError('Invalid argument: usb_connected in SystemPower');
      }
      if (checkConsistency && !this._brick_validIntialized) {
        throw new TypeError('Invalid argument: brick_valid in SystemPower');
      }
      if (checkConsistency && !this._usb_validIntialized) {
        throw new TypeError('Invalid argument: usb_valid in SystemPower');
      }
      if (checkConsistency && !this._servo_validIntialized) {
        throw new TypeError('Invalid argument: servo_valid in SystemPower');
      }
      if (checkConsistency && !this._periph_5v_ocIntialized) {
        throw new TypeError('Invalid argument: periph_5v_oc in SystemPower');
      }
      if (checkConsistency && !this._hipower_5v_ocIntialized) {
        throw new TypeError('Invalid argument: hipower_5v_oc in SystemPower');
      }
      if (checkConsistency && !this._comp_5v_validIntialized) {
        throw new TypeError('Invalid argument: comp_5v_valid in SystemPower');
      }
      if (checkConsistency && !this._can1_gps1_5v_validIntialized) {
        throw new TypeError('Invalid argument: can1_gps1_5v_valid in SystemPower');
      }
    this._refObject.sensors3v3 = Array.from(this._wrapperFields.sensors3v3.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._voltage5v_vIntialized = true;
    this._refObject.voltage5v_v = refObject.voltage5v_v;
    this._wrapperFields.sensors3v3.fill(refObject.sensors3v3.toArray());
    this._sensors3v3_validIntialized = true;
    this._refObject.sensors3v3_valid = refObject.sensors3v3_valid;
    this._usb_connectedIntialized = true;
    this._refObject.usb_connected = refObject.usb_connected;
    this._brick_validIntialized = true;
    this._refObject.brick_valid = refObject.brick_valid;
    this._usb_validIntialized = true;
    this._refObject.usb_valid = refObject.usb_valid;
    this._servo_validIntialized = true;
    this._refObject.servo_valid = refObject.servo_valid;
    this._periph_5v_ocIntialized = true;
    this._refObject.periph_5v_oc = refObject.periph_5v_oc;
    this._hipower_5v_ocIntialized = true;
    this._refObject.hipower_5v_oc = refObject.hipower_5v_oc;
    this._comp_5v_validIntialized = true;
    this._refObject.comp_5v_valid = refObject.comp_5v_valid;
    this._can1_gps1_5v_validIntialized = true;
    this._refObject.can1_gps1_5v_valid = refObject.can1_gps1_5v_valid;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SystemPowerWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'SystemPower'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get voltage5v_v() {
    if (!this._voltage5v_vIntialized) {
      return undefined;
    }
    return this._refObject.voltage5v_v;
  }
  set voltage5v_v(value) {
    this._voltage5v_vIntialized = true;
    this._refObject.voltage5v_v = value;
  }
  get sensors3v3() {
    return this._wrapperFields['sensors3v3'].data;
  }
  set sensors3v3(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields['sensors3v3'].fill(value);
  }
  get sensors3v3_valid() {
    if (!this._sensors3v3_validIntialized) {
      return undefined;
    }
    return this._refObject.sensors3v3_valid;
  }
  set sensors3v3_valid(value) {
    this._sensors3v3_validIntialized = true;
    this._refObject.sensors3v3_valid = value;
  }
  get usb_connected() {
    if (!this._usb_connectedIntialized) {
      return undefined;
    }
    return this._refObject.usb_connected;
  }
  set usb_connected(value) {
    this._usb_connectedIntialized = true;
    this._refObject.usb_connected = value;
  }
  get brick_valid() {
    if (!this._brick_validIntialized) {
      return undefined;
    }
    return this._refObject.brick_valid;
  }
  set brick_valid(value) {
    this._brick_validIntialized = true;
    this._refObject.brick_valid = value;
  }
  get usb_valid() {
    if (!this._usb_validIntialized) {
      return undefined;
    }
    return this._refObject.usb_valid;
  }
  set usb_valid(value) {
    this._usb_validIntialized = true;
    this._refObject.usb_valid = value;
  }
  get servo_valid() {
    if (!this._servo_validIntialized) {
      return undefined;
    }
    return this._refObject.servo_valid;
  }
  set servo_valid(value) {
    this._servo_validIntialized = true;
    this._refObject.servo_valid = value;
  }
  get periph_5v_oc() {
    if (!this._periph_5v_ocIntialized) {
      return undefined;
    }
    return this._refObject.periph_5v_oc;
  }
  set periph_5v_oc(value) {
    this._periph_5v_ocIntialized = true;
    this._refObject.periph_5v_oc = value;
  }
  get hipower_5v_oc() {
    if (!this._hipower_5v_ocIntialized) {
      return undefined;
    }
    return this._refObject.hipower_5v_oc;
  }
  set hipower_5v_oc(value) {
    this._hipower_5v_ocIntialized = true;
    this._refObject.hipower_5v_oc = value;
  }
  get comp_5v_valid() {
    if (!this._comp_5v_validIntialized) {
      return undefined;
    }
    return this._refObject.comp_5v_valid;
  }
  set comp_5v_valid(value) {
    this._comp_5v_validIntialized = true;
    this._refObject.comp_5v_valid = value;
  }
  get can1_gps1_5v_valid() {
    if (!this._can1_gps1_5v_validIntialized) {
      return undefined;
    }
    return this._refObject.can1_gps1_5v_valid;
  }
  set can1_gps1_5v_valid(value) {
    this._can1_gps1_5v_validIntialized = true;
    this._refObject.can1_gps1_5v_valid = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SystemPowerRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._voltage5v_vIntialized = true;
    this._wrapperFields.sensors3v3.fill(refObject.sensors3v3.toArray());
    this._sensors3v3_validIntialized = true;
    this._usb_connectedIntialized = true;
    this._brick_validIntialized = true;
    this._usb_validIntialized = true;
    this._servo_validIntialized = true;
    this._periph_5v_ocIntialized = true;
    this._hipower_5v_ocIntialized = true;
    this._comp_5v_validIntialized = true;
    this._can1_gps1_5v_validIntialized = true;
  }
  copy(other) {
    this._refObject = new SystemPowerRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._voltage5v_vIntialized = true;
    this._wrapperFields.sensors3v3.copy(other._wrapperFields.sensors3v3);
    this._sensors3v3_validIntialized = true;
    this._usb_connectedIntialized = true;
    this._brick_validIntialized = true;
    this._usb_validIntialized = true;
    this._servo_validIntialized = true;
    this._periph_5v_ocIntialized = true;
    this._hipower_5v_ocIntialized = true;
    this._comp_5v_validIntialized = true;
    this._can1_gps1_5v_validIntialized = true;
  }
  static get classType() {
    return SystemPowerWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"BRICK1_VALID_SHIFTS","value":0},{"type":"uint8","name":"BRICK1_VALID_MASK","value":1},{"type":"uint8","name":"BRICK2_VALID_SHIFTS","value":1},{"type":"uint8","name":"BRICK2_VALID_MASK","value":2},{"type":"uint8","name":"BRICK3_VALID_SHIFTS","value":2},{"type":"uint8","name":"BRICK3_VALID_MASK","value":4},{"type":"uint8","name":"BRICK4_VALID_SHIFTS","value":3},{"type":"uint8","name":"BRICK4_VALID_MASK","value":8}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"voltage5v_v","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sensors3v3","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sensors3v3_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"usb_connected","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"brick_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"usb_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"servo_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"periph_5v_oc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hipower_5v_oc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"comp_5v_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"can1_gps1_5v_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"SystemPower","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SystemPower"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","voltage5v_v","sensors3v3","sensors3v3_valid","usb_connected","brick_valid","usb_valid","servo_valid","periph_5v_oc","hipower_5v_oc","comp_5v_valid","can1_gps1_5v_valid"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SystemPowerArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SystemPowerWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SystemPowerWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SystemPowerArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SystemPowerArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SystemPowerRefArray(size);
    this._refObject = new SystemPowerRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SystemPowerWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SystemPowerArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SystemPowerArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SystemPowerWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SystemPowerWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SystemPowerArrayWrapper;
  }
}
// Define constants (8 in total)
Object.defineProperty(SystemPowerWrapper, "BRICK1_VALID_SHIFTS", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SystemPowerWrapper, "BRICK1_VALID_MASK", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SystemPowerWrapper, "BRICK2_VALID_SHIFTS", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SystemPowerWrapper, "BRICK2_VALID_MASK", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SystemPowerWrapper, "BRICK3_VALID_SHIFTS", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SystemPowerWrapper, "BRICK3_VALID_MASK", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SystemPowerWrapper, "BRICK4_VALID_SHIFTS", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(SystemPowerWrapper, "BRICK4_VALID_MASK", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = SystemPowerWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "BRICK1_VALID_SHIFTS",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "BRICK1_VALID_MASK",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "BRICK2_VALID_SHIFTS",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "BRICK2_VALID_MASK",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "BRICK3_VALID_SHIFTS",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "BRICK3_VALID_MASK",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "BRICK4_VALID_SHIFTS",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "BRICK4_VALID_MASK",
      "value": 8
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "voltage5v_v",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sensors3v3",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sensors3v3_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "usb_connected",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "brick_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "usb_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "servo_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "periph_5v_oc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hipower_5v_oc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "comp_5v_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "can1_gps1_5v_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "SystemPower",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SystemPower"
}
*/
