// rclnodejs/generated/px4_msgs/px4_msgs__msg__PowerButtonState.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PowerButtonStateRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  event: primitiveTypes.uint8,
});
const PowerButtonStateRefArray = ArrayType(PowerButtonStateRefStruct);
const PowerButtonStateRefStructArray = StructType({
  data: PowerButtonStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PowerButtonStateWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PowerButtonStateRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._eventIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PowerButtonStateRefStruct();
    this._timestampIntialized = false;
    this._eventIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PowerButtonStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PowerButtonStateArrayWrapper;
  }
  static get ArrayType() {
    return PowerButtonStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return PowerButtonStateRefStructArray
  }
  static get refObjectType() {
    return PowerButtonStateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in PowerButtonState');
      }
      if (checkConsistency && !this._eventIntialized) {
        throw new TypeError('Invalid argument: event in PowerButtonState');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._eventIntialized = true;
    this._refObject.event = refObject.event;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PowerButtonStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'PowerButtonState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get event() {
    if (!this._eventIntialized) {
      return undefined;
    }
    return this._refObject.event;
  }
  set event(value) {
    this._eventIntialized = true;
    this._refObject.event = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PowerButtonStateRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._eventIntialized = true;
  }
  copy(other) {
    this._refObject = new PowerButtonStateRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._eventIntialized = true;
  }
  static get classType() {
    return PowerButtonStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"PWR_BUTTON_STATE_IDEL","value":0},{"type":"uint8","name":"PWR_BUTTON_STATE_DOWN","value":1},{"type":"uint8","name":"PWR_BUTTON_STATE_UP","value":2},{"type":"uint8","name":"PWR_BUTTON_STATE_REQUEST_SHUTDOWN","value":3}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"event","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"PowerButtonState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PowerButtonState"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","event"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PowerButtonStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PowerButtonStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PowerButtonStateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PowerButtonStateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PowerButtonStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PowerButtonStateRefArray(size);
    this._refObject = new PowerButtonStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PowerButtonStateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PowerButtonStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PowerButtonStateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PowerButtonStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PowerButtonStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PowerButtonStateArrayWrapper;
  }
}
// Define constants (4 in total)
Object.defineProperty(PowerButtonStateWrapper, "PWR_BUTTON_STATE_IDEL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PowerButtonStateWrapper, "PWR_BUTTON_STATE_DOWN", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PowerButtonStateWrapper, "PWR_BUTTON_STATE_UP", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PowerButtonStateWrapper, "PWR_BUTTON_STATE_REQUEST_SHUTDOWN", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = PowerButtonStateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "PWR_BUTTON_STATE_IDEL",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "PWR_BUTTON_STATE_DOWN",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "PWR_BUTTON_STATE_UP",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "PWR_BUTTON_STATE_REQUEST_SHUTDOWN",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "event",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "PowerButtonState",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PowerButtonState"
}
*/
