// rclnodejs/generated/px4_msgs/px4_msgs__msg__AirspeedValidated.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const AirspeedValidatedRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  indicated_airspeed_m_s: primitiveTypes.float32,
  calibrated_airspeed_m_s: primitiveTypes.float32,
  true_airspeed_m_s: primitiveTypes.float32,
  calibrated_ground_minus_wind_m_s: primitiveTypes.float32,
  true_ground_minus_wind_m_s: primitiveTypes.float32,
  airspeed_sensor_measurement_valid: primitiveTypes.bool,
  selected_airspeed_index: primitiveTypes.int8,
});
const AirspeedValidatedRefArray = ArrayType(AirspeedValidatedRefStruct);
const AirspeedValidatedRefStructArray = StructType({
  data: AirspeedValidatedRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AirspeedValidatedWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AirspeedValidatedRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._indicated_airspeed_m_sIntialized = true;
      this._calibrated_airspeed_m_sIntialized = true;
      this._true_airspeed_m_sIntialized = true;
      this._calibrated_ground_minus_wind_m_sIntialized = true;
      this._true_ground_minus_wind_m_sIntialized = true;
      this._airspeed_sensor_measurement_validIntialized = true;
      this._selected_airspeed_indexIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AirspeedValidatedRefStruct();
    this._timestampIntialized = false;
    this._indicated_airspeed_m_sIntialized = false;
    this._calibrated_airspeed_m_sIntialized = false;
    this._true_airspeed_m_sIntialized = false;
    this._calibrated_ground_minus_wind_m_sIntialized = false;
    this._true_ground_minus_wind_m_sIntialized = false;
    this._airspeed_sensor_measurement_validIntialized = false;
    this._selected_airspeed_indexIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new AirspeedValidatedWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AirspeedValidatedArrayWrapper;
  }
  static get ArrayType() {
    return AirspeedValidatedArrayWrapper;
  }
  static get refObjectArrayType() {
    return AirspeedValidatedRefStructArray
  }
  static get refObjectType() {
    return AirspeedValidatedRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in AirspeedValidated');
      }
      if (checkConsistency && !this._indicated_airspeed_m_sIntialized) {
        throw new TypeError('Invalid argument: indicated_airspeed_m_s in AirspeedValidated');
      }
      if (checkConsistency && !this._calibrated_airspeed_m_sIntialized) {
        throw new TypeError('Invalid argument: calibrated_airspeed_m_s in AirspeedValidated');
      }
      if (checkConsistency && !this._true_airspeed_m_sIntialized) {
        throw new TypeError('Invalid argument: true_airspeed_m_s in AirspeedValidated');
      }
      if (checkConsistency && !this._calibrated_ground_minus_wind_m_sIntialized) {
        throw new TypeError('Invalid argument: calibrated_ground_minus_wind_m_s in AirspeedValidated');
      }
      if (checkConsistency && !this._true_ground_minus_wind_m_sIntialized) {
        throw new TypeError('Invalid argument: true_ground_minus_wind_m_s in AirspeedValidated');
      }
      if (checkConsistency && !this._airspeed_sensor_measurement_validIntialized) {
        throw new TypeError('Invalid argument: airspeed_sensor_measurement_valid in AirspeedValidated');
      }
      if (checkConsistency && !this._selected_airspeed_indexIntialized) {
        throw new TypeError('Invalid argument: selected_airspeed_index in AirspeedValidated');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._indicated_airspeed_m_sIntialized = true;
    this._refObject.indicated_airspeed_m_s = refObject.indicated_airspeed_m_s;
    this._calibrated_airspeed_m_sIntialized = true;
    this._refObject.calibrated_airspeed_m_s = refObject.calibrated_airspeed_m_s;
    this._true_airspeed_m_sIntialized = true;
    this._refObject.true_airspeed_m_s = refObject.true_airspeed_m_s;
    this._calibrated_ground_minus_wind_m_sIntialized = true;
    this._refObject.calibrated_ground_minus_wind_m_s = refObject.calibrated_ground_minus_wind_m_s;
    this._true_ground_minus_wind_m_sIntialized = true;
    this._refObject.true_ground_minus_wind_m_s = refObject.true_ground_minus_wind_m_s;
    this._airspeed_sensor_measurement_validIntialized = true;
    this._refObject.airspeed_sensor_measurement_valid = refObject.airspeed_sensor_measurement_valid;
    this._selected_airspeed_indexIntialized = true;
    this._refObject.selected_airspeed_index = refObject.selected_airspeed_index;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    AirspeedValidatedWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'AirspeedValidated'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get indicated_airspeed_m_s() {
    if (!this._indicated_airspeed_m_sIntialized) {
      return undefined;
    }
    return this._refObject.indicated_airspeed_m_s;
  }
  set indicated_airspeed_m_s(value) {
    this._indicated_airspeed_m_sIntialized = true;
    this._refObject.indicated_airspeed_m_s = value;
  }
  get calibrated_airspeed_m_s() {
    if (!this._calibrated_airspeed_m_sIntialized) {
      return undefined;
    }
    return this._refObject.calibrated_airspeed_m_s;
  }
  set calibrated_airspeed_m_s(value) {
    this._calibrated_airspeed_m_sIntialized = true;
    this._refObject.calibrated_airspeed_m_s = value;
  }
  get true_airspeed_m_s() {
    if (!this._true_airspeed_m_sIntialized) {
      return undefined;
    }
    return this._refObject.true_airspeed_m_s;
  }
  set true_airspeed_m_s(value) {
    this._true_airspeed_m_sIntialized = true;
    this._refObject.true_airspeed_m_s = value;
  }
  get calibrated_ground_minus_wind_m_s() {
    if (!this._calibrated_ground_minus_wind_m_sIntialized) {
      return undefined;
    }
    return this._refObject.calibrated_ground_minus_wind_m_s;
  }
  set calibrated_ground_minus_wind_m_s(value) {
    this._calibrated_ground_minus_wind_m_sIntialized = true;
    this._refObject.calibrated_ground_minus_wind_m_s = value;
  }
  get true_ground_minus_wind_m_s() {
    if (!this._true_ground_minus_wind_m_sIntialized) {
      return undefined;
    }
    return this._refObject.true_ground_minus_wind_m_s;
  }
  set true_ground_minus_wind_m_s(value) {
    this._true_ground_minus_wind_m_sIntialized = true;
    this._refObject.true_ground_minus_wind_m_s = value;
  }
  get airspeed_sensor_measurement_valid() {
    if (!this._airspeed_sensor_measurement_validIntialized) {
      return undefined;
    }
    return this._refObject.airspeed_sensor_measurement_valid;
  }
  set airspeed_sensor_measurement_valid(value) {
    this._airspeed_sensor_measurement_validIntialized = true;
    this._refObject.airspeed_sensor_measurement_valid = value;
  }
  get selected_airspeed_index() {
    if (!this._selected_airspeed_indexIntialized) {
      return undefined;
    }
    return this._refObject.selected_airspeed_index;
  }
  set selected_airspeed_index(value) {
    this._selected_airspeed_indexIntialized = true;
    this._refObject.selected_airspeed_index = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AirspeedValidatedRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._indicated_airspeed_m_sIntialized = true;
    this._calibrated_airspeed_m_sIntialized = true;
    this._true_airspeed_m_sIntialized = true;
    this._calibrated_ground_minus_wind_m_sIntialized = true;
    this._true_ground_minus_wind_m_sIntialized = true;
    this._airspeed_sensor_measurement_validIntialized = true;
    this._selected_airspeed_indexIntialized = true;
  }
  copy(other) {
    this._refObject = new AirspeedValidatedRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._indicated_airspeed_m_sIntialized = true;
    this._calibrated_airspeed_m_sIntialized = true;
    this._true_airspeed_m_sIntialized = true;
    this._calibrated_ground_minus_wind_m_sIntialized = true;
    this._true_ground_minus_wind_m_sIntialized = true;
    this._airspeed_sensor_measurement_validIntialized = true;
    this._selected_airspeed_indexIntialized = true;
  }
  static get classType() {
    return AirspeedValidatedWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"indicated_airspeed_m_s","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"calibrated_airspeed_m_s","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"true_airspeed_m_s","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"calibrated_ground_minus_wind_m_s","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"true_ground_minus_wind_m_s","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"airspeed_sensor_measurement_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"selected_airspeed_index","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"AirspeedValidated","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AirspeedValidated"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","indicated_airspeed_m_s","calibrated_airspeed_m_s","true_airspeed_m_s","calibrated_ground_minus_wind_m_s","true_ground_minus_wind_m_s","airspeed_sensor_measurement_valid","selected_airspeed_index"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AirspeedValidatedArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AirspeedValidatedWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AirspeedValidatedWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AirspeedValidatedArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AirspeedValidatedArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AirspeedValidatedRefArray(size);
    this._refObject = new AirspeedValidatedRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AirspeedValidatedWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AirspeedValidatedArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AirspeedValidatedArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AirspeedValidatedWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AirspeedValidatedWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AirspeedValidatedArrayWrapper;
  }
}
module.exports = AirspeedValidatedWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "indicated_airspeed_m_s",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "calibrated_airspeed_m_s",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "true_airspeed_m_s",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "calibrated_ground_minus_wind_m_s",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "true_ground_minus_wind_m_s",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "airspeed_sensor_measurement_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "selected_airspeed_index",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "AirspeedValidated",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "AirspeedValidated"
}
*/
