// rclnodejs/generated/px4_msgs/px4_msgs__msg__Cpuload.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CpuloadRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  load: primitiveTypes.float32,
  ram_usage: primitiveTypes.float32,
});
const CpuloadRefArray = ArrayType(CpuloadRefStruct);
const CpuloadRefStructArray = StructType({
  data: CpuloadRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CpuloadWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CpuloadRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._loadIntialized = true;
      this._ram_usageIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CpuloadRefStruct();
    this._timestampIntialized = false;
    this._loadIntialized = false;
    this._ram_usageIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CpuloadWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CpuloadArrayWrapper;
  }
  static get ArrayType() {
    return CpuloadArrayWrapper;
  }
  static get refObjectArrayType() {
    return CpuloadRefStructArray
  }
  static get refObjectType() {
    return CpuloadRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in Cpuload');
      }
      if (checkConsistency && !this._loadIntialized) {
        throw new TypeError('Invalid argument: load in Cpuload');
      }
      if (checkConsistency && !this._ram_usageIntialized) {
        throw new TypeError('Invalid argument: ram_usage in Cpuload');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._loadIntialized = true;
    this._refObject.load = refObject.load;
    this._ram_usageIntialized = true;
    this._refObject.ram_usage = refObject.ram_usage;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CpuloadWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'Cpuload'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get load() {
    if (!this._loadIntialized) {
      return undefined;
    }
    return this._refObject.load;
  }
  set load(value) {
    this._loadIntialized = true;
    this._refObject.load = value;
  }
  get ram_usage() {
    if (!this._ram_usageIntialized) {
      return undefined;
    }
    return this._refObject.ram_usage;
  }
  set ram_usage(value) {
    this._ram_usageIntialized = true;
    this._refObject.ram_usage = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CpuloadRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._loadIntialized = true;
    this._ram_usageIntialized = true;
  }
  copy(other) {
    this._refObject = new CpuloadRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._loadIntialized = true;
    this._ram_usageIntialized = true;
  }
  static get classType() {
    return CpuloadWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"load","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ram_usage","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"Cpuload","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Cpuload"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","load","ram_usage"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CpuloadArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CpuloadWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CpuloadWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CpuloadArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CpuloadArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CpuloadRefArray(size);
    this._refObject = new CpuloadRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CpuloadWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CpuloadArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CpuloadArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CpuloadWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CpuloadWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CpuloadArrayWrapper;
  }
}
module.exports = CpuloadWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "load",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ram_usage",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "Cpuload",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Cpuload"
}
*/
