// rclnodejs/generated/px4_msgs/px4_msgs__msg__AutotuneAttitudeControlStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const AutotuneAttitudeControlStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  coeff: ArrayType(primitiveTypes.float32, 5),
  coeff_var: ArrayType(primitiveTypes.float32, 5),
  fitness: primitiveTypes.float32,
  innov: primitiveTypes.float32,
  dt_model: primitiveTypes.float32,
  kc: primitiveTypes.float32,
  ki: primitiveTypes.float32,
  kd: primitiveTypes.float32,
  kff: primitiveTypes.float32,
  att_p: primitiveTypes.float32,
  rate_sp: ArrayType(primitiveTypes.float32, 3),
  u_filt: primitiveTypes.float32,
  y_filt: primitiveTypes.float32,
  state: primitiveTypes.uint8,
});
const AutotuneAttitudeControlStatusRefArray = ArrayType(AutotuneAttitudeControlStatusRefStruct);
const AutotuneAttitudeControlStatusRefStructArray = StructType({
  data: AutotuneAttitudeControlStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class AutotuneAttitudeControlStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new AutotuneAttitudeControlStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._wrapperFields.coeff = Float32Wrapper.createArray();
      this._wrapperFields.coeff.copy(other._wrapperFields.coeff);
      this._wrapperFields.coeff_var = Float32Wrapper.createArray();
      this._wrapperFields.coeff_var.copy(other._wrapperFields.coeff_var);
      this._fitnessIntialized = true;
      this._innovIntialized = true;
      this._dt_modelIntialized = true;
      this._kcIntialized = true;
      this._kiIntialized = true;
      this._kdIntialized = true;
      this._kffIntialized = true;
      this._att_pIntialized = true;
      this._wrapperFields.rate_sp = Float32Wrapper.createArray();
      this._wrapperFields.rate_sp.copy(other._wrapperFields.rate_sp);
      this._u_filtIntialized = true;
      this._y_filtIntialized = true;
      this._stateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new AutotuneAttitudeControlStatusRefStruct();
    this._timestampIntialized = false;
    this._wrapperFields.coeff = Float32Wrapper.createArray();
    this._wrapperFields.coeff_var = Float32Wrapper.createArray();
    this._fitnessIntialized = false;
    this._innovIntialized = false;
    this._dt_modelIntialized = false;
    this._kcIntialized = false;
    this._kiIntialized = false;
    this._kdIntialized = false;
    this._kffIntialized = false;
    this._att_pIntialized = false;
    this._wrapperFields.rate_sp = Float32Wrapper.createArray();
    this._u_filtIntialized = false;
    this._y_filtIntialized = false;
    this._stateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new AutotuneAttitudeControlStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new AutotuneAttitudeControlStatusArrayWrapper;
  }
  static get ArrayType() {
    return AutotuneAttitudeControlStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return AutotuneAttitudeControlStatusRefStructArray
  }
  static get refObjectType() {
    return AutotuneAttitudeControlStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._fitnessIntialized) {
        throw new TypeError('Invalid argument: fitness in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._innovIntialized) {
        throw new TypeError('Invalid argument: innov in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._dt_modelIntialized) {
        throw new TypeError('Invalid argument: dt_model in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._kcIntialized) {
        throw new TypeError('Invalid argument: kc in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._kiIntialized) {
        throw new TypeError('Invalid argument: ki in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._kdIntialized) {
        throw new TypeError('Invalid argument: kd in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._kffIntialized) {
        throw new TypeError('Invalid argument: kff in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._att_pIntialized) {
        throw new TypeError('Invalid argument: att_p in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._u_filtIntialized) {
        throw new TypeError('Invalid argument: u_filt in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._y_filtIntialized) {
        throw new TypeError('Invalid argument: y_filt in AutotuneAttitudeControlStatus');
      }
      if (checkConsistency && !this._stateIntialized) {
        throw new TypeError('Invalid argument: state in AutotuneAttitudeControlStatus');
      }
    this._refObject.coeff = Array.from(this._wrapperFields.coeff.data);
    this._refObject.coeff_var = Array.from(this._wrapperFields.coeff_var.data);
    this._refObject.rate_sp = Array.from(this._wrapperFields.rate_sp.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._wrapperFields.coeff.fill(refObject.coeff.toArray());
    this._wrapperFields.coeff_var.fill(refObject.coeff_var.toArray());
    this._fitnessIntialized = true;
    this._refObject.fitness = refObject.fitness;
    this._innovIntialized = true;
    this._refObject.innov = refObject.innov;
    this._dt_modelIntialized = true;
    this._refObject.dt_model = refObject.dt_model;
    this._kcIntialized = true;
    this._refObject.kc = refObject.kc;
    this._kiIntialized = true;
    this._refObject.ki = refObject.ki;
    this._kdIntialized = true;
    this._refObject.kd = refObject.kd;
    this._kffIntialized = true;
    this._refObject.kff = refObject.kff;
    this._att_pIntialized = true;
    this._refObject.att_p = refObject.att_p;
    this._wrapperFields.rate_sp.fill(refObject.rate_sp.toArray());
    this._u_filtIntialized = true;
    this._refObject.u_filt = refObject.u_filt;
    this._y_filtIntialized = true;
    this._refObject.y_filt = refObject.y_filt;
    this._stateIntialized = true;
    this._refObject.state = refObject.state;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    AutotuneAttitudeControlStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'AutotuneAttitudeControlStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get coeff() {
    return this._wrapperFields['coeff'].data;
  }
  set coeff(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields['coeff'].fill(value);
  }
  get coeff_var() {
    return this._wrapperFields['coeff_var'].data;
  }
  set coeff_var(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields['coeff_var'].fill(value);
  }
  get fitness() {
    if (!this._fitnessIntialized) {
      return undefined;
    }
    return this._refObject.fitness;
  }
  set fitness(value) {
    this._fitnessIntialized = true;
    this._refObject.fitness = value;
  }
  get innov() {
    if (!this._innovIntialized) {
      return undefined;
    }
    return this._refObject.innov;
  }
  set innov(value) {
    this._innovIntialized = true;
    this._refObject.innov = value;
  }
  get dt_model() {
    if (!this._dt_modelIntialized) {
      return undefined;
    }
    return this._refObject.dt_model;
  }
  set dt_model(value) {
    this._dt_modelIntialized = true;
    this._refObject.dt_model = value;
  }
  get kc() {
    if (!this._kcIntialized) {
      return undefined;
    }
    return this._refObject.kc;
  }
  set kc(value) {
    this._kcIntialized = true;
    this._refObject.kc = value;
  }
  get ki() {
    if (!this._kiIntialized) {
      return undefined;
    }
    return this._refObject.ki;
  }
  set ki(value) {
    this._kiIntialized = true;
    this._refObject.ki = value;
  }
  get kd() {
    if (!this._kdIntialized) {
      return undefined;
    }
    return this._refObject.kd;
  }
  set kd(value) {
    this._kdIntialized = true;
    this._refObject.kd = value;
  }
  get kff() {
    if (!this._kffIntialized) {
      return undefined;
    }
    return this._refObject.kff;
  }
  set kff(value) {
    this._kffIntialized = true;
    this._refObject.kff = value;
  }
  get att_p() {
    if (!this._att_pIntialized) {
      return undefined;
    }
    return this._refObject.att_p;
  }
  set att_p(value) {
    this._att_pIntialized = true;
    this._refObject.att_p = value;
  }
  get rate_sp() {
    return this._wrapperFields['rate_sp'].data;
  }
  set rate_sp(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['rate_sp'].fill(value);
  }
  get u_filt() {
    if (!this._u_filtIntialized) {
      return undefined;
    }
    return this._refObject.u_filt;
  }
  set u_filt(value) {
    this._u_filtIntialized = true;
    this._refObject.u_filt = value;
  }
  get y_filt() {
    if (!this._y_filtIntialized) {
      return undefined;
    }
    return this._refObject.y_filt;
  }
  set y_filt(value) {
    this._y_filtIntialized = true;
    this._refObject.y_filt = value;
  }
  get state() {
    if (!this._stateIntialized) {
      return undefined;
    }
    return this._refObject.state;
  }
  set state(value) {
    this._stateIntialized = true;
    this._refObject.state = value;
  }
  copyRefObject(refObject) {
    this._refObject = new AutotuneAttitudeControlStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.coeff.fill(refObject.coeff.toArray());
    this._wrapperFields.coeff_var.fill(refObject.coeff_var.toArray());
    this._fitnessIntialized = true;
    this._innovIntialized = true;
    this._dt_modelIntialized = true;
    this._kcIntialized = true;
    this._kiIntialized = true;
    this._kdIntialized = true;
    this._kffIntialized = true;
    this._att_pIntialized = true;
    this._wrapperFields.rate_sp.fill(refObject.rate_sp.toArray());
    this._u_filtIntialized = true;
    this._y_filtIntialized = true;
    this._stateIntialized = true;
  }
  copy(other) {
    this._refObject = new AutotuneAttitudeControlStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.coeff.copy(other._wrapperFields.coeff);
    this._wrapperFields.coeff_var.copy(other._wrapperFields.coeff_var);
    this._fitnessIntialized = true;
    this._innovIntialized = true;
    this._dt_modelIntialized = true;
    this._kcIntialized = true;
    this._kiIntialized = true;
    this._kdIntialized = true;
    this._kffIntialized = true;
    this._att_pIntialized = true;
    this._wrapperFields.rate_sp.copy(other._wrapperFields.rate_sp);
    this._u_filtIntialized = true;
    this._y_filtIntialized = true;
    this._stateIntialized = true;
  }
  static get classType() {
    return AutotuneAttitudeControlStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"STATE_IDLE","value":0},{"type":"uint8","name":"STATE_INIT","value":1},{"type":"uint8","name":"STATE_ROLL","value":2},{"type":"uint8","name":"STATE_ROLL_PAUSE","value":3},{"type":"uint8","name":"STATE_PITCH","value":4},{"type":"uint8","name":"STATE_PITCH_PAUSE","value":5},{"type":"uint8","name":"STATE_YAW","value":6},{"type":"uint8","name":"STATE_YAW_PAUSE","value":7},{"type":"uint8","name":"STATE_VERIFICATION","value":8},{"type":"uint8","name":"STATE_APPLY","value":9},{"type":"uint8","name":"STATE_TEST","value":10},{"type":"uint8","name":"STATE_COMPLETE","value":11},{"type":"uint8","name":"STATE_FAIL","value":12},{"type":"uint8","name":"STATE_WAIT_FOR_DISARM","value":13}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"coeff","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"coeff_var","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fitness","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dt_model","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"kc","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ki","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"kd","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"kff","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"att_p","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rate_sp","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"u_filt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y_filt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"AutotuneAttitudeControlStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"AutotuneAttitudeControlStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","coeff","coeff_var","fitness","innov","dt_model","kc","ki","kd","kff","att_p","rate_sp","u_filt","y_filt","state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class AutotuneAttitudeControlStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof AutotuneAttitudeControlStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new AutotuneAttitudeControlStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AutotuneAttitudeControlStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to AutotuneAttitudeControlStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new AutotuneAttitudeControlStatusRefArray(size);
    this._refObject = new AutotuneAttitudeControlStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new AutotuneAttitudeControlStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof AutotuneAttitudeControlStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "AutotuneAttitudeControlStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      AutotuneAttitudeControlStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return AutotuneAttitudeControlStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return AutotuneAttitudeControlStatusArrayWrapper;
  }
}
// Define constants (14 in total)
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_IDLE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_INIT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_ROLL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_ROLL_PAUSE", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_PITCH", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_PITCH_PAUSE", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_YAW", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_YAW_PAUSE", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_VERIFICATION", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_APPLY", {value: 9, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_TEST", {value: 10, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_COMPLETE", {value: 11, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_FAIL", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(AutotuneAttitudeControlStatusWrapper, "STATE_WAIT_FOR_DISARM", {value: 13, writable: false, enumerable: true, configurable: true});
module.exports = AutotuneAttitudeControlStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "STATE_IDLE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "STATE_INIT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "STATE_ROLL",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "STATE_ROLL_PAUSE",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "STATE_PITCH",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "STATE_PITCH_PAUSE",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "STATE_YAW",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "STATE_YAW_PAUSE",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "STATE_VERIFICATION",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "STATE_APPLY",
      "value": 9
    },
    {
      "type": "uint8",
      "name": "STATE_TEST",
      "value": 10
    },
    {
      "type": "uint8",
      "name": "STATE_COMPLETE",
      "value": 11
    },
    {
      "type": "uint8",
      "name": "STATE_FAIL",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "STATE_WAIT_FOR_DISARM",
      "value": 13
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "coeff",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "coeff_var",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fitness",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dt_model",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "kc",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ki",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "kd",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "kff",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "att_p",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rate_sp",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "u_filt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y_filt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "AutotuneAttitudeControlStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "AutotuneAttitudeControlStatus"
}
*/
