// rclnodejs/generated/px4_msgs/px4_msgs__msg__EstimatorInnovations.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const EstimatorInnovationsRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  gps_hvel: ArrayType(primitiveTypes.float32, 2),
  gps_vvel: primitiveTypes.float32,
  gps_hpos: ArrayType(primitiveTypes.float32, 2),
  gps_vpos: primitiveTypes.float32,
  ev_hvel: ArrayType(primitiveTypes.float32, 2),
  ev_vvel: primitiveTypes.float32,
  ev_hpos: ArrayType(primitiveTypes.float32, 2),
  ev_vpos: primitiveTypes.float32,
  rng_vpos: primitiveTypes.float32,
  baro_vpos: primitiveTypes.float32,
  aux_hvel: ArrayType(primitiveTypes.float32, 2),
  aux_vvel: primitiveTypes.float32,
  flow: ArrayType(primitiveTypes.float32, 2),
  terr_flow: ArrayType(primitiveTypes.float32, 2),
  heading: primitiveTypes.float32,
  mag_field: ArrayType(primitiveTypes.float32, 3),
  gravity: ArrayType(primitiveTypes.float32, 3),
  drag: ArrayType(primitiveTypes.float32, 2),
  airspeed: primitiveTypes.float32,
  beta: primitiveTypes.float32,
  hagl: primitiveTypes.float32,
  hagl_rate: primitiveTypes.float32,
});
const EstimatorInnovationsRefArray = ArrayType(EstimatorInnovationsRefStruct);
const EstimatorInnovationsRefStructArray = StructType({
  data: EstimatorInnovationsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EstimatorInnovationsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EstimatorInnovationsRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._wrapperFields.gps_hvel = Float32Wrapper.createArray();
      this._wrapperFields.gps_hvel.copy(other._wrapperFields.gps_hvel);
      this._gps_vvelIntialized = true;
      this._wrapperFields.gps_hpos = Float32Wrapper.createArray();
      this._wrapperFields.gps_hpos.copy(other._wrapperFields.gps_hpos);
      this._gps_vposIntialized = true;
      this._wrapperFields.ev_hvel = Float32Wrapper.createArray();
      this._wrapperFields.ev_hvel.copy(other._wrapperFields.ev_hvel);
      this._ev_vvelIntialized = true;
      this._wrapperFields.ev_hpos = Float32Wrapper.createArray();
      this._wrapperFields.ev_hpos.copy(other._wrapperFields.ev_hpos);
      this._ev_vposIntialized = true;
      this._rng_vposIntialized = true;
      this._baro_vposIntialized = true;
      this._wrapperFields.aux_hvel = Float32Wrapper.createArray();
      this._wrapperFields.aux_hvel.copy(other._wrapperFields.aux_hvel);
      this._aux_vvelIntialized = true;
      this._wrapperFields.flow = Float32Wrapper.createArray();
      this._wrapperFields.flow.copy(other._wrapperFields.flow);
      this._wrapperFields.terr_flow = Float32Wrapper.createArray();
      this._wrapperFields.terr_flow.copy(other._wrapperFields.terr_flow);
      this._headingIntialized = true;
      this._wrapperFields.mag_field = Float32Wrapper.createArray();
      this._wrapperFields.mag_field.copy(other._wrapperFields.mag_field);
      this._wrapperFields.gravity = Float32Wrapper.createArray();
      this._wrapperFields.gravity.copy(other._wrapperFields.gravity);
      this._wrapperFields.drag = Float32Wrapper.createArray();
      this._wrapperFields.drag.copy(other._wrapperFields.drag);
      this._airspeedIntialized = true;
      this._betaIntialized = true;
      this._haglIntialized = true;
      this._hagl_rateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EstimatorInnovationsRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._wrapperFields.gps_hvel = Float32Wrapper.createArray();
    this._gps_vvelIntialized = false;
    this._wrapperFields.gps_hpos = Float32Wrapper.createArray();
    this._gps_vposIntialized = false;
    this._wrapperFields.ev_hvel = Float32Wrapper.createArray();
    this._ev_vvelIntialized = false;
    this._wrapperFields.ev_hpos = Float32Wrapper.createArray();
    this._ev_vposIntialized = false;
    this._rng_vposIntialized = false;
    this._baro_vposIntialized = false;
    this._wrapperFields.aux_hvel = Float32Wrapper.createArray();
    this._aux_vvelIntialized = false;
    this._wrapperFields.flow = Float32Wrapper.createArray();
    this._wrapperFields.terr_flow = Float32Wrapper.createArray();
    this._headingIntialized = false;
    this._wrapperFields.mag_field = Float32Wrapper.createArray();
    this._wrapperFields.gravity = Float32Wrapper.createArray();
    this._wrapperFields.drag = Float32Wrapper.createArray();
    this._airspeedIntialized = false;
    this._betaIntialized = false;
    this._haglIntialized = false;
    this._hagl_rateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new EstimatorInnovationsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EstimatorInnovationsArrayWrapper;
  }
  static get ArrayType() {
    return EstimatorInnovationsArrayWrapper;
  }
  static get refObjectArrayType() {
    return EstimatorInnovationsRefStructArray
  }
  static get refObjectType() {
    return EstimatorInnovationsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in EstimatorInnovations');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in EstimatorInnovations');
      }
      if (checkConsistency && !this._gps_vvelIntialized) {
        throw new TypeError('Invalid argument: gps_vvel in EstimatorInnovations');
      }
      if (checkConsistency && !this._gps_vposIntialized) {
        throw new TypeError('Invalid argument: gps_vpos in EstimatorInnovations');
      }
      if (checkConsistency && !this._ev_vvelIntialized) {
        throw new TypeError('Invalid argument: ev_vvel in EstimatorInnovations');
      }
      if (checkConsistency && !this._ev_vposIntialized) {
        throw new TypeError('Invalid argument: ev_vpos in EstimatorInnovations');
      }
      if (checkConsistency && !this._rng_vposIntialized) {
        throw new TypeError('Invalid argument: rng_vpos in EstimatorInnovations');
      }
      if (checkConsistency && !this._baro_vposIntialized) {
        throw new TypeError('Invalid argument: baro_vpos in EstimatorInnovations');
      }
      if (checkConsistency && !this._aux_vvelIntialized) {
        throw new TypeError('Invalid argument: aux_vvel in EstimatorInnovations');
      }
      if (checkConsistency && !this._headingIntialized) {
        throw new TypeError('Invalid argument: heading in EstimatorInnovations');
      }
      if (checkConsistency && !this._airspeedIntialized) {
        throw new TypeError('Invalid argument: airspeed in EstimatorInnovations');
      }
      if (checkConsistency && !this._betaIntialized) {
        throw new TypeError('Invalid argument: beta in EstimatorInnovations');
      }
      if (checkConsistency && !this._haglIntialized) {
        throw new TypeError('Invalid argument: hagl in EstimatorInnovations');
      }
      if (checkConsistency && !this._hagl_rateIntialized) {
        throw new TypeError('Invalid argument: hagl_rate in EstimatorInnovations');
      }
    this._refObject.gps_hvel = Array.from(this._wrapperFields.gps_hvel.data);
    this._refObject.gps_hpos = Array.from(this._wrapperFields.gps_hpos.data);
    this._refObject.ev_hvel = Array.from(this._wrapperFields.ev_hvel.data);
    this._refObject.ev_hpos = Array.from(this._wrapperFields.ev_hpos.data);
    this._refObject.aux_hvel = Array.from(this._wrapperFields.aux_hvel.data);
    this._refObject.flow = Array.from(this._wrapperFields.flow.data);
    this._refObject.terr_flow = Array.from(this._wrapperFields.terr_flow.data);
    this._refObject.mag_field = Array.from(this._wrapperFields.mag_field.data);
    this._refObject.gravity = Array.from(this._wrapperFields.gravity.data);
    this._refObject.drag = Array.from(this._wrapperFields.drag.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._wrapperFields.gps_hvel.fill(refObject.gps_hvel.toArray());
    this._gps_vvelIntialized = true;
    this._refObject.gps_vvel = refObject.gps_vvel;
    this._wrapperFields.gps_hpos.fill(refObject.gps_hpos.toArray());
    this._gps_vposIntialized = true;
    this._refObject.gps_vpos = refObject.gps_vpos;
    this._wrapperFields.ev_hvel.fill(refObject.ev_hvel.toArray());
    this._ev_vvelIntialized = true;
    this._refObject.ev_vvel = refObject.ev_vvel;
    this._wrapperFields.ev_hpos.fill(refObject.ev_hpos.toArray());
    this._ev_vposIntialized = true;
    this._refObject.ev_vpos = refObject.ev_vpos;
    this._rng_vposIntialized = true;
    this._refObject.rng_vpos = refObject.rng_vpos;
    this._baro_vposIntialized = true;
    this._refObject.baro_vpos = refObject.baro_vpos;
    this._wrapperFields.aux_hvel.fill(refObject.aux_hvel.toArray());
    this._aux_vvelIntialized = true;
    this._refObject.aux_vvel = refObject.aux_vvel;
    this._wrapperFields.flow.fill(refObject.flow.toArray());
    this._wrapperFields.terr_flow.fill(refObject.terr_flow.toArray());
    this._headingIntialized = true;
    this._refObject.heading = refObject.heading;
    this._wrapperFields.mag_field.fill(refObject.mag_field.toArray());
    this._wrapperFields.gravity.fill(refObject.gravity.toArray());
    this._wrapperFields.drag.fill(refObject.drag.toArray());
    this._airspeedIntialized = true;
    this._refObject.airspeed = refObject.airspeed;
    this._betaIntialized = true;
    this._refObject.beta = refObject.beta;
    this._haglIntialized = true;
    this._refObject.hagl = refObject.hagl;
    this._hagl_rateIntialized = true;
    this._refObject.hagl_rate = refObject.hagl_rate;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    EstimatorInnovationsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'EstimatorInnovations'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get gps_hvel() {
    return this._wrapperFields['gps_hvel'].data;
  }
  set gps_hvel(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['gps_hvel'].fill(value);
  }
  get gps_vvel() {
    if (!this._gps_vvelIntialized) {
      return undefined;
    }
    return this._refObject.gps_vvel;
  }
  set gps_vvel(value) {
    this._gps_vvelIntialized = true;
    this._refObject.gps_vvel = value;
  }
  get gps_hpos() {
    return this._wrapperFields['gps_hpos'].data;
  }
  set gps_hpos(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['gps_hpos'].fill(value);
  }
  get gps_vpos() {
    if (!this._gps_vposIntialized) {
      return undefined;
    }
    return this._refObject.gps_vpos;
  }
  set gps_vpos(value) {
    this._gps_vposIntialized = true;
    this._refObject.gps_vpos = value;
  }
  get ev_hvel() {
    return this._wrapperFields['ev_hvel'].data;
  }
  set ev_hvel(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['ev_hvel'].fill(value);
  }
  get ev_vvel() {
    if (!this._ev_vvelIntialized) {
      return undefined;
    }
    return this._refObject.ev_vvel;
  }
  set ev_vvel(value) {
    this._ev_vvelIntialized = true;
    this._refObject.ev_vvel = value;
  }
  get ev_hpos() {
    return this._wrapperFields['ev_hpos'].data;
  }
  set ev_hpos(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['ev_hpos'].fill(value);
  }
  get ev_vpos() {
    if (!this._ev_vposIntialized) {
      return undefined;
    }
    return this._refObject.ev_vpos;
  }
  set ev_vpos(value) {
    this._ev_vposIntialized = true;
    this._refObject.ev_vpos = value;
  }
  get rng_vpos() {
    if (!this._rng_vposIntialized) {
      return undefined;
    }
    return this._refObject.rng_vpos;
  }
  set rng_vpos(value) {
    this._rng_vposIntialized = true;
    this._refObject.rng_vpos = value;
  }
  get baro_vpos() {
    if (!this._baro_vposIntialized) {
      return undefined;
    }
    return this._refObject.baro_vpos;
  }
  set baro_vpos(value) {
    this._baro_vposIntialized = true;
    this._refObject.baro_vpos = value;
  }
  get aux_hvel() {
    return this._wrapperFields['aux_hvel'].data;
  }
  set aux_hvel(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['aux_hvel'].fill(value);
  }
  get aux_vvel() {
    if (!this._aux_vvelIntialized) {
      return undefined;
    }
    return this._refObject.aux_vvel;
  }
  set aux_vvel(value) {
    this._aux_vvelIntialized = true;
    this._refObject.aux_vvel = value;
  }
  get flow() {
    return this._wrapperFields['flow'].data;
  }
  set flow(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['flow'].fill(value);
  }
  get terr_flow() {
    return this._wrapperFields['terr_flow'].data;
  }
  set terr_flow(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['terr_flow'].fill(value);
  }
  get heading() {
    if (!this._headingIntialized) {
      return undefined;
    }
    return this._refObject.heading;
  }
  set heading(value) {
    this._headingIntialized = true;
    this._refObject.heading = value;
  }
  get mag_field() {
    return this._wrapperFields['mag_field'].data;
  }
  set mag_field(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['mag_field'].fill(value);
  }
  get gravity() {
    return this._wrapperFields['gravity'].data;
  }
  set gravity(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['gravity'].fill(value);
  }
  get drag() {
    return this._wrapperFields['drag'].data;
  }
  set drag(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['drag'].fill(value);
  }
  get airspeed() {
    if (!this._airspeedIntialized) {
      return undefined;
    }
    return this._refObject.airspeed;
  }
  set airspeed(value) {
    this._airspeedIntialized = true;
    this._refObject.airspeed = value;
  }
  get beta() {
    if (!this._betaIntialized) {
      return undefined;
    }
    return this._refObject.beta;
  }
  set beta(value) {
    this._betaIntialized = true;
    this._refObject.beta = value;
  }
  get hagl() {
    if (!this._haglIntialized) {
      return undefined;
    }
    return this._refObject.hagl;
  }
  set hagl(value) {
    this._haglIntialized = true;
    this._refObject.hagl = value;
  }
  get hagl_rate() {
    if (!this._hagl_rateIntialized) {
      return undefined;
    }
    return this._refObject.hagl_rate;
  }
  set hagl_rate(value) {
    this._hagl_rateIntialized = true;
    this._refObject.hagl_rate = value;
  }
  copyRefObject(refObject) {
    this._refObject = new EstimatorInnovationsRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._wrapperFields.gps_hvel.fill(refObject.gps_hvel.toArray());
    this._gps_vvelIntialized = true;
    this._wrapperFields.gps_hpos.fill(refObject.gps_hpos.toArray());
    this._gps_vposIntialized = true;
    this._wrapperFields.ev_hvel.fill(refObject.ev_hvel.toArray());
    this._ev_vvelIntialized = true;
    this._wrapperFields.ev_hpos.fill(refObject.ev_hpos.toArray());
    this._ev_vposIntialized = true;
    this._rng_vposIntialized = true;
    this._baro_vposIntialized = true;
    this._wrapperFields.aux_hvel.fill(refObject.aux_hvel.toArray());
    this._aux_vvelIntialized = true;
    this._wrapperFields.flow.fill(refObject.flow.toArray());
    this._wrapperFields.terr_flow.fill(refObject.terr_flow.toArray());
    this._headingIntialized = true;
    this._wrapperFields.mag_field.fill(refObject.mag_field.toArray());
    this._wrapperFields.gravity.fill(refObject.gravity.toArray());
    this._wrapperFields.drag.fill(refObject.drag.toArray());
    this._airspeedIntialized = true;
    this._betaIntialized = true;
    this._haglIntialized = true;
    this._hagl_rateIntialized = true;
  }
  copy(other) {
    this._refObject = new EstimatorInnovationsRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._wrapperFields.gps_hvel.copy(other._wrapperFields.gps_hvel);
    this._gps_vvelIntialized = true;
    this._wrapperFields.gps_hpos.copy(other._wrapperFields.gps_hpos);
    this._gps_vposIntialized = true;
    this._wrapperFields.ev_hvel.copy(other._wrapperFields.ev_hvel);
    this._ev_vvelIntialized = true;
    this._wrapperFields.ev_hpos.copy(other._wrapperFields.ev_hpos);
    this._ev_vposIntialized = true;
    this._rng_vposIntialized = true;
    this._baro_vposIntialized = true;
    this._wrapperFields.aux_hvel.copy(other._wrapperFields.aux_hvel);
    this._aux_vvelIntialized = true;
    this._wrapperFields.flow.copy(other._wrapperFields.flow);
    this._wrapperFields.terr_flow.copy(other._wrapperFields.terr_flow);
    this._headingIntialized = true;
    this._wrapperFields.mag_field.copy(other._wrapperFields.mag_field);
    this._wrapperFields.gravity.copy(other._wrapperFields.gravity);
    this._wrapperFields.drag.copy(other._wrapperFields.drag);
    this._airspeedIntialized = true;
    this._betaIntialized = true;
    this._haglIntialized = true;
    this._hagl_rateIntialized = true;
  }
  static get classType() {
    return EstimatorInnovationsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_hvel","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_vvel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_hpos","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gps_vpos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ev_hvel","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ev_vvel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ev_hpos","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ev_vpos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rng_vpos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baro_vpos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux_hvel","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"aux_vvel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flow","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"terr_flow","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"heading","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mag_field","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gravity","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"drag","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"airspeed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"beta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hagl","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hagl_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"EstimatorInnovations","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EstimatorInnovations"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","gps_hvel","gps_vvel","gps_hpos","gps_vpos","ev_hvel","ev_vvel","ev_hpos","ev_vpos","rng_vpos","baro_vpos","aux_hvel","aux_vvel","flow","terr_flow","heading","mag_field","gravity","drag","airspeed","beta","hagl","hagl_rate"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EstimatorInnovationsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EstimatorInnovationsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EstimatorInnovationsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorInnovationsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EstimatorInnovationsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EstimatorInnovationsRefArray(size);
    this._refObject = new EstimatorInnovationsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EstimatorInnovationsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EstimatorInnovationsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EstimatorInnovationsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EstimatorInnovationsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EstimatorInnovationsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EstimatorInnovationsArrayWrapper;
  }
}
module.exports = EstimatorInnovationsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_hvel",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_vvel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_hpos",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gps_vpos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ev_hvel",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ev_vvel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ev_hpos",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ev_vpos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rng_vpos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baro_vpos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux_hvel",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "aux_vvel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flow",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "terr_flow",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "heading",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mag_field",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gravity",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "drag",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "airspeed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "beta",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hagl",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hagl_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "EstimatorInnovations",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EstimatorInnovations"
}
*/
