// rclnodejs/generated/px4_msgs/px4_msgs__msg__RtlTimeEstimate.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RtlTimeEstimateRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  valid: primitiveTypes.bool,
  time_estimate: primitiveTypes.float32,
  safe_time_estimate: primitiveTypes.float32,
});
const RtlTimeEstimateRefArray = ArrayType(RtlTimeEstimateRefStruct);
const RtlTimeEstimateRefStructArray = StructType({
  data: RtlTimeEstimateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RtlTimeEstimateWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RtlTimeEstimateRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._validIntialized = true;
      this._time_estimateIntialized = true;
      this._safe_time_estimateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RtlTimeEstimateRefStruct();
    this._timestampIntialized = false;
    this._validIntialized = false;
    this._time_estimateIntialized = false;
    this._safe_time_estimateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new RtlTimeEstimateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RtlTimeEstimateArrayWrapper;
  }
  static get ArrayType() {
    return RtlTimeEstimateArrayWrapper;
  }
  static get refObjectArrayType() {
    return RtlTimeEstimateRefStructArray
  }
  static get refObjectType() {
    return RtlTimeEstimateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in RtlTimeEstimate');
      }
      if (checkConsistency && !this._validIntialized) {
        throw new TypeError('Invalid argument: valid in RtlTimeEstimate');
      }
      if (checkConsistency && !this._time_estimateIntialized) {
        throw new TypeError('Invalid argument: time_estimate in RtlTimeEstimate');
      }
      if (checkConsistency && !this._safe_time_estimateIntialized) {
        throw new TypeError('Invalid argument: safe_time_estimate in RtlTimeEstimate');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._validIntialized = true;
    this._refObject.valid = refObject.valid;
    this._time_estimateIntialized = true;
    this._refObject.time_estimate = refObject.time_estimate;
    this._safe_time_estimateIntialized = true;
    this._refObject.safe_time_estimate = refObject.safe_time_estimate;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    RtlTimeEstimateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'RtlTimeEstimate'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get valid() {
    if (!this._validIntialized) {
      return undefined;
    }
    return this._refObject.valid;
  }
  set valid(value) {
    this._validIntialized = true;
    this._refObject.valid = value;
  }
  get time_estimate() {
    if (!this._time_estimateIntialized) {
      return undefined;
    }
    return this._refObject.time_estimate;
  }
  set time_estimate(value) {
    this._time_estimateIntialized = true;
    this._refObject.time_estimate = value;
  }
  get safe_time_estimate() {
    if (!this._safe_time_estimateIntialized) {
      return undefined;
    }
    return this._refObject.safe_time_estimate;
  }
  set safe_time_estimate(value) {
    this._safe_time_estimateIntialized = true;
    this._refObject.safe_time_estimate = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RtlTimeEstimateRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._validIntialized = true;
    this._time_estimateIntialized = true;
    this._safe_time_estimateIntialized = true;
  }
  copy(other) {
    this._refObject = new RtlTimeEstimateRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._validIntialized = true;
    this._time_estimateIntialized = true;
    this._safe_time_estimateIntialized = true;
  }
  static get classType() {
    return RtlTimeEstimateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_estimate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"safe_time_estimate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"RtlTimeEstimate","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RtlTimeEstimate"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","valid","time_estimate","safe_time_estimate"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RtlTimeEstimateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RtlTimeEstimateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RtlTimeEstimateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RtlTimeEstimateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RtlTimeEstimateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RtlTimeEstimateRefArray(size);
    this._refObject = new RtlTimeEstimateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RtlTimeEstimateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RtlTimeEstimateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RtlTimeEstimateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RtlTimeEstimateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RtlTimeEstimateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RtlTimeEstimateArrayWrapper;
  }
}
module.exports = RtlTimeEstimateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_estimate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "safe_time_estimate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "RtlTimeEstimate",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RtlTimeEstimate"
}
*/
