// rclnodejs/generated/px4_msgs/px4_msgs__msg__CollisionReport.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CollisionReportRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  src: primitiveTypes.uint8,
  id: primitiveTypes.uint32,
  action: primitiveTypes.uint8,
  threat_level: primitiveTypes.uint8,
  time_to_minimum_delta: primitiveTypes.float32,
  altitude_minimum_delta: primitiveTypes.float32,
  horizontal_minimum_delta: primitiveTypes.float32,
});
const CollisionReportRefArray = ArrayType(CollisionReportRefStruct);
const CollisionReportRefStructArray = StructType({
  data: CollisionReportRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CollisionReportWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CollisionReportRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._srcIntialized = true;
      this._idIntialized = true;
      this._actionIntialized = true;
      this._threat_levelIntialized = true;
      this._time_to_minimum_deltaIntialized = true;
      this._altitude_minimum_deltaIntialized = true;
      this._horizontal_minimum_deltaIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CollisionReportRefStruct();
    this._timestampIntialized = false;
    this._srcIntialized = false;
    this._idIntialized = false;
    this._actionIntialized = false;
    this._threat_levelIntialized = false;
    this._time_to_minimum_deltaIntialized = false;
    this._altitude_minimum_deltaIntialized = false;
    this._horizontal_minimum_deltaIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CollisionReportWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CollisionReportArrayWrapper;
  }
  static get ArrayType() {
    return CollisionReportArrayWrapper;
  }
  static get refObjectArrayType() {
    return CollisionReportRefStructArray
  }
  static get refObjectType() {
    return CollisionReportRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in CollisionReport');
      }
      if (checkConsistency && !this._srcIntialized) {
        throw new TypeError('Invalid argument: src in CollisionReport');
      }
      if (checkConsistency && !this._idIntialized) {
        throw new TypeError('Invalid argument: id in CollisionReport');
      }
      if (checkConsistency && !this._actionIntialized) {
        throw new TypeError('Invalid argument: action in CollisionReport');
      }
      if (checkConsistency && !this._threat_levelIntialized) {
        throw new TypeError('Invalid argument: threat_level in CollisionReport');
      }
      if (checkConsistency && !this._time_to_minimum_deltaIntialized) {
        throw new TypeError('Invalid argument: time_to_minimum_delta in CollisionReport');
      }
      if (checkConsistency && !this._altitude_minimum_deltaIntialized) {
        throw new TypeError('Invalid argument: altitude_minimum_delta in CollisionReport');
      }
      if (checkConsistency && !this._horizontal_minimum_deltaIntialized) {
        throw new TypeError('Invalid argument: horizontal_minimum_delta in CollisionReport');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._srcIntialized = true;
    this._refObject.src = refObject.src;
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._actionIntialized = true;
    this._refObject.action = refObject.action;
    this._threat_levelIntialized = true;
    this._refObject.threat_level = refObject.threat_level;
    this._time_to_minimum_deltaIntialized = true;
    this._refObject.time_to_minimum_delta = refObject.time_to_minimum_delta;
    this._altitude_minimum_deltaIntialized = true;
    this._refObject.altitude_minimum_delta = refObject.altitude_minimum_delta;
    this._horizontal_minimum_deltaIntialized = true;
    this._refObject.horizontal_minimum_delta = refObject.horizontal_minimum_delta;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CollisionReportWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'CollisionReport'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get src() {
    if (!this._srcIntialized) {
      return undefined;
    }
    return this._refObject.src;
  }
  set src(value) {
    this._srcIntialized = true;
    this._refObject.src = value;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get action() {
    if (!this._actionIntialized) {
      return undefined;
    }
    return this._refObject.action;
  }
  set action(value) {
    this._actionIntialized = true;
    this._refObject.action = value;
  }
  get threat_level() {
    if (!this._threat_levelIntialized) {
      return undefined;
    }
    return this._refObject.threat_level;
  }
  set threat_level(value) {
    this._threat_levelIntialized = true;
    this._refObject.threat_level = value;
  }
  get time_to_minimum_delta() {
    if (!this._time_to_minimum_deltaIntialized) {
      return undefined;
    }
    return this._refObject.time_to_minimum_delta;
  }
  set time_to_minimum_delta(value) {
    this._time_to_minimum_deltaIntialized = true;
    this._refObject.time_to_minimum_delta = value;
  }
  get altitude_minimum_delta() {
    if (!this._altitude_minimum_deltaIntialized) {
      return undefined;
    }
    return this._refObject.altitude_minimum_delta;
  }
  set altitude_minimum_delta(value) {
    this._altitude_minimum_deltaIntialized = true;
    this._refObject.altitude_minimum_delta = value;
  }
  get horizontal_minimum_delta() {
    if (!this._horizontal_minimum_deltaIntialized) {
      return undefined;
    }
    return this._refObject.horizontal_minimum_delta;
  }
  set horizontal_minimum_delta(value) {
    this._horizontal_minimum_deltaIntialized = true;
    this._refObject.horizontal_minimum_delta = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CollisionReportRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._srcIntialized = true;
    this._idIntialized = true;
    this._actionIntialized = true;
    this._threat_levelIntialized = true;
    this._time_to_minimum_deltaIntialized = true;
    this._altitude_minimum_deltaIntialized = true;
    this._horizontal_minimum_deltaIntialized = true;
  }
  copy(other) {
    this._refObject = new CollisionReportRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._srcIntialized = true;
    this._idIntialized = true;
    this._actionIntialized = true;
    this._threat_levelIntialized = true;
    this._time_to_minimum_deltaIntialized = true;
    this._altitude_minimum_deltaIntialized = true;
    this._horizontal_minimum_deltaIntialized = true;
  }
  static get classType() {
    return CollisionReportWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"src","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"threat_level","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"time_to_minimum_delta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude_minimum_delta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"horizontal_minimum_delta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"CollisionReport","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CollisionReport"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","src","id","action","threat_level","time_to_minimum_delta","altitude_minimum_delta","horizontal_minimum_delta"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CollisionReportArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CollisionReportWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CollisionReportWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CollisionReportArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CollisionReportArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CollisionReportRefArray(size);
    this._refObject = new CollisionReportRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CollisionReportWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CollisionReportArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CollisionReportArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CollisionReportWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CollisionReportWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CollisionReportArrayWrapper;
  }
}
module.exports = CollisionReportWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "src",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "threat_level",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "time_to_minimum_delta",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude_minimum_delta",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "horizontal_minimum_delta",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "CollisionReport",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CollisionReport"
}
*/
