// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleRoi.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const VehicleRoiRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  mode: primitiveTypes.uint8,
  lat: primitiveTypes.float64,
  lon: primitiveTypes.float64,
  alt: primitiveTypes.float32,
  roll_offset: primitiveTypes.float32,
  pitch_offset: primitiveTypes.float32,
  yaw_offset: primitiveTypes.float32,
});
const VehicleRoiRefArray = ArrayType(VehicleRoiRefStruct);
const VehicleRoiRefStructArray = StructType({
  data: VehicleRoiRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleRoiWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleRoiRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._modeIntialized = true;
      this._latIntialized = true;
      this._lonIntialized = true;
      this._altIntialized = true;
      this._roll_offsetIntialized = true;
      this._pitch_offsetIntialized = true;
      this._yaw_offsetIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleRoiRefStruct();
    this._timestampIntialized = false;
    this._modeIntialized = false;
    this._latIntialized = false;
    this._lonIntialized = false;
    this._altIntialized = false;
    this._roll_offsetIntialized = false;
    this._pitch_offsetIntialized = false;
    this._yaw_offsetIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleRoiWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleRoiArrayWrapper;
  }
  static get ArrayType() {
    return VehicleRoiArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleRoiRefStructArray
  }
  static get refObjectType() {
    return VehicleRoiRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleRoi');
      }
      if (checkConsistency && !this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in VehicleRoi');
      }
      if (checkConsistency && !this._latIntialized) {
        throw new TypeError('Invalid argument: lat in VehicleRoi');
      }
      if (checkConsistency && !this._lonIntialized) {
        throw new TypeError('Invalid argument: lon in VehicleRoi');
      }
      if (checkConsistency && !this._altIntialized) {
        throw new TypeError('Invalid argument: alt in VehicleRoi');
      }
      if (checkConsistency && !this._roll_offsetIntialized) {
        throw new TypeError('Invalid argument: roll_offset in VehicleRoi');
      }
      if (checkConsistency && !this._pitch_offsetIntialized) {
        throw new TypeError('Invalid argument: pitch_offset in VehicleRoi');
      }
      if (checkConsistency && !this._yaw_offsetIntialized) {
        throw new TypeError('Invalid argument: yaw_offset in VehicleRoi');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._latIntialized = true;
    this._refObject.lat = refObject.lat;
    this._lonIntialized = true;
    this._refObject.lon = refObject.lon;
    this._altIntialized = true;
    this._refObject.alt = refObject.alt;
    this._roll_offsetIntialized = true;
    this._refObject.roll_offset = refObject.roll_offset;
    this._pitch_offsetIntialized = true;
    this._refObject.pitch_offset = refObject.pitch_offset;
    this._yaw_offsetIntialized = true;
    this._refObject.yaw_offset = refObject.yaw_offset;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleRoiWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleRoi'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get lat() {
    if (!this._latIntialized) {
      return undefined;
    }
    return this._refObject.lat;
  }
  set lat(value) {
    this._latIntialized = true;
    this._refObject.lat = value;
  }
  get lon() {
    if (!this._lonIntialized) {
      return undefined;
    }
    return this._refObject.lon;
  }
  set lon(value) {
    this._lonIntialized = true;
    this._refObject.lon = value;
  }
  get alt() {
    if (!this._altIntialized) {
      return undefined;
    }
    return this._refObject.alt;
  }
  set alt(value) {
    this._altIntialized = true;
    this._refObject.alt = value;
  }
  get roll_offset() {
    if (!this._roll_offsetIntialized) {
      return undefined;
    }
    return this._refObject.roll_offset;
  }
  set roll_offset(value) {
    this._roll_offsetIntialized = true;
    this._refObject.roll_offset = value;
  }
  get pitch_offset() {
    if (!this._pitch_offsetIntialized) {
      return undefined;
    }
    return this._refObject.pitch_offset;
  }
  set pitch_offset(value) {
    this._pitch_offsetIntialized = true;
    this._refObject.pitch_offset = value;
  }
  get yaw_offset() {
    if (!this._yaw_offsetIntialized) {
      return undefined;
    }
    return this._refObject.yaw_offset;
  }
  set yaw_offset(value) {
    this._yaw_offsetIntialized = true;
    this._refObject.yaw_offset = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleRoiRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._modeIntialized = true;
    this._latIntialized = true;
    this._lonIntialized = true;
    this._altIntialized = true;
    this._roll_offsetIntialized = true;
    this._pitch_offsetIntialized = true;
    this._yaw_offsetIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleRoiRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._modeIntialized = true;
    this._latIntialized = true;
    this._lonIntialized = true;
    this._altIntialized = true;
    this._roll_offsetIntialized = true;
    this._pitch_offsetIntialized = true;
    this._yaw_offsetIntialized = true;
  }
  static get classType() {
    return VehicleRoiWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ROI_NONE","value":0},{"type":"uint8","name":"ROI_WPNEXT","value":1},{"type":"uint8","name":"ROI_WPINDEX","value":2},{"type":"uint8","name":"ROI_LOCATION","value":3},{"type":"uint8","name":"ROI_TARGET","value":4},{"type":"uint8","name":"ROI_ENUM_END","value":5}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lat","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lon","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"alt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"roll_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleRoi","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleRoi"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","mode","lat","lon","alt","roll_offset","pitch_offset","yaw_offset"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleRoiArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleRoiWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleRoiWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleRoiArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleRoiArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleRoiRefArray(size);
    this._refObject = new VehicleRoiRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleRoiWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleRoiArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleRoiArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleRoiWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleRoiWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleRoiArrayWrapper;
  }
}
// Define constants (6 in total)
Object.defineProperty(VehicleRoiWrapper, "ROI_NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleRoiWrapper, "ROI_WPNEXT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleRoiWrapper, "ROI_WPINDEX", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleRoiWrapper, "ROI_LOCATION", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleRoiWrapper, "ROI_TARGET", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleRoiWrapper, "ROI_ENUM_END", {value: 5, writable: false, enumerable: true, configurable: true});
module.exports = VehicleRoiWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ROI_NONE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ROI_WPNEXT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "ROI_WPINDEX",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ROI_LOCATION",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "ROI_TARGET",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "ROI_ENUM_END",
      "value": 5
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lat",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lon",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "alt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "roll_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleRoi",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleRoi"
}
*/
