// rclnodejs/generated/px4_msgs/px4_msgs__msg__LaunchDetectionStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LaunchDetectionStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  launch_detection_state: primitiveTypes.uint8,
});
const LaunchDetectionStatusRefArray = ArrayType(LaunchDetectionStatusRefStruct);
const LaunchDetectionStatusRefStructArray = StructType({
  data: LaunchDetectionStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LaunchDetectionStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LaunchDetectionStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._launch_detection_stateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LaunchDetectionStatusRefStruct();
    this._timestampIntialized = false;
    this._launch_detection_stateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LaunchDetectionStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LaunchDetectionStatusArrayWrapper;
  }
  static get ArrayType() {
    return LaunchDetectionStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return LaunchDetectionStatusRefStructArray
  }
  static get refObjectType() {
    return LaunchDetectionStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in LaunchDetectionStatus');
      }
      if (checkConsistency && !this._launch_detection_stateIntialized) {
        throw new TypeError('Invalid argument: launch_detection_state in LaunchDetectionStatus');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._launch_detection_stateIntialized = true;
    this._refObject.launch_detection_state = refObject.launch_detection_state;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LaunchDetectionStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'LaunchDetectionStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get launch_detection_state() {
    if (!this._launch_detection_stateIntialized) {
      return undefined;
    }
    return this._refObject.launch_detection_state;
  }
  set launch_detection_state(value) {
    this._launch_detection_stateIntialized = true;
    this._refObject.launch_detection_state = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LaunchDetectionStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._launch_detection_stateIntialized = true;
  }
  copy(other) {
    this._refObject = new LaunchDetectionStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._launch_detection_stateIntialized = true;
  }
  static get classType() {
    return LaunchDetectionStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"STATE_WAITING_FOR_LAUNCH","value":0},{"type":"uint8","name":"STATE_LAUNCH_DETECTED_DISABLED_MOTOR","value":1},{"type":"uint8","name":"STATE_FLYING","value":2}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"launch_detection_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"LaunchDetectionStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LaunchDetectionStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","launch_detection_state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LaunchDetectionStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LaunchDetectionStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LaunchDetectionStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LaunchDetectionStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LaunchDetectionStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LaunchDetectionStatusRefArray(size);
    this._refObject = new LaunchDetectionStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LaunchDetectionStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LaunchDetectionStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LaunchDetectionStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LaunchDetectionStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LaunchDetectionStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LaunchDetectionStatusArrayWrapper;
  }
}
// Define constants (3 in total)
Object.defineProperty(LaunchDetectionStatusWrapper, "STATE_WAITING_FOR_LAUNCH", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LaunchDetectionStatusWrapper, "STATE_LAUNCH_DETECTED_DISABLED_MOTOR", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LaunchDetectionStatusWrapper, "STATE_FLYING", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = LaunchDetectionStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "STATE_WAITING_FOR_LAUNCH",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "STATE_LAUNCH_DETECTED_DISABLED_MOTOR",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "STATE_FLYING",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "launch_detection_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "LaunchDetectionStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LaunchDetectionStatus"
}
*/
