// rclnodejs/generated/px4_msgs/px4_msgs__msg__TrajectorySetpoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const TrajectorySetpointRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  position: ArrayType(primitiveTypes.float32, 3),
  velocity: ArrayType(primitiveTypes.float32, 3),
  acceleration: ArrayType(primitiveTypes.float32, 3),
  jerk: ArrayType(primitiveTypes.float32, 3),
  yaw: primitiveTypes.float32,
  yawspeed: primitiveTypes.float32,
});
const TrajectorySetpointRefArray = ArrayType(TrajectorySetpointRefStruct);
const TrajectorySetpointRefStructArray = StructType({
  data: TrajectorySetpointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TrajectorySetpointWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TrajectorySetpointRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._wrapperFields.position = Float32Wrapper.createArray();
      this._wrapperFields.position.copy(other._wrapperFields.position);
      this._wrapperFields.velocity = Float32Wrapper.createArray();
      this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
      this._wrapperFields.acceleration = Float32Wrapper.createArray();
      this._wrapperFields.acceleration.copy(other._wrapperFields.acceleration);
      this._wrapperFields.jerk = Float32Wrapper.createArray();
      this._wrapperFields.jerk.copy(other._wrapperFields.jerk);
      this._yawIntialized = true;
      this._yawspeedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TrajectorySetpointRefStruct();
    this._timestampIntialized = false;
    this._wrapperFields.position = Float32Wrapper.createArray();
    this._wrapperFields.velocity = Float32Wrapper.createArray();
    this._wrapperFields.acceleration = Float32Wrapper.createArray();
    this._wrapperFields.jerk = Float32Wrapper.createArray();
    this._yawIntialized = false;
    this._yawspeedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new TrajectorySetpointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TrajectorySetpointArrayWrapper;
  }
  static get ArrayType() {
    return TrajectorySetpointArrayWrapper;
  }
  static get refObjectArrayType() {
    return TrajectorySetpointRefStructArray
  }
  static get refObjectType() {
    return TrajectorySetpointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in TrajectorySetpoint');
      }
      if (checkConsistency && !this._yawIntialized) {
        throw new TypeError('Invalid argument: yaw in TrajectorySetpoint');
      }
      if (checkConsistency && !this._yawspeedIntialized) {
        throw new TypeError('Invalid argument: yawspeed in TrajectorySetpoint');
      }
    this._refObject.position = Array.from(this._wrapperFields.position.data);
    this._refObject.velocity = Array.from(this._wrapperFields.velocity.data);
    this._refObject.acceleration = Array.from(this._wrapperFields.acceleration.data);
    this._refObject.jerk = Array.from(this._wrapperFields.jerk.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._wrapperFields.position.fill(refObject.position.toArray());
    this._wrapperFields.velocity.fill(refObject.velocity.toArray());
    this._wrapperFields.acceleration.fill(refObject.acceleration.toArray());
    this._wrapperFields.jerk.fill(refObject.jerk.toArray());
    this._yawIntialized = true;
    this._refObject.yaw = refObject.yaw;
    this._yawspeedIntialized = true;
    this._refObject.yawspeed = refObject.yawspeed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    TrajectorySetpointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'TrajectorySetpoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get position() {
    return this._wrapperFields['position'].data;
  }
  set position(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['position'].fill(value);
  }
  get velocity() {
    return this._wrapperFields['velocity'].data;
  }
  set velocity(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['velocity'].fill(value);
  }
  get acceleration() {
    return this._wrapperFields['acceleration'].data;
  }
  set acceleration(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['acceleration'].fill(value);
  }
  get jerk() {
    return this._wrapperFields['jerk'].data;
  }
  set jerk(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['jerk'].fill(value);
  }
  get yaw() {
    if (!this._yawIntialized) {
      return undefined;
    }
    return this._refObject.yaw;
  }
  set yaw(value) {
    this._yawIntialized = true;
    this._refObject.yaw = value;
  }
  get yawspeed() {
    if (!this._yawspeedIntialized) {
      return undefined;
    }
    return this._refObject.yawspeed;
  }
  set yawspeed(value) {
    this._yawspeedIntialized = true;
    this._refObject.yawspeed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TrajectorySetpointRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.position.fill(refObject.position.toArray());
    this._wrapperFields.velocity.fill(refObject.velocity.toArray());
    this._wrapperFields.acceleration.fill(refObject.acceleration.toArray());
    this._wrapperFields.jerk.fill(refObject.jerk.toArray());
    this._yawIntialized = true;
    this._yawspeedIntialized = true;
  }
  copy(other) {
    this._refObject = new TrajectorySetpointRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.position.copy(other._wrapperFields.position);
    this._wrapperFields.velocity.copy(other._wrapperFields.velocity);
    this._wrapperFields.acceleration.copy(other._wrapperFields.acceleration);
    this._wrapperFields.jerk.copy(other._wrapperFields.jerk);
    this._yawIntialized = true;
    this._yawspeedIntialized = true;
  }
  static get classType() {
    return TrajectorySetpointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"velocity","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"acceleration","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"jerk","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yawspeed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"TrajectorySetpoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TrajectorySetpoint"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","position","velocity","acceleration","jerk","yaw","yawspeed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TrajectorySetpointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TrajectorySetpointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TrajectorySetpointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectorySetpointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TrajectorySetpointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TrajectorySetpointRefArray(size);
    this._refObject = new TrajectorySetpointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TrajectorySetpointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TrajectorySetpointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TrajectorySetpointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TrajectorySetpointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TrajectorySetpointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TrajectorySetpointArrayWrapper;
  }
}
module.exports = TrajectorySetpointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "velocity",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "acceleration",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "jerk",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yawspeed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "TrajectorySetpoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TrajectorySetpoint"
}
*/
