// rclnodejs/generated/px4_msgs/px4_msgs__msg__GimbalDeviceInformation.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const GimbalDeviceInformationRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  vendor_name: ArrayType(primitiveTypes.uint8, 32),
  model_name: ArrayType(primitiveTypes.uint8, 32),
  custom_name: ArrayType(primitiveTypes.uint8, 32),
  firmware_version: primitiveTypes.uint32,
  hardware_version: primitiveTypes.uint32,
  uid: primitiveTypes.uint64,
  cap_flags: primitiveTypes.uint16,
  custom_cap_flags: primitiveTypes.uint16,
  roll_min: primitiveTypes.float32,
  roll_max: primitiveTypes.float32,
  pitch_min: primitiveTypes.float32,
  pitch_max: primitiveTypes.float32,
  yaw_min: primitiveTypes.float32,
  yaw_max: primitiveTypes.float32,
  gimbal_device_compid: primitiveTypes.uint8,
});
const GimbalDeviceInformationRefArray = ArrayType(GimbalDeviceInformationRefStruct);
const GimbalDeviceInformationRefStructArray = StructType({
  data: GimbalDeviceInformationRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GimbalDeviceInformationWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GimbalDeviceInformationRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._wrapperFields.vendor_name = UInt8Wrapper.createArray();
      this._wrapperFields.vendor_name.copy(other._wrapperFields.vendor_name);
      this._wrapperFields.model_name = UInt8Wrapper.createArray();
      this._wrapperFields.model_name.copy(other._wrapperFields.model_name);
      this._wrapperFields.custom_name = UInt8Wrapper.createArray();
      this._wrapperFields.custom_name.copy(other._wrapperFields.custom_name);
      this._firmware_versionIntialized = true;
      this._hardware_versionIntialized = true;
      this._uidIntialized = true;
      this._cap_flagsIntialized = true;
      this._custom_cap_flagsIntialized = true;
      this._roll_minIntialized = true;
      this._roll_maxIntialized = true;
      this._pitch_minIntialized = true;
      this._pitch_maxIntialized = true;
      this._yaw_minIntialized = true;
      this._yaw_maxIntialized = true;
      this._gimbal_device_compidIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GimbalDeviceInformationRefStruct();
    this._timestampIntialized = false;
    this._wrapperFields.vendor_name = UInt8Wrapper.createArray();
    this._wrapperFields.model_name = UInt8Wrapper.createArray();
    this._wrapperFields.custom_name = UInt8Wrapper.createArray();
    this._firmware_versionIntialized = false;
    this._hardware_versionIntialized = false;
    this._uidIntialized = false;
    this._cap_flagsIntialized = false;
    this._custom_cap_flagsIntialized = false;
    this._roll_minIntialized = false;
    this._roll_maxIntialized = false;
    this._pitch_minIntialized = false;
    this._pitch_maxIntialized = false;
    this._yaw_minIntialized = false;
    this._yaw_maxIntialized = false;
    this._gimbal_device_compidIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GimbalDeviceInformationWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GimbalDeviceInformationArrayWrapper;
  }
  static get ArrayType() {
    return GimbalDeviceInformationArrayWrapper;
  }
  static get refObjectArrayType() {
    return GimbalDeviceInformationRefStructArray
  }
  static get refObjectType() {
    return GimbalDeviceInformationRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._firmware_versionIntialized) {
        throw new TypeError('Invalid argument: firmware_version in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._hardware_versionIntialized) {
        throw new TypeError('Invalid argument: hardware_version in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._uidIntialized) {
        throw new TypeError('Invalid argument: uid in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._cap_flagsIntialized) {
        throw new TypeError('Invalid argument: cap_flags in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._custom_cap_flagsIntialized) {
        throw new TypeError('Invalid argument: custom_cap_flags in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._roll_minIntialized) {
        throw new TypeError('Invalid argument: roll_min in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._roll_maxIntialized) {
        throw new TypeError('Invalid argument: roll_max in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._pitch_minIntialized) {
        throw new TypeError('Invalid argument: pitch_min in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._pitch_maxIntialized) {
        throw new TypeError('Invalid argument: pitch_max in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._yaw_minIntialized) {
        throw new TypeError('Invalid argument: yaw_min in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._yaw_maxIntialized) {
        throw new TypeError('Invalid argument: yaw_max in GimbalDeviceInformation');
      }
      if (checkConsistency && !this._gimbal_device_compidIntialized) {
        throw new TypeError('Invalid argument: gimbal_device_compid in GimbalDeviceInformation');
      }
    this._refObject.vendor_name = Array.from(this._wrapperFields.vendor_name.data);
    this._refObject.model_name = Array.from(this._wrapperFields.model_name.data);
    this._refObject.custom_name = Array.from(this._wrapperFields.custom_name.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._wrapperFields.vendor_name.fill(refObject.vendor_name.toArray());
    this._wrapperFields.model_name.fill(refObject.model_name.toArray());
    this._wrapperFields.custom_name.fill(refObject.custom_name.toArray());
    this._firmware_versionIntialized = true;
    this._refObject.firmware_version = refObject.firmware_version;
    this._hardware_versionIntialized = true;
    this._refObject.hardware_version = refObject.hardware_version;
    this._uidIntialized = true;
    this._refObject.uid = refObject.uid;
    this._cap_flagsIntialized = true;
    this._refObject.cap_flags = refObject.cap_flags;
    this._custom_cap_flagsIntialized = true;
    this._refObject.custom_cap_flags = refObject.custom_cap_flags;
    this._roll_minIntialized = true;
    this._refObject.roll_min = refObject.roll_min;
    this._roll_maxIntialized = true;
    this._refObject.roll_max = refObject.roll_max;
    this._pitch_minIntialized = true;
    this._refObject.pitch_min = refObject.pitch_min;
    this._pitch_maxIntialized = true;
    this._refObject.pitch_max = refObject.pitch_max;
    this._yaw_minIntialized = true;
    this._refObject.yaw_min = refObject.yaw_min;
    this._yaw_maxIntialized = true;
    this._refObject.yaw_max = refObject.yaw_max;
    this._gimbal_device_compidIntialized = true;
    this._refObject.gimbal_device_compid = refObject.gimbal_device_compid;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GimbalDeviceInformationWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'GimbalDeviceInformation'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get vendor_name() {
    return this._wrapperFields['vendor_name'].data;
  }
  set vendor_name(value) {
    if (value.length !== 32) {
      throw new RangeError('The length of the array must be 32.');
    }
    this._wrapperFields['vendor_name'].fill(value);
  }
  get model_name() {
    return this._wrapperFields['model_name'].data;
  }
  set model_name(value) {
    if (value.length !== 32) {
      throw new RangeError('The length of the array must be 32.');
    }
    this._wrapperFields['model_name'].fill(value);
  }
  get custom_name() {
    return this._wrapperFields['custom_name'].data;
  }
  set custom_name(value) {
    if (value.length !== 32) {
      throw new RangeError('The length of the array must be 32.');
    }
    this._wrapperFields['custom_name'].fill(value);
  }
  get firmware_version() {
    if (!this._firmware_versionIntialized) {
      return undefined;
    }
    return this._refObject.firmware_version;
  }
  set firmware_version(value) {
    this._firmware_versionIntialized = true;
    this._refObject.firmware_version = value;
  }
  get hardware_version() {
    if (!this._hardware_versionIntialized) {
      return undefined;
    }
    return this._refObject.hardware_version;
  }
  set hardware_version(value) {
    this._hardware_versionIntialized = true;
    this._refObject.hardware_version = value;
  }
  get uid() {
    if (!this._uidIntialized) {
      return undefined;
    }
    return this._refObject.uid;
  }
  set uid(value) {
    this._uidIntialized = true;
    this._refObject.uid = value;
  }
  get cap_flags() {
    if (!this._cap_flagsIntialized) {
      return undefined;
    }
    return this._refObject.cap_flags;
  }
  set cap_flags(value) {
    this._cap_flagsIntialized = true;
    this._refObject.cap_flags = value;
  }
  get custom_cap_flags() {
    if (!this._custom_cap_flagsIntialized) {
      return undefined;
    }
    return this._refObject.custom_cap_flags;
  }
  set custom_cap_flags(value) {
    this._custom_cap_flagsIntialized = true;
    this._refObject.custom_cap_flags = value;
  }
  get roll_min() {
    if (!this._roll_minIntialized) {
      return undefined;
    }
    return this._refObject.roll_min;
  }
  set roll_min(value) {
    this._roll_minIntialized = true;
    this._refObject.roll_min = value;
  }
  get roll_max() {
    if (!this._roll_maxIntialized) {
      return undefined;
    }
    return this._refObject.roll_max;
  }
  set roll_max(value) {
    this._roll_maxIntialized = true;
    this._refObject.roll_max = value;
  }
  get pitch_min() {
    if (!this._pitch_minIntialized) {
      return undefined;
    }
    return this._refObject.pitch_min;
  }
  set pitch_min(value) {
    this._pitch_minIntialized = true;
    this._refObject.pitch_min = value;
  }
  get pitch_max() {
    if (!this._pitch_maxIntialized) {
      return undefined;
    }
    return this._refObject.pitch_max;
  }
  set pitch_max(value) {
    this._pitch_maxIntialized = true;
    this._refObject.pitch_max = value;
  }
  get yaw_min() {
    if (!this._yaw_minIntialized) {
      return undefined;
    }
    return this._refObject.yaw_min;
  }
  set yaw_min(value) {
    this._yaw_minIntialized = true;
    this._refObject.yaw_min = value;
  }
  get yaw_max() {
    if (!this._yaw_maxIntialized) {
      return undefined;
    }
    return this._refObject.yaw_max;
  }
  set yaw_max(value) {
    this._yaw_maxIntialized = true;
    this._refObject.yaw_max = value;
  }
  get gimbal_device_compid() {
    if (!this._gimbal_device_compidIntialized) {
      return undefined;
    }
    return this._refObject.gimbal_device_compid;
  }
  set gimbal_device_compid(value) {
    this._gimbal_device_compidIntialized = true;
    this._refObject.gimbal_device_compid = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GimbalDeviceInformationRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.vendor_name.fill(refObject.vendor_name.toArray());
    this._wrapperFields.model_name.fill(refObject.model_name.toArray());
    this._wrapperFields.custom_name.fill(refObject.custom_name.toArray());
    this._firmware_versionIntialized = true;
    this._hardware_versionIntialized = true;
    this._uidIntialized = true;
    this._cap_flagsIntialized = true;
    this._custom_cap_flagsIntialized = true;
    this._roll_minIntialized = true;
    this._roll_maxIntialized = true;
    this._pitch_minIntialized = true;
    this._pitch_maxIntialized = true;
    this._yaw_minIntialized = true;
    this._yaw_maxIntialized = true;
    this._gimbal_device_compidIntialized = true;
  }
  copy(other) {
    this._refObject = new GimbalDeviceInformationRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.vendor_name.copy(other._wrapperFields.vendor_name);
    this._wrapperFields.model_name.copy(other._wrapperFields.model_name);
    this._wrapperFields.custom_name.copy(other._wrapperFields.custom_name);
    this._firmware_versionIntialized = true;
    this._hardware_versionIntialized = true;
    this._uidIntialized = true;
    this._cap_flagsIntialized = true;
    this._custom_cap_flagsIntialized = true;
    this._roll_minIntialized = true;
    this._roll_maxIntialized = true;
    this._pitch_minIntialized = true;
    this._pitch_maxIntialized = true;
    this._yaw_minIntialized = true;
    this._yaw_maxIntialized = true;
    this._gimbal_device_compidIntialized = true;
  }
  static get classType() {
    return GimbalDeviceInformationWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_RETRACT","value":1},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_NEUTRAL","value":2},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_ROLL_AXIS","value":4},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_ROLL_FOLLOW","value":8},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_ROLL_LOCK","value":16},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_PITCH_AXIS","value":32},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_PITCH_FOLLOW","value":64},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_PITCH_LOCK","value":128},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_YAW_AXIS","value":256},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_YAW_FOLLOW","value":512},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_HAS_YAW_LOCK","value":1024},{"type":"uint32","name":"GIMBAL_DEVICE_CAP_FLAGS_SUPPORTS_INFINITE_YAW","value":2048}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vendor_name","type":{"isArray":true,"arraySize":32,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"model_name","type":{"isArray":true,"arraySize":32,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"custom_name","type":{"isArray":true,"arraySize":32,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"firmware_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hardware_version","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"uid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cap_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"custom_cap_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"roll_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"roll_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pitch_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_min","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_max","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gimbal_device_compid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"GimbalDeviceInformation","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GimbalDeviceInformation"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","vendor_name","model_name","custom_name","firmware_version","hardware_version","uid","cap_flags","custom_cap_flags","roll_min","roll_max","pitch_min","pitch_max","yaw_min","yaw_max","gimbal_device_compid"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GimbalDeviceInformationArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GimbalDeviceInformationWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GimbalDeviceInformationWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalDeviceInformationArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalDeviceInformationArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GimbalDeviceInformationRefArray(size);
    this._refObject = new GimbalDeviceInformationRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GimbalDeviceInformationWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GimbalDeviceInformationArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GimbalDeviceInformationArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GimbalDeviceInformationWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GimbalDeviceInformationWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GimbalDeviceInformationArrayWrapper;
  }
}
// Define constants (12 in total)
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_RETRACT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_NEUTRAL", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_ROLL_AXIS", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_ROLL_FOLLOW", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_ROLL_LOCK", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_PITCH_AXIS", {value: 32, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_PITCH_FOLLOW", {value: 64, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_PITCH_LOCK", {value: 128, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_YAW_AXIS", {value: 256, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_YAW_FOLLOW", {value: 512, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_HAS_YAW_LOCK", {value: 1024, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GimbalDeviceInformationWrapper, "GIMBAL_DEVICE_CAP_FLAGS_SUPPORTS_INFINITE_YAW", {value: 2048, writable: false, enumerable: true, configurable: true});
module.exports = GimbalDeviceInformationWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_RETRACT",
      "value": 1
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_NEUTRAL",
      "value": 2
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_ROLL_AXIS",
      "value": 4
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_ROLL_FOLLOW",
      "value": 8
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_ROLL_LOCK",
      "value": 16
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_PITCH_AXIS",
      "value": 32
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_PITCH_FOLLOW",
      "value": 64
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_PITCH_LOCK",
      "value": 128
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_YAW_AXIS",
      "value": 256
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_YAW_FOLLOW",
      "value": 512
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_HAS_YAW_LOCK",
      "value": 1024
    },
    {
      "type": "uint32",
      "name": "GIMBAL_DEVICE_CAP_FLAGS_SUPPORTS_INFINITE_YAW",
      "value": 2048
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vendor_name",
      "type": {
        "isArray": true,
        "arraySize": 32,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "model_name",
      "type": {
        "isArray": true,
        "arraySize": 32,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "custom_name",
      "type": {
        "isArray": true,
        "arraySize": 32,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "firmware_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hardware_version",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "uid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cap_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "custom_cap_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "roll_min",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "roll_max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch_min",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pitch_max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_min",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_max",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gimbal_device_compid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "GimbalDeviceInformation",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GimbalDeviceInformation"
}
*/
