// rclnodejs/generated/px4_msgs/px4_msgs__msg__SensorAccelFifo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int16Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int16.js');
const SensorAccelFifoRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  device_id: primitiveTypes.uint32,
  dt: primitiveTypes.float32,
  scale: primitiveTypes.float32,
  samples: primitiveTypes.uint8,
  x: ArrayType(primitiveTypes.int16, 32),
  y: ArrayType(primitiveTypes.int16, 32),
  z: ArrayType(primitiveTypes.int16, 32),
});
const SensorAccelFifoRefArray = ArrayType(SensorAccelFifoRefStruct);
const SensorAccelFifoRefStructArray = StructType({
  data: SensorAccelFifoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SensorAccelFifoWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SensorAccelFifoRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._device_idIntialized = true;
      this._dtIntialized = true;
      this._scaleIntialized = true;
      this._samplesIntialized = true;
      this._wrapperFields.x = Int16Wrapper.createArray();
      this._wrapperFields.x.copy(other._wrapperFields.x);
      this._wrapperFields.y = Int16Wrapper.createArray();
      this._wrapperFields.y.copy(other._wrapperFields.y);
      this._wrapperFields.z = Int16Wrapper.createArray();
      this._wrapperFields.z.copy(other._wrapperFields.z);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SensorAccelFifoRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._device_idIntialized = false;
    this._dtIntialized = false;
    this._scaleIntialized = false;
    this._samplesIntialized = false;
    this._wrapperFields.x = Int16Wrapper.createArray();
    this._wrapperFields.y = Int16Wrapper.createArray();
    this._wrapperFields.z = Int16Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SensorAccelFifoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SensorAccelFifoArrayWrapper;
  }
  static get ArrayType() {
    return SensorAccelFifoArrayWrapper;
  }
  static get refObjectArrayType() {
    return SensorAccelFifoRefStructArray
  }
  static get refObjectType() {
    return SensorAccelFifoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in SensorAccelFifo');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in SensorAccelFifo');
      }
      if (checkConsistency && !this._device_idIntialized) {
        throw new TypeError('Invalid argument: device_id in SensorAccelFifo');
      }
      if (checkConsistency && !this._dtIntialized) {
        throw new TypeError('Invalid argument: dt in SensorAccelFifo');
      }
      if (checkConsistency && !this._scaleIntialized) {
        throw new TypeError('Invalid argument: scale in SensorAccelFifo');
      }
      if (checkConsistency && !this._samplesIntialized) {
        throw new TypeError('Invalid argument: samples in SensorAccelFifo');
      }
    this._refObject.x = Array.from(this._wrapperFields.x.data);
    this._refObject.y = Array.from(this._wrapperFields.y.data);
    this._refObject.z = Array.from(this._wrapperFields.z.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._device_idIntialized = true;
    this._refObject.device_id = refObject.device_id;
    this._dtIntialized = true;
    this._refObject.dt = refObject.dt;
    this._scaleIntialized = true;
    this._refObject.scale = refObject.scale;
    this._samplesIntialized = true;
    this._refObject.samples = refObject.samples;
    this._wrapperFields.x.fill(refObject.x.toArray());
    this._wrapperFields.y.fill(refObject.y.toArray());
    this._wrapperFields.z.fill(refObject.z.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SensorAccelFifoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'SensorAccelFifo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get device_id() {
    if (!this._device_idIntialized) {
      return undefined;
    }
    return this._refObject.device_id;
  }
  set device_id(value) {
    this._device_idIntialized = true;
    this._refObject.device_id = value;
  }
  get dt() {
    if (!this._dtIntialized) {
      return undefined;
    }
    return this._refObject.dt;
  }
  set dt(value) {
    this._dtIntialized = true;
    this._refObject.dt = value;
  }
  get scale() {
    if (!this._scaleIntialized) {
      return undefined;
    }
    return this._refObject.scale;
  }
  set scale(value) {
    this._scaleIntialized = true;
    this._refObject.scale = value;
  }
  get samples() {
    if (!this._samplesIntialized) {
      return undefined;
    }
    return this._refObject.samples;
  }
  set samples(value) {
    this._samplesIntialized = true;
    this._refObject.samples = value;
  }
  get x() {
    return this._wrapperFields['x'].data;
  }
  set x(value) {
    if (value.length !== 32) {
      throw new RangeError('The length of the array must be 32.');
    }
    this._wrapperFields['x'].fill(value);
  }
  get y() {
    return this._wrapperFields['y'].data;
  }
  set y(value) {
    if (value.length !== 32) {
      throw new RangeError('The length of the array must be 32.');
    }
    this._wrapperFields['y'].fill(value);
  }
  get z() {
    return this._wrapperFields['z'].data;
  }
  set z(value) {
    if (value.length !== 32) {
      throw new RangeError('The length of the array must be 32.');
    }
    this._wrapperFields['z'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SensorAccelFifoRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._device_idIntialized = true;
    this._dtIntialized = true;
    this._scaleIntialized = true;
    this._samplesIntialized = true;
    this._wrapperFields.x.fill(refObject.x.toArray());
    this._wrapperFields.y.fill(refObject.y.toArray());
    this._wrapperFields.z.fill(refObject.z.toArray());
  }
  copy(other) {
    this._refObject = new SensorAccelFifoRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._device_idIntialized = true;
    this._dtIntialized = true;
    this._scaleIntialized = true;
    this._samplesIntialized = true;
    this._wrapperFields.x.copy(other._wrapperFields.x);
    this._wrapperFields.y.copy(other._wrapperFields.y);
    this._wrapperFields.z.copy(other._wrapperFields.z);
  }
  static get classType() {
    return SensorAccelFifoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scale","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"samples","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":true,"arraySize":32,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":true,"arraySize":32,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":true,"arraySize":32,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"SensorAccelFifo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SensorAccelFifo"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","device_id","dt","scale","samples","x","y","z"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SensorAccelFifoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SensorAccelFifoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SensorAccelFifoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorAccelFifoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorAccelFifoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SensorAccelFifoRefArray(size);
    this._refObject = new SensorAccelFifoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SensorAccelFifoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SensorAccelFifoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SensorAccelFifoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SensorAccelFifoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SensorAccelFifoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SensorAccelFifoArrayWrapper;
  }
}
module.exports = SensorAccelFifoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "scale",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "samples",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": true,
        "arraySize": 32,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": true,
        "arraySize": 32,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z",
      "type": {
        "isArray": true,
        "arraySize": 32,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "SensorAccelFifo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SensorAccelFifo"
}
*/
