// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleAirData.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const VehicleAirDataRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  baro_device_id: primitiveTypes.uint32,
  baro_alt_meter: primitiveTypes.float32,
  baro_temp_celcius: primitiveTypes.float32,
  baro_pressure_pa: primitiveTypes.float32,
  rho: primitiveTypes.float32,
  calibration_count: primitiveTypes.uint8,
});
const VehicleAirDataRefArray = ArrayType(VehicleAirDataRefStruct);
const VehicleAirDataRefStructArray = StructType({
  data: VehicleAirDataRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleAirDataWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleAirDataRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._baro_device_idIntialized = true;
      this._baro_alt_meterIntialized = true;
      this._baro_temp_celciusIntialized = true;
      this._baro_pressure_paIntialized = true;
      this._rhoIntialized = true;
      this._calibration_countIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleAirDataRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._baro_device_idIntialized = false;
    this._baro_alt_meterIntialized = false;
    this._baro_temp_celciusIntialized = false;
    this._baro_pressure_paIntialized = false;
    this._rhoIntialized = false;
    this._calibration_countIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleAirDataWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleAirDataArrayWrapper;
  }
  static get ArrayType() {
    return VehicleAirDataArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleAirDataRefStructArray
  }
  static get refObjectType() {
    return VehicleAirDataRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleAirData');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in VehicleAirData');
      }
      if (checkConsistency && !this._baro_device_idIntialized) {
        throw new TypeError('Invalid argument: baro_device_id in VehicleAirData');
      }
      if (checkConsistency && !this._baro_alt_meterIntialized) {
        throw new TypeError('Invalid argument: baro_alt_meter in VehicleAirData');
      }
      if (checkConsistency && !this._baro_temp_celciusIntialized) {
        throw new TypeError('Invalid argument: baro_temp_celcius in VehicleAirData');
      }
      if (checkConsistency && !this._baro_pressure_paIntialized) {
        throw new TypeError('Invalid argument: baro_pressure_pa in VehicleAirData');
      }
      if (checkConsistency && !this._rhoIntialized) {
        throw new TypeError('Invalid argument: rho in VehicleAirData');
      }
      if (checkConsistency && !this._calibration_countIntialized) {
        throw new TypeError('Invalid argument: calibration_count in VehicleAirData');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._baro_device_idIntialized = true;
    this._refObject.baro_device_id = refObject.baro_device_id;
    this._baro_alt_meterIntialized = true;
    this._refObject.baro_alt_meter = refObject.baro_alt_meter;
    this._baro_temp_celciusIntialized = true;
    this._refObject.baro_temp_celcius = refObject.baro_temp_celcius;
    this._baro_pressure_paIntialized = true;
    this._refObject.baro_pressure_pa = refObject.baro_pressure_pa;
    this._rhoIntialized = true;
    this._refObject.rho = refObject.rho;
    this._calibration_countIntialized = true;
    this._refObject.calibration_count = refObject.calibration_count;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleAirDataWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleAirData'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get baro_device_id() {
    if (!this._baro_device_idIntialized) {
      return undefined;
    }
    return this._refObject.baro_device_id;
  }
  set baro_device_id(value) {
    this._baro_device_idIntialized = true;
    this._refObject.baro_device_id = value;
  }
  get baro_alt_meter() {
    if (!this._baro_alt_meterIntialized) {
      return undefined;
    }
    return this._refObject.baro_alt_meter;
  }
  set baro_alt_meter(value) {
    this._baro_alt_meterIntialized = true;
    this._refObject.baro_alt_meter = value;
  }
  get baro_temp_celcius() {
    if (!this._baro_temp_celciusIntialized) {
      return undefined;
    }
    return this._refObject.baro_temp_celcius;
  }
  set baro_temp_celcius(value) {
    this._baro_temp_celciusIntialized = true;
    this._refObject.baro_temp_celcius = value;
  }
  get baro_pressure_pa() {
    if (!this._baro_pressure_paIntialized) {
      return undefined;
    }
    return this._refObject.baro_pressure_pa;
  }
  set baro_pressure_pa(value) {
    this._baro_pressure_paIntialized = true;
    this._refObject.baro_pressure_pa = value;
  }
  get rho() {
    if (!this._rhoIntialized) {
      return undefined;
    }
    return this._refObject.rho;
  }
  set rho(value) {
    this._rhoIntialized = true;
    this._refObject.rho = value;
  }
  get calibration_count() {
    if (!this._calibration_countIntialized) {
      return undefined;
    }
    return this._refObject.calibration_count;
  }
  set calibration_count(value) {
    this._calibration_countIntialized = true;
    this._refObject.calibration_count = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleAirDataRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._baro_device_idIntialized = true;
    this._baro_alt_meterIntialized = true;
    this._baro_temp_celciusIntialized = true;
    this._baro_pressure_paIntialized = true;
    this._rhoIntialized = true;
    this._calibration_countIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleAirDataRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._baro_device_idIntialized = true;
    this._baro_alt_meterIntialized = true;
    this._baro_temp_celciusIntialized = true;
    this._baro_pressure_paIntialized = true;
    this._rhoIntialized = true;
    this._calibration_countIntialized = true;
  }
  static get classType() {
    return VehicleAirDataWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baro_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baro_alt_meter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baro_temp_celcius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"baro_pressure_pa","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rho","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"calibration_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleAirData","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleAirData"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","baro_device_id","baro_alt_meter","baro_temp_celcius","baro_pressure_pa","rho","calibration_count"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleAirDataArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleAirDataWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleAirDataWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleAirDataArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleAirDataArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleAirDataRefArray(size);
    this._refObject = new VehicleAirDataRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleAirDataWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleAirDataArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleAirDataArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleAirDataWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleAirDataWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleAirDataArrayWrapper;
  }
}
module.exports = VehicleAirDataWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baro_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baro_alt_meter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baro_temp_celcius",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "baro_pressure_pa",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rho",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "calibration_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleAirData",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleAirData"
}
*/
