// rclnodejs/generated/px4_msgs/px4_msgs__msg__ActuatorTest.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ActuatorTestRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  action: primitiveTypes.uint8,
  function: primitiveTypes.uint16,
  value: primitiveTypes.float32,
  timeout_ms: primitiveTypes.uint32,
});
const ActuatorTestRefArray = ArrayType(ActuatorTestRefStruct);
const ActuatorTestRefStructArray = StructType({
  data: ActuatorTestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ActuatorTestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ActuatorTestRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._actionIntialized = true;
      this._functionIntialized = true;
      this._valueIntialized = true;
      this._timeout_msIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ActuatorTestRefStruct();
    this._timestampIntialized = false;
    this._actionIntialized = false;
    this._functionIntialized = false;
    this._valueIntialized = false;
    this._timeout_msIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ActuatorTestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ActuatorTestArrayWrapper;
  }
  static get ArrayType() {
    return ActuatorTestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ActuatorTestRefStructArray
  }
  static get refObjectType() {
    return ActuatorTestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in ActuatorTest');
      }
      if (checkConsistency && !this._actionIntialized) {
        throw new TypeError('Invalid argument: action in ActuatorTest');
      }
      if (checkConsistency && !this._functionIntialized) {
        throw new TypeError('Invalid argument: function in ActuatorTest');
      }
      if (checkConsistency && !this._valueIntialized) {
        throw new TypeError('Invalid argument: value in ActuatorTest');
      }
      if (checkConsistency && !this._timeout_msIntialized) {
        throw new TypeError('Invalid argument: timeout_ms in ActuatorTest');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._actionIntialized = true;
    this._refObject.action = refObject.action;
    this._functionIntialized = true;
    this._refObject.function = refObject.function;
    this._valueIntialized = true;
    this._refObject.value = refObject.value;
    this._timeout_msIntialized = true;
    this._refObject.timeout_ms = refObject.timeout_ms;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ActuatorTestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'ActuatorTest'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get action() {
    if (!this._actionIntialized) {
      return undefined;
    }
    return this._refObject.action;
  }
  set action(value) {
    this._actionIntialized = true;
    this._refObject.action = value;
  }
  get function() {
    if (!this._functionIntialized) {
      return undefined;
    }
    return this._refObject.function;
  }
  set function(value) {
    this._functionIntialized = true;
    this._refObject.function = value;
  }
  get value() {
    if (!this._valueIntialized) {
      return undefined;
    }
    return this._refObject.value;
  }
  set value(value) {
    this._valueIntialized = true;
    this._refObject.value = value;
  }
  get timeout_ms() {
    if (!this._timeout_msIntialized) {
      return undefined;
    }
    return this._refObject.timeout_ms;
  }
  set timeout_ms(value) {
    this._timeout_msIntialized = true;
    this._refObject.timeout_ms = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ActuatorTestRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._actionIntialized = true;
    this._functionIntialized = true;
    this._valueIntialized = true;
    this._timeout_msIntialized = true;
  }
  copy(other) {
    this._refObject = new ActuatorTestRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._actionIntialized = true;
    this._functionIntialized = true;
    this._valueIntialized = true;
    this._timeout_msIntialized = true;
  }
  static get classType() {
    return ActuatorTestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ACTION_RELEASE_CONTROL","value":0},{"type":"uint8","name":"ACTION_DO_CONTROL","value":1},{"type":"uint8","name":"FUNCTION_MOTOR1","value":101},{"type":"uint8","name":"MAX_NUM_MOTORS","value":12},{"type":"uint8","name":"FUNCTION_SERVO1","value":201},{"type":"uint8","name":"MAX_NUM_SERVOS","value":8},{"type":"uint8","name":"ORB_QUEUE_LENGTH","value":12}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"function","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timeout_ms","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"ActuatorTest","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ActuatorTest"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","action","function","value","timeout_ms"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ActuatorTestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ActuatorTestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ActuatorTestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorTestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActuatorTestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ActuatorTestRefArray(size);
    this._refObject = new ActuatorTestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ActuatorTestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ActuatorTestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ActuatorTestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ActuatorTestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ActuatorTestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ActuatorTestArrayWrapper;
  }
}
// Define constants (7 in total)
Object.defineProperty(ActuatorTestWrapper, "ACTION_RELEASE_CONTROL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorTestWrapper, "ACTION_DO_CONTROL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorTestWrapper, "FUNCTION_MOTOR1", {value: 101, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorTestWrapper, "MAX_NUM_MOTORS", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorTestWrapper, "FUNCTION_SERVO1", {value: 201, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorTestWrapper, "MAX_NUM_SERVOS", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ActuatorTestWrapper, "ORB_QUEUE_LENGTH", {value: 12, writable: false, enumerable: true, configurable: true});
module.exports = ActuatorTestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ACTION_RELEASE_CONTROL",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ACTION_DO_CONTROL",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "FUNCTION_MOTOR1",
      "value": 101
    },
    {
      "type": "uint8",
      "name": "MAX_NUM_MOTORS",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "FUNCTION_SERVO1",
      "value": 201
    },
    {
      "type": "uint8",
      "name": "MAX_NUM_SERVOS",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "ORB_QUEUE_LENGTH",
      "value": 12
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "function",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timeout_ms",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "ActuatorTest",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ActuatorTest"
}
*/
