// rclnodejs/generated/px4_msgs/px4_msgs__msg__LandingTargetInnovations.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LandingTargetInnovationsRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  innov_x: primitiveTypes.float32,
  innov_y: primitiveTypes.float32,
  innov_cov_x: primitiveTypes.float32,
  innov_cov_y: primitiveTypes.float32,
});
const LandingTargetInnovationsRefArray = ArrayType(LandingTargetInnovationsRefStruct);
const LandingTargetInnovationsRefStructArray = StructType({
  data: LandingTargetInnovationsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LandingTargetInnovationsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LandingTargetInnovationsRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._innov_xIntialized = true;
      this._innov_yIntialized = true;
      this._innov_cov_xIntialized = true;
      this._innov_cov_yIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LandingTargetInnovationsRefStruct();
    this._timestampIntialized = false;
    this._innov_xIntialized = false;
    this._innov_yIntialized = false;
    this._innov_cov_xIntialized = false;
    this._innov_cov_yIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LandingTargetInnovationsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LandingTargetInnovationsArrayWrapper;
  }
  static get ArrayType() {
    return LandingTargetInnovationsArrayWrapper;
  }
  static get refObjectArrayType() {
    return LandingTargetInnovationsRefStructArray
  }
  static get refObjectType() {
    return LandingTargetInnovationsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in LandingTargetInnovations');
      }
      if (checkConsistency && !this._innov_xIntialized) {
        throw new TypeError('Invalid argument: innov_x in LandingTargetInnovations');
      }
      if (checkConsistency && !this._innov_yIntialized) {
        throw new TypeError('Invalid argument: innov_y in LandingTargetInnovations');
      }
      if (checkConsistency && !this._innov_cov_xIntialized) {
        throw new TypeError('Invalid argument: innov_cov_x in LandingTargetInnovations');
      }
      if (checkConsistency && !this._innov_cov_yIntialized) {
        throw new TypeError('Invalid argument: innov_cov_y in LandingTargetInnovations');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._innov_xIntialized = true;
    this._refObject.innov_x = refObject.innov_x;
    this._innov_yIntialized = true;
    this._refObject.innov_y = refObject.innov_y;
    this._innov_cov_xIntialized = true;
    this._refObject.innov_cov_x = refObject.innov_cov_x;
    this._innov_cov_yIntialized = true;
    this._refObject.innov_cov_y = refObject.innov_cov_y;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LandingTargetInnovationsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'LandingTargetInnovations'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get innov_x() {
    if (!this._innov_xIntialized) {
      return undefined;
    }
    return this._refObject.innov_x;
  }
  set innov_x(value) {
    this._innov_xIntialized = true;
    this._refObject.innov_x = value;
  }
  get innov_y() {
    if (!this._innov_yIntialized) {
      return undefined;
    }
    return this._refObject.innov_y;
  }
  set innov_y(value) {
    this._innov_yIntialized = true;
    this._refObject.innov_y = value;
  }
  get innov_cov_x() {
    if (!this._innov_cov_xIntialized) {
      return undefined;
    }
    return this._refObject.innov_cov_x;
  }
  set innov_cov_x(value) {
    this._innov_cov_xIntialized = true;
    this._refObject.innov_cov_x = value;
  }
  get innov_cov_y() {
    if (!this._innov_cov_yIntialized) {
      return undefined;
    }
    return this._refObject.innov_cov_y;
  }
  set innov_cov_y(value) {
    this._innov_cov_yIntialized = true;
    this._refObject.innov_cov_y = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LandingTargetInnovationsRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._innov_xIntialized = true;
    this._innov_yIntialized = true;
    this._innov_cov_xIntialized = true;
    this._innov_cov_yIntialized = true;
  }
  copy(other) {
    this._refObject = new LandingTargetInnovationsRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._innov_xIntialized = true;
    this._innov_yIntialized = true;
    this._innov_cov_xIntialized = true;
    this._innov_cov_yIntialized = true;
  }
  static get classType() {
    return LandingTargetInnovationsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov_cov_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"innov_cov_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"LandingTargetInnovations","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LandingTargetInnovations"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","innov_x","innov_y","innov_cov_x","innov_cov_y"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LandingTargetInnovationsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LandingTargetInnovationsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LandingTargetInnovationsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandingTargetInnovationsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandingTargetInnovationsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LandingTargetInnovationsRefArray(size);
    this._refObject = new LandingTargetInnovationsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LandingTargetInnovationsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LandingTargetInnovationsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LandingTargetInnovationsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LandingTargetInnovationsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LandingTargetInnovationsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LandingTargetInnovationsArrayWrapper;
  }
}
module.exports = LandingTargetInnovationsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov_cov_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "innov_cov_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "LandingTargetInnovations",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LandingTargetInnovations"
}
*/
