// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleAttitude.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VehicleAttitudeRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  q: ArrayType(primitiveTypes.float32, 4),
  delta_q_reset: ArrayType(primitiveTypes.float32, 4),
  quat_reset_counter: primitiveTypes.uint8,
});
const VehicleAttitudeRefArray = ArrayType(VehicleAttitudeRefStruct);
const VehicleAttitudeRefStructArray = StructType({
  data: VehicleAttitudeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleAttitudeWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleAttitudeRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._wrapperFields.q = Float32Wrapper.createArray();
      this._wrapperFields.q.copy(other._wrapperFields.q);
      this._wrapperFields.delta_q_reset = Float32Wrapper.createArray();
      this._wrapperFields.delta_q_reset.copy(other._wrapperFields.delta_q_reset);
      this._quat_reset_counterIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleAttitudeRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._wrapperFields.q = Float32Wrapper.createArray();
    this._wrapperFields.delta_q_reset = Float32Wrapper.createArray();
    this._quat_reset_counterIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleAttitudeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleAttitudeArrayWrapper;
  }
  static get ArrayType() {
    return VehicleAttitudeArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleAttitudeRefStructArray
  }
  static get refObjectType() {
    return VehicleAttitudeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleAttitude');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in VehicleAttitude');
      }
      if (checkConsistency && !this._quat_reset_counterIntialized) {
        throw new TypeError('Invalid argument: quat_reset_counter in VehicleAttitude');
      }
    this._refObject.q = Array.from(this._wrapperFields.q.data);
    this._refObject.delta_q_reset = Array.from(this._wrapperFields.delta_q_reset.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._wrapperFields.q.fill(refObject.q.toArray());
    this._wrapperFields.delta_q_reset.fill(refObject.delta_q_reset.toArray());
    this._quat_reset_counterIntialized = true;
    this._refObject.quat_reset_counter = refObject.quat_reset_counter;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleAttitudeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleAttitude'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get q() {
    return this._wrapperFields['q'].data;
  }
  set q(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields['q'].fill(value);
  }
  get delta_q_reset() {
    return this._wrapperFields['delta_q_reset'].data;
  }
  set delta_q_reset(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields['delta_q_reset'].fill(value);
  }
  get quat_reset_counter() {
    if (!this._quat_reset_counterIntialized) {
      return undefined;
    }
    return this._refObject.quat_reset_counter;
  }
  set quat_reset_counter(value) {
    this._quat_reset_counterIntialized = true;
    this._refObject.quat_reset_counter = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleAttitudeRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._wrapperFields.q.fill(refObject.q.toArray());
    this._wrapperFields.delta_q_reset.fill(refObject.delta_q_reset.toArray());
    this._quat_reset_counterIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleAttitudeRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._wrapperFields.q.copy(other._wrapperFields.q);
    this._wrapperFields.delta_q_reset.copy(other._wrapperFields.delta_q_reset);
    this._quat_reset_counterIntialized = true;
  }
  static get classType() {
    return VehicleAttitudeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"q","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_q_reset","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"quat_reset_counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleAttitude","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleAttitude"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","q","delta_q_reset","quat_reset_counter"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleAttitudeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleAttitudeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleAttitudeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleAttitudeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleAttitudeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleAttitudeRefArray(size);
    this._refObject = new VehicleAttitudeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleAttitudeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleAttitudeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleAttitudeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleAttitudeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleAttitudeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleAttitudeArrayWrapper;
  }
}
module.exports = VehicleAttitudeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "q",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_q_reset",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "quat_reset_counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleAttitude",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleAttitude"
}
*/
