// rclnodejs/generated/px4_msgs/px4_msgs__msg__EscStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EscReportWrapper = require('../../generated/px4_msgs/px4_msgs__msg__EscReport.js');
const EscStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  counter: primitiveTypes.uint16,
  esc_count: primitiveTypes.uint8,
  esc_connectiontype: primitiveTypes.uint8,
  esc_online_flags: primitiveTypes.uint8,
  esc_armed_flags: primitiveTypes.uint8,
  esc: ArrayType(EscReportWrapper.refObjectType, 8),
});
const EscStatusRefArray = ArrayType(EscStatusRefStruct);
const EscStatusRefStructArray = StructType({
  data: EscStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class EscStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new EscStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._counterIntialized = true;
      this._esc_countIntialized = true;
      this._esc_connectiontypeIntialized = true;
      this._esc_online_flagsIntialized = true;
      this._esc_armed_flagsIntialized = true;
      this._wrapperFields.esc = EscReportWrapper.createArray();
      this._wrapperFields.esc.copy(other._wrapperFields.esc);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new EscStatusRefStruct();
    this._timestampIntialized = false;
    this._counterIntialized = false;
    this._esc_countIntialized = false;
    this._esc_connectiontypeIntialized = false;
    this._esc_online_flagsIntialized = false;
    this._esc_armed_flagsIntialized = false;
    this._wrapperFields.esc = EscReportWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new EscStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new EscStatusArrayWrapper;
  }
  static get ArrayType() {
    return EscStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return EscStatusRefStructArray
  }
  static get refObjectType() {
    return EscStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in EscStatus');
      }
      if (checkConsistency && !this._counterIntialized) {
        throw new TypeError('Invalid argument: counter in EscStatus');
      }
      if (checkConsistency && !this._esc_countIntialized) {
        throw new TypeError('Invalid argument: esc_count in EscStatus');
      }
      if (checkConsistency && !this._esc_connectiontypeIntialized) {
        throw new TypeError('Invalid argument: esc_connectiontype in EscStatus');
      }
      if (checkConsistency && !this._esc_online_flagsIntialized) {
        throw new TypeError('Invalid argument: esc_online_flags in EscStatus');
      }
      if (checkConsistency && !this._esc_armed_flagsIntialized) {
        throw new TypeError('Invalid argument: esc_armed_flags in EscStatus');
      }
    for (let i = 0; i < 8; i++) {
      if (this._wrapperFields.esc.data[i]) {
        this._wrapperFields.esc.data[i].freeze(own, checkConsistency);
        this._refObject.esc[i] = this._wrapperFields.esc.data[i].refObject;
      }
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._counterIntialized = true;
    this._refObject.counter = refObject.counter;
    this._esc_countIntialized = true;
    this._refObject.esc_count = refObject.esc_count;
    this._esc_connectiontypeIntialized = true;
    this._refObject.esc_connectiontype = refObject.esc_connectiontype;
    this._esc_online_flagsIntialized = true;
    this._refObject.esc_online_flags = refObject.esc_online_flags;
    this._esc_armed_flagsIntialized = true;
    this._refObject.esc_armed_flags = refObject.esc_armed_flags;
    this._refObject.esc = refObject.esc;
    this._wrapperFields.esc.size = 8
    for (let i = 0; i < 8; i++) {
      this._wrapperFields.esc.data[i].copyRefObject(refObject.esc[i]);
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    for (let i = 0; i < 8; i++) {
      EscReportWrapper.freeStruct(refObject.esc[i]);
    }
  }
  static destoryRawROS(msg) {
    EscStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'EscStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get counter() {
    if (!this._counterIntialized) {
      return undefined;
    }
    return this._refObject.counter;
  }
  set counter(value) {
    this._counterIntialized = true;
    this._refObject.counter = value;
  }
  get esc_count() {
    if (!this._esc_countIntialized) {
      return undefined;
    }
    return this._refObject.esc_count;
  }
  set esc_count(value) {
    this._esc_countIntialized = true;
    this._refObject.esc_count = value;
  }
  get esc_connectiontype() {
    if (!this._esc_connectiontypeIntialized) {
      return undefined;
    }
    return this._refObject.esc_connectiontype;
  }
  set esc_connectiontype(value) {
    this._esc_connectiontypeIntialized = true;
    this._refObject.esc_connectiontype = value;
  }
  get esc_online_flags() {
    if (!this._esc_online_flagsIntialized) {
      return undefined;
    }
    return this._refObject.esc_online_flags;
  }
  set esc_online_flags(value) {
    this._esc_online_flagsIntialized = true;
    this._refObject.esc_online_flags = value;
  }
  get esc_armed_flags() {
    if (!this._esc_armed_flagsIntialized) {
      return undefined;
    }
    return this._refObject.esc_armed_flags;
  }
  set esc_armed_flags(value) {
    this._esc_armed_flagsIntialized = true;
    this._refObject.esc_armed_flags = value;
  }
  get esc() {
    return this._wrapperFields.esc;
  }
  set esc(value) {
    if (value.length !== 8) {
      throw new RangeError('The length of the array must be 8.');
    }
    this._wrapperFields.esc.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new EscStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._counterIntialized = true;
    this._esc_countIntialized = true;
    this._esc_connectiontypeIntialized = true;
    this._esc_online_flagsIntialized = true;
    this._esc_armed_flagsIntialized = true;
    this._refObject.esc = refObject.esc;
    this._wrapperFields.esc.size = 8
    for (let i = 0; i < 8; i++) {
      this._wrapperFields.esc.data[i].copyRefObject(refObject.esc[i]);
    }
  }
  copy(other) {
    this._refObject = new EscStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._counterIntialized = true;
    this._esc_countIntialized = true;
    this._esc_connectiontypeIntialized = true;
    this._esc_online_flagsIntialized = true;
    this._esc_armed_flagsIntialized = true;
    this._wrapperFields.esc.copy(other._wrapperFields.esc);
  }
  static get classType() {
    return EscStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"CONNECTED_ESC_MAX","value":8},{"type":"uint8","name":"ESC_CONNECTION_TYPE_PPM","value":0},{"type":"uint8","name":"ESC_CONNECTION_TYPE_SERIAL","value":1},{"type":"uint8","name":"ESC_CONNECTION_TYPE_ONESHOT","value":2},{"type":"uint8","name":"ESC_CONNECTION_TYPE_I2C","value":3},{"type":"uint8","name":"ESC_CONNECTION_TYPE_CAN","value":4},{"type":"uint8","name":"ESC_CONNECTION_TYPE_DSHOT","value":5}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"esc_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"esc_connectiontype","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"esc_online_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"esc_armed_flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"esc","type":{"isArray":true,"arraySize":8,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":"px4_msgs","type":"EscReport","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"EscStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"EscStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","counter","esc_count","esc_connectiontype","esc_online_flags","esc_armed_flags","esc"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class EscStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof EscStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new EscStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EscStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to EscStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new EscStatusRefArray(size);
    this._refObject = new EscStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new EscStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof EscStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "EscStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      EscStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return EscStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return EscStatusArrayWrapper;
  }
}
// Define constants (7 in total)
Object.defineProperty(EscStatusWrapper, "CONNECTED_ESC_MAX", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EscStatusWrapper, "ESC_CONNECTION_TYPE_PPM", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EscStatusWrapper, "ESC_CONNECTION_TYPE_SERIAL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EscStatusWrapper, "ESC_CONNECTION_TYPE_ONESHOT", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EscStatusWrapper, "ESC_CONNECTION_TYPE_I2C", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EscStatusWrapper, "ESC_CONNECTION_TYPE_CAN", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(EscStatusWrapper, "ESC_CONNECTION_TYPE_DSHOT", {value: 5, writable: false, enumerable: true, configurable: true});
module.exports = EscStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "CONNECTED_ESC_MAX",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "ESC_CONNECTION_TYPE_PPM",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ESC_CONNECTION_TYPE_SERIAL",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "ESC_CONNECTION_TYPE_ONESHOT",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ESC_CONNECTION_TYPE_I2C",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "ESC_CONNECTION_TYPE_CAN",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "ESC_CONNECTION_TYPE_DSHOT",
      "value": 5
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "esc_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "esc_connectiontype",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "esc_online_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "esc_armed_flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "esc",
      "type": {
        "isArray": true,
        "arraySize": 8,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": "px4_msgs",
        "type": "EscReport",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "EscStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "EscStatus"
}
*/
