// rclnodejs/generated/px4_msgs/px4_msgs__msg__PositionSetpointTriplet.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PositionSetpointWrapper = require('../../generated/px4_msgs/px4_msgs__msg__PositionSetpoint.js');
const PositionSetpointTripletRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  previous: PositionSetpointWrapper.refObjectType,
  current: PositionSetpointWrapper.refObjectType,
  next: PositionSetpointWrapper.refObjectType,
});
const PositionSetpointTripletRefArray = ArrayType(PositionSetpointTripletRefStruct);
const PositionSetpointTripletRefStructArray = StructType({
  data: PositionSetpointTripletRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PositionSetpointTripletWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PositionSetpointTripletRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._wrapperFields.previous =  new PositionSetpointWrapper(other._wrapperFields.previous);
      this._wrapperFields.current =  new PositionSetpointWrapper(other._wrapperFields.current);
      this._wrapperFields.next =  new PositionSetpointWrapper(other._wrapperFields.next);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PositionSetpointTripletRefStruct();
    this._timestampIntialized = false;
    this._wrapperFields.previous = new PositionSetpointWrapper();
    this._wrapperFields.current = new PositionSetpointWrapper();
    this._wrapperFields.next = new PositionSetpointWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new PositionSetpointTripletWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PositionSetpointTripletArrayWrapper;
  }
  static get ArrayType() {
    return PositionSetpointTripletArrayWrapper;
  }
  static get refObjectArrayType() {
    return PositionSetpointTripletRefStructArray
  }
  static get refObjectType() {
    return PositionSetpointTripletRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in PositionSetpointTriplet');
      }
    this._wrapperFields.previous.freeze(own, checkConsistency);
    this._refObject.previous = this._wrapperFields.previous.refObject;
    this._wrapperFields.current.freeze(own, checkConsistency);
    this._refObject.current = this._wrapperFields.current.refObject;
    this._wrapperFields.next.freeze(own, checkConsistency);
    this._refObject.next = this._wrapperFields.next.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._wrapperFields.previous.copyRefObject(refObject.previous);
    this._wrapperFields.current.copyRefObject(refObject.current);
    this._wrapperFields.next.copyRefObject(refObject.next);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PositionSetpointWrapper.freeStruct(refObject.previous);
    PositionSetpointWrapper.freeStruct(refObject.current);
    PositionSetpointWrapper.freeStruct(refObject.next);
  }
  static destoryRawROS(msg) {
    PositionSetpointTripletWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'PositionSetpointTriplet'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get previous() {
    return this._wrapperFields.previous;
  }
  set previous(value) {
    if (value instanceof PositionSetpointWrapper) {
      this._wrapperFields.previous.copy(value);
    } else {
      this._wrapperFields.previous.copy(new PositionSetpointWrapper(value));
    }
  }
  get current() {
    return this._wrapperFields.current;
  }
  set current(value) {
    if (value instanceof PositionSetpointWrapper) {
      this._wrapperFields.current.copy(value);
    } else {
      this._wrapperFields.current.copy(new PositionSetpointWrapper(value));
    }
  }
  get next() {
    return this._wrapperFields.next;
  }
  set next(value) {
    if (value instanceof PositionSetpointWrapper) {
      this._wrapperFields.next.copy(value);
    } else {
      this._wrapperFields.next.copy(new PositionSetpointWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new PositionSetpointTripletRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.previous.copyRefObject(this._refObject.previous);
    this._wrapperFields.current.copyRefObject(this._refObject.current);
    this._wrapperFields.next.copyRefObject(this._refObject.next);
  }
  copy(other) {
    this._refObject = new PositionSetpointTripletRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.previous.copy(other._wrapperFields.previous);
    this._wrapperFields.current.copy(other._wrapperFields.current);
    this._wrapperFields.next.copy(other._wrapperFields.next);
  }
  static get classType() {
    return PositionSetpointTripletWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"previous","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"px4_msgs","type":"PositionSetpoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"current","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"px4_msgs","type":"PositionSetpoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"next","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"px4_msgs","type":"PositionSetpoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"PositionSetpointTriplet","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PositionSetpointTriplet"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","previous","current","next"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PositionSetpointTripletArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PositionSetpointTripletWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PositionSetpointTripletWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PositionSetpointTripletArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PositionSetpointTripletArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PositionSetpointTripletRefArray(size);
    this._refObject = new PositionSetpointTripletRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PositionSetpointTripletWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PositionSetpointTripletArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PositionSetpointTripletArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PositionSetpointTripletWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PositionSetpointTripletWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PositionSetpointTripletArrayWrapper;
  }
}
module.exports = PositionSetpointTripletWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "previous",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "px4_msgs",
        "type": "PositionSetpoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "current",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "px4_msgs",
        "type": "PositionSetpoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "next",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "px4_msgs",
        "type": "PositionSetpoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "PositionSetpointTriplet",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PositionSetpointTriplet"
}
*/
