// rclnodejs/generated/px4_msgs/px4_msgs__msg__PpsCapture.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PpsCaptureRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  rtc_timestamp: primitiveTypes.uint64,
  pps_rate_exceeded_counter: primitiveTypes.uint8,
});
const PpsCaptureRefArray = ArrayType(PpsCaptureRefStruct);
const PpsCaptureRefStructArray = StructType({
  data: PpsCaptureRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PpsCaptureWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PpsCaptureRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._rtc_timestampIntialized = true;
      this._pps_rate_exceeded_counterIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PpsCaptureRefStruct();
    this._timestampIntialized = false;
    this._rtc_timestampIntialized = false;
    this._pps_rate_exceeded_counterIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PpsCaptureWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PpsCaptureArrayWrapper;
  }
  static get ArrayType() {
    return PpsCaptureArrayWrapper;
  }
  static get refObjectArrayType() {
    return PpsCaptureRefStructArray
  }
  static get refObjectType() {
    return PpsCaptureRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in PpsCapture');
      }
      if (checkConsistency && !this._rtc_timestampIntialized) {
        throw new TypeError('Invalid argument: rtc_timestamp in PpsCapture');
      }
      if (checkConsistency && !this._pps_rate_exceeded_counterIntialized) {
        throw new TypeError('Invalid argument: pps_rate_exceeded_counter in PpsCapture');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._rtc_timestampIntialized = true;
    this._refObject.rtc_timestamp = refObject.rtc_timestamp;
    this._pps_rate_exceeded_counterIntialized = true;
    this._refObject.pps_rate_exceeded_counter = refObject.pps_rate_exceeded_counter;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PpsCaptureWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'PpsCapture'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get rtc_timestamp() {
    if (!this._rtc_timestampIntialized) {
      return undefined;
    }
    return this._refObject.rtc_timestamp;
  }
  set rtc_timestamp(value) {
    this._rtc_timestampIntialized = true;
    this._refObject.rtc_timestamp = value;
  }
  get pps_rate_exceeded_counter() {
    if (!this._pps_rate_exceeded_counterIntialized) {
      return undefined;
    }
    return this._refObject.pps_rate_exceeded_counter;
  }
  set pps_rate_exceeded_counter(value) {
    this._pps_rate_exceeded_counterIntialized = true;
    this._refObject.pps_rate_exceeded_counter = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PpsCaptureRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._rtc_timestampIntialized = true;
    this._pps_rate_exceeded_counterIntialized = true;
  }
  copy(other) {
    this._refObject = new PpsCaptureRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._rtc_timestampIntialized = true;
    this._pps_rate_exceeded_counterIntialized = true;
  }
  static get classType() {
    return PpsCaptureWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rtc_timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pps_rate_exceeded_counter","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"PpsCapture","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PpsCapture"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","rtc_timestamp","pps_rate_exceeded_counter"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PpsCaptureArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PpsCaptureWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PpsCaptureWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PpsCaptureArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PpsCaptureArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PpsCaptureRefArray(size);
    this._refObject = new PpsCaptureRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PpsCaptureWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PpsCaptureArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PpsCaptureArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PpsCaptureWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PpsCaptureWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PpsCaptureArrayWrapper;
  }
}
module.exports = PpsCaptureWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rtc_timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pps_rate_exceeded_counter",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "PpsCapture",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PpsCapture"
}
*/
