// rclnodejs/generated/px4_msgs/px4_msgs__msg__QshellReq.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CharWrapper = require('../../generated/std_msgs/std_msgs__msg__Char.js');
const QshellReqRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  cmd: ArrayType(primitiveTypes.char, 100),
  strlen: primitiveTypes.uint32,
  request_sequence: primitiveTypes.uint32,
});
const QshellReqRefArray = ArrayType(QshellReqRefStruct);
const QshellReqRefStructArray = StructType({
  data: QshellReqRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class QshellReqWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new QshellReqRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._wrapperFields.cmd = CharWrapper.createArray();
      this._wrapperFields.cmd.copy(other._wrapperFields.cmd);
      this._strlenIntialized = true;
      this._request_sequenceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new QshellReqRefStruct();
    this._timestampIntialized = false;
    this._wrapperFields.cmd = CharWrapper.createArray();
    this._strlenIntialized = false;
    this._request_sequenceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new QshellReqWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new QshellReqArrayWrapper;
  }
  static get ArrayType() {
    return QshellReqArrayWrapper;
  }
  static get refObjectArrayType() {
    return QshellReqRefStructArray
  }
  static get refObjectType() {
    return QshellReqRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in QshellReq');
      }
      if (checkConsistency && !this._strlenIntialized) {
        throw new TypeError('Invalid argument: strlen in QshellReq');
      }
      if (checkConsistency && !this._request_sequenceIntialized) {
        throw new TypeError('Invalid argument: request_sequence in QshellReq');
      }
    this._refObject.cmd = Array.from(this._wrapperFields.cmd.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._wrapperFields.cmd.fill(refObject.cmd.toArray());
    this._strlenIntialized = true;
    this._refObject.strlen = refObject.strlen;
    this._request_sequenceIntialized = true;
    this._refObject.request_sequence = refObject.request_sequence;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    QshellReqWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'QshellReq'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get cmd() {
    return this._wrapperFields['cmd'].data;
  }
  set cmd(value) {
    if (value.length !== 100) {
      throw new RangeError('The length of the array must be 100.');
    }
    this._wrapperFields['cmd'].fill(value);
  }
  get strlen() {
    if (!this._strlenIntialized) {
      return undefined;
    }
    return this._refObject.strlen;
  }
  set strlen(value) {
    this._strlenIntialized = true;
    this._refObject.strlen = value;
  }
  get request_sequence() {
    if (!this._request_sequenceIntialized) {
      return undefined;
    }
    return this._refObject.request_sequence;
  }
  set request_sequence(value) {
    this._request_sequenceIntialized = true;
    this._refObject.request_sequence = value;
  }
  copyRefObject(refObject) {
    this._refObject = new QshellReqRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.cmd.fill(refObject.cmd.toArray());
    this._strlenIntialized = true;
    this._request_sequenceIntialized = true;
  }
  copy(other) {
    this._refObject = new QshellReqRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.cmd.copy(other._wrapperFields.cmd);
    this._strlenIntialized = true;
    this._request_sequenceIntialized = true;
  }
  static get classType() {
    return QshellReqWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint32","name":"MAX_STRLEN","value":100}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cmd","type":{"isArray":true,"arraySize":100,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"char","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"strlen","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"request_sequence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"QshellReq","stringUpperBound":null,"isPrimitiveType":false},"msgName":"QshellReq"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","cmd","strlen","request_sequence"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class QshellReqArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof QshellReqWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new QshellReqWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QshellReqArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to QshellReqArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new QshellReqRefArray(size);
    this._refObject = new QshellReqRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new QshellReqWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof QshellReqArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "QshellReqArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      QshellReqWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return QshellReqWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return QshellReqArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(QshellReqWrapper, "MAX_STRLEN", {value: 100, writable: false, enumerable: true, configurable: true});
module.exports = QshellReqWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint32",
      "name": "MAX_STRLEN",
      "value": 100
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cmd",
      "type": {
        "isArray": true,
        "arraySize": 100,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "char",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "strlen",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "request_sequence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "QshellReq",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "QshellReq"
}
*/
