// rclnodejs/generated/px4_msgs/px4_msgs__msg__ObstacleDistance.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt16Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt16.js');
const ObstacleDistanceRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  frame: primitiveTypes.uint8,
  sensor_type: primitiveTypes.uint8,
  distances: ArrayType(primitiveTypes.uint16, 72),
  increment: primitiveTypes.float32,
  min_distance: primitiveTypes.uint16,
  max_distance: primitiveTypes.uint16,
  angle_offset: primitiveTypes.float32,
});
const ObstacleDistanceRefArray = ArrayType(ObstacleDistanceRefStruct);
const ObstacleDistanceRefStructArray = StructType({
  data: ObstacleDistanceRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ObstacleDistanceWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObstacleDistanceRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._frameIntialized = true;
      this._sensor_typeIntialized = true;
      this._wrapperFields.distances = UInt16Wrapper.createArray();
      this._wrapperFields.distances.copy(other._wrapperFields.distances);
      this._incrementIntialized = true;
      this._min_distanceIntialized = true;
      this._max_distanceIntialized = true;
      this._angle_offsetIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ObstacleDistanceRefStruct();
    this._timestampIntialized = false;
    this._frameIntialized = false;
    this._sensor_typeIntialized = false;
    this._wrapperFields.distances = UInt16Wrapper.createArray();
    this._incrementIntialized = false;
    this._min_distanceIntialized = false;
    this._max_distanceIntialized = false;
    this._angle_offsetIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ObstacleDistanceWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObstacleDistanceArrayWrapper;
  }
  static get ArrayType() {
    return ObstacleDistanceArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObstacleDistanceRefStructArray
  }
  static get refObjectType() {
    return ObstacleDistanceRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in ObstacleDistance');
      }
      if (checkConsistency && !this._frameIntialized) {
        throw new TypeError('Invalid argument: frame in ObstacleDistance');
      }
      if (checkConsistency && !this._sensor_typeIntialized) {
        throw new TypeError('Invalid argument: sensor_type in ObstacleDistance');
      }
      if (checkConsistency && !this._incrementIntialized) {
        throw new TypeError('Invalid argument: increment in ObstacleDistance');
      }
      if (checkConsistency && !this._min_distanceIntialized) {
        throw new TypeError('Invalid argument: min_distance in ObstacleDistance');
      }
      if (checkConsistency && !this._max_distanceIntialized) {
        throw new TypeError('Invalid argument: max_distance in ObstacleDistance');
      }
      if (checkConsistency && !this._angle_offsetIntialized) {
        throw new TypeError('Invalid argument: angle_offset in ObstacleDistance');
      }
    this._refObject.distances = Array.from(this._wrapperFields.distances.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._frameIntialized = true;
    this._refObject.frame = refObject.frame;
    this._sensor_typeIntialized = true;
    this._refObject.sensor_type = refObject.sensor_type;
    this._wrapperFields.distances.fill(refObject.distances.toArray());
    this._incrementIntialized = true;
    this._refObject.increment = refObject.increment;
    this._min_distanceIntialized = true;
    this._refObject.min_distance = refObject.min_distance;
    this._max_distanceIntialized = true;
    this._refObject.max_distance = refObject.max_distance;
    this._angle_offsetIntialized = true;
    this._refObject.angle_offset = refObject.angle_offset;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ObstacleDistanceWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'ObstacleDistance'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get frame() {
    if (!this._frameIntialized) {
      return undefined;
    }
    return this._refObject.frame;
  }
  set frame(value) {
    this._frameIntialized = true;
    this._refObject.frame = value;
  }
  get sensor_type() {
    if (!this._sensor_typeIntialized) {
      return undefined;
    }
    return this._refObject.sensor_type;
  }
  set sensor_type(value) {
    this._sensor_typeIntialized = true;
    this._refObject.sensor_type = value;
  }
  get distances() {
    return this._wrapperFields['distances'].data;
  }
  set distances(value) {
    if (value.length !== 72) {
      throw new RangeError('The length of the array must be 72.');
    }
    this._wrapperFields['distances'].fill(value);
  }
  get increment() {
    if (!this._incrementIntialized) {
      return undefined;
    }
    return this._refObject.increment;
  }
  set increment(value) {
    this._incrementIntialized = true;
    this._refObject.increment = value;
  }
  get min_distance() {
    if (!this._min_distanceIntialized) {
      return undefined;
    }
    return this._refObject.min_distance;
  }
  set min_distance(value) {
    this._min_distanceIntialized = true;
    this._refObject.min_distance = value;
  }
  get max_distance() {
    if (!this._max_distanceIntialized) {
      return undefined;
    }
    return this._refObject.max_distance;
  }
  set max_distance(value) {
    this._max_distanceIntialized = true;
    this._refObject.max_distance = value;
  }
  get angle_offset() {
    if (!this._angle_offsetIntialized) {
      return undefined;
    }
    return this._refObject.angle_offset;
  }
  set angle_offset(value) {
    this._angle_offsetIntialized = true;
    this._refObject.angle_offset = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ObstacleDistanceRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._frameIntialized = true;
    this._sensor_typeIntialized = true;
    this._wrapperFields.distances.fill(refObject.distances.toArray());
    this._incrementIntialized = true;
    this._min_distanceIntialized = true;
    this._max_distanceIntialized = true;
    this._angle_offsetIntialized = true;
  }
  copy(other) {
    this._refObject = new ObstacleDistanceRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._frameIntialized = true;
    this._sensor_typeIntialized = true;
    this._wrapperFields.distances.copy(other._wrapperFields.distances);
    this._incrementIntialized = true;
    this._min_distanceIntialized = true;
    this._max_distanceIntialized = true;
    this._angle_offsetIntialized = true;
  }
  static get classType() {
    return ObstacleDistanceWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"MAV_FRAME_GLOBAL","value":0},{"type":"uint8","name":"MAV_FRAME_LOCAL_NED","value":1},{"type":"uint8","name":"MAV_FRAME_BODY_FRD","value":12},{"type":"uint8","name":"MAV_DISTANCE_SENSOR_LASER","value":0},{"type":"uint8","name":"MAV_DISTANCE_SENSOR_ULTRASOUND","value":1},{"type":"uint8","name":"MAV_DISTANCE_SENSOR_INFRARED","value":2},{"type":"uint8","name":"MAV_DISTANCE_SENSOR_RADAR","value":3}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sensor_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"distances","type":{"isArray":true,"arraySize":72,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"increment","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"min_distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angle_offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"ObstacleDistance","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObstacleDistance"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","frame","sensor_type","distances","increment","min_distance","max_distance","angle_offset"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ObstacleDistanceArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObstacleDistanceWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObstacleDistanceWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObstacleDistanceArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObstacleDistanceArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ObstacleDistanceRefArray(size);
    this._refObject = new ObstacleDistanceRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObstacleDistanceWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObstacleDistanceArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObstacleDistanceArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObstacleDistanceWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObstacleDistanceWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObstacleDistanceArrayWrapper;
  }
}
// Define constants (7 in total)
Object.defineProperty(ObstacleDistanceWrapper, "MAV_FRAME_GLOBAL", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ObstacleDistanceWrapper, "MAV_FRAME_LOCAL_NED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ObstacleDistanceWrapper, "MAV_FRAME_BODY_FRD", {value: 12, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ObstacleDistanceWrapper, "MAV_DISTANCE_SENSOR_LASER", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ObstacleDistanceWrapper, "MAV_DISTANCE_SENSOR_ULTRASOUND", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ObstacleDistanceWrapper, "MAV_DISTANCE_SENSOR_INFRARED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ObstacleDistanceWrapper, "MAV_DISTANCE_SENSOR_RADAR", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = ObstacleDistanceWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "MAV_FRAME_GLOBAL",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "MAV_FRAME_LOCAL_NED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "MAV_FRAME_BODY_FRD",
      "value": 12
    },
    {
      "type": "uint8",
      "name": "MAV_DISTANCE_SENSOR_LASER",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "MAV_DISTANCE_SENSOR_ULTRASOUND",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "MAV_DISTANCE_SENSOR_INFRARED",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "MAV_DISTANCE_SENSOR_RADAR",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sensor_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "distances",
      "type": {
        "isArray": true,
        "arraySize": 72,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "increment",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "min_distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angle_offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "ObstacleDistance",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ObstacleDistance"
}
*/
