// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleConstraints.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const VehicleConstraintsRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  speed_up: primitiveTypes.float32,
  speed_down: primitiveTypes.float32,
  want_takeoff: primitiveTypes.bool,
});
const VehicleConstraintsRefArray = ArrayType(VehicleConstraintsRefStruct);
const VehicleConstraintsRefStructArray = StructType({
  data: VehicleConstraintsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleConstraintsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleConstraintsRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._speed_upIntialized = true;
      this._speed_downIntialized = true;
      this._want_takeoffIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleConstraintsRefStruct();
    this._timestampIntialized = false;
    this._speed_upIntialized = false;
    this._speed_downIntialized = false;
    this._want_takeoffIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleConstraintsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleConstraintsArrayWrapper;
  }
  static get ArrayType() {
    return VehicleConstraintsArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleConstraintsRefStructArray
  }
  static get refObjectType() {
    return VehicleConstraintsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleConstraints');
      }
      if (checkConsistency && !this._speed_upIntialized) {
        throw new TypeError('Invalid argument: speed_up in VehicleConstraints');
      }
      if (checkConsistency && !this._speed_downIntialized) {
        throw new TypeError('Invalid argument: speed_down in VehicleConstraints');
      }
      if (checkConsistency && !this._want_takeoffIntialized) {
        throw new TypeError('Invalid argument: want_takeoff in VehicleConstraints');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._speed_upIntialized = true;
    this._refObject.speed_up = refObject.speed_up;
    this._speed_downIntialized = true;
    this._refObject.speed_down = refObject.speed_down;
    this._want_takeoffIntialized = true;
    this._refObject.want_takeoff = refObject.want_takeoff;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleConstraintsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleConstraints'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get speed_up() {
    if (!this._speed_upIntialized) {
      return undefined;
    }
    return this._refObject.speed_up;
  }
  set speed_up(value) {
    this._speed_upIntialized = true;
    this._refObject.speed_up = value;
  }
  get speed_down() {
    if (!this._speed_downIntialized) {
      return undefined;
    }
    return this._refObject.speed_down;
  }
  set speed_down(value) {
    this._speed_downIntialized = true;
    this._refObject.speed_down = value;
  }
  get want_takeoff() {
    if (!this._want_takeoffIntialized) {
      return undefined;
    }
    return this._refObject.want_takeoff;
  }
  set want_takeoff(value) {
    this._want_takeoffIntialized = true;
    this._refObject.want_takeoff = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleConstraintsRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._speed_upIntialized = true;
    this._speed_downIntialized = true;
    this._want_takeoffIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleConstraintsRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._speed_upIntialized = true;
    this._speed_downIntialized = true;
    this._want_takeoffIntialized = true;
  }
  static get classType() {
    return VehicleConstraintsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"speed_up","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"speed_down","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"want_takeoff","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleConstraints","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleConstraints"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","speed_up","speed_down","want_takeoff"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleConstraintsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleConstraintsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleConstraintsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleConstraintsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleConstraintsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleConstraintsRefArray(size);
    this._refObject = new VehicleConstraintsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleConstraintsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleConstraintsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleConstraintsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleConstraintsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleConstraintsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleConstraintsArrayWrapper;
  }
}
module.exports = VehicleConstraintsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "speed_up",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "speed_down",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "want_takeoff",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleConstraints",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleConstraints"
}
*/
