// rclnodejs/generated/px4_msgs/px4_msgs__msg__DebugKeyValue.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CharWrapper = require('../../generated/std_msgs/std_msgs__msg__Char.js');
const DebugKeyValueRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  key: ArrayType(primitiveTypes.char, 10),
  value: primitiveTypes.float32,
});
const DebugKeyValueRefArray = ArrayType(DebugKeyValueRefStruct);
const DebugKeyValueRefStructArray = StructType({
  data: DebugKeyValueRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DebugKeyValueWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DebugKeyValueRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._wrapperFields.key = CharWrapper.createArray();
      this._wrapperFields.key.copy(other._wrapperFields.key);
      this._valueIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DebugKeyValueRefStruct();
    this._timestampIntialized = false;
    this._wrapperFields.key = CharWrapper.createArray();
    this._valueIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new DebugKeyValueWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DebugKeyValueArrayWrapper;
  }
  static get ArrayType() {
    return DebugKeyValueArrayWrapper;
  }
  static get refObjectArrayType() {
    return DebugKeyValueRefStructArray
  }
  static get refObjectType() {
    return DebugKeyValueRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in DebugKeyValue');
      }
      if (checkConsistency && !this._valueIntialized) {
        throw new TypeError('Invalid argument: value in DebugKeyValue');
      }
    this._refObject.key = Array.from(this._wrapperFields.key.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._wrapperFields.key.fill(refObject.key.toArray());
    this._valueIntialized = true;
    this._refObject.value = refObject.value;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    DebugKeyValueWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'DebugKeyValue'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get key() {
    return this._wrapperFields['key'].data;
  }
  set key(value) {
    if (value.length !== 10) {
      throw new RangeError('The length of the array must be 10.');
    }
    this._wrapperFields['key'].fill(value);
  }
  get value() {
    if (!this._valueIntialized) {
      return undefined;
    }
    return this._refObject.value;
  }
  set value(value) {
    this._valueIntialized = true;
    this._refObject.value = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DebugKeyValueRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.key.fill(refObject.key.toArray());
    this._valueIntialized = true;
  }
  copy(other) {
    this._refObject = new DebugKeyValueRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.key.copy(other._wrapperFields.key);
    this._valueIntialized = true;
  }
  static get classType() {
    return DebugKeyValueWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"key","type":{"isArray":true,"arraySize":10,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"char","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"DebugKeyValue","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DebugKeyValue"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","key","value"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DebugKeyValueArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DebugKeyValueWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DebugKeyValueWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DebugKeyValueArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DebugKeyValueArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DebugKeyValueRefArray(size);
    this._refObject = new DebugKeyValueRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DebugKeyValueWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DebugKeyValueArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DebugKeyValueArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DebugKeyValueWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DebugKeyValueWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DebugKeyValueArrayWrapper;
  }
}
module.exports = DebugKeyValueWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "key",
      "type": {
        "isArray": true,
        "arraySize": 10,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "char",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "DebugKeyValue",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DebugKeyValue"
}
*/
