// rclnodejs/generated/px4_msgs/px4_msgs__msg__GeofenceResult.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GeofenceResultRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  geofence_violation_reason: primitiveTypes.uint8,
  primary_geofence_breached: primitiveTypes.bool,
  primary_geofence_action: primitiveTypes.uint8,
  home_required: primitiveTypes.bool,
});
const GeofenceResultRefArray = ArrayType(GeofenceResultRefStruct);
const GeofenceResultRefStructArray = StructType({
  data: GeofenceResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GeofenceResultWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GeofenceResultRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._geofence_violation_reasonIntialized = true;
      this._primary_geofence_breachedIntialized = true;
      this._primary_geofence_actionIntialized = true;
      this._home_requiredIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GeofenceResultRefStruct();
    this._timestampIntialized = false;
    this._geofence_violation_reasonIntialized = false;
    this._primary_geofence_breachedIntialized = false;
    this._primary_geofence_actionIntialized = false;
    this._home_requiredIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GeofenceResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GeofenceResultArrayWrapper;
  }
  static get ArrayType() {
    return GeofenceResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return GeofenceResultRefStructArray
  }
  static get refObjectType() {
    return GeofenceResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in GeofenceResult');
      }
      if (checkConsistency && !this._geofence_violation_reasonIntialized) {
        throw new TypeError('Invalid argument: geofence_violation_reason in GeofenceResult');
      }
      if (checkConsistency && !this._primary_geofence_breachedIntialized) {
        throw new TypeError('Invalid argument: primary_geofence_breached in GeofenceResult');
      }
      if (checkConsistency && !this._primary_geofence_actionIntialized) {
        throw new TypeError('Invalid argument: primary_geofence_action in GeofenceResult');
      }
      if (checkConsistency && !this._home_requiredIntialized) {
        throw new TypeError('Invalid argument: home_required in GeofenceResult');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._geofence_violation_reasonIntialized = true;
    this._refObject.geofence_violation_reason = refObject.geofence_violation_reason;
    this._primary_geofence_breachedIntialized = true;
    this._refObject.primary_geofence_breached = refObject.primary_geofence_breached;
    this._primary_geofence_actionIntialized = true;
    this._refObject.primary_geofence_action = refObject.primary_geofence_action;
    this._home_requiredIntialized = true;
    this._refObject.home_required = refObject.home_required;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GeofenceResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'GeofenceResult'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get geofence_violation_reason() {
    if (!this._geofence_violation_reasonIntialized) {
      return undefined;
    }
    return this._refObject.geofence_violation_reason;
  }
  set geofence_violation_reason(value) {
    this._geofence_violation_reasonIntialized = true;
    this._refObject.geofence_violation_reason = value;
  }
  get primary_geofence_breached() {
    if (!this._primary_geofence_breachedIntialized) {
      return undefined;
    }
    return this._refObject.primary_geofence_breached;
  }
  set primary_geofence_breached(value) {
    this._primary_geofence_breachedIntialized = true;
    this._refObject.primary_geofence_breached = value;
  }
  get primary_geofence_action() {
    if (!this._primary_geofence_actionIntialized) {
      return undefined;
    }
    return this._refObject.primary_geofence_action;
  }
  set primary_geofence_action(value) {
    this._primary_geofence_actionIntialized = true;
    this._refObject.primary_geofence_action = value;
  }
  get home_required() {
    if (!this._home_requiredIntialized) {
      return undefined;
    }
    return this._refObject.home_required;
  }
  set home_required(value) {
    this._home_requiredIntialized = true;
    this._refObject.home_required = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GeofenceResultRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._geofence_violation_reasonIntialized = true;
    this._primary_geofence_breachedIntialized = true;
    this._primary_geofence_actionIntialized = true;
    this._home_requiredIntialized = true;
  }
  copy(other) {
    this._refObject = new GeofenceResultRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._geofence_violation_reasonIntialized = true;
    this._primary_geofence_breachedIntialized = true;
    this._primary_geofence_actionIntialized = true;
    this._home_requiredIntialized = true;
  }
  static get classType() {
    return GeofenceResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"GF_ACTION_NONE","value":0},{"type":"uint8","name":"GF_ACTION_WARN","value":1},{"type":"uint8","name":"GF_ACTION_LOITER","value":2},{"type":"uint8","name":"GF_ACTION_RTL","value":3},{"type":"uint8","name":"GF_ACTION_TERMINATE","value":4},{"type":"uint8","name":"GF_ACTION_LAND","value":5}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"geofence_violation_reason","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"primary_geofence_breached","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"primary_geofence_action","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"home_required","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"GeofenceResult","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GeofenceResult"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","geofence_violation_reason","primary_geofence_breached","primary_geofence_action","home_required"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GeofenceResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GeofenceResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GeofenceResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GeofenceResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GeofenceResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GeofenceResultRefArray(size);
    this._refObject = new GeofenceResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GeofenceResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GeofenceResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GeofenceResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GeofenceResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GeofenceResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GeofenceResultArrayWrapper;
  }
}
// Define constants (6 in total)
Object.defineProperty(GeofenceResultWrapper, "GF_ACTION_NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GeofenceResultWrapper, "GF_ACTION_WARN", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GeofenceResultWrapper, "GF_ACTION_LOITER", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GeofenceResultWrapper, "GF_ACTION_RTL", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GeofenceResultWrapper, "GF_ACTION_TERMINATE", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GeofenceResultWrapper, "GF_ACTION_LAND", {value: 5, writable: false, enumerable: true, configurable: true});
module.exports = GeofenceResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "GF_ACTION_NONE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "GF_ACTION_WARN",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "GF_ACTION_LOITER",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "GF_ACTION_RTL",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "GF_ACTION_TERMINATE",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "GF_ACTION_LAND",
      "value": 5
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "geofence_violation_reason",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "primary_geofence_breached",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "primary_geofence_action",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "home_required",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "GeofenceResult",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GeofenceResult"
}
*/
