// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleControlMode.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const VehicleControlModeRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  flag_armed: primitiveTypes.bool,
  flag_multicopter_position_control_enabled: primitiveTypes.bool,
  flag_control_manual_enabled: primitiveTypes.bool,
  flag_control_auto_enabled: primitiveTypes.bool,
  flag_control_offboard_enabled: primitiveTypes.bool,
  flag_control_rates_enabled: primitiveTypes.bool,
  flag_control_attitude_enabled: primitiveTypes.bool,
  flag_control_acceleration_enabled: primitiveTypes.bool,
  flag_control_velocity_enabled: primitiveTypes.bool,
  flag_control_position_enabled: primitiveTypes.bool,
  flag_control_altitude_enabled: primitiveTypes.bool,
  flag_control_climb_rate_enabled: primitiveTypes.bool,
  flag_control_termination_enabled: primitiveTypes.bool,
});
const VehicleControlModeRefArray = ArrayType(VehicleControlModeRefStruct);
const VehicleControlModeRefStructArray = StructType({
  data: VehicleControlModeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleControlModeWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleControlModeRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._flag_armedIntialized = true;
      this._flag_multicopter_position_control_enabledIntialized = true;
      this._flag_control_manual_enabledIntialized = true;
      this._flag_control_auto_enabledIntialized = true;
      this._flag_control_offboard_enabledIntialized = true;
      this._flag_control_rates_enabledIntialized = true;
      this._flag_control_attitude_enabledIntialized = true;
      this._flag_control_acceleration_enabledIntialized = true;
      this._flag_control_velocity_enabledIntialized = true;
      this._flag_control_position_enabledIntialized = true;
      this._flag_control_altitude_enabledIntialized = true;
      this._flag_control_climb_rate_enabledIntialized = true;
      this._flag_control_termination_enabledIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleControlModeRefStruct();
    this._timestampIntialized = false;
    this._flag_armedIntialized = false;
    this._flag_multicopter_position_control_enabledIntialized = false;
    this._flag_control_manual_enabledIntialized = false;
    this._flag_control_auto_enabledIntialized = false;
    this._flag_control_offboard_enabledIntialized = false;
    this._flag_control_rates_enabledIntialized = false;
    this._flag_control_attitude_enabledIntialized = false;
    this._flag_control_acceleration_enabledIntialized = false;
    this._flag_control_velocity_enabledIntialized = false;
    this._flag_control_position_enabledIntialized = false;
    this._flag_control_altitude_enabledIntialized = false;
    this._flag_control_climb_rate_enabledIntialized = false;
    this._flag_control_termination_enabledIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleControlModeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleControlModeArrayWrapper;
  }
  static get ArrayType() {
    return VehicleControlModeArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleControlModeRefStructArray
  }
  static get refObjectType() {
    return VehicleControlModeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_armedIntialized) {
        throw new TypeError('Invalid argument: flag_armed in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_multicopter_position_control_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_multicopter_position_control_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_manual_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_manual_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_auto_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_auto_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_offboard_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_offboard_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_rates_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_rates_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_attitude_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_attitude_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_acceleration_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_acceleration_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_velocity_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_velocity_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_position_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_position_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_altitude_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_altitude_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_climb_rate_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_climb_rate_enabled in VehicleControlMode');
      }
      if (checkConsistency && !this._flag_control_termination_enabledIntialized) {
        throw new TypeError('Invalid argument: flag_control_termination_enabled in VehicleControlMode');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._flag_armedIntialized = true;
    this._refObject.flag_armed = refObject.flag_armed;
    this._flag_multicopter_position_control_enabledIntialized = true;
    this._refObject.flag_multicopter_position_control_enabled = refObject.flag_multicopter_position_control_enabled;
    this._flag_control_manual_enabledIntialized = true;
    this._refObject.flag_control_manual_enabled = refObject.flag_control_manual_enabled;
    this._flag_control_auto_enabledIntialized = true;
    this._refObject.flag_control_auto_enabled = refObject.flag_control_auto_enabled;
    this._flag_control_offboard_enabledIntialized = true;
    this._refObject.flag_control_offboard_enabled = refObject.flag_control_offboard_enabled;
    this._flag_control_rates_enabledIntialized = true;
    this._refObject.flag_control_rates_enabled = refObject.flag_control_rates_enabled;
    this._flag_control_attitude_enabledIntialized = true;
    this._refObject.flag_control_attitude_enabled = refObject.flag_control_attitude_enabled;
    this._flag_control_acceleration_enabledIntialized = true;
    this._refObject.flag_control_acceleration_enabled = refObject.flag_control_acceleration_enabled;
    this._flag_control_velocity_enabledIntialized = true;
    this._refObject.flag_control_velocity_enabled = refObject.flag_control_velocity_enabled;
    this._flag_control_position_enabledIntialized = true;
    this._refObject.flag_control_position_enabled = refObject.flag_control_position_enabled;
    this._flag_control_altitude_enabledIntialized = true;
    this._refObject.flag_control_altitude_enabled = refObject.flag_control_altitude_enabled;
    this._flag_control_climb_rate_enabledIntialized = true;
    this._refObject.flag_control_climb_rate_enabled = refObject.flag_control_climb_rate_enabled;
    this._flag_control_termination_enabledIntialized = true;
    this._refObject.flag_control_termination_enabled = refObject.flag_control_termination_enabled;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleControlModeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleControlMode'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get flag_armed() {
    if (!this._flag_armedIntialized) {
      return undefined;
    }
    return this._refObject.flag_armed;
  }
  set flag_armed(value) {
    this._flag_armedIntialized = true;
    this._refObject.flag_armed = value;
  }
  get flag_multicopter_position_control_enabled() {
    if (!this._flag_multicopter_position_control_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_multicopter_position_control_enabled;
  }
  set flag_multicopter_position_control_enabled(value) {
    this._flag_multicopter_position_control_enabledIntialized = true;
    this._refObject.flag_multicopter_position_control_enabled = value;
  }
  get flag_control_manual_enabled() {
    if (!this._flag_control_manual_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_manual_enabled;
  }
  set flag_control_manual_enabled(value) {
    this._flag_control_manual_enabledIntialized = true;
    this._refObject.flag_control_manual_enabled = value;
  }
  get flag_control_auto_enabled() {
    if (!this._flag_control_auto_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_auto_enabled;
  }
  set flag_control_auto_enabled(value) {
    this._flag_control_auto_enabledIntialized = true;
    this._refObject.flag_control_auto_enabled = value;
  }
  get flag_control_offboard_enabled() {
    if (!this._flag_control_offboard_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_offboard_enabled;
  }
  set flag_control_offboard_enabled(value) {
    this._flag_control_offboard_enabledIntialized = true;
    this._refObject.flag_control_offboard_enabled = value;
  }
  get flag_control_rates_enabled() {
    if (!this._flag_control_rates_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_rates_enabled;
  }
  set flag_control_rates_enabled(value) {
    this._flag_control_rates_enabledIntialized = true;
    this._refObject.flag_control_rates_enabled = value;
  }
  get flag_control_attitude_enabled() {
    if (!this._flag_control_attitude_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_attitude_enabled;
  }
  set flag_control_attitude_enabled(value) {
    this._flag_control_attitude_enabledIntialized = true;
    this._refObject.flag_control_attitude_enabled = value;
  }
  get flag_control_acceleration_enabled() {
    if (!this._flag_control_acceleration_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_acceleration_enabled;
  }
  set flag_control_acceleration_enabled(value) {
    this._flag_control_acceleration_enabledIntialized = true;
    this._refObject.flag_control_acceleration_enabled = value;
  }
  get flag_control_velocity_enabled() {
    if (!this._flag_control_velocity_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_velocity_enabled;
  }
  set flag_control_velocity_enabled(value) {
    this._flag_control_velocity_enabledIntialized = true;
    this._refObject.flag_control_velocity_enabled = value;
  }
  get flag_control_position_enabled() {
    if (!this._flag_control_position_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_position_enabled;
  }
  set flag_control_position_enabled(value) {
    this._flag_control_position_enabledIntialized = true;
    this._refObject.flag_control_position_enabled = value;
  }
  get flag_control_altitude_enabled() {
    if (!this._flag_control_altitude_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_altitude_enabled;
  }
  set flag_control_altitude_enabled(value) {
    this._flag_control_altitude_enabledIntialized = true;
    this._refObject.flag_control_altitude_enabled = value;
  }
  get flag_control_climb_rate_enabled() {
    if (!this._flag_control_climb_rate_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_climb_rate_enabled;
  }
  set flag_control_climb_rate_enabled(value) {
    this._flag_control_climb_rate_enabledIntialized = true;
    this._refObject.flag_control_climb_rate_enabled = value;
  }
  get flag_control_termination_enabled() {
    if (!this._flag_control_termination_enabledIntialized) {
      return undefined;
    }
    return this._refObject.flag_control_termination_enabled;
  }
  set flag_control_termination_enabled(value) {
    this._flag_control_termination_enabledIntialized = true;
    this._refObject.flag_control_termination_enabled = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleControlModeRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._flag_armedIntialized = true;
    this._flag_multicopter_position_control_enabledIntialized = true;
    this._flag_control_manual_enabledIntialized = true;
    this._flag_control_auto_enabledIntialized = true;
    this._flag_control_offboard_enabledIntialized = true;
    this._flag_control_rates_enabledIntialized = true;
    this._flag_control_attitude_enabledIntialized = true;
    this._flag_control_acceleration_enabledIntialized = true;
    this._flag_control_velocity_enabledIntialized = true;
    this._flag_control_position_enabledIntialized = true;
    this._flag_control_altitude_enabledIntialized = true;
    this._flag_control_climb_rate_enabledIntialized = true;
    this._flag_control_termination_enabledIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleControlModeRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._flag_armedIntialized = true;
    this._flag_multicopter_position_control_enabledIntialized = true;
    this._flag_control_manual_enabledIntialized = true;
    this._flag_control_auto_enabledIntialized = true;
    this._flag_control_offboard_enabledIntialized = true;
    this._flag_control_rates_enabledIntialized = true;
    this._flag_control_attitude_enabledIntialized = true;
    this._flag_control_acceleration_enabledIntialized = true;
    this._flag_control_velocity_enabledIntialized = true;
    this._flag_control_position_enabledIntialized = true;
    this._flag_control_altitude_enabledIntialized = true;
    this._flag_control_climb_rate_enabledIntialized = true;
    this._flag_control_termination_enabledIntialized = true;
  }
  static get classType() {
    return VehicleControlModeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_armed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_multicopter_position_control_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_manual_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_auto_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_offboard_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_rates_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_attitude_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_acceleration_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_velocity_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_position_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_altitude_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_climb_rate_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flag_control_termination_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleControlMode","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleControlMode"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","flag_armed","flag_multicopter_position_control_enabled","flag_control_manual_enabled","flag_control_auto_enabled","flag_control_offboard_enabled","flag_control_rates_enabled","flag_control_attitude_enabled","flag_control_acceleration_enabled","flag_control_velocity_enabled","flag_control_position_enabled","flag_control_altitude_enabled","flag_control_climb_rate_enabled","flag_control_termination_enabled"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleControlModeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleControlModeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleControlModeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleControlModeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleControlModeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleControlModeRefArray(size);
    this._refObject = new VehicleControlModeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleControlModeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleControlModeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleControlModeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleControlModeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleControlModeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleControlModeArrayWrapper;
  }
}
module.exports = VehicleControlModeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_armed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_multicopter_position_control_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_manual_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_auto_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_offboard_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_rates_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_attitude_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_acceleration_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_velocity_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_position_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_altitude_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_climb_rate_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flag_control_termination_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleControlMode",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleControlMode"
}
*/
