// rclnodejs/generated/px4_msgs/px4_msgs__msg__TaskStackInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CharWrapper = require('../../generated/std_msgs/std_msgs__msg__Char.js');
const TaskStackInfoRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  stack_free: primitiveTypes.uint16,
  task_name: ArrayType(primitiveTypes.char, 24),
});
const TaskStackInfoRefArray = ArrayType(TaskStackInfoRefStruct);
const TaskStackInfoRefStructArray = StructType({
  data: TaskStackInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TaskStackInfoWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TaskStackInfoRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._stack_freeIntialized = true;
      this._wrapperFields.task_name = CharWrapper.createArray();
      this._wrapperFields.task_name.copy(other._wrapperFields.task_name);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TaskStackInfoRefStruct();
    this._timestampIntialized = false;
    this._stack_freeIntialized = false;
    this._wrapperFields.task_name = CharWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new TaskStackInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TaskStackInfoArrayWrapper;
  }
  static get ArrayType() {
    return TaskStackInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return TaskStackInfoRefStructArray
  }
  static get refObjectType() {
    return TaskStackInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in TaskStackInfo');
      }
      if (checkConsistency && !this._stack_freeIntialized) {
        throw new TypeError('Invalid argument: stack_free in TaskStackInfo');
      }
    this._refObject.task_name = Array.from(this._wrapperFields.task_name.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._stack_freeIntialized = true;
    this._refObject.stack_free = refObject.stack_free;
    this._wrapperFields.task_name.fill(refObject.task_name.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    TaskStackInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'TaskStackInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get stack_free() {
    if (!this._stack_freeIntialized) {
      return undefined;
    }
    return this._refObject.stack_free;
  }
  set stack_free(value) {
    this._stack_freeIntialized = true;
    this._refObject.stack_free = value;
  }
  get task_name() {
    return this._wrapperFields['task_name'].data;
  }
  set task_name(value) {
    if (value.length !== 24) {
      throw new RangeError('The length of the array must be 24.');
    }
    this._wrapperFields['task_name'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new TaskStackInfoRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._stack_freeIntialized = true;
    this._wrapperFields.task_name.fill(refObject.task_name.toArray());
  }
  copy(other) {
    this._refObject = new TaskStackInfoRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._stack_freeIntialized = true;
    this._wrapperFields.task_name.copy(other._wrapperFields.task_name);
  }
  static get classType() {
    return TaskStackInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ORB_QUEUE_LENGTH","value":2}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stack_free","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"task_name","type":{"isArray":true,"arraySize":24,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"char","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"TaskStackInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TaskStackInfo"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","stack_free","task_name"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TaskStackInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TaskStackInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TaskStackInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TaskStackInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TaskStackInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TaskStackInfoRefArray(size);
    this._refObject = new TaskStackInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TaskStackInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TaskStackInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TaskStackInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TaskStackInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TaskStackInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TaskStackInfoArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(TaskStackInfoWrapper, "ORB_QUEUE_LENGTH", {value: 2, writable: false, enumerable: true, configurable: true});
module.exports = TaskStackInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ORB_QUEUE_LENGTH",
      "value": 2
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stack_free",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "task_name",
      "type": {
        "isArray": true,
        "arraySize": 24,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "char",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "TaskStackInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TaskStackInfo"
}
*/
