// rclnodejs/generated/px4_msgs/px4_msgs__msg__LandingTargetPose.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LandingTargetPoseRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  is_static: primitiveTypes.bool,
  rel_pos_valid: primitiveTypes.bool,
  rel_vel_valid: primitiveTypes.bool,
  x_rel: primitiveTypes.float32,
  y_rel: primitiveTypes.float32,
  z_rel: primitiveTypes.float32,
  vx_rel: primitiveTypes.float32,
  vy_rel: primitiveTypes.float32,
  cov_x_rel: primitiveTypes.float32,
  cov_y_rel: primitiveTypes.float32,
  cov_vx_rel: primitiveTypes.float32,
  cov_vy_rel: primitiveTypes.float32,
  abs_pos_valid: primitiveTypes.bool,
  x_abs: primitiveTypes.float32,
  y_abs: primitiveTypes.float32,
  z_abs: primitiveTypes.float32,
});
const LandingTargetPoseRefArray = ArrayType(LandingTargetPoseRefStruct);
const LandingTargetPoseRefStructArray = StructType({
  data: LandingTargetPoseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LandingTargetPoseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LandingTargetPoseRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._is_staticIntialized = true;
      this._rel_pos_validIntialized = true;
      this._rel_vel_validIntialized = true;
      this._x_relIntialized = true;
      this._y_relIntialized = true;
      this._z_relIntialized = true;
      this._vx_relIntialized = true;
      this._vy_relIntialized = true;
      this._cov_x_relIntialized = true;
      this._cov_y_relIntialized = true;
      this._cov_vx_relIntialized = true;
      this._cov_vy_relIntialized = true;
      this._abs_pos_validIntialized = true;
      this._x_absIntialized = true;
      this._y_absIntialized = true;
      this._z_absIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LandingTargetPoseRefStruct();
    this._timestampIntialized = false;
    this._is_staticIntialized = false;
    this._rel_pos_validIntialized = false;
    this._rel_vel_validIntialized = false;
    this._x_relIntialized = false;
    this._y_relIntialized = false;
    this._z_relIntialized = false;
    this._vx_relIntialized = false;
    this._vy_relIntialized = false;
    this._cov_x_relIntialized = false;
    this._cov_y_relIntialized = false;
    this._cov_vx_relIntialized = false;
    this._cov_vy_relIntialized = false;
    this._abs_pos_validIntialized = false;
    this._x_absIntialized = false;
    this._y_absIntialized = false;
    this._z_absIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LandingTargetPoseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LandingTargetPoseArrayWrapper;
  }
  static get ArrayType() {
    return LandingTargetPoseArrayWrapper;
  }
  static get refObjectArrayType() {
    return LandingTargetPoseRefStructArray
  }
  static get refObjectType() {
    return LandingTargetPoseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in LandingTargetPose');
      }
      if (checkConsistency && !this._is_staticIntialized) {
        throw new TypeError('Invalid argument: is_static in LandingTargetPose');
      }
      if (checkConsistency && !this._rel_pos_validIntialized) {
        throw new TypeError('Invalid argument: rel_pos_valid in LandingTargetPose');
      }
      if (checkConsistency && !this._rel_vel_validIntialized) {
        throw new TypeError('Invalid argument: rel_vel_valid in LandingTargetPose');
      }
      if (checkConsistency && !this._x_relIntialized) {
        throw new TypeError('Invalid argument: x_rel in LandingTargetPose');
      }
      if (checkConsistency && !this._y_relIntialized) {
        throw new TypeError('Invalid argument: y_rel in LandingTargetPose');
      }
      if (checkConsistency && !this._z_relIntialized) {
        throw new TypeError('Invalid argument: z_rel in LandingTargetPose');
      }
      if (checkConsistency && !this._vx_relIntialized) {
        throw new TypeError('Invalid argument: vx_rel in LandingTargetPose');
      }
      if (checkConsistency && !this._vy_relIntialized) {
        throw new TypeError('Invalid argument: vy_rel in LandingTargetPose');
      }
      if (checkConsistency && !this._cov_x_relIntialized) {
        throw new TypeError('Invalid argument: cov_x_rel in LandingTargetPose');
      }
      if (checkConsistency && !this._cov_y_relIntialized) {
        throw new TypeError('Invalid argument: cov_y_rel in LandingTargetPose');
      }
      if (checkConsistency && !this._cov_vx_relIntialized) {
        throw new TypeError('Invalid argument: cov_vx_rel in LandingTargetPose');
      }
      if (checkConsistency && !this._cov_vy_relIntialized) {
        throw new TypeError('Invalid argument: cov_vy_rel in LandingTargetPose');
      }
      if (checkConsistency && !this._abs_pos_validIntialized) {
        throw new TypeError('Invalid argument: abs_pos_valid in LandingTargetPose');
      }
      if (checkConsistency && !this._x_absIntialized) {
        throw new TypeError('Invalid argument: x_abs in LandingTargetPose');
      }
      if (checkConsistency && !this._y_absIntialized) {
        throw new TypeError('Invalid argument: y_abs in LandingTargetPose');
      }
      if (checkConsistency && !this._z_absIntialized) {
        throw new TypeError('Invalid argument: z_abs in LandingTargetPose');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._is_staticIntialized = true;
    this._refObject.is_static = refObject.is_static;
    this._rel_pos_validIntialized = true;
    this._refObject.rel_pos_valid = refObject.rel_pos_valid;
    this._rel_vel_validIntialized = true;
    this._refObject.rel_vel_valid = refObject.rel_vel_valid;
    this._x_relIntialized = true;
    this._refObject.x_rel = refObject.x_rel;
    this._y_relIntialized = true;
    this._refObject.y_rel = refObject.y_rel;
    this._z_relIntialized = true;
    this._refObject.z_rel = refObject.z_rel;
    this._vx_relIntialized = true;
    this._refObject.vx_rel = refObject.vx_rel;
    this._vy_relIntialized = true;
    this._refObject.vy_rel = refObject.vy_rel;
    this._cov_x_relIntialized = true;
    this._refObject.cov_x_rel = refObject.cov_x_rel;
    this._cov_y_relIntialized = true;
    this._refObject.cov_y_rel = refObject.cov_y_rel;
    this._cov_vx_relIntialized = true;
    this._refObject.cov_vx_rel = refObject.cov_vx_rel;
    this._cov_vy_relIntialized = true;
    this._refObject.cov_vy_rel = refObject.cov_vy_rel;
    this._abs_pos_validIntialized = true;
    this._refObject.abs_pos_valid = refObject.abs_pos_valid;
    this._x_absIntialized = true;
    this._refObject.x_abs = refObject.x_abs;
    this._y_absIntialized = true;
    this._refObject.y_abs = refObject.y_abs;
    this._z_absIntialized = true;
    this._refObject.z_abs = refObject.z_abs;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LandingTargetPoseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'LandingTargetPose'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get is_static() {
    if (!this._is_staticIntialized) {
      return undefined;
    }
    return this._refObject.is_static;
  }
  set is_static(value) {
    this._is_staticIntialized = true;
    this._refObject.is_static = value;
  }
  get rel_pos_valid() {
    if (!this._rel_pos_validIntialized) {
      return undefined;
    }
    return this._refObject.rel_pos_valid;
  }
  set rel_pos_valid(value) {
    this._rel_pos_validIntialized = true;
    this._refObject.rel_pos_valid = value;
  }
  get rel_vel_valid() {
    if (!this._rel_vel_validIntialized) {
      return undefined;
    }
    return this._refObject.rel_vel_valid;
  }
  set rel_vel_valid(value) {
    this._rel_vel_validIntialized = true;
    this._refObject.rel_vel_valid = value;
  }
  get x_rel() {
    if (!this._x_relIntialized) {
      return undefined;
    }
    return this._refObject.x_rel;
  }
  set x_rel(value) {
    this._x_relIntialized = true;
    this._refObject.x_rel = value;
  }
  get y_rel() {
    if (!this._y_relIntialized) {
      return undefined;
    }
    return this._refObject.y_rel;
  }
  set y_rel(value) {
    this._y_relIntialized = true;
    this._refObject.y_rel = value;
  }
  get z_rel() {
    if (!this._z_relIntialized) {
      return undefined;
    }
    return this._refObject.z_rel;
  }
  set z_rel(value) {
    this._z_relIntialized = true;
    this._refObject.z_rel = value;
  }
  get vx_rel() {
    if (!this._vx_relIntialized) {
      return undefined;
    }
    return this._refObject.vx_rel;
  }
  set vx_rel(value) {
    this._vx_relIntialized = true;
    this._refObject.vx_rel = value;
  }
  get vy_rel() {
    if (!this._vy_relIntialized) {
      return undefined;
    }
    return this._refObject.vy_rel;
  }
  set vy_rel(value) {
    this._vy_relIntialized = true;
    this._refObject.vy_rel = value;
  }
  get cov_x_rel() {
    if (!this._cov_x_relIntialized) {
      return undefined;
    }
    return this._refObject.cov_x_rel;
  }
  set cov_x_rel(value) {
    this._cov_x_relIntialized = true;
    this._refObject.cov_x_rel = value;
  }
  get cov_y_rel() {
    if (!this._cov_y_relIntialized) {
      return undefined;
    }
    return this._refObject.cov_y_rel;
  }
  set cov_y_rel(value) {
    this._cov_y_relIntialized = true;
    this._refObject.cov_y_rel = value;
  }
  get cov_vx_rel() {
    if (!this._cov_vx_relIntialized) {
      return undefined;
    }
    return this._refObject.cov_vx_rel;
  }
  set cov_vx_rel(value) {
    this._cov_vx_relIntialized = true;
    this._refObject.cov_vx_rel = value;
  }
  get cov_vy_rel() {
    if (!this._cov_vy_relIntialized) {
      return undefined;
    }
    return this._refObject.cov_vy_rel;
  }
  set cov_vy_rel(value) {
    this._cov_vy_relIntialized = true;
    this._refObject.cov_vy_rel = value;
  }
  get abs_pos_valid() {
    if (!this._abs_pos_validIntialized) {
      return undefined;
    }
    return this._refObject.abs_pos_valid;
  }
  set abs_pos_valid(value) {
    this._abs_pos_validIntialized = true;
    this._refObject.abs_pos_valid = value;
  }
  get x_abs() {
    if (!this._x_absIntialized) {
      return undefined;
    }
    return this._refObject.x_abs;
  }
  set x_abs(value) {
    this._x_absIntialized = true;
    this._refObject.x_abs = value;
  }
  get y_abs() {
    if (!this._y_absIntialized) {
      return undefined;
    }
    return this._refObject.y_abs;
  }
  set y_abs(value) {
    this._y_absIntialized = true;
    this._refObject.y_abs = value;
  }
  get z_abs() {
    if (!this._z_absIntialized) {
      return undefined;
    }
    return this._refObject.z_abs;
  }
  set z_abs(value) {
    this._z_absIntialized = true;
    this._refObject.z_abs = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LandingTargetPoseRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._is_staticIntialized = true;
    this._rel_pos_validIntialized = true;
    this._rel_vel_validIntialized = true;
    this._x_relIntialized = true;
    this._y_relIntialized = true;
    this._z_relIntialized = true;
    this._vx_relIntialized = true;
    this._vy_relIntialized = true;
    this._cov_x_relIntialized = true;
    this._cov_y_relIntialized = true;
    this._cov_vx_relIntialized = true;
    this._cov_vy_relIntialized = true;
    this._abs_pos_validIntialized = true;
    this._x_absIntialized = true;
    this._y_absIntialized = true;
    this._z_absIntialized = true;
  }
  copy(other) {
    this._refObject = new LandingTargetPoseRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._is_staticIntialized = true;
    this._rel_pos_validIntialized = true;
    this._rel_vel_validIntialized = true;
    this._x_relIntialized = true;
    this._y_relIntialized = true;
    this._z_relIntialized = true;
    this._vx_relIntialized = true;
    this._vy_relIntialized = true;
    this._cov_x_relIntialized = true;
    this._cov_y_relIntialized = true;
    this._cov_vx_relIntialized = true;
    this._cov_vy_relIntialized = true;
    this._abs_pos_validIntialized = true;
    this._x_absIntialized = true;
    this._y_absIntialized = true;
    this._z_absIntialized = true;
  }
  static get classType() {
    return LandingTargetPoseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"is_static","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rel_pos_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rel_vel_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x_rel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y_rel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z_rel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vx_rel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vy_rel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cov_x_rel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cov_y_rel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cov_vx_rel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cov_vy_rel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"abs_pos_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x_abs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y_abs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z_abs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"LandingTargetPose","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LandingTargetPose"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","is_static","rel_pos_valid","rel_vel_valid","x_rel","y_rel","z_rel","vx_rel","vy_rel","cov_x_rel","cov_y_rel","cov_vx_rel","cov_vy_rel","abs_pos_valid","x_abs","y_abs","z_abs"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LandingTargetPoseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LandingTargetPoseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LandingTargetPoseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandingTargetPoseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LandingTargetPoseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LandingTargetPoseRefArray(size);
    this._refObject = new LandingTargetPoseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LandingTargetPoseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LandingTargetPoseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LandingTargetPoseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LandingTargetPoseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LandingTargetPoseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LandingTargetPoseArrayWrapper;
  }
}
module.exports = LandingTargetPoseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "is_static",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rel_pos_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rel_vel_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x_rel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y_rel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z_rel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vx_rel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vy_rel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cov_x_rel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cov_y_rel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cov_vx_rel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cov_vy_rel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "abs_pos_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x_abs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y_abs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z_abs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "LandingTargetPose",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LandingTargetPose"
}
*/
