// rclnodejs/generated/px4_msgs/px4_msgs__msg__GimbalManagerStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GimbalManagerStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  flags: primitiveTypes.uint32,
  gimbal_device_id: primitiveTypes.uint8,
  primary_control_sysid: primitiveTypes.uint8,
  primary_control_compid: primitiveTypes.uint8,
  secondary_control_sysid: primitiveTypes.uint8,
  secondary_control_compid: primitiveTypes.uint8,
});
const GimbalManagerStatusRefArray = ArrayType(GimbalManagerStatusRefStruct);
const GimbalManagerStatusRefStructArray = StructType({
  data: GimbalManagerStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GimbalManagerStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GimbalManagerStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._flagsIntialized = true;
      this._gimbal_device_idIntialized = true;
      this._primary_control_sysidIntialized = true;
      this._primary_control_compidIntialized = true;
      this._secondary_control_sysidIntialized = true;
      this._secondary_control_compidIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GimbalManagerStatusRefStruct();
    this._timestampIntialized = false;
    this._flagsIntialized = false;
    this._gimbal_device_idIntialized = false;
    this._primary_control_sysidIntialized = false;
    this._primary_control_compidIntialized = false;
    this._secondary_control_sysidIntialized = false;
    this._secondary_control_compidIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GimbalManagerStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GimbalManagerStatusArrayWrapper;
  }
  static get ArrayType() {
    return GimbalManagerStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return GimbalManagerStatusRefStructArray
  }
  static get refObjectType() {
    return GimbalManagerStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in GimbalManagerStatus');
      }
      if (checkConsistency && !this._flagsIntialized) {
        throw new TypeError('Invalid argument: flags in GimbalManagerStatus');
      }
      if (checkConsistency && !this._gimbal_device_idIntialized) {
        throw new TypeError('Invalid argument: gimbal_device_id in GimbalManagerStatus');
      }
      if (checkConsistency && !this._primary_control_sysidIntialized) {
        throw new TypeError('Invalid argument: primary_control_sysid in GimbalManagerStatus');
      }
      if (checkConsistency && !this._primary_control_compidIntialized) {
        throw new TypeError('Invalid argument: primary_control_compid in GimbalManagerStatus');
      }
      if (checkConsistency && !this._secondary_control_sysidIntialized) {
        throw new TypeError('Invalid argument: secondary_control_sysid in GimbalManagerStatus');
      }
      if (checkConsistency && !this._secondary_control_compidIntialized) {
        throw new TypeError('Invalid argument: secondary_control_compid in GimbalManagerStatus');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._flagsIntialized = true;
    this._refObject.flags = refObject.flags;
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = refObject.gimbal_device_id;
    this._primary_control_sysidIntialized = true;
    this._refObject.primary_control_sysid = refObject.primary_control_sysid;
    this._primary_control_compidIntialized = true;
    this._refObject.primary_control_compid = refObject.primary_control_compid;
    this._secondary_control_sysidIntialized = true;
    this._refObject.secondary_control_sysid = refObject.secondary_control_sysid;
    this._secondary_control_compidIntialized = true;
    this._refObject.secondary_control_compid = refObject.secondary_control_compid;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GimbalManagerStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'GimbalManagerStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get flags() {
    if (!this._flagsIntialized) {
      return undefined;
    }
    return this._refObject.flags;
  }
  set flags(value) {
    this._flagsIntialized = true;
    this._refObject.flags = value;
  }
  get gimbal_device_id() {
    if (!this._gimbal_device_idIntialized) {
      return undefined;
    }
    return this._refObject.gimbal_device_id;
  }
  set gimbal_device_id(value) {
    this._gimbal_device_idIntialized = true;
    this._refObject.gimbal_device_id = value;
  }
  get primary_control_sysid() {
    if (!this._primary_control_sysidIntialized) {
      return undefined;
    }
    return this._refObject.primary_control_sysid;
  }
  set primary_control_sysid(value) {
    this._primary_control_sysidIntialized = true;
    this._refObject.primary_control_sysid = value;
  }
  get primary_control_compid() {
    if (!this._primary_control_compidIntialized) {
      return undefined;
    }
    return this._refObject.primary_control_compid;
  }
  set primary_control_compid(value) {
    this._primary_control_compidIntialized = true;
    this._refObject.primary_control_compid = value;
  }
  get secondary_control_sysid() {
    if (!this._secondary_control_sysidIntialized) {
      return undefined;
    }
    return this._refObject.secondary_control_sysid;
  }
  set secondary_control_sysid(value) {
    this._secondary_control_sysidIntialized = true;
    this._refObject.secondary_control_sysid = value;
  }
  get secondary_control_compid() {
    if (!this._secondary_control_compidIntialized) {
      return undefined;
    }
    return this._refObject.secondary_control_compid;
  }
  set secondary_control_compid(value) {
    this._secondary_control_compidIntialized = true;
    this._refObject.secondary_control_compid = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GimbalManagerStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._flagsIntialized = true;
    this._gimbal_device_idIntialized = true;
    this._primary_control_sysidIntialized = true;
    this._primary_control_compidIntialized = true;
    this._secondary_control_sysidIntialized = true;
    this._secondary_control_compidIntialized = true;
  }
  copy(other) {
    this._refObject = new GimbalManagerStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._flagsIntialized = true;
    this._gimbal_device_idIntialized = true;
    this._primary_control_sysidIntialized = true;
    this._primary_control_compidIntialized = true;
    this._secondary_control_sysidIntialized = true;
    this._secondary_control_compidIntialized = true;
  }
  static get classType() {
    return GimbalManagerStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flags","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gimbal_device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"primary_control_sysid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"primary_control_compid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"secondary_control_sysid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"secondary_control_compid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"GimbalManagerStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GimbalManagerStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","flags","gimbal_device_id","primary_control_sysid","primary_control_compid","secondary_control_sysid","secondary_control_compid"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GimbalManagerStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GimbalManagerStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GimbalManagerStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GimbalManagerStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GimbalManagerStatusRefArray(size);
    this._refObject = new GimbalManagerStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GimbalManagerStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GimbalManagerStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GimbalManagerStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GimbalManagerStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GimbalManagerStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GimbalManagerStatusArrayWrapper;
  }
}
module.exports = GimbalManagerStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gimbal_device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "primary_control_sysid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "primary_control_compid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "secondary_control_sysid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "secondary_control_compid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "GimbalManagerStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GimbalManagerStatus"
}
*/
