// rclnodejs/generated/px4_msgs/px4_msgs__msg__FollowTargetStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const FollowTargetStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  tracked_target_course: primitiveTypes.float32,
  follow_angle: primitiveTypes.float32,
  orbit_angle_setpoint: primitiveTypes.float32,
  angular_rate_setpoint: primitiveTypes.float32,
  desired_position_raw: ArrayType(primitiveTypes.float32, 3),
  in_emergency_ascent: primitiveTypes.bool,
  gimbal_pitch: primitiveTypes.float32,
});
const FollowTargetStatusRefArray = ArrayType(FollowTargetStatusRefStruct);
const FollowTargetStatusRefStructArray = StructType({
  data: FollowTargetStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FollowTargetStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FollowTargetStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._tracked_target_courseIntialized = true;
      this._follow_angleIntialized = true;
      this._orbit_angle_setpointIntialized = true;
      this._angular_rate_setpointIntialized = true;
      this._wrapperFields.desired_position_raw = Float32Wrapper.createArray();
      this._wrapperFields.desired_position_raw.copy(other._wrapperFields.desired_position_raw);
      this._in_emergency_ascentIntialized = true;
      this._gimbal_pitchIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FollowTargetStatusRefStruct();
    this._timestampIntialized = false;
    this._tracked_target_courseIntialized = false;
    this._follow_angleIntialized = false;
    this._orbit_angle_setpointIntialized = false;
    this._angular_rate_setpointIntialized = false;
    this._wrapperFields.desired_position_raw = Float32Wrapper.createArray();
    this._in_emergency_ascentIntialized = false;
    this._gimbal_pitchIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new FollowTargetStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FollowTargetStatusArrayWrapper;
  }
  static get ArrayType() {
    return FollowTargetStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return FollowTargetStatusRefStructArray
  }
  static get refObjectType() {
    return FollowTargetStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in FollowTargetStatus');
      }
      if (checkConsistency && !this._tracked_target_courseIntialized) {
        throw new TypeError('Invalid argument: tracked_target_course in FollowTargetStatus');
      }
      if (checkConsistency && !this._follow_angleIntialized) {
        throw new TypeError('Invalid argument: follow_angle in FollowTargetStatus');
      }
      if (checkConsistency && !this._orbit_angle_setpointIntialized) {
        throw new TypeError('Invalid argument: orbit_angle_setpoint in FollowTargetStatus');
      }
      if (checkConsistency && !this._angular_rate_setpointIntialized) {
        throw new TypeError('Invalid argument: angular_rate_setpoint in FollowTargetStatus');
      }
      if (checkConsistency && !this._in_emergency_ascentIntialized) {
        throw new TypeError('Invalid argument: in_emergency_ascent in FollowTargetStatus');
      }
      if (checkConsistency && !this._gimbal_pitchIntialized) {
        throw new TypeError('Invalid argument: gimbal_pitch in FollowTargetStatus');
      }
    this._refObject.desired_position_raw = Array.from(this._wrapperFields.desired_position_raw.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._tracked_target_courseIntialized = true;
    this._refObject.tracked_target_course = refObject.tracked_target_course;
    this._follow_angleIntialized = true;
    this._refObject.follow_angle = refObject.follow_angle;
    this._orbit_angle_setpointIntialized = true;
    this._refObject.orbit_angle_setpoint = refObject.orbit_angle_setpoint;
    this._angular_rate_setpointIntialized = true;
    this._refObject.angular_rate_setpoint = refObject.angular_rate_setpoint;
    this._wrapperFields.desired_position_raw.fill(refObject.desired_position_raw.toArray());
    this._in_emergency_ascentIntialized = true;
    this._refObject.in_emergency_ascent = refObject.in_emergency_ascent;
    this._gimbal_pitchIntialized = true;
    this._refObject.gimbal_pitch = refObject.gimbal_pitch;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    FollowTargetStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'FollowTargetStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get tracked_target_course() {
    if (!this._tracked_target_courseIntialized) {
      return undefined;
    }
    return this._refObject.tracked_target_course;
  }
  set tracked_target_course(value) {
    this._tracked_target_courseIntialized = true;
    this._refObject.tracked_target_course = value;
  }
  get follow_angle() {
    if (!this._follow_angleIntialized) {
      return undefined;
    }
    return this._refObject.follow_angle;
  }
  set follow_angle(value) {
    this._follow_angleIntialized = true;
    this._refObject.follow_angle = value;
  }
  get orbit_angle_setpoint() {
    if (!this._orbit_angle_setpointIntialized) {
      return undefined;
    }
    return this._refObject.orbit_angle_setpoint;
  }
  set orbit_angle_setpoint(value) {
    this._orbit_angle_setpointIntialized = true;
    this._refObject.orbit_angle_setpoint = value;
  }
  get angular_rate_setpoint() {
    if (!this._angular_rate_setpointIntialized) {
      return undefined;
    }
    return this._refObject.angular_rate_setpoint;
  }
  set angular_rate_setpoint(value) {
    this._angular_rate_setpointIntialized = true;
    this._refObject.angular_rate_setpoint = value;
  }
  get desired_position_raw() {
    return this._wrapperFields['desired_position_raw'].data;
  }
  set desired_position_raw(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['desired_position_raw'].fill(value);
  }
  get in_emergency_ascent() {
    if (!this._in_emergency_ascentIntialized) {
      return undefined;
    }
    return this._refObject.in_emergency_ascent;
  }
  set in_emergency_ascent(value) {
    this._in_emergency_ascentIntialized = true;
    this._refObject.in_emergency_ascent = value;
  }
  get gimbal_pitch() {
    if (!this._gimbal_pitchIntialized) {
      return undefined;
    }
    return this._refObject.gimbal_pitch;
  }
  set gimbal_pitch(value) {
    this._gimbal_pitchIntialized = true;
    this._refObject.gimbal_pitch = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FollowTargetStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._tracked_target_courseIntialized = true;
    this._follow_angleIntialized = true;
    this._orbit_angle_setpointIntialized = true;
    this._angular_rate_setpointIntialized = true;
    this._wrapperFields.desired_position_raw.fill(refObject.desired_position_raw.toArray());
    this._in_emergency_ascentIntialized = true;
    this._gimbal_pitchIntialized = true;
  }
  copy(other) {
    this._refObject = new FollowTargetStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._tracked_target_courseIntialized = true;
    this._follow_angleIntialized = true;
    this._orbit_angle_setpointIntialized = true;
    this._angular_rate_setpointIntialized = true;
    this._wrapperFields.desired_position_raw.copy(other._wrapperFields.desired_position_raw);
    this._in_emergency_ascentIntialized = true;
    this._gimbal_pitchIntialized = true;
  }
  static get classType() {
    return FollowTargetStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"tracked_target_course","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"follow_angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"orbit_angle_setpoint","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular_rate_setpoint","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"desired_position_raw","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"in_emergency_ascent","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gimbal_pitch","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"FollowTargetStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FollowTargetStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","tracked_target_course","follow_angle","orbit_angle_setpoint","angular_rate_setpoint","desired_position_raw","in_emergency_ascent","gimbal_pitch"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FollowTargetStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FollowTargetStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FollowTargetStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowTargetStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FollowTargetStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FollowTargetStatusRefArray(size);
    this._refObject = new FollowTargetStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FollowTargetStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FollowTargetStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FollowTargetStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FollowTargetStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FollowTargetStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FollowTargetStatusArrayWrapper;
  }
}
module.exports = FollowTargetStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "tracked_target_course",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "follow_angle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "orbit_angle_setpoint",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angular_rate_setpoint",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "desired_position_raw",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "in_emergency_ascent",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gimbal_pitch",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "FollowTargetStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FollowTargetStatus"
}
*/
