// rclnodejs/generated/px4_msgs/px4_msgs__msg__Ping.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PingRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  ping_time: primitiveTypes.uint64,
  ping_sequence: primitiveTypes.uint32,
  dropped_packets: primitiveTypes.uint32,
  rtt_ms: primitiveTypes.float32,
  system_id: primitiveTypes.uint8,
  component_id: primitiveTypes.uint8,
});
const PingRefArray = ArrayType(PingRefStruct);
const PingRefStructArray = StructType({
  data: PingRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PingWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PingRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._ping_timeIntialized = true;
      this._ping_sequenceIntialized = true;
      this._dropped_packetsIntialized = true;
      this._rtt_msIntialized = true;
      this._system_idIntialized = true;
      this._component_idIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PingRefStruct();
    this._timestampIntialized = false;
    this._ping_timeIntialized = false;
    this._ping_sequenceIntialized = false;
    this._dropped_packetsIntialized = false;
    this._rtt_msIntialized = false;
    this._system_idIntialized = false;
    this._component_idIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PingWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PingArrayWrapper;
  }
  static get ArrayType() {
    return PingArrayWrapper;
  }
  static get refObjectArrayType() {
    return PingRefStructArray
  }
  static get refObjectType() {
    return PingRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in Ping');
      }
      if (checkConsistency && !this._ping_timeIntialized) {
        throw new TypeError('Invalid argument: ping_time in Ping');
      }
      if (checkConsistency && !this._ping_sequenceIntialized) {
        throw new TypeError('Invalid argument: ping_sequence in Ping');
      }
      if (checkConsistency && !this._dropped_packetsIntialized) {
        throw new TypeError('Invalid argument: dropped_packets in Ping');
      }
      if (checkConsistency && !this._rtt_msIntialized) {
        throw new TypeError('Invalid argument: rtt_ms in Ping');
      }
      if (checkConsistency && !this._system_idIntialized) {
        throw new TypeError('Invalid argument: system_id in Ping');
      }
      if (checkConsistency && !this._component_idIntialized) {
        throw new TypeError('Invalid argument: component_id in Ping');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._ping_timeIntialized = true;
    this._refObject.ping_time = refObject.ping_time;
    this._ping_sequenceIntialized = true;
    this._refObject.ping_sequence = refObject.ping_sequence;
    this._dropped_packetsIntialized = true;
    this._refObject.dropped_packets = refObject.dropped_packets;
    this._rtt_msIntialized = true;
    this._refObject.rtt_ms = refObject.rtt_ms;
    this._system_idIntialized = true;
    this._refObject.system_id = refObject.system_id;
    this._component_idIntialized = true;
    this._refObject.component_id = refObject.component_id;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PingWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'Ping'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get ping_time() {
    if (!this._ping_timeIntialized) {
      return undefined;
    }
    return this._refObject.ping_time;
  }
  set ping_time(value) {
    this._ping_timeIntialized = true;
    this._refObject.ping_time = value;
  }
  get ping_sequence() {
    if (!this._ping_sequenceIntialized) {
      return undefined;
    }
    return this._refObject.ping_sequence;
  }
  set ping_sequence(value) {
    this._ping_sequenceIntialized = true;
    this._refObject.ping_sequence = value;
  }
  get dropped_packets() {
    if (!this._dropped_packetsIntialized) {
      return undefined;
    }
    return this._refObject.dropped_packets;
  }
  set dropped_packets(value) {
    this._dropped_packetsIntialized = true;
    this._refObject.dropped_packets = value;
  }
  get rtt_ms() {
    if (!this._rtt_msIntialized) {
      return undefined;
    }
    return this._refObject.rtt_ms;
  }
  set rtt_ms(value) {
    this._rtt_msIntialized = true;
    this._refObject.rtt_ms = value;
  }
  get system_id() {
    if (!this._system_idIntialized) {
      return undefined;
    }
    return this._refObject.system_id;
  }
  set system_id(value) {
    this._system_idIntialized = true;
    this._refObject.system_id = value;
  }
  get component_id() {
    if (!this._component_idIntialized) {
      return undefined;
    }
    return this._refObject.component_id;
  }
  set component_id(value) {
    this._component_idIntialized = true;
    this._refObject.component_id = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PingRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._ping_timeIntialized = true;
    this._ping_sequenceIntialized = true;
    this._dropped_packetsIntialized = true;
    this._rtt_msIntialized = true;
    this._system_idIntialized = true;
    this._component_idIntialized = true;
  }
  copy(other) {
    this._refObject = new PingRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._ping_timeIntialized = true;
    this._ping_sequenceIntialized = true;
    this._dropped_packetsIntialized = true;
    this._rtt_msIntialized = true;
    this._system_idIntialized = true;
    this._component_idIntialized = true;
  }
  static get classType() {
    return PingWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ping_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ping_sequence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"dropped_packets","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rtt_ms","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"system_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"component_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"Ping","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Ping"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","ping_time","ping_sequence","dropped_packets","rtt_ms","system_id","component_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PingArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PingWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PingWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PingArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PingArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PingRefArray(size);
    this._refObject = new PingRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PingWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PingArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PingArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PingWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PingWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PingArrayWrapper;
  }
}
module.exports = PingWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ping_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ping_sequence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "dropped_packets",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rtt_ms",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "system_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "component_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "Ping",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Ping"
}
*/
