// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleTrajectoryWaypoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TrajectoryWaypointWrapper = require('../../generated/px4_msgs/px4_msgs__msg__TrajectoryWaypoint.js');
const VehicleTrajectoryWaypointRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  type: primitiveTypes.uint8,
  waypoints: ArrayType(TrajectoryWaypointWrapper.refObjectType, 5),
});
const VehicleTrajectoryWaypointRefArray = ArrayType(VehicleTrajectoryWaypointRefStruct);
const VehicleTrajectoryWaypointRefStructArray = StructType({
  data: VehicleTrajectoryWaypointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleTrajectoryWaypointWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleTrajectoryWaypointRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._typeIntialized = true;
      this._wrapperFields.waypoints = TrajectoryWaypointWrapper.createArray();
      this._wrapperFields.waypoints.copy(other._wrapperFields.waypoints);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleTrajectoryWaypointRefStruct();
    this._timestampIntialized = false;
    this._typeIntialized = false;
    this._wrapperFields.waypoints = TrajectoryWaypointWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new VehicleTrajectoryWaypointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleTrajectoryWaypointArrayWrapper;
  }
  static get ArrayType() {
    return VehicleTrajectoryWaypointArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleTrajectoryWaypointRefStructArray
  }
  static get refObjectType() {
    return VehicleTrajectoryWaypointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleTrajectoryWaypoint');
      }
      if (checkConsistency && !this._typeIntialized) {
        throw new TypeError('Invalid argument: type in VehicleTrajectoryWaypoint');
      }
    for (let i = 0; i < 5; i++) {
      if (this._wrapperFields.waypoints.data[i]) {
        this._wrapperFields.waypoints.data[i].freeze(own, checkConsistency);
        this._refObject.waypoints[i] = this._wrapperFields.waypoints.data[i].refObject;
      }
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
    this._refObject.waypoints = refObject.waypoints;
    this._wrapperFields.waypoints.size = 5
    for (let i = 0; i < 5; i++) {
      this._wrapperFields.waypoints.data[i].copyRefObject(refObject.waypoints[i]);
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    for (let i = 0; i < 5; i++) {
      TrajectoryWaypointWrapper.freeStruct(refObject.waypoints[i]);
    }
  }
  static destoryRawROS(msg) {
    VehicleTrajectoryWaypointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleTrajectoryWaypoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  get waypoints() {
    return this._wrapperFields.waypoints;
  }
  set waypoints(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields.waypoints.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleTrajectoryWaypointRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._typeIntialized = true;
    this._refObject.waypoints = refObject.waypoints;
    this._wrapperFields.waypoints.size = 5
    for (let i = 0; i < 5; i++) {
      this._wrapperFields.waypoints.data[i].copyRefObject(refObject.waypoints[i]);
    }
  }
  copy(other) {
    this._refObject = new VehicleTrajectoryWaypointRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._typeIntialized = true;
    this._wrapperFields.waypoints.copy(other._wrapperFields.waypoints);
  }
  static get classType() {
    return VehicleTrajectoryWaypointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"MAV_TRAJECTORY_REPRESENTATION_WAYPOINTS","value":0},{"type":"uint8","name":"POINT_0","value":0},{"type":"uint8","name":"POINT_1","value":1},{"type":"uint8","name":"POINT_2","value":2},{"type":"uint8","name":"POINT_3","value":3},{"type":"uint8","name":"POINT_4","value":4},{"type":"uint8","name":"NUMBER_POINTS","value":5}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"waypoints","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":"px4_msgs","type":"TrajectoryWaypoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleTrajectoryWaypoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleTrajectoryWaypoint"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","type","waypoints"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleTrajectoryWaypointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleTrajectoryWaypointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleTrajectoryWaypointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleTrajectoryWaypointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleTrajectoryWaypointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleTrajectoryWaypointRefArray(size);
    this._refObject = new VehicleTrajectoryWaypointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleTrajectoryWaypointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleTrajectoryWaypointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleTrajectoryWaypointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleTrajectoryWaypointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleTrajectoryWaypointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleTrajectoryWaypointArrayWrapper;
  }
}
// Define constants (7 in total)
Object.defineProperty(VehicleTrajectoryWaypointWrapper, "MAV_TRAJECTORY_REPRESENTATION_WAYPOINTS", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryWaypointWrapper, "POINT_0", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryWaypointWrapper, "POINT_1", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryWaypointWrapper, "POINT_2", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryWaypointWrapper, "POINT_3", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryWaypointWrapper, "POINT_4", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryWaypointWrapper, "NUMBER_POINTS", {value: 5, writable: false, enumerable: true, configurable: true});
module.exports = VehicleTrajectoryWaypointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "MAV_TRAJECTORY_REPRESENTATION_WAYPOINTS",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "POINT_0",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "POINT_1",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "POINT_2",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "POINT_3",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "POINT_4",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "NUMBER_POINTS",
      "value": 5
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "waypoints",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": "px4_msgs",
        "type": "TrajectoryWaypoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleTrajectoryWaypoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleTrajectoryWaypoint"
}
*/
