// rclnodejs/generated/px4_msgs/px4_msgs__msg__SatelliteInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const SatelliteInfoRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  count: primitiveTypes.uint8,
  svid: ArrayType(primitiveTypes.uint8, 20),
  used: ArrayType(primitiveTypes.uint8, 20),
  elevation: ArrayType(primitiveTypes.uint8, 20),
  azimuth: ArrayType(primitiveTypes.uint8, 20),
  snr: ArrayType(primitiveTypes.uint8, 20),
  prn: ArrayType(primitiveTypes.uint8, 20),
});
const SatelliteInfoRefArray = ArrayType(SatelliteInfoRefStruct);
const SatelliteInfoRefStructArray = StructType({
  data: SatelliteInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SatelliteInfoWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SatelliteInfoRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._countIntialized = true;
      this._wrapperFields.svid = UInt8Wrapper.createArray();
      this._wrapperFields.svid.copy(other._wrapperFields.svid);
      this._wrapperFields.used = UInt8Wrapper.createArray();
      this._wrapperFields.used.copy(other._wrapperFields.used);
      this._wrapperFields.elevation = UInt8Wrapper.createArray();
      this._wrapperFields.elevation.copy(other._wrapperFields.elevation);
      this._wrapperFields.azimuth = UInt8Wrapper.createArray();
      this._wrapperFields.azimuth.copy(other._wrapperFields.azimuth);
      this._wrapperFields.snr = UInt8Wrapper.createArray();
      this._wrapperFields.snr.copy(other._wrapperFields.snr);
      this._wrapperFields.prn = UInt8Wrapper.createArray();
      this._wrapperFields.prn.copy(other._wrapperFields.prn);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SatelliteInfoRefStruct();
    this._timestampIntialized = false;
    this._countIntialized = false;
    this._wrapperFields.svid = UInt8Wrapper.createArray();
    this._wrapperFields.used = UInt8Wrapper.createArray();
    this._wrapperFields.elevation = UInt8Wrapper.createArray();
    this._wrapperFields.azimuth = UInt8Wrapper.createArray();
    this._wrapperFields.snr = UInt8Wrapper.createArray();
    this._wrapperFields.prn = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SatelliteInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SatelliteInfoArrayWrapper;
  }
  static get ArrayType() {
    return SatelliteInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return SatelliteInfoRefStructArray
  }
  static get refObjectType() {
    return SatelliteInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in SatelliteInfo');
      }
      if (checkConsistency && !this._countIntialized) {
        throw new TypeError('Invalid argument: count in SatelliteInfo');
      }
    this._refObject.svid = Array.from(this._wrapperFields.svid.data);
    this._refObject.used = Array.from(this._wrapperFields.used.data);
    this._refObject.elevation = Array.from(this._wrapperFields.elevation.data);
    this._refObject.azimuth = Array.from(this._wrapperFields.azimuth.data);
    this._refObject.snr = Array.from(this._wrapperFields.snr.data);
    this._refObject.prn = Array.from(this._wrapperFields.prn.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._countIntialized = true;
    this._refObject.count = refObject.count;
    this._wrapperFields.svid.fill(refObject.svid.toArray());
    this._wrapperFields.used.fill(refObject.used.toArray());
    this._wrapperFields.elevation.fill(refObject.elevation.toArray());
    this._wrapperFields.azimuth.fill(refObject.azimuth.toArray());
    this._wrapperFields.snr.fill(refObject.snr.toArray());
    this._wrapperFields.prn.fill(refObject.prn.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SatelliteInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'SatelliteInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get count() {
    if (!this._countIntialized) {
      return undefined;
    }
    return this._refObject.count;
  }
  set count(value) {
    this._countIntialized = true;
    this._refObject.count = value;
  }
  get svid() {
    return this._wrapperFields['svid'].data;
  }
  set svid(value) {
    if (value.length !== 20) {
      throw new RangeError('The length of the array must be 20.');
    }
    this._wrapperFields['svid'].fill(value);
  }
  get used() {
    return this._wrapperFields['used'].data;
  }
  set used(value) {
    if (value.length !== 20) {
      throw new RangeError('The length of the array must be 20.');
    }
    this._wrapperFields['used'].fill(value);
  }
  get elevation() {
    return this._wrapperFields['elevation'].data;
  }
  set elevation(value) {
    if (value.length !== 20) {
      throw new RangeError('The length of the array must be 20.');
    }
    this._wrapperFields['elevation'].fill(value);
  }
  get azimuth() {
    return this._wrapperFields['azimuth'].data;
  }
  set azimuth(value) {
    if (value.length !== 20) {
      throw new RangeError('The length of the array must be 20.');
    }
    this._wrapperFields['azimuth'].fill(value);
  }
  get snr() {
    return this._wrapperFields['snr'].data;
  }
  set snr(value) {
    if (value.length !== 20) {
      throw new RangeError('The length of the array must be 20.');
    }
    this._wrapperFields['snr'].fill(value);
  }
  get prn() {
    return this._wrapperFields['prn'].data;
  }
  set prn(value) {
    if (value.length !== 20) {
      throw new RangeError('The length of the array must be 20.');
    }
    this._wrapperFields['prn'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SatelliteInfoRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._countIntialized = true;
    this._wrapperFields.svid.fill(refObject.svid.toArray());
    this._wrapperFields.used.fill(refObject.used.toArray());
    this._wrapperFields.elevation.fill(refObject.elevation.toArray());
    this._wrapperFields.azimuth.fill(refObject.azimuth.toArray());
    this._wrapperFields.snr.fill(refObject.snr.toArray());
    this._wrapperFields.prn.fill(refObject.prn.toArray());
  }
  copy(other) {
    this._refObject = new SatelliteInfoRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._countIntialized = true;
    this._wrapperFields.svid.copy(other._wrapperFields.svid);
    this._wrapperFields.used.copy(other._wrapperFields.used);
    this._wrapperFields.elevation.copy(other._wrapperFields.elevation);
    this._wrapperFields.azimuth.copy(other._wrapperFields.azimuth);
    this._wrapperFields.snr.copy(other._wrapperFields.snr);
    this._wrapperFields.prn.copy(other._wrapperFields.prn);
  }
  static get classType() {
    return SatelliteInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"SAT_INFO_MAX_SATELLITES","value":20}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"svid","type":{"isArray":true,"arraySize":20,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"used","type":{"isArray":true,"arraySize":20,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"elevation","type":{"isArray":true,"arraySize":20,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"azimuth","type":{"isArray":true,"arraySize":20,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"snr","type":{"isArray":true,"arraySize":20,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"prn","type":{"isArray":true,"arraySize":20,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"SatelliteInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SatelliteInfo"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","count","svid","used","elevation","azimuth","snr","prn"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SatelliteInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SatelliteInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SatelliteInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SatelliteInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SatelliteInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SatelliteInfoRefArray(size);
    this._refObject = new SatelliteInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SatelliteInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SatelliteInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SatelliteInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SatelliteInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SatelliteInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SatelliteInfoArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(SatelliteInfoWrapper, "SAT_INFO_MAX_SATELLITES", {value: 20, writable: false, enumerable: true, configurable: true});
module.exports = SatelliteInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "SAT_INFO_MAX_SATELLITES",
      "value": 20
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "svid",
      "type": {
        "isArray": true,
        "arraySize": 20,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "used",
      "type": {
        "isArray": true,
        "arraySize": 20,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "elevation",
      "type": {
        "isArray": true,
        "arraySize": 20,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "azimuth",
      "type": {
        "isArray": true,
        "arraySize": 20,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "snr",
      "type": {
        "isArray": true,
        "arraySize": 20,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "prn",
      "type": {
        "isArray": true,
        "arraySize": 20,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "SatelliteInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SatelliteInfo"
}
*/
