// rclnodejs/generated/px4_msgs/px4_msgs__msg__OrbitStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const OrbitStatusRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  radius: primitiveTypes.float32,
  frame: primitiveTypes.uint8,
  x: primitiveTypes.float64,
  y: primitiveTypes.float64,
  z: primitiveTypes.float32,
  yaw_behaviour: primitiveTypes.uint8,
});
const OrbitStatusRefArray = ArrayType(OrbitStatusRefStruct);
const OrbitStatusRefStructArray = StructType({
  data: OrbitStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OrbitStatusWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OrbitStatusRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._radiusIntialized = true;
      this._frameIntialized = true;
      this._xIntialized = true;
      this._yIntialized = true;
      this._zIntialized = true;
      this._yaw_behaviourIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OrbitStatusRefStruct();
    this._timestampIntialized = false;
    this._radiusIntialized = false;
    this._frameIntialized = false;
    this._xIntialized = false;
    this._yIntialized = false;
    this._zIntialized = false;
    this._yaw_behaviourIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new OrbitStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OrbitStatusArrayWrapper;
  }
  static get ArrayType() {
    return OrbitStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return OrbitStatusRefStructArray
  }
  static get refObjectType() {
    return OrbitStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in OrbitStatus');
      }
      if (checkConsistency && !this._radiusIntialized) {
        throw new TypeError('Invalid argument: radius in OrbitStatus');
      }
      if (checkConsistency && !this._frameIntialized) {
        throw new TypeError('Invalid argument: frame in OrbitStatus');
      }
      if (checkConsistency && !this._xIntialized) {
        throw new TypeError('Invalid argument: x in OrbitStatus');
      }
      if (checkConsistency && !this._yIntialized) {
        throw new TypeError('Invalid argument: y in OrbitStatus');
      }
      if (checkConsistency && !this._zIntialized) {
        throw new TypeError('Invalid argument: z in OrbitStatus');
      }
      if (checkConsistency && !this._yaw_behaviourIntialized) {
        throw new TypeError('Invalid argument: yaw_behaviour in OrbitStatus');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._radiusIntialized = true;
    this._refObject.radius = refObject.radius;
    this._frameIntialized = true;
    this._refObject.frame = refObject.frame;
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._zIntialized = true;
    this._refObject.z = refObject.z;
    this._yaw_behaviourIntialized = true;
    this._refObject.yaw_behaviour = refObject.yaw_behaviour;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    OrbitStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'OrbitStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get radius() {
    if (!this._radiusIntialized) {
      return undefined;
    }
    return this._refObject.radius;
  }
  set radius(value) {
    this._radiusIntialized = true;
    this._refObject.radius = value;
  }
  get frame() {
    if (!this._frameIntialized) {
      return undefined;
    }
    return this._refObject.frame;
  }
  set frame(value) {
    this._frameIntialized = true;
    this._refObject.frame = value;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get z() {
    if (!this._zIntialized) {
      return undefined;
    }
    return this._refObject.z;
  }
  set z(value) {
    this._zIntialized = true;
    this._refObject.z = value;
  }
  get yaw_behaviour() {
    if (!this._yaw_behaviourIntialized) {
      return undefined;
    }
    return this._refObject.yaw_behaviour;
  }
  set yaw_behaviour(value) {
    this._yaw_behaviourIntialized = true;
    this._refObject.yaw_behaviour = value;
  }
  copyRefObject(refObject) {
    this._refObject = new OrbitStatusRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._radiusIntialized = true;
    this._frameIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._yaw_behaviourIntialized = true;
  }
  copy(other) {
    this._refObject = new OrbitStatusRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._radiusIntialized = true;
    this._frameIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._yaw_behaviourIntialized = true;
  }
  static get classType() {
    return OrbitStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"ORBIT_YAW_BEHAVIOUR_HOLD_FRONT_TO_CIRCLE_CENTER","value":0},{"type":"uint8","name":"ORBIT_YAW_BEHAVIOUR_HOLD_INITIAL_HEADING","value":1},{"type":"uint8","name":"ORBIT_YAW_BEHAVIOUR_UNCONTROLLED","value":2},{"type":"uint8","name":"ORBIT_YAW_BEHAVIOUR_HOLD_FRONT_TANGENT_TO_CIRCLE","value":3},{"type":"uint8","name":"ORBIT_YAW_BEHAVIOUR_RC_CONTROLLED","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"radius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"yaw_behaviour","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"OrbitStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OrbitStatus"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","radius","frame","x","y","z","yaw_behaviour"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OrbitStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OrbitStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OrbitStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OrbitStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OrbitStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OrbitStatusRefArray(size);
    this._refObject = new OrbitStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OrbitStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OrbitStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OrbitStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OrbitStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OrbitStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OrbitStatusArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(OrbitStatusWrapper, "ORBIT_YAW_BEHAVIOUR_HOLD_FRONT_TO_CIRCLE_CENTER", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OrbitStatusWrapper, "ORBIT_YAW_BEHAVIOUR_HOLD_INITIAL_HEADING", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OrbitStatusWrapper, "ORBIT_YAW_BEHAVIOUR_UNCONTROLLED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OrbitStatusWrapper, "ORBIT_YAW_BEHAVIOUR_HOLD_FRONT_TANGENT_TO_CIRCLE", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(OrbitStatusWrapper, "ORBIT_YAW_BEHAVIOUR_RC_CONTROLLED", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = OrbitStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "ORBIT_YAW_BEHAVIOUR_HOLD_FRONT_TO_CIRCLE_CENTER",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "ORBIT_YAW_BEHAVIOUR_HOLD_INITIAL_HEADING",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "ORBIT_YAW_BEHAVIOUR_UNCONTROLLED",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "ORBIT_YAW_BEHAVIOUR_HOLD_FRONT_TANGENT_TO_CIRCLE",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "ORBIT_YAW_BEHAVIOUR_RC_CONTROLLED",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "radius",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "yaw_behaviour",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "OrbitStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OrbitStatus"
}
*/
