// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleCommand.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const VehicleCommandRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  param1: primitiveTypes.float32,
  param2: primitiveTypes.float32,
  param3: primitiveTypes.float32,
  param4: primitiveTypes.float32,
  param5: primitiveTypes.float64,
  param6: primitiveTypes.float64,
  param7: primitiveTypes.float32,
  command: primitiveTypes.uint32,
  target_system: primitiveTypes.uint8,
  target_component: primitiveTypes.uint8,
  source_system: primitiveTypes.uint8,
  source_component: primitiveTypes.uint8,
  confirmation: primitiveTypes.uint8,
  from_external: primitiveTypes.bool,
});
const VehicleCommandRefArray = ArrayType(VehicleCommandRefStruct);
const VehicleCommandRefStructArray = StructType({
  data: VehicleCommandRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleCommandWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleCommandRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._param1Intialized = true;
      this._param2Intialized = true;
      this._param3Intialized = true;
      this._param4Intialized = true;
      this._param5Intialized = true;
      this._param6Intialized = true;
      this._param7Intialized = true;
      this._commandIntialized = true;
      this._target_systemIntialized = true;
      this._target_componentIntialized = true;
      this._source_systemIntialized = true;
      this._source_componentIntialized = true;
      this._confirmationIntialized = true;
      this._from_externalIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleCommandRefStruct();
    this._timestampIntialized = false;
    this._param1Intialized = false;
    this._param2Intialized = false;
    this._param3Intialized = false;
    this._param4Intialized = false;
    this._param5Intialized = false;
    this._param6Intialized = false;
    this._param7Intialized = false;
    this._commandIntialized = false;
    this._target_systemIntialized = false;
    this._target_componentIntialized = false;
    this._source_systemIntialized = false;
    this._source_componentIntialized = false;
    this._confirmationIntialized = false;
    this._from_externalIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleCommandWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleCommandArrayWrapper;
  }
  static get ArrayType() {
    return VehicleCommandArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleCommandRefStructArray
  }
  static get refObjectType() {
    return VehicleCommandRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleCommand');
      }
      if (checkConsistency && !this._param1Intialized) {
        throw new TypeError('Invalid argument: param1 in VehicleCommand');
      }
      if (checkConsistency && !this._param2Intialized) {
        throw new TypeError('Invalid argument: param2 in VehicleCommand');
      }
      if (checkConsistency && !this._param3Intialized) {
        throw new TypeError('Invalid argument: param3 in VehicleCommand');
      }
      if (checkConsistency && !this._param4Intialized) {
        throw new TypeError('Invalid argument: param4 in VehicleCommand');
      }
      if (checkConsistency && !this._param5Intialized) {
        throw new TypeError('Invalid argument: param5 in VehicleCommand');
      }
      if (checkConsistency && !this._param6Intialized) {
        throw new TypeError('Invalid argument: param6 in VehicleCommand');
      }
      if (checkConsistency && !this._param7Intialized) {
        throw new TypeError('Invalid argument: param7 in VehicleCommand');
      }
      if (checkConsistency && !this._commandIntialized) {
        throw new TypeError('Invalid argument: command in VehicleCommand');
      }
      if (checkConsistency && !this._target_systemIntialized) {
        throw new TypeError('Invalid argument: target_system in VehicleCommand');
      }
      if (checkConsistency && !this._target_componentIntialized) {
        throw new TypeError('Invalid argument: target_component in VehicleCommand');
      }
      if (checkConsistency && !this._source_systemIntialized) {
        throw new TypeError('Invalid argument: source_system in VehicleCommand');
      }
      if (checkConsistency && !this._source_componentIntialized) {
        throw new TypeError('Invalid argument: source_component in VehicleCommand');
      }
      if (checkConsistency && !this._confirmationIntialized) {
        throw new TypeError('Invalid argument: confirmation in VehicleCommand');
      }
      if (checkConsistency && !this._from_externalIntialized) {
        throw new TypeError('Invalid argument: from_external in VehicleCommand');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._param1Intialized = true;
    this._refObject.param1 = refObject.param1;
    this._param2Intialized = true;
    this._refObject.param2 = refObject.param2;
    this._param3Intialized = true;
    this._refObject.param3 = refObject.param3;
    this._param4Intialized = true;
    this._refObject.param4 = refObject.param4;
    this._param5Intialized = true;
    this._refObject.param5 = refObject.param5;
    this._param6Intialized = true;
    this._refObject.param6 = refObject.param6;
    this._param7Intialized = true;
    this._refObject.param7 = refObject.param7;
    this._commandIntialized = true;
    this._refObject.command = refObject.command;
    this._target_systemIntialized = true;
    this._refObject.target_system = refObject.target_system;
    this._target_componentIntialized = true;
    this._refObject.target_component = refObject.target_component;
    this._source_systemIntialized = true;
    this._refObject.source_system = refObject.source_system;
    this._source_componentIntialized = true;
    this._refObject.source_component = refObject.source_component;
    this._confirmationIntialized = true;
    this._refObject.confirmation = refObject.confirmation;
    this._from_externalIntialized = true;
    this._refObject.from_external = refObject.from_external;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleCommandWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleCommand'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get param1() {
    if (!this._param1Intialized) {
      return undefined;
    }
    return this._refObject.param1;
  }
  set param1(value) {
    this._param1Intialized = true;
    this._refObject.param1 = value;
  }
  get param2() {
    if (!this._param2Intialized) {
      return undefined;
    }
    return this._refObject.param2;
  }
  set param2(value) {
    this._param2Intialized = true;
    this._refObject.param2 = value;
  }
  get param3() {
    if (!this._param3Intialized) {
      return undefined;
    }
    return this._refObject.param3;
  }
  set param3(value) {
    this._param3Intialized = true;
    this._refObject.param3 = value;
  }
  get param4() {
    if (!this._param4Intialized) {
      return undefined;
    }
    return this._refObject.param4;
  }
  set param4(value) {
    this._param4Intialized = true;
    this._refObject.param4 = value;
  }
  get param5() {
    if (!this._param5Intialized) {
      return undefined;
    }
    return this._refObject.param5;
  }
  set param5(value) {
    this._param5Intialized = true;
    this._refObject.param5 = value;
  }
  get param6() {
    if (!this._param6Intialized) {
      return undefined;
    }
    return this._refObject.param6;
  }
  set param6(value) {
    this._param6Intialized = true;
    this._refObject.param6 = value;
  }
  get param7() {
    if (!this._param7Intialized) {
      return undefined;
    }
    return this._refObject.param7;
  }
  set param7(value) {
    this._param7Intialized = true;
    this._refObject.param7 = value;
  }
  get command() {
    if (!this._commandIntialized) {
      return undefined;
    }
    return this._refObject.command;
  }
  set command(value) {
    this._commandIntialized = true;
    this._refObject.command = value;
  }
  get target_system() {
    if (!this._target_systemIntialized) {
      return undefined;
    }
    return this._refObject.target_system;
  }
  set target_system(value) {
    this._target_systemIntialized = true;
    this._refObject.target_system = value;
  }
  get target_component() {
    if (!this._target_componentIntialized) {
      return undefined;
    }
    return this._refObject.target_component;
  }
  set target_component(value) {
    this._target_componentIntialized = true;
    this._refObject.target_component = value;
  }
  get source_system() {
    if (!this._source_systemIntialized) {
      return undefined;
    }
    return this._refObject.source_system;
  }
  set source_system(value) {
    this._source_systemIntialized = true;
    this._refObject.source_system = value;
  }
  get source_component() {
    if (!this._source_componentIntialized) {
      return undefined;
    }
    return this._refObject.source_component;
  }
  set source_component(value) {
    this._source_componentIntialized = true;
    this._refObject.source_component = value;
  }
  get confirmation() {
    if (!this._confirmationIntialized) {
      return undefined;
    }
    return this._refObject.confirmation;
  }
  set confirmation(value) {
    this._confirmationIntialized = true;
    this._refObject.confirmation = value;
  }
  get from_external() {
    if (!this._from_externalIntialized) {
      return undefined;
    }
    return this._refObject.from_external;
  }
  set from_external(value) {
    this._from_externalIntialized = true;
    this._refObject.from_external = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleCommandRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._param1Intialized = true;
    this._param2Intialized = true;
    this._param3Intialized = true;
    this._param4Intialized = true;
    this._param5Intialized = true;
    this._param6Intialized = true;
    this._param7Intialized = true;
    this._commandIntialized = true;
    this._target_systemIntialized = true;
    this._target_componentIntialized = true;
    this._source_systemIntialized = true;
    this._source_componentIntialized = true;
    this._confirmationIntialized = true;
    this._from_externalIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleCommandRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._param1Intialized = true;
    this._param2Intialized = true;
    this._param3Intialized = true;
    this._param4Intialized = true;
    this._param5Intialized = true;
    this._param6Intialized = true;
    this._param7Intialized = true;
    this._commandIntialized = true;
    this._target_systemIntialized = true;
    this._target_componentIntialized = true;
    this._source_systemIntialized = true;
    this._source_componentIntialized = true;
    this._confirmationIntialized = true;
    this._from_externalIntialized = true;
  }
  static get classType() {
    return VehicleCommandWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint16","name":"VEHICLE_CMD_CUSTOM_0","value":0},{"type":"uint16","name":"VEHICLE_CMD_CUSTOM_1","value":1},{"type":"uint16","name":"VEHICLE_CMD_CUSTOM_2","value":2},{"type":"uint16","name":"VEHICLE_CMD_NAV_WAYPOINT","value":16},{"type":"uint16","name":"VEHICLE_CMD_NAV_LOITER_UNLIM","value":17},{"type":"uint16","name":"VEHICLE_CMD_NAV_LOITER_TURNS","value":18},{"type":"uint16","name":"VEHICLE_CMD_NAV_LOITER_TIME","value":19},{"type":"uint16","name":"VEHICLE_CMD_NAV_RETURN_TO_LAUNCH","value":20},{"type":"uint16","name":"VEHICLE_CMD_NAV_LAND","value":21},{"type":"uint16","name":"VEHICLE_CMD_NAV_TAKEOFF","value":22},{"type":"uint16","name":"VEHICLE_CMD_NAV_PRECLAND","value":23},{"type":"uint16","name":"VEHICLE_CMD_DO_ORBIT","value":34},{"type":"uint16","name":"VEHICLE_CMD_NAV_ROI","value":80},{"type":"uint16","name":"VEHICLE_CMD_NAV_PATHPLANNING","value":81},{"type":"uint16","name":"VEHICLE_CMD_NAV_VTOL_TAKEOFF","value":84},{"type":"uint16","name":"VEHICLE_CMD_NAV_VTOL_LAND","value":85},{"type":"uint16","name":"VEHICLE_CMD_NAV_GUIDED_LIMITS","value":90},{"type":"uint16","name":"VEHICLE_CMD_NAV_GUIDED_MASTER","value":91},{"type":"uint16","name":"VEHICLE_CMD_NAV_DELAY","value":93},{"type":"uint16","name":"VEHICLE_CMD_NAV_LAST","value":95},{"type":"uint16","name":"VEHICLE_CMD_CONDITION_DELAY","value":112},{"type":"uint16","name":"VEHICLE_CMD_CONDITION_CHANGE_ALT","value":113},{"type":"uint16","name":"VEHICLE_CMD_CONDITION_DISTANCE","value":114},{"type":"uint16","name":"VEHICLE_CMD_CONDITION_YAW","value":115},{"type":"uint16","name":"VEHICLE_CMD_CONDITION_LAST","value":159},{"type":"uint16","name":"VEHICLE_CMD_CONDITION_GATE","value":4501},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_MODE","value":176},{"type":"uint16","name":"VEHICLE_CMD_DO_JUMP","value":177},{"type":"uint16","name":"VEHICLE_CMD_DO_CHANGE_SPEED","value":178},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_HOME","value":179},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_PARAMETER","value":180},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_RELAY","value":181},{"type":"uint16","name":"VEHICLE_CMD_DO_REPEAT_RELAY","value":182},{"type":"uint16","name":"VEHICLE_CMD_DO_REPEAT_SERVO","value":184},{"type":"uint16","name":"VEHICLE_CMD_DO_FLIGHTTERMINATION","value":185},{"type":"uint16","name":"VEHICLE_CMD_DO_CHANGE_ALTITUDE","value":186},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_ACTUATOR","value":187},{"type":"uint16","name":"VEHICLE_CMD_DO_LAND_START","value":189},{"type":"uint16","name":"VEHICLE_CMD_DO_GO_AROUND","value":191},{"type":"uint16","name":"VEHICLE_CMD_DO_REPOSITION","value":192},{"type":"uint16","name":"VEHICLE_CMD_DO_PAUSE_CONTINUE","value":193},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_ROI_LOCATION","value":195},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_ROI_WPNEXT_OFFSET","value":196},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_ROI_NONE","value":197},{"type":"uint16","name":"VEHICLE_CMD_DO_CONTROL_VIDEO","value":200},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_ROI","value":201},{"type":"uint16","name":"VEHICLE_CMD_DO_DIGICAM_CONTROL","value":203},{"type":"uint16","name":"VEHICLE_CMD_DO_MOUNT_CONFIGURE","value":204},{"type":"uint16","name":"VEHICLE_CMD_DO_MOUNT_CONTROL","value":205},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_CAM_TRIGG_DIST","value":206},{"type":"uint16","name":"VEHICLE_CMD_DO_FENCE_ENABLE","value":207},{"type":"uint16","name":"VEHICLE_CMD_DO_PARACHUTE","value":208},{"type":"uint16","name":"VEHICLE_CMD_DO_MOTOR_TEST","value":209},{"type":"uint16","name":"VEHICLE_CMD_DO_INVERTED_FLIGHT","value":210},{"type":"uint16","name":"VEHICLE_CMD_DO_GRIPPER","value":211},{"type":"uint16","name":"VEHICLE_CMD_DO_SET_CAM_TRIGG_INTERVAL","value":214},{"type":"uint16","name":"VEHICLE_CMD_DO_MOUNT_CONTROL_QUAT","value":220},{"type":"uint16","name":"VEHICLE_CMD_DO_GUIDED_MASTER","value":221},{"type":"uint16","name":"VEHICLE_CMD_DO_GUIDED_LIMITS","value":222},{"type":"uint16","name":"VEHICLE_CMD_DO_LAST","value":240},{"type":"uint16","name":"VEHICLE_CMD_PREFLIGHT_CALIBRATION","value":241},{"type":"uint16","name":"PREFLIGHT_CALIBRATION_TEMPERATURE_CALIBRATION","value":3},{"type":"uint16","name":"VEHICLE_CMD_PREFLIGHT_SET_SENSOR_OFFSETS","value":242},{"type":"uint16","name":"VEHICLE_CMD_PREFLIGHT_UAVCAN","value":243},{"type":"uint16","name":"VEHICLE_CMD_PREFLIGHT_STORAGE","value":245},{"type":"uint16","name":"VEHICLE_CMD_PREFLIGHT_REBOOT_SHUTDOWN","value":246},{"type":"uint16","name":"VEHICLE_CMD_OBLIQUE_SURVEY","value":260},{"type":"uint16","name":"VEHICLE_CMD_GIMBAL_DEVICE_INFORMATION","value":283},{"type":"uint16","name":"VEHICLE_CMD_MISSION_START","value":300},{"type":"uint16","name":"VEHICLE_CMD_ACTUATOR_TEST","value":310},{"type":"uint16","name":"VEHICLE_CMD_CONFIGURE_ACTUATOR","value":311},{"type":"uint16","name":"VEHICLE_CMD_COMPONENT_ARM_DISARM","value":400},{"type":"uint16","name":"VEHICLE_CMD_RUN_PREARM_CHECKS","value":401},{"type":"uint16","name":"VEHICLE_CMD_INJECT_FAILURE","value":420},{"type":"uint16","name":"VEHICLE_CMD_START_RX_PAIR","value":500},{"type":"uint16","name":"VEHICLE_CMD_REQUEST_MESSAGE","value":512},{"type":"uint16","name":"VEHICLE_CMD_SET_CAMERA_MODE","value":530},{"type":"uint16","name":"VEHICLE_CMD_SET_CAMERA_ZOOM","value":531},{"type":"uint16","name":"VEHICLE_CMD_SET_CAMERA_FOCUS","value":532},{"type":"uint16","name":"VEHICLE_CMD_DO_GIMBAL_MANAGER_PITCHYAW","value":1000},{"type":"uint16","name":"VEHICLE_CMD_DO_GIMBAL_MANAGER_CONFIGURE","value":1001},{"type":"uint16","name":"VEHICLE_CMD_IMAGE_START_CAPTURE","value":2000},{"type":"uint16","name":"VEHICLE_CMD_DO_TRIGGER_CONTROL","value":2003},{"type":"uint16","name":"VEHICLE_CMD_VIDEO_START_CAPTURE","value":2500},{"type":"uint16","name":"VEHICLE_CMD_VIDEO_STOP_CAPTURE","value":2501},{"type":"uint16","name":"VEHICLE_CMD_LOGGING_START","value":2510},{"type":"uint16","name":"VEHICLE_CMD_LOGGING_STOP","value":2511},{"type":"uint16","name":"VEHICLE_CMD_CONTROL_HIGH_LATENCY","value":2600},{"type":"uint16","name":"VEHICLE_CMD_DO_VTOL_TRANSITION","value":3000},{"type":"uint16","name":"VEHICLE_CMD_ARM_AUTHORIZATION_REQUEST","value":3001},{"type":"uint16","name":"VEHICLE_CMD_PAYLOAD_PREPARE_DEPLOY","value":30001},{"type":"uint16","name":"VEHICLE_CMD_PAYLOAD_CONTROL_DEPLOY","value":30002},{"type":"uint16","name":"VEHICLE_CMD_FIXED_MAG_CAL_YAW","value":42006},{"type":"uint16","name":"VEHICLE_CMD_DO_WINCH","value":42600},{"type":"uint32","name":"VEHICLE_CMD_PX4_INTERNAL_START","value":65537},{"type":"uint32","name":"VEHICLE_CMD_SET_GPS_GLOBAL_ORIGIN","value":100000},{"type":"uint8","name":"VEHICLE_MOUNT_MODE_RETRACT","value":0},{"type":"uint8","name":"VEHICLE_MOUNT_MODE_NEUTRAL","value":1},{"type":"uint8","name":"VEHICLE_MOUNT_MODE_MAVLINK_TARGETING","value":2},{"type":"uint8","name":"VEHICLE_MOUNT_MODE_RC_TARGETING","value":3},{"type":"uint8","name":"VEHICLE_MOUNT_MODE_GPS_POINT","value":4},{"type":"uint8","name":"VEHICLE_MOUNT_MODE_ENUM_END","value":5},{"type":"uint8","name":"VEHICLE_ROI_NONE","value":0},{"type":"uint8","name":"VEHICLE_ROI_WPNEXT","value":1},{"type":"uint8","name":"VEHICLE_ROI_WPINDEX","value":2},{"type":"uint8","name":"VEHICLE_ROI_LOCATION","value":3},{"type":"uint8","name":"VEHICLE_ROI_TARGET","value":4},{"type":"uint8","name":"VEHICLE_ROI_ENUM_END","value":5},{"type":"uint8","name":"PARACHUTE_ACTION_DISABLE","value":0},{"type":"uint8","name":"PARACHUTE_ACTION_ENABLE","value":1},{"type":"uint8","name":"PARACHUTE_ACTION_RELEASE","value":2},{"type":"uint8","name":"FAILURE_UNIT_SENSOR_GYRO","value":0},{"type":"uint8","name":"FAILURE_UNIT_SENSOR_ACCEL","value":1},{"type":"uint8","name":"FAILURE_UNIT_SENSOR_MAG","value":2},{"type":"uint8","name":"FAILURE_UNIT_SENSOR_BARO","value":3},{"type":"uint8","name":"FAILURE_UNIT_SENSOR_GPS","value":4},{"type":"uint8","name":"FAILURE_UNIT_SENSOR_OPTICAL_FLOW","value":5},{"type":"uint8","name":"FAILURE_UNIT_SENSOR_VIO","value":6},{"type":"uint8","name":"FAILURE_UNIT_SENSOR_DISTANCE_SENSOR","value":7},{"type":"uint8","name":"FAILURE_UNIT_SENSOR_AIRSPEED","value":8},{"type":"uint8","name":"FAILURE_UNIT_SYSTEM_BATTERY","value":100},{"type":"uint8","name":"FAILURE_UNIT_SYSTEM_MOTOR","value":101},{"type":"uint8","name":"FAILURE_UNIT_SYSTEM_SERVO","value":102},{"type":"uint8","name":"FAILURE_UNIT_SYSTEM_AVOIDANCE","value":103},{"type":"uint8","name":"FAILURE_UNIT_SYSTEM_RC_SIGNAL","value":104},{"type":"uint8","name":"FAILURE_UNIT_SYSTEM_MAVLINK_SIGNAL","value":105},{"type":"uint8","name":"FAILURE_TYPE_OK","value":0},{"type":"uint8","name":"FAILURE_TYPE_OFF","value":1},{"type":"uint8","name":"FAILURE_TYPE_STUCK","value":2},{"type":"uint8","name":"FAILURE_TYPE_GARBAGE","value":3},{"type":"uint8","name":"FAILURE_TYPE_WRONG","value":4},{"type":"uint8","name":"FAILURE_TYPE_SLOW","value":5},{"type":"uint8","name":"FAILURE_TYPE_DELAYED","value":6},{"type":"uint8","name":"FAILURE_TYPE_INTERMITTENT","value":7},{"type":"uint8","name":"SPEED_TYPE_AIRSPEED","value":0},{"type":"uint8","name":"SPEED_TYPE_GROUNDSPEED","value":1},{"type":"uint8","name":"SPEED_TYPE_CLIMB_SPEED","value":2},{"type":"uint8","name":"SPEED_TYPE_DESCEND_SPEED","value":3},{"type":"int8","name":"ARMING_ACTION_DISARM","value":0},{"type":"int8","name":"ARMING_ACTION_ARM","value":1},{"type":"uint8","name":"GRIPPER_ACTION_RELEASE","value":0},{"type":"uint8","name":"GRIPPER_ACTION_GRAB","value":1},{"type":"uint8","name":"ORB_QUEUE_LENGTH","value":8}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param5","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param6","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"param7","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"command","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"target_system","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"target_component","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"source_system","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"source_component","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"confirmation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"from_external","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleCommand","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleCommand"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","param1","param2","param3","param4","param5","param6","param7","command","target_system","target_component","source_system","source_component","confirmation","from_external"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleCommandArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleCommandWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleCommandWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleCommandArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleCommandArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleCommandRefArray(size);
    this._refObject = new VehicleCommandRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleCommandWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleCommandArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleCommandArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleCommandWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleCommandWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleCommandArrayWrapper;
  }
}
// Define constants (143 in total)
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CUSTOM_0", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CUSTOM_1", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CUSTOM_2", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_WAYPOINT", {value: 16, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_LOITER_UNLIM", {value: 17, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_LOITER_TURNS", {value: 18, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_LOITER_TIME", {value: 19, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_RETURN_TO_LAUNCH", {value: 20, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_LAND", {value: 21, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_TAKEOFF", {value: 22, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_PRECLAND", {value: 23, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_ORBIT", {value: 34, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_ROI", {value: 80, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_PATHPLANNING", {value: 81, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_VTOL_TAKEOFF", {value: 84, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_VTOL_LAND", {value: 85, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_GUIDED_LIMITS", {value: 90, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_GUIDED_MASTER", {value: 91, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_DELAY", {value: 93, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_NAV_LAST", {value: 95, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CONDITION_DELAY", {value: 112, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CONDITION_CHANGE_ALT", {value: 113, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CONDITION_DISTANCE", {value: 114, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CONDITION_YAW", {value: 115, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CONDITION_LAST", {value: 159, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CONDITION_GATE", {value: 4501, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_MODE", {value: 176, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_JUMP", {value: 177, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_CHANGE_SPEED", {value: 178, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_HOME", {value: 179, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_PARAMETER", {value: 180, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_RELAY", {value: 181, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_REPEAT_RELAY", {value: 182, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_REPEAT_SERVO", {value: 184, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_FLIGHTTERMINATION", {value: 185, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_CHANGE_ALTITUDE", {value: 186, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_ACTUATOR", {value: 187, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_LAND_START", {value: 189, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_GO_AROUND", {value: 191, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_REPOSITION", {value: 192, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_PAUSE_CONTINUE", {value: 193, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_ROI_LOCATION", {value: 195, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_ROI_WPNEXT_OFFSET", {value: 196, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_ROI_NONE", {value: 197, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_CONTROL_VIDEO", {value: 200, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_ROI", {value: 201, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_DIGICAM_CONTROL", {value: 203, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_MOUNT_CONFIGURE", {value: 204, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_MOUNT_CONTROL", {value: 205, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_CAM_TRIGG_DIST", {value: 206, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_FENCE_ENABLE", {value: 207, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_PARACHUTE", {value: 208, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_MOTOR_TEST", {value: 209, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_INVERTED_FLIGHT", {value: 210, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_GRIPPER", {value: 211, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_SET_CAM_TRIGG_INTERVAL", {value: 214, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_MOUNT_CONTROL_QUAT", {value: 220, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_GUIDED_MASTER", {value: 221, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_GUIDED_LIMITS", {value: 222, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_LAST", {value: 240, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_PREFLIGHT_CALIBRATION", {value: 241, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "PREFLIGHT_CALIBRATION_TEMPERATURE_CALIBRATION", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_PREFLIGHT_SET_SENSOR_OFFSETS", {value: 242, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_PREFLIGHT_UAVCAN", {value: 243, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_PREFLIGHT_STORAGE", {value: 245, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_PREFLIGHT_REBOOT_SHUTDOWN", {value: 246, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_OBLIQUE_SURVEY", {value: 260, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_GIMBAL_DEVICE_INFORMATION", {value: 283, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_MISSION_START", {value: 300, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_ACTUATOR_TEST", {value: 310, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CONFIGURE_ACTUATOR", {value: 311, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_COMPONENT_ARM_DISARM", {value: 400, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_RUN_PREARM_CHECKS", {value: 401, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_INJECT_FAILURE", {value: 420, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_START_RX_PAIR", {value: 500, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_REQUEST_MESSAGE", {value: 512, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_SET_CAMERA_MODE", {value: 530, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_SET_CAMERA_ZOOM", {value: 531, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_SET_CAMERA_FOCUS", {value: 532, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_GIMBAL_MANAGER_PITCHYAW", {value: 1000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_GIMBAL_MANAGER_CONFIGURE", {value: 1001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_IMAGE_START_CAPTURE", {value: 2000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_TRIGGER_CONTROL", {value: 2003, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_VIDEO_START_CAPTURE", {value: 2500, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_VIDEO_STOP_CAPTURE", {value: 2501, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_LOGGING_START", {value: 2510, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_LOGGING_STOP", {value: 2511, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_CONTROL_HIGH_LATENCY", {value: 2600, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_VTOL_TRANSITION", {value: 3000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_ARM_AUTHORIZATION_REQUEST", {value: 3001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_PAYLOAD_PREPARE_DEPLOY", {value: 30001, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_PAYLOAD_CONTROL_DEPLOY", {value: 30002, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_FIXED_MAG_CAL_YAW", {value: 42006, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_DO_WINCH", {value: 42600, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_PX4_INTERNAL_START", {value: 65537, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_CMD_SET_GPS_GLOBAL_ORIGIN", {value: 100000, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_MOUNT_MODE_RETRACT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_MOUNT_MODE_NEUTRAL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_MOUNT_MODE_MAVLINK_TARGETING", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_MOUNT_MODE_RC_TARGETING", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_MOUNT_MODE_GPS_POINT", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_MOUNT_MODE_ENUM_END", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_ROI_NONE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_ROI_WPNEXT", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_ROI_WPINDEX", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_ROI_LOCATION", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_ROI_TARGET", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "VEHICLE_ROI_ENUM_END", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "PARACHUTE_ACTION_DISABLE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "PARACHUTE_ACTION_ENABLE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "PARACHUTE_ACTION_RELEASE", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SENSOR_GYRO", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SENSOR_ACCEL", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SENSOR_MAG", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SENSOR_BARO", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SENSOR_GPS", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SENSOR_OPTICAL_FLOW", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SENSOR_VIO", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SENSOR_DISTANCE_SENSOR", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SENSOR_AIRSPEED", {value: 8, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SYSTEM_BATTERY", {value: 100, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SYSTEM_MOTOR", {value: 101, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SYSTEM_SERVO", {value: 102, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SYSTEM_AVOIDANCE", {value: 103, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SYSTEM_RC_SIGNAL", {value: 104, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_UNIT_SYSTEM_MAVLINK_SIGNAL", {value: 105, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_TYPE_OK", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_TYPE_OFF", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_TYPE_STUCK", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_TYPE_GARBAGE", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_TYPE_WRONG", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_TYPE_SLOW", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_TYPE_DELAYED", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "FAILURE_TYPE_INTERMITTENT", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "SPEED_TYPE_AIRSPEED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "SPEED_TYPE_GROUNDSPEED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "SPEED_TYPE_CLIMB_SPEED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "SPEED_TYPE_DESCEND_SPEED", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "ARMING_ACTION_DISARM", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "ARMING_ACTION_ARM", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "GRIPPER_ACTION_RELEASE", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "GRIPPER_ACTION_GRAB", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleCommandWrapper, "ORB_QUEUE_LENGTH", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = VehicleCommandWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CUSTOM_0",
      "value": 0
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CUSTOM_1",
      "value": 1
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CUSTOM_2",
      "value": 2
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_WAYPOINT",
      "value": 16
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_LOITER_UNLIM",
      "value": 17
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_LOITER_TURNS",
      "value": 18
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_LOITER_TIME",
      "value": 19
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_RETURN_TO_LAUNCH",
      "value": 20
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_LAND",
      "value": 21
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_TAKEOFF",
      "value": 22
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_PRECLAND",
      "value": 23
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_ORBIT",
      "value": 34
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_ROI",
      "value": 80
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_PATHPLANNING",
      "value": 81
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_VTOL_TAKEOFF",
      "value": 84
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_VTOL_LAND",
      "value": 85
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_GUIDED_LIMITS",
      "value": 90
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_GUIDED_MASTER",
      "value": 91
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_DELAY",
      "value": 93
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_NAV_LAST",
      "value": 95
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CONDITION_DELAY",
      "value": 112
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CONDITION_CHANGE_ALT",
      "value": 113
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CONDITION_DISTANCE",
      "value": 114
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CONDITION_YAW",
      "value": 115
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CONDITION_LAST",
      "value": 159
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CONDITION_GATE",
      "value": 4501
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_MODE",
      "value": 176
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_JUMP",
      "value": 177
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_CHANGE_SPEED",
      "value": 178
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_HOME",
      "value": 179
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_PARAMETER",
      "value": 180
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_RELAY",
      "value": 181
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_REPEAT_RELAY",
      "value": 182
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_REPEAT_SERVO",
      "value": 184
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_FLIGHTTERMINATION",
      "value": 185
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_CHANGE_ALTITUDE",
      "value": 186
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_ACTUATOR",
      "value": 187
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_LAND_START",
      "value": 189
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_GO_AROUND",
      "value": 191
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_REPOSITION",
      "value": 192
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_PAUSE_CONTINUE",
      "value": 193
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_ROI_LOCATION",
      "value": 195
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_ROI_WPNEXT_OFFSET",
      "value": 196
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_ROI_NONE",
      "value": 197
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_CONTROL_VIDEO",
      "value": 200
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_ROI",
      "value": 201
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_DIGICAM_CONTROL",
      "value": 203
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_MOUNT_CONFIGURE",
      "value": 204
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_MOUNT_CONTROL",
      "value": 205
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_CAM_TRIGG_DIST",
      "value": 206
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_FENCE_ENABLE",
      "value": 207
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_PARACHUTE",
      "value": 208
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_MOTOR_TEST",
      "value": 209
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_INVERTED_FLIGHT",
      "value": 210
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_GRIPPER",
      "value": 211
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_SET_CAM_TRIGG_INTERVAL",
      "value": 214
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_MOUNT_CONTROL_QUAT",
      "value": 220
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_GUIDED_MASTER",
      "value": 221
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_GUIDED_LIMITS",
      "value": 222
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_LAST",
      "value": 240
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_PREFLIGHT_CALIBRATION",
      "value": 241
    },
    {
      "type": "uint16",
      "name": "PREFLIGHT_CALIBRATION_TEMPERATURE_CALIBRATION",
      "value": 3
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_PREFLIGHT_SET_SENSOR_OFFSETS",
      "value": 242
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_PREFLIGHT_UAVCAN",
      "value": 243
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_PREFLIGHT_STORAGE",
      "value": 245
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_PREFLIGHT_REBOOT_SHUTDOWN",
      "value": 246
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_OBLIQUE_SURVEY",
      "value": 260
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_GIMBAL_DEVICE_INFORMATION",
      "value": 283
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_MISSION_START",
      "value": 300
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_ACTUATOR_TEST",
      "value": 310
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CONFIGURE_ACTUATOR",
      "value": 311
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_COMPONENT_ARM_DISARM",
      "value": 400
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_RUN_PREARM_CHECKS",
      "value": 401
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_INJECT_FAILURE",
      "value": 420
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_START_RX_PAIR",
      "value": 500
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_REQUEST_MESSAGE",
      "value": 512
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_SET_CAMERA_MODE",
      "value": 530
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_SET_CAMERA_ZOOM",
      "value": 531
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_SET_CAMERA_FOCUS",
      "value": 532
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_GIMBAL_MANAGER_PITCHYAW",
      "value": 1000
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_GIMBAL_MANAGER_CONFIGURE",
      "value": 1001
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_IMAGE_START_CAPTURE",
      "value": 2000
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_TRIGGER_CONTROL",
      "value": 2003
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_VIDEO_START_CAPTURE",
      "value": 2500
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_VIDEO_STOP_CAPTURE",
      "value": 2501
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_LOGGING_START",
      "value": 2510
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_LOGGING_STOP",
      "value": 2511
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_CONTROL_HIGH_LATENCY",
      "value": 2600
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_VTOL_TRANSITION",
      "value": 3000
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_ARM_AUTHORIZATION_REQUEST",
      "value": 3001
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_PAYLOAD_PREPARE_DEPLOY",
      "value": 30001
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_PAYLOAD_CONTROL_DEPLOY",
      "value": 30002
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_FIXED_MAG_CAL_YAW",
      "value": 42006
    },
    {
      "type": "uint16",
      "name": "VEHICLE_CMD_DO_WINCH",
      "value": 42600
    },
    {
      "type": "uint32",
      "name": "VEHICLE_CMD_PX4_INTERNAL_START",
      "value": 65537
    },
    {
      "type": "uint32",
      "name": "VEHICLE_CMD_SET_GPS_GLOBAL_ORIGIN",
      "value": 100000
    },
    {
      "type": "uint8",
      "name": "VEHICLE_MOUNT_MODE_RETRACT",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "VEHICLE_MOUNT_MODE_NEUTRAL",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "VEHICLE_MOUNT_MODE_MAVLINK_TARGETING",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "VEHICLE_MOUNT_MODE_RC_TARGETING",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "VEHICLE_MOUNT_MODE_GPS_POINT",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "VEHICLE_MOUNT_MODE_ENUM_END",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "VEHICLE_ROI_NONE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "VEHICLE_ROI_WPNEXT",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "VEHICLE_ROI_WPINDEX",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "VEHICLE_ROI_LOCATION",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "VEHICLE_ROI_TARGET",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "VEHICLE_ROI_ENUM_END",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "PARACHUTE_ACTION_DISABLE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "PARACHUTE_ACTION_ENABLE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "PARACHUTE_ACTION_RELEASE",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SENSOR_GYRO",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SENSOR_ACCEL",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SENSOR_MAG",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SENSOR_BARO",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SENSOR_GPS",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SENSOR_OPTICAL_FLOW",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SENSOR_VIO",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SENSOR_DISTANCE_SENSOR",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SENSOR_AIRSPEED",
      "value": 8
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SYSTEM_BATTERY",
      "value": 100
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SYSTEM_MOTOR",
      "value": 101
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SYSTEM_SERVO",
      "value": 102
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SYSTEM_AVOIDANCE",
      "value": 103
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SYSTEM_RC_SIGNAL",
      "value": 104
    },
    {
      "type": "uint8",
      "name": "FAILURE_UNIT_SYSTEM_MAVLINK_SIGNAL",
      "value": 105
    },
    {
      "type": "uint8",
      "name": "FAILURE_TYPE_OK",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "FAILURE_TYPE_OFF",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "FAILURE_TYPE_STUCK",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "FAILURE_TYPE_GARBAGE",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "FAILURE_TYPE_WRONG",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "FAILURE_TYPE_SLOW",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "FAILURE_TYPE_DELAYED",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "FAILURE_TYPE_INTERMITTENT",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "SPEED_TYPE_AIRSPEED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "SPEED_TYPE_GROUNDSPEED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "SPEED_TYPE_CLIMB_SPEED",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "SPEED_TYPE_DESCEND_SPEED",
      "value": 3
    },
    {
      "type": "int8",
      "name": "ARMING_ACTION_DISARM",
      "value": 0
    },
    {
      "type": "int8",
      "name": "ARMING_ACTION_ARM",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "GRIPPER_ACTION_RELEASE",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "GRIPPER_ACTION_GRAB",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "ORB_QUEUE_LENGTH",
      "value": 8
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param5",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param6",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "param7",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "command",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "target_system",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "target_component",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "source_system",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "source_component",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "confirmation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "from_external",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleCommand",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleCommand"
}
*/
