// rclnodejs/generated/px4_msgs/px4_msgs__msg__MagWorkerData.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const MagWorkerDataRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  done_count: primitiveTypes.uint32,
  calibration_points_perside: primitiveTypes.uint32,
  calibration_interval_perside_us: primitiveTypes.uint64,
  calibration_counter_total: ArrayType(primitiveTypes.uint32, 4),
  side_data_collected: ArrayType(primitiveTypes.bool, 4),
  x: ArrayType(primitiveTypes.float32, 4),
  y: ArrayType(primitiveTypes.float32, 4),
  z: ArrayType(primitiveTypes.float32, 4),
});
const MagWorkerDataRefArray = ArrayType(MagWorkerDataRefStruct);
const MagWorkerDataRefStructArray = StructType({
  data: MagWorkerDataRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MagWorkerDataWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._side_data_collectedArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MagWorkerDataRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._done_countIntialized = true;
      this._calibration_points_persideIntialized = true;
      this._calibration_interval_perside_usIntialized = true;
      this._wrapperFields.calibration_counter_total = UInt32Wrapper.createArray();
      this._wrapperFields.calibration_counter_total.copy(other._wrapperFields.calibration_counter_total);
      this._wrapperFields.side_data_collected = BoolWrapper.createArray();
      this._wrapperFields.side_data_collected.copy(other._wrapperFields.side_data_collected);
      this.side_data_collected = other.side_data_collected;
      this._wrapperFields.x = Float32Wrapper.createArray();
      this._wrapperFields.x.copy(other._wrapperFields.x);
      this._wrapperFields.y = Float32Wrapper.createArray();
      this._wrapperFields.y.copy(other._wrapperFields.y);
      this._wrapperFields.z = Float32Wrapper.createArray();
      this._wrapperFields.z.copy(other._wrapperFields.z);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MagWorkerDataRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._done_countIntialized = false;
    this._calibration_points_persideIntialized = false;
    this._calibration_interval_perside_usIntialized = false;
    this._wrapperFields.calibration_counter_total = UInt32Wrapper.createArray();
    this._wrapperFields.side_data_collected = BoolWrapper.createArray();
    this._wrapperFields.x = Float32Wrapper.createArray();
    this._wrapperFields.y = Float32Wrapper.createArray();
    this._wrapperFields.z = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MagWorkerDataWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MagWorkerDataArrayWrapper;
  }
  static get ArrayType() {
    return MagWorkerDataArrayWrapper;
  }
  static get refObjectArrayType() {
    return MagWorkerDataRefStructArray
  }
  static get refObjectType() {
    return MagWorkerDataRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in MagWorkerData');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in MagWorkerData');
      }
      if (checkConsistency && !this._done_countIntialized) {
        throw new TypeError('Invalid argument: done_count in MagWorkerData');
      }
      if (checkConsistency && !this._calibration_points_persideIntialized) {
        throw new TypeError('Invalid argument: calibration_points_perside in MagWorkerData');
      }
      if (checkConsistency && !this._calibration_interval_perside_usIntialized) {
        throw new TypeError('Invalid argument: calibration_interval_perside_us in MagWorkerData');
      }
    this._refObject.calibration_counter_total = Array.from(this._wrapperFields.calibration_counter_total.data);
    this._refObject.side_data_collected = this._side_data_collectedArray;
    this._refObject.x = Array.from(this._wrapperFields.x.data);
    this._refObject.y = Array.from(this._wrapperFields.y.data);
    this._refObject.z = Array.from(this._wrapperFields.z.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._done_countIntialized = true;
    this._refObject.done_count = refObject.done_count;
    this._calibration_points_persideIntialized = true;
    this._refObject.calibration_points_perside = refObject.calibration_points_perside;
    this._calibration_interval_perside_usIntialized = true;
    this._refObject.calibration_interval_perside_us = refObject.calibration_interval_perside_us;
    this._wrapperFields.calibration_counter_total.fill(refObject.calibration_counter_total.toArray());
    this._side_data_collectedArray = refObject.side_data_collected.toArray();
    this._wrapperFields.x.fill(refObject.x.toArray());
    this._wrapperFields.y.fill(refObject.y.toArray());
    this._wrapperFields.z.fill(refObject.z.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    MagWorkerDataWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'MagWorkerData'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get done_count() {
    if (!this._done_countIntialized) {
      return undefined;
    }
    return this._refObject.done_count;
  }
  set done_count(value) {
    this._done_countIntialized = true;
    this._refObject.done_count = value;
  }
  get calibration_points_perside() {
    if (!this._calibration_points_persideIntialized) {
      return undefined;
    }
    return this._refObject.calibration_points_perside;
  }
  set calibration_points_perside(value) {
    this._calibration_points_persideIntialized = true;
    this._refObject.calibration_points_perside = value;
  }
  get calibration_interval_perside_us() {
    if (!this._calibration_interval_perside_usIntialized) {
      return undefined;
    }
    return this._refObject.calibration_interval_perside_us;
  }
  set calibration_interval_perside_us(value) {
    this._calibration_interval_perside_usIntialized = true;
    this._refObject.calibration_interval_perside_us = value;
  }
  get calibration_counter_total() {
    return this._wrapperFields['calibration_counter_total'].data;
  }
  set calibration_counter_total(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields['calibration_counter_total'].fill(value);
  }
  get side_data_collected() {
    return this._side_data_collectedArray;
  }
  set side_data_collected(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._side_data_collectedArray = value;
  }
  get x() {
    return this._wrapperFields['x'].data;
  }
  set x(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields['x'].fill(value);
  }
  get y() {
    return this._wrapperFields['y'].data;
  }
  set y(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields['y'].fill(value);
  }
  get z() {
    return this._wrapperFields['z'].data;
  }
  set z(value) {
    if (value.length !== 4) {
      throw new RangeError('The length of the array must be 4.');
    }
    this._wrapperFields['z'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MagWorkerDataRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._done_countIntialized = true;
    this._calibration_points_persideIntialized = true;
    this._calibration_interval_perside_usIntialized = true;
    this._wrapperFields.calibration_counter_total.fill(refObject.calibration_counter_total.toArray());
    refObject.side_data_collected.data.length = refObject.side_data_collected.size;
    for (let index = 0; index < refObject.side_data_collected.size; index++) {
      this._side_data_collectedArray[index] = refObject.side_data_collected.data[index].data;
    }
    this._wrapperFields.x.fill(refObject.x.toArray());
    this._wrapperFields.y.fill(refObject.y.toArray());
    this._wrapperFields.z.fill(refObject.z.toArray());
  }
  copy(other) {
    this._refObject = new MagWorkerDataRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._done_countIntialized = true;
    this._calibration_points_persideIntialized = true;
    this._calibration_interval_perside_usIntialized = true;
    this._wrapperFields.calibration_counter_total.copy(other._wrapperFields.calibration_counter_total);
    this._side_data_collectedArray = other._side_data_collectedArray.slice();
    this._wrapperFields.x.copy(other._wrapperFields.x);
    this._wrapperFields.y.copy(other._wrapperFields.y);
    this._wrapperFields.z.copy(other._wrapperFields.z);
  }
  static get classType() {
    return MagWorkerDataWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"MAX_MAGS","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"done_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"calibration_points_perside","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"calibration_interval_perside_us","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"calibration_counter_total","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"side_data_collected","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":true,"arraySize":4,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"MagWorkerData","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MagWorkerData"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","done_count","calibration_points_perside","calibration_interval_perside_us","calibration_counter_total","side_data_collected","x","y","z"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MagWorkerDataArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MagWorkerDataWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MagWorkerDataWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MagWorkerDataArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MagWorkerDataArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MagWorkerDataRefArray(size);
    this._refObject = new MagWorkerDataRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MagWorkerDataWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MagWorkerDataArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MagWorkerDataArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MagWorkerDataWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MagWorkerDataWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MagWorkerDataArrayWrapper;
  }
}
// Define constants (1 in total)
Object.defineProperty(MagWorkerDataWrapper, "MAX_MAGS", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = MagWorkerDataWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "MAX_MAGS",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "done_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "calibration_points_perside",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "calibration_interval_perside_us",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "calibration_counter_total",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "side_data_collected",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z",
      "type": {
        "isArray": true,
        "arraySize": 4,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "MagWorkerData",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MagWorkerData"
}
*/
