// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleTrajectoryBezier.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TrajectoryBezierWrapper = require('../../generated/px4_msgs/px4_msgs__msg__TrajectoryBezier.js');
const VehicleTrajectoryBezierRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  control_points: ArrayType(TrajectoryBezierWrapper.refObjectType, 5),
  bezier_order: primitiveTypes.uint8,
});
const VehicleTrajectoryBezierRefArray = ArrayType(VehicleTrajectoryBezierRefStruct);
const VehicleTrajectoryBezierRefStructArray = StructType({
  data: VehicleTrajectoryBezierRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleTrajectoryBezierWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleTrajectoryBezierRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._wrapperFields.control_points = TrajectoryBezierWrapper.createArray();
      this._wrapperFields.control_points.copy(other._wrapperFields.control_points);
      this._bezier_orderIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleTrajectoryBezierRefStruct();
    this._timestampIntialized = false;
    this._wrapperFields.control_points = TrajectoryBezierWrapper.createArray();
    this._bezier_orderIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleTrajectoryBezierWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleTrajectoryBezierArrayWrapper;
  }
  static get ArrayType() {
    return VehicleTrajectoryBezierArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleTrajectoryBezierRefStructArray
  }
  static get refObjectType() {
    return VehicleTrajectoryBezierRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleTrajectoryBezier');
      }
      if (checkConsistency && !this._bezier_orderIntialized) {
        throw new TypeError('Invalid argument: bezier_order in VehicleTrajectoryBezier');
      }
    for (let i = 0; i < 5; i++) {
      if (this._wrapperFields.control_points.data[i]) {
        this._wrapperFields.control_points.data[i].freeze(own, checkConsistency);
        this._refObject.control_points[i] = this._wrapperFields.control_points.data[i].refObject;
      }
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._refObject.control_points = refObject.control_points;
    this._wrapperFields.control_points.size = 5
    for (let i = 0; i < 5; i++) {
      this._wrapperFields.control_points.data[i].copyRefObject(refObject.control_points[i]);
    }
    this._bezier_orderIntialized = true;
    this._refObject.bezier_order = refObject.bezier_order;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    for (let i = 0; i < 5; i++) {
      TrajectoryBezierWrapper.freeStruct(refObject.control_points[i]);
    }
  }
  static destoryRawROS(msg) {
    VehicleTrajectoryBezierWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleTrajectoryBezier'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get control_points() {
    return this._wrapperFields.control_points;
  }
  set control_points(value) {
    if (value.length !== 5) {
      throw new RangeError('The length of the array must be 5.');
    }
    this._wrapperFields.control_points.fill(value);
  }
  get bezier_order() {
    if (!this._bezier_orderIntialized) {
      return undefined;
    }
    return this._refObject.bezier_order;
  }
  set bezier_order(value) {
    this._bezier_orderIntialized = true;
    this._refObject.bezier_order = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleTrajectoryBezierRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._refObject.control_points = refObject.control_points;
    this._wrapperFields.control_points.size = 5
    for (let i = 0; i < 5; i++) {
      this._wrapperFields.control_points.data[i].copyRefObject(refObject.control_points[i]);
    }
    this._bezier_orderIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleTrajectoryBezierRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._wrapperFields.control_points.copy(other._wrapperFields.control_points);
    this._bezier_orderIntialized = true;
  }
  static get classType() {
    return VehicleTrajectoryBezierWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"POINT_0","value":0},{"type":"uint8","name":"POINT_1","value":1},{"type":"uint8","name":"POINT_2","value":2},{"type":"uint8","name":"POINT_3","value":3},{"type":"uint8","name":"POINT_4","value":4},{"type":"uint8","name":"NUMBER_POINTS","value":5}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"control_points","type":{"isArray":true,"arraySize":5,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":"px4_msgs","type":"TrajectoryBezier","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"bezier_order","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleTrajectoryBezier","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleTrajectoryBezier"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","control_points","bezier_order"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleTrajectoryBezierArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleTrajectoryBezierWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleTrajectoryBezierWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleTrajectoryBezierArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleTrajectoryBezierArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleTrajectoryBezierRefArray(size);
    this._refObject = new VehicleTrajectoryBezierRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleTrajectoryBezierWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleTrajectoryBezierArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleTrajectoryBezierArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleTrajectoryBezierWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleTrajectoryBezierWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleTrajectoryBezierArrayWrapper;
  }
}
// Define constants (6 in total)
Object.defineProperty(VehicleTrajectoryBezierWrapper, "POINT_0", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryBezierWrapper, "POINT_1", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryBezierWrapper, "POINT_2", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryBezierWrapper, "POINT_3", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryBezierWrapper, "POINT_4", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(VehicleTrajectoryBezierWrapper, "NUMBER_POINTS", {value: 5, writable: false, enumerable: true, configurable: true});
module.exports = VehicleTrajectoryBezierWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "POINT_0",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "POINT_1",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "POINT_2",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "POINT_3",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "POINT_4",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "NUMBER_POINTS",
      "value": 5
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "control_points",
      "type": {
        "isArray": true,
        "arraySize": 5,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": "px4_msgs",
        "type": "TrajectoryBezier",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "bezier_order",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleTrajectoryBezier",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleTrajectoryBezier"
}
*/
