// rclnodejs/generated/px4_msgs/px4_msgs__msg__VehicleOpticalFlow.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const VehicleOpticalFlowRefStruct = StructType({
  timestamp: primitiveTypes.uint64,
  timestamp_sample: primitiveTypes.uint64,
  device_id: primitiveTypes.uint32,
  pixel_flow: ArrayType(primitiveTypes.float32, 2),
  delta_angle: ArrayType(primitiveTypes.float32, 3),
  distance_m: primitiveTypes.float32,
  integration_timespan_us: primitiveTypes.uint32,
  quality: primitiveTypes.uint8,
  max_flow_rate: primitiveTypes.float32,
  min_ground_distance: primitiveTypes.float32,
  max_ground_distance: primitiveTypes.float32,
});
const VehicleOpticalFlowRefArray = ArrayType(VehicleOpticalFlowRefStruct);
const VehicleOpticalFlowRefStructArray = StructType({
  data: VehicleOpticalFlowRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleOpticalFlowWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleOpticalFlowRefStruct(other._refObject.toObject());
      this._timestampIntialized = true;
      this._timestamp_sampleIntialized = true;
      this._device_idIntialized = true;
      this._wrapperFields.pixel_flow = Float32Wrapper.createArray();
      this._wrapperFields.pixel_flow.copy(other._wrapperFields.pixel_flow);
      this._wrapperFields.delta_angle = Float32Wrapper.createArray();
      this._wrapperFields.delta_angle.copy(other._wrapperFields.delta_angle);
      this._distance_mIntialized = true;
      this._integration_timespan_usIntialized = true;
      this._qualityIntialized = true;
      this._max_flow_rateIntialized = true;
      this._min_ground_distanceIntialized = true;
      this._max_ground_distanceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleOpticalFlowRefStruct();
    this._timestampIntialized = false;
    this._timestamp_sampleIntialized = false;
    this._device_idIntialized = false;
    this._wrapperFields.pixel_flow = Float32Wrapper.createArray();
    this._wrapperFields.delta_angle = Float32Wrapper.createArray();
    this._distance_mIntialized = false;
    this._integration_timespan_usIntialized = false;
    this._qualityIntialized = false;
    this._max_flow_rateIntialized = false;
    this._min_ground_distanceIntialized = false;
    this._max_ground_distanceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new VehicleOpticalFlowWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleOpticalFlowArrayWrapper;
  }
  static get ArrayType() {
    return VehicleOpticalFlowArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleOpticalFlowRefStructArray
  }
  static get refObjectType() {
    return VehicleOpticalFlowRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._timestampIntialized) {
        throw new TypeError('Invalid argument: timestamp in VehicleOpticalFlow');
      }
      if (checkConsistency && !this._timestamp_sampleIntialized) {
        throw new TypeError('Invalid argument: timestamp_sample in VehicleOpticalFlow');
      }
      if (checkConsistency && !this._device_idIntialized) {
        throw new TypeError('Invalid argument: device_id in VehicleOpticalFlow');
      }
      if (checkConsistency && !this._distance_mIntialized) {
        throw new TypeError('Invalid argument: distance_m in VehicleOpticalFlow');
      }
      if (checkConsistency && !this._integration_timespan_usIntialized) {
        throw new TypeError('Invalid argument: integration_timespan_us in VehicleOpticalFlow');
      }
      if (checkConsistency && !this._qualityIntialized) {
        throw new TypeError('Invalid argument: quality in VehicleOpticalFlow');
      }
      if (checkConsistency && !this._max_flow_rateIntialized) {
        throw new TypeError('Invalid argument: max_flow_rate in VehicleOpticalFlow');
      }
      if (checkConsistency && !this._min_ground_distanceIntialized) {
        throw new TypeError('Invalid argument: min_ground_distance in VehicleOpticalFlow');
      }
      if (checkConsistency && !this._max_ground_distanceIntialized) {
        throw new TypeError('Invalid argument: max_ground_distance in VehicleOpticalFlow');
      }
    this._refObject.pixel_flow = Array.from(this._wrapperFields.pixel_flow.data);
    this._refObject.delta_angle = Array.from(this._wrapperFields.delta_angle.data);
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._timestampIntialized = true;
    this._refObject.timestamp = refObject.timestamp;
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = refObject.timestamp_sample;
    this._device_idIntialized = true;
    this._refObject.device_id = refObject.device_id;
    this._wrapperFields.pixel_flow.fill(refObject.pixel_flow.toArray());
    this._wrapperFields.delta_angle.fill(refObject.delta_angle.toArray());
    this._distance_mIntialized = true;
    this._refObject.distance_m = refObject.distance_m;
    this._integration_timespan_usIntialized = true;
    this._refObject.integration_timespan_us = refObject.integration_timespan_us;
    this._qualityIntialized = true;
    this._refObject.quality = refObject.quality;
    this._max_flow_rateIntialized = true;
    this._refObject.max_flow_rate = refObject.max_flow_rate;
    this._min_ground_distanceIntialized = true;
    this._refObject.min_ground_distance = refObject.min_ground_distance;
    this._max_ground_distanceIntialized = true;
    this._refObject.max_ground_distance = refObject.max_ground_distance;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    VehicleOpticalFlowWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'px4_msgs', subFolder: 'msg', interfaceName: 'VehicleOpticalFlow'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    if (!this._timestampIntialized) {
      return undefined;
    }
    return this._refObject.timestamp;
  }
  set timestamp(value) {
    this._timestampIntialized = true;
    this._refObject.timestamp = value;
  }
  get timestamp_sample() {
    if (!this._timestamp_sampleIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_sample;
  }
  set timestamp_sample(value) {
    this._timestamp_sampleIntialized = true;
    this._refObject.timestamp_sample = value;
  }
  get device_id() {
    if (!this._device_idIntialized) {
      return undefined;
    }
    return this._refObject.device_id;
  }
  set device_id(value) {
    this._device_idIntialized = true;
    this._refObject.device_id = value;
  }
  get pixel_flow() {
    return this._wrapperFields['pixel_flow'].data;
  }
  set pixel_flow(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['pixel_flow'].fill(value);
  }
  get delta_angle() {
    return this._wrapperFields['delta_angle'].data;
  }
  set delta_angle(value) {
    if (value.length !== 3) {
      throw new RangeError('The length of the array must be 3.');
    }
    this._wrapperFields['delta_angle'].fill(value);
  }
  get distance_m() {
    if (!this._distance_mIntialized) {
      return undefined;
    }
    return this._refObject.distance_m;
  }
  set distance_m(value) {
    this._distance_mIntialized = true;
    this._refObject.distance_m = value;
  }
  get integration_timespan_us() {
    if (!this._integration_timespan_usIntialized) {
      return undefined;
    }
    return this._refObject.integration_timespan_us;
  }
  set integration_timespan_us(value) {
    this._integration_timespan_usIntialized = true;
    this._refObject.integration_timespan_us = value;
  }
  get quality() {
    if (!this._qualityIntialized) {
      return undefined;
    }
    return this._refObject.quality;
  }
  set quality(value) {
    this._qualityIntialized = true;
    this._refObject.quality = value;
  }
  get max_flow_rate() {
    if (!this._max_flow_rateIntialized) {
      return undefined;
    }
    return this._refObject.max_flow_rate;
  }
  set max_flow_rate(value) {
    this._max_flow_rateIntialized = true;
    this._refObject.max_flow_rate = value;
  }
  get min_ground_distance() {
    if (!this._min_ground_distanceIntialized) {
      return undefined;
    }
    return this._refObject.min_ground_distance;
  }
  set min_ground_distance(value) {
    this._min_ground_distanceIntialized = true;
    this._refObject.min_ground_distance = value;
  }
  get max_ground_distance() {
    if (!this._max_ground_distanceIntialized) {
      return undefined;
    }
    return this._refObject.max_ground_distance;
  }
  set max_ground_distance(value) {
    this._max_ground_distanceIntialized = true;
    this._refObject.max_ground_distance = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleOpticalFlowRefStruct(refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._device_idIntialized = true;
    this._wrapperFields.pixel_flow.fill(refObject.pixel_flow.toArray());
    this._wrapperFields.delta_angle.fill(refObject.delta_angle.toArray());
    this._distance_mIntialized = true;
    this._integration_timespan_usIntialized = true;
    this._qualityIntialized = true;
    this._max_flow_rateIntialized = true;
    this._min_ground_distanceIntialized = true;
    this._max_ground_distanceIntialized = true;
  }
  copy(other) {
    this._refObject = new VehicleOpticalFlowRefStruct(other._refObject.toObject());
    this._timestampIntialized = true;
    this._timestamp_sampleIntialized = true;
    this._device_idIntialized = true;
    this._wrapperFields.pixel_flow.copy(other._wrapperFields.pixel_flow);
    this._wrapperFields.delta_angle.copy(other._wrapperFields.delta_angle);
    this._distance_mIntialized = true;
    this._integration_timespan_usIntialized = true;
    this._qualityIntialized = true;
    this._max_flow_rateIntialized = true;
    this._min_ground_distanceIntialized = true;
    this._max_ground_distanceIntialized = true;
  }
  static get classType() {
    return VehicleOpticalFlowWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_sample","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"device_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pixel_flow","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"delta_angle","type":{"isArray":true,"arraySize":3,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"distance_m","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"integration_timespan_us","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"quality","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_flow_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"min_ground_distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_ground_distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"px4_msgs","type":"VehicleOpticalFlow","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleOpticalFlow"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","timestamp_sample","device_id","pixel_flow","delta_angle","distance_m","integration_timespan_us","quality","max_flow_rate","min_ground_distance","max_ground_distance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleOpticalFlowArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleOpticalFlowWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleOpticalFlowWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleOpticalFlowArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleOpticalFlowArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleOpticalFlowRefArray(size);
    this._refObject = new VehicleOpticalFlowRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleOpticalFlowWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleOpticalFlowArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleOpticalFlowArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleOpticalFlowWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleOpticalFlowWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleOpticalFlowArrayWrapper;
  }
}
module.exports = VehicleOpticalFlowWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_sample",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "device_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pixel_flow",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "delta_angle",
      "type": {
        "isArray": true,
        "arraySize": 3,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "distance_m",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "integration_timespan_us",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "quality",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_flow_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "min_ground_distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_ground_distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "px4_msgs",
    "type": "VehicleOpticalFlow",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleOpticalFlow"
}
*/
